--- a/net/minecraft/core/dispenser/MinecartDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/MinecartDispenseItemBehavior.java
@@ -34,7 +_,7 @@
         BlockState blockFront = level.getBlockState(front);
         double yOffset;
         if (blockFront.is(BlockTags.RAILS)) {
-            if (getRailShape(blockFront).isSlope()) {
+            if (getRailShape(blockFront, level,front).isSlope()) {
                 yOffset = 0.6;
             } else {
                 yOffset = 0.1;
@@ -49,7 +_,7 @@
                 return this.defaultDispenseItemBehavior.dispense(source, dispensed);
             }
 
-            if (direction != Direction.DOWN && getRailShape(blockBelow).isSlope()) {
+            if (direction != Direction.DOWN && getRailShape(blockBelow, level, front.below()).isSlope()) {
                 yOffset = -0.4;
             } else {
                 yOffset = -0.9;
@@ -68,8 +_,8 @@
         return dispensed;
     }
 
-    private static RailShape getRailShape(BlockState blockFront) {
-        return blockFront.getBlock() instanceof BaseRailBlock railBlock ? blockFront.getValue(railBlock.getShapeProperty()) : RailShape.NORTH_SOUTH;
+    private static RailShape getRailShape(BlockState blockFront, ServerLevel level, BlockPos pos) {
+        return blockFront.getBlock() instanceof BaseRailBlock baserailblock ? baserailblock.getRailDirection(blockFront, level, pos, null) : RailShape.NORTH_SOUTH;
     }
 
     @Override
