--- a/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/ShearsDispenseItemBehavior.java
@@ -22,7 +_,7 @@
         ServerLevel level = source.level();
         if (!level.isClientSide()) {
             BlockPos pos = source.pos().relative(source.state().getValue(DispenserBlock.FACING));
-            this.setSuccess(tryShearBeehive(level, dispensed, pos) || tryShearEntity(level, pos, dispensed));
+            this.setSuccess(net.neoforged.neoforge.common.CommonHooks.tryDispenseShearsHarvestBlock(source, dispensed, level, pos) || tryShearBeehive(level, dispensed, pos) || tryShearEntity(level, pos, dispensed));
             if (this.isSuccess()) {
                 dispensed.hurtAndBreak(1, level, null, item -> {});
             }
@@ -53,8 +_,9 @@
                 return true;
             }
 
-            if (entity instanceof Shearable shearable && shearable.readyForShearing()) {
-                shearable.shear(level, SoundSource.BLOCKS, tool);
+            if (entity instanceof net.neoforged.neoforge.common.IShearable shearable && shearable.isShearable(null, tool, level, pos)) {
+                shearable.onSheared(null, tool, level, pos)
+                        .forEach(drop -> shearable.spawnShearedDrop(level, pos, drop));
                 level.gameEvent(null, GameEvent.SHEAR, pos);
                 return true;
             }
