--- a/net/minecraft/world/item/enchantment/effects/ReplaceBlock.java
+++ b/net/minecraft/world/item/enchantment/effects/ReplaceBlock.java
@@ -26,7 +26,7 @@
 
         if ((Boolean) this.predicate.map((blockpredicate) -> {
             return blockpredicate.test(serverLevel, blockpos);
-        }).orElse(true) && serverLevel.setBlockAndUpdate(blockpos, this.blockState.getState(serverLevel, entity.getRandom(), blockpos))) {
+        }).orElse(true) && org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(serverLevel, blockpos, this.blockState.getState(serverLevel, entity.getRandom(), blockpos), entity)) { // CraftBukkit - Call EntityBlockFormEvent
             this.triggerGameEvent.ifPresent((holder) -> {
                 serverLevel.gameEvent(entity, holder, blockpos);
             });
