--- a/net/minecraft/world/item/enchantment/effects/SummonEntityEffect.java
+++ b/net/minecraft/world/item/enchantment/effects/SummonEntityEffect.java
@@ -20,6 +20,12 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import net.minecraft.world.item.Items;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.weather.LightningStrikeEvent;
+// CraftBukkit end
+
 public record SummonEntityEffect(HolderSet<EntityType<?>> entityTypes, boolean joinTeam) implements EnchantmentEntityEffect {
 
     public static final MapCodec<SummonEntityEffect> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
@@ -34,7 +40,7 @@
             Optional<Holder<EntityType<?>>> optional = this.entityTypes().getRandomElement(serverLevel.getRandom());
 
             if (!optional.isEmpty()) {
-                Entity entity1 = ((EntityType) ((Holder) optional.get()).value()).spawn(serverLevel, blockpos, EntitySpawnReason.TRIGGERED);
+                Entity entity1 = ((EntityType) ((Holder) optional.get()).value()).create(serverLevel, null, blockpos, EntitySpawnReason.TRIGGERED, false, false); // CraftBukkit
 
                 if (entity1 != null) {
                     if (entity1 instanceof LightningBolt) {
@@ -46,6 +52,11 @@
 
                             lightningbolt.setCause(serverplayer);
                         }
+                        // CraftBukkit start
+                        serverLevel.strikeLightning(entity1, (item.itemStack().getItem() == Items.TRIDENT) ? LightningStrikeEvent.Cause.TRIDENT : LightningStrikeEvent.Cause.ENCHANTMENT);
+                    } else {
+                        serverLevel.addFreshEntityWithPassengers(entity1, CreatureSpawnEvent.SpawnReason.ENCHANTMENT);
+                        // CraftBukkit end
                     }
 
                     if (this.joinTeam && entity.getTeam() != null) {
