--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -684,9 +697,13 @@
                 this.submitBlockEntities(poseStack, levelRenderState, this.submitNodeStorage);
                 profiler.popPush("submitParticles");
                 this.particlesRenderState.submit(this.submitNodeStorage, levelRenderState.cameraRenderState);
+                profiler.popPush("neoforge_submit_custom");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.SubmitCustomGeometryEvent(this.levelRenderState, this.submitNodeStorage, poseStack, this.getRenderableSections()));
                 profiler.popPush("solidFeatures");
                 this.featureRenderDispatcher.renderSolidFeatures();
                 bufferSource.endBatch();
+                profiler.popPush("neoforge_render_after_opaque_features");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterOpaqueFeatures(this, this.levelRenderState, poseStack, modelViewMatrix, this.getRenderableSections()));
                 profiler.pop();
                 this.checkPoseStack(poseStack);
                 if (translucentTarget != null) {
@@ -730,13 +751,24 @@
                 bufferSource.endBatch();
                 this.featureRenderDispatcher.clearSubmitNodes();
                 this.particlesRenderState.reset();
+                profiler.push("neoforge_render_after_translucent_particles");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterTranslucentParticles(this, this.levelRenderState, poseStack, modelViewMatrix, this.getRenderableSections()));
             }
         );
     }
 
+    /**
+     * @deprecated Neo: use {@link #addCloudsPass(FrameGraphBuilder, CloudStatus, Vec3, long, float, int, float, Matrix4f)} instead
+     */
+    @Deprecated
     private void addCloudsPass(
         FrameGraphBuilder frame, CloudStatus cloudsType, Vec3 cameraPosition, long gameTime, float partialTicks, int cloudColor, float cloudHeight
     ) {
+        addCloudsPass(frame, cloudsType, cameraPosition, gameTime, partialTicks, cloudColor, cloudHeight, RenderSystem.getModelViewMatrix());
+    }
+
+    // Neo: Add modelViewMatrix for use in our event
+    private void addCloudsPass(FrameGraphBuilder frame, CloudStatus cloudsType, Vec3 cameraPosition, long gameTime, float partialTicks, int cloudColor, float cloudHeight, Matrix4f modelViewMatrix) {
         FramePass pass = frame.addPass("clouds");
         if (this.targets.clouds != null) {
             this.targets.clouds = pass.readsAndWrites(this.targets.clouds);
