--- a/net/minecraft/server/packs/OverlayMetadataSection.java
+++ b/net/minecraft/server/packs/OverlayMetadataSection.java
@@ -16,6 +_,10 @@
         "overlays", codecForPackType(PackType.CLIENT_RESOURCES)
     );
     public static final MetadataSectionType<OverlayMetadataSection> SERVER_TYPE = new MetadataSectionType<>("overlays", codecForPackType(PackType.SERVER_DATA));
+    // Neo: alternative metadata section types that will only be loaded on neoforged. Useful for making datapacks with special
+    // logic on different modloaders, or when running on neo vs in vanilla, without having to invert the main pack and overlays
+    public static final MetadataSectionType<OverlayMetadataSection> NEOFORGE_CLIENT_TYPE = new MetadataSectionType<>("neoforge:overlays", CLIENT_TYPE.codec());
+    public static final MetadataSectionType<OverlayMetadataSection> NEOFORGE_SERVER_TYPE = new MetadataSectionType<>("neoforge:overlays", SERVER_TYPE.codec());
 
     private static DataResult<String> validateOverlayDir(String path) {
         return !DIR_VALIDATOR.matcher(path).matches() ? DataResult.error(() -> path + " is not accepted directory name") : DataResult.success(path);
@@ -36,6 +_,13 @@
         };
     }
 
+    public static MetadataSectionType<OverlayMetadataSection> forPackTypeNeoForge(PackType packType) {
+        return switch (packType) {
+            case CLIENT_RESOURCES -> NEOFORGE_CLIENT_TYPE;
+            case SERVER_DATA -> NEOFORGE_SERVER_TYPE;
+        };
+    }
+
     public List<String> overlaysForVersion(PackFormat version) {
         return this.overlays.stream().filter(entry -> entry.isApplicable(version)).map(OverlayMetadataSection.OverlayEntry::overlay).toList();
     }
@@ -43,8 +_,7 @@
     public record OverlayEntry(InclusiveRange<PackFormat> format, String overlay) {
         private static Codec<List<OverlayMetadataSection.OverlayEntry>> listCodecForPackType(PackType packType) {
             int lastPreMinorVersion = PackFormat.lastPreMinorVersion(packType);
-            return OverlayMetadataSection.OverlayEntry.IntermediateEntry.CODEC
-                .listOf()
+            return net.neoforged.neoforge.common.conditions.ConditionalOps.decodeListWithElementConditions(OverlayMetadataSection.OverlayEntry.IntermediateEntry.CODEC)
                 .flatXmap(
                     list -> PackFormat.validateHolderList(
                         (List<OverlayMetadataSection.OverlayEntry.IntermediateEntry>)list,
