--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -17,6 +17,11 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.BossEvent;
 
+// CraftBukkit start
+import org.bukkit.boss.KeyedBossBar;
+import org.bukkit.craftbukkit.boss.CraftKeyedBossbar;
+// CraftBukkit end
+
 public class CustomBossEvent extends ServerBossEvent {
 
     private static final int DEFAULT_MAX = 100;
@@ -25,6 +30,16 @@
     private int value;
     private int max = 100;
     private final Runnable dirtyCallback;
+    // CraftBukkit start
+    private KeyedBossBar bossBar;
+
+    public KeyedBossBar getBukkitEntity() {
+        if (bossBar == null) {
+            bossBar = new CraftKeyedBossbar(this);
+        }
+        return bossBar;
+    }
+    // CraftBukkit end
 
     public CustomBossEvent(UUID id, Identifier customId, Component name, Runnable dirtyCallback) {
         super(id, name, BossEvent.BossBarColor.WHITE, BossEvent.BossBarOverlay.PROGRESS);
