--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -38,6 +_,7 @@
     private @Nullable Quaternionfc leftRotation;
     private @Nullable Vector3fc scale;
     private @Nullable Quaternionfc rightRotation;
+    private org.joml.@Nullable Matrix3f normalTransform = null;
     public static final Transformation IDENTITY = Util.make(() -> {
         Transformation identity = new Transformation(new Matrix4f());
         identity.translation = ZERO_TRANSLATION;
@@ -157,6 +_,16 @@
     @Override
     public int hashCode() {
         return Objects.hash(this.matrix);
+    }
+
+    /// Neo: {@return the normal matrix of this transformation}
+    public org.joml.Matrix3fc getNormalMatrix() {
+        if (this.normalTransform == null) {
+            this.normalTransform = new org.joml.Matrix3f(this.matrix);
+            this.normalTransform.invert();
+            this.normalTransform.transpose();
+        }
+        return this.normalTransform;
     }
 
     public Transformation slerp(Transformation that, float progress) {
