--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -161,6 +_,19 @@
         return Objects.hash(this.matrix);
     }
 
+    private Matrix3f normalTransform = null;
+    public org.joml.Matrix3fc getNormalMatrix() {
+        checkNormalTransform();
+        return normalTransform;
+    }
+    private void checkNormalTransform() {
+        if (normalTransform == null) {
+            normalTransform = new Matrix3f(matrix);
+            normalTransform.invert();
+            normalTransform.transpose();
+        }
+    }
+
     public Transformation slerp(Transformation p_175938_, float p_175939_) {
         return new Transformation(
             this.getTranslation().lerp(p_175938_.getTranslation(), p_175939_, new Vector3f()),
