--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -164,6 +_,19 @@
         return Objects.hash(this.matrix);
     }
 
+    private Matrix3f normalTransform = null;
+    public Matrix3f getNormalMatrix() {
+        checkNormalTransform();
+        return normalTransform;
+    }
+    private void checkNormalTransform() {
+        if (normalTransform == null) {
+            normalTransform = new Matrix3f(matrix);
+            normalTransform.invert();
+            normalTransform.transpose();
+        }
+    }
+
     public Transformation slerp(Transformation p_175938_, float p_175939_) {
         Vector3f vector3f = this.getTranslation();
         Quaternionf quaternionf = this.getLeftRotation();
