--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -87,6 +_,15 @@
         OptionSpec<Void> optionspec13 = optionparser.accepts("jfrProfile");
         OptionSpec<Path> optionspec14 = optionparser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
         OptionSpec<String> optionspec15 = optionparser.nonOptions();
+        optionparser.accepts("allowUpdates").withRequiredArg().ofType(Boolean.class).defaultsTo(Boolean.TRUE); // Forge: allow mod updates to proceed
+        optionparser.accepts("gameDir").withRequiredArg().ofType(File.class).defaultsTo(new File(".")); //Forge: Consume this argument, we use it in the launcher, and the client side.
+        final OptionSpec<net.minecraft.core.BlockPos> spawnPosOpt;
+        boolean gametestEnabled = Boolean.getBoolean("neoforge.gameTestServer");
+        if (gametestEnabled) {
+            spawnPosOpt = optionparser.accepts("spawnPos").withRequiredArg().withValuesConvertedBy(new net.neoforged.neoforge.gametest.BlockPosValueConverter()).defaultsTo(new net.minecraft.core.BlockPos(0, 60, 0));
+        } else {
+             spawnPosOpt = null;
+        }
 
         try {
             OptionSet optionset = optionparser.parse(p_129699_);
@@ -94,6 +_,13 @@
                 optionparser.printHelpOn(System.err);
                 return;
             }
+            Path path2 = Paths.get("eula.txt");
+            Eula eula = new Eula(path2);
+
+            if (!eula.hasAgreedToEULA()) {
+                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+                return;
+            }
 
             Path path = optionset.valueOf(optionspec14);
             if (path != null) {
@@ -109,24 +_,22 @@
             Bootstrap.validate();
             Util.startTimerHackThread();
             Path path1 = Paths.get("server.properties");
+            if (!optionset.has(optionspec1)) net.neoforged.neoforge.server.loading.ServerModLoader.load(); // Load mods before we load almost anything else anymore. Single spot now. Only loads if they haven't passed the initserver param
             DedicatedServerSettings dedicatedserversettings = new DedicatedServerSettings(path1);
             dedicatedserversettings.forceSave();
             RegionFileVersion.configure(dedicatedserversettings.getProperties().regionFileComression);
-            Path path2 = Paths.get("eula.txt");
-            Eula eula = new Eula(path2);
             if (optionset.has(optionspec1)) {
                 LOGGER.info("Initialized '{}' and '{}'", path1.toAbsolutePath(), path2.toAbsolutePath());
                 return;
             }
 
-            if (!eula.hasAgreedToEULA()) {
-                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-                return;
-            }
-
             File file1 = new File(optionset.valueOf(optionspec9));
             Services services = Services.create(new YggdrasilAuthenticationService(Proxy.NO_PROXY), file1);
             String s = Optional.ofNullable(optionset.valueOf(optionspec10)).orElse(dedicatedserversettings.getProperties().levelName);
+            if (s == null || s.isEmpty() || new File(file1, s).getAbsolutePath().equals(new File(s).getAbsolutePath())) {
+                LOGGER.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! " + s);
+                return;
+            }
             LevelStorageSource levelstoragesource = LevelStorageSource.createDefault(file1.toPath());
             LevelStorageSource.LevelStorageAccess levelstoragesource$levelstorageaccess = levelstoragesource.validateAndCreateAccess(s);
             Dynamic<?> dynamic;
@@ -135,6 +_,7 @@
                 try {
                     dynamic = levelstoragesource$levelstorageaccess.getDataTag();
                     levelsummary = levelstoragesource$levelstorageaccess.getSummary(dynamic);
+                    levelstoragesource$levelstorageaccess.readAdditionalLevelSaveData(false);
                 } catch (NbtException | ReportedNbtException | IOException ioexception1) {
                     LevelStorageSource.LevelDirectory levelstoragesource$leveldirectory = levelstoragesource$levelstorageaccess.getLevelDirectory();
                     LOGGER.warn("Failed to load world data from {}", levelstoragesource$leveldirectory.dataFile(), ioexception1);
@@ -143,6 +_,7 @@
                     try {
                         dynamic = levelstoragesource$levelstorageaccess.getDataTagFallback();
                         levelsummary = levelstoragesource$levelstorageaccess.getSummary(dynamic);
+                        levelstoragesource$levelstorageaccess.readAdditionalLevelSaveData(true);
                     } catch (NbtException | ReportedNbtException | IOException ioexception) {
                         LOGGER.error("Failed to load world data from {}", levelstoragesource$leveldirectory.oldDataFile(), ioexception);
                         LOGGER.error(
@@ -177,6 +_,15 @@
 
             PackRepository packrepository = ServerPacksSource.createPackRepository(levelstoragesource$levelstorageaccess);
 
+            if (gametestEnabled) {
+                net.neoforged.neoforge.gametest.GameTestHooks.registerGametests();
+                net.minecraft.core.BlockPos spawnPos = optionset.valueOf(spawnPosOpt);
+                MinecraftServer.spin(thread -> net.minecraft.gametest.framework.GameTestServer.create(thread, levelstoragesource$levelstorageaccess, packrepository, net.minecraft.gametest.framework.GameTestRegistry.getAllTestFunctions(), spawnPos));
+                // If we're running a gametest server we don't need to load the resources normally (GameTestServer#create does it using a flat world)
+                // or create a shutdown thread as the gametest server will always exit itself
+                return;
+            }
+
             WorldStem worldstem;
             try {
                 WorldLoader.InitConfig worldloader$initconfig = loadOrCreateConfig(dedicatedserversettings.getProperties(), dynamic1, flag, packrepository);
@@ -218,6 +_,9 @@
                                             worlddimensions = dedicatedserverproperties.createDimensions(p_359487_.datapackWorldgen());
                                         }
 
+                                        // Neo: Do a write-read-cycle to inject modded dimensions on first start of a dedicated server into its generated world dimensions list.
+                                        var registryOps = net.minecraft.resources.RegistryOps.create(net.minecraft.nbt.NbtOps.INSTANCE, p_359487_.datapackWorldgen());
+                                        worlddimensions = WorldDimensions.CODEC.encoder().encodeStart(registryOps, worlddimensions).flatMap((writtenPayloadWithModdedDimensions) -> WorldDimensions.CODEC.decoder().parse(registryOps, writtenPayloadWithModdedDimensions)).resultOrPartial(LOGGER::error).orElse(worlddimensions);
                                         WorldDimensions.Complete worlddimensions$complete = worlddimensions.bake(registry);
                                         Lifecycle lifecycle = worlddimensions$complete.lifecycle().add(p_359487_.datapackWorldgen().allRegistriesLifecycle());
                                         return new WorldLoader.DataLoadOutput<>(
@@ -277,6 +_,7 @@
                 @Override
                 public void run() {
                     dedicatedserver.halt(true);
+                    org.apache.logging.log4j.LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
                 }
             };
             thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
