--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -296,7 +_,7 @@
 
     public static <S extends MinecraftServer> S spin(Function<Thread, S> factory) {
         AtomicReference<S> serverReference = new AtomicReference<>();
-        Thread thread = new Thread(() -> serverReference.get().runServer(), "Server thread");
+        Thread thread = new Thread(net.neoforged.fml.util.thread.SidedThreadGroups.SERVER, () -> serverReference.get().runServer(), "Server thread");
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         if (Runtime.getRuntime().availableProcessors() > 4) {
             thread.setPriority(8);
@@ -350,9 +_,9 @@
         this.structureTemplateManager = new StructureTemplateManager(worldStem.resourceManager(), storageSource, fixerUpper, blockLookup);
         this.serverThread = serverThread;
         this.executor = Util.backgroundExecutor();
-        this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures());
+        this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures(), this.registryAccess());
         this.resources.managers.getRecipeManager().finalizeRecipeLoading(this.worldData.enabledFeatures());
-        this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+        this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registries.compositeAccess(), this.worldData.enabledFeatures());
         this.tickFrame = TracyClient.createDiscontinuousFrame("Server Tick");
         this.notificationManager = notificationManager;
         this.serverActivityMonitor = new ServerActivityMonitor(notificationManager, 30);
@@ -436,6 +_,7 @@
         this.scoreboard.load(this.savedDataStorage.computeIfAbsent(ScoreboardSaveData.TYPE).getData());
         this.commandStorage = new CommandStorage(this.savedDataStorage);
         this.stopwatches = this.savedDataStorage.computeIfAbsent(Stopwatches.TYPE);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Load(levels.get(Level.OVERWORLD)));
         if (!levelData.isInitialized()) {
             try {
                 setInitialSpawn(overworld, levelData, worldOptions.generateBonusChest(), isDebug, this.levelLoadListener);
@@ -470,6 +_,7 @@
                     this, this.executor, this.storageSource, derivedLevelData, dimension, entry.getValue(), isDebug, biomeZoomSeed, ImmutableList.of(), false
                 );
                 this.levels.put(dimension, level);
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Load(levels.get(name)));
             } else {
                 level = overworld;
             }
@@ -488,6 +_,7 @@
             levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), BlockPos.ZERO.above(80), 0.0F, 0.0F));
         } else {
             ServerChunkCache chunkSource = level.getChunkSource();
+            if (net.neoforged.neoforge.event.EventHooks.onCreateWorldSpawn(level, levelData)) return;
             ChunkPos spawnChunk = ChunkPos.containing(chunkSource.randomState().sampler().findSpawnPosition());
             levelLoadListener.start(LevelLoadListener.Stage.PREPARE_GLOBAL_SPAWN, 0);
             levelLoadListener.updateFocus(level.dimension(), spawnChunk);
@@ -550,6 +_,7 @@
                 TicketStorage savedTickets = level.getDataStorage().get(TicketStorage.TYPE);
                 if (savedTickets != null) {
                     savedTickets.activateAllDeactivatedTickets();
+                    net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.activateAllDeactivatedTickets(level, savedTickets);
                 }
             });
         }
@@ -675,6 +_,7 @@
             if (level != null) {
                 try {
                     level.close();
+                    net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(level));
                 } catch (IOException e) {
                     LOGGER.error("Exception closing the level", e);
                 }
@@ -721,9 +_,11 @@
                 throw new IllegalStateException("Failed to initialize server");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarted(this);
             this.nextTickTimeNanos = Util.getNanos();
             this.statusIcon = this.loadStatusIcon().orElse(null);
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
 
             while (this.running) {
                 long thisTickNanos;
@@ -777,6 +_,8 @@
                 this.isReady = true;
                 JvmProfiler.INSTANCE.onServerTick(this.smoothedTickTimeMillis);
             }
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
         } catch (Throwable t) {
             LOGGER.error("Encountered an unexpected exception", t);
             CrashReport report = constructOrExtractCrashReport(t);
@@ -788,6 +_,7 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
             this.onServerCrash(report);
         } finally {
             try {
@@ -796,6 +_,7 @@
             } catch (Throwable t) {
                 LOGGER.error("Exception stopping the server", t);
             } finally {
+                net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopped(this);
                 this.onServerExit();
             }
         }
@@ -977,11 +_,13 @@
         }
 
         this.tickCount++;
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPre(haveTime, this);
         this.tickRateManager.tick();
         this.tickChildren(haveTime);
         if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
         }
 
         this.ticksUntilAutosave--;
@@ -999,6 +_,7 @@
         this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)tickTime / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
         this.logTickMethodTime(nano);
         profiler.pop();
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPost(haveTime, this);
     }
 
     protected void processPacketsAndTick(boolean sprinting) {
@@ -1029,6 +_,16 @@
         }
     }
 
+    private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+    private String cachedServerStatus; // Neo: cache the server status json in case a client spams requests
+    private void resetStatusCache(ServerStatus status) {
+        this.cachedServerStatus = GSON.toJson(ServerStatus.CODEC.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status)
+                .result().orElseThrow());
+    }
+    public String getStatusJson() {
+        return cachedServerStatus;
+    }
+
     private int computeNextAutosaveInterval() {
         float ticksPerSecond;
         if (this.tickRateManager.isSprinting()) {
@@ -1060,7 +_,8 @@
             Optional.of(players),
             Optional.of(ServerStatus.Version.current()),
             Optional.ofNullable(this.statusIcon),
-            this.enforceSecureProfile()
+            this.enforceSecureProfile(),
+            true //TODO Neo: Possible build a system which indicates what the status of the modded server is.
         );
     }
 
@@ -1105,9 +_,11 @@
         profiler.push("levels");
         this.updateEffectiveRespawnData();
 
-        for (ServerLevel level : this.getAllLevels()) {
+        for (ServerLevel level : this.getWorldArray()) {
+            long tickStart = Util.getNanos();
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPre(level, haveTime);
 
             try {
                 level.tick(haveTime);
@@ -1116,9 +_,11 @@
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPost(level, haveTime);
 
             profiler.pop();
             profiler.pop();
+            perWorldTickTimes.computeIfAbsent(level.dimension(), k -> new long[100])[this.tickCount % 100] = Util.getNanos() - tickStart;
         }
 
         profiler.popPush("connection");
@@ -1127,8 +_,8 @@
         this.playerList.tick();
         profiler.popPush("debugSubscribers");
         this.debugSubscribers.tick();
-        if (this.tickRateManager.runsNormally()) {
             profiler.popPush("gameTests");
+        if (net.neoforged.neoforge.gametest.GameTestHooks.isGametestEnabled() && this.tickRateManager.runsNormally()) {
             GameTestTicker.SINGLETON.tick();
         }
 
@@ -1160,6 +_,9 @@
     }
 
     public void forceGameTimeSynchronization() {
+        if (this.overworld() == null) {
+            return; // Neo: Our ephemeral JUnit server has no overworld
+        }
         ProfilerFiller profiler = Profiler.get();
         profiler.push("timeSync");
         this.playerList.broadcastAll(new ClientboundSetTimePacket(this.overworld().getGameTime(), Map.of()));
@@ -1213,7 +_,7 @@
     }
 
     public String getServerModName() {
-        return "vanilla";
+        return net.neoforged.neoforge.internal.BrandingControl.getServerBranding();
     }
 
     public ServerClockManager clockManager() {
@@ -1515,9 +_,9 @@
 
     public CompletableFuture<Void> reloadResources(Collection<String> packsToEnable) {
         CompletableFuture<Void> result = CompletableFuture.<ImmutableList>supplyAsync(
-                () -> packsToEnable.stream()
-                    .map(this.packRepository::getPack)
-                    .filter(Objects::nonNull)
+                () -> this.packRepository
+                    .rebuildSelected(packsToEnable)
+                    .stream()
                     .map(Pack::open)
                     .collect(ImmutableList.toImmutableList()),
                 this
@@ -1556,7 +_,8 @@
                 this.getPlayerList().reloadResources();
                 this.functionManager.replaceLibrary(this.resources.managers.getFunctionLibrary());
                 this.structureTemplateManager.onResourceManagerReload(this.resources.resourceManager);
-                this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.getPlayerList().getPlayers().forEach(this.getPlayerList()::sendPlayerPermissionLevel); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
             }, this);
         if (this.isSameThread()) {
             this.managedBlock(result::isDone);
@@ -1572,8 +_,10 @@
         FeatureFlagSet forcedFeatures = initMode ? FeatureFlagSet.of() : initialDataConfig.enabledFeatures();
         FeatureFlagSet allowedFeatures = initMode ? FeatureFlags.REGISTRY.allFlags() : initialDataConfig.enabledFeatures();
         packRepository.reload();
+        DataPackConfig.DEFAULT.addModPacks(net.neoforged.neoforge.common.CommonHooks.getModDataPacks());
+        dataPackConfig.addModPacks(net.neoforged.neoforge.common.CommonHooks.getModDataPacks());
         if (safeMode) {
-            return configureRepositoryWithSelection(packRepository, List.of("vanilla"), forcedFeatures, false);
+            return configureRepositoryWithSelection(packRepository, net.neoforged.neoforge.common.CommonHooks.getModDataPacksWithVanilla(), forcedFeatures, false);
         }
 
         Set<String> selected = Sets.newLinkedHashSet();
@@ -1620,6 +_,8 @@
             selected.add("vanilla");
         }
 
+        net.neoforged.neoforge.resource.ResourcePackLoader.reorderNewlyDiscoveredPacks(selected, dataPackConfig.getEnabled(), packRepository);
+
         return configureRepositoryWithSelection(packRepository, selected, forcedFeatures, true);
     }
 
@@ -1843,6 +_,31 @@
 
     public abstract boolean isSingleplayerOwner(NameAndId nameAndId);
 
+    private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+    @Nullable
+    public long[] getTickTime(ResourceKey<Level> dim) {
+        return perWorldTickTimes.get(dim);
+    }
+
+    @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+    public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+        return this.levels;
+    }
+    private int worldArrayMarker = 0;
+    private int worldArrayLast = -1;
+    private ServerLevel[] worldArray;
+    @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+    public synchronized void markWorldsDirty() {
+        worldArrayMarker++;
+    }
+    private ServerLevel[] getWorldArray() {
+        if (worldArrayMarker == worldArrayLast && worldArray != null)
+            return worldArray;
+        worldArray = this.levels.values().stream().toArray(x -> new ServerLevel[x]);
+        worldArrayLast = worldArrayMarker;
+        return worldArray;
+    }
+
     public void dumpServerProperties(Path path) throws IOException {
     }
 
@@ -2002,6 +_,10 @@
         return this.worldData;
     }
 
+    public MinecraftServer.ReloadableResources getServerResources() {
+         return resources;
+    }
+
     public RegistryAccess.Frozen registryAccess() {
         return this.registries.compositeAccess();
     }
@@ -2073,6 +_,13 @@
         return true;
     }
 
+    public void handleCustomClickAction(Identifier id, Optional<Tag> payload, @Nullable ServerPlayer player, GameProfile profile) {
+        if (!net.neoforged.neoforge.common.CommonHooks.onCustomClickAction(player, profile, id, payload))
+            handleCustomClickAction(id, payload);
+    }
+
+    /// @deprecated Neo: Use the overload that has the player context and triggers the appropriate event.
+    @Deprecated
     public void handleCustomClickAction(Identifier id, Optional<Tag> payload) {
         LOGGER.debug("Received custom click action {} with payload {}", id, payload.orElse(null));
     }
@@ -2132,6 +_,9 @@
         } else if (rule == GameRules.ADVANCE_TIME) {
             this.getPlayerList().broadcastAll(this.clockManager().createFullSyncPacket());
         }
+
+        // Neo: Allow mods to react to game rule state changes
+        net.neoforged.neoforge.event.EventHooks.onGameRuleChanged(this, rule, value);
     }
 
     @Deprecated
