--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -215,7 +_,7 @@
         Function<Thread, S> p_129873_
     ) {
         AtomicReference<S> atomicreference = new AtomicReference<>();
-        Thread thread = new Thread(() -> atomicreference.get().runServer(p_422046_, p_422613_, p_423185_, p_423070_), "Server thread");
+        Thread thread = new Thread(net.neoforged.fml.util.thread.SidedThreadGroups.SERVER, () -> atomicreference.get().runServer(p_422046_, p_422613_, p_423185_, p_423070_), "Server thread");
         thread.setUncaughtExceptionHandler((p_177909_, p_177910_) -> LOGGER.error("Uncaught exception in server thread", p_177910_));
         if (Runtime.getRuntime().availableProcessors() > 4) {
             thread.setPriority(8);
@@ -345,6 +_,7 @@
 
     protected void runServer(PackRepository p_423231_, WorldStem p_421508_, MinecraftServer.WorldReloader p_421684_, ChunkProgressListenerFactory p_419541_) {
         try {
+            net.neoforged.neoforge.server.ServerLifecycleHooks.loadConfigs(this);
             if (!this.initServer()) {
                 throw new IllegalStateException("Failed to initialize server");
             }
@@ -352,12 +_,16 @@
             while (this.running) {
                 TheGame thegame = this.initGame(p_423231_, p_421508_, p_419541_);
                 this.theGame = thegame;
+                net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(thegame);
                 if (this.reloadFuture != null) {
                     this.restoreAllPlayersConfigPlayers(thegame);
                     this.rejoinFuture = this.reloadFuture;
                     this.reloadFuture = null;
                 }
 
+                this.status = this.buildServerStatus(thegame);
+                resetStatusCache(status);
+
                 this.innerServerLoopWowo(thegame);
                 if (this.reloadFuture != null) {
                     this.waitForPlayersToEnterConfig();
@@ -381,6 +_,7 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
             this.onServerCrash(crashreport);
         } finally {
             try {
@@ -393,6 +_,7 @@
                     this.services.profileCache().clearExecutor();
                 }
 
+                net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopped(this);
                 this.onServerExit();
             }
         }
@@ -447,6 +_,8 @@
         this.status = this.buildServerStatus(p_420688_);
         ServerTickRateManager servertickratemanager = p_420688_.tickRateManager();
 
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarted(this);
+
         while (this.running && this.reloadFuture == null) {
             if (this.rejoinFuture != null
                 && this.connection
@@ -506,6 +_,8 @@
 
             this.isReady = true;
             JvmProfiler.INSTANCE.onServerTick(this.smoothedTickTimeMillis);
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
         }
     }
 
@@ -704,11 +_,13 @@
         }
 
         this.tickCount++;
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPre(p_129871_, this);
         p_420275_.tickRateManager().tick();
         this.tickChildren(p_420275_, p_129871_);
         if (i - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = i;
             this.status = this.buildServerStatus(p_420275_);
+            resetStatusCache(status);
         }
 
         this.ticksUntilAutosave--;
@@ -726,6 +_,7 @@
         this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)k / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
         this.logTickMethodTime(i);
         profilerfiller.pop();
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPost(p_129871_, this);
     }
 
     private void autoSave(TheGame p_421580_) {
@@ -776,7 +_,8 @@
             Optional.of(serverstatus$players),
             Optional.of(ServerStatus.Version.current()),
             Optional.ofNullable(this.statusIcon),
-            this.enforceSecureProfile()
+            this.enforceSecureProfile(),
+            true //TODO Neo: Possible build a system which indicates what the status of the modded server is.
         );
     }
 
@@ -808,7 +_,8 @@
         p_420534_.getFunctions().tick();
         profilerfiller.popPush("levels");
 
-        for (ServerLevel serverlevel : p_420534_.getAllLevels()) {
+        for (ServerLevel serverlevel : this.getWorldArray()) {
+            long tickStart = Util.getNanos();
             profilerfiller.push(() -> serverlevel + " " + serverlevel.dimension().location());
             if (this.tickCount % 20 == 0) {
                 profilerfiller.push("timeSync");
@@ -817,6 +_,7 @@
             }
 
             profilerfiller.push("tick");
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPre(serverlevel, p_129954_);
 
             try {
                 serverlevel.tick(p_129954_);
@@ -825,16 +_,18 @@
                 serverlevel.fillReportDetails(crashreport);
                 throw new ReportedException(crashreport);
             }
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPost(serverlevel, p_129954_);
 
             profilerfiller.pop();
             profilerfiller.pop();
+            perWorldTickTimes.computeIfAbsent(serverlevel.dimension(), k -> new long[100])[this.tickCount % 100] = Util.getNanos() - tickStart;
         }
 
         profilerfiller.popPush("connection");
         this.tickConnection();
         profilerfiller.popPush("players");
         playerlist.tick();
-        if (p_420534_.tickRateManager().runsNormally()) {
+        if (net.neoforged.neoforge.gametest.GameTestHooks.isGametestEnabled() && this.theGame.tickRateManager().runsNormally()) {
             GameTestTicker.SINGLETON.tick();
         }
 
@@ -897,7 +_,7 @@
 
     @DontObfuscate
     public String getServerModName() {
-        return "vanilla";
+        return net.neoforged.neoforge.internal.BrandingControl.getServerBranding();
     }
 
     public SystemReport fillSystemReport(SystemReport p_177936_) {
@@ -1231,6 +_,45 @@
     }
 
     public abstract boolean isSingleplayerOwner(GameProfile p_129840_);
+
+    private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+    private String cachedServerStatus; // Neo: cache the server status json in case a client spams requests
+    private void resetStatusCache(ServerStatus status) {
+        this.cachedServerStatus = GSON.toJson(ServerStatus.CODEC.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status)
+                .result().orElseThrow());
+    }
+    public String getStatusJson() {
+        return cachedServerStatus;
+    }
+
+    private java.util.Map<net.minecraft.resources.ResourceKey<Level>, long[]> perWorldTickTimes = com.google.common.collect.Maps.newIdentityHashMap();
+    @Nullable
+    public long[] getTickTime(net.minecraft.resources.ResourceKey<Level> dim) {
+        return perWorldTickTimes.get(dim);
+    }
+
+    @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+    public synchronized java.util.Map<net.minecraft.resources.ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+        var map = new java.util.HashMap<net.minecraft.resources.ResourceKey<Level>, ServerLevel>();
+        for (ServerLevel level : this.theGame.getAllLevels()) {
+            map.put(level.dimension(), level);
+        }
+        return map;
+    }
+    private int worldArrayMarker = 0;
+    private int worldArrayLast = -1;
+    private ServerLevel[] worldArray;
+    @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+    public synchronized void markWorldsDirty() {
+        worldArrayMarker++;
+    }
+    private ServerLevel[] getWorldArray() {
+        if (worldArrayMarker == worldArrayLast && worldArray != null)
+            return worldArray;
+        worldArray = this.theGame.getAllLevels().stream().toArray(x -> new ServerLevel[x]);
+        worldArrayLast = worldArrayMarker;
+        return worldArray;
+    }
 
     public void dumpServerProperties(Path p_177911_) throws IOException {
     }
