--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -166,6 +_,8 @@
     }
 
     public boolean award(AdvancementHolder advancement, String criterionKey) {
+        // Forge: don't grant advancements for fake players
+        if (this.player instanceof net.neoforged.neoforge.common.util.FakePlayer) return false;
         boolean flag = false;
         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
         boolean flag1 = advancementprogress.isDone();
@@ -173,12 +_,14 @@
             this.unregisterListeners(advancement);
             this.progressChanged.add(advancement);
             flag = true;
+            net.neoforged.neoforge.event.EventHooks.onAdvancementProgressedEvent(this.player, advancement, advancementprogress, criterionKey, net.neoforged.neoforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.GRANT);
             if (!flag1 && advancementprogress.isDone()) {
                 advancement.value().rewards().grant(this.player);
                 advancement.value().display().ifPresent(p_352686_ -> {
                     if (p_352686_.shouldAnnounceChat() && this.player.level().getGameRules().getBoolean(GameRules.RULE_ANNOUNCE_ADVANCEMENTS)) {
                         this.playerList.broadcastSystemMessage(p_352686_.getType().createAnnouncement(advancement, this.player), false);
                     }
+                    net.neoforged.neoforge.event.EventHooks.onAdvancementEarnedEvent(this.player, advancement);
                 });
             }
         }
@@ -198,6 +_,7 @@
             this.registerListeners(advancement);
             this.progressChanged.add(advancement);
             flag = true;
+            net.neoforged.neoforge.event.EventHooks.onAdvancementProgressedEvent(this.player, advancement, advancementprogress, criterionKey, net.neoforged.neoforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.REVOKE);
         }
 
         if (flag1 && !advancementprogress.isDone()) {
