--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -167,6 +167,8 @@
    }
 
    public boolean award(AdvancementHolder p_300979_, String p_135990_) {
+      // Forge: don't grant advancements for fake players
+      if (this.player instanceof net.neoforged.neoforge.common.util.FakePlayer) return false;
       boolean flag = false;
       AdvancementProgress advancementprogress = this.getOrStartProgress(p_300979_);
       boolean flag1 = advancementprogress.isDone();
@@ -174,6 +176,7 @@
          this.unregisterListeners(p_300979_);
          this.progressChanged.add(p_300979_);
          flag = true;
+         net.neoforged.neoforge.event.ForgeEventFactory.onAdvancementProgressedEvent(this.player, p_300979_, advancementprogress, p_135990_, net.neoforged.neoforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.GRANT);
          if (!flag1 && advancementprogress.isDone()) {
             p_300979_.value().rewards().grant(this.player);
             p_300979_.value()
@@ -189,6 +192,7 @@
                               false
                            );
                      }
+                     net.neoforged.neoforge.event.ForgeEventFactory.onAdvancementEarnedEvent(this.player, p_300979_);
                   }
                );
          }
@@ -209,6 +213,7 @@
          this.registerListeners(p_301073_);
          this.progressChanged.add(p_301073_);
          flag = true;
+         net.neoforged.neoforge.event.ForgeEventFactory.onAdvancementProgressedEvent(this.player, p_301073_, advancementprogress, p_136000_, net.neoforged.neoforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.REVOKE);
       }
 
       if (flag1 && !advancementprogress.isDone()) {
