--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -107,7 +_,7 @@
         }
     }
 
-    protected void load(ServerAdvancementManager manager) {
+    private void load(ServerAdvancementManager manager) {
         if (Files.isRegularFile(this.playerSavePath)) {
             try (Reader reader = Files.newBufferedReader(this.playerSavePath, StandardCharsets.UTF_8)) {
                 JsonElement json = StrictJsonParser.parse(reader);
@@ -142,6 +_,7 @@
         data.forEach((id, progress) -> {
             AdvancementHolder advancement = manager.get(id);
             if (advancement == null) {
+                if (!id.getNamespace().equals("minecraft")) return; // CraftBukkit
                 LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", id, this.playerSavePath);
             } else {
                 this.startProgress(advancement, progress);
@@ -170,6 +_,7 @@
             this.progressChanged.add(holder);
             result = true;
             if (!wasDone && progress.isDone()) {
+                this.player.level().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.player.PlayerAdvancementDoneEvent(this.player.getBukkitEntity(), holder.toBukkit())); // CraftBukkit
                 holder.value().rewards().grant(this.player);
                 holder.value().display().ifPresent(display -> {
                     if (display.shouldAnnounceChat() && this.player.level().getGameRules().get(GameRules.SHOW_ADVANCEMENT_MESSAGES)) {
