--- a/net/minecraft/server/ReloadableServerRegistries.java
+++ b/net/minecraft/server/ReloadableServerRegistries.java
@@ -60,7 +_,16 @@
         return CompletableFuture.supplyAsync(() -> {
             WritableRegistry<T> registry = new MappedRegistry<>(type.registryKey(), Lifecycle.experimental());
             Map<Identifier, T> elements = new HashMap<>();
-            SimpleJsonResourceReloadListener.scanDirectory(manager, type.registryKey(), ops, type.codec(), elements);
+            var provider = net.neoforged.neoforge.common.CommonHooks.extractLookupProvider(ops);
+            Map<Identifier, Optional<T>> optionalMap = new HashMap<>();
+            SimpleJsonResourceReloadListener.scanDirectoryWithOptionalValues(manager, type.registryKey(), ops, type.conditionalCodec(), optionalMap);
+            optionalMap.forEach((rl, optionalEntry) -> {
+                optionalEntry.ifPresent(entry -> type.idSetter().accept(entry, rl));
+                T value = optionalEntry.orElse(type.defaultValue());
+                if (value instanceof LootTable lootTable) value = (T) net.neoforged.neoforge.event.EventHooks.loadLootTable(provider, rl, lootTable);
+                if (value != null)
+                    elements.put(rl, value);
+            });
             elements.forEach((id, element) -> registry.register(ResourceKey.create(type.registryKey(), id), (T)element, DEFAULT_REGISTRATION_INFO));
             TagLoader.loadTagsForRegistry(manager, registry);
             return registry;
