--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -47,6 +_,9 @@
         this.commands = new Commands(commandSelection, CommandBuildContext.simple(loadingContext, enabledFeatures));
         this.advancements = new ServerAdvancementManager(loadingContext);
         this.functionLibrary = new ServerFunctionLibrary(functionCompilationPermissions, this.commands.getDispatcher());
+        // Neo: Store registries and create context object
+        this.registryLookup = loadingContext;
+        this.context = new net.neoforged.neoforge.common.conditions.ConditionContext(this.postponedTags, fullLayers.compositeAccess(), enabledFeatures);
     }
 
     public ServerFunctionLibrary getFunctionLibrary() {
@@ -71,6 +_,25 @@
 
     public List<PreparableReloadListener> listeners() {
         return List.of(this.recipes, this.functionLibrary, this.advancements);
+    }
+
+    private final HolderLookup.Provider registryLookup;
+    private final net.neoforged.neoforge.common.conditions.ConditionContext context;
+
+    /**
+     * Exposes the current condition context for usage in other reload listeners.<br>
+     * This is not useful outside the reloading stage.
+     * @return The condition context for the currently active reload.
+     */
+    public net.neoforged.neoforge.common.conditions.ICondition.IContext getConditionContext() {
+        return this.context;
+    }
+
+    /**
+      * {@return the lookup provider access for the currently active reload}
+      */
+    public HolderLookup.Provider getRegistryLookup() {
+        return this.registryLookup;
     }
 
     public static CompletableFuture<ReloadableServerResources> loadResources(
