--- a/net/minecraft/server/ReloadableServerRegistries.java
+++ b/net/minecraft/server/ReloadableServerRegistries.java
@@ -60,12 +_,13 @@
     ) {
         return CompletableFuture.supplyAsync(
             () -> {
+                var provider = net.neoforged.neoforge.common.CommonHooks.extractLookupProvider(p_336173_);
                 WritableRegistry<T> writableregistry = new MappedRegistry<>(p_335741_.registryKey(), Lifecycle.experimental());
                 Map<ResourceLocation, JsonElement> map = new HashMap<>();
                 String s = Registries.elementsDirPath(p_335741_.registryKey());
                 SimpleJsonResourceReloadListener.scanDirectory(p_335893_, s, GSON, map);
                 map.forEach(
-                    (p_335614_, p_335474_) -> p_335741_.deserialize(p_335614_, p_336173_, p_335474_)
+                    (p_335614_, p_335474_) -> p_335741_.deserialize(provider, p_335614_, p_336173_, p_335474_)
                             .ifPresent(
                                 p_335683_ -> writableregistry.register(
                                         ResourceKey.create(p_335741_.registryKey(), p_335614_), (T)p_335683_, DEFAULT_REGISTRATION_INFO
