--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -47,6 +_,9 @@
         this.commands = new Commands(commandSelection, CommandBuildContext.simple(loadingContext, enabledFeatures));
         this.advancements = new ServerAdvancementManager(loadingContext);
         this.functionLibrary = new ServerFunctionLibrary(functionCompilationPermissions, this.commands.getDispatcher());
+        // Neo: Store registries and create context object
+        this.registryLookup = loadingContext;
+        this.context = new net.neoforged.neoforge.common.conditions.ConditionContext(this.postponedTags, fullLayers.compositeAccess(), enabledFeatures);
     }
 
     public ServerFunctionLibrary getFunctionLibrary() {
@@ -73,6 +_,25 @@
         return List.of(this.recipes, this.functionLibrary, this.advancements);
     }
 
+    private final HolderLookup.Provider registryLookup;
+    private final net.neoforged.neoforge.common.conditions.ConditionContext context;
+
+    /**
+     * Exposes the current condition context for usage in other reload listeners.<br>
+     * This is not useful outside the reloading stage.
+     * @return The condition context for the currently active reload.
+     */
+    public net.neoforged.neoforge.common.conditions.ICondition.IContext getConditionContext() {
+        return this.context;
+    }
+
+    /**
+      * {@return the lookup provider access for the currently active reload}
+      */
+    public HolderLookup.Provider getRegistryLookup() {
+        return this.registryLookup;
+    }
+
     public static CompletableFuture<ReloadableServerResources> loadResources(
         ResourceManager resourceManager,
         LayeredRegistryAccess<RegistryLayer> contextLayers,
@@ -99,15 +_,35 @@
                                 functionCompilationPermissions,
                                 (List<DataComponentInitializers.PendingComponents<?>>)pendingComponents
                             );
+
+                            // Neo: Fire the AddServerReloadListenersEvent and use the resulting listeners instead of the vanilla listener list.
+                            List<PreparableReloadListener> listeners = net.neoforged.neoforge.event.EventHooks.onResourceReload(result, fullRegistries.layers().compositeAccess());
+
+                            // Neo: Inject the ConditionContext and RegistryLookup to any resource listener that requests it.
+                            for (PreparableReloadListener rl : listeners) {
+                                if (rl instanceof net.neoforged.neoforge.resource.ContextAwareReloadListener carl) {
+                                    carl.injectContext(result.context, result.registryLookup);
+                                }
+                            }
+
                             return SimpleReloadInstance.create(
                                     resourceManager,
-                                    result.listeners(),
+                                    listeners,
                                     backgroundExecutor,
                                     mainThreadExecutor,
                                     DATA_RELOAD_INITIAL_TASK,
                                     LOGGER.isDebugEnabled()
                                 )
                                 .done()
+                                .thenRun(() -> {
+                                    // Neo: Clear context after reload completes
+                                    result.context.clear();
+                                    listeners.forEach(rl -> {
+                                        if (rl instanceof net.neoforged.neoforge.resource.ContextAwareReloadListener srl) {
+                                            srl.injectContext(net.neoforged.neoforge.common.conditions.ICondition.IContext.EMPTY, net.minecraft.core.RegistryAccess.EMPTY);
+                                        }
+                                    });
+                                })
                                 .thenApply(ignore -> result);
                         }
                     )
@@ -116,6 +_,8 @@
 
     public void updateComponentsAndStaticRegistryTags() {
         this.postponedTags.forEach(Registry.PendingTags::apply);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.TagsUpdatedEvent(registryLookup, false, false));
         this.newComponents.forEach(DataComponentInitializers.PendingComponents::apply);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.DefaultDataComponentsBoundEvent(false, false));
     }
 }
