--- a/net/minecraft/server/WorldLoader.java
+++ b/net/minecraft/server/WorldLoader.java
@@ -33,7 +_,7 @@
             CloseableResourceManager closeableresourcemanager = pair.getSecond();
             LayeredRegistryAccess<RegistryLayer> layeredregistryaccess = RegistryLayer.createRegistryAccess();
             LayeredRegistryAccess<RegistryLayer> layeredregistryaccess1 = loadAndReplaceLayer(
-                closeableresourcemanager, layeredregistryaccess, RegistryLayer.WORLDGEN, RegistryDataLoader.WORLDGEN_REGISTRIES
+                closeableresourcemanager, layeredregistryaccess, RegistryLayer.WORLDGEN, net.neoforged.neoforge.registries.DataPackRegistriesHooks.getDataPackRegistries()
             );
             RegistryAccess.Frozen registryaccess$frozen = layeredregistryaccess1.getAccessForLoading(RegistryLayer.DIMENSIONS);
             RegistryAccess.Frozen registryaccess$frozen1 = RegistryDataLoader.load(
@@ -53,7 +_,8 @@
                     p_214363_.commandSelection(),
                     p_214363_.functionCompilationLevel(),
                     p_214366_,
-                    p_214367_
+                    p_214367_,
+                    worlddataconfiguration.enabledModdedFeatures()
                 )
                 .whenComplete((p_214370_, p_214371_) -> {
                     if (p_214371_ != null) {
