--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -66,6 +66,7 @@
 
     public void setEntityId(EntityType<?> type, @Nullable Level level, RandomSource random, BlockPos pos) {
         this.getOrCreateNextSpawnData(level, random, pos).getEntityToSpawn().putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(type).toString());
+        this.spawnPotentials = WeightedList.of(); // CraftBukkit - SPIGOT-3496, MC-92282
     }
 
     private boolean isNearPlayer(Level level, BlockPos pos) {
@@ -168,13 +169,18 @@
                                     ((Mob) entity).finalizeSpawn(level, level.getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.SPAWNER, (SpawnGroupData) null);
                                 }
 
-                                Optional optional1 = spawndata.getEquipment();
+                                Optional<net.minecraft.world.entity.EquipmentTable> optional1 = spawndata.getEquipment(); // CraftBukkit - decompile error
 
                                 Objects.requireNonNull(mob);
                                 optional1.ifPresent(mob::equip);
                             }
 
-                            if (!level.tryAddFreshEntityWithPassengers(entity)) {
+                            // CraftBukkit start
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callSpawnerSpawnEvent(entity, pos).isCancelled()) {
+                                continue;
+                            }
+                            if (!level.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER)) {
+                                // CraftBukkit end
                                 this.delay(level, pos);
                                 return;
                             }
