--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -37,6 +37,13 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.entity.boss.enderdragon.EnderDragonPart;
+import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
+import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class ServerExplosion implements Explosion {
 
     private static final ExplosionDamageCalculator EXPLOSION_DAMAGE_CALCULATOR = new ExplosionDamageCalculator();
@@ -51,16 +58,21 @@
     private final DamageSource damageSource;
     private final ExplosionDamageCalculator damageCalculator;
     private final Map<Player, Vec3> hitPlayers = new HashMap();
+    // CraftBukkit - add field
+    public boolean wasCanceled = false;
+    public float yield;
+    // CraftBukkit end
 
     public ServerExplosion(ServerLevel level, @Nullable Entity source, @Nullable DamageSource damageSource, @Nullable ExplosionDamageCalculator damageCalculator, Vec3 center, float radius, boolean fire, Explosion.BlockInteraction blockInteraction) {
         this.level = level;
         this.source = source;
-        this.radius = radius;
+        this.radius = (float) Math.max(radius, 0.0); // CraftBukkit - clamp bad values
         this.center = center;
         this.fire = fire;
         this.blockInteraction = blockInteraction;
         this.damageSource = damageSource == null ? level.damageSources().explosion(this) : damageSource;
         this.damageCalculator = damageCalculator == null ? this.makeDamageCalculator(source) : damageCalculator;
+        this.yield = this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F; // CraftBukkit
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(@Nullable Entity source) {
@@ -174,7 +186,10 @@
             int i1 = Mth.floor(this.center.z - (double) f - 1.0D);
             int j1 = Mth.floor(this.center.z + (double) f + 1.0D);
 
-            for (Entity entity : this.level.getEntities(this.source, new AABB((double) i, (double) k, (double) i1, (double) j, (double) l, (double) j1))) {
+            // CraftBukkit start
+            List<Entity> list = this.level.getEntities(this.source, new AABB((double) i, (double) k, (double) i1, (double) j, (double) l, (double) j1));
+            for (Entity entity : list) {
+                // CraftBukkit end
                 if (!entity.ignoreExplosion(this)) {
                     double d0 = Math.sqrt(entity.distanceToSqr(this.center)) / (double) f;
 
@@ -186,7 +201,35 @@
                         float f2 = !flag && f1 == 0.0F ? 0.0F : getSeenPercent(this.center, entity);
 
                         if (flag) {
-                            entity.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, f2));
+                            // CraftBukkit start
+
+                            // Special case ender dragon only give knockback if no damage is cancelled
+                            // Thinks to note:
+                            // - Setting a velocity to a ComplexEntityPart is ignored (and therefore not needed)
+                            // - Damaging ComplexEntityPart while forward the damage to EntityEnderDragon
+                            // - Damaging EntityEnderDragon does nothing
+                            // - EntityEnderDragon hitbock always covers the other parts and is therefore always present
+                            if (entity instanceof EnderDragonPart) {
+                                continue;
+                            }
+
+                            entity.lastDamageCancelled = false;
+
+                            if (entity instanceof EnderDragon) {
+                                for (EnderDragonPart entityComplexPart : ((EnderDragon) entity).subEntities) {
+                                    // Calculate damage separately for each EntityComplexPart
+                                    if (list.contains(entityComplexPart)) {
+                                        entityComplexPart.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, f2));
+                                    }
+                                }
+                            } else {
+                                entity.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, f2));
+                            }
+
+                            if (entity.lastDamageCancelled) { // SPIGOT-5339, SPIGOT-6252, SPIGOT-6777: Skip entity if damage event was cancelled
+                                continue;
+                            }
+                            // CraftBukkit end
                         }
 
                         double d1;
@@ -202,6 +245,17 @@
                         double d2 = d1;
                         double d3 = (1.0D - d0) * (double) f2 * (double) f1 * (1.0D - d2);
                         Vec3 vec32 = vec31.scale(d3);
+                        // CraftBukkit start - Call EntityKnockbackEvent
+                        if (entity instanceof LivingEntity) {
+                            Vec3 result = entity.getDeltaMovement().add(vec32);
+                            org.bukkit.event.entity.EntityKnockbackEvent event = CraftEventFactory.callEntityKnockbackEvent((org.bukkit.craftbukkit.entity.CraftLivingEntity) entity.getBukkitEntity(), source, org.bukkit.event.entity.EntityKnockbackEvent.KnockbackCause.EXPLOSION, d3, vec32, result.x, result.y, result.z);
+
+                            // SPIGOT-7640: Need to subtract entity movement from the event result,
+                            // since the code below (the setDeltaMovement call as well as the hitPlayers map)
+                            // want the vector to be the relative velocity will the event provides the absolute velocity
+                            vec32 = (event.isCancelled()) ? Vec3.ZERO : new Vec3(event.getFinalKnockback().getX(), event.getFinalKnockback().getY(), event.getFinalKnockback().getZ()).subtract(entity.getDeltaMovement());
+                        }
+                        // CraftBukkit end
 
                         entity.push(vec32);
                         if (entity.is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile) {
@@ -228,8 +282,30 @@
         List<ServerExplosion.StackCollector> list1 = new ArrayList();
 
         Util.shuffle(targetBlocks, this.level.random);
+        // CraftBukkit start
+        List<org.bukkit.block.Block> bukkitBlocks = CraftEventFactory.handleExplodeEvent(this, targetBlocks);
+
+        targetBlocks.clear();
+        targetBlocks.addAll(bukkitBlocks.stream().map(bblock -> new BlockPos(bblock.getX(), bblock.getY(), bblock.getZ())).toList());
+
+        if (this.wasCanceled) {
+            return;
+        }
+        // CraftBukkit end
 
         for (BlockPos blockpos : targetBlocks) {
+            // CraftBukkit start - TNTPrimeEvent
+            BlockState blockstate = this.level.getBlockState(blockpos);
+            Block block = blockstate.getBlock();
+            if (block instanceof net.minecraft.world.level.block.TntBlock) {
+                BlockPos sourceBlock = source == null ? BlockPos.containing(this.center) : null;
+                if (!CraftEventFactory.callTNTPrimeEvent(this.level, blockpos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.EXPLOSION, source, sourceBlock)) {
+                    this.level.sendBlockUpdated(blockpos, Blocks.AIR.defaultBlockState(), blockstate, 3); // Update the block on the client
+                    continue;
+                }
+            }
+            // CraftBukkit end
+
             this.level.getBlockState(blockpos).onExplosionHit(this.level, blockpos, this, (itemstack, blockpos1) -> {
                 addOrAppendStack(list1, itemstack, blockpos1);
             });
@@ -244,13 +320,22 @@
     private void createFire(List<BlockPos> targetBlocks) {
         for (BlockPos blockpos : targetBlocks) {
             if (this.level.random.nextInt(3) == 0 && this.level.getBlockState(blockpos).isAir() && this.level.getBlockState(blockpos.below()).isSolidRender()) {
-                this.level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level, blockpos));
+                // CraftBukkit start - Ignition by explosion
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(this.level, blockpos, this).isCancelled()) {
+                    this.level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level, blockpos));
+                }
+                // CraftBukkit end
             }
         }
 
     }
 
     public int explode() {
+        // CraftBukkit start
+        if (this.radius < 0.1F) {
+            return 0;
+        }
+        // CraftBukkit end
         this.level.gameEvent(this.source, (Holder) GameEvent.EXPLODE, this.center);
         List<BlockPos> list = this.calculateExplodedPositions();
 
@@ -261,7 +346,15 @@
             profilerfiller.push("explosion_blocks");
             this.interactWithBlocks(list);
             profilerfiller.pop();
+            // CraftBukkit start - handle KEEP effect
+        } else {
+            Util.shuffle(list, this.level.random); // CraftBukkit - Copy from calculateExplodedPositions
+            List<org.bukkit.block.Block> bukkitBlocks = CraftEventFactory.handleExplodeEvent(this, list);
+
+            list.clear();
+            list.addAll(bukkitBlocks.stream().map(bblock -> new BlockPos(bblock.getX(), bblock.getY(), bblock.getZ())).toList());
         }
+        // CraftBukkit end
 
         if (this.fire) {
             this.createFire(list);
@@ -271,6 +364,7 @@
     }
 
     private static void addOrAppendStack(List<ServerExplosion.StackCollector> stacks, ItemStack stack, BlockPos pos) {
+        if (stack.isEmpty()) return; // CraftBukkit - SPIGOT-5425
         for (ServerExplosion.StackCollector serverexplosion_stackcollector : stacks) {
             serverexplosion_stackcollector.tryMerge(stack);
             if (stack.isEmpty()) {
@@ -315,13 +409,13 @@
 
     @Override
     public boolean canTriggerBlocks() {
-        return this.blockInteraction != Explosion.BlockInteraction.TRIGGER_BLOCK ? false : (this.source != null && this.source.is((Object) EntityTypes.BREEZE_WIND_CHARGE) ? (Boolean) this.level.getGameRules().get(GameRules.MOB_GRIEFING) : true);
+        return this.blockInteraction != Explosion.BlockInteraction.TRIGGER_BLOCK ? false : (this.source != null && this.source.is(EntityTypes.BREEZE_WIND_CHARGE) ? (Boolean) this.level.getGameRules().get(GameRules.MOB_GRIEFING) : true); // CraftBukkit - decompile error
     }
 
     @Override
     public boolean shouldAffectBlocklikeEntities() {
         boolean flag = (Boolean) this.level.getGameRules().get(GameRules.MOB_GRIEFING);
-        boolean flag1 = this.source == null || !this.source.is((Object) EntityTypes.BREEZE_WIND_CHARGE) && !this.source.is((Object) EntityTypes.WIND_CHARGE);
+        boolean flag1 = this.source == null || !this.source.is(EntityTypes.BREEZE_WIND_CHARGE) && !this.source.is(EntityTypes.WIND_CHARGE); // CraftBukkit - decompile error
 
         return flag ? flag1 : this.blockInteraction.shouldAffectBlocklikeEntities() && flag1;
     }
