--- a/com/mojang/blaze3d/systems/RenderSystem.java
+++ b/com/mojang/blaze3d/systems/RenderSystem.java
@@ -71,6 +71,7 @@
     private static @Nullable DynamicUniforms dynamicUniforms;
     private static final ScissorState scissorStateForRenderTypeDraws = new ScissorState();
     private static final SamplerCache samplerCache = new SamplerCache();
+    private static final net.neoforged.neoforge.client.pipeline.PipelineModifierStack PIPELINE_MODIFIERS = new net.neoforged.neoforge.client.pipeline.PipelineModifierStack();
 
     public static SamplerCache getSamplerCache() {
         return samplerCache;
