--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -51,7 +51,16 @@
 
             Vec3 vec3 = user.position();
 
-            if (user.randomTeleport(d0, d1, d2, true)) {
+            // CraftBukkit start - handle canceled status of teleport event
+            java.util.Optional<Boolean> status = user.randomTeleport(d0, d1, d2, true, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.CHORUS_FRUIT);
+
+            if (!status.isPresent()) {
+                // teleport event was canceled, no more tries
+                break;
+            }
+
+            if (status.get()) {
+                // CraftBukkit end
                 level.gameEvent(GameEvent.TELEPORT, vec3, GameEvent.Context.of((Entity) user));
                 SoundSource soundsource;
                 SoundEvent soundevent;
