--- a/com/mojang/blaze3d/vertex/BufferBuilder.java
+++ b/com/mojang/blaze3d/vertex/BufferBuilder.java
@@ -495,4 +_,22 @@
             this.sorting = p_277855_;
         }
     }
+
+    // Forge start
+    public void putBulkData(ByteBuffer buffer) {
+        ensureCapacity(buffer.limit() + this.format.getVertexSize());
+        this.buffer.position(this.nextElementByte);
+        this.buffer.put(buffer);
+        this.buffer.position(0);
+        this.vertices += buffer.limit() / this.format.getVertexSize();
+        this.nextElementByte += buffer.limit();
+    }
+
+    public ByteBuffer getBuffer() {
+        return this.buffer;
+    }
+
+    public int getNextElementByte() {
+        return this.nextElementByte;
+    }
 }
