--- a/com/mojang/blaze3d/vertex/VertexConsumer.java
+++ b/com/mojang/blaze3d/vertex/VertexConsumer.java
@@ -12,7 +_,7 @@
 import org.lwjgl.system.MemoryStack;
 
 @OnlyIn(Dist.CLIENT)
-public interface VertexConsumer {
+public interface VertexConsumer extends net.neoforged.neoforge.client.extensions.IVertexConsumerExtension {
     VertexConsumer addVertex(float x, float y, float z);
 
     VertexConsumer setColor(int red, int green, int blue, int alpha);
@@ -131,11 +_,14 @@
                     f5 = brightness[l] * blue * 255.0F;
                 }
 
-                int i1 = FastColor.ARGB32.color(k, (int)f3, (int)f4, (int)f5);
-                int j1 = lightmap[l];
+                // Neo: also apply alpha that's coming from the baked quad
+                int vertexAlpha = readAlpha ? (int)((alpha * (float) (bytebuffer.get(15) & 255) / 255.0F) * 255) : k;
+                int i1 = FastColor.ARGB32.color(vertexAlpha, (int)f3, (int)f4, (int)f5);
+                int j1 = applyBakedLighting(lightmap[l], bytebuffer);
                 float f10 = bytebuffer.getFloat(16);
                 float f9 = bytebuffer.getFloat(20);
                 Vector3f vector3f1 = matrix4f.transformPosition(f, f1, f2, new Vector3f());
+                applyBakedNormals(vector3f, bytebuffer, pose.normal());
                 this.addVertex(vector3f1.x(), vector3f1.y(), vector3f1.z(), i1, f10, f9, packedOverlay, j1, vector3f.x(), vector3f.y(), vector3f.z());
             }
         }
