--- a/com/mojang/blaze3d/vertex/VertexConsumer.java
+++ b/com/mojang/blaze3d/vertex/VertexConsumer.java
@@ -71,10 +_,12 @@
             long packedUv = quad.packedUV(vertex);
             float brightnessForVertex = brightness[vertex];
             int color = ARGB.colorFromFloat(a, brightnessForVertex * r, brightnessForVertex * g, brightnessForVertex * b);
+            color = ARGB.multiply(color, quad.bakedColors().color(vertex)); // Neo: apply baked color from the quad
             int light = LightCoordsUtil.lightCoordsWithEmission(lightmapCoord[vertex], lightEmission);
             Vector3f pos = matrix.transformPosition(position, new Vector3f());
             float u = UVPair.unpackU(packedUv);
             float v = UVPair.unpackV(packedUv);
+            applyBakedNormals(normal, quad.bakedNormals(), vertex, pose.normal()); // Neo: apply baked normals from the quad
             this.addVertex(pos.x(), pos.y(), pos.z(), color, u, v, overlayCoords, light, normal.x(), normal.y(), normal.z());
         }
     }
