--- a/com/mojang/blaze3d/vertex/VertexFormatElement.java
+++ b/com/mojang/blaze3d/vertex/VertexFormatElement.java
@@ -75,6 +_,15 @@
         return ELEMENTS.stream().filter(p_350573_ -> p_350573_ != null && (p_350349_ & p_350573_.mask()) != 0);
     }
 
+    public static int findNextId() {
+        for (int i = 0; i < BY_ID.length; i++) {
+            if (BY_ID[i] == null) {
+                return i;
+            }
+        }
+        throw new IllegalStateException("VertexFormatElement count limit exceeded");
+    }
+
     @OnlyIn(Dist.CLIENT)
     @DontObfuscate
     public static enum Type {
