--- a/com/mojang/blaze3d/vertex/VertexFormat.java
+++ b/com/mojang/blaze3d/vertex/VertexFormat.java
@@ -128,6 +_,7 @@
     }
 
     private GpuBuffer uploadToBufferWithWorkaround(@Nullable GpuBuffer p_428850_, ByteBuffer p_428822_, int p_428845_, Supplier<String> p_428834_) {
+        p_428845_ |= GpuBuffer.USAGE_PERF_FRAME_TRANSIENT;
         if (USE_STAGING_BUFFER_WORKAROUND) {
             GpuDevice gpudevice = RenderSystem.getDevice();
             if (p_428850_ == null) {
@@ -237,5 +_,29 @@
                 default -> 0;
             };
         }
+    }
+
+    public ImmutableMap<String, VertexFormatElement> getElementMapping() {
+        ImmutableMap.Builder<String, VertexFormatElement> builder = ImmutableMap.builder();
+        for (int i = 0; i < elements.size(); i++) {
+            builder.put(names.get(i), elements.get(i));
+        }
+        return builder.build();
+    }
+
+    public boolean hasPosition() {
+        return elements.stream().anyMatch(e -> e.usage() == VertexFormatElement.Usage.POSITION);
+    }
+
+    public boolean hasNormal() {
+        return elements.stream().anyMatch(e -> e.usage() == VertexFormatElement.Usage.NORMAL);
+    }
+
+    public boolean hasColor() {
+        return elements.stream().anyMatch(e -> e.usage() == VertexFormatElement.Usage.COLOR);
+    }
+
+    public boolean hasUV(int which) {
+        return elements.stream().anyMatch(e -> e.usage() == VertexFormatElement.Usage.UV && e.index() == which);
     }
 }
