--- a/com/mojang/blaze3d/vertex/VertexConsumer.java
+++ b/com/mojang/blaze3d/vertex/VertexConsumer.java
@@ -59,11 +_,12 @@
         for (int vertex = 0; vertex < 4; vertex++) {
             Vector3fc position = quad.position(vertex);
             long packedUv = quad.packedUV(vertex);
-            int vertexColor = brightness.scaleColor(vertex, color);
+            int vertexColor = brightness.scaleColor(vertex, net.minecraft.util.ARGB.multiply(color, quad.bakedColors().color(vertex))); // Neo: apply baked color from the quad
             int light = lightmapCoord.composeWithEmission(vertex, lightEmission);
             Vector3f pos = matrix.transformPosition(position, new Vector3f());
             float u = UVPair.unpackU(packedUv);
             float v = UVPair.unpackV(packedUv);
+            applyBakedNormals(normal, quad.bakedNormals(), vertex, pose.normal()); // Neo: apply baked normals from the quad
             this.addVertex(pos.x(), pos.y(), pos.z(), vertexColor, u, v, overlayCoords, light, normal.x(), normal.y(), normal.z());
         }
     }
