--- a/net/minecraft/world/entity/ai/behavior/GiveGiftToHero.java
+++ b/net/minecraft/world/entity/ai/behavior/GiveGiftToHero.java
@@ -21,6 +_,8 @@
     private static final int MIN_TIME_BETWEEN_GIFTS = 600;
     private static final int MAX_TIME_BETWEEN_GIFTS = 6600;
     private static final int TIME_TO_DELAY_FOR_HEAD_TO_FINISH_TURNING = 20;
+    /** @deprecated Neo: use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#RAID_HERO_GIFTS data map} instead */
+    @Deprecated
     private static final Map<ResourceKey<VillagerProfession>, ResourceKey<LootTable>> GIFTS = ImmutableMap.<ResourceKey<VillagerProfession>, ResourceKey<LootTable>>builder()
         .put(VillagerProfession.ARMORER, BuiltInLootTables.ARMORER_GIFT)
         .put(VillagerProfession.BUTCHER, BuiltInLootTables.BUTCHER_GIFT)
@@ -107,6 +_,11 @@
     private static ResourceKey<LootTable> getLootTableToThrow(Villager villager) {
         if (villager.isBaby()) {
             return BuiltInLootTables.BABY_VILLAGER_GIFT;
+        }
+
+        var gift = villager.getVillagerData().profession().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.RAID_HERO_GIFTS);
+        if (gift != null) {
+            return gift.lootTable();
         }
 
         Optional<ResourceKey<VillagerProfession>> profession = villager.getVillagerData().profession().unwrapKey();
