--- a/net/minecraft/world/entity/ai/behavior/GiveGiftToHero.java
+++ b/net/minecraft/world/entity/ai/behavior/GiveGiftToHero.java
@@ -22,6 +_,8 @@
     private static final int MIN_TIME_BETWEEN_GIFTS = 600;
     private static final int MAX_TIME_BETWEEN_GIFTS = 6600;
     private static final int TIME_TO_DELAY_FOR_HEAD_TO_FINISH_TURNING = 20;
+    /** @deprecated Neo: use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#RAID_HERO_GIFTS data map} instead */
+    @Deprecated
     private static final Map<VillagerProfession, ResourceKey<LootTable>> GIFTS = ImmutableMap.<VillagerProfession, ResourceKey<LootTable>>builder()
         .put(VillagerProfession.ARMORER, BuiltInLootTables.ARMORER_GIFT)
         .put(VillagerProfession.BUTCHER, BuiltInLootTables.BUTCHER_GIFT)
@@ -112,7 +_,8 @@
             return BuiltInLootTables.BABY_VILLAGER_GIFT;
         } else {
             VillagerProfession villagerprofession = p_372869_.getVillagerData().getProfession();
-            return GIFTS.getOrDefault(villagerprofession, BuiltInLootTables.UNEMPLOYED_GIFT);
+            var gift = net.minecraft.core.registries.BuiltInRegistries.VILLAGER_PROFESSION.wrapAsHolder(villagerprofession).getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.RAID_HERO_GIFTS);
+            return gift != null ? gift.lootTable() : BuiltInLootTables.UNEMPLOYED_GIFT;
         }
     }
 
