--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -46,7 +_,7 @@
     }
 
     protected boolean checkExtraStartConditions(ServerLevel level, Villager body) {
-        if (!level.getGameRules().get(GameRules.MOB_GRIEFING)) {
+        if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(level, body)) {
             return false;
         } else if (!body.getVillagerData().profession().is(VillagerProfession.FARMER)) {
             return false;
@@ -80,7 +_,7 @@
         BlockState state = level.getBlockState(blockPos);
         Block block = state.getBlock();
         Block blockBelow = level.getBlockState(blockPos.below()).getBlock();
-        return block instanceof CropBlock && ((CropBlock)block).isMaxAge(state) || state.isAir() && blockBelow instanceof FarmlandBlock;
+        return block instanceof CropBlock && ((CropBlock)block).isMaxAge(state) || state.isAir() && (blockBelow instanceof FarmlandBlock || blockBelow.builtInRegistryHolder().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS));
     }
 
     protected void start(ServerLevel level, Villager body, long timestamp) {
@@ -107,7 +_,7 @@
                     level.destroyBlock(this.aboveFarmlandPos, true, body);
                 }
 
-                if (blockState.isAir() && blockBelow instanceof FarmlandBlock && body.hasFarmSeeds()) {
+                if (blockState.isAir() && (blockBelow instanceof FarmlandBlock || blockBelow.builtInRegistryHolder().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS)) && body.hasFarmSeeds()) {
                     SimpleContainer inventory = body.getInventory();
 
                     for (int i = 0; i < inventory.getContainerSize(); i++) {
@@ -118,6 +_,11 @@
                             level.setBlockAndUpdate(this.aboveFarmlandPos, place);
                             level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(body, place));
                             ok = true;
+                        } else if (itemStack.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(body)) {
+                            if (specialPlantable.canPlacePlantAtPosition(itemStack, level, aboveFarmlandPos, net.minecraft.core.Direction.DOWN)) {
+                                specialPlantable.spawnPlantAtPosition(itemStack, level, aboveFarmlandPos, net.minecraft.core.Direction.DOWN);
+                                ok = true;
+                            }
                         }
 
                         if (ok) {
