--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -47,7 +_,7 @@
     }
 
     protected boolean checkExtraStartConditions(ServerLevel level, Villager owner) {
-        if (!level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+        if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(level, owner)) {
             return false;
         } else if (owner.getVillagerData().getProfession() != VillagerProfession.FARMER) {
             return false;
@@ -82,7 +_,7 @@
         BlockState blockstate = serverLevel.getBlockState(pos);
         Block block = blockstate.getBlock();
         Block block1 = serverLevel.getBlockState(pos.below()).getBlock();
-        return block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockstate) || blockstate.isAir() && block1 instanceof FarmBlock;
+        return block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockstate) || blockstate.isAir() && (block1 instanceof FarmBlock || block1.builtInRegistryHolder().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS));
     }
 
     protected void start(ServerLevel level, Villager entity, long gameTime) {
@@ -109,7 +_,7 @@
                     level.destroyBlock(this.aboveFarmlandPos, true, owner);
                 }
 
-                if (blockstate.isAir() && block1 instanceof FarmBlock && owner.hasFarmSeeds()) {
+                if (blockstate.isAir() && (block1 instanceof FarmBlock || block1.builtInRegistryHolder().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS)) && owner.hasFarmSeeds()) {
                     SimpleContainer simplecontainer = owner.getInventory();
 
                     for (int i = 0; i < simplecontainer.getContainerSize(); i++) {
@@ -120,6 +_,11 @@
                             level.setBlockAndUpdate(this.aboveFarmlandPos, blockstate1);
                             level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(owner, blockstate1));
                             flag = true;
+                        } else if (itemstack.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(owner)) {
+                            if (specialPlantable.canPlacePlantAtPosition(itemstack, level, aboveFarmlandPos, net.minecraft.core.Direction.DOWN)) {
+                                specialPlantable.spawnPlantAtPosition(itemstack, level, aboveFarmlandPos, net.minecraft.core.Direction.DOWN);
+                                flag = true;
+                            }
                         }
 
                         if (flag) {
