--- a/net/minecraft/world/entity/ai/behavior/StartAttacking.java
+++ b/net/minecraft/world/entity/ai/behavior/StartAttacking.java
@@ -28,7 +_,11 @@
                             if (!p_375685_.canAttack(livingentity)) {
                                 return false;
                             } else {
-                                p_258778_.set(livingentity);
+                                net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.neoforged.neoforge.common.CommonHooks.onLivingChangeTarget(p_375685_, livingentity, net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.BEHAVIOR_TARGET);
+                                if (changeTargetEvent.isCanceled() || changeTargetEvent.getNewAboutToBeSetTarget() == null)
+                                    return false;
+
+                                p_258778_.set(changeTargetEvent.getNewAboutToBeSetTarget());
                                 p_258779_.erase();
                                 return true;
                             }
