--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -108,6 +_,8 @@
             child.setAge(-24000);
             child.snapTo(source.getX(), source.getY(), source.getZ(), 0.0F, 0.0F);
             level.addFreshEntityWithPassengers(child);
+            // Neo: If villager is blocked from spawning (e.g., FinalizeSpawnEvent), then breed should be unsuccessful
+            if (!child.isAddedToLevel()) return Optional.empty();
             level.broadcastEntityEvent(child, (byte)12);
             return Optional.of(child);
         }
