--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -13,6 +13,8 @@
 import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
 
+import org.bukkit.craftbukkit.inventory.view.CraftMerchantView; // CraftBukkit
+
 public class MerchantMenu extends AbstractContainerMenu {
 
     protected static final int PAYMENT1_SLOT = 0;
@@ -32,6 +34,19 @@
     private boolean showProgressBar;
     private boolean canRestock;
 
+    // CraftBukkit start
+    private CraftMerchantView bukkitEntity = null;
+    private Inventory player;
+
+    @Override
+    public CraftMerchantView getBukkitView() {
+        if (bukkitEntity == null) {
+            bukkitEntity = new CraftMerchantView(this.player.player.getBukkitEntity(), new org.bukkit.craftbukkit.inventory.CraftInventoryMerchant(trader, tradeContainer), this, trader);
+        }
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+
     public MerchantMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, new ClientSideMerchant(inventory.player));
     }
@@ -43,6 +58,7 @@
         this.addSlot(new Slot(this.tradeContainer, 0, 136, 37));
         this.addSlot(new Slot(this.tradeContainer, 1, 162, 37));
         this.addSlot(new MerchantResultSlot(inventory.player, merchant, this.tradeContainer, 2, 220, 37));
+        this.player = inventory; // CraftBukkit - save player
         this.addStandardInventorySlots(inventory, 108, 84);
     }
 
@@ -62,6 +78,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!checkReachable) return true; // CraftBukkit - checkReachable
         return this.trader.stillValid(player);
     }
 
@@ -143,7 +160,7 @@
     }
 
     private void playTradeSound() {
-        if (!this.trader.isClientSide()) {
+        if (!this.trader.isClientSide() && this.trader instanceof Entity) { // CraftBukkit - SPIGOT-5035
             Entity entity = (Entity) this.trader;
 
             entity.level().playLocalSound(entity.getX(), entity.getY(), entity.getZ(), this.trader.getNotifyTradeSound(), SoundSource.NEUTRAL, 1.0F, 1.0F, false);
