--- a/net/minecraft/world/inventory/SmithingMenu.java
+++ b/net/minecraft/world/inventory/SmithingMenu.java
@@ -18,6 +18,8 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 
+import org.bukkit.craftbukkit.inventory.CraftInventoryView; // CraftBukkit
+
 public class SmithingMenu extends ItemCombinerMenu {
 
     public static final int TEMPLATE_SLOT = 0;
@@ -34,6 +36,9 @@
     private final RecipePropertySet templateItemTest;
     private final RecipePropertySet additionItemTest;
     private final DataSlot hasRecipeError;
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity;
+    // CraftBukkit end
 
     public SmithingMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, ContainerLevelAccess.NULL);
@@ -129,7 +134,9 @@
             ItemStack itemstack = ((SmithingRecipe) recipeholder.value()).assemble(smithingrecipeinput);
 
             this.resultSlots.setRecipeUsed(recipeholder);
-            this.resultSlots.setItem(0, itemstack);
+            // CraftBukkit start
+            org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), itemstack);
+            // CraftBukkit end
         }, () -> {
             this.resultSlots.setRecipeUsed((RecipeHolder) null);
             this.resultSlots.setItem(0, ItemStack.EMPTY);
@@ -149,4 +156,18 @@
     public boolean hasRecipeError() {
         return this.hasRecipeError.get() > 0;
     }
+
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventorySmithing(
+                access.getLocation(), this.inputSlots, this.resultSlots);
+        bukkitEntity = new CraftInventoryView(this.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 }
