--- a/net/minecraft/world/entity/ai/goal/RunAroundLikeCrazyGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RunAroundLikeCrazyGoal.java
@@ -58,7 +_,7 @@
             if (passenger instanceof Player player) {
                 int temper = this.horse.getTemper();
                 int maxTemper = this.horse.getMaxTemper();
-                if (maxTemper > 0 && this.horse.getRandom().nextInt(maxTemper) < temper) {
+                if (maxTemper > 0 && this.horse.getRandom().nextInt(maxTemper) < temper && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(horse, player)) {
                     this.horse.tameWithName(player);
                     return;
                 }
