--- a/net/minecraft/util/thread/BlockableEventLoop.java
+++ b/net/minecraft/util/thread/BlockableEventLoop.java
@@ -64,10 +_,15 @@
     }
 
     public CompletableFuture<Void> submitAsync(Runnable runnable) {
-        return CompletableFuture.supplyAsync(() -> {
+        return CompletableFuture.<Void>supplyAsync(() -> {
             runnable.run();
             return null;
-        }, this);
+        }, this)
+        .exceptionallyCompose(ex -> {
+            // Neo: Log since this is usually swallowed
+            LOGGER.error(LogUtils.FATAL_MARKER, "Error executing task on {}", name(), ex);
+            return CompletableFuture.failedStage(ex);
+        });
     }
 
     @CheckReturnValue
