--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -26,8 +_,10 @@
         this.itemsById = itemsById;
     }
 
+    private static final StackWalker STACK_WALKER = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
     public static <T> EntityDataAccessor<T> defineId(Class<? extends SyncedDataHolder> clazz, EntityDataSerializer<T> type) {
-        if (LOGGER.isDebugEnabled()) {
+        net.neoforged.neoforge.common.CommonHooks.verifyEntityDataAccessorRegistration(STACK_WALKER.getCallerClass(), clazz);
+        if (false && LOGGER.isDebugEnabled()) { // Neo: disable this check as we are handling it in the above hook
             try {
                 Class<?> aClass = Class.forName(Thread.currentThread().getStackTrace()[2].getClassName());
                 if (!aClass.equals(clazz)) {
