++++ Target missing
--- a/net/minecraft/world/level/storage/DimensionDataStorage.java
+++ b/net/minecraft/world/level/storage/DimensionDataStorage.java
@@ -38,13 +38,21 @@
 
 public class DimensionDataStorage implements AutoCloseable {
     private static final Logger LOGGER = LogUtils.getLogger();
+    private final net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerLevel level;
     private final Map<SavedDataType<?>, Optional<SavedData>> cache = new HashMap<>();
     private final DataFixer fixerUpper;
     private final HolderLookup.Provider registries;
     private final Path dataFolder;
     private CompletableFuture<?> pendingWriteFuture = CompletableFuture.completedFuture(null);
 
+    // Neo: Use the version that has a level
+    @Deprecated
     public DimensionDataStorage(Path dataFolder, DataFixer fixerUpper, HolderLookup.Provider registries) {
+        this(null, dataFolder, fixerUpper, registries);
+    }
+
+    public DimensionDataStorage(net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerLevel level, Path dataFolder, DataFixer fixerUpper, HolderLookup.Provider registries) {
+        this.level = level;
         this.fixerUpper = fixerUpper;
         this.dataFolder = dataFolder;
         this.registries = registries;
@@ -59,7 +67,7 @@
         if (data != null) {
             return data;
         } else {
-            T newData = (T)type.constructor().get();
+            T newData = (T)type.factory().create(this.level);
             this.set(type, newData);
             return newData;
         }
@@ -81,7 +89,7 @@
             if (Files.exists(file)) {
                 CompoundTag tag = this.readTagFromDisk(type.id(), type.dataFixType(), SharedConstants.getCurrentVersion().dataVersion().version());
                 RegistryOps<Tag> ops = this.registries.createSerializationContext(NbtOps.INSTANCE);
-                return type.codec()
+                return type.codecFactory().create(this.level)
                     .parse(ops, tag.get("data"))
                     .resultOrPartial(error -> LOGGER.error("Failed to parse saved data for '{}': {}", type, error))
                     .orElse(null);
@@ -98,7 +106,7 @@
         data.setDirty();
     }
 
-    public CompoundTag readTagFromDisk(String id, DataFixTypes type, int newVersion) throws IOException {
+    public CompoundTag readTagFromDisk(String id, @Nullable DataFixTypes type, int newVersion) throws IOException {
         CompoundTag var8;
         try (
             InputStream in = Files.newInputStream(this.getDataFile(id));
@@ -113,9 +121,19 @@
                 }
             }
 
-            int version = NbtUtils.getDataVersion(tag, 1343);
-            var8 = type.update(this.fixerUpper, tag, version, newVersion);
+            if (type != null) {
+                int version = NbtUtils.getDataVersion(tag, 1343);
+                var8 = type.update(this.fixerUpper, tag, version, newVersion);
+            } else {
+                var8 = tag;
+            }
         }
+
+        // Neo: delete any temporary files so that we don't inflate disk space unnecessarily.
+        this.pendingWriteFuture = this.pendingWriteFuture.thenRunAsync(
+                () -> net.neoforged.neoforge.common.IOUtilities.tryCleanupTempFiles(this.dataFolder, id),
+                Util.ioPool()
+        );
 
         return var8;
     }
@@ -187,7 +205,7 @@
     }
 
     private <T extends SavedData> CompoundTag encodeUnchecked(SavedDataType<T> type, SavedData data, RegistryOps<Tag> ops) {
-        Codec<T> codec = type.codec();
+        Codec<T> codec = type.codecFactory().create(this.level);
         CompoundTag tag = new CompoundTag();
         tag.put("data", codec.encodeStart(ops, (T)data).getOrThrow());
         NbtUtils.addCurrentDataVersion(tag);
@@ -198,7 +216,11 @@
         Path path = this.getDataFile(type.id());
 
         try {
-            NbtIo.writeCompressed(tag, path);
+            // Neo: ensure parent directories exist if the SavedData's path contains slashes
+            if (!Files.exists(path)) {
+                Files.createDirectories(path.getParent());
+            }
+            net.neoforged.neoforge.common.IOUtilities.writeNbtCompressed(tag, path);
         } catch (IOException var5) {
             LOGGER.error("Could not save data to {}", path.getFileName(), var5);
         }
