--- a/com/mojang/renderpearl/frontend/FrontendGpuDevice.java
+++ b/com/mojang/renderpearl/frontend/FrontendGpuDevice.java
@@ -36,7 +_,7 @@
 @OnlyIn(Dist.CLIENT)
 public class FrontendGpuDevice implements GpuDevice {
     private static final Logger LOGGER = LogUtils.getLogger();
-    public static final boolean STRICT_VALIDATION = SharedConstants.IS_RUNNING_IN_IDE;
+    public static final boolean STRICT_VALIDATION = SharedConstants.IS_RUNNING_IN_IDE && !Boolean.getBoolean("neoforge.disableBlaze3DValidation");
     private final GpuDeviceBackend backend;
     private final @Nullable TracyGpuProfiler profiler;
     private final CommandEncoder encoder;
