--- a/net/minecraft/world/level/border/WorldBorder.java
+++ b/net/minecraft/world/level/border/WorldBorder.java
@@ -218,6 +218,7 @@
     }
 
     public void addListener(BorderChangeListener listener) {
+        if (listeners.contains(listener)) return; // CraftBukkit
         this.listeners.add(listener);
     }
 
@@ -299,7 +300,13 @@
     }
 
     public void applyInitialSettings(long gameTime) {
-        if (!this.initialized) {
+        // CraftBukkit start
+        this.applyInitialSettings(gameTime, false);
+    }
+
+    public void applyInitialSettings(long gameTime, boolean force) {
+        if (!this.initialized || force) {
+            // CraftBukkit end
             this.setCenter(this.settings.centerX(), this.settings.centerZ());
             this.setDamagePerBlock(this.settings.damagePerBlock());
             this.setSafeZone(this.settings.safeZone());
