--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -54,6 +_,7 @@
         ),
         Advancement::validate
     );
+    public static final Codec<Optional<net.neoforged.neoforge.common.conditions.WithConditions<Advancement>>> CONDITIONAL_ADVANCEMENT_CODEC = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodecWithConditions(CODEC);
 
     public Advancement(
         Optional<ResourceLocation> p_300893_,
@@ -113,7 +_,7 @@
         });
     }
 
-    public static class Builder {
+    public static class Builder implements net.neoforged.neoforge.common.extensions.IAdvancementBuilderExtension {
         private Optional<ResourceLocation> parent = Optional.empty();
         private Optional<DisplayInfo> display = Optional.empty();
         private AdvancementRewards rewards = AdvancementRewards.EMPTY;
