--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -65,6 +_,7 @@
             parent, display, AdvancementRewards.EMPTY, Map.of(), requirements, sendsTelemetryEvent
         )
     );
+    public static final Codec<Optional<net.neoforged.neoforge.common.conditions.WithConditions<Advancement>>> CONDITIONAL_CODEC = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodecWithConditions(CODEC);
 
     public Advancement(
         Optional<Identifier> parent,
@@ -225,8 +_,12 @@
         }
 
         public AdvancementHolder save(BootstrapContext<Advancement> output, String name) {
+            return this.save(output, name, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+        }
+
+        public AdvancementHolder save(BootstrapContext<Advancement> output, String name, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
             AdvancementHolder advancement = this.build(Identifier.parse(name));
-            advancement.register(output);
+            advancement.register(output, conditions);
             return advancement;
         }
     }
