--- a/net/minecraft/advancements/AdvancementRewards.java
+++ b/net/minecraft/advancements/AdvancementRewards.java
@@ -46,6 +_,7 @@
         LootParams params = new LootParams.Builder(level)
             .withParameter(LootContextParams.THIS_ENTITY, player)
             .withParameter(LootContextParams.ORIGIN, player.position())
+            .withLuck(player.getLuck()) // Neo: Add luck to LootParams
             .create(LootContextParamSets.ADVANCEMENT_REWARD);
         boolean changes = false;
 
