--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -52,6 +_,7 @@
         )
         .validate(Advancement::validate);
     public static final StreamCodec<RegistryFriendlyByteBuf, Advancement> STREAM_CODEC = StreamCodec.ofMember(Advancement::write, Advancement::read);
+    public static final Codec<Optional<net.neoforged.neoforge.common.conditions.WithConditions<Advancement>>> CONDITIONAL_CODEC = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodecWithConditions(CODEC);
 
     public Advancement(
         Optional<ResourceLocation> p_300893_,
@@ -111,7 +_,7 @@
         });
     }
 
-    public static class Builder {
+    public static class Builder implements net.neoforged.neoforge.common.extensions.IAdvancementBuilderExtension {
         private Optional<ResourceLocation> parent = Optional.empty();
         private Optional<DisplayInfo> display = Optional.empty();
         private AdvancementRewards rewards = AdvancementRewards.EMPTY;
