--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -662,6 +662,8 @@
 
                 if (chunkBB.isInside(blockpos1)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     level.setBlock(blockpos1, Blocks.SPAWNER.defaultBlockState(), 2);
                     BlockEntity blockentity = level.getBlockEntity(blockpos1);
 
@@ -670,6 +672,9 @@
 
                         spawnerblockentity.setEntityId(EntityTypes.BLAZE, random);
                     }
+                    */
+                    placeCraftSpawner(level, blockpos1, org.bukkit.entity.EntityType.BLAZE, 2);
+                    // CraftBukkit end
                 }
             }
 
