--- a/net/minecraft/client/renderer/fog/FogRenderer.java
+++ b/net/minecraft/client/renderer/fog/FogRenderer.java
@@ -157,7 +_,7 @@
             fogBlue = Mth.lerp(brightenFactor, fogBlue, fogBlue * scale);
         }
 
-        dest.set(fogRed, fogGreen, fogBlue, 1.0F);
+        net.neoforged.neoforge.client.ClientHooks.getFogColor(camera, partialTicks, level, renderDistance, darkenWorldAmount, fogRed, fogGreen, fogBlue, dest);
     }
 
     public static boolean toggleFog() {
@@ -172,9 +_,11 @@
         FogData fog = new FogData();
         this.computeFogColor(camera, partialTickTime, level, renderDistanceInChunks, darkenWorldAmount, fog.color);
 
+        FogEnvironment chosenEnvironment = null;
         for (FogEnvironment fogEnvironment : FOG_ENVIRONMENTS) {
             if (fogEnvironment.isApplicable(fogType, entity)) {
                 fogEnvironment.setupFog(fog, camera, level, renderDistanceInBlocks, deltaTracker);
+                chosenEnvironment = fogEnvironment;
                 break;
             }
         }
@@ -182,6 +_,7 @@
         float renderDistanceFogSpan = Mth.clamp(renderDistanceInBlocks / 10.0F, 4.0F, 64.0F);
         fog.renderDistanceStart = renderDistanceInBlocks - renderDistanceFogSpan;
         fog.renderDistanceEnd = renderDistanceInBlocks;
+        net.neoforged.neoforge.client.ClientHooks.onSetupFog(chosenEnvironment, fogType, camera, partialTickTime, renderDistanceInChunks, fog);
         return fog;
     }
 
