--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -23,6 +23,8 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
+
 public abstract class BasePressurePlateBlock extends Block {
 
     private static final VoxelShape SHAPE_PRESSED = Block.column(14.0D, 0.0D, 0.5D);
@@ -91,6 +93,19 @@
         boolean flag = oldSignal > 0;
         boolean flag1 = j > 0;
 
+        // CraftBukkit start - Interact Pressure Plate
+        org.bukkit.World bworld = level.getWorld();
+        org.bukkit.plugin.PluginManager manager = level.getCraftServer().getPluginManager();
+
+        if (flag != flag1) {
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()), oldSignal, j);
+            manager.callEvent(eventRedstone);
+
+            flag1 = eventRedstone.getNewCurrent() > 0;
+            j = eventRedstone.getNewCurrent();
+        }
+        // CraftBukkit end
+
         if (oldSignal != j) {
             BlockState blockstate1 = this.setSignalForState(state, j);
 
@@ -142,9 +157,15 @@
     }
 
     protected static int getEntityCount(Level level, AABB entityDetectionBox, Class<? extends Entity> entityClass) {
+        // CraftBukkit start
+        return getEntities(level, entityDetectionBox, entityClass).size();
+    }
+
+    protected static <T extends Entity> java.util.List<T> getEntities(Level level, AABB entityDetectionBox, Class<T> entityClass) {
+        // CraftBukkit end
         return level.getEntitiesOfClass(entityClass, entityDetectionBox, EntitySelector.NO_SPECTATORS.and((entity) -> {
             return !entity.isIgnoringBlockTriggers();
-        })).size();
+        })); // CraftBukkit
     }
 
     protected abstract int getSignalStrength(Level level, BlockPos pos);
