--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -23,6 +23,8 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class ChorusFlowerBlock extends Block {
 
     public static final MapCodec<ChorusFlowerBlock> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
@@ -102,8 +104,12 @@
                 }
 
                 if (flag && allNeighborsEmpty(level, blockpos1, (Direction) null) && level.isEmptyBlock(pos.above(2))) {
-                    level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
-                    this.placeGrownFlower(level, blockpos1, i);
+                    // CraftBukkit start - add event
+                    if (CraftEventFactory.handleBlockSpreadEvent(level, pos, blockpos1, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(i)), 2)) {
+                        level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
+                        this.placeGrownFlower(level, blockpos1, i);
+                    }
+                    // CraftBukkit end
                 } else if (i < 4) {
                     int l = random.nextInt(4);
 
@@ -118,18 +124,30 @@
                         BlockPos blockpos2 = pos.relative(direction);
 
                         if (level.isEmptyBlock(blockpos2) && level.isEmptyBlock(blockpos2.below()) && allNeighborsEmpty(level, blockpos2, direction.getOpposite())) {
-                            this.placeGrownFlower(level, blockpos2, i + 1);
-                            flag2 = true;
+                            // CraftBukkit start - add event
+                            if (CraftEventFactory.handleBlockSpreadEvent(level, pos, blockpos2, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(i + 1)), 2)) {
+                                this.placeGrownFlower(level, blockpos2, i + 1);
+                                flag2 = true;
+                            }
+                            // CraftBukkit end
                         }
                     }
 
                     if (flag2) {
                         level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
                     } else {
-                        this.placeDeadFlower(level, pos);
+                        // CraftBukkit start - add event
+                        if (CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
+                            this.placeDeadFlower(level, pos);
+                        }
+                        // CraftBukkit end
                     }
                 } else {
-                    this.placeDeadFlower(level, pos);
+                    // CraftBukkit start - add event
+                    if (CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
+                        this.placeDeadFlower(level, pos);
+                    }
+                    // CraftBukkit end
                 }
 
             }
@@ -259,6 +277,11 @@
 
         if (level instanceof ServerLevel serverlevel) {
             if (projectile.mayInteract(serverlevel, blockpos) && projectile.mayBreak(serverlevel)) {
+                // CraftBukkit
+                if (!CraftEventFactory.callEntityChangeBlockEvent(projectile, blockpos, Blocks.AIR.defaultBlockState())) {
+                    return;
+                }
+                // CraftBukkit end
                 level.destroyBlock(blockpos, true, projectile);
             }
         }
