--- a/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -27,6 +27,8 @@
 import net.minecraft.world.level.redstone.Orientation;
 import net.minecraft.world.phys.AABB;
 
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
+
 public class DetectorRailBlock extends BaseRailBlock {
 
     public static final MapCodec<DetectorRailBlock> CODEC = simpleCodec(DetectorRailBlock::new);
@@ -87,6 +89,17 @@
                 flag1 = true;
             }
 
+            // CraftBukkit start
+            if (flag != flag1) {
+                org.bukkit.block.Block block = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+
+                BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, flag ? 15 : 0, flag1 ? 15 : 0);
+                level.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+                flag1 = eventRedstone.getNewCurrent() > 0;
+            }
+            // CraftBukkit end
+
             if (flag1 && !flag) {
                 BlockState blockstate1 = (BlockState) state.setValue(DetectorRailBlock.POWERED, true);
 
