--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -45,7 +45,7 @@
 public class LiquidBlock extends Block implements BucketPickup {
 
     private static final Codec<FlowingFluid> FLOWING_FLUID = BuiltInRegistries.FLUID.byNameCodec().comapFlatMap((fluid) -> {
-        DataResult dataresult;
+        DataResult<FlowingFluid> dataresult; // CraftBukkit - decompile error
 
         if (fluid instanceof FlowingFluid flowingfluid) {
             dataresult = DataResult.success(flowingfluid);
@@ -229,14 +229,20 @@
                 if (level.getFluidState(blockpos1).is(FluidTags.WATER)) {
                     Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
 
-                    level.setBlockAndUpdate(pos, block.defaultBlockState());
-                    this.fizz(level, pos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState())) {
+                        this.fizz(level, pos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
 
                 if (flag && level.getBlockState(blockpos1).is(Blocks.BLUE_ICE)) {
-                    level.setBlockAndUpdate(pos, Blocks.BASALT.defaultBlockState());
-                    this.fizz(level, pos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, Blocks.BASALT.defaultBlockState())) {
+                        this.fizz(level, pos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
             }
