--- a/net/minecraft/world/level/block/NetherFungusBlock.java
+++ b/net/minecraft/world/level/block/NetherFungusBlock.java
@@ -78,6 +78,13 @@
     @Override
     public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
         this.getFeature(level).ifPresent((holder) -> {
+            // CraftBukkit start
+            if (this == Blocks.WARPED_FUNGUS) {
+                SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+            } else if (this == Blocks.CRIMSON_FUNGUS) {
+                SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+            }
+            // CraftBukkit end
             ((ConfiguredFeature) holder.value()).place(level, level.getChunkSource().getGenerator(), random, pos);
         });
     }
