--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -23,6 +23,8 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class VineBlock extends Block {
 
     public static final MapCodec<VineBlock> CODEC = simpleCodec(VineBlock::new);
@@ -178,30 +180,34 @@
                             BlockPos blockpos3 = blockpos2.relative(direction1);
                             BlockPos blockpos4 = blockpos2.relative(direction2);
 
+                            // CraftBukkit start - Call BlockSpreadEvent
+                            BlockPos source = pos;
+
                             if (flag && isAcceptableNeighbour(level, blockpos3, direction1)) {
-                                level.setBlock(blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction1), true), 2);
+                                CraftEventFactory.handleBlockSpreadEvent(level, source, blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction1), true), 2);
                             } else if (flag1 && isAcceptableNeighbour(level, blockpos4, direction2)) {
-                                level.setBlock(blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction2), true), 2);
+                                CraftEventFactory.handleBlockSpreadEvent(level, source, blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction2), true), 2);
                             } else {
                                 Direction direction3 = direction.getOpposite();
 
                                 if (flag && level.isEmptyBlock(blockpos3) && isAcceptableNeighbour(level, pos.relative(direction1), direction3)) {
-                                    level.setBlock(blockpos3, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
+                                    CraftEventFactory.handleBlockSpreadEvent(level, source, blockpos3, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
                                 } else if (flag1 && level.isEmptyBlock(blockpos4) && isAcceptableNeighbour(level, pos.relative(direction2), direction3)) {
-                                    level.setBlock(blockpos4, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
+                                    CraftEventFactory.handleBlockSpreadEvent(level, source, blockpos4, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
                                 } else if ((double) random.nextFloat() < 0.05D && isAcceptableNeighbour(level, blockpos2.above(), Direction.UP)) {
-                                    level.setBlock(blockpos2, (BlockState) this.defaultBlockState().setValue(VineBlock.UP, true), 2);
+                                    CraftEventFactory.handleBlockSpreadEvent(level, source, blockpos2, (BlockState) this.defaultBlockState().setValue(VineBlock.UP, true), 2);
                                 }
+                                // CraftBukkit end
                             }
                         } else if (isAcceptableNeighbour(level, blockpos2, direction)) {
-                            level.setBlock(pos, (BlockState) state.setValue(getPropertyForFace(direction), true), 2);
+                            CraftEventFactory.handleBlockGrowEvent(level, pos, (BlockState) state.setValue(getPropertyForFace(direction), true), 2); // CraftBukkit
                         }
 
                     }
                 } else {
                     if (direction == Direction.UP && pos.getY() < level.getMaxY()) {
                         if (this.canSupportAtFace(level, pos, direction)) {
-                            level.setBlock(pos, (BlockState) state.setValue(VineBlock.UP, true), 2);
+                            CraftEventFactory.handleBlockGrowEvent(level, pos, (BlockState) state.setValue(VineBlock.UP, true), 2); // CraftBukkit
                             return;
                         }
 
@@ -219,7 +225,7 @@
                             }
 
                             if (this.hasHorizontalConnection(blockstate2)) {
-                                level.setBlock(blockpos1, blockstate2, 2);
+                                CraftEventFactory.handleBlockSpreadEvent(level, pos, blockpos1, blockstate2, 2); // CraftBukkit
                             }
 
                             return;
@@ -235,7 +241,7 @@
                             BlockState blockstate5 = this.copyRandomFaces(state, blockstate4, random);
 
                             if (blockstate4 != blockstate5 && this.hasHorizontalConnection(blockstate5)) {
-                                level.setBlock(blockpos5, blockstate5, 2);
+                                CraftEventFactory.handleBlockSpreadEvent(level, pos, blockpos5, blockstate5, 2); // CraftBukkit
                             }
                         }
                     }
