--- a/net/minecraft/world/level/block/WitherRoseBlock.java
+++ b/net/minecraft/world/level/block/WitherRoseBlock.java
@@ -68,7 +68,7 @@
         if (level instanceof ServerLevel serverlevel) {
             if (level.getDifficulty() != Difficulty.PEACEFUL && entity instanceof LivingEntity livingentity) {
                 if (!livingentity.isInvulnerableTo(serverlevel, level.damageSources().wither())) {
-                    livingentity.addEffect(this.getBeeInteractionEffect());
+                    livingentity.addEffect(this.getBeeInteractionEffect(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WITHER_ROSE); // CraftBukkit
                 }
             }
         }
