--- a/net/minecraft/world/item/component/BundleContents.java
+++ b/net/minecraft/world/item/component/BundleContents.java
@@ -77,7 +_,8 @@
     }
 
     public static boolean canItemBeInBundle(ItemStack itemToAdd) {
-        return !itemToAdd.isEmpty() && itemToAdd.getItem().canFitInsideContainerItems();
+        // Neo: stack-aware placeability check
+        return !itemToAdd.isEmpty() && itemToAdd.canFitInsideContainerItems();
     }
 
     public int getNumberOfItemsToShow() {
