--- a/net/minecraft/world/item/component/FireworkExplosion.java
+++ b/net/minecraft/world/item/component/FireworkExplosion.java
@@ -104,6 +_,9 @@
         return new FireworkExplosion(this.shape, this.colors, new IntArrayList(p_330678_), this.hasTrail, this.hasTwinkle);
     }
 
+    @net.neoforged.fml.common.asm.enumextension.IndexedEnum
+    @net.neoforged.fml.common.asm.enumextension.NamedEnum(1)
+    @net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.BIDIRECTIONAL)
     public static enum Shape implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
         SMALL_BALL(0, "small_ball"),
         LARGE_BALL(1, "large_ball"),
@@ -139,6 +_,10 @@
         @Override
         public String getSerializedName() {
             return this.name;
+        }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(FireworkExplosion.Shape.class);
         }
     }
 }
