--- a/net/minecraft/world/item/component/ItemAttributeModifiers.java
+++ b/net/minecraft/world/item/component/ItemAttributeModifiers.java
@@ -38,7 +_,7 @@
     public static final Codec<ItemAttributeModifiers> CODEC = ItemAttributeModifiers.Entry.CODEC
         .listOf()
         .xmap(ItemAttributeModifiers::new, ItemAttributeModifiers::modifiers);
-    public static final StreamCodec<RegistryFriendlyByteBuf, ItemAttributeModifiers> STREAM_CODEC = StreamCodec.composite(
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemAttributeModifiers> STREAM_CODEC = net.neoforged.neoforge.common.CommonHooks.makeItemAttributesStreamCodec(
         ItemAttributeModifiers.Entry.STREAM_CODEC.apply(ByteBufCodecs.list()), ItemAttributeModifiers::modifiers, ItemAttributeModifiers::new
     );
     public static final DecimalFormat ATTRIBUTE_MODIFIER_FORMAT = new DecimalFormat("#.##", DecimalFormatSymbols.getInstance(Locale.ROOT));
