--- a/net/minecraft/client/renderer/feature/BlockFeatureRenderer.java
+++ b/net/minecraft/client/renderer/feature/BlockFeatureRenderer.java
@@ -69,7 +_,7 @@
             MovingBlockRenderState movingBlockRenderState = submit.movingBlockRenderState();
             BlockState blockState = movingBlockRenderState.blockState;
             BlockStateModel model = blockRenderDispatcher.getBlockModel(blockState);
-            if (model.hasTranslucency() == translucent) {
+            if (model.hasTranslucency(movingBlockRenderState, movingBlockRenderState.blockPos, blockState) == translucent) {
                 poseStack.setIdentity();
                 poseStack.mulPose(submit.pose());
                 BlockQuadOutput blockOutput = ModelBlockRenderer.forceOpaque(cutoutLeaves, blockState) ? solidOutput : output;
