--- a/net/minecraft/client/renderer/feature/BlockFeatureRenderer.java
+++ b/net/minecraft/client/renderer/feature/BlockFeatureRenderer.java
@@ -65,7 +_,7 @@
             MovingBlockRenderState movingBlockRenderState = submit.movingBlockRenderState();
             BlockState blockState = movingBlockRenderState.blockState;
             BlockStateModel model = blockRenderDispatcher.getBlockModel(blockState);
-            if (model.hasTranslucency() == translucent) {
+            if (model.hasTranslucency(movingBlockRenderState, movingBlockRenderState.blockPos, blockState) == translucent) {
                 poseStack.setIdentity();
                 poseStack.mulPose(submit.pose());
                 BlockQuadOutput blockOutput = ModelBlockRenderer.forceOpaque(cutoutLeaves, blockState) ? solidOutput : output;
@@ -118,7 +_,7 @@
                 quadInstance.setOverlayCoords(submit.overlayCoords());
                 this.random.setSeed(42L);
 
-                for (BlockModelPart part : submit.model().collectParts(this.random)) {
+                for (BlockModelPart part : submit.model().collectParts(net.minecraft.client.renderer.block.BlockAndTintGetter.EMPTY, net.minecraft.core.BlockPos.ZERO, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), this.random)) {
                     putPartQuads(part, submit.pose(), quadInstance, submit.tintLayers(), buffer, outlineBuffer);
                 }
             }
