--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -209,7 +_,7 @@
             result.add(
                 jigsaw.withInfo(
                     new StructureTemplate.StructureBlockInfo(
-                        calculateRelativePosition(settings, blockInfo.pos()).offset(position), blockInfo.state.rotate(settings.getRotation()), blockInfo.nbt
+                        calculateRelativePosition(settings, blockInfo.pos()).offset(position), blockInfo.state.rotate(settings.getRotation()), net.neoforged.neoforge.common.CommonHooks.captureStructureRotation(blockInfo.nbt, rotation)
                     )
                 )
             );
@@ -228,7 +_,7 @@
         for (StructureTemplate.StructureBlockInfo blockInfo : settings.getRandomPalette(this.palettes, position).blocks(block)) {
             BlockPos blockPos = absolute ? calculateRelativePosition(settings, blockInfo.pos).offset(position) : blockInfo.pos;
             if (boundingBox == null || boundingBox.isInside(blockPos)) {
-                result.add(new StructureTemplate.StructureBlockInfo(blockPos, blockInfo.state.rotate(settings.getRotation()), blockInfo.nbt));
+                result.add(new StructureTemplate.StructureBlockInfo(blockPos, blockInfo.state.rotate(settings.getRotation()), net.neoforged.neoforge.common.CommonHooks.captureStructureRotation(blockInfo.nbt, settings.getRotation())));
             }
         }
 
@@ -245,6 +_,10 @@
         return transform(pos, settings.getMirror(), settings.getRotation(), settings.getRotationPivot());
     }
 
+    public static Vec3 transformedVec3d(StructurePlaceSettings placementIn, Vec3 pos) {
+        return transform(pos, placementIn.getMirror(), placementIn.getRotation(), placementIn.getRotationPivot());
+    }
+
     public boolean placeInWorld(
         ServerLevelAccessor level,
         BlockPos position,
@@ -272,7 +_,7 @@
             int maxX = Integer.MIN_VALUE;
             int maxY = Integer.MIN_VALUE;
             int maxZ = Integer.MIN_VALUE;
-            List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = processBlockInfos(level, position, referencePos, settings, blockInfoList);
+            List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = processBlockInfos(level, position, referencePos, settings, blockInfoList, this);
 
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(LOGGER)) {
                 for (StructureTemplate.StructureBlockInfo blockInfo : processedBlockInfoList) {
@@ -302,6 +_,7 @@
                                     blockEntity.loadWithComponents(
                                         TagValueInput.create(reporter.forChild(blockEntity.problemPath()), level.registryAccess(), blockInfo.nbt)
                                     );
+                                    net.neoforged.neoforge.common.CommonHooks.applyStructureRotation(blockEntity, blockInfo.nbt, settings);
                                 }
                             }
 
@@ -386,16 +_,7 @@
                 }
 
                 if (!settings.isIgnoreEntities()) {
-                    this.placeEntities(
-                        level,
-                        position,
-                        settings.getMirror(),
-                        settings.getRotation(),
-                        settings.getRotationPivot(),
-                        boundingBox,
-                        settings.shouldFinalizeEntities(),
-                        reporter
-                    );
+                    this.addEntitiesToWorld(level, position, settings, reporter);
                 }
             }
 
@@ -429,6 +_,10 @@
         });
     }
 
+    /**
+     * @deprecated Forge: Use {@link #processBlockInfos(ServerLevelAccessor, BlockPos, BlockPos, StructurePlaceSettings, List, StructureTemplate)} instead.
+     */
+    @Deprecated
     public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(
         ServerLevelAccessor level,
         BlockPos position,
@@ -436,6 +_,10 @@
         StructurePlaceSettings settings,
         List<StructureTemplate.StructureBlockInfo> blockInfoList
     ) {
+        return processBlockInfos(level, position, referencePos, settings, blockInfoList, null);
+    }
+
+    public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(ServerLevelAccessor level, BlockPos position, BlockPos referencePos, StructurePlaceSettings settings, List<StructureTemplate.StructureBlockInfo> blockInfoList, @Nullable StructureTemplate template) {
         List<StructureTemplate.StructureBlockInfo> originalBlockInfoList = new ArrayList<>();
         List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = new ArrayList<>();
         boolean processOnlyInCurrentChunk = true;
@@ -458,7 +_,7 @@
                 Iterator<StructureProcessor> iterator = settings.getProcessors().iterator();
 
                 while (processedBlockInfo != null && iterator.hasNext()) {
-                    processedBlockInfo = iterator.next().processBlock(level, position, referencePos, blockInfo.pos, processedBlockInfo, settings);
+                    processedBlockInfo = iterator.next().process(level, position, referencePos, blockInfo, processedBlockInfo, settings, template);
                 }
 
                 if (processedBlockInfo != null) {
@@ -475,22 +_,30 @@
         return processedBlockInfoList;
     }
 
-    private void placeEntities(
-        ServerLevelAccessor level,
-        BlockPos position,
-        Mirror mirror,
-        Rotation rotation,
-        BlockPos pivot,
-        @Nullable BoundingBox boundingBox,
-        boolean finalizeEntities,
-        ProblemReporter problemReporter
-    ) {
-        for (StructureTemplate.StructureEntityInfo entityInfo : this.entityInfoList) {
-            BlockPos blockPos = transform(entityInfo.blockPos, mirror, rotation, pivot).offset(position);
-            if (boundingBox == null || boundingBox.isInside(blockPos)) {
+
+    public static List<StructureTemplate.StructureEntityInfo> processEntityInfos(@Nullable StructureTemplate template, LevelAccessor p_215387_0_, BlockPos p_215387_1_, StructurePlaceSettings p_215387_2_, List<StructureTemplate.StructureEntityInfo> p_215387_3_) {
+        List<StructureTemplate.StructureEntityInfo> list = Lists.newArrayList();
+        for(StructureTemplate.StructureEntityInfo entityInfo : p_215387_3_) {
+            Vec3 pos = transformedVec3d(p_215387_2_, entityInfo.pos).add(Vec3.atLowerCornerOf(p_215387_1_));
+            BlockPos blockpos = calculateRelativePosition(p_215387_2_, entityInfo.blockPos).offset(p_215387_1_);
+            StructureTemplate.StructureEntityInfo info = new StructureTemplate.StructureEntityInfo(pos, blockpos, entityInfo.nbt);
+            for (StructureProcessor proc : p_215387_2_.getProcessors()) {
+                info = proc.processEntity(p_215387_0_, p_215387_1_, entityInfo, info, p_215387_2_, template);
+                if (info == null)
+                    break;
+            }
+            if (info != null)
+                list.add(info);
+        }
+        return list;
+    }
+
+    private void addEntitiesToWorld(ServerLevelAccessor level, BlockPos position, StructurePlaceSettings placementIn, ProblemReporter problemReporter) {
+        for(StructureTemplate.StructureEntityInfo entityInfo : processEntityInfos(this, level, position, placementIn, this.entityInfoList)) {
+            BlockPos blockpos = entityInfo.blockPos; // FORGE: Position will have already been transformed by processEntityInfos
+            if (placementIn.getBoundingBox() == null || placementIn.getBoundingBox().isInside(blockpos)) {
                 CompoundTag tag = entityInfo.nbt.copy();
-                Vec3 relativePos = transform(entityInfo.pos, mirror, rotation, pivot);
-                Vec3 pos = relativePos.add(position.getX(), position.getY(), position.getZ());
+                Vec3 pos = entityInfo.pos; // FORGE: Position will have already been transformed by processEntityInfos
                 ListTag posTag = new ListTag();
                 posTag.add(DoubleTag.valueOf(pos.x));
                 posTag.add(DoubleTag.valueOf(pos.y));
@@ -498,12 +_,12 @@
                 tag.put("Pos", posTag);
                 tag.remove("UUID");
                 createEntityIgnoreException(problemReporter, level, tag).ifPresent(entity -> {
-                    float yRot = entity.rotate(rotation);
-                    yRot += entity.mirror(mirror) - entity.getYRot();
+                    float yRot = entity.rotate(placementIn.getRotation());
+                    yRot += entity.mirror(placementIn.getMirror()) - entity.getYRot();
                     entity.snapTo(pos.x, pos.y, pos.z, yRot, entity.getXRot());
                     entity.setYBodyRot(yRot);
                     entity.setYHeadRot(yRot);
-                    if (finalizeEntities && entity instanceof Mob mob) {
+                    if (placementIn.shouldFinalizeEntities() && entity instanceof Mob mob) {
                         mob.finalizeSpawn(level, level.getCurrentDifficultyAt(BlockPos.containing(pos)), EntitySpawnReason.STRUCTURE, null);
                     }
 
@@ -817,7 +_,7 @@
 
     public static final class Palette {
         private final List<StructureTemplate.StructureBlockInfo> blocks;
-        private final Map<Block, List<StructureTemplate.StructureBlockInfo>> cache = Maps.newHashMap();
+        private final Map<Block, List<StructureTemplate.StructureBlockInfo>> cache = Maps.newConcurrentMap(); // Neo: Fixes MC-271899 - Make the global StructureTemplate's palette caches now thread safe for worldgen
         private @Nullable List<StructureTemplate.JigsawBlockInfo> cachedJigsaws;
 
         private Palette(List<StructureTemplate.StructureBlockInfo> blocks) {
