--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessor.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureProcessor.java
@@ -8,6 +_,10 @@
 import org.jspecify.annotations.Nullable;
 
 public interface StructureProcessor {
+    /**
+     * @deprecated Neo: Use {@link #process(LevelReader, BlockPos, BlockPos, StructureTemplate.StructureBlockInfo, StructureTemplate.StructureBlockInfo, StructurePlaceSettings, StructureTemplate)} instead.
+     */
+    @Deprecated
     default StructureTemplate.@Nullable StructureBlockInfo processBlock(
         LevelReader level,
         BlockPos targetPosition,
@@ -17,6 +_,29 @@
         StructurePlaceSettings settings
     ) {
         return processedBlockInfo;
+    }
+
+    default StructureTemplate.@Nullable StructureBlockInfo process(LevelReader level, BlockPos targetPosition, BlockPos referencePos, StructureTemplate.StructureBlockInfo originalBlockInfo, StructureTemplate.StructureBlockInfo processedBlockInfo, StructurePlaceSettings settings, @Nullable StructureTemplate template) {
+        return processBlock(level, targetPosition, referencePos, originalBlockInfo.pos(), processedBlockInfo, settings);
+    }
+
+    /**
+     * FORGE: Add entity processing.
+     * <p>
+     * Use this method to process entities from a structure in much the same way as
+     * blocks, parameters are analogous.
+     *
+     * @param world
+     * @param seedPos
+     * @param rawEntityInfo
+     * @param entityInfo
+     * @param placementSettings
+     * @param template
+     *
+     * @see #process(LevelReader, BlockPos, BlockPos, StructureTemplate.StructureBlockInfo, StructureTemplate.StructureBlockInfo, StructurePlaceSettings, StructureTemplate)
+     */
+    default StructureTemplate.StructureEntityInfo processEntity(LevelReader world, BlockPos seedPos, StructureTemplate.StructureEntityInfo rawEntityInfo, StructureTemplate.StructureEntityInfo entityInfo, StructurePlaceSettings placementSettings, StructureTemplate template) {
+        return entityInfo;
     }
 
     MapCodec<? extends StructureProcessor> codec();
