--- a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -72,7 +_,16 @@
         return this.addEntity(entity, false);
     }
 
+    public boolean addNewEntityWithoutEvent(T entity) {
+        return this.addEntityWithoutEvent(entity, false);
+    }
+
     private boolean addEntity(T entity, boolean loaded) {
+        if (entity instanceof Entity actualEntity && net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityJoinLevelEvent(actualEntity, actualEntity.level(), loaded)).isCanceled()) return false;
+        return addEntityWithoutEvent(entity, loaded);
+    }
+
+    private boolean addEntityWithoutEvent(T entity, boolean loaded) {
         if (!this.addEntityUuid(entity)) {
             return false;
         } else {
@@ -106,11 +_,17 @@
     }
 
     public void addLegacyChunkEntities(Stream<T> entities) {
-        entities.forEach(e -> this.addEntity((T)e, true));
+        entities.forEach(e -> {
+            this.addEntity(e, true);
+            if (e instanceof Entity entity) entity.onAddedToLevel();
+        });
     }
 
     public void addWorldGenChunkEntities(Stream<T> entities) {
-        entities.forEach(e -> this.addEntity((T)e, false));
+        entities.forEach(e -> {
+            this.addEntity(e, false);
+            if (e instanceof Entity entity) entity.onAddedToLevel();
+        });
     }
 
     private void startTicking(T entity) {
@@ -233,7 +_,10 @@
     public void processPendingLoads() {
         ChunkEntities<T> loadedChunk;
         while ((loadedChunk = this.loadingInbox.poll()) != null) {
-            loadedChunk.getEntities().forEach(e -> this.addEntity((T)e, true));
+            loadedChunk.getEntities().forEach(e -> {
+                this.addEntity(e, true);
+                if (e instanceof Entity entity) entity.onAddedToLevel();
+            });
             this.chunkLoadStatuses.put(loadedChunk.getPos().pack(), PersistentEntitySectionManager.ChunkLoadStatus.LOADED);
         }
     }
@@ -370,6 +_,7 @@
 
     private class Callback implements EntityInLevelCallback {
         private final T entity;
+        private final Entity realEntity;
         private long currentSectionKey;
         private EntitySection<T> currentSection;
 
@@ -377,6 +_,7 @@
             Objects.requireNonNull(PersistentEntitySectionManager.this);
             super();
             this.entity = entity;
+            this.realEntity = entity instanceof Entity ? (Entity) entity : null;
             this.currentSectionKey = currentSectionKey;
             this.currentSection = currentSection;
         }
@@ -395,9 +_,11 @@
                 PersistentEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
                 EntitySection<T> newSection = PersistentEntitySectionManager.this.sectionStorage.getOrCreateSection(newSectionPos);
                 newSection.add(this.entity);
+                long oldSectionKey = currentSectionKey;
                 this.currentSection = newSection;
                 this.currentSectionKey = newSectionPos;
                 this.updateStatus(previousStatus, newSection.getStatus());
+                if (this.realEntity != null) net.neoforged.neoforge.common.CommonHooks.onEntityEnterSection(this.realEntity, oldSectionKey, newSectionPos);
             }
         }
 
