--- a/net/minecraft/world/level/entity/TransientEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/TransientEntitySectionManager.java
@@ -84,11 +_,13 @@
 
     class Callback implements EntityInLevelCallback {
         private final T entity;
+        private final Entity realEntity;
         private long currentSectionKey;
         private EntitySection<T> currentSection;
 
         Callback(T entity, long section, EntitySection<T> currentSection) {
             this.entity = entity;
+            this.realEntity = entity instanceof Entity ? (Entity) entity : null;
             this.currentSectionKey = section;
             this.currentSection = currentSection;
         }
@@ -107,6 +_,7 @@
                 TransientEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
                 EntitySection<T> entitysection = TransientEntitySectionManager.this.sectionStorage.getOrCreateSection(i);
                 entitysection.add(this.entity);
+                long oldSectionKey = currentSectionKey;
                 this.currentSection = entitysection;
                 this.currentSectionKey = i;
                 TransientEntitySectionManager.this.callbacks.onSectionChange(this.entity);
@@ -119,6 +_,7 @@
                         TransientEntitySectionManager.this.callbacks.onTickingStart(this.entity);
                     }
                 }
+                if (this.realEntity != null) net.neoforged.neoforge.common.CommonHooks.onEntityEnterSection(this.realEntity, oldSectionKey, i);
             }
         }
 
