--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModelPart.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModelPart.java
@@ -16,6 +_,8 @@
 public interface BlockStateModelPart extends BlockStateModelPartExtension {
     List<BakedQuad> getQuads(@Nullable Direction direction);
 
+    /// @deprecated Neo: Use [#ambientOcclusion()] instead.
+    @Deprecated
     boolean useAmbientOcclusion();
 
     Material.Baked particleMaterial();
