--- a/net/minecraft/data/models/blockstates/VariantProperties.java
+++ b/net/minecraft/data/models/blockstates/VariantProperties.java
@@ -10,6 +_,10 @@
     public static final VariantProperty<Boolean> UV_LOCK = new VariantProperty<>("uvlock", JsonPrimitive::new);
     public static final VariantProperty<Integer> WEIGHT = new VariantProperty<>("weight", JsonPrimitive::new);
 
+    // Neo: Allow passing model builders into block state variants
+    // Not to be used in conjunction with 'MODEL'
+    public static final VariantProperty<net.neoforged.neoforge.client.model.generators.BlockModelBuilder> MODEL_BUILDER = new VariantProperty<>("model", net.neoforged.neoforge.client.model.generators.BlockModelBuilder::toJson);
+
     public static enum Rotation {
         R0(0),
         R90(90),
