--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1160,7 +1180,7 @@
                                         && !playerStandsOnSomething
                                         && !this.player.isSpectator()
                                         && !this.server.allowFlight()
-                                        && !this.player.getAbilities().mayfly
+                                        && !this.player.mayFly()
                                         && !this.player.hasEffect(MobEffects.LEVITATION)
                                         && !isFallFlying
                                         && !isAutoSpinAttack
