--- a/net/minecraft/client/gui/screens/packs/PackSelectionModel.java
+++ b/net/minecraft/client/gui/screens/packs/PackSelectionModel.java
@@ -31,9 +_,9 @@
         this.onListChanged = onListChanged;
         this.iconGetter = iconGetter;
         this.repository = repository;
-        this.selected = Lists.newArrayList(repository.getSelectedPacks());
+        this.selected = Lists.newArrayList(repository.getSelectedPacks().stream().filter(p -> !p.isHidden()).toList());
         Collections.reverse(this.selected);
-        this.unselected = Lists.newArrayList(repository.getAvailablePacks());
+        this.unselected = Lists.newArrayList(repository.getAvailablePacks().stream().filter(p -> !p.isHidden()).toList());
         this.unselected.removeAll(this.selected);
         this.output = output;
     }
@@ -59,7 +_,7 @@
         this.repository.reload();
         this.selected.retainAll(this.repository.getAvailablePacks());
         this.unselected.clear();
-        this.unselected.addAll(this.repository.getAvailablePacks());
+        this.unselected.addAll(this.repository.getAvailablePacks().stream().filter(p -> !p.isHidden()).toList());
         this.unselected.removeAll(this.selected);
     }
 
