--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -17,8 +17,9 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import net.neoforged.neoforge.common.extensions.IBlockGetterExtension;
 
-public interface BlockGetter extends LevelHeightAccessor {
+public interface BlockGetter extends LevelHeightAccessor, IBlockGetterExtension {
    @Nullable
    BlockEntity getBlockEntity(BlockPos p_45570_);
 
@@ -32,7 +33,7 @@
    FluidState getFluidState(BlockPos p_45569_);
 
    default int getLightEmission(BlockPos p_45572_) {
-      return this.getBlockState(p_45572_).getLightEmission();
+      return this.getBlockState(p_45572_).getLightEmission(this, p_45572_);
    }
 
    default int getMaxLightLevel() {
