--- a/net/minecraft/world/level/BlockGetter.java
+++ b/net/minecraft/world/level/BlockGetter.java
@@ -18,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public interface BlockGetter extends LevelHeightAccessor {
+public interface BlockGetter extends LevelHeightAccessor, net.neoforged.neoforge.common.extensions.IBlockGetterExtension {
     @Nullable
     BlockEntity getBlockEntity(BlockPos pos);
 
@@ -32,7 +_,7 @@
     FluidState getFluidState(BlockPos pos);
 
     default int getLightEmission(BlockPos pos) {
-        return this.getBlockState(pos).getLightEmission();
+        return this.getBlockState(pos).getLightEmission(this, pos);
     }
 
     default int getMaxLightLevel() {
