--- a/net/minecraft/world/level/ExplosionDamageCalculator.java
+++ b/net/minecraft/world/level/ExplosionDamageCalculator.java
@@ -11,7 +_,7 @@
     public Optional<Float> getBlockExplosionResistance(Explosion explosion, BlockGetter reader, BlockPos pos, BlockState state, FluidState fluid) {
         return state.isAir() && fluid.isEmpty()
             ? Optional.empty()
-            : Optional.of(Math.max(state.getBlock().getExplosionResistance(), fluid.getExplosionResistance()));
+            : Optional.of(Math.max(state.getExplosionResistance(reader, pos, explosion), fluid.getExplosionResistance(reader, pos, explosion)));
     }
 
     public boolean shouldBlockExplode(Explosion explosion, BlockGetter reader, BlockPos pos, BlockState state, float power) {
