--- a/net/minecraft/world/level/ForcedChunksSavedData.java
+++ b/net/minecraft/world/level/ForcedChunksSavedData.java
@@ -25,16 +_,31 @@
     }
 
     public static ForcedChunksSavedData load(CompoundTag tag, HolderLookup.Provider registries) {
-        return new ForcedChunksSavedData(new LongOpenHashSet(tag.getLongArray("Forced")));
+        ForcedChunksSavedData savedData = new ForcedChunksSavedData(new LongOpenHashSet(tag.getLongArray("Forced")));
+        net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.readModForcedChunks(tag, savedData.blockForcedChunks, savedData.entityForcedChunks);
+        return savedData;
     }
 
     @Override
     public CompoundTag save(CompoundTag tag, HolderLookup.Provider registries) {
         tag.putLongArray("Forced", this.chunks.toLongArray());
+        net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.writeModForcedChunks(tag, this.blockForcedChunks, this.entityForcedChunks);
         return tag;
     }
 
     public LongSet getChunks() {
         return this.chunks;
+    }
+
+    // Neo: Keep track of forced loaded chunks caused by entities or blocks.
+    private final net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<net.minecraft.core.BlockPos> blockForcedChunks = new net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<>();
+    private final net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<java.util.UUID> entityForcedChunks = new net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<>();
+
+    public net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<net.minecraft.core.BlockPos> getBlockForcedChunks() {
+        return this.blockForcedChunks;
+    }
+
+    public net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<java.util.UUID> getEntityForcedChunks() {
+        return this.entityForcedChunks;
     }
 }
