--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -22,7 +_,7 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.AABB;
 
-public interface LevelReader extends BlockAndTintGetter, CollisionGetter, SignalGetter, BiomeManager.NoiseBiomeSource {
+public interface LevelReader extends BlockAndTintGetter, CollisionGetter, SignalGetter, BiomeManager.NoiseBiomeSource, net.neoforged.neoforge.common.extensions.ILevelReaderExtension {
     @Nullable
     ChunkAccess getChunk(int p_46823_, int p_46824_, ChunkStatus p_330944_, boolean p_46826_);
 
@@ -181,6 +_,10 @@
     @Deprecated
     default boolean hasChunkAt(BlockPos p_46806_) {
         return this.hasChunkAt(p_46806_.getX(), p_46806_.getZ());
+    }
+
+    default boolean isAreaLoaded(BlockPos center, int range) {
+        return this.hasChunksAt(center.offset(-range, -range, -range), center.offset(range, range, range));
     }
 
     @Deprecated
