--- a/net/minecraft/world/level/SignalGetter.java
+++ b/net/minecraft/world/level/SignalGetter.java
@@ -65,7 +_,7 @@
     default int getSignal(BlockPos pos, Direction direction) {
         BlockState blockstate = this.getBlockState(pos);
         int i = blockstate.getSignal(this, pos, direction);
-        return blockstate.isRedstoneConductor(this, pos) ? Math.max(i, this.getDirectSignalTo(pos)) : i;
+        return blockstate.shouldCheckWeakPower(this, pos, direction) ? Math.max(i, this.getDirectSignalTo(pos)) : i;
     }
 
     default boolean hasNeighborSignal(BlockPos pos) {
