--- a/net/minecraft/world/level/TicketStorage.java
+++ b/net/minecraft/world/level/TicketStorage.java
@@ -35,7 +_,9 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Codec<Pair<ChunkPos, Ticket>> TICKET_ENTRY = Codec.mapPair(ChunkPos.CODEC.fieldOf("chunk_pos"), Ticket.CODEC).codec();
     public static final Codec<TicketStorage> CODEC = RecordCodecBuilder.create(
-        i -> i.group(TICKET_ENTRY.listOf().optionalFieldOf("tickets", List.of()).forGetter(TicketStorage::packTickets)).apply(i, TicketStorage::fromPacked)
+        i -> i.group(TICKET_ENTRY.listOf().optionalFieldOf("tickets", List.of()).forGetter(TicketStorage::packTickets))
+                .and(net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.defineExtraStorageParams())
+                .apply(i, (tickets, neoData) -> net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.readStoredTickets(TicketStorage::fromPacked, tickets, neoData))
     );
     public static final SavedDataType<TicketStorage> TYPE = new SavedDataType<>(
         Identifier.withDefaultNamespace("chunk_tickets"), TicketStorage::new, CODEC, DataFixTypes.SAVED_DATA_FORCED_CHUNKS
@@ -50,6 +_,7 @@
         this.tickets = tickets;
         this.deactivatedTickets = deactivatedTickets;
         this.updateForcedChunks();
+        this.updateForcedNaturalSpawning();
     }
 
     public TicketStorage() {
@@ -173,6 +_,7 @@
         if (ticket.getType().equals(TicketType.FORCED)) {
             this.chunksWithForcedTickets.add(key);
         }
+        if (ticket.getType().forceNaturalSpawning()) chunksWithForceNaturalSpawning.add(key);
 
         this.setDirty();
         return true;
@@ -261,6 +_,7 @@
         if (ticket.getType().equals(TicketType.FORCED)) {
             this.updateForcedChunks();
         }
+        if (ticket.getType().forceNaturalSpawning()) this.updateForcedNaturalSpawning();
 
         this.setDirty();
         return true;
@@ -303,11 +_,14 @@
 
     public void deactivateTicketsOnClosing() {
         this.removeTicketIf((ticket, chunkPos) -> ticket.getType() != TicketType.UNKNOWN, this.deactivatedTickets);
+        blockForcedChunks.deactivateTicketsOnClosing();
+        entityForcedChunks.deactivateTicketsOnClosing();
     }
 
     public void removeTicketIf(TicketStorage.TicketPredicate predicate, @Nullable Long2ObjectOpenHashMap<List<Ticket>> removedTickets) {
         ObjectIterator<Entry<List<Ticket>>> ticketsPerChunkIterator = this.tickets.long2ObjectEntrySet().fastIterator();
         boolean removedForced = false;
+        boolean updateNaturalSpawning = false;
 
         while (ticketsPerChunkIterator.hasNext()) {
             Entry<List<Ticket>> entry = ticketsPerChunkIterator.next();
@@ -336,6 +_,7 @@
                     if (ticket.getType().equals(TicketType.FORCED)) {
                         removedForced = true;
                     }
+                    if (ticket.getType().forceNaturalSpawning()) updateNaturalSpawning = true;
                 }
             }
 
@@ -358,6 +_,7 @@
         if (removedForced) {
             this.updateForcedChunks();
         }
+        if (updateNaturalSpawning) this.updateForcedNaturalSpawning();
     }
 
     public void replaceTicketLevelOfType(int newLevel, TicketType ticketType) {
@@ -411,5 +_,26 @@
 
     public interface TicketPredicate {
         boolean test(Ticket ticket, long chunkPos);
+    }
+
+    // Neo: Keep track of forced loaded chunks caused by entities or blocks.
+    private final net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<net.minecraft.core.BlockPos> blockForcedChunks = new net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<>(this, net.neoforged.neoforge.common.NeoForgeMod.BLOCK_TICKET, net.neoforged.neoforge.common.NeoForgeMod.BLOCK_WITH_NATURAL_SPAWNING_TICKET);
+    private final net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<java.util.UUID> entityForcedChunks = new net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<>(this, net.neoforged.neoforge.common.NeoForgeMod.ENTITY_TICKET, net.neoforged.neoforge.common.NeoForgeMod.ENTITY_WITH_NATURAL_SPAWNING_TICKET);
+    private LongSet chunksWithForceNaturalSpawning = new LongOpenHashSet();
+
+    public net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<net.minecraft.core.BlockPos> getBlockForcedChunks() {
+        return this.blockForcedChunks;
+    }
+
+    public net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.TicketTracker<java.util.UUID> getEntityForcedChunks() {
+        return this.entityForcedChunks;
+    }
+
+    private void updateForcedNaturalSpawning() {
+        this.chunksWithForceNaturalSpawning = this.getAllChunksWithTicketThat(ticket -> ticket.getType().forceNaturalSpawning());
+    }
+
+    public boolean shouldForceNaturalSpawning(ChunkPos chunkPos) {
+        return chunksWithForceNaturalSpawning.contains(chunkPos.pack());
     }
 }
