--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1068,7 +1104,7 @@
             }
 
             if (!forced) {
-                return Optional.empty();
+                return blockState.getRespawnPosition(EntityTypes.PLAYER, level, pos, yaw);
             }
 
             boolean freeBottom = block.isPossibleToRespawnInThis(blockState);
@@ -1184,8 +1225,19 @@