--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -47,6 +47,8 @@
     public static final int MIN_LEVEL = 0;
     public static final int MAX_LEVEL = 7;
     public static final IntegerProperty LEVEL = BlockStateProperties.LEVEL_COMPOSTER;
+    /** @deprecated Neo: Use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#COMPOSTABLES compostable} data map instead */
+    @Deprecated
     public static final Object2FloatMap<ItemLike> COMPOSTABLES = new Object2FloatOpenHashMap<>();
     private static final int HOLE_WIDTH = 12;
     private static final VoxelShape[] SHAPES = Util.make(
@@ -249,7 +260,7 @@
         ItemStack itemStack, BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult
     ) {
         int fillLevel = state.getValue(LEVEL);
-        if (fillLevel < 8 && COMPOSTABLES.containsKey(itemStack.getItem())) {
+        if (fillLevel < 8 && getValue(itemStack) > 0) {
             if (fillLevel < 7 && !level.isClientSide()) {
                 BlockState newState = addItem(player, state, level, pos, itemStack);
                 level.levelEvent(1500, pos, state != newState ? 1 : 0);
@@ -276,7 +287,7 @@
 
     public static BlockState insertItem(Entity sourceEntity, BlockState state, ServerLevel level, ItemStack itemStack, BlockPos pos) {
         int fillLevel = state.getValue(LEVEL);
-        if (fillLevel < 7 && COMPOSTABLES.containsKey(itemStack.getItem())) {
+        if (fillLevel < 7 && getValue(itemStack) > 0) {
             BlockState newState = addItem(sourceEntity, state, level, pos, itemStack);
             itemStack.shrink(1);
             return newState;
@@ -307,7 +318,7 @@
 
     private static BlockState addItem(@Nullable Entity sourceEntity, BlockState state, LevelAccessor level, BlockPos pos, ItemStack itemStack) {
         int fillLevel = state.getValue(LEVEL);
-        float chance = COMPOSTABLES.getFloat(itemStack.getItem());
+        float chance = getValue(itemStack);
         if ((fillLevel != 0 || !(chance > 0.0F)) && !(level.getRandom().nextDouble() < chance)) {
             return state;
         }
