--- a/net/minecraft/world/level/block/FarmlandBlock.java
+++ b/net/minecraft/world/level/block/FarmlandBlock.java
@@ -129,13 +126,14 @@
     }
 
     private static boolean isNearWater(LevelReader level, BlockPos pos) {
+        BlockState state = level.getBlockState(pos);
         for (BlockPos blockPos : BlockPos.betweenClosed(pos.offset(-4, 0, -4), pos.offset(4, 1, 4))) {
-            if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
+            if (state.canBeHydrated(level, pos, level.getFluidState(blockPos), blockPos)) {
                 return true;
             }
         }
 
-        return false;
+        return net.neoforged.neoforge.common.FarmlandWaterManager.hasBlockWaterTicket(level, pos);
     }
 
     @Override
