--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -320,10 +320,14 @@
 
     @Override
     public final void findBlockLightSources(BiConsumer<BlockPos, BlockState> consumer) {
-        this.findBlocks(state -> state.getLightEmission() != 0, consumer);
+        this.findBlocks(p_284897_ -> p_284897_.hasDynamicLightEmission() || p_284897_.getLightEmission(net.minecraft.world.level.EmptyBlockGetter.INSTANCE, BlockPos.ZERO) != 0, (p_284897_, pos) -> p_284897_.getLightEmission(this, pos) != 0, consumer);
     }
 
     public void findBlocks(Predicate<BlockState> predicate, BiConsumer<BlockPos, BlockState> consumer) {
+        findBlocks(predicate, (state, pos) -> predicate.test(state), consumer);
+    }
+
+    public void findBlocks(Predicate<BlockState> predicate, java.util.function.BiPredicate<BlockState, BlockPos> fineFilter, BiConsumer<BlockPos, BlockState> consumer) {
         BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
 
         for (int sectionY = this.getMinSectionY(); sectionY <= this.getMaxSectionY(); sectionY++) {
@@ -335,8 +339,9 @@
                     for (int z = 0; z < 16; z++) {
                         for (int x = 0; x < 16; x++) {
                             BlockState state = section.getBlockState(x, y, z);
-                            if (predicate.test(state)) {
-                                consumer.accept(mutablePos.setWithOffset(origin, x, y, z), state);
+                            mutablePos.setWithOffset(origin, x, y, z);
+                            if (fineFilter.test(state, mutablePos)) {
+                                consumer.accept(mutablePos, state);
                             }
                         }
                     }
