--- a/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
+++ b/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
@@ -29,6 +29,7 @@
     private final RecipeCategory category;
     private final CookingBookCategory bookCategory;
     private final Item result;
+    private final ItemStack stackResult; // Neo: add stack result support
     private final Ingredient ingredient;
     private final float experience;
     private final int cookingTime;
@@ -45,9 +46,22 @@
         int cookingTime,
         AbstractCookingRecipe.Factory<?> factory
     ) {
+        this(category, bookCategory, new ItemStack(result), ingredient, experience, cookingTime, factory);
+    }
+
+    private SimpleCookingRecipeBuilder(
+            RecipeCategory category,
+            CookingBookCategory bookCategory,
+            ItemStack result,
+            Ingredient ingredient,
+            float experience,
+            int cookingTime,
+            AbstractCookingRecipe.Factory<?> factory
+    ) {
         this.category = category;
         this.bookCategory = bookCategory;
-        this.result = result.asItem();
+        this.result = result.getItem();
+        this.stackResult = result;
         this.ingredient = ingredient;
         this.experience = experience;
         this.cookingTime = cookingTime;
@@ -111,7 +157,7 @@
         this.criteria.forEach(advancement::addCriterion);
         AbstractCookingRecipe recipe = this.factory
             .create(
-                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, new ItemStack(this.result), this.experience, this.cookingTime
+                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, this.stackResult, this.experience, this.cookingTime
             );
         output.accept(id, recipe, advancement.build(id.identifier().withPrefix("recipes/" + this.category.getFolderName() + "/")));
     }
