--- a/net/minecraft/server/packs/repository/BuiltInPackSource.java
+++ b/net/minecraft/server/packs/repository/BuiltInPackSource.java
@@ -106,4 +_,18 @@
             }
         };
     }
+
+    public static Pack.ResourcesSupplier fromName(final Function<PackLocationInfo, PackResources> onName) {
+        return new Pack.ResourcesSupplier() {
+            @Override
+            public PackResources openPrimary(PackLocationInfo p_294636_) {
+                return onName.apply(p_294636_);
+            }
+
+            @Override
+            public PackResources openFull(PackLocationInfo p_251717_, Pack.Metadata p_294956_) {
+                return onName.apply(p_251717_);
+            }
+        };
+    }
 }
