++++ Target missing
--- a/net/minecraft/client/renderer/block/model/BlockElementFace.java
+++ b/net/minecraft/client/renderer/block/model/BlockElementFace.java
@@ -15,7 +15,7 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public record BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, BlockElementFace.@Nullable UVs uvs, Quadrant rotation) {
+public record BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, BlockElementFace.@Nullable UVs uvs, Quadrant rotation, net.neoforged.neoforge.client.model.@Nullable ExtraFaceData faceData, org.apache.commons.lang3.mutable.MutableObject<BlockElement> parent) {
     public static final int NO_TINT = -1;
 
     public static float getU(BlockElementFace.UVs uvs, Quadrant rotation, int vertex) {
@@ -26,6 +26,21 @@
         return uvs.getVertexV(rotation.rotateVertexIndex(index)) / 16.0F;
     }
 
+    public BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, BlockElementFace.@Nullable UVs uvs, Quadrant rotation) {
+        this(cullForDirection, tintIndex, texture, uvs, rotation, null, new org.apache.commons.lang3.mutable.MutableObject<>());
+    }
+
+    @Override
+    public net.neoforged.neoforge.client.model.ExtraFaceData faceData() {
+        if (this.faceData != null) {
+            return this.faceData;
+        } else if (this.parent.getValue() != null) {
+            return this.parent.getValue().faceData();
+        } else {
+            return net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT;
+        }
+    }
+
     @OnlyIn(Dist.CLIENT)
     public static class Deserializer implements JsonDeserializer<BlockElementFace> {
         private static final int DEFAULT_TINT_INDEX = -1;
@@ -38,7 +53,9 @@
             String texture = getTexture(object);
             BlockElementFace.UVs uvs = getUVs(object);
             Quadrant rotation = getRotation(object);
-            return new BlockElementFace(cullDirection, tintIndex, texture, uvs, rotation);
+            if (object.has("forge_data")) throw new JsonParseException("forge_data should be replaced by neoforge_data"); // TODO 1.22: Remove
+            var faceData = net.neoforged.neoforge.client.model.ExtraFaceData.read(object.get("neoforge_data"), null);
+            return new BlockElementFace(cullDirection, tintIndex, texture, uvs, rotation, faceData, new org.apache.commons.lang3.mutable.MutableObject<>());
         }
 
         private static int getTintIndex(JsonObject object) {
