++++ Target missing
--- a/net/minecraft/client/renderer/block/model/BlockElement.java
+++ b/net/minecraft/client/renderer/block/model/BlockElement.java
@@ -21,12 +21,20 @@
 
 @OnlyIn(Dist.CLIENT)
 public record BlockElement(
-    Vector3fc from, Vector3fc to, Map<Direction, BlockElementFace> faces, @Nullable BlockElementRotation rotation, boolean shade, int lightEmission
+    Vector3fc from, Vector3fc to, Map<Direction, BlockElementFace> faces, @Nullable BlockElementRotation rotation, boolean shade, int lightEmission, net.neoforged.neoforge.client.model.ExtraFaceData faceData
 ) {
     private static final boolean DEFAULT_RESCALE = false;
     private static final float MIN_EXTENT = -16.0F;
     private static final float MAX_EXTENT = 32.0F;
 
+    public BlockElement {
+        faces.values().forEach(face -> face.parent().setValue(this));
+    }
+
+    public BlockElement(Vector3fc from, Vector3fc to, Map<Direction, BlockElementFace> faces, @Nullable BlockElementRotation rotation, boolean shade, int lightEmission) {
+        this(from, to, faces, rotation, shade, lightEmission, net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT);
+    }
+
     public BlockElement(Vector3fc from, Vector3fc to, Map<Direction, BlockElementFace> faces) {
         this(from, to, faces, null, true, 0);
     }
@@ -71,7 +79,8 @@
                     }
                 }
 
-                return new BlockElement(from, to, faces, rotation, shade, lightEmission);
+                var faceData = net.neoforged.neoforge.client.model.ExtraFaceData.read(object.get("neoforge_data"), net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT);
+                return new BlockElement(from, to, faces, rotation, shade, lightEmission, faceData);
             }
         }
 
