--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -209,7 +209,7 @@
             result.add(
                 jigsaw.withInfo(
                     new StructureTemplate.StructureBlockInfo(
-                        calculateRelativePosition(settings, blockInfo.pos()).offset(position), blockInfo.state.rotate(settings.getRotation()), blockInfo.nbt
+                        calculateRelativePosition(settings, blockInfo.pos()).offset(position), blockInfo.state.rotate(settings.getRotation()), net.neoforged.neoforge.common.CommonHooks.captureStructureRotation(blockInfo.nbt, rotation)
                     )
                 )
             );
@@ -498,12 +510,12 @@
                 tag.put("Pos", posTag);
                 tag.remove("UUID");
                 createEntityIgnoreException(problemReporter, level, tag).ifPresent(entity -> {
-                    float yRot = entity.rotate(rotation);
-                    yRot += entity.mirror(mirror) - entity.getYRot();
+                    float yRot = entity.rotate(placementIn.getRotation());
+                    yRot += entity.mirror(placementIn.getMirror()) - entity.getYRot();
                     entity.snapTo(pos.x, pos.y, pos.z, yRot, entity.getXRot());
                     entity.setYBodyRot(yRot);
                     entity.setYHeadRot(yRot);
-                    if (finalizeEntities && entity instanceof Mob mob) {
+                    if (placementIn.shouldFinalizeEntities() && entity instanceof Mob mob) {
                         mob.finalizeSpawn(level, level.getCurrentDifficultyAt(BlockPos.containing(pos)), EntitySpawnReason.STRUCTURE, null);
                     }
 
