--- a/net/minecraft/world/flag/FeatureFlagRegistry.java
+++ b/net/minecraft/world/flag/FeatureFlagRegistry.java
@@ -75,6 +_,18 @@
         }, set -> List.copyOf(this.toNames(set)));
     }
 
+    public FeatureFlag getFlag(Identifier name) {
+        return com.google.common.base.Preconditions.checkNotNull(this.names.get(name), "Flag %s was not registered", name);
+    }
+
+    public Map<Identifier, FeatureFlag> getAllFlags() {
+        return this.names;
+    }
+
+    public boolean hasAnyModdedFlags() {
+        return this.names.values().stream().anyMatch(FeatureFlag::isModded);
+    }
+
     public static class Builder {
         private final FeatureFlagUniverse universe;
         private int id;
@@ -88,14 +_,23 @@
             return this.create(Identifier.withDefaultNamespace(name));
         }
 
+        /**
+         * @deprecated Neo: use {@link #create(Identifier, boolean)} instead
+         */
+        @Deprecated
         public FeatureFlag create(Identifier name) {
-            if (this.id >= 64) {
+            return create(name, false);
+        }
+
+        public FeatureFlag create(Identifier p_250098_, boolean modded) {
+            if (this.id >= 64 && false) {
                 throw new IllegalStateException("Too many feature flags");
             } else {
-                FeatureFlag result = new FeatureFlag(this.universe, this.id++);
-                FeatureFlag previous = this.flags.put(name, result);
+                FeatureFlag result = new FeatureFlag(this.universe, this.id % FeatureFlagSet.MAX_CONTAINER_SIZE, this.id / FeatureFlagSet.MAX_CONTAINER_SIZE, modded);
+                this.id++;
+                FeatureFlag previous = this.flags.put(p_250098_, result);
                 if (previous != null) {
-                    throw new IllegalStateException("Duplicate feature flag " + name);
+                    throw new IllegalStateException("Duplicate feature flag " + p_250098_);
                 } else {
                     return result;
                 }
