--- a/net/minecraft/world/flag/FeatureFlags.java
+++ b/net/minecraft/world/flag/FeatureFlags.java
@@ -35,6 +_,7 @@
         TRADE_REBALANCE = builder.createVanilla("trade_rebalance");
         REDSTONE_EXPERIMENTS = builder.createVanilla("redstone_experiments");
         MINECART_IMPROVEMENTS = builder.createVanilla("minecart_improvements");
+        net.neoforged.neoforge.common.util.flag.FeatureFlagLoader.loadModdedFlags(builder);
         REGISTRY = builder.build();
         CODEC = REGISTRY.codec();
         VANILLA_SET = FeatureFlagSet.of(VANILLA);
