--- a/net/minecraft/world/flag/FeatureFlags.java
+++ b/net/minecraft/world/flag/FeatureFlags.java
@@ -10,6 +_,9 @@
     public static final FeatureFlag BUNDLE;
     public static final FeatureFlag UPDATE_1_21;
     public static final FeatureFlag TRADE_REBALANCE;
+    // Neo: FeatureFlag used to mark modded elements as experimental
+    // Modders should not be making their own flags, use this one instead
+    public static final FeatureFlag MOD_EXPERIMENTAL;
     public static final FeatureFlagRegistry REGISTRY;
     public static final Codec<FeatureFlagSet> CODEC;
     public static final FeatureFlagSet VANILLA_SET;
@@ -35,6 +_,7 @@
         BUNDLE = featureflagregistry$builder.createVanilla("bundle");
         TRADE_REBALANCE = featureflagregistry$builder.createVanilla("trade_rebalance");
         UPDATE_1_21 = featureflagregistry$builder.createVanilla("update_1_21");
+        MOD_EXPERIMENTAL = featureflagregistry$builder.create(new ResourceLocation(net.neoforged.neoforge.internal.versions.neoforge.NeoForgeVersion.MOD_ID, "experimental"));
         REGISTRY = featureflagregistry$builder.build();
         CODEC = REGISTRY.codec();
         VANILLA_SET = FeatureFlagSet.of(VANILLA);
