--- a/net/minecraft/server/network/config/SynchronizeRegistriesTask.java
+++ b/net/minecraft/server/network/config/SynchronizeRegistriesTask.java
@@ -45,11 +_,10 @@
     }
 
     public void handleResponse(List<KnownPack> packs, Consumer<Packet<?>> packetSender) {
-        if (packs.equals(this.requestedPacks)) {
-            this.sendRegistries(packetSender, Set.copyOf(this.requestedPacks));
-        } else {
-            this.sendRegistries(packetSender, Set.of());
-        }
+        // Neo: instead of using either all available KnownPacks or none, allow partial fallback to normal syncing
+        Set<KnownPack> requested = new java.util.HashSet<>(this.requestedPacks);
+        requested.retainAll(packs);
+        this.sendRegistries(packetSender, requested);
     }
 
     @Override
