--- a/net/minecraft/world/damagesource/DamageSources.java
+++ b/net/minecraft/world/damagesource/DamageSources.java
@@ -43,9 +43,15 @@
     private final DamageSource stalagmite;
     private final DamageSource outsideBorder;
     private final DamageSource genericKill;
+    // CraftBukkit start
+    private final DamageSource melting;
+    private final DamageSource poison;
 
     public DamageSources(RegistryAccess registries) {
         this.damageTypes = registries.lookupOrThrow(Registries.DAMAGE_TYPE);
+        this.melting = this.source(DamageTypes.ON_FIRE).melting();
+        this.poison = this.source(DamageTypes.MAGIC).poison();
+        // CraftBukkit end
         this.inFire = this.source(DamageTypes.IN_FIRE);
         this.campfire = this.source(DamageTypes.CAMPFIRE);
         this.lightningBolt = this.source(DamageTypes.LIGHTNING_BOLT);
@@ -85,6 +91,16 @@
         return new DamageSource(this.damageTypes.getOrThrow(key), directEntity, causingEntity);
     }
 
+    // CraftBukkit start
+    public DamageSource melting() {
+        return this.melting;
+    }
+
+    public DamageSource poison() {
+        return this.poison;
+    }
+    // CraftBukkit end
+
     public DamageSource inFire() {
         return this.inFire;
     }
@@ -254,7 +270,13 @@
     }
 
     public DamageSource explosion(@Nullable Entity entity, @Nullable Entity cause) {
-        return this.source(cause != null && entity != null ? DamageTypes.PLAYER_EXPLOSION : DamageTypes.EXPLOSION, entity, cause);
+        // CraftBukkit start
+        return this.explosion(entity, cause, cause != null && entity != null ? DamageTypes.PLAYER_EXPLOSION : DamageTypes.EXPLOSION);
+    }
+
+    public DamageSource explosion(@Nullable Entity entity, @Nullable Entity entity1, ResourceKey<DamageType> resourceKey) {
+        return this.source(resourceKey, entity, entity1);
+        // CraftBukkit end
     }
 
     public DamageSource sonicBoom(Entity entity) {
@@ -262,7 +284,13 @@
     }
 
     public DamageSource badRespawnPointExplosion(Vec3 boomPos) {
-        return new DamageSource(this.damageTypes.getOrThrow(DamageTypes.BAD_RESPAWN_POINT), boomPos);
+        // CraftBukkit start
+        return badRespawnPointExplosion(boomPos, null);
+    }
+
+    public DamageSource badRespawnPointExplosion(Vec3 vec3, org.bukkit.block.BlockState blockState) {
+        return new DamageSource(this.damageTypes.getOrThrow(DamageTypes.BAD_RESPAWN_POINT), vec3).directBlockState(blockState);
+        // CraftBukkit end
     }
 
     public DamageSource outOfBorder() {
