--- a/net/minecraft/world/entity/animal/Cat.java
+++ b/net/minecraft/world/entity/animal/Cat.java
@@ -378,8 +_,8 @@
                     }
                 } else if (item.isEdible() && this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
                     if (!this.level().isClientSide()) {
+                        this.heal((float)itemstack.getFoodProperties(this).getNutrition());
                         this.usePlayerItem(p_28153_, p_28154_, itemstack);
-                        this.heal((float)item.getFoodProperties().getNutrition());
                     }
 
                     return InteractionResult.sidedSuccess(this.level().isClientSide());
@@ -439,7 +_,7 @@
     }
 
     private void tryToTame(Player p_333858_) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, p_333858_)) {
             this.tame(p_333858_);
             this.setOrderedToSit(true);
             this.level().broadcastEntityEvent(this, (byte)7);
