--- a/net/minecraft/world/entity/animal/Cat.java
+++ b/net/minecraft/world/entity/animal/Cat.java
@@ -368,9 +_,9 @@
                     }
                 } else if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
                     if (!this.level().isClientSide()) {
-                        this.usePlayerItem(player, hand, itemstack);
-                        FoodProperties foodproperties = itemstack.get(DataComponents.FOOD);
+                        FoodProperties foodproperties = itemstack.getFoodProperties(this);
                         this.heal(foodproperties != null ? (float)foodproperties.nutrition() : 1.0F);
+                        this.usePlayerItem(player, hand, itemstack);
                     }
 
                     return InteractionResult.sidedSuccess(this.level().isClientSide());
@@ -430,7 +_,7 @@
     }
 
     private void tryToTame(Player player) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, player)) {
             this.tame(player);
             this.setOrderedToSit(true);
             this.level().broadcastEntityEvent(this, (byte)7);
