--- a/net/minecraft/world/entity/animal/Parrot.java
+++ b/net/minecraft/world/entity/animal/Parrot.java
@@ -73,13 +_,15 @@
     private static final EntityDataAccessor<Integer> DATA_VARIANT_ID = SynchedEntityData.defineId(Parrot.class, EntityDataSerializers.INT);
     private static final Predicate<Mob> NOT_PARROT_PREDICATE = new Predicate<Mob>() {
         public boolean test(@Nullable Mob p_29453_) {
-            return p_29453_ != null && Parrot.MOB_SOUND_MAP.containsKey(p_29453_.getType());
+            return p_29453_ != null && (Parrot.MOB_SOUND_MAP.containsKey(p_29453_.getType()) || p_29453_.getType().builtInRegistryHolder().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.PARROT_IMITATIONS) != null);
         }
     };
     private static final Item POISONOUS_FOOD = Items.COOKIE;
     private static final Set<Item> TAME_FOOD = Sets.newHashSet(
         Items.WHEAT_SEEDS, Items.MELON_SEEDS, Items.PUMPKIN_SEEDS, Items.BEETROOT_SEEDS, Items.TORCHFLOWER_SEEDS, Items.PITCHER_POD
     );
+    /** @deprecated Neo: use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#PARROT_IMITATIONS data map} instead */
+    @Deprecated
     static final Map<EntityType<?>, SoundEvent> MOB_SOUND_MAP = Util.make(Maps.newHashMap(), p_326764_ -> {
         p_326764_.put(EntityType.BLAZE, SoundEvents.PARROT_IMITATE_BLAZE);
         p_326764_.put(EntityType.BOGGED, SoundEvents.PARROT_IMITATE_BOGGED);
@@ -260,7 +_,7 @@
             }
 
             if (!this.level().isClientSide) {
-                if (this.random.nextInt(10) == 0) {
+                if (this.random.nextInt(10) == 0 && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, p_29414_)) {
                     this.tame(p_29414_);
                     this.level().broadcastEntityEvent(this, (byte)7);
                 } else {
@@ -329,7 +_,10 @@
 
     public static SoundEvent getAmbient(Level p_218239_, RandomSource p_218240_) {
         if (p_218239_.getDifficulty() != Difficulty.PEACEFUL && p_218240_.nextInt(1000) == 0) {
-            List<EntityType<?>> list = Lists.newArrayList(MOB_SOUND_MAP.keySet());
+            final var entities = new java.util.HashSet<>(MOB_SOUND_MAP.keySet());
+            final var registry = p_218239_.registryAccess().registryOrThrow(net.minecraft.core.registries.Registries.ENTITY_TYPE);
+            registry.getDataMap(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.PARROT_IMITATIONS).keySet().forEach(key -> entities.add(registry.get(key)));
+            List<EntityType<?>> list = Lists.newArrayList(entities);
             return getImitatedSound(list.get(p_218240_.nextInt(list.size())));
         } else {
             return SoundEvents.PARROT_AMBIENT;
@@ -337,6 +_,8 @@
     }
 
     private static SoundEvent getImitatedSound(EntityType<?> p_29409_) {
+        var imitation = p_29409_.builtInRegistryHolder().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.PARROT_IMITATIONS);
+        if (imitation != null) return imitation.sound();
         return MOB_SOUND_MAP.getOrDefault(p_29409_, SoundEvents.PARROT_AMBIENT);
     }
 
