--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -206,7 +_,7 @@
 
     @Override
     public void thunderHit(ServerLevel level, LightningBolt lightning) {
-        if (level.getDifficulty() != Difficulty.PEACEFUL) {
+        if (level.getDifficulty() != Difficulty.PEACEFUL && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> {})) {
             ZombifiedPiglin zombifiedpiglin = EntityType.ZOMBIFIED_PIGLIN.create(level);
             if (zombifiedpiglin != null) {
                 zombifiedpiglin.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
@@ -219,6 +_,7 @@
                 }
 
                 zombifiedpiglin.setPersistenceRequired();
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, zombifiedpiglin);
                 level.addFreshEntity(zombifiedpiglin);
                 this.discard();
             } else {
