--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -92,7 +_,7 @@
                 this.hurt(this.damageSources().onFire(), 1.0F);
             }
 
-            if (!this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(this.level(), this)) {
                 return;
             }
 
