--- a/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
@@ -64,7 +_,7 @@
     protected static void setDirtAt(
         LevelSimulatedReader level, BiConsumer<BlockPos, BlockState> blockSetter, RandomSource random, BlockPos pos, TreeConfiguration config
     ) {
-        if (config.forceDirt || !isDirt(level, pos)) {
+        if (!(((net.minecraft.world.level.LevelReader) level).getBlockState(pos).onTreeGrow((net.minecraft.world.level.LevelReader) level, blockSetter, random, pos, config)) && (config.forceDirt || !isDirt(level, pos))) {
             blockSetter.accept(pos, config.dirtProvider.getState(random, pos));
         }
     }
