--- a/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
@@ -58,6 +_,10 @@
     protected static void placeBelowTrunkBlock(
         WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos pos, TreeConfiguration config
     ) {
+        if (level.getBlockState(pos).onTreeGrow(level, trunkSetter, random, pos, config)) {
+            return;
+        }
+
         BlockState blockBelowTrunk = config.belowTrunkProvider.getState(level, random, pos);
         if (blockBelowTrunk != null) {
             trunkSetter.accept(pos, blockBelowTrunk);
