--- a/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
+++ b/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
@@ -262,7 +_,7 @@
 
         for (RecipeBookTabButton tabButton : this.tabButtons) {
             ExtendedRecipeBookCategory category = tabButton.getCategory();
-            if (category instanceof SearchRecipeBookCategory) {
+            if (category instanceof SearchRecipeBookCategory || net.neoforged.neoforge.client.RecipeBookManager.getSearchCategories().containsKey(category)) {
                 tabButton.visible = true;
                 tabButton.setPosition(xPosTab, yPosTab + 27 * index++);
             } else if (tabButton.updateVisibility(this.book)) {
