--- a/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
+++ b/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
@@ -260,6 +_,7 @@
 
         for (RecipeBookTabButton recipebooktabbutton : this.tabButtons) {
             ExtendedRecipeBookCategory extendedrecipebookcategory = recipebooktabbutton.getCategory();
+            // TODO 1.21.2: make this accessible to mods.
             if (extendedrecipebookcategory instanceof SearchRecipeBookCategory) {
                 recipebooktabbutton.visible = true;
                 recipebooktabbutton.setPosition(i, j + 27 * l++);
