--- a/net/minecraft/client/gui/screens/controls/KeyBindsList.java
+++ b/net/minecraft/client/gui/screens/controls/KeyBindsList.java
@@ -165,6 +_,7 @@
                 )
                 .build();
             this.resetButton = Button.builder(Component.translatable("controls.reset"), p_269616_ -> {
+                this.key.setToDefault();
                 KeyBindsList.this.minecraft.options.setKey(p_193916_, p_193916_.getDefaultKey());
                 KeyBindsList.this.resetMappingAndUpdateButtons();
             }).bounds(0, 0, 50, 20).createNarration(p_253313_ -> Component.translatable("narrator.controls.reset", p_193917_)).build();
@@ -184,8 +_,9 @@
             boolean p_282605_,
             float p_281432_
         ) {
-            int k = p_281373_ + 90 - KeyBindsList.this.maxNameWidth;
-            p_281805_.drawString(KeyBindsList.this.minecraft.font, this.name, k, p_282357_ + p_281932_ / 2 - 9 / 2, 16777215, false);
+            int textMaxWidth = p_281805_.guiWidth() / 2;
+            int k = p_281373_ + 90 - Math.min(KeyBindsList.this.maxNameWidth, textMaxWidth - 40);
+            net.minecraft.client.gui.components.AbstractWidget.renderScrollingString(p_281805_, KeyBindsList.this.minecraft.font, this.name, k, k, p_282357_ + p_281932_ / 2 - 9 / 2, textMaxWidth, p_282357_ + p_281932_ / 2 - 9 / 2 + 16, 16777215); // Neo: Makes descriptions start scrolling if too long
             this.resetButton.setX(p_281373_ + 190);
             this.resetButton.setY(p_282357_);
             this.resetButton.render(p_281805_, p_282224_, p_282053_, p_281432_);
@@ -218,7 +_,7 @@
             MutableComponent mutablecomponent = Component.empty();
             if (!this.key.isUnbound()) {
                 for(KeyMapping keymapping : KeyBindsList.this.minecraft.options.keyMappings) {
-                    if (keymapping != this.key && this.key.same(keymapping)) {
+                    if ((keymapping != this.key && this.key.same(keymapping)) || keymapping.hasKeyModifierConflict(this.key)) { // Neo: gracefully handle conflicts like SHIFT vs SHIFT+G
                         if (this.hasCollision) {
                             mutablecomponent.append(", ");
                         }
