--- a/net/minecraft/world/level/material/EmptyFluid.java
+++ b/net/minecraft/world/level/material/EmptyFluid.java
@@ -2,6 +_,9 @@
 
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.BlockGetter;
@@ -13,6 +_,11 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class EmptyFluid extends Fluid {
+    public EmptyFluid() {
+        super(net.neoforged.neoforge.common.world.FluidBehaviour.Properties.of()
+            .setId(ResourceKey.create(BuiltInRegistries.FLUID.key(), ResourceLocation.withDefaultNamespace("empty"))));
+    }
+
     @Override
     public Item getBucket() {
         return Items.AIR;
