--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -164,7 +_,7 @@
             BlockState blockstate = level.getBlockState(blockpos);
             FluidState fluidstate = blockstate.getFluidState();
             if (fluidstate.getType().isSame(this) && this.canPassThroughWall(direction, level, pos, blockState, blockpos, blockstate)) {
-                if (fluidstate.isSource()) {
+                if (fluidstate.isSource() && net.neoforged.neoforge.event.EventHooks.canCreateFluidSource(level, blockpos, blockstate)) {
                     j++;
                 }
 
@@ -172,7 +_,7 @@
             }
         }
 
-        if (this.canConvertToSource(level) && j >= 2) {
+        if (j >= 2) {
             BlockState blockstate1 = level.getBlockState(pos.below());
             FluidState fluidstate1 = blockstate1.getFluidState();
             if (blockstate1.isSolid() || this.isSourceBlockOfThisType(fluidstate1)) {
@@ -238,6 +_,15 @@
         return this.getSource().defaultFluidState().setValue(FALLING, Boolean.valueOf(falling));
     }
 
+    @Override
+    public boolean canConvertToSource(FluidState state, Level level, BlockPos pos) {
+        return this.canConvertToSource(level);
+    }
+
+    /**
+     * @deprecated Forge: Use {@link #canConvertToSource(FluidState, Level, BlockPos)} instead.
+     */
+    @Deprecated
     protected abstract boolean canConvertToSource(Level level);
 
     protected void spreadTo(LevelAccessor level, BlockPos pos, BlockState blockState, Direction direction, FluidState fluidState) {
