--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -122,6 +_,8 @@
         return this.getType().builtInRegistryHolder();
     }
 
+    /// @deprecated Neo: use [#getExplosionResistance(BlockGetter, BlockPos, net.minecraft.world.level.Explosion)] instead
+    @Deprecated
     public float getExplosionResistance() {
         return this.getType().getExplosionResistance();
     }
