--- a/net/minecraft/world/level/material/Fluid.java
+++ b/net/minecraft/world/level/material/Fluid.java
@@ -23,8 +_,10 @@
 
 public abstract class Fluid implements net.neoforged.neoforge.common.extensions.IFluidExtension {
     public static final IdMapper<FluidState> FLUID_STATE_REGISTRY = new IdMapper<>();
+    public final net.neoforged.neoforge.transfer.resources.FluidResource defaultResource;
     protected final StateDefinition<Fluid, FluidState> stateDefinition;
     private FluidState defaultFluidState;
+
     private final Holder.Reference<Fluid> builtInRegistryHolder = BuiltInRegistries.FLUID.createIntrusiveHolder(this);
 
     protected Fluid() {
@@ -32,6 +_,7 @@
         this.createFluidStateDefinition(builder);
         this.stateDefinition = builder.create(Fluid::defaultFluidState, FluidState::new);
         this.registerDefaultState(this.stateDefinition.any());
+        this.defaultResource = net.neoforged.neoforge.transfer.resources.FluidResource.of(this);
     }
 
     protected void createFluidStateDefinition(StateDefinition.Builder<Fluid, FluidState> p_76121_) {
@@ -101,6 +_,13 @@
     }
 
     public abstract VoxelShape getShape(FluidState p_76137_, BlockGetter p_76138_, BlockPos p_76139_);
+
+    private net.neoforged.neoforge.fluids.FluidType forgeFluidType;
+    @Override
+    public net.neoforged.neoforge.fluids.FluidType getFluidType() {
+        if (forgeFluidType == null) forgeFluidType = net.neoforged.neoforge.common.CommonHooks.getVanillaFluidType(this);
+        return forgeFluidType;
+    }
 
     public Optional<SoundEvent> getPickupSound() {
         return Optional.empty();
