--- a/net/minecraft/world/level/material/Fluid.java
+++ b/net/minecraft/world/level/material/Fluid.java
@@ -23,14 +_,16 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import net.neoforged.neoforge.fluids.FluidStack;
 
-public abstract class Fluid implements net.neoforged.neoforge.common.extensions.IFluidExtension {
+public abstract class Fluid extends net.neoforged.neoforge.common.world.FluidBehaviour implements net.neoforged.neoforge.common.extensions.IFluidExtension {
     public static final IdMapper<FluidState> FLUID_STATE_REGISTRY = new IdMapper<>();
     protected final StateDefinition<Fluid, FluidState> stateDefinition;
     private FluidState defaultFluidState;
     private final Holder.Reference<Fluid> builtInRegistryHolder = BuiltInRegistries.FLUID.createIntrusiveHolder(this);
 
-    protected Fluid() {
+    protected Fluid(net.neoforged.neoforge.common.world.FluidBehaviour.Properties properties) {
+        super(properties);
         StateDefinition.Builder<Fluid, FluidState> builder = new StateDefinition.Builder<>(this);
         this.createFluidStateDefinition(builder);
         this.stateDefinition = builder.create(Fluid::defaultFluidState, FluidState::new);
@@ -85,8 +_,6 @@
         return false;
     }
 
-    protected abstract float getExplosionResistance();
-
     public abstract float getHeight(FluidState p_76124_, BlockGetter p_76125_, BlockPos p_76126_);
 
     public abstract float getOwnHeight(FluidState p_76123_);
@@ -125,5 +_,27 @@
     @Deprecated
     public Holder.Reference<Fluid> builtInRegistryHolder() {
         return this.builtInRegistryHolder;
+    }
+
+    @Override
+    public String toString() {// Neo: Patch in a toString similar to Item#toString so that FluidStack#toStringis more readable
+        return BuiltInRegistries.FLUID.wrapAsHolder(this).getRegisteredName();
+    }
+
+    @Override
+    public FluidState getStateForPlacement(BlockGetter level, BlockPos pos, FluidStack stack) {
+        return defaultFluidState();
+    }
+
+    @Override
+    public BlockState getBlockStateForPlacement(BlockGetter level, BlockPos pos) {
+        return defaultFluidState().createLegacyBlock();
+    }
+
+    @Override
+    public boolean shouldHideAdjacentFluidFace(Direction selfFace, FluidState adjacentFluid) {
+        return getFamily()
+            .map(adjacentFluid::is)
+            .orElse(adjacentFluid.is(this));
     }
 }
