--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -93,7 +_,7 @@
                     BlockState blockState = level.getBlockState(testPos);
                     if (blockState.isAir()) {
                         if (this.hasFlammableNeighbours(level, testPos)) {
-                            level.setBlockAndUpdate(testPos, BaseFireBlock.getState(level, testPos));
+                            level.setBlockAndUpdate(testPos, net.neoforged.neoforge.event.EventHooks.fireFluidPlaceBlockEvent(level, testPos, pos, BaseFireBlock.getState(level, testPos)));
                             return;
                         }
                     } else if (blockState.blocksMotion()) {
@@ -107,8 +_,8 @@
                         return;
                     }
 
-                    if (level.isEmptyBlock(testPos.above()) && this.isFlammable(level, testPos)) {
-                        level.setBlockAndUpdate(testPos.above(), BaseFireBlock.getState(level, testPos));
+                    if (level.isEmptyBlock(testPos.above()) && this.isFlammable(level, testPos, Direction.UP)) {
+                        level.setBlockAndUpdate(testPos.above(), net.neoforged.neoforge.event.EventHooks.fireFluidPlaceBlockEvent(level, testPos.above(), pos, BaseFireBlock.getState(level, testPos)));
                     }
                 }
             }
@@ -124,7 +_,7 @@
 
     private boolean hasFlammableNeighbours(LevelReader level, BlockPos pos) {
         for (Direction direction : Direction.values()) {
-            if (this.isFlammable(level, pos.relative(direction))) {
+            if (this.isFlammable(level, pos.relative(direction), direction.getOpposite())) {
                 return true;
             }
         }
@@ -132,10 +_,20 @@
         return false;
     }
 
+    /** @deprecated Forge: use {@link LavaFluid#isFlammable(LevelReader,BlockPos,Direction)} instead */
+    @Deprecated
     private boolean isFlammable(LevelReader level, BlockPos pos) {
         return level.isInsideBuildHeight(pos.getY()) && !level.hasChunkAt(pos) ? false : level.getBlockState(pos).ignitedByLava();
     }
 
+    private boolean isFlammable(LevelReader level, BlockPos pos, Direction face) {
+        if (level.isInsideBuildHeight(pos.getY()) && !level.hasChunkAt(pos)) {
+            return false;
+        }
+        BlockState state = level.getBlockState(pos);
+        return state.ignitedByLava(level, pos, face);
+    }
+
     @Override
     public @Nullable ParticleOptions getDripParticle() {
         return ParticleTypes.DRIPPING_LAVA;
@@ -206,7 +_,7 @@
             FluidState fluidState = level.getFluidState(pos);
             if (this.is(FluidTags.LAVA) && fluidState.is(FluidTags.WATER)) {
                 if (state.getBlock() instanceof LiquidBlock) {
-                    level.setBlock(pos, Blocks.STONE.defaultBlockState(), 3);
+                    level.setBlock(pos, net.neoforged.neoforge.event.EventHooks.fireFluidPlaceBlockEvent(level, pos, pos, Blocks.STONE.defaultBlockState()), 3);
                 }
 
                 this.fizz(level, pos);
