--- a/net/minecraft/client/resources/language/LanguageManager.java
+++ b/net/minecraft/client/resources/language/LanguageManager.java
@@ -30,11 +_,12 @@
     private final Minecraft minecraft;
     private Map<String, LanguageInfo> languages = ImmutableMap.of("en_us", DEFAULT_LANGUAGE);
     private String currentCode;
+    private java.util.Locale javaLocale; // Forge: add locale information for modders
     private final Consumer<ClientLanguage> reloadCallback;
 
     public LanguageManager(Minecraft minecraft, String languageCode, Consumer<ClientLanguage> reloadCallback) {
         this.minecraft = minecraft;
-        this.currentCode = languageCode;
+        this.setSelected(languageCode);
         this.reloadCallback = reloadCallback;
     }
 
@@ -86,8 +_,14 @@
         }
     }
 
+    public java.util.Locale getJavaLocale() {
+        return this.javaLocale;
+    }
+
     public void setSelected(String code) {
         this.currentCode = code;
+        final String[] langSplit = code.split("_", 2);
+        this.javaLocale = langSplit.length == 1 ? new java.util.Locale(langSplit[0]) : new java.util.Locale(langSplit[0], langSplit[1]);
     }
 
     public String getSelected() {
