--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -700,13 +_,18 @@
 
     @Override
     protected void dropAllDeathLoot(ServerLevel level, DamageSource source) {
+        super.dropAllDeathLoot(level, source);
+    }
+
+    @Override
+    protected void dropEquipment(ServerLevel level) { // Forge: move extra drops to dropEquipment to allow them to be captured by LivingDropsEvent
+        super.dropEquipment(level);
         ItemStack itemStack = this.getItemBySlot(EquipmentSlot.MAINHAND);
         if (!itemStack.isEmpty()) {
             this.spawnAtLocation(level, itemStack);
             this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
         }
 
-        super.dropAllDeathLoot(level, source);
     }
 
     public static boolean isPathClear(Fox fox, LivingEntity target) {
@@ -873,6 +_,17 @@
         @Override
         protected void breed() {
             Fox offspring = (Fox)this.animal.getBreedOffspring(this.level, this.partner);
+            final net.neoforged.neoforge.event.entity.living.BabyEntitySpawnEvent event = new net.neoforged.neoforge.event.entity.living.BabyEntitySpawnEvent(animal, partner, offspring);
+            final boolean cancelled = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled();
+            offspring = (Fox) event.getChild();
+            if (cancelled) {
+                //Reset the "inLove" state for the animals
+                this.animal.setAge(6000);
+                this.partner.setAge(6000);
+                this.animal.resetLove();
+                this.partner.resetLove();
+                return;
+            }
             if (offspring != null) {
                 ServerPlayer animalLoveCause = this.animal.getLoveCause();
                 ServerPlayer partnerLoveCause = this.partner.getLoveCause();
@@ -951,7 +_,7 @@
         }
 
         protected void onReachedTarget() {
-            if (getServerLevel(Fox.this.level()).getGameRules().get(GameRules.MOB_GRIEFING)) {
+            if (net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(Fox.this.level()), Fox.this)) {
                 BlockState state = Fox.this.level().getBlockState(this.blockPos);
                 if (state.is(Blocks.SWEET_BERRY_BUSH)) {
                     this.pickSweetBerries(state);
