--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -238,6 +_,7 @@
     }
 
     protected void doUnderWaterConversion(ServerLevel level) {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.DROWNED, (timer) -> this.conversionTime = timer)) return;
         this.convertToZombieType(level, EntityType.DROWNED);
         if (!this.isSilent()) {
             level.levelEvent(null, 1040, this.blockPosition(), 0);
@@ -248,9 +_,10 @@
         this.convertTo(
             zombieType,
             ConversionParams.single(this, true, true),
-            newZombie -> newZombie.handleAttributes(
-                level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION
-            )
+            newZombie -> {
+                newZombie.handleAttributes(level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION);
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, newZombie);
+            }
         );
     }
 
@@ -267,6 +_,7 @@
                 zombie.setGossips(villager.getGossips().copy());
                 zombie.setTradeOffers(villager.getOffers().copy());
                 zombie.setVillagerXp(villager.getVillagerXp());
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(villager, zombie);
                 if (!this.isSilent()) {
                     level.levelEvent(null, 1026, this.blockPosition(), 0);
                 }
@@ -424,7 +_,7 @@
     @Override
     public boolean killedEntity(ServerLevel level, LivingEntity entity, DamageSource source) {
         boolean perished = super.killedEntity(level, entity, source);
-        if ((level.getDifficulty() == Difficulty.NORMAL || level.getDifficulty() == Difficulty.HARD) && entity instanceof Villager villager) {
+        if ((level.getDifficulty() == Difficulty.NORMAL || level.getDifficulty() == Difficulty.HARD) && entity instanceof Villager villager && net.neoforged.neoforge.event.EventHooks.canLivingConvert(entity, EntityType.ZOMBIE_VILLAGER, (timer) -> {})) {
             if (level.getDifficulty() != Difficulty.HARD && this.random.nextBoolean()) {
                 return perished;
             }
