--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -293,6 +_,8 @@
         } catch (Throwable t) {
             CrashReport report = CrashReport.forThrowable(t, "Adding item to inventory");
             CrashReportCategory category = report.addCategory("Item being added");
+            category.setDetail("Registry Name", () -> String.valueOf(net.minecraft.core.registries.BuiltInRegistries.ITEM.getKey(itemStack.getItem())));
+            category.setDetail("Item Class", () -> itemStack.getItem().getClass().getName());
             category.setDetail("Item ID", Item.getId(itemStack.getItem()));
             category.setDetail("Item data", itemStack.getDamageValue());
             category.setDetail("Item name", () -> itemStack.getHoverName().getString());
