--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -137,7 +137,7 @@
 
       for(int k = 0; k < 9; ++k) {
          int l = (this.selected + k) % 9;
-         if (!this.items.get(l).isEnchanted()) {
+         if (!this.items.get(l).isNotReplaceableByPickAction(this.player, l)) {
             return l;
          }
       }
@@ -186,7 +186,8 @@
       int i = p_36049_.getCount();
       ItemStack itemstack = this.getItem(p_36048_);
       if (itemstack.isEmpty()) {
-         itemstack = new ItemStack(item, 0);
+         itemstack = p_36049_.copy(); // Forge: Replace Item clone above to preserve item capabilities when picking the item up.
+         itemstack.setCount(0);
          if (p_36049_.hasTag()) {
             itemstack.setTag(p_36049_.getTag().copy());
          }
@@ -237,6 +238,7 @@
             }
          }
       }
+      armor.forEach(e -> e.onArmorTick(player.level(), player));
    }
 
    public boolean add(ItemStack p_36055_) {
@@ -284,6 +286,8 @@
          } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Adding item to inventory");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Item being added");
+            crashreportcategory.setDetail("Registry Name", () -> String.valueOf(net.neoforged.neoforge.registries.ForgeRegistries.ITEMS.getKey(p_36042_.getItem())));
+            crashreportcategory.setDetail("Item Class", () -> p_36042_.getItem().getClass().getName());
             crashreportcategory.setDetail("Item ID", Item.getId(p_36042_.getItem()));
             crashreportcategory.setDetail("Item data", p_36042_.getDamageValue());
             crashreportcategory.setDetail("Item name", () -> p_36042_.getHoverName().getString());
