--- a/net/minecraft/world/item/trading/MerchantOffer.java
+++ b/net/minecraft/world/item/trading/MerchantOffer.java
@@ -107,7 +_,7 @@
     private int getModifiedCostCount(ItemCost p_330556_) {
         int i = p_330556_.count();
         int j = Math.max(0, Mth.floor((float)(i * this.demand) * this.priceMultiplier));
-        return Mth.clamp(i + j + this.specialPriceDiff, 1, p_330556_.item().value().getMaxStackSize());
+        return Mth.clamp(i + j + this.specialPriceDiff, 1, p_330556_.itemStack().getMaxStackSize());
     }
 
     public ItemStack getCostB() {
