--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -116,10 +123,10 @@
         super.loadAdditional(input);
         this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
         ContainerHelper.loadAllItems(input, this.items);
-        this.cookingTimer = input.getShortOr("cooking_time_spent", (short)0);
-        this.cookingTotalTime = input.getShortOr("cooking_total_time", (short)0);
-        this.litTimeRemaining = input.getShortOr("lit_time_remaining", (short)0);
-        this.litTotalTime = input.getShortOr("lit_total_time", (short)0);
+        this.cookingTimer = input.getIntOr("cooking_time_spent", (short)0);
+        this.cookingTotalTime = input.getIntOr("cooking_total_time", (short)0);
+        this.litTimeRemaining = input.getIntOr("lit_time_remaining", (short)0);
+        this.litTotalTime = input.getIntOr("lit_total_time", (short)0);
         this.recipesUsed.clear();
         this.recipesUsed.putAll(input.read("RecipesUsed", RECIPES_USED_CODEC).orElse(Map.of()));
     }
@@ -127,10 +134,10 @@
     @Override
     protected void saveAdditional(ValueOutput output) {
         super.saveAdditional(output);
-        output.putShort("cooking_time_spent", (short)this.cookingTimer);
-        output.putShort("cooking_total_time", (short)this.cookingTotalTime);
-        output.putShort("lit_time_remaining", (short)this.litTimeRemaining);
-        output.putShort("lit_total_time", (short)this.litTotalTime);
+        output.putInt("cooking_time_spent", this.cookingTimer);
+        output.putInt("cooking_total_time", this.cookingTotalTime);
+        output.putInt("lit_time_remaining", this.litTimeRemaining);
+        output.putInt("lit_total_time", this.litTotalTime);
         ContainerHelper.saveAllItems(output, this.items);
         output.store("RecipesUsed", RECIPES_USED_CODEC, this.recipesUsed);
     }
@@ -245,7 +253,7 @@
     }
 
     protected int getBurnDuration(FuelValues fuelValues, ItemStack itemStack) {
-        return fuelValues.burnDuration(itemStack);
+        return itemStack.getBurnTime(this.recipeType, fuelValues);
     }
 
     private static int getTotalCookTime(ServerLevel level, AbstractFurnaceBlockEntity entity) {
