--- a/net/minecraft/world/entity/vehicle/minecart/MinecartCommandBlock.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartCommandBlock.java
@@ -136,5 +136,12 @@
         public boolean isValid() {
             return !MinecartCommandBlock.this.isRemoved();
         }
+
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return (org.bukkit.craftbukkit.entity.CraftMinecartCommand) MinecartCommandBlock.this.getBukkitEntity();
+        }
+        // CraftBukkit end
     }
 }
