--- a/net/minecraft/client/gui/components/AbstractButton.java
+++ b/net/minecraft/client/gui/components/AbstractButton.java
@@ -14,7 +14,7 @@
 @OnlyIn(Dist.CLIENT)
 public abstract class AbstractButton extends AbstractWidget {
    protected static final int TEXT_MARGIN = 2;
-   private static final WidgetSprites SPRITES = new WidgetSprites(
+   protected static final WidgetSprites SPRITES = new WidgetSprites(
       new ResourceLocation("widget/button"), new ResourceLocation("widget/button_disabled"), new ResourceLocation("widget/button_highlighted")
    );
 
@@ -32,7 +32,7 @@
       RenderSystem.enableDepthTest();
       p_281670_.blitSprite(SPRITES.get(this.active, this.isHoveredOrFocused()), this.getX(), this.getY(), this.getWidth(), this.getHeight());
       p_281670_.setColor(1.0F, 1.0F, 1.0F, 1.0F);
-      int i = this.active ? 16777215 : 10526880;
+      int i = getFGColor();
       this.renderString(p_281670_, minecraft.font, i | Mth.ceil(this.alpha * 255.0F) << 24);
    }
 
