--- a/net/minecraft/client/gui/components/AbstractButton.java
+++ b/net/minecraft/client/gui/components/AbstractButton.java
@@ -40,7 +_,13 @@
     protected abstract void extractContents(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float a);
 
     protected void extractDefaultLabel(ActiveTextCollector output) {
-        this.extractScrollingStringOverContents(output, this.getMessage(), 2);
+        // Neo: Apply forced foreground color if set
+        var message = this.getMessage();
+        if (getFGColor() != UNSET_FG_COLOR) {
+            message = message.copy().withStyle(style -> style.withColor(getFGColor()));
+        }
+
+        this.extractScrollingStringOverContents(output, message, 2);
     }
 
     protected final void extractDefaultSprite(GuiGraphicsExtractor graphics) {
