--- a/net/minecraft/client/gui/components/AbstractWidget.java
+++ b/net/minecraft/client/gui/components/AbstractWidget.java
@@ -25,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public abstract class AbstractWidget implements Renderable, GuiEventListener, LayoutElement, NarratableEntry {
+public abstract class AbstractWidget implements Renderable, GuiEventListener, LayoutElement, NarratableEntry, net.neoforged.neoforge.client.extensions.IAbstractWidgetExtension {
     private static final double PERIOD_PER_SCROLLED_PIXEL = 0.5;
     private static final double MIN_SCROLL_PERIOD = 3.0;
     protected int width;
@@ -123,6 +_,8 @@
         renderScrollingString(p_281857_, p_282790_, this.getMessage(), i, this.getY(), j, this.getY() + this.getHeight(), p_282944_);
     }
 
+    /** @deprecated Neo: Use {@link #onClick(double, double, int)} instead. */
+    @Deprecated
     public void onClick(double p_93634_, double p_93635_) {
     }
 
@@ -139,7 +_,7 @@
                 boolean flag = this.clicked(p_93641_, p_93642_);
                 if (flag) {
                     this.playDownSound(Minecraft.getInstance().getSoundManager());
-                    this.onClick(p_93641_, p_93642_);
+                    this.onClick(p_93641_, p_93642_, p_93643_);
                     return true;
                 }
             }
@@ -253,6 +_,19 @@
     @Override
     public void setFocused(boolean p_93693_) {
         this.focused = p_93693_;
+    }
+
+    public static final int UNSET_FG_COLOR = -1;
+    protected int packedFGColor = UNSET_FG_COLOR;
+    public int getFGColor() {
+        if (packedFGColor != UNSET_FG_COLOR) return packedFGColor;
+        return this.active ? 16777215 : 10526880; // White : Light Grey
+    }
+    public void setFGColor(int color) {
+        this.packedFGColor = color;
+    }
+    public void clearFGColor() {
+        this.packedFGColor = UNSET_FG_COLOR;
     }
 
     @Override
