--- a/net/minecraft/client/gui/components/AbstractWidget.java
+++ b/net/minecraft/client/gui/components/AbstractWidget.java
@@ -225,6 +_,32 @@
         this.focused = focused;
     }
 
+    // Neo Begin
+    // Support for custom forced text colors. If the color is not set, it will use the vanilla defaults (white when active, light grey when inactive).
+
+    /// Neo: Stores the label text color override. See [#getLabelTextColor()] and [#setLabelTextColor].
+    protected @Nullable Integer labelTextColor = null;
+
+    /// Neo: {@return the widget's label text color} If a custom text color has been set, that color will be used.
+    /// If a custom text color has not been set, [net.minecraft.util.CommonColors#WHITE] will be used when active,
+    /// and [net.minecraft.util.CommonColors#LIGHT_GRAY] otherwise.
+    ///
+    /// Subclasses may override this to provide more advanced text color handling.
+    public int getLabelTextColor() {
+        return this.labelTextColor != null
+                ? this.labelTextColor
+                : this.active ? net.minecraft.util.CommonColors.WHITE : net.minecraft.util.CommonColors.LIGHT_GRAY;
+    }
+
+    /// Neo: Sets the label text color for this widget based on the given ARGB value
+    ///
+    /// @param color the new label text color as an ARGB value
+    public void setLabelTextColor(int color) {
+        this.labelTextColor = color;
+    }
+
+    // Neo End
+
     @Override
     public NarratableEntry.NarrationPriority narrationPriority() {
         if (this.isFocused()) {
