--- a/net/minecraft/client/gui/components/AbstractWidget.java
+++ b/net/minecraft/client/gui/components/AbstractWidget.java
@@ -25,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public abstract class AbstractWidget implements Renderable, GuiEventListener, LayoutElement, NarratableEntry {
+public abstract class AbstractWidget implements Renderable, GuiEventListener, LayoutElement, NarratableEntry, net.neoforged.neoforge.client.extensions.IAbstractWidgetExtension {
     private static final double PERIOD_PER_SCROLLED_PIXEL = 0.5;
     private static final double MIN_SCROLL_PERIOD = 3.0;
     protected int width;
@@ -91,13 +_,25 @@
     protected abstract void renderWidget(GuiGraphics p_282139_, int p_268034_, int p_268009_, float p_268085_);
 
     public static void renderScrollingString(
+            GuiGraphics p_281620_, Font p_282651_, Component p_281467_, int p_283621_, int p_282084_, int p_283398_, int p_281938_, int p_283471_, boolean dropShadow
+    ) {
+        renderScrollingString(p_281620_, p_282651_, p_281467_, (p_283621_ + p_283398_) / 2, p_283621_, p_282084_, p_283398_, p_281938_, p_283471_, dropShadow);
+    }
+
+    public static void renderScrollingString(
+            GuiGraphics p_296355_, Font p_295317_, Component p_294875_, int p_294289_, int p_295475_, int p_294243_, int p_296428_, int p_294696_, int p_295687_
+    ){
+        renderScrollingString(p_296355_, p_295317_, p_294875_, p_294289_, p_295475_, p_294243_, p_296428_, p_294696_, p_295687_, true);
+    }
+
+    public static void renderScrollingString(
         GuiGraphics p_281620_, Font p_282651_, Component p_281467_, int p_283621_, int p_282084_, int p_283398_, int p_281938_, int p_283471_
     ) {
-        renderScrollingString(p_281620_, p_282651_, p_281467_, (p_283621_ + p_283398_) / 2, p_283621_, p_282084_, p_283398_, p_281938_, p_283471_);
+        renderScrollingString(p_281620_, p_282651_, p_281467_, (p_283621_ + p_283398_) / 2, p_283621_, p_282084_, p_283398_, p_281938_, p_283471_, true);
     }
 
     public static void renderScrollingString(
-        GuiGraphics p_296355_, Font p_295317_, Component p_294875_, int p_294289_, int p_295475_, int p_294243_, int p_296428_, int p_294696_, int p_295687_
+        GuiGraphics p_296355_, Font p_295317_, Component p_294875_, int p_294289_, int p_295475_, int p_294243_, int p_296428_, int p_294696_, int p_295687_, boolean dropShadow
     ) {
         int i = p_295317_.width(p_294875_);
         int j = (p_294243_ + p_294696_ - 9) / 2 + 1;
@@ -109,11 +_,11 @@
             double d2 = Math.sin((Math.PI / 2) * Math.cos((Math.PI * 2) * d0 / d1)) / 2.0 + 0.5;
             double d3 = Mth.lerp(d2, 0.0, (double)l);
             p_296355_.enableScissor(p_295475_, p_294243_, p_296428_, p_294696_);
-            p_296355_.drawString(p_295317_, p_294875_, p_295475_ - (int)d3, j, p_295687_);
+            p_296355_.drawString(p_295317_, p_294875_, p_295475_ - (int)d3, j, p_295687_, dropShadow);
             p_296355_.disableScissor();
         } else {
             int i1 = Mth.clamp(p_294289_, p_295475_ + i / 2, p_296428_ - i / 2);
-            p_296355_.drawCenteredString(p_295317_, p_294875_, i1, j, p_295687_);
+            p_296355_.drawCenteredString(p_295317_, p_294875_, i1, j, p_295687_, dropShadow);
         }
     }
 
@@ -123,6 +_,8 @@
         renderScrollingString(p_281857_, p_282790_, this.getMessage(), i, this.getY(), j, this.getY() + this.getHeight(), p_282944_);
     }
 
+    /** @deprecated Neo: Use {@link #onClick(double, double, int)} instead. */
+    @Deprecated
     public void onClick(double p_93634_, double p_93635_) {
     }
 
@@ -139,7 +_,7 @@
                 boolean flag = this.clicked(p_93641_, p_93642_);
                 if (flag) {
                     this.playDownSound(Minecraft.getInstance().getSoundManager());
-                    this.onClick(p_93641_, p_93642_);
+                    this.onClick(p_93641_, p_93642_, p_93643_);
                     return true;
                 }
             }
@@ -253,6 +_,19 @@
     @Override
     public void setFocused(boolean p_93693_) {
         this.focused = p_93693_;
+    }
+
+    public static final int UNSET_FG_COLOR = -1;
+    protected int packedFGColor = UNSET_FG_COLOR;
+    public int getFGColor() {
+        if (packedFGColor != UNSET_FG_COLOR) return packedFGColor;
+        return this.active ? 16777215 : 10526880; // White : Light Grey
+    }
+    public void setFGColor(int color) {
+        this.packedFGColor = color;
+    }
+    public void clearFGColor() {
+        this.packedFGColor = UNSET_FG_COLOR;
     }
 
     @Override
