--- a/net/minecraft/client/gui/components/Button.java
+++ b/net/minecraft/client/gui/components/Button.java
@@ -33,6 +_,11 @@
         this.createNarration = p_259552_;
     }
 
+    protected Button(Builder builder) {
+        this(builder.x, builder.y, builder.width, builder.height, builder.message, builder.onPress, builder.createNarration);
+        setTooltip(builder.tooltip); // Forge: Make use of the Builder tooltip
+    }
+
     @Override
     public void onPress(InputWithModifiers p_446034_) {
         this.onPress.onPress(this);
@@ -100,6 +_,10 @@
             button.setTooltip(this.tooltip);
             return button;
         }
+
+        public Button build(java.util.function.Function<Builder, Button> builder) {
+            return builder.apply(this);
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
@@ -118,6 +_,10 @@
             int p_457831_, int p_458177_, int p_458246_, int p_457590_, Component p_457882_, Button.OnPress p_457728_, Button.CreateNarration p_458093_
         ) {
             super(p_457831_, p_458177_, p_458246_, p_457590_, p_457882_, p_457728_, p_458093_);
+        }
+
+        protected Plain(Builder builder) {
+            super(builder);
         }
 
         @Override
