--- a/net/minecraft/client/gui/components/DebugScreenOverlay.java
+++ b/net/minecraft/client/gui/components/DebugScreenOverlay.java
@@ -134,8 +_,13 @@
         this.block = entity.pick(20.0, 0.0F, false);
         this.liquid = entity.pick(20.0, 0.0F, true);
         p_281427_.drawManaged(() -> {
-            this.drawGameInformation(p_281427_);
-            this.drawSystemInformation(p_281427_);
+            final List<String> gameInformation = new java.util.ArrayList<>();
+            final List<String> systemInformation = new java.util.ArrayList<>();
+            var event = new net.neoforged.neoforge.client.event.CustomizeGuiOverlayEvent.DebugText(minecraft.getWindow(), p_281427_, minecraft.getFrameTime(), gameInformation, systemInformation);
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event);
+
+            this.drawGameInformation(p_281427_, gameInformation);
+            this.drawSystemInformation(p_281427_, systemInformation);
             if (this.renderFpsCharts) {
                 int i = p_281427_.guiWidth();
                 int j = i / 2;
@@ -160,7 +_,7 @@
         this.minecraft.getProfiler().pop();
     }
 
-    protected void drawGameInformation(GuiGraphics p_281525_) {
+    protected void drawGameInformation(GuiGraphics p_281525_, List<String> gameInformation) {
         List<String> list = this.getGameInformation();
         list.add("");
         boolean flag = this.minecraft.getSingleplayerServer() != null;
@@ -175,11 +_,13 @@
                 + (this.renderNetworkCharts ? " visible" : " hidden")
         );
         list.add("For help: press F3 + Q");
+        list.addAll(gameInformation);
         this.renderLines(p_281525_, list, true);
     }
 
-    protected void drawSystemInformation(GuiGraphics p_281261_) {
+    protected void drawSystemInformation(GuiGraphics p_281261_, List<String> systemInformation) {
         List<String> list = this.getSystemInformation();
+        list.addAll(systemInformation);
         this.renderLines(p_281261_, list, false);
     }
 
@@ -508,6 +_,13 @@
             GlUtil.getRenderer(),
             GlUtil.getOpenGLVersion()
         );
+        if (this.minecraft.isDemo()) {
+            if (this.minecraft.level.getGameTime() >= 120500L) {
+                list.add(0, net.minecraft.network.chat.Component.translatable("demo.demoExpired").getString());
+            } else {
+                list.add(0, net.minecraft.network.chat.Component.translatable("demo.remainingTime", net.minecraft.util.StringUtil.formatTickDuration((int)(120500L - this.minecraft.level.getGameTime()), this.minecraft.level.tickRateManager().tickrate())).getString());
+            }
+        }
         if (this.minecraft.showOnlyReducedInfo()) {
             return list;
         } else {
@@ -544,6 +_,7 @@
                 list.add("");
                 list.add(ChatFormatting.UNDERLINE + "Targeted Entity");
                 list.add(String.valueOf(BuiltInRegistries.ENTITY_TYPE.getKey(entity.getType())));
+                entity.getType().builtInRegistryHolder().tags().forEach(t -> list.add("#" + t.location()));
             }
 
             return list;
