--- a/net/minecraft/client/multiplayer/chat/ChatListener.java
+++ b/net/minecraft/client/multiplayer/chat/ChatListener.java
@@ -118,7 +_,9 @@
         Instant instant = Instant.now();
         this.handleMessage(null, () -> {
             Component component = boundChatType.decorate(message);
-            this.minecraft.gui.getChat().addMessage(component);
+            Component forgeComponent = net.neoforged.neoforge.client.ClientHooks.onClientChat(boundChatType, component, Util.NIL_UUID);
+            if (forgeComponent == null) return false;
+            this.minecraft.gui.getChat().addMessage(forgeComponent);
             this.narrateChatMessage(boundChatType, message);
             this.logSystemMessage(component, instant);
             this.previousMessageTime = Util.getMillis();
@@ -137,12 +_,16 @@
             MessageSignature messagesignature = chatMessage.signature();
             FilterMask filtermask = chatMessage.filterMask();
             if (filtermask.isEmpty()) {
-                this.minecraft.gui.getChat().addMessage(decoratedServerContent, messagesignature, guimessagetag);
+                Component forgeComponent = net.neoforged.neoforge.client.ClientHooks.onClientPlayerChat(boundChatType, decoratedServerContent, chatMessage, chatMessage.sender());
+                if (forgeComponent == null) return false;
+                this.minecraft.gui.getChat().addMessage(forgeComponent, messagesignature, guimessagetag);
                 this.narrateChatMessage(boundChatType, chatMessage.decoratedContent());
             } else {
                 Component component = filtermask.applyWithFormatting(chatMessage.signedContent());
                 if (component != null) {
-                    this.minecraft.gui.getChat().addMessage(boundChatType.decorate(component), messagesignature, guimessagetag);
+                    Component forgeComponent = net.neoforged.neoforge.client.ClientHooks.onClientPlayerChat(boundChatType, boundChatType.decorate(component), chatMessage, chatMessage.sender());
+                    if (forgeComponent == null) return false;
+                    this.minecraft.gui.getChat().addMessage(forgeComponent, messagesignature, guimessagetag);
                     this.narrateChatMessage(boundChatType, component);
                 }
             }
@@ -175,6 +_,8 @@
 
     public void handleSystemMessage(Component message, boolean isOverlay) {
         if (!this.minecraft.options.hideMatchedNames().get() || !this.minecraft.isBlocked(this.guessChatUUID(message))) {
+            message = net.neoforged.neoforge.client.ClientHooks.onClientSystemChat(message, isOverlay);
+            if (message == null) return;
             if (isOverlay) {
                 this.minecraft.gui.setOverlayMessage(message, false);
             } else {
