--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -78,6 +_,8 @@
     @Nullable
     private RemoteSampleLogger tickTimeLogger;
     private final ServerLinks serverLinks;
+    @Nullable
+    private net.minecraft.client.server.LanServerPinger dediLanPinger;
 
     public DedicatedServer(
         Thread p_214789_, LevelStorageSource.LevelStorageAccess p_214790_, DedicatedServerSettings p_214793_, DataFixer p_214794_, Services p_214795_
@@ -94,6 +_,7 @@
         Thread thread = new Thread("Server console handler") {
             @Override
             public void run() {
+                if (net.neoforged.neoforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
                 BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
 
                 String s;
@@ -187,6 +_,11 @@
                 LOGGER.info("JMX monitoring enabled");
             }
 
+            if (net.neoforged.neoforge.common.config.NeoForgeServerConfig.INSTANCE.advertiseDedicatedServerToLan.get()) {
+                this.dediLanPinger = new net.minecraft.client.server.LanServerPinger(this.getMotd(), String.valueOf(this.getServerPort()));
+                this.dediLanPinger.start();
+            }
+
             return true;
         }
     }
@@ -208,6 +_,7 @@
         long j = Util.getNanos() - i;
         String s = String.format(Locale.ROOT, "%.3fs", j / 1.0E9);
         LOGGER.info("Done ({})! For help, type \"help\"", s);
+        this.nextTickTimeNanos = Util.getNanos(); // Neo: Update server time to prevent watchdog/spaming during long load.
         if (dedicatedserverproperties.announcePlayerAchievements != null) {
             thegame.getGameRules().getRule(GameRules.RULE_ANNOUNCE_ADVANCEMENTS).set(dedicatedserverproperties.announcePlayerAchievements, thegame);
         }
@@ -281,6 +_,13 @@
         if (this.queryThreadGs4 != null) {
             this.queryThreadGs4.stop();
         }
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
+
+        // Neo: Forcibly stop the FileWatcher default instance, to prevent it from blocking normal JVM exit; see #1626
+        com.electronwill.nightconfig.core.file.FileWatcher.defaultInstance().stop();
     }
 
     @Override
@@ -530,7 +_,12 @@
 
     @Override
     public void stopServer(TheGame p_421148_) {
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         super.stopServer(p_421148_);
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
         Util.shutdownExecutors();
         SkullBlockEntity.clear();
     }
