--- a/net/minecraft/server/dedicated/Settings.java
+++ b/net/minecraft/server/dedicated/Settings.java
@@ -66,7 +_,7 @@
 
     public void store(Path output) {
         try (Writer os = Files.newBufferedWriter(output, StandardCharsets.UTF_8)) {
-            this.properties.store(os, "Minecraft server properties");
+            net.neoforged.neoforge.common.util.SortedProperties.store(this.properties, os, "Minecraft server properties");
         } catch (IOException var7) {
             LOGGER.error("Failed to store properties to file: {}", output);
         }
