--- a/net/minecraft/network/protocol/common/ClientboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/common/ClientboundCustomPayloadPacket.java
@@ -53,12 +_,16 @@
                 new CustomPacketPayload.TypeAndCodec<>(StructuresDebugPayload.TYPE, StructuresDebugPayload.STREAM_CODEC),
                 new CustomPacketPayload.TypeAndCodec<>(VillageSectionsDebugPayload.TYPE, VillageSectionsDebugPayload.STREAM_CODEC),
                 new CustomPacketPayload.TypeAndCodec<>(WorldGenAttemptDebugPayload.TYPE, WorldGenAttemptDebugPayload.STREAM_CODEC)
-            )
+            ),
+            net.minecraft.network.ConnectionProtocol.PLAY,
+            net.minecraft.network.protocol.PacketFlow.CLIENTBOUND
         )
         .map(ClientboundCustomPayloadPacket::new, ClientboundCustomPayloadPacket::payload);
     public static final StreamCodec<FriendlyByteBuf, ClientboundCustomPayloadPacket> CONFIG_STREAM_CODEC = CustomPacketPayload.<FriendlyByteBuf>codec(
             p_320228_ -> DiscardedPayload.codec(p_320228_, 1048576),
-            List.of(new CustomPacketPayload.TypeAndCodec<>(BrandPayload.TYPE, BrandPayload.STREAM_CODEC))
+            List.of(new CustomPacketPayload.TypeAndCodec<>(BrandPayload.TYPE, BrandPayload.STREAM_CODEC)),
+            net.minecraft.network.ConnectionProtocol.CONFIGURATION,
+            net.minecraft.network.protocol.PacketFlow.CLIENTBOUND
         )
         .map(ClientboundCustomPayloadPacket::new, ClientboundCustomPayloadPacket::payload);
 
