--- a/net/minecraft/network/protocol/common/ClientboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/common/ClientboundCustomPayloadPacket.java
@@ -2,6 +_,9 @@
 
 import com.google.common.collect.ImmutableMap;
 import java.util.Map;
+
+import io.netty.channel.ChannelHandlerContext;
+import net.minecraft.network.ConnectionProtocol;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.common.custom.BeeDebugPayload;
@@ -29,7 +_,7 @@
 
 public record ClientboundCustomPayloadPacket(CustomPacketPayload payload) implements Packet<ClientCommonPacketListener> {
     private static final int MAX_PAYLOAD_SIZE = 1048576;
-    private static final Map<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>> KNOWN_TYPES = ImmutableMap.<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>>builder(
+    public static final Map<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>> KNOWN_TYPES = ImmutableMap.<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>>builder(
             
         )
         .put(BrandPayload.ID, BrandPayload::new)
@@ -52,16 +_,53 @@
         .put(VillageSectionsDebugPayload.ID, VillageSectionsDebugPayload::new)
         .put(WorldGenAttemptDebugPayload.ID, WorldGenAttemptDebugPayload::new)
         .build();
-
+    
+    /**
+     * @deprecated Use {@link #ClientboundCustomPayloadPacket(FriendlyByteBuf, ChannelHandlerContext, ConnectionProtocol)} instead, this variant can only read vanilla payloads.
+     * @param p_295835_ The buffer to read from
+     */
+    @Deprecated()
     public ClientboundCustomPayloadPacket(FriendlyByteBuf p_295835_) {
         this(readPayload(p_295835_.readResourceLocation(), p_295835_));
     }
-
+    
+    /**
+     * Reads a custom payload packet from the given buffer.
+     *
+     * @param p_295835_ The buffer to read from
+     * @param context The context of the channel in which this packet is read
+     * @param protocol The protocol of the connection
+     */
+    public ClientboundCustomPayloadPacket(FriendlyByteBuf p_295835_, ChannelHandlerContext context, ConnectionProtocol protocol) {
+        this(readPayload(p_295835_.readResourceLocation(), p_295835_, context, protocol));
+    }
+    
+    /**
+     * Reads the payload from the given buffer.
+     *
+     * @param p_294367_ The id of the payload
+     * @param p_294321_ The buffer to read from
+     * @param context The context of the channel in which this packet is read
+     * @param protocol The protocol of the connection
+     * @return The payload
+     */
+    private static CustomPacketPayload readPayload(ResourceLocation p_294367_, FriendlyByteBuf p_294321_, ChannelHandlerContext context, ConnectionProtocol protocol) {
+        FriendlyByteBuf.Reader<? extends CustomPacketPayload> reader = net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().getReader(p_294367_, context, protocol, KNOWN_TYPES);
+        return (CustomPacketPayload)(reader != null ? reader.apply(p_294321_) : readUnknownPayload(p_294367_, p_294321_));
+    }
+    
+    /**
+     * @deprecated Use {@link #readPayload(ResourceLocation, FriendlyByteBuf, ChannelHandlerContext, ConnectionProtocol)} instead, this variant can only read vanilla payloads.
+     * @param p_294802_ The id of the payload
+     * @param p_294886_ The buffer to read from
+     * @return The payload
+     */
+    @Deprecated
     private static CustomPacketPayload readPayload(ResourceLocation p_294802_, FriendlyByteBuf p_294886_) {
         FriendlyByteBuf.Reader<? extends CustomPacketPayload> reader = KNOWN_TYPES.get(p_294802_);
-        return (CustomPacketPayload)(reader != null ? reader.apply(p_294886_) : readUnknownPayload(p_294802_, p_294886_));
+        return reader != null ? reader.apply(p_294886_) : readUnknownPayload(p_294802_, p_294886_);
     }
-
+    
     private static DiscardedPayload readUnknownPayload(ResourceLocation p_295991_, FriendlyByteBuf p_295803_) {
         int i = p_295803_.readableBytes();
         if (i >= 0 && i <= 1048576) {
