--- a/net/minecraft/network/protocol/common/ServerCommonPacketListener.java
+++ b/net/minecraft/network/protocol/common/ServerCommonPacketListener.java
@@ -2,7 +_,7 @@
 
 import net.minecraft.network.protocol.game.ServerPacketListener;
 
-public interface ServerCommonPacketListener extends ServerPacketListener {
+public interface ServerCommonPacketListener extends ServerPacketListener, net.neoforged.neoforge.common.extensions.IClientboundPacketSenderExtension {
     void handleKeepAlive(ServerboundKeepAlivePacket p_296457_);
 
     void handlePong(ServerboundPongPacket p_294309_);
