--- a/com/mojang/blaze3d/shaders/ShaderType.java
+++ b/com/mojang/blaze3d/shaders/ShaderType.java
@@ -9,6 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 @DontObfuscate
+@net.neoforged.neoforge.internal.NonExhaustiveEnum(reason = "Compute shader support will be later added, Task and Mesh shaders may also be later defined")
 public enum ShaderType {
     VERTEX("vertex", ".vsh"),
     FRAGMENT("fragment", ".fsh");
