--- a/com/mojang/blaze3d/shaders/UniformType.java
+++ b/com/mojang/blaze3d/shaders/UniformType.java
@@ -4,6 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
+@net.neoforged.neoforge.internal.NonExhaustiveEnum(reason = "Further uniform types such as SSBO and StorageTexelBuffer may be added")
 public enum UniformType {
     UNIFORM_BUFFER("ubo"),
     TEXEL_BUFFER("utb");
