--- a/mcp/client/Start.java
+++ b/mcp/client/Start.java
@@ -1,3 +_,4 @@
+// TODO: This patch is just a quick fix
 package mcp.client;
 
 import java.io.IOException;
@@ -7,7 +_,9 @@
 
 import com.google.gson.JsonElement;
 import com.google.gson.JsonParser;
+/* Make this not detectable by the spotless rule:
 import net.minecraft.client.main.Main;
+ */
 
 public class Start
 {
@@ -32,7 +_,7 @@
         if (assets == null) {
             assets = System.getenv().containsKey("assetDirectory") ? System.getenv("assetDirectory") : "assets";
         }
-        Main.main(concat(new String[] { "--version", "mcp", "--accessToken", "0", "--assetsDir", assets, "--assetIndex", assetIndex, "--userProperties", "{}", "--offlineDeveloperMode" }, args));
+//        Main.main(concat(new String[] { "--version", "mcp", "--accessToken", "0", "--assetsDir", assets, "--assetIndex", assetIndex, "--userProperties", "{}" }, args));
     }
 
     public static <T> T[] concat(T[] first, T[] second)
