--- a/net/minecraft/client/Camera.java
+++ b/net/minecraft/client/Camera.java
@@ -56,7 +_,7 @@
                 this.setRotation(this.yRot + 180.0F, -this.xRot);
             }
 
-            this.move(-this.getMaxZoom(4.0), 0.0, 0.0);
+            this.move(-this.getMaxZoom(net.neoforged.neoforge.client.ClientHooks.getDetachedCameraDistance(this, p_90579_)), 0.0, 0.0);
         } else if (p_90577_ instanceof LivingEntity && ((LivingEntity)p_90577_).isSleeping()) {
             Direction direction = ((LivingEntity)p_90577_).getBedOrientation();
             this.setRotation(direction != null ? direction.toYRot() - 180.0F : 0.0F, 0.0F);
@@ -223,6 +_,18 @@
 
     public float getPartialTickTime() {
         return this.partialTickTime;
+    }
+
+    public void setAnglesInternal(float yaw, float pitch) {
+        this.yRot = yaw;
+        this.xRot = pitch;
+    }
+
+    public net.minecraft.world.level.block.state.BlockState getBlockAtCamera() {
+        if (!this.initialized)
+            return net.minecraft.world.level.block.Blocks.AIR.defaultBlockState();
+        else
+            return this.level.getBlockState(this.blockPosition).getStateAtViewpoint(this.level, this.blockPosition, this.position);
     }
 
     @OnlyIn(Dist.CLIENT)
