--- a/net/minecraft/client/Camera.java
+++ b/net/minecraft/client/Camera.java
@@ -325,10 +_,17 @@
         this.setPosition(new Vec3(this.position.x + offset.x, this.position.y + offset.y, this.position.z + offset.z));
     }
 
+    /** @deprecated Neo: call {@link #setRotation(float, float, float)} instead */
+    @Deprecated
     protected void setRotation(float yRot, float xRot) {
+        setRotation(yRot, xRot, 0F);
+    }
+
+    protected void setRotation(float yRot, float xRot, float roll) {
         this.xRot = xRot;
         this.yRot = yRot;
-        this.rotation.rotationYXZ((float) Math.PI - yRot * (float) (Math.PI / 180.0), -xRot * (float) (Math.PI / 180.0), 0.0F);
+        this.roll = roll;
+        this.rotation.rotationYXZ((float) Math.PI - yRot * (float) (Math.PI / 180.0), -xRot * (float) (Math.PI / 180.0), -roll * (float) (Math.PI / 180.0));
         FORWARDS.rotate(this.rotation, this.forwards);
         UP.rotate(this.rotation, this.up);
         LEFT.rotate(this.rotation, this.left);
