--- a/net/minecraft/client/Camera.java
+++ b/net/minecraft/client/Camera.java
@@ -45,6 +_,7 @@
     private float eyeHeight;
     private float eyeHeightOld;
     private float partialTickTime;
+    private float roll;
 
     public void setup(BlockGetter p_90576_, Entity p_90577_, boolean p_90578_, boolean p_90579_, float p_90580_) {
         this.initialized = true;
@@ -63,7 +_,10 @@
             this.setRotation(p_90577_.getViewYRot(p_90580_), p_90577_.getViewXRot(p_90580_));
             this.setPosition(newminecartbehavior.getCartLerpPosition(p_90580_).add(vec3));
         } else {
-            this.setRotation(p_90577_.getViewYRot(p_90580_), p_90577_.getViewXRot(p_90580_));
+            var cameraSetup = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ViewportEvent.ComputeCameraAngles(
+                    this, p_90580_, p_90577_.getViewYRot(p_90580_), p_90577_.getViewXRot(p_90580_), 0)
+            );
+            this.setRotation(cameraSetup.getYaw(), cameraSetup.getPitch(), cameraSetup.getRoll());
             this.setPosition(
                 Mth.lerp((double)p_90580_, p_90577_.xo, p_90577_.getX()),
                 Mth.lerp((double)p_90580_, p_90577_.yo, p_90577_.getY()) + Mth.lerp(p_90580_, this.eyeHeightOld, this.eyeHeight),
@@ -73,7 +_,7 @@
 
         if (p_90578_) {
             if (p_90579_) {
-                this.setRotation(this.yRot + 180.0F, -this.xRot);
+                this.setRotation(this.yRot + 180.0F, -this.xRot, -this.roll);
             }
 
             float f1 = 4.0F;
@@ -90,7 +_,7 @@
                 f = (float)livingentity.getAttributeValue(Attributes.CAMERA_DISTANCE);
             }
 
-            this.move(-this.getMaxZoom(Math.max(f2 * f1, f3 * f)), 0.0F, 0.0F);
+            this.move(-this.getMaxZoom(net.neoforged.neoforge.client.ClientHooks.getDetachedCameraDistance(this, p_90579_, f1, f2, f3, f)), 0.0F, 0.0F);
         } else if (p_90577_ instanceof LivingEntity && ((LivingEntity)p_90577_).isSleeping()) {
             Direction direction = ((LivingEntity)p_90577_).getBedOrientation();
             this.setRotation(direction != null ? direction.toYRot() - 180.0F : 0.0F, 0.0F);
@@ -131,10 +_,17 @@
         this.setPosition(new Vec3(this.position.x + vector3f.x, this.position.y + vector3f.y, this.position.z + vector3f.z));
     }
 
+    /** @deprecated Neo: call {@link #setRotation(float, float, float)} instead */
+    @Deprecated
     protected void setRotation(float p_90573_, float p_90574_) {
+        setRotation(p_90573_, p_90574_, 0F);
+    }
+
+    protected void setRotation(float p_90573_, float p_90574_, float roll) {
         this.xRot = p_90574_;
         this.yRot = p_90573_;
-        this.rotation.rotationYXZ((float) Math.PI - p_90573_ * (float) (Math.PI / 180.0), -p_90574_ * (float) (Math.PI / 180.0), 0.0F);
+        this.roll = roll;
+        this.rotation.rotationYXZ((float) Math.PI - p_90573_ * (float) (Math.PI / 180.0), -p_90574_ * (float) (Math.PI / 180.0), -roll * (float) (Math.PI / 180.0));
         FORWARDS.rotate(this.rotation, this.forwards);
         UP.rotate(this.rotation, this.up);
         LEFT.rotate(this.rotation, this.left);
@@ -259,6 +_,17 @@
     @Override
     public Vec3 position() {
         return this.getPosition();
+    }
+
+    public float getRoll() {
+        return this.roll;
+    }
+
+    public net.minecraft.world.level.block.state.BlockState getBlockAtCamera() {
+        if (!this.initialized)
+            return net.minecraft.world.level.block.Blocks.AIR.defaultBlockState();
+        else
+            return this.level.getBlockState(this.blockPosition).getStateAtViewpoint(this.level, this.blockPosition, this.position);
     }
 
     @OnlyIn(Dist.CLIENT)
