--- a/net/minecraft/client/ClientClockManager.java
+++ b/net/minecraft/client/ClientClockManager.java
@@ -46,6 +_,12 @@
         return this.getInstance(definition).totalTicks;
     }
 
+
+    @Override
+    public float getPartialTick(Holder<WorldClock> definition) {
+        return this.getInstance(definition).partialTick;
+    }
+
     private static class ClockInstance {
         private long totalTicks;
         private float partialTick;
