--- a/net/minecraft/client/ClientClockManager.java
+++ b/net/minecraft/client/ClientClockManager.java
@@ -24,7 +_,11 @@
 
         for (ClientClockManager.ClockInstance instance : this.clocks.values()) {
             if (!instance.paused) {
-                instance.totalTicks += gameTimeDelta;
+                // Neo: support variable-speed clocks
+                float timeStep = instance.fractionalTick + gameTimeDelta * instance.speed;
+                long result = (long)timeStep;
+                instance.fractionalTick = (timeStep - result);
+                instance.totalTicks += result;
             }
         }
     }
@@ -38,6 +_,18 @@
         });
     }
 
+    // Neo: Handle extended state sent by server
+    public void handleExtendedUpdates(long gameTime, Map<Holder<WorldClock>, net.neoforged.neoforge.network.payload.ExtendedClockState> updates) {
+        this.tick(gameTime);
+        updates.forEach((definition, state) -> {
+            ClientClockManager.ClockInstance clock = this.getInstance((Holder<WorldClock>)definition);
+            clock.totalTicks = state.state().totalTicks();
+            clock.paused = state.state().paused();
+            clock.speed = state.speed();
+            clock.fractionalTick = state.fractionalTick();
+        });
+    }
+
     @Override
     public long getTotalTicks(Holder<WorldClock> definition) {
         return this.getInstance(definition).totalTicks;
@@ -47,5 +_,7 @@
     private static class ClockInstance {
         private long totalTicks;
         private boolean paused;
+        private float speed = 1;
+        private float fractionalTick = 0;
     }
 }
