--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -291,6 +_,7 @@
     public final LevelExtractor levelExtractor;
     public final LevelRenderer levelRenderer;
     private final EntityRenderDispatcher entityRenderDispatcher;
+    private final BlockModelResolver blockModelResolver;
     private final ItemModelResolver itemModelResolver;
     private final MapRenderer mapRenderer;
     public final ParticleEngine particleEngine;
@@ -422,7 +_,6 @@
             }
         }, Util.nonCriticalIoPool());
         LOGGER.info("Setting user: {}", this.user.getName());
-        LOGGER.debug("(Session ID is {})", this.user.getSessionId());
         this.demo = gameConfig.game.demo;
         this.allowsMultiplayer = !gameConfig.game.disableMultiplayer;
         this.allowsChat = !gameConfig.game.disableChat;
@@ -484,6 +_,8 @@
         }
 
         String initialWindowTitle = this.createTitle();
+        // Neo: Enable GL debug labels and synchronous GL debug output via FML config
+        boolean enableGlDebug = net.neoforged.fml.loading.FMLConfig.getBoolConfigValue(net.neoforged.fml.loading.FMLConfig.ConfigValue.DEBUG_OPENGL);
 
         for (GpuBackend backendToTry : backendsToTry) {
             boolean libraryLoaded = false;
@@ -494,8 +_,8 @@
                 device = backendToTry.createDevice(
                     new GpuDebugOptions(
                         this.options.glDebugVerbosity,
-                        SharedConstants.DEBUG_SYNCHRONOUS_GL_LOGS,
-                        gameConfig.game.renderDebugLabels,
+                        SharedConstants.DEBUG_SYNCHRONOUS_GL_LOGS || enableGlDebug,
+                        gameConfig.game.renderDebugLabels || enableGlDebug,
                         gameConfig.game.vulkanValidation
                     )
                 );
@@ -570,6 +_,9 @@
         this.keyboardHandler = new KeyboardHandler(this);
         this.options.applyGraphicsPreset(this.options.graphicsPreset().get());
         this.resourceManager = new ReloadableResourceManager(PackType.CLIENT_RESOURCES);
+        net.neoforged.neoforge.client.loading.ClientModLoader.setupModResourcePacks(this.resourcePackRepository);
+        //Move client bootstrap to after mod loading so that events can be fired for it.
+        ClientBootstrap.bootstrap();
         this.resourcePackRepository.reload();
         this.options.loadSelectedResourcePacks(this.resourcePackRepository);
         this.languageManager = new LanguageManager(this, this.options.languageCode, var1 -> {
@@ -603,11 +_,13 @@
         this.resourceManager.registerReloadListener(new DryFoliageColorReloadListener());
         RenderSystem.setupDefaultState();
         this.blockColors = BlockColors.createDefault();
+        net.neoforged.neoforge.client.model.UnbakedModelParser.init();
         this.modelManager = new ModelManager(this.blockColors, this.atlasManager, this.playerSkinRenderCache);
         this.resourceManager.registerReloadListener(this.modelManager);
         EquipmentAssetManager equipmentAssets = new EquipmentAssetManager();
         this.resourceManager.registerReloadListener(equipmentAssets);
         BlockModelResolver blockModelResolver = new BlockModelResolver(this.modelManager);
+        this.blockModelResolver = blockModelResolver;
         this.itemModelResolver = new ItemModelResolver(this.modelManager);
         this.palettedTextureManager = new PalettedTextureManager(this.resourceManager, this.textureManager);
         this.resourceManager.registerReloadListener(this.palettedTextureManager);
@@ -648,6 +_,7 @@
         this.resourceManager.registerReloadListener(particleResources);
         this.particleEngine = new ParticleEngine(this.level, particleResources);
         particleResources.onReload(this.particleEngine::clearParticles);
+        net.neoforged.neoforge.client.ClientHooks.onRegisterParticleProviders(particleResources);
         this.gameRenderer = new GameRenderer(this, new FirstPersonHandsAndItemsRenderer(this), this.modelManager, this.itemModelResolver);
         this.resourceManager.registerReloadListener(this.gameRenderer);
         WindowRenderState windowRenderState = this.gameRenderer.gameRenderState().windowRenderState;
@@ -696,6 +_,7 @@
             MessageBox.error(message.toString());
         }
 
+        net.neoforged.neoforge.client.ClientHooks.initClientHooks(this, this.resourceManager);
         this.window.setAllowCursorChanges(this.options.allowCursorChanges().get());
         this.window.setQuitShortcuts(this.options.quitShortcuts().get());
         if (InputQuirks.EMULATE_RIGHT_CLICK_WITH_CTRL_KEY) {
@@ -715,6 +_,7 @@
         TitleScreen.registerTextures(this.textureManager);
         LoadingOverlay.registerTextures(this.textureManager);
         this.gameRenderer.registerPanoramaTextures(this.textureManager);
+        net.neoforged.neoforge.client.loading.ClientModLoader.finish();
         this.timerQuery = new TimerQuery();
         this.gui.setScreen(new GenericMessageScreen(Component.translatable("gui.loadingMinecraft")));
         List<PackResources> packs = this.resourcePackRepository.openAllSelected();
@@ -724,7 +_,7 @@
         GameLoadTimesEvent.INSTANCE.beginStep(TelemetryProperty.LOAD_TIME_LOADING_OVERLAY_MS);
         GameLoadCookie loadCookie = new GameLoadCookie(realmsClient, gameConfig.quickPlay);
         this.gui.setOverlay(new LoadingOverlay(this, reloadInstance, maybeT -> Util.ifElse(maybeT, t -> this.rollbackResourcePacks(t, loadCookie), () -> {
-            if (SharedConstants.IS_RUNNING_IN_IDE) {
+            if (SharedConstants.IS_RUNNING_IN_IDE && false /* Neo: Disable self-test */) {
                 this.selfTest();
             }
 
@@ -778,6 +_,7 @@
     }
 
     private void onResourceLoadFinished(@Nullable GameLoadCookie loadCookie) {
+        net.neoforged.neoforge.client.ClientHooks.fireResourceLoadFinishedEvent(!this.gameLoadFinished);
         if (!this.gameLoadFinished) {
             this.gameLoadFinished = true;
             this.onGameLoadFinished(loadCookie);
@@ -813,7 +_,7 @@
     private String createTitle() {
         StringBuilder builder = new StringBuilder("Minecraft");
         if (checkModStatus().shouldReportAsModified()) {
-            builder.append("*");
+            builder.append(' ').append(net.neoforged.neoforge.internal.BrandingControl.BRANDING_NAME).append('*');
         }
 
         builder.append(" ");
@@ -853,7 +_,7 @@
     }
 
     private void rollbackResourcePacks(Throwable t, @Nullable GameLoadCookie loadCookie) {
-        if (this.resourcePackRepository.getSelectedIds().size() > 1) {
+        if (this.resourcePackRepository.getSelectedPacks().stream().anyMatch(e -> !e.isRequired())) { //Forge: This caused infinite loop if any resource packs are forced. Such as mod resources. So check if we can disable any.
             this.clearResourcePacksOnError(t, null, loadCookie);
         } else {
             Util.throwAsRuntime(t);
@@ -909,6 +_,8 @@
         DiscontinuousFrame tickFrame = TracyClient.createDiscontinuousFrame("Client Tick");
 
         try {
+            net.neoforged.neoforge.client.ClientLifecycleHooks.handleClientStarted(this);
+
             boolean oomRecovery = false;
 
             while (this.running) {
@@ -965,7 +_,7 @@
         CrashReport finalReport = this.fillReport(partialReport);
         int exitCode = saveReportAndShutdownSoundManager(this, this.gameDirectory, finalReport, -1);
         this.emergencySave();
-        System.exit(exitCode);
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleExit(exitCode);
     }
 
     public static void saveReport(File gameDirectory, CrashReport crash) {
@@ -1002,7 +_,7 @@
 
     public static void crash(@Nullable Minecraft minecraft, File gameDirectory, CrashReport crash, int exitCode) {
         int actualExitCode = saveReportAndShutdownSoundManager(minecraft, gameDirectory, crash, exitCode);
-        System.exit(actualExitCode);
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleExit(actualExitCode);
     }
 
     public static int saveReportAndShutdownSoundManager(@Nullable Minecraft minecraft, File gameDirectory, CrashReport crash, int exitCode) {
@@ -1106,6 +_,7 @@
 
     public void exitWorldAndClose() {
         try {
+            net.neoforged.neoforge.client.ClientLifecycleHooks.handleClientStopping(this);
             LOGGER.info("Stopping!");
             if (this.level != null) {
                 this.level.disconnect(ClientLevel.DEFAULT_QUIT_MESSAGE);
@@ -1148,6 +_,8 @@
             FreeTypeUtil.destroy();
             this.windowSurface.close();
             RenderSystem.shutdownRenderer();
+
+            net.neoforged.neoforge.client.ClientLifecycleHooks.handleClientStopped(this);
         } catch (Throwable t) {
             LOGGER.error("Shutdown failure!", t);
             throw t;
@@ -1229,7 +_,11 @@
 
         this.window.setErrorSection("Post render");
         boolean var28 = this.pause;
-        this.pause = this.hasSingleplayerServer() && this.gui.isPausing() && !this.singleplayerServer.isPublished();
+        boolean pause = this.hasSingleplayerServer() && this.gui.isPausing() && !this.singleplayerServer.isPublished();
+        if (pause != this.pause && !net.neoforged.neoforge.client.ClientHooks.onClientPauseChangePre(pause)) {
+            this.pause = pause;
+            net.neoforged.neoforge.client.ClientHooks.onClientPauseChangePost(pause);
+        }
         if (!var28 && this.pause) {
             this.soundManager.pauseAllExcept(SoundSource.MUSIC, SoundSource.UI);
         }
@@ -1305,7 +_,9 @@
                 profiler.popPush("gpuAsync");
                 RenderSystem.executePendingTasks();
                 profiler.pop();
+                net.neoforged.neoforge.client.ClientHooks.fireRenderFramePre(this.deltaTracker);
                 this.gameRenderer.render();
+                net.neoforged.neoforge.client.ClientHooks.fireRenderFramePost(this.deltaTracker);
             }
 
             profiler.push("swapchainBlit");
@@ -1343,6 +_,7 @@
 
             RenderSystem.getDynamicUniforms().reset();
             this.levelRenderer.endFrame();
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.FlipFrameEvent());
             profiler.popPush("frameLimiter");
             int framerateLimit = this.gameRenderer.gameRenderState().framerateLimit;
             if (framerateLimit < 260) {
@@ -1430,8 +_,10 @@
     public void resizeGui() {
         int guiScale = this.window.calculateScale(this.options.guiScale().get(), this.isEnforceUnicode());
         this.window.setGuiScale(guiScale);
+        net.neoforged.neoforge.client.ClientHooks.fireWindowResize(this.window);
         if (this.gui.screen() != null) {
             this.gui.screen().resize(this.window.getGuiScaledWidth(), this.window.getGuiScaledHeight());
+            this.gui.resizeScreenLayers(this.window.getGuiScaledWidth(), this.window.getGuiScaledHeight());
         }
 
         this.mouseHandler.resyncMousePosition();
@@ -1595,6 +_,7 @@
     }
 
     public void stop() {
+        if (this.isRunning()) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         this.running = false;
     }
 
@@ -1621,7 +_,8 @@
                     BlockPos pos = blockHit.getBlockPos();
                     if (!this.level.getBlockState(pos).isAir()) {
                         Direction direction = blockHit.getDirection();
-                        if (this.gameMode.continueDestroyBlock(pos, direction)) {
+                        var inputEvent = net.neoforged.neoforge.client.ClientHooks.onClickInput(0, this.options.keyAttack, InteractionHand.MAIN_HAND);
+                        if ((inputEvent.isCanceled() || this.gameMode.continueDestroyBlock(pos, direction)) && inputEvent.shouldSwingHand()) {
                             this.player.swing(InteractionHand.MAIN_HAND, swingAnimation, false);
                             this.player.connection.send(ServerboundPunchPacket.INSTANCE);
                         }
@@ -1670,8 +_,16 @@
 
                 SwingAnimation attackAnimation = heldItem.getAttackAnimation();
                 boolean endAttack = false;
+                var inputEvent = net.neoforged.neoforge.client.ClientHooks.onClickInput(0, this.options.keyAttack, InteractionHand.MAIN_HAND);
+                if (inputEvent.isCanceled()) {
+                    if (inputEvent.shouldSwingHand()) {
+                        this.player.swing(InteractionHand.MAIN_HAND, attackAnimation, false);
+                    }
+                    return endAttack;
+                }
                 PiercingWeapon piercingWeapon = heldItem.get(DataComponents.PIERCING_WEAPON);
                 if (piercingWeapon != null) {
+                    if (inputEvent.shouldSwingHand())
                     this.gameMode.piercingAttack(attackAnimation, piercingWeapon);
                     return true;
                 }
@@ -1699,10 +_,13 @@
                         }
 
                         this.player.resetAttackStrengthTicker();
+                        net.neoforged.neoforge.common.CommonHooks.onEmptyLeftClick(this.player);
                 }
 
+                if (inputEvent.shouldSwingHand()) {
                 this.player.swing(InteractionHand.MAIN_HAND, attackAnimation, false);
                 this.player.connection.send(ServerboundPunchPacket.INSTANCE);
+                }
                 return endAttack;
             }
         }
@@ -1723,6 +_,13 @@
                     }
 
                     SwingAnimation swingAnimation = heldItem.getInteractAnimation();
+                    var inputEvent = net.neoforged.neoforge.client.ClientHooks.onClickInput(1, this.options.keyUse, hand);
+                    if (inputEvent.isCanceled()) {
+                        if (inputEvent.shouldSwingHand()) {
+                            this.player.swing(hand, swingAnimation, false);
+                        }
+                        return;
+                    }
                     if (this.hitResult != null) {
                         switch (this.hitResult.getType()) {
                             case ENTITY:
@@ -1734,7 +_,7 @@
 
                                 if (this.player.isWithinEntityInteractionRange(entity, 0.0)
                                     && this.gameMode.interact(this.player, entity, entityHit, hand) instanceof InteractionResult.Success success) {
-                                    if (success.swingSource() == InteractionResult.SwingSource.PREDICTED) {
+                                    if (success.swingSource() == InteractionResult.SwingSource.PREDICTED && inputEvent.shouldSwingHand()) {
                                         this.player.swing(hand, swingAnimation, false);
                                     }
 
@@ -1746,7 +_,7 @@
                                 int oldCount = heldItem.getCount();
                                 InteractionResult useResult = this.gameMode.useItemOn(this.player, hand, blockHit);
                                 if (useResult instanceof InteractionResult.Success success) {
-                                    if (success.swingSource() == InteractionResult.SwingSource.PREDICTED) {
+                                    if (success.swingSource() == InteractionResult.SwingSource.PREDICTED && inputEvent.shouldSwingHand()) {
                                         this.player.swing(hand, swingAnimation, false);
                                         if (!heldItem.isEmpty() && (heldItem.getCount() != oldCount || this.player.hasInfiniteMaterials())) {
                                             this.player.itemUsed(hand);
@@ -1762,8 +_,12 @@
                         }
                     }
 
+                    if (heldItem.isEmpty() && (this.hitResult == null || this.hitResult.getType() == HitResult.Type.MISS)) {
+                        net.neoforged.neoforge.common.CommonHooks.onEmptyClick(this.player, hand);
+                    }
+
                     if (!heldItem.isEmpty() && this.gameMode.useItem(this.player, hand) instanceof InteractionResult.Success success) {
-                        if (success.swingSource() == InteractionResult.SwingSource.PREDICTED) {
+                        if (success.swingSource() == InteractionResult.SwingSource.PREDICTED && inputEvent.shouldSwingHand()) {
                             this.player.swing(hand, swingAnimation, false);
                         }
 
@@ -1781,6 +_,10 @@
 
     public void tick() {
         this.clientTickCount++;
+        if (this.gameLoadFinished) {
+            net.neoforged.neoforge.client.ClientHooks.fireClientTickPre();
+        }
+
         if (this.level != null && !this.pause) {
             this.level.tickRateManager().tick();
         }
@@ -1842,6 +_,7 @@
 
                 this.tutorial.tick();
 
+                net.neoforged.neoforge.event.EventHooks.fireLevelTickPre(this.level, () -> true);
                 try {
                     this.level.tick(() -> true);
                 } catch (Throwable t) {
@@ -1855,6 +_,7 @@
 
                     throw new ReportedException(report);
                 }
+                net.neoforged.neoforge.event.EventHooks.fireLevelTickPost(this.level, () -> true);
             }
 
             profiler.popPush("animateTick");
@@ -1879,6 +_,10 @@
         profiler.popPush("keyboard");
         this.keyboardHandler.tick();
         profiler.pop();
+
+        if (this.gameLoadFinished) {
+            net.neoforged.neoforge.client.ClientHooks.fireClientTickPost();
+        }
     }
 
     private boolean isLevelRunningNormally() {
@@ -2085,6 +_,7 @@
     }
 
     public void setLevel(ClientLevel level) {
+        if (this.level != null) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
         this.level = level;
         this.updateLevelInEngines(level);
     }
@@ -2146,13 +_,16 @@
         IntegratedServer server = this.singleplayerServer;
         this.singleplayerServer = null;
         this.gameRenderer.resetData();
+        net.neoforged.neoforge.client.ClientHooks.firePlayerLogout(this.gameMode, this.player);
         this.gameMode = null;
         this.narrator.clear();
         this.gui.setClientLevelTeardownInProgress(true);
 
+        var shouldRevertRegistriesToFrozen = this.getConnection() != null && this.getConnection().getConnection() != null; // Neo: Track whether to revert registries after disconnect
         try {
             if (this.level != null) {
                 this.gui.hud.onDisconnected();
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
             }
 
             this.level = null;
@@ -2177,6 +_,7 @@
         } finally {
             this.gui.setClientLevelTeardownInProgress(false);
         }
+        if (shouldRevertRegistriesToFrozen) net.neoforged.neoforge.registries.RegistryManager.revertToFrozen(); // Neo: Revert registries to frozen on disconnect
     }
 
     public void clearDownloadedResourcePacks() {
@@ -2202,6 +_,7 @@
         try {
             this.setScreenAndShow(screen);
             this.gui.hud.onDisconnected();
+            if (this.level != null) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
             this.level = null;
             this.updateLevelInEngines(null);
             this.player = null;
@@ -2400,6 +_,7 @@
 
     private void pickBlockOrEntity() {
         if (this.hitResult != null && this.hitResult.getType() != HitResult.Type.MISS) {
+            if (net.neoforged.neoforge.client.ClientHooks.onClickInput(2, this.options.keyPickItem, InteractionHand.MAIN_HAND).isCanceled()) return;
             boolean includeData = this.hasControlDown();
             switch (this.hitResult) {
                 case BlockHitResult blockHitResult:
@@ -2954,6 +_,10 @@
         }
 
         return defaultTickTargetMillis;
+    }
+
+    public BlockModelResolver getBlockModelResolver() {
+        return this.blockModelResolver;
     }
 
     public ItemModelResolver getItemModelResolver() {
