--- a/net/minecraft/client/Screenshot.java
+++ b/net/minecraft/client/Screenshot.java
@@ -43,6 +_,13 @@
                     file2 = new File(file1, p_92297_);
                 }
 
+                net.neoforged.neoforge.client.event.ScreenshotEvent event = net.neoforged.neoforge.client.ClientHooks.onScreenshot(p_406180_, file2);
+                if (event.isCanceled()) {
+                    p_92301_.accept(event.getCancelMessage());
+                    return;
+                }
+                final File target = event.getScreenshotFile();
+
                 Util.ioPool()
                     .execute(
                         () -> {
@@ -50,11 +_,15 @@
                                 NativeImage $$4x = p_406180_;
 
                                 try {
-                                    p_406180_.writeToFile(file2);
-                                    Component component = Component.literal(file2.getName())
+                                    p_406180_.writeToFile(target);
+                                    if (event.getResultMessage() != null) {
+                                        p_92301_.accept(event.getResultMessage());
+                                    } else {
+                                    Component component = Component.literal(target.getName())
                                         .withStyle(ChatFormatting.UNDERLINE)
-                                        .withStyle(p_392495_ -> p_392495_.withClickEvent(new ClickEvent.OpenFile(file2.getAbsoluteFile())));
+                                        .withStyle(p_392495_ -> p_392495_.withClickEvent(new ClickEvent.OpenFile(target.getAbsoluteFile())));
                                     p_92301_.accept(Component.translatable("screenshot.success", component));
+                                    }
                                 } catch (Throwable throwable1) {
                                     if (p_406180_ != null) {
                                         try {
