--- a/net/minecraft/client/ToggleKeyMapping.java
+++ b/net/minecraft/client/ToggleKeyMapping.java
@@ -17,13 +_,14 @@
     @Override
     public void setDown(boolean value) {
         if (this.needsToggle.getAsBoolean()) {
-            if (value) {
+            if (value && isConflictContextAndModifierActive()) {
                 super.setDown(!this.isDown());
             }
         } else {
             super.setDown(value);
         }
     }
+    @Override public boolean isDown() { return this.isDown && (isConflictContextAndModifierActive() || needsToggle.getAsBoolean()); }
 
     protected void reset() {
         super.setDown(false);
