--- a/net/minecraft/client/User.java
+++ b/net/minecraft/client/User.java
@@ -22,6 +_,19 @@
     private final User.Type type;
 
     public User(String name, UUID uuid, String accessToken, Optional<String> xuid, Optional<String> clientId, User.Type type) {
+        if (name == null || name.isEmpty()) {
+            name = "MissingName";
+            uuid = UUID.fromString("00000000-0000-0000-0000-000000000000");
+            accessToken = "NotValid";
+            org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(getClass().getName());
+            logger.warn("=========================================================");
+            logger.warn("WARNING!! the username was not set for this session, typically");
+            logger.warn("this means you installed Forge incorrectly. We have set your");
+            logger.warn("name to \"MissingName\" and your session to nothing. Please");
+            logger.warn("check your installation and post a console log from the launcher");
+            logger.warn("when asking for help!");
+            logger.warn("=========================================================");
+        }
         this.name = name;
         this.uuid = uuid;
         this.accessToken = accessToken;
