--- a/net/minecraft/client/User.java
+++ b/net/minecraft/client/User.java
@@ -22,6 +_,19 @@
    private final User.Type type;
 
    public User(String p_193799_, UUID p_296409_, String p_193800_, Optional<String> p_193802_, Optional<String> p_193803_, User.Type p_193804_) {
+      if (p_193799_ == null || p_193799_.isEmpty()) {
+         p_193799_ = "MissingName";
+         p_296409_ = UUID.fromString("00000000-0000-0000-0000-000000000000");
+         p_193800_ = "NotValid";
+         org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(getClass().getName());
+         logger.warn("=========================================================");
+         logger.warn("WARNING!! the username was not set for this session, typically");
+         logger.warn("this means you installed Forge incorrectly. We have set your");
+         logger.warn("name to \"MissingName\" and your session to nothing. Please");
+         logger.warn("check your installation and post a console log from the launcher");
+         logger.warn("when asking for help!");
+         logger.warn("=========================================================");
+      }
       this.name = p_193799_;
       this.uuid = p_296409_;
       this.accessToken = p_193800_;
