--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -18,6 +18,7 @@
     private static final int CHECK_TICK_DELAY = 5;
     private int openCount;
     private double maxInteractionRange;
+    public boolean opened; // CraftBukkit
 
     public ContainerOpenersCounter() {}
 
@@ -27,11 +28,36 @@
 
     protected abstract void openerCountChanged(Level level, BlockPos pos, BlockState blockState, int previous, int current);
 
+    // CraftBukkit start
+    public void onAPIOpen(Level level, BlockPos blockpos, BlockState blockstate) {
+        onOpen(level, blockpos, blockstate);
+    }
+
+    public void onAPIClose(Level level, BlockPos blockpos, BlockState blockstate) {
+        onClose(level, blockpos, blockstate);
+    }
+
+    public void openerAPICountChanged(Level level, BlockPos blockpos, BlockState blockstate, int i, int j) {
+        openerCountChanged(level, blockpos, blockstate, i, j);
+    }
+    // CraftBukkit end
+
     public abstract boolean isOwnContainer(Player player);
 
     public void incrementOpeners(LivingEntity entity, Level level, BlockPos pos, BlockState blockState, double maxInteractionRange) {
+        int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         int i = this.openCount++;
 
+        // CraftBukkit start - Call redstone event
+        if (level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
+            int newPower = Math.max(0, Math.min(15, this.openCount));
+
+            if (oldPower != newPower) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, oldPower, newPower);
+            }
+        }
+        // CraftBukkit end
+
         if (i == 0) {
             this.onOpen(level, pos, blockState);
             level.gameEvent(entity, (Holder) GameEvent.CONTAINER_OPEN, pos);
@@ -43,8 +69,19 @@
     }
 
     public void decrementOpeners(LivingEntity entity, Level level, BlockPos pos, BlockState blockState) {
+        int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         int i = this.openCount--;
 
+        // CraftBukkit start - Call redstone event
+        if (level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
+            int newPower = Math.max(0, Math.min(15, this.openCount));
+
+            if (oldPower != newPower) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, oldPower, newPower);
+            }
+        }
+        // CraftBukkit end
+
         if (this.openCount == 0) {
             this.onClose(level, pos, blockState);
             level.gameEvent(entity, (Holder) GameEvent.CONTAINER_CLOSE, pos);
@@ -85,6 +122,7 @@
         }
 
         int i = list.size();
+        if (opened) i++; // CraftBukkit - add dummy count from API
         int j = this.openCount;
 
         if (j != i) {
