--- a/net/minecraft/server/jsonrpc/api/Schema.java
+++ b/net/minecraft/server/jsonrpc/api/Schema.java
@@ -32,7 +_,7 @@
                 .apply(p_451636_, Schema::new)
         )
     );
-    private static final List<SchemaComponent> SCHEMA_REGISTRY = new ArrayList<>();
+    private static final List<SchemaComponent> SCHEMA_REGISTRY = net.neoforged.neoforge.server.ServerLifecycleHooks.getSchemaRegistry();
     public static final Schema BOOL_SCHEMA = ofType("boolean");
     public static final Schema INT_SCHEMA = ofType("integer");
     public static final Schema NUMBER_SCHEMA = ofType("number");
@@ -96,7 +_,7 @@
     }
 
     public static List<SchemaComponent> getSchemaRegistry() {
-        return SCHEMA_REGISTRY;
+        return java.util.Collections.unmodifiableList(SCHEMA_REGISTRY);
     }
 
     public static Schema ofRef(URI p_451180_) {
