--- a/net/minecraft/server/jsonrpc/api/SchemaComponent.java
+++ b/net/minecraft/server/jsonrpc/api/SchemaComponent.java
@@ -4,6 +_,10 @@
 import java.util.List;
 
 public record SchemaComponent<T>(String name, URI ref, Schema<T> schema) {
+    public SchemaComponent(net.minecraft.resources.Identifier name, Schema<T> schema) {
+        this(name.toString(), ReferenceUtil.createLocalReference(name.toString()), schema);
+    }
+
     public Schema<T> asRef() {
         return Schema.ofRef(this.ref, this.schema.codec());
     }
