--- a/com/mojang/blaze3d/pipeline/MainTarget.java
+++ b/com/mojang/blaze3d/pipeline/MainTarget.java
@@ -18,7 +_,11 @@
     private static final MainTarget.Dimension DEFAULT_DIMENSIONS = new MainTarget.Dimension(854, 480);
 
     public MainTarget(int desiredWidth, int desiredHeight) {
-        super("Main", GpuFormat.RGBA8_UNORM, GpuFormat.D32_FLOAT);
+        this(desiredWidth, desiredHeight, GpuFormat.D32_FLOAT);
+    }
+
+    public MainTarget(int desiredWidth, int desiredHeight, GpuFormat depthFormat) {
+        super("Main", GpuFormat.RGBA8_UNORM, depthFormat);
         this.createFrameBuffer(desiredWidth, desiredHeight);
     }
 
