--- a/com/mojang/blaze3d/pipeline/RenderPipeline.java
+++ b/com/mojang/blaze3d/pipeline/RenderPipeline.java
@@ -280,6 +_,16 @@
             return this;
         }
 
+        public RenderPipeline.Builder withStencilTest(net.neoforged.neoforge.client.stencil.StencilTest stencilTest) {
+            this.stencilTest = Optional.of(stencilTest);
+            return this;
+        }
+
+        public RenderPipeline.Builder withoutStencilTest(){
+            this.stencilTest = Optional.empty();
+            return this;
+        }
+
         private void withSnippet(RenderPipeline.Snippet snippet) {
             if (snippet.vertexShader.isPresent()) {
                 this.vertexShader = snippet.vertexShader;
@@ -341,6 +_,10 @@
 
             if (snippet.polygonMode.isPresent()) {
                 this.polygonMode = snippet.polygonMode;
+            }
+
+            if (snippet.stencilTest.isPresent()) {
+                this.stencilTest = snippet.stencilTest;
             }
         }
 
