--- a/com/mojang/blaze3d/pipeline/TextureTarget.java
+++ b/com/mojang/blaze3d/pipeline/TextureTarget.java
@@ -9,7 +_,11 @@
 @OnlyIn(Dist.CLIENT)
 public class TextureTarget extends RenderTarget {
     public TextureTarget(@Nullable String label, int width, int height, boolean useDepth, GpuFormat format) {
-        super(label, useDepth, format);
+        this(label, width, height, useDepth, false, format);
+    }
+
+    public TextureTarget(@Nullable String label, int width, int height, boolean useDepth, boolean useStencil, GpuFormat format) {
+        super(label, useDepth, useStencil, format);
         RenderSystem.assertOnRenderThread();
         this.resize(width, height);
     }
