--- a/net/minecraft/world/entity/ConversionTracker.java
+++ b/net/minecraft/world/entity/ConversionTracker.java
@@ -53,6 +_,7 @@
     public void tick() {
         if (this.mob.level() instanceof ServerLevel serverLevel && this.mob.isAlive() && !this.mob.isNoAi()) {
             if (this.isAfflicted.getAsBoolean()) {
+                if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this.mob, this.convertsTo.get(), (timer) -> this.conversionTime = timer)) return;
                 if (this.isConverting()) {
                     this.conversionTime--;
                     if (this.conversionTime < 0) {
@@ -96,6 +_,7 @@
 
     private void doConversion(ServerLevel serverLevel) {
         this.mob.convertTo(this.convertsTo.get(), ConversionParams.single(this.mob, true, true), converted -> {
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this.mob, converted);
             if (this.postConversionStep != null) {
                 this.postConversionStep.accept((M)converted, serverLevel);
             }
