--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -112,7 +_,8 @@
         this.move(MoverType.SELF, this.getDeltaMovement());
         float f = 0.98F;
         if (this.onGround()) {
-            f = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+            BlockPos pos = getBlockPosBelowThatAffectsMyMovement();
+            f = this.level().getBlockState(pos).getFriction(this.level(), pos, this) * 0.98F;
         }
 
         this.setDeltaMovement(this.getDeltaMovement().multiply((double)f, 0.98, (double)f));
@@ -228,6 +_,7 @@
     public void playerTouch(Player entity) {
         if (entity instanceof ServerPlayer serverplayer) {
             if (entity.takeXpDelay == 0) {
+                if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.PlayerXpEvent.PickupXp(entity, this)).isCanceled()) return;
                 entity.takeXpDelay = 2;
                 entity.take(this, 1);
                 int i = this.repairPlayerItems(serverplayer, this.value);
@@ -247,7 +_,7 @@
         Optional<EnchantedItemInUse> optional = EnchantmentHelper.getRandomItemWith(EnchantmentEffectComponents.REPAIR_WITH_XP, player, ItemStack::isDamaged);
         if (optional.isPresent()) {
             ItemStack itemstack = optional.get().itemStack();
-            int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(player.serverLevel(), itemstack, value);
+            int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(player.serverLevel(), itemstack, (int) (value * itemstack.getXpRepairRatio()));
             int j = Math.min(i, itemstack.getDamageValue());
             itemstack.setDamageValue(itemstack.getDamageValue() - j);
             if (j > 0) {
