--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -136,7 +_,8 @@
             this.applyEffectsFromBlocks();
             float friction = 0.98F;
             if (this.onGround()) {
-                friction = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                BlockPos pos = this.getBlockPosBelowThatAffectsMyMovement();
+                friction = this.level().getBlockState(pos).getFriction(this.level(), pos, this) * 0.98F;
             }
 
             this.setDeltaMovement(this.getDeltaMovement().scale(friction));
@@ -153,7 +_,7 @@
 
     private void followNearbyPlayer() {
         if (this.followingPlayer == null || this.followingPlayer.isSpectator() || this.followingPlayer.distanceToSqr(this) > 64.0) {
-            Player nearestPlayer = this.level().getNearestPlayer(this, 8.0);
+            Player nearestPlayer = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.entity.XpOrbTargetingEvent(this, 8.0)).getFollowingPlayer();
             if (nearestPlayer != null && !nearestPlayer.isSpectator() && !nearestPlayer.isDeadOrDying()) {
                 this.followingPlayer = nearestPlayer;
             } else {
@@ -277,6 +_,7 @@
     public void playerTouch(Player player) {
         if (player instanceof ServerPlayer serverPlayer) {
             if (player.takeXpDelay == 0) {
+                if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.PlayerXpEvent.PickupXp(player, this)).isCanceled()) return;
                 player.takeXpDelay = 2;
                 player.take(this, 1);
                 int remaining = this.repairPlayerItems(serverPlayer, this.getValue());
@@ -296,7 +_,7 @@
         Optional<EnchantedItemInUse> selected = EnchantmentHelper.getRandomItemWith(EnchantmentEffectComponents.REPAIR_WITH_XP, player, ItemStack::isDamaged);
         if (selected.isPresent()) {
             ItemStack itemStack = selected.get().itemStack();
-            int toRepairFromXpAmount = EnchantmentHelper.modifyDurabilityToRepairFromXp(player.level(), itemStack, amount);
+            int toRepairFromXpAmount = EnchantmentHelper.modifyDurabilityToRepairFromXp(player.level(), itemStack, (int) (amount * itemStack.getXpRepairRatio()));
             int repair = Math.min(toRepairFromXpAmount, itemStack.getDamageValue());
             itemStack.setDamageValue(itemStack.getDamageValue() - repair);
             if (repair > 0) {
