--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -112,7 +_,8 @@
         this.move(MoverType.SELF, this.getDeltaMovement());
         float f = 0.98F;
         if (this.onGround()) {
-            f = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+            BlockPos pos = getBlockPosBelowThatAffectsMyMovement();
+            f = this.level().getBlockState(pos).getFriction(this.level(), pos, this) * 0.98F;
         }
 
         this.setDeltaMovement(this.getDeltaMovement().multiply((double)f, 0.98, (double)f));
@@ -228,6 +_,7 @@
     public void playerTouch(Player p_20792_) {
         if (p_20792_ instanceof ServerPlayer serverplayer) {
             if (p_20792_.takeXpDelay == 0) {
+                if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.PlayerXpEvent.PickupXp(p_20792_, this)).isCanceled()) return;
                 p_20792_.takeXpDelay = 2;
                 p_20792_.take(this, 1);
                 int i = this.repairPlayerItems(serverplayer, this.value);
@@ -247,7 +_,7 @@
         Optional<EnchantedItemInUse> optional = EnchantmentHelper.getRandomItemWith(EnchantmentEffectComponents.REPAIR_WITH_XP, p_344821_, ItemStack::isDamaged);
         if (optional.isPresent()) {
             ItemStack itemstack = optional.get().itemStack();
-            int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(p_344821_.serverLevel(), itemstack, p_147094_);
+            int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(p_344821_.serverLevel(), itemstack, (int) (p_147094_ * itemstack.getXpRepairRatio()));
             int j = Math.min(i, itemstack.getDamageValue());
             itemstack.setDamageValue(itemstack.getDamageValue() - j);
             if (j > 0) {
