--- a/net/minecraft/world/entity/LightningBolt.java
+++ b/net/minecraft/world/entity/LightningBolt.java
@@ -38,6 +_,7 @@
     private ServerPlayer cause;
     private final Set<Entity> hitEntities = Sets.newHashSet();
     private int blocksSetOnFire;
+    private float damage = 5.0F;
 
     public LightningBolt(EntityType<? extends LightningBolt> p_20865_, Level p_20866_) {
         super(p_20865_, p_20866_);
@@ -68,11 +_,20 @@
     private void powerLightningRod() {
         BlockPos blockpos = this.getStrikePosition();
         BlockState blockstate = this.level().getBlockState(blockpos);
-        if (blockstate.is(Blocks.LIGHTNING_ROD)) {
+        // Neo: support custom LightningRodBlocks
+        if (blockstate.getBlock() instanceof LightningRodBlock) {
             ((LightningRodBlock)blockstate.getBlock()).onLightningStrike(blockstate, this.level(), blockpos);
         }
     }
 
+    public void setDamage(float damage) {
+        this.damage = damage;
+    }
+
+    public float getDamage() {
+        return this.damage;
+    }
+
     @Override
     public void tick() {
         super.tick();
@@ -151,6 +_,7 @@
                     );
 
                 for (Entity entity : list1) {
+                    if (!net.neoforged.neoforge.event.EventHooks.onEntityStruckByLightning(entity, this))
                     entity.thunderHit((ServerLevel)this.level(), this);
                 }
 
@@ -191,7 +_,8 @@
         BlockState blockstate = p_147151_.getBlockState(p_147152_);
         BlockPos blockpos;
         BlockState blockstate1;
-        if (blockstate.is(Blocks.LIGHTNING_ROD)) {
+        // Neo: support custom LightningRodBlocks
+        if (blockstate.getBlock() instanceof LightningRodBlock) {
             blockpos = p_147152_.relative(blockstate.getValue(LightningRodBlock.FACING).getOpposite());
             blockstate1 = p_147151_.getBlockState(blockpos);
         } else {
