--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -197,14 +_,20 @@
     private static final int CURRENT_IMPULSE_CONTEXT_RESET_GRACE_TIME_TICKS = 40;
     private static final int DEFAULT_CURRENT_IMPULSE_CONTEXT_RESET_GRACE_TIME = 0;
     private int currentImpulseContextResetGraceTime = 0;
-    public static final Predicate<LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM = livingEntity -> {
+    // Neo: Support IItemExtension#isGazeDisguise
+    public static final java.util.function.BiPredicate<LivingEntity, @org.jspecify.annotations.Nullable LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET = (livingEntity, target) -> {
         if (livingEntity instanceof Player player) {
             ItemStack helmet = player.getItemBySlot(EquipmentSlot.HEAD);
-            return !helmet.is(ItemTags.GAZE_DISGUISE_EQUIPMENT);
+            return !helmet.isGazeDisguise(player, target);
         } else {
             return true;
         }
     };
+    /** @deprecated Neo: use {@link #PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET} with target info instead */
+    @Deprecated
+    public static final Predicate<LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM = livingEntity -> {
+        return PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET.test(livingEntity, null);
+     };
     private static final Dynamic<?> EMPTY_BRAIN = new Dynamic<>(JavaOps.INSTANCE, Map.of("memories", Map.of()));
     private final AttributeMap attributes;
     private final CombatTracker combatTracker = new CombatTracker(this);
@@ -272,6 +_,13 @@
     private Waypoint.Icon locatorBarIcon = new Waypoint.Icon();
     public @Nullable Vec3 currentImpulseImpactPos;
     public @Nullable Entity currentExplosionCause;
+    /**
+     * This field stores information about damage dealt to this entity.
+     * a new {@link net.neoforged.neoforge.common.damagesource.DamageContainer} is instantiated
+     * via {@link #hurt(DamageSource, float)} after invulnerability checks, and is removed from
+     * the stack before the method's return.
+     **/
+    protected java.util.@Nullable Stack<net.neoforged.neoforge.common.damagesource.DamageContainer> damageContainers = new java.util.Stack<>();
 
     protected LivingEntity(EntityType<? extends LivingEntity> type, Level level) {
         super(type, level);
@@ -348,13 +_,15 @@
             .add(Attributes.MOVEMENT_EFFICIENCY)
             .add(Attributes.ATTACK_KNOCKBACK)
             .add(Attributes.CAMERA_DISTANCE)
-            .add(Attributes.WAYPOINT_TRANSMIT_RANGE);
+            .add(Attributes.WAYPOINT_TRANSMIT_RANGE)
+            .add(net.neoforged.neoforge.common.NeoForgeMod.SWIM_SPEED)
+            .add(net.neoforged.neoforge.common.NeoForgeMod.NAMETAG_DISTANCE);
     }
 
     @Override
     protected void checkFallDamage(double ya, boolean onGround, BlockState onState, BlockPos pos) {
         if (!this.isInWater()) {
-            this.updateInWaterStateAndDoWaterCurrentPushing();
+            this.updateInWaterStateAndDoWaterCurrentPushing(false);
         }
 
         if (this.level() instanceof ServerLevel level && onGround && this.fallDistance > 0.0) {
@@ -375,7 +_,8 @@
 
                 double scale = Math.min(0.2F + power / 15.0, 2.5);
                 int particles = (int)(150.0 * scale);
-                level.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
+                if (!onState.addLandingEffects(level, pos, onState, this, particles))
+                level.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState, pos), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
             }
         }
 
@@ -385,6 +_,7 @@
         }
     }
 
+    @Deprecated //FORGE: Use canDrownInFluidType instead
     public boolean canBreatheUnderwater() {
         return this.is(EntityTypeTags.CAN_BREATHE_UNDER_WATER);
     }
@@ -425,6 +_,9 @@
                 }
             }
 
+            int airSupply = this.getAirSupply();
+            net.neoforged.neoforge.common.CommonHooks.onLivingBreathe(this, airSupply - decreaseAirSupply(airSupply), increaseAirSupply(airSupply) - airSupply);
+            if (false) // Forge: Handled in ForgeHooks#onLivingBreathe(LivingEntity, int, int)
             if (this.isEyeInFluid(FluidTags.WATER)
                 && !level.getBlockState(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())).is(Blocks.BUBBLE_COLUMN)) {
                 boolean canDrownInWater = !this.canBreatheUnderwater()
@@ -775,7 +_,11 @@
         } else {
             ItemEntity entity = this.createItemStackToDrop(itemStack, randomly, thrownFromHand);
             if (entity != null) {
-                this.level().addFreshEntity(entity);
+                if (captureDrops() != null) {
+                    captureDrops().add(entity);
+                } else {
+                    this.level().addFreshEntity(entity);
+                }
             }
 
             return entity;
@@ -843,8 +_,10 @@
                     Holder<MobEffect> mobEffect = iterator.next();
                     MobEffectInstance effect = this.activeEffects.get(mobEffect);
                     if (!effect.tickServer(serverLevel, this, () -> this.onEffectUpdated(effect, true, null))) {
+                        if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.MobEffectEvent.Expired(this, effect)).isCanceled()) {
                         iterator.remove();
                         this.onEffectsRemoved(List.of(effect));
+                        }
                     } else if (effect.getDuration() % 600 == 0) {
                         this.onEffectUpdated(effect, false, null);
                     }
@@ -891,8 +_,9 @@
         List<ParticleOptions> visibleEffectParticles = this.activeEffects
             .values()
             .stream()
-            .filter(MobEffectInstance::isVisible)
-            .map(MobEffectInstance::getParticleOptions)
+            .map(effect -> net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.EffectParticleModificationEvent(this, effect)))
+            .filter(net.neoforged.neoforge.event.entity.living.EffectParticleModificationEvent::isVisible)
+            .map(net.neoforged.neoforge.event.entity.living.EffectParticleModificationEvent::getParticleOptions)
             .toList();
         this.entityData.set(DATA_EFFECT_PARTICLES, visibleEffectParticles);
         this.entityData.set(DATA_EFFECT_AMBIENCE_ID, areAllEffectsAmbient(this.activeEffects.values()));
@@ -931,6 +_,7 @@
             }
         }
 
+        visibilityPercent = net.neoforged.neoforge.common.CommonHooks.getEntityVisibilityMultiplier(this, targetingEntity, visibilityPercent);
         return visibilityPercent;
     }
 
@@ -966,9 +_,14 @@
         } else if (this.activeEffects.isEmpty()) {
             return false;
         } else {
-            Map<Holder<MobEffect>, MobEffectInstance> copy = Maps.newHashMap(this.activeEffects);
-            this.activeEffects.clear();
-            this.onEffectsRemoved(copy.values());
+            Map<Holder<MobEffect>, MobEffectInstance> map = new java.util.HashMap<>(this.activeEffects.size());
+            for (Map.Entry<Holder<MobEffect>, MobEffectInstance> entry : this.activeEffects.entrySet()) {
+                if (!net.neoforged.neoforge.event.EventHooks.onEffectRemoved(this, entry.getValue())) {
+                    map.put(entry.getKey(), entry.getValue());
+                }
+            }
+            map.keySet().forEach(this.activeEffects::remove);
+            this.onEffectsRemoved(map.values());
             return true;
         }
     }
@@ -999,11 +_,12 @@
     }
 
     public boolean addEffect(MobEffectInstance newEffect, @Nullable Entity source) {
-        if (!this.canBeAffected(newEffect)) {
+        if (!net.neoforged.neoforge.common.CommonHooks.canMobEffectBeApplied(this, newEffect, source)) {
             return false;
         } else {
             MobEffectInstance effect = this.activeEffects.get(newEffect.getEffect());
             boolean changed = false;
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.MobEffectEvent.Added(this, effect, newEffect, source));
             if (effect == null) {
                 this.activeEffects.put(newEffect.getEffect(), newEffect);
                 this.onEffectAdded(newEffect, source);
@@ -1019,6 +_,14 @@
         }
     }
 
+    /**
+     * Neo: Override-Only. Call via {@link net.neoforged.neoforge.common.CommonHooks#canMobEffectBeApplied(LivingEntity, MobEffectInstance, Entity)}
+     *
+     * @param newEffect A mob effect instance
+     * @return If the mob effect instance can be applied to this entity
+     */
+    @Deprecated
+    @org.jetbrains.annotations.ApiStatus.OverrideOnly
     public boolean canBeAffected(MobEffectInstance newEffect) {
         if (this.is(EntityTypeTags.IMMUNE_TO_INFESTED)) {
             return !newEffect.is(MobEffects.INFESTED);
@@ -1030,7 +_,7 @@
     }
 
     public void forceAddEffect(MobEffectInstance newEffect, @Nullable Entity source) {
-        if (this.canBeAffected(newEffect)) {
+        if (net.neoforged.neoforge.common.CommonHooks.canMobEffectBeApplied(this, newEffect, source)) {
             MobEffectInstance previousEffect = this.activeEffects.put(newEffect.getEffect(), newEffect);
             if (previousEffect == null) {
                 this.onEffectAdded(newEffect, source);
@@ -1050,6 +_,7 @@
     }
 
     public boolean removeEffect(Holder<MobEffect> effect) {
+        if (net.neoforged.neoforge.event.EventHooks.onEffectRemoved(this, effect)) return false;
         MobEffectInstance effectInstance = this.removeEffectNoUpdate(effect);
         if (effectInstance != null) {
             this.onEffectsRemoved(List.of(effectInstance));
@@ -1141,6 +_,8 @@
     }
 
     public void heal(float heal) {
+        heal = net.neoforged.neoforge.event.EventHooks.onLivingHeal(this, heal);
+        if (heal <= 0) return;
         float health = this.getHealth();
         if (health > 0.0F) {
             this.setHealth(health + heal);
@@ -1168,11 +_,14 @@
         } else if (source.is(DamageTypeTags.IS_FIRE) && this.hasEffect(MobEffects.FIRE_RESISTANCE)) {
             return false;
         } else {
+            this.damageContainers.push(new net.neoforged.neoforge.common.damagesource.DamageContainer(source, damage));
+            if (net.neoforged.neoforge.common.CommonHooks.onEntityIncomingDamage(this, this.damageContainers.peek())) return false;
             if (this.isSleeping()) {
                 this.stopSleeping();
             }
 
             this.noActionTime = 0;
+            damage = this.damageContainers.peek().getNewDamage(); //Neo: enforce damage container as source of truth for damage amount
             if (damage < 0.0F) {
                 damage = 0.0F;
             }
@@ -1193,24 +_,28 @@
             if (Float.isNaN(damage) || Float.isInfinite(damage)) {
                 damage = Float.MAX_VALUE;
             }
+            this.damageContainers.peek().setNewDamage(damage); //update container with vanilla changes
 
             boolean tookFullDamage = true;
             if (this.invulnerableTime > 10.0F && !source.is(DamageTypeTags.BYPASSES_COOLDOWN)) {
                 if (damage <= this.lastHurt) {
+                    this.damageContainers.pop();
                     return false;
                 }
 
+                this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.INVULNERABILITY, this.lastHurt);
                 this.actuallyHurt(level, source, damage - this.lastHurt);
                 this.lastHurt = damage;
                 tookFullDamage = false;
             } else {
                 this.lastHurt = damage;
-                this.invulnerableTime = 20;
+                this.invulnerableTime = this.damageContainers.peek().getPostAttackInvulnerabilityTicks();
                 this.actuallyHurt(level, source, damage);
                 this.hurtDuration = 10;
                 this.hurtTime = this.hurtDuration;
             }
 
+            damage = this.damageContainers.peek().getNewDamage(); //update local with container value
             this.resolveMobResponsibleForDamage(source);
             this.resolvePlayerResponsibleForDamage(source);
             if (tookFullDamage) {
@@ -1279,6 +_,7 @@
                 CriteriaTriggers.PLAYER_HURT_ENTITY.trigger(sourcePlayer, this, source, damage, damage, blocked);
             }
 
+            this.damageContainers.pop();
             return success;
         }
     }
@@ -1292,7 +_,7 @@
                 return 0.0F;
             } else {
                 BlocksAttacks blocksAttacks = blockingWith.get(DataComponents.BLOCKS_ATTACKS);
-                if (blocksAttacks != null && !blocksAttacks.bypassedBy().map(source::is).orElse(false)) {
+                if (blocksAttacks != null) {
                     if (source.getDirectEntity() instanceof AbstractArrow abstractArrow && abstractArrow.getPierceLevel() > 0) {
                         return 0.0F;
                     } else {
@@ -1308,7 +_,13 @@
                         }
 
                         float damageBlocked = blocksAttacks.resolveBlockedDamage(source, damage, angle);
-                        blocksAttacks.hurtBlockingItem(this.level(), blockingWith, this, this.getUsedItemHand(), damageBlocked);
+
+                        var ev = net.neoforged.neoforge.common.CommonHooks.onDamageBlock(this, this.damageContainers.peek(), damageBlocked, !blocksAttacks.bypassedBy().map(source::is).orElse(false));
+                        if (!ev.getBlocked()) return 0.0F;
+
+                        damageBlocked = ev.getBlockedDamage();
+                        this.damageContainers.peek().setBlockedDamage(ev);
+                        blocksAttacks.hurtBlockingItem(this.level(), blockingWith, this, this.getUsedItemHand(), damageBlocked, ev.shieldDamage());
                         if (damageBlocked > 0.0F && !source.is(DamageTypeTags.IS_PROJECTILE) && source.getDirectEntity() instanceof LivingEntity livingEntity) {
                             this.blockUsingItem(level, livingEntity);
                         }
@@ -1341,9 +_,9 @@
         Entity sourceEntity = source.getEntity();
         if (sourceEntity instanceof Player playerSource) {
             this.setLastHurtByPlayer(playerSource, 100);
-        } else if (sourceEntity instanceof Wolf wolf && wolf.isTame()) {
-            if (wolf.getOwnerReference() != null) {
-                this.setLastHurtByPlayer(wolf.getOwnerReference().getUUID(), 100);
+        } else if (sourceEntity instanceof TamableAnimal tamableAnimal && tamableAnimal.isTame()) {
+            if (tamableAnimal.getOwnerReference() != null) {
+                this.setLastHurtByPlayer(tamableAnimal.getOwnerReference().getUUID(), 100);
             } else {
                 this.lastHurtByPlayer = null;
                 this.lastHurtByPlayerMemoryTime = 0;
@@ -1371,7 +_,7 @@
             for (InteractionHand hand : InteractionHand.values()) {
                 ItemStack itemStack = this.getItemInHand(hand);
                 protection = itemStack.get(DataComponents.DEATH_PROTECTION);
-                if (protection != null) {
+                if (protection != null && net.neoforged.neoforge.common.CommonHooks.onLivingUseTotem(this, killingDamage, itemStack, hand)) {
                     protectionItem = itemStack.copy();
                     itemStack.shrink(1);
                     break;
@@ -1427,6 +_,7 @@
     }
 
     public void die(DamageSource source) {
+        if (net.neoforged.neoforge.common.CommonHooks.onLivingDeath(this, source)) return;
         if (!this.isRemoved() && !this.dead) {
             Entity sourceEntity = source.getEntity();
             LivingEntity killer = this.getKillCredit();
@@ -1463,7 +_,7 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
             if (killer instanceof WitherBoss) {
-                if (serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverLevel, killer)) {
                     BlockPos pos = this.blockPosition();
                     BlockState state = Blocks.WITHER_ROSE.defaultBlockState();
                     if (this.level().getBlockState(pos).isAir() && state.canSurvive(this.level(), pos)) {
@@ -1481,6 +_,7 @@
     }
 
     protected void dropAllDeathLoot(ServerLevel level, DamageSource source) {
+        this.captureDrops(new java.util.ArrayList<>());
         boolean playerKilled = this.lastHurtByPlayerMemoryTime > 0;
         if (this.shouldDropLoot(level)) {
             this.dropFromLootTable(level, source, playerKilled);
@@ -1489,6 +_,10 @@
 
         this.dropEquipment(level);
         this.dropExperience(level, source.getEntity());
+
+        Collection<ItemEntity> drops = captureDrops(null);
+        if (!net.neoforged.neoforge.common.CommonHooks.onLivingDrops(this, source, drops, lastHurtByPlayerMemoryTime > 0))
+            drops.forEach(e -> level().addFreshEntity(e));
     }
 
     protected void dropEquipment(ServerLevel level) {
@@ -1500,7 +_,8 @@
                 this.isAlwaysExperienceDropper()
                     || this.lastHurtByPlayerMemoryTime > 0 && this.shouldDropExperience() && level.getGameRules().get(GameRules.MOB_DROPS)
             )) {
-            ExperienceOrb.award(level, this.position(), this.getExperienceReward(level, killer));
+            int reward = net.neoforged.neoforge.event.EventHooks.getExperienceDrop(this, this.getLastHurtByPlayer(), this.getExperienceReward(level, killer));
+            ExperienceOrb.award((ServerLevel) this.level(), this.position(), reward);
         }
     }
 
@@ -1600,6 +_,11 @@
     }
 
     public void knockback(double power, double xd, double zd) {
+        net.neoforged.neoforge.event.entity.living.LivingKnockBackEvent event = net.neoforged.neoforge.common.CommonHooks.onLivingKnockBack(this, (float) power, xd, zd);
+        if(event.isCanceled()) return;
+        power = event.getStrength();
+        xd = event.getRatioX();
+        zd = event.getRatioZ();
         power *= 1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
         if (!(power <= 0.0)) {
             this.needsSync = true;
@@ -1681,17 +_,11 @@
         } else {
             BlockPos ladderCheckPos = this.blockPosition();
             BlockState state = this.getInBlockState();
-            if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH)) {
-                return false;
-            } else if (state.is(BlockTags.CLIMBABLE)) {
-                this.lastClimbablePos = Optional.of(ladderCheckPos);
-                return true;
-            } else if (state.getBlock() instanceof TrapDoorBlock && this.trapdoorUsableAsLadder(ladderCheckPos, state)) {
-                this.lastClimbablePos = Optional.of(ladderCheckPos);
-                return true;
-            } else {
-                return false;
-            }
+            if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH))
+                return false;
+            Optional<BlockPos> ladderPos = net.neoforged.neoforge.common.CommonHooks.isLivingOnLadder(state, level(), ladderCheckPos, this);
+            if (ladderPos.isPresent()) this.lastClimbablePos = ladderPos;
+            return ladderPos.isPresent();
         }
     }
 
@@ -1752,6 +_,11 @@
             effectiveFallDistance = fallDistance;
         }
 
+        var event = net.neoforged.neoforge.common.CommonHooks.onLivingFall(this, effectiveFallDistance, damageModifier);
+        if (event.isCanceled()) return false;
+        effectiveFallDistance = event.getDistance();
+        damageModifier = event.getDamageMultiplier();
+
         boolean damaged = super.causeFallDamage(effectiveFallDistance, damageModifier, damageSource);
         int dmg = this.calculateFallDamage(effectiveFallDistance, damageModifier);
         if (dmg > 0) {
@@ -1816,10 +_,10 @@
             int xx = Mth.floor(this.getX());
             int yy = Mth.floor(this.getY() - 0.2F);
             int zz = Mth.floor(this.getZ());
-            BlockState state = this.level().getBlockState(new BlockPos(xx, yy, zz));
+            var pos = new BlockPos(xx, yy, zz);
+            BlockState state = this.level().getBlockState(pos);
             if (!state.isAir()) {
-                SoundType soundType = state.getSoundType();
-                this.playSound(soundType.getFallSound(), soundType.getVolume() * 0.5F, soundType.getPitch() * 0.75F);
+                state.playFallSound(this.level(), pos, this);
             }
         }
     }
@@ -1844,6 +_,8 @@
         if (!(damage <= 0.0F)) {
             int durabilityDamage = (int)Math.max(1.0F, damage / 4.0F);
 
+            net.neoforged.neoforge.common.CommonHooks.onArmorHurt(damageSource, slots, durabilityDamage, this);
+            if (true) return; //Neo: Invalidates the loop. Armor damage happens in common hook.
             for (EquipmentSlot slot : slots) {
                 ItemStack itemStack = this.getItemBySlot(slot);
                 Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
@@ -1877,6 +_,7 @@
                 damage = Math.max(v / 25.0F, 0.0F);
                 float damageResisted = oldDamage - damage;
                 if (damageResisted > 0.0F && damageResisted < 3.4028235E37F) {
+                    this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.MOB_EFFECTS, damageResisted);
                     if (this instanceof ServerPlayer) {
                         ((ServerPlayer)this).awardStat(Stats.DAMAGE_RESISTED, Math.round(damageResisted * 10.0F));
                     } else if (damageSource.getEntity() instanceof ServerPlayer) {
@@ -1899,6 +_,7 @@
 
                 if (enchantmentArmor > 0.0F) {
                     damage = CombatRules.getDamageAfterMagicAbsorb(damage, enchantmentArmor);
+                    this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ENCHANTMENTS,this.damageContainers.peek().getNewDamage() - damage);
                 }
 
                 return damage;
@@ -1908,11 +_,13 @@
 
     protected void actuallyHurt(ServerLevel level, DamageSource source, float dmg) {
         if (!this.isInvulnerableTo(level, source)) {
-            dmg = this.getDamageAfterArmorAbsorb(source, dmg);
-            dmg = this.getDamageAfterMagicAbsorb(source, dmg);
-            float var10 = Math.max(dmg - this.getAbsorptionAmount(), 0.0F);
-            this.setAbsorptionAmount(this.getAbsorptionAmount() - (dmg - var10));
-            float absorbedDamage = dmg - var10;
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ARMOR, this.damageContainers.peek().getNewDamage() - this.getDamageAfterArmorAbsorb(source, this.damageContainers.peek().getNewDamage()));
+            this.getDamageAfterMagicAbsorb(source, this.damageContainers.peek().getNewDamage());
+            float damage = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION, Math.min(this.getAbsorptionAmount(), damage));
+            float absorbedDamage = Math.min(damage, this.damageContainers.peek().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION));
+            this.setAbsorptionAmount(Math.max(0, this.getAbsorptionAmount() - absorbedDamage));
+            float var10 = this.damageContainers.peek().getNewDamage();
             if (absorbedDamage > 0.0F && absorbedDamage < 3.4028235E37F && source.getEntity() instanceof ServerPlayer serverPlayer) {
                 serverPlayer.awardStat(Stats.DAMAGE_DEALT_ABSORBED, Math.round(absorbedDamage * 10.0F));
             }
@@ -1920,9 +_,10 @@
             if (var10 != 0.0F) {
                 this.getCombatTracker().recordDamage(source, var10);
                 this.setHealth(this.getHealth() - var10);
-                this.setAbsorptionAmount(this.getAbsorptionAmount() - var10);
                 this.gameEvent(GameEvent.ENTITY_DAMAGE);
+                this.onDamageTaken(this.damageContainers.peek());
             }
+            net.neoforged.neoforge.common.CommonHooks.onLivingDamagePost(this, this.damageContainers.peek());
         }
     }
 
@@ -1979,6 +_,8 @@
     }
 
     public void swing(InteractionHand hand, boolean sendToSwingingEntity) {
+        ItemStack stack = this.getItemInHand(hand);
+        if (!stack.isEmpty() && stack.onEntitySwing(this, hand)) return;
         if (!this.swinging || this.swingTime >= this.getCurrentSwingDuration() / 2 || this.swingTime < 0) {
             this.swingTime = -1;
             this.swinging = true;
@@ -2121,8 +_,10 @@
 
     private void swapHandItems() {
         ItemStack tmp = this.getItemBySlot(EquipmentSlot.OFFHAND);
-        this.setItemSlot(EquipmentSlot.OFFHAND, this.getItemBySlot(EquipmentSlot.MAINHAND));
-        this.setItemSlot(EquipmentSlot.MAINHAND, tmp);
+        var event = net.neoforged.neoforge.common.CommonHooks.onLivingSwapHandItems(this);
+        if (event.isCanceled()) return;
+        this.setItemSlot(EquipmentSlot.OFFHAND, event.getItemSwappedToOffHand());
+        this.setItemSlot(EquipmentSlot.MAINHAND, event.getItemSwappedToMainHand());
     }
 
     @Override
@@ -2230,7 +_,13 @@
     }
 
     public void setItemSlot(EquipmentSlot slot, ItemStack itemStack) {
-        this.onEquipItem(slot, this.equipment.set(slot, itemStack), itemStack);
+        setItemSlot(slot, itemStack, false);
+    }
+    /** Neo: Passing {@code true} for {@code insideTransaction} sets the item without side-effects (calling {@link #onEquipItem}).
+     * Callers are responsible for performing the deferred side-effects, when the active transaction commits. */
+    public void setItemSlot(EquipmentSlot slot, ItemStack itemStack, boolean insideTransaction) {
+        var oldStack = this.equipment.set(slot, itemStack);
+        if (!insideTransaction) this.onEquipItem(slot, oldStack, itemStack);
     }
 
     public float getArmorCoverPercentage() {
@@ -2335,15 +_,18 @@
             }
 
             this.needsSync = true;
+            net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
         }
     }
 
+    @Deprecated // FORGE: use sinkInFluid instead
     protected void goDownInWater() {
-        this.setDeltaMovement(this.getDeltaMovement().add(0.0, -0.04F, 0.0));
+        this.sinkInFluid(net.neoforged.neoforge.common.NeoForgeMod.WATER_TYPE.value());
     }
 
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(TagKey<Fluid> type) {
-        this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.04F, 0.0));
+        this.setDeltaMovement(this.getDeltaMovement().add(0.0D, (double)0.04F * this.getAttributeValue(net.neoforged.neoforge.common.NeoForgeMod.SWIM_SPEED), 0.0D));
     }
 
     protected float getWaterSlowDown() {
@@ -2375,7 +_,8 @@
     }
 
     protected boolean shouldTravelInFluid(FluidState fluidState) {
-        return (this.isInWater() || this.isInLava()) && this.isAffectedByFluids() && !this.canStandOnFluid(fluidState);
+        // Neo: Call (patched-in overload of) #travelInFluid for custom fluid types
+        return (this.isInWater() || this.isInLava() || this.isInFluidType(fluidState)) && this.isAffectedByFluids() && !this.canStandOnFluid(fluidState);
     }
 
     protected void travelFlying(Vec3 input, float speed) {
@@ -2400,7 +_,7 @@
 
     private void travelInAir(Vec3 input) {
         BlockPos posBelow = this.getBlockPosBelowThatAffectsMyMovement();
-        float blockFriction = this.onGround() ? this.level().getBlockState(posBelow).getBlock().getFriction() : 1.0F;
+        float blockFriction = this.onGround() ? this.level().getBlockState(posBelow).getFriction(this.level(), posBelow, this) : 1.0F;
         float friction = blockFriction * 0.91F;
         Vec3 movement = this.handleRelativeFrictionAndCalculateMovement(input, blockFriction);
         double movementY = movement.y;
@@ -2423,11 +_,20 @@
         }
     }
 
+    /**
+     * @deprecated Neo: use {@link #travelInFluid(Vec3, FluidState)} instead
+     */
+    @Deprecated
     private void travelInFluid(Vec3 input) {
+        travelInFluid(input, net.minecraft.world.level.material.Fluids.EMPTY.defaultFluidState());
+    }
+
+    private void travelInFluid(Vec3 input, FluidState fluidState) {
         boolean isFalling = this.getDeltaMovement().y <= 0.0;
         double oldY = this.getY();
         double baseGravity = this.getEffectiveGravity();
-        if (this.isInWater()) {
+        // Neo: patched in to use the fluid the eye is in because fluidState is always EMPTY fluid.
+        if (this.isInWater() || (this.isInFluidType(this.getEyeInFluidType()) && !this.moveInFluid(this.level().getFluidState(BlockPos.containing(this.getEyePosition())), input, baseGravity))) {
             this.travelInWater(input, baseGravity, isFalling, oldY);
             this.floatInWaterWhileRidden();
         } else {
@@ -2452,6 +_,7 @@
             slowDown = 0.96F;
         }
 
+        speed *= (float)this.getAttributeValue(net.neoforged.neoforge.common.NeoForgeMod.SWIM_SPEED);
         this.moveRelative(speed, input);
         this.move(MoverType.SELF, this.getDeltaMovement());
         Vec3 ladderMovement = this.getDeltaMovement();
@@ -2625,7 +_,7 @@
             double xd = Mth.clamp(delta.x, -0.15F, 0.15F);
             double zd = Mth.clamp(delta.z, -0.15F, 0.15F);
             double yd = Math.max(delta.y, -0.15F);
-            if (yd < 0.0 && !this.getInBlockState().is(Blocks.SCAFFOLDING) && this.isSuppressingSlidingDownLadder() && this instanceof Player) {
+            if (yd < 0.0D && !this.getInBlockState().isScaffolding(this) && this.isSuppressingSlidingDownLadder() && this instanceof Player) {
                 yd = 0.0;
             }
 
@@ -2859,6 +_,7 @@
             ItemStack previous = this.lastEquipmentItems.get(slot);
             ItemStack current = this.getItemBySlot(slot);
             if (this.equipmentHasChanged(previous, current)) {
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.LivingEquipmentChangeEvent(this, slot, previous, current));
                 if (changedItems == null) {
                     changedItems = Maps.newEnumMap(EquipmentSlot.class);
                 }
@@ -2994,6 +_,9 @@
         profiler.push("jump");
         if (this.jumping && this.isAffectedByFluids()) {
             double fluidHeight;
+            net.neoforged.neoforge.fluids.FluidType fluidType = this.getMaxHeightFluidType();
+            if (!fluidType.isAir()) fluidHeight = this.getFluidTypeHeight(fluidType);
+            else
             if (this.isInLava()) {
                 fluidHeight = this.getFluidHeight(FluidTags.LAVA);
             } else {
@@ -3004,15 +_,17 @@
             double fluidJumpThreshold = this.getFluidJumpThreshold();
             if (!inWaterAndHasFluidHeight || this.onGround() && !(fluidHeight > fluidJumpThreshold)) {
                 if (!this.isInLava() || this.onGround() && !(fluidHeight > fluidJumpThreshold)) {
+                    if (fluidType.isAir() || this.onGround() && !(fluidHeight > fluidJumpThreshold)) {
                     if ((this.onGround() || inWaterAndHasFluidHeight && fluidHeight <= fluidJumpThreshold) && this.noJumpDelay == 0) {
                         this.jumpFromGround();
                         this.noJumpDelay = 10;
                     }
+                    } else this.jumpInFluid(fluidType);
                 } else {
-                    this.jumpInLiquid(FluidTags.LAVA);
+                    this.jumpInFluid(net.neoforged.neoforge.common.NeoForgeMod.LAVA_TYPE.value());
                 }
             } else {
-                this.jumpInLiquid(FluidTags.WATER);
+                this.jumpInFluid(net.neoforged.neoforge.common.NeoForgeMod.WATER_TYPE.value());
             }
         } else {
             this.noJumpDelay = 0;
@@ -3330,8 +_,11 @@
 
     private void updatingUsingItem() {
         if (this.isUsingItem()) {
-            if (ItemStack.isSameItem(this.getItemInHand(this.getUsedItemHand()), this.useItem)) {
-                this.useItem = this.getItemInHand(this.getUsedItemHand());
+            ItemStack itemStack = this.getItemInHand(this.getUsedItemHand());
+            if (net.neoforged.neoforge.common.CommonHooks.canContinueUsing(this.useItem, itemStack)) {
+                this.useItem = itemStack;
+            }
+            if (itemStack == this.useItem) {
                 this.updateUsingItem(this.useItem);
             } else {
                 this.stopUsingItem();
@@ -3374,8 +_,11 @@
     }
 
     protected void updateUsingItem(ItemStack useItem) {
+        if (!useItem.isEmpty())
+            this.useItemRemaining = net.neoforged.neoforge.event.EventHooks.onItemUseTick(this, useItem, this.getUseItemRemainingTicks());
+        if (this.getUseItemRemainingTicks() > 0)
         useItem.onUseTick(this.level(), this, this.getUseItemRemainingTicks());
-        if (--this.useItemRemaining == 0 && !this.level().isClientSide() && !useItem.useOnRelease()) {
+        if (--this.useItemRemaining <= 0 && !this.level().isClientSide() && !useItem.useOnRelease()) {
             this.completeUsingItem();
         }
     }
@@ -3403,8 +_,10 @@
     public void startUsingItem(InteractionHand hand) {
         ItemStack itemStack = this.getItemInHand(hand);
         if (!itemStack.isEmpty() && !this.isUsingItem()) {
+            int duration = net.neoforged.neoforge.event.EventHooks.onItemUseStart(this, itemStack, hand, itemStack.getUseDuration(this));
+            if (duration < 0) return; // Neo: Early return for negative values, as that indicates event cancellation.
             this.useItem = itemStack;
-            this.useItemRemaining = itemStack.getUseDuration(this);
+            this.useItemRemaining = duration;
             if (!this.level().isClientSide()) {
                 this.setLivingEntityFlag(1, true);
                 this.setLivingEntityFlag(2, hand == InteractionHand.OFF_HAND);
@@ -3470,7 +_,8 @@
                 this.releaseUsingItem();
             } else {
                 if (!this.useItem.isEmpty() && this.isUsingItem()) {
-                    ItemStack result = this.useItem.finishUsingItem(this.level(), this);
+                    ItemStack copy = this.useItem.copy();
+                    ItemStack result = net.neoforged.neoforge.event.EventHooks.onItemUseFinish(this, copy, getUseItemRemainingTicks(), this.useItem.finishUsingItem(this.level(), this));
                     if (result != this.useItem) {
                         this.setItemInHand(hand, result);
                     }
@@ -3504,7 +_,11 @@
         ItemStack itemInUsedHand = this.getItemInHand(this.getUsedItemHand());
         if (!this.useItem.isEmpty() && ItemStack.isSameItem(itemInUsedHand, this.useItem)) {
             this.useItem = itemInUsedHand;
+            if (!net.neoforged.neoforge.event.EventHooks.onUseItemStop(this, useItem, this.getUseItemRemainingTicks())) {
+            ItemStack copy = this instanceof Player ? useItem.copy() : null;
             this.useItem.releaseUsing(this.level(), this, this.getUseItemRemainingTicks());
+            if (copy != null && useItem.isEmpty()) net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem((Player)this, copy, getUsedItemHand());
+            }
             if (this.useItem.useOnRelease()) {
                 this.updatingUsingItem();
             }
@@ -3514,6 +_,7 @@
     }
 
     public void stopUsingItem() {
+        if (this.isUsingItem() && !this.useItem.isEmpty()) this.useItem.onStopUsing(this, useItemRemaining);
         if (!this.level().isClientSide()) {
             boolean wasUsingItem = this.isUsingItem();
             this.recentKineticEnemies = null;
@@ -3675,8 +_,8 @@
         }
 
         BlockState blockState = this.level().getBlockState(bedPosition);
-        if (blockState.getBlock() instanceof BedBlock) {
-            this.level().setBlock(bedPosition, blockState.setValue(BedBlock.OCCUPIED, true), 3);
+        if (blockState.isBed(level(), bedPosition, this)) {
+            blockState.setBedOccupied(level(), bedPosition, this, true);
         }
 
         this.setPose(Pose.SLEEPING);
@@ -3691,15 +_,17 @@
     }
 
     private boolean checkBedExists() {
-        return this.getSleepingPos().map(bedPosition -> this.level().getBlockState(bedPosition).getBlock() instanceof BedBlock).orElse(false);
+        // Neo: Overwrite the vanilla instanceof BedBlock check with isBed and fire the CanContinueSleepingEvent.
+        boolean hasBed = this.getSleepingPos().map(pos -> this.level().getBlockState(pos).isBed(this.level(), pos, this)).orElse(false);
+        return net.neoforged.neoforge.event.EventHooks.canEntityContinueSleeping(this, hasBed ? null : Player.BedSleepingProblem.OTHER_PROBLEM);
     }
 
     public void stopSleeping() {
         this.getSleepingPos().filter(this.level()::hasChunkAt).ifPresent(bedPosition -> {
             BlockState state = this.level().getBlockState(bedPosition);
-            if (state.getBlock() instanceof BedBlock) {
+            if (state.isBed(level(), bedPosition, this)) {
                 Direction facing = state.getValue(BedBlock.FACING);
-                this.level().setBlock(bedPosition, state.setValue(BedBlock.OCCUPIED, false), 3);
+                state.setBedOccupied(level(), bedPosition, this, false);
                 Vec3 standUp = BedBlock.findStandUpPosition(this.getType(), this.level(), bedPosition, facing, this.getYRot()).orElseGet(() -> {
                     BlockPos above = bedPosition.above();
                     return new Vec3(above.getX() + 0.5, above.getY() + 0.1, above.getZ() + 0.5);
@@ -3719,7 +_,9 @@
 
     public @Nullable Direction getBedOrientation() {
         BlockPos bedPos = this.getSleepingPos().orElse(null);
-        return bedPos != null ? BedBlock.getBedOrientation(this.level(), bedPos) : null;
+        if (bedPos == null) return null;
+        BlockState state = this.level().getBlockState(bedPos);
+        return !state.isBed(level(), bedPos, this) ? null : state.getBedDirection(level(), bedPos);
     }
 
     @Override
@@ -3728,7 +_,7 @@
     }
 
     public ItemStack getProjectile(ItemStack heldWeapon) {
-        return ItemStack.EMPTY;
+        return net.neoforged.neoforge.common.CommonHooks.getProjectile(this, heldWeapon, ItemStack.EMPTY);
     }
 
     private static byte entityEventForEquipmentBreak(EquipmentSlot equipmentSlot) {
@@ -3780,6 +_,8 @@
     }
 
     public final EquipmentSlot getEquipmentSlotForItem(ItemStack itemStack) {
+        final EquipmentSlot slot = itemStack.getEquipmentSlot();
+        if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
         Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
         return equippable != null && this.canUseSlot(equippable.slot()) ? equippable.slot() : EquipmentSlot.MAINHAND;
     }
