--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -145,6 +_,11 @@
     private Leashable.@Nullable LeashData leashData;
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
+    @Nullable
+    private EntitySpawnReason spawnType;
+    private boolean spawnCancelled = false;
+    /** Neo: Prevent immediate spawning from conversions to capture conversion results for events */
+    protected boolean preventConversionSpawns = false;
 
     protected Mob(EntityType<? extends Mob> type, Level level) {
         super(type, level);
@@ -250,7 +_,11 @@
     }
 
     public void setTarget(@Nullable LivingEntity target) {
-        this.target = this.asValidTarget(target);
+        target = this.asValidTarget(target);
+        net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.neoforged.neoforge.common.CommonHooks.onLivingChangeTarget(this, target, net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.MOB_TARGET);
+        if(!changeTargetEvent.isCanceled()) {
+             this.target = changeTargetEvent.getNewAboutToBeSetTarget();
+        }
     }
 
     @Override
@@ -346,6 +_,12 @@
         if (!this.level().isClientSide() && this.tickCount % 5 == 0) {
             this.updateControlFlags();
         }
+
+        // Neo: Animal armor tick patch
+        if (this.canUseSlot(EquipmentSlot.BODY)) {
+            ItemStack stack = this.getBodyArmorItem();
+            if (stack.has(DataComponents.EQUIPPABLE)) stack.onAnimalArmorTick(level(), this);
+        }
     }
 
     protected void updateControlFlags() {
@@ -389,6 +_,9 @@
         if (this.isNoAi()) {
             output.putBoolean("NoAI", this.isNoAi());
         }
+        if (this.spawnType != null) {
+            output.putString("neoforge:spawn_type", this.spawnType.name());
+        }
     }
 
     @Override
@@ -407,6 +_,14 @@
         this.lootTable = input.read("DeathLootTable", LootTable.KEY_CODEC);
         this.lootTableSeed = input.getLongOr("DeathLootTableSeed", 0L);
         this.setNoAi(input.getBooleanOr("NoAI", false));
+
+        input.getString("neoforge:spawn_type")
+                .ifPresent(spawnType -> {
+                    try {
+                        this.spawnType = EntitySpawnReason.valueOf(spawnType);
+                    } catch (Exception ignored) {
+                    }
+                });
     }
 
     @Override
@@ -465,7 +_,7 @@
             && this.canPickUpLoot()
             && this.isAlive()
             && !this.dead
-            && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+            && net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverLevel, this)) {
             Vec3i pickupReach = this.getPickupReach();
 
             for (ItemEntity entity : this.level()
@@ -647,6 +_,8 @@
     private double getApproximateAttributeWith(ItemStack itemStack, Holder<Attribute> attribute, EquipmentSlot slot) {
         double baseValue = this.getAttributes().hasAttribute(attribute) ? this.getAttributeBaseValue(attribute) : 0.0;
         ItemAttributeModifiers attributeModifiers = itemStack.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
+        // Neo: Respect gameplay modifiers
+        attributeModifiers = itemStack.getAttributeModifiers();
         return attributeModifiers.compute(attribute, baseValue, slot);
     }
 
@@ -1082,6 +_,11 @@
         }
     }
 
+    /**
+     * @deprecated Override-Only. External callers should call via {@link net.neoforged.neoforge.event.EventHooks#finalizeMobSpawn}.
+     */
+    @Deprecated
+    @org.jetbrains.annotations.ApiStatus.OverrideOnly
     public @Nullable SpawnGroupData finalizeSpawn(
         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData groupData
     ) {
@@ -1094,6 +_,7 @@
         }
 
         this.setLeftHanded(random.nextFloat() < 0.05F);
+        this.spawnType = spawnReason;
         return groupData;
     }
 
@@ -1242,6 +_,7 @@
 
         params.type().convert(this, newMob, params);
         afterConversion.finalizeConversion(newMob);
+        if (!preventConversionSpawns)
         if (this.level() instanceof ServerLevel serverLevel) {
             serverLevel.addFreshEntity(newMob);
         }
@@ -1403,12 +_,22 @@
     }
 
     @Override
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(TagKey<Fluid> type) {
+        this.jumpInLiquidInternal(() -> super.jumpInLiquid(type));
+    }
+
+    private void jumpInLiquidInternal(Runnable onSuper) {
         if (this.getNavigation().canFloat()) {
-            super.jumpInLiquid(type);
+            onSuper.run();
         } else {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.3, 0.0));
         }
+    }
+
+    @Override
+    public void jumpInFluid(net.neoforged.neoforge.fluids.FluidType type) {
+        this.jumpInLiquidInternal(() -> super.jumpInFluid(type));
     }
 
     @VisibleForTesting
