--- a/net/minecraft/world/entity/Shearable.java
+++ b/net/minecraft/world/entity/Shearable.java
@@ -6,8 +_,14 @@
 
 import net.neoforged.neoforge.common.IShearable;
 
+/// @deprecated Neo: Use [net.neoforged.neoforge.common.IShearable] instead.
+@Deprecated
 public interface Shearable extends IShearable {
+    /// @deprecated Neo: Use [net.neoforged.neoforge.common.IShearable#onSheared(net.minecraft.world.entity.player.Player, ItemStack, net.minecraft.world.level.Level, net.minecraft.core.BlockPos)] instead.
+    @Deprecated
     void shear(ServerLevel level, SoundSource soundSource, ItemStack tool);
 
+    /// @deprecated Neo: Use [net.neoforged.neoforge.common.IShearable#isShearable(net.minecraft.world.entity.player.Player, ItemStack, net.minecraft.world.level.Level, net.minecraft.core.BlockPos)] instead.
+    @Deprecated
     boolean readyForShearing();
 }
