--- a/net/minecraft/gametest/framework/GameTestInfo.java
+++ b/net/minecraft/gametest/framework/GameTestInfo.java
@@ -206,6 +_,7 @@
       this.structureBlockEntity = StructureUtils.spawnStructure(this.getStructureName(), p_127620_, this.getRotation(), p_127621_, this.level, false);
       this.structureBlockPos = this.structureBlockEntity.getBlockPos();
       this.structureBlockEntity.setStructureName(this.getTestName());
+      this.structureBlockEntity.setMetaData(this.getTestName()); // Neo: set the test name as metadata as it isn't always RL-compatible
       StructureUtils.addCommandBlockAndButtonToStartTest(this.structureBlockPos, new BlockPos(1, 0, -1), this.getRotation(), this.level);
       this.listeners.forEach(p_127630_ -> p_127630_.testStructureLoaded(this));
    }
