--- a/net/minecraft/gametest/framework/GameTestMainUtil.java
+++ b/net/minecraft/gametest/framework/GameTestMainUtil.java
@@ -69,6 +_,7 @@
 
             Bootstrap.bootStrap();
             Util.startTimerHackThread();
+            net.neoforged.neoforge.server.loading.ServerModLoader.load(true); // Load mods before the level and packs are loaded
             String s = optionset.valueOf(universe);
             createOrResetDir(s);
             p_397859_.accept(s);
