--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -149,6 +_,7 @@
     public boolean initServer() {
         this.setPlayerList(new PlayerList(this, this.registries(), this.playerDataStorage, 1) {
         });
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
         this.loadLevel();
         ServerLevel serverlevel = this.overworld();
         this.testBatches = Lists.newArrayList(GameTestBatchFactory.fromTestFunction(this.testFunctions, serverlevel));
@@ -156,6 +_,7 @@
         int i = 20000000;
         serverlevel.setWeatherParameters(20000000, 20000000, false, false);
         LOGGER.info("Started game test server");
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
 
