--- a/net/minecraft/gametest/framework/GameTestSequence.java
+++ b/net/minecraft/gametest/framework/GameTestSequence.java
@@ -70,6 +_,15 @@
       return gametestsequence$condition;
    }
 
+   public <T> net.neoforged.neoforge.gametest.ParametrizedGameTestSequence<T> thenMap(Supplier<T> value) {
+      return new net.neoforged.neoforge.gametest.ParametrizedGameTestSequence<>(parent, this, value);
+   }
+
+   public GameTestSequence thenSequence(java.util.function.Consumer<GameTestSequence> consumer) {
+      consumer.accept(parent.createSequence());
+      return this;
+   }
+
    public void tickAndContinue(long p_127778_) {
       try {
          this.tick(p_127778_);
