--- a/net/minecraft/gametest/framework/GameTestHelper.java
+++ b/net/minecraft/gametest/framework/GameTestHelper.java
@@ -836,7 +_,8 @@
 
     public void forEveryBlockInStructure(Consumer<BlockPos> p_177293_) {
         AABB aabb = this.getRelativeBounds();
-        BlockPos.MutableBlockPos.betweenClosedStream(aabb.move(0.0, 1.0, 0.0)).forEach(p_177293_);
+        // Neo: Shrink AABB by one as an aabb surrounding two blocks is increased by one compared to the actual positions
+        BlockPos.MutableBlockPos.betweenClosedStream(aabb.contract(1.0, 1.0, 1.0)).forEach(p_177293_);
     }
 
     public void onEachTick(Runnable p_177424_) {
