--- a/net/minecraft/gametest/framework/TestCommand.java
+++ b/net/minecraft/gametest/framework/TestCommand.java
@@ -328,7 +_,7 @@
             } else {
                 BlockPos blockpos1 = blockpos.subtract(optional.get());
                 String s = blockpos1.getX() + ", " + blockpos1.getY() + ", " + blockpos1.getZ();
-                String s1 = structureblockentity.getMetaData();
+                String s1 = structureblockentity.getMetaData().isBlank() ? structureblockentity.getStructureName() : structureblockentity.getMetaData(); // Neo: use the metadata for the structure name
                 Component component = Component.literal(s)
                     .setStyle(
                         Style.EMPTY
