--- a/net/minecraft/client/gui/font/providers/GlyphProviderType.java
+++ b/net/minecraft/client/gui/font/providers/GlyphProviderType.java
@@ -8,6 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
+@net.neoforged.fml.common.asm.enumextension.NamedEnum
 public enum GlyphProviderType implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     BITMAP("bitmap", BitmapProvider.Definition.CODEC),
     TTF("ttf", TrueTypeGlyphProviderDefinition.CODEC),
@@ -31,5 +_,9 @@
 
     public MapCodec<? extends GlyphProviderDefinition> mapCodec() {
         return this.codec;
+    }
+
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(GlyphProviderType.class);
     }
 }
