--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -163,6 +201,16 @@
         @Override
         public BakedQuad.SpriteInfo spriteInfo(BakedQuad.SpriteInfo sprite) {
             return this.spriteInfos.intern(sprite);
+        }
+
+        @Override
+        public net.neoforged.neoforge.client.model.quad.BakedNormals normals(net.neoforged.neoforge.client.model.quad.BakedNormals normals) {
+            return this.normals.intern(normals);
+        }
+
+        @Override
+        public net.neoforged.neoforge.client.model.quad.BakedColors colors(net.neoforged.neoforge.client.model.quad.BakedColors colors) {
+            return this.colors.intern(colors);
         }
     }
 
