--- a/net/minecraft/core/Direction.java
+++ b/net/minecraft/core/Direction.java
@@ -463,6 +_,22 @@
       return direction;
    }
 
+   /** NeoForge: Version of {@link #getNearest(float, float, float)} that accounts for potential rounding errors. */
+   public static Direction getNearestStable(float p_122373_, float p_122374_, float p_122375_) {
+      Direction direction = NORTH;
+      float f = Float.MIN_VALUE;
+
+      for(Direction direction1 : f_122346_) {
+         float f1 = p_122373_ * (float)direction1.f_122345_.m_123341_() + p_122374_ * (float)direction1.f_122345_.m_123342_() + p_122375_ * (float)direction1.f_122345_.m_123343_();
+         if (f1 > f + com.mojang.math.Constants.f_142770_) {
+            f = f1;
+            direction = direction1;
+         }
+      }
+
+      return direction;
+   }
+
    public String toString() {
       return this.f_122342_;
    }
