--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -11,7 +_,22 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.TagKey;
 
-public interface Holder<T> {
+public interface Holder<T> extends java.util.function.Supplier<T>, net.minecraftforge.registries.tags.IReverseTag<T>, net.minecraftforge.registries.attachment.IWithAttachments<T> {
+   @Override
+   default boolean containsTag(TagKey<T> key) {
+      return this.m_203656_(key);
+   }
+
+   @Override
+   default Stream<TagKey<T>> getTagKeys() {
+      return this.m_203616_();
+   }
+
+   @Override
+   default T get() {
+      return this.m_203334_();
+   }
+
    T m_203334_();
 
    boolean m_203633_();
@@ -191,8 +_,17 @@
          this.f_205749_ = Set.copyOf(p_205770_);
       }
 
+      /** {@inheritDoc} **/ @Override
+      public <A> @Nullable A getAttachment(net.minecraftforge.registries.attachment.AttachmentTypeKey<A> type) {
+         return ((HolderLookup.RegistryLookup<T>) f_254697_).getAttachment(type, this);
+      }
+
       public Stream<TagKey<T>> m_203616_() {
          return this.f_205749_.stream();
+      }
+
+      public Type getType() {
+         return this.f_205750_;
       }
 
       public String toString() {
