--- a/net/minecraft/nbt/CompoundTag.java
+++ b/net/minecraft/nbt/CompoundTag.java
@@ -51,13 +_,19 @@
             return var3;
         }
 
+        private static byte readNamedTagType(DataInput input, NbtAccounter accounter) throws IOException {
+            accounter.accountBytes(2);
+            return input.readByte();
+        }
+
         private static CompoundTag loadCompound(DataInput input, NbtAccounter accounter) throws IOException {
             accounter.accountBytes(48L);
             Map<String, Tag> values = Maps.newHashMap();
 
             byte tagType;
-            while ((tagType = input.readByte()) != 0) {
-                String key = readString(input, accounter);
+            while ((tagType = readNamedTagType(input, accounter)) != 0) {
+                String key = accounter.readUTF(input.readUTF());
+                accounter.accountBytes(4); //Forge: 4 extra bytes for the object allocation.
                 Tag tag = CompoundTag.readNamedTagData(TagTypes.getType(tagType), key, input, accounter);
                 if (values.put(key, tag) == null) {
                     accounter.accountBytes(36L);
@@ -173,6 +_,17 @@
         this(new HashMap<>());
     }
 
+    /**
+     * Neo: create a compound tag that is generally suitable to hold the given amount of entries
+     * without needing to resize the internal map.
+     *
+     * @param expectedEntries the expected number of entries that the compound tag will have
+     * @see HashMap#newHashMap(int)
+     */
+    public CompoundTag(int expectedEntries) {
+        this(HashMap.newHashMap(expectedEntries));
+    }
+
     @Override
     public void write(DataOutput output) throws IOException {
         for (String key : this.tags.keySet()) {
@@ -227,6 +_,7 @@
     }
 
     public @Nullable Tag put(String name, Tag tag) {
+        if (tag == null) throw new IllegalArgumentException("Invalid null NBT value with key " + name);
         return this.tags.put(name, tag);
     }
 
