--- a/net/minecraft/nbt/NbtIo.java
+++ b/net/minecraft/nbt/NbtIo.java
@@ -178,10 +_,12 @@
 
     private static Tag readUnnamedTag(DataInput input, NbtAccounter accounter) throws IOException {
         byte b0 = input.readByte();
+        accounter.accountBytes(1); // Forge: Count everything!
         if (b0 == 0) {
             return EndTag.INSTANCE;
         } else {
-            StringTag.skipString(input);
+            accounter.readUTF(input.readUTF()); //Forge: Count this string.
+            accounter.accountBytes(4); //Forge: 4 extra bytes for the object allocation.
             return readTagSafe(input, accounter, b0);
         }
     }
