--- a/net/minecraft/nbt/NbtIo.java
+++ b/net/minecraft/nbt/NbtIo.java
@@ -169,11 +_,13 @@
     @VisibleForTesting
     public static Tag readUnnamedTag(DataInput input, NbtAccounter accounter) throws IOException {
         byte type = input.readByte();
+        accounter.accountBytes(1); // Neo: Count the tag type
         if (type == 0) {
             return EndTag.INSTANCE;
         }
 
-        StringTag.skipString(input);
+        accounter.readUTF(input.readUTF()); // Neo: Count this string.
+        accounter.accountBytes(4); // Neo: 4 extra bytes for the object allocation.
         return readTagSafe(input, accounter, type);
     }
 
