--- a/net/minecraft/nbt/StringTag.java
+++ b/net/minecraft/nbt/StringTag.java
@@ -20,7 +_,7 @@
         private static String readAccounted(DataInput input, NbtAccounter accounter) throws IOException {
             accounter.accountBytes(36L);
             String data = input.readUTF();
-            accounter.accountBytes(2L, data.length());
+            accounter.readUTF(data);
             return data;
         }
 
