--- a/net/minecraft/nbt/StringTag.java
+++ b/net/minecraft/nbt/StringTag.java
@@ -20,7 +_,7 @@
         private static String readAccounted(DataInput input, NbtAccounter nbtAccounter) throws IOException {
             nbtAccounter.accountBytes(36L);
             String s = input.readUTF();
-            nbtAccounter.accountBytes(2L, (long)s.length());
+            nbtAccounter.readUTF(s);
             return s;
         }
 
