--- a/net/minecraft/server/rcon/thread/RconClient.java
+++ b/net/minecraft/server/rcon/thread/RconClient.java
@@ -123,13 +_,14 @@
     }
 
     private void sendCmdResponse(int id, String message) throws IOException {
-        int i = message.length();
-
+        byte[] whole = message.getBytes(StandardCharsets.UTF_8);
+        int i = whole.length;
+        int start = 0;
         do {
             int j = 4096 <= i ? 4096 : i;
-            this.send(id, 0, message.substring(0, j));
-            message = message.substring(j);
-            i = message.length();
+            this.send(id, 0, new String(java.util.Arrays.copyOfRange(whole, start, j+start), StandardCharsets.UTF_8));
+            i -= j;
+            start += j;
         } while (0 != i);
     }
 
