--- a/net/minecraft/data/models/model/TexturedModel.java
+++ b/net/minecraft/data/models/model/TexturedModel.java
@@ -8,7 +_,7 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.level.block.Block;
 
-public class TexturedModel {
+public class TexturedModel implements net.neoforged.neoforge.common.extensions.ITexturedModelExtension {
     public static final TexturedModel.Provider CUBE = createDefault(TextureMapping::cube, ModelTemplates.CUBE_ALL);
     public static final TexturedModel.Provider CUBE_INNER_FACES = createDefault(TextureMapping::cube, ModelTemplates.CUBE_ALL_INNER_FACES);
     public static final TexturedModel.Provider CUBE_MIRRORED = createDefault(TextureMapping::cube, ModelTemplates.CUBE_MIRRORED_ALL);
@@ -74,7 +_,7 @@
     }
 
     @FunctionalInterface
-    public interface Provider {
+    public interface Provider extends net.neoforged.neoforge.common.extensions.ITexturedModelExtension.Provider {
         TexturedModel get(Block p_125965_);
 
         default ResourceLocation create(Block p_125957_, BiConsumer<ResourceLocation, Supplier<JsonElement>> p_125958_) {
