--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -306,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.BRUSH) && this.brushOffScute(player, itemStack)) {
+        if (itemStack.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.BRUSH_BRUSH) && this.brushOffScute(player, itemStack)) {
             itemStack.hurtAndBreak(16, player, hand.asEquipmentSlot());
             return InteractionResult.SUCCESS;
         } else {
