--- a/net/minecraft/world/entity/npc/villager/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/villager/VillagerTrades.java
@@ -1389,6 +_,7 @@
         private final int villagerXp;
 
         public EmeraldsForVillagerTypeItem(int p_480321_, int p_480095_, int p_478325_, Map<ResourceKey<VillagerType>, Item> p_481157_) {
+            if (false) // Neo: disable this check so that mods can add custom villager types
             BuiltInRegistries.VILLAGER_TYPE.registryKeySet().stream().filter(p_480286_ -> !p_481157_.containsKey(p_480286_)).findAny().ifPresent(p_478321_ -> {
                 throw new IllegalStateException("Missing trade for villager type: " + p_478321_);
             });
@@ -1405,7 +_,9 @@
                 if (resourcekey == null) {
                     return null;
                 } else {
-                    ItemCost itemcost = new ItemCost(this.trades.get(resourcekey), this.cost);
+                    Item item = this.trades.get(resourcekey);
+                    if (item == null) return null;
+                    ItemCost itemcost = new ItemCost(item, this.cost);
                     return new MerchantOffer(itemcost, new ItemStack(Items.EMERALD), this.maxUses, this.villagerXp, 0.05F);
                 }
             } else {
