--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -701,6 +718,8 @@
                 profiler.push("renderTranslucentFeatures");
                 this.featureRenderDispatcher.renderTranslucentFeatures();
                 bufferSource.endBatch();
+                profiler.popPush("neoforge_render_after_translucent_features");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterTranslucentFeatures(this, levelRenderState, poseStack, modelViewMatrix, this.getRenderableSections()));
                 profiler.popPush("destroyProgress");
                 this.renderBlockDestroyAnimation(poseStack, crumblingBufferSource, levelRenderState);
                 crumblingBufferSource.endBatch();
@@ -943,7 +1015,7 @@
             poseStack.pushPose();
             BlockPos pos = state.blockPos;
             poseStack.translate(pos.getX() - camX, pos.getY() - camY, pos.getZ() - camZ);
-            this.minecraft.getBlockRenderer().renderBreakingTexture(state.blockState, pos, poseStack, bufferSource, state.progress);
+            this.minecraft.getBlockRenderer().renderBreakingTexture(state, state.blockState, pos, poseStack, bufferSource, state.progress);
             poseStack.popPose();
         }
     }
