--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -95,6 +158,9 @@
     }
 
     public SlotDisplay display() {
+        if (this.customIngredient != null) {
+            return this.customIngredient.display();
+        }
         return (SlotDisplay)this.values
             .unwrap()
             .map(SlotDisplay.TagSlotDisplay::new, l -> new SlotDisplay.Composite(l.stream().map(Ingredient::displayForSingleItem).toList()));
