--- a/net/minecraft/world/item/crafting/RecipeMap.java
+++ b/net/minecraft/world/item/crafting/RecipeMap.java
@@ -31,6 +31,25 @@
         }
 
         return new RecipeMap(byType.build(), byKey.build());
+    }
+
+    // Neo: handle ordering according to any recipe priorities
+    public void order(it.unimi.dsi.fastutil.objects.Object2IntMap<ResourceKey<net.minecraft.world.item.crafting.Recipe<?>>> recipePriorities) {
+        it.unimi.dsi.fastutil.ints.Int2ObjectRBTreeMap<java.util.List<RecipeHolder<?>>> priorityBuilder = new it.unimi.dsi.fastutil.ints.Int2ObjectRBTreeMap<>();
+        com.google.common.collect.LinkedListMultimap<RecipeType<?>, RecipeHolder<?>> finalBuilder = com.google.common.collect.LinkedListMultimap.create();
+
+        for (RecipeHolder<?> recipeholder : this.byKey.values()) {
+            int priority = recipePriorities.getOrDefault(recipeholder.id(), 0);
+            priorityBuilder.computeIfAbsent(priority, p -> new java.util.ArrayList<>()).add(recipeholder);
+        }
+
+        for (var list : priorityBuilder.reversed().values()) {
+            for (RecipeHolder<?> recipeHolder : list) {
+                finalBuilder.put(recipeHolder.value().getType(), recipeHolder);
+            }
+        }
+
+        this.byType = ImmutableMultimap.copyOf(finalBuilder);
     }
 
     public <I extends RecipeInput, T extends Recipe<I>> Collection<RecipeHolder<T>> byType(RecipeType<T> type) {
