--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -188,7 +_,7 @@
             this.shakeTime--;
         }
 
-        if (this.isInWaterOrRain() || blockstate.is(Blocks.POWDER_SNOW)) {
+        if (this.isInWaterOrRain() || blockstate.is(Blocks.POWDER_SNOW) || this.isInFluidType((fluidType, height) -> this.canFluidExtinguish(fluidType))) {
             this.clearFire();
         }
 
@@ -224,7 +_,9 @@
                     }
                 }
 
-                if (hitresult != null && !flag) {
+                if (hitresult != null && hitresult.getType() != HitResult.Type.MISS && !flag) {
+                    if (net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, hitresult))
+                        break;
                     ProjectileDeflection projectiledeflection = this.hitTargetOrDeflectSelf(hitresult);
                     this.hasImpulse = true;
                     if (projectiledeflection != ProjectileDeflection.NONE) {
@@ -590,7 +_,7 @@
     protected boolean tryPickup(Player p_150121_) {
         return switch (this.pickup) {
             case DISALLOWED -> false;
-            case ALLOWED -> p_150121_.getInventory().add(this.getPickupItem());
+            case ALLOWED -> net.neoforged.neoforge.event.EventHooks.addToInventory(p_150121_, this.getPickupItem(), new net.neoforged.neoforge.event.entity.item.ObtainItemEvent.ProjectileSource(this));
             case CREATIVE_ONLY -> p_150121_.hasInfiniteMaterials();
         };
     }
