--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -191,7 +_,7 @@
             this.shakeTime--;
         }
 
-        if (this.isInWaterOrRain() || blockstate.is(Blocks.POWDER_SNOW)) {
+        if (this.isInWaterOrRain() || blockstate.is(Blocks.POWDER_SNOW) || this.isInFluidType((fluidType, height) -> this.canFluidExtinguish(fluidType))) {
             this.clearFire();
         }
 
@@ -275,11 +_,15 @@
 
             if (entityhitresult == null) {
                 if (this.isAlive() && p_371761_.getType() != HitResult.Type.MISS) {
+                    if (net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, p_371761_))
+                        break;
                     this.hitTargetOrDeflectSelf(p_371761_);
                     this.hasImpulse = true;
                 }
                 break;
-            } else if (this.isAlive() && !this.noPhysics) {
+            } else if (this.isAlive() && !this.noPhysics && entityhitresult.getType() != HitResult.Type.MISS) {
+                if (net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, entityhitresult))
+                    break;
                 ProjectileDeflection projectiledeflection = this.hitTargetOrDeflectSelf(entityhitresult);
                 this.hasImpulse = true;
                 if (this.getPierceLevel() > 0 && projectiledeflection == ProjectileDeflection.NONE) {
