--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -72,7 +_,8 @@
         if (this.level().isClientSide || (entity == null || !entity.isRemoved()) && this.level().hasChunkAt(this.blockPosition())) {
             HitResult hitresult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity, this.getClipType());
             Vec3 vec3;
-            if (hitresult.getType() != HitResult.Type.MISS) {
+            boolean impacted = hitresult.getType() != HitResult.Type.MISS && !net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, hitresult);
+            if (impacted) {
                 vec3 = hitresult.getLocation();
             } else {
                 vec3 = this.position().add(this.getDeltaMovement());
@@ -86,7 +_,7 @@
                 this.igniteForSeconds(1.0F);
             }
 
-            if (hitresult.getType() != HitResult.Type.MISS && this.isAlive()) {
+            if (hitresult.getType() != HitResult.Type.MISS && this.isAlive() && impacted) {
                 this.hitTargetOrDeflectSelf(hitresult);
             }
 
