--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -352,7 +_,7 @@
     @Override
     public boolean mayInteract(ServerLevel p_376318_, BlockPos p_150168_) {
         Entity entity = this.getOwner();
-        return entity instanceof Player ? entity.mayInteract(p_376318_, p_150168_) : entity == null || p_376318_.getGameRules().get(GameRules.MOB_GRIEFING);
+        return entity instanceof Player ? entity.mayInteract(p_376318_, p_150168_) : entity == null || net.neoforged.neoforge.event.EventHooks.canEntityGrief(p_376318_, entity);
     }
 
     public boolean mayBreak(ServerLevel p_376471_) {
