--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -284,7 +_,7 @@
         Entity entity = this.getOwner();
         return entity instanceof Player
             ? entity.mayInteract(level, pos)
-            : entity == null || level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+            : entity == null || net.neoforged.neoforge.event.EventHooks.canEntityGrief(level, entity);
     }
 
     public boolean mayBreak(Level level) {
