--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -364,7 +_,7 @@
         Entity entity = this.getOwner();
         return entity instanceof Player
             ? entity.mayInteract(p_376318_, p_150168_)
-            : entity == null || p_376318_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+            : entity == null || net.neoforged.neoforge.event.EventHooks.canEntityGrief(p_376318_, entity);
     }
 
     public boolean mayBreak(ServerLevel p_376471_) {
