--- a/net/minecraft/world/entity/projectile/ProjectileUtil.java
+++ b/net/minecraft/world/entity/projectile/ProjectileUtil.java
@@ -113,7 +_,7 @@
                 Vec3 location = clipPoint.get();
                 double dd = from.distanceToSqr(location);
                 if (dd < nearest || nearest == 0.0) {
-                    if (entity.getRootVehicle() == except.getRootVehicle()) {
+                    if (entity.getRootVehicle() == except.getRootVehicle() && !entity.canRiderInteract()) {
                         if (nearest == 0.0) {
                             hovered = entity;
                             hoveredPos = location;
@@ -242,8 +_,13 @@
         }
     }
 
+    @Deprecated // Forge: Use the version below that takes in a Predicate<Item> instead of an Item
     public static InteractionHand getWeaponHoldingHand(LivingEntity mob, Item weaponItem) {
         return mob.getMainHandItem().is(weaponItem) ? InteractionHand.MAIN_HAND : InteractionHand.OFF_HAND;
+    }
+
+    public static InteractionHand getWeaponHoldingHand(LivingEntity livingEntity, Predicate<Item> itemPredicate) {
+        return itemPredicate.test(livingEntity.getMainHandItem().getItem()) ? InteractionHand.MAIN_HAND : InteractionHand.OFF_HAND;
     }
 
     public static AbstractArrow getMobArrow(LivingEntity mob, ItemStack projectile, float power, @Nullable ItemStack firedFromWeapon) {
