--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -104,6 +_,8 @@
                 Vec3 vec3 = this.oldPosition();
                 if (entity instanceof ServerPlayer serverplayer) {
                     if (serverplayer.connection.isAcceptingMessages()) {
+                        net.neoforged.neoforge.event.entity.EntityTeleportEvent.EnderPearl event = net.neoforged.neoforge.event.EventHooks.onEnderPearlLand(serverplayer, this.getX(), this.getY(), this.getZ(), this, 5.0F, p_37504_);
+                        if (!event.isCanceled()) { // Don't indent to lower patch size
                         if (this.random.nextFloat() < 0.05F && serverlevel.isSpawningMonsters()) {
                             Endermite endermite = EntityType.ENDERMITE.create(serverlevel, EntitySpawnReason.TRIGGERED);
                             if (endermite != null) {
@@ -118,16 +_,17 @@
 
                         ServerPlayer serverplayer1 = serverplayer.teleport(
                             new TeleportTransition(
-                                serverlevel, vec3, Vec3.ZERO, 0.0F, 0.0F, Relative.union(Relative.ROTATION, Relative.DELTA), TeleportTransition.DO_NOTHING
+                                serverlevel, event.getTarget(), entity.getDeltaMovement(), entity.getYRot(), entity.getXRot(), TeleportTransition.DO_NOTHING
                             )
                         );
                         if (serverplayer1 != null) {
                             serverplayer1.resetFallDistance();
                             serverplayer1.resetCurrentImpulseContext();
-                            serverplayer1.hurtServer(serverplayer.level(), this.damageSources().enderPearl(), 5.0F);
+                            serverplayer1.hurtServer(serverplayer.level(), this.damageSources().fall(), event.getAttackDamage());
                         }
 
                         this.playSound(serverlevel, vec3);
+                        } //Forge: End
                     }
                 } else {
                     Entity entity1 = entity.teleport(
