--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -66,6 +_,8 @@
 
                 if (entity instanceof ServerPlayer serverplayer) {
                     if (serverplayer.connection.isAcceptingMessages()) {
+                        net.neoforged.neoforge.event.entity.EntityTeleportEvent.EnderPearl event = net.neoforged.neoforge.event.EventHooks.onEnderPearlLand(serverplayer, this.getX(), this.getY(), this.getZ(), this, 5.0F, p_37504_);
+                        if (!event.isCanceled()) { // Don't indent to lower patch size
                         if (this.random.nextFloat() < 0.05F && serverlevel.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
                             Endermite endermite = EntityType.ENDERMITE.create(serverlevel);
                             if (endermite != null) {
@@ -76,13 +_,14 @@
 
                         entity.changeDimension(
                             new DimensionTransition(
-                                serverlevel, this.position(), entity.getDeltaMovement(), entity.getYRot(), entity.getXRot(), DimensionTransition.DO_NOTHING
+                                serverlevel, event.getTarget(), entity.getDeltaMovement(), entity.getYRot(), entity.getXRot(), DimensionTransition.DO_NOTHING
                             )
                         );
                         entity.resetFallDistance();
                         serverplayer.resetCurrentImpulseContext();
-                        entity.hurt(this.damageSources().fall(), 5.0F);
+                        entity.hurt(this.damageSources().fall(), event.getAttackDamage());
                         this.playSound(serverlevel, this.position());
+                        } //Forge: End
                     }
                 } else {
                     entity.changeDimension(
