--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -49,7 +_,7 @@
         this.applyInertia();
         HitResult result = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
         Vec3 newPosition;
-        if (result.getType() != HitResult.Type.MISS) {
+        if (result.getType() != HitResult.Type.MISS && !net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, result)) {
             newPosition = result.getLocation();
         } else {
             newPosition = this.position().add(this.getDeltaMovement());
