--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -60,7 +60,7 @@
          }
       }
 
-      if (hitresult.getType() != HitResult.Type.MISS && !flag) {
+      if (hitresult.getType() != HitResult.Type.MISS && !flag && !net.neoforged.neoforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
          this.onHit(hitresult);
       }
 
