--- a/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -49,7 +_,7 @@
     public float getBlockExplosionResistance(
         Explosion explosion, BlockGetter level, BlockPos pos, BlockState blockState, FluidState fluidState, float explosionPower
     ) {
-        return this.isDangerous() && WitherBoss.canDestroy(blockState) ? Math.min(0.8F, explosionPower) : explosionPower;
+        return this.isDangerous() && blockState.canEntityDestroy(level, pos, this) ? Math.min(0.8F, explosionPower) : explosionPower;
     }
 
     @Override
