--- a/com/mojang/realmsclient/gui/screens/RealmsSelectWorldTemplateScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsSelectWorldTemplateScreen.java
@@ -366,9 +_,9 @@
             this(Collections.emptyList());
         }
 
-        public WorldTemplateList(Iterable<WorldTemplate> templates) {
+        public WorldTemplateList(Iterable<WorldTemplate> p_313798_) {
             super(RealmsSelectWorldTemplateScreen.this.width, RealmsSelectWorldTemplateScreen.this.height - 33 - RealmsSelectWorldTemplateScreen.this.getHeaderHeight(), RealmsSelectWorldTemplateScreen.this.getHeaderHeight(), 46);
-            templates.forEach(this::addEntry);
+            p_313798_.forEach(this::addEntry);
         }
 
         public void addEntry(WorldTemplate template) {
