--- a/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
@@ -128,7 +_,7 @@
     private void drawIcons(GuiGraphics p_282966_) {
         int i = this.numberOfPendingInvites;
         int j = 24;
-        int k = this.height / 4 + 48;
+        int k = this.height / 4 + (inTitleScreen() ? 32 : 48);
         int l = this.width / 2 + 100;
         int i1 = k + 48 + 2;
         int j1 = l - 3;
