--- a/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsNotificationsScreen.java
@@ -131,7 +_,7 @@
     private void extractIcons(GuiGraphicsExtractor graphics) {
         int pendingInvitesCount = this.numberOfPendingInvites;
         int spacing = 24;
-        int topPos = this.height / 4 + 48;
+        int topPos = this.height / 4 + (inTitleScreen() ? 32 : 48);
         int buttonRight = this.width / 2 + 100;
         int baseY = topPos + 48 + 2;
         int iconRight = buttonRight - 3;
