--- a/com/mojang/blaze3d/systems/CommandEncoder.java
+++ b/com/mojang/blaze3d/systems/CommandEncoder.java
@@ -32,6 +32,8 @@
 
     void clearDepthTexture(GpuTexture depthTexture, double clearDepth);
 
+    void clearStencilTexture(GpuTexture texture, int value);
+
     void writeToBuffer(GpuBufferSlice destination, ByteBuffer data);
 
     GpuBuffer.MappedView mapBuffer(GpuBuffer buffer, boolean read, boolean write);
