--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -121,8 +_,7 @@
             CrashReport.preload();
             logger = LogUtils.getLogger();
             s1 = "Bootstrap";
-            Bootstrap.bootStrap();
-            ClientBootstrap.bootstrap();
+            net.neoforged.fml.loading.BackgroundWaiter.runAndTick(() -> Bootstrap.bootStrap(), net.neoforged.fml.loading.FMLLoader.progressWindowTick);
             GameLoadTimesEvent.INSTANCE.setBootstrapTime(Bootstrap.bootstrapDuration.get());
             Bootstrap.validate();
             s1 = "Argument parsing";
