--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -76,12 +_,13 @@
         OptionSpec<Integer> optionspec12 = optionparser.accepts("proxyPort").withRequiredArg().defaultsTo("8080").ofType(Integer.class);
         OptionSpec<String> optionspec13 = optionparser.accepts("proxyUser").withRequiredArg();
         OptionSpec<String> optionspec14 = optionparser.accepts("proxyPass").withRequiredArg();
-        OptionSpec<String> optionspec15 = optionparser.accepts("username").withRequiredArg().defaultsTo("Player" + System.currentTimeMillis() % 1000L);
+        OptionSpec<String> optionspec15 = optionparser.accepts("username").withRequiredArg().defaultsTo(net.neoforged.fml.loading.FMLEnvironment.isProduction() ? "Player" + System.currentTimeMillis() % 1000L : "Dev").withValuesConvertedBy(net.neoforged.neoforge.client.ClientHooks.convertUsername());
         OptionSpec<Void> optionspec16 = optionparser.accepts("offlineDeveloperMode");
         OptionSpec<String> optionspec17 = optionparser.accepts("uuid").withRequiredArg();
         OptionSpec<String> optionspec18 = optionparser.accepts("xuid").withOptionalArg().defaultsTo("");
         OptionSpec<String> optionspec19 = optionparser.accepts("clientId").withOptionalArg().defaultsTo("");
-        OptionSpec<String> optionspec20 = optionparser.accepts("accessToken").withRequiredArg().required();
+        // Neo: Make accessToken optional in dev
+        OptionSpec<String> optionspec20 = net.neoforged.neoforge.client.ClientHooks.optionalInDev(optionparser.accepts("accessToken").withRequiredArg(), "0");
         OptionSpec<String> optionspec21 = optionparser.accepts("version").withRequiredArg().required();
         OptionSpec<Integer> optionspec22 = optionparser.accepts("width").withRequiredArg().ofType(Integer.class).defaultsTo(854);
         OptionSpec<Integer> optionspec23 = optionparser.accepts("height").withRequiredArg().ofType(Integer.class).defaultsTo(480);
@@ -116,10 +_,12 @@
             CrashReport.preload();
             logger = LogUtils.getLogger();
             s1 = "Bootstrap";
-            Bootstrap.bootStrap();
-            ClientBootstrap.bootstrap();
+            net.neoforged.fml.loading.BackgroundWaiter.runAndTick(() -> Bootstrap.bootStrap(), net.neoforged.fml.loading.ImmediateWindowHandler::renderTick);
             GameLoadTimesEvent.INSTANCE.setBootstrapTime(Bootstrap.bootstrapDuration.get());
             Bootstrap.validate();
+
+            net.neoforged.neoforge.client.loading.ClientModLoader.begin(); // Mirroring server mod-loading, we construct immediately after Bootstrap.validate()
+
             s1 = "Argument parsing";
             List<String> list = optionset.valuesOf(optionspec28);
             if (!list.isEmpty()) {
@@ -174,7 +_,8 @@
                 new GameConfig.UserData(user, proxy),
                 new DisplayData(i, j, optionalint, optionalint1, flag),
                 new GameConfig.FolderData(file1, file3, file2, s6),
-                new GameConfig.GameData(flag1, s, s5, flag2, flag3, flag4, flag5, optionset.has(optionspec16)),
+                // Neo: Auto-enable offlineDeveloperMode if no accessToken is passed in dev
+                new GameConfig.GameData(flag1, s, s5, flag2, flag3, flag4, flag5, optionset.has(optionspec16) || !optionset.has(optionspec20)),
                 new GameConfig.QuickPlayData(s9, gameconfig$quickplayvariant)
             );
             Util.startTimerHackThread();
