--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -133,7 +_,8 @@
     }
 
     protected boolean isFuel(ItemStack itemStack) {
-        return this.level.fuelValues().isFuel(itemStack);
+        // TODO 26.2 - get the recipe type... somehow
+        return itemStack.getBurnTime(null, this.level.fuelValues()) > 0;
     }
 
     public float getBurnProgress() {
