--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -167,7 +167,7 @@
    }
 
    protected boolean isFuel(ItemStack p_38989_) {
-      return AbstractFurnaceBlockEntity.isFuel(p_38989_);
+      return net.neoforged.neoforge.common.CommonHooks.getBurnTime(p_38989_, this.recipeType) > 0;
    }
 
    public float getBurnProgress() {
