--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -133,7 +_,13 @@
     }
 
     protected boolean isFuel(ItemStack itemStack) {
-        return this.level.fuelValues().isFuel(itemStack);
+        var recipeType = switch (recipeBookType) {
+            case SMOKER -> net.minecraft.world.item.crafting.RecipeType.SMOKING;
+            case FURNACE -> net.minecraft.world.item.crafting.RecipeType.SMELTING;
+            case BLAST_FURNACE -> net.minecraft.world.item.crafting.RecipeType.BLASTING;
+            default -> null;
+        };
+        return itemStack.getBurnTime(recipeType, this.level.fuelValues()) > 0;
     }
 
     public float getBurnProgress() {
