--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -91,7 +_,7 @@
     }
 
     protected static boolean stillValid(ContainerLevelAccess access, Player player, Block block) {
-        return access.evaluate((level, pos) -> !level.getBlockState(pos).is(block) ? false : player.isWithinBlockInteractionRange(pos, 4.0), true);
+        return access.evaluate((level, pos) -> !level.getBlockState(pos).is(block) ? false : net.neoforged.neoforge.common.CommonHooks.menuRangeValidity(player, level, pos, 4.0), true);
     }
 
     public MenuType<?> getType() {
@@ -558,6 +_,13 @@
     }
 
     private boolean tryItemClickBehaviourOverride(Player player, ClickAction clickAction, Slot slot, ItemStack clicked, ItemStack carried) {
+        // Neo: Fire the ItemStackedOnOtherEvent, and return the event's cancel result if it was cancelled. Returning true will trigger the container to stop processing further logic.
+        // The first parameter to onItemStackedOn is the "carried" (under-mouse) item, which is the second ItemStack parameter to this method.
+        var event = net.neoforged.neoforge.common.CommonHooks.onItemStackedOn(carried, clicked, slot, clickAction, player, createCarriedSlotAccess());
+        if (event.isCanceled()) {
+            return event.getCancellationResult();
+        }
+
         FeatureFlagSet enabledFeatures = player.level().enabledFeatures();
         return carried.isItemEnabled(enabledFeatures) && carried.overrideStackedOnOther(slot, clickAction, player)
             ? true
