--- a/net/minecraft/world/inventory/AnvilMenu.java
+++ b/net/minecraft/world/inventory/AnvilMenu.java
@@ -79,6 +_,8 @@
             p_150474_.giveExperienceLevels(-this.cost.get());
         }
 
+        float breakChance = net.neoforged.neoforge.common.CommonHooks.onAnvilRepair(p_150474_, p_150475_, AnvilMenu.this.inputSlots.getItem(0), AnvilMenu.this.inputSlots.getItem(1));
+
         if (this.repairItemCountCost > 0) {
             ItemStack itemstack = this.inputSlots.getItem(1);
             if (!itemstack.isEmpty() && itemstack.getCount() > this.repairItemCountCost) {
@@ -95,7 +_,7 @@
         this.inputSlots.setItem(0, ItemStack.EMPTY);
         this.access.execute((p_150479_, p_150480_) -> {
             BlockState blockstate = p_150479_.getBlockState(p_150480_);
-            if (!p_150474_.hasInfiniteMaterials() && blockstate.is(BlockTags.ANVIL) && p_150474_.getRandom().nextFloat() < 0.12F) {
+            if (!p_150474_.getAbilities().instabuild && blockstate.is(BlockTags.ANVIL) && p_150474_.getRandom().nextFloat() < breakChance) {
                 BlockState blockstate1 = AnvilBlock.damage(blockstate);
                 if (blockstate1 == null) {
                     p_150479_.removeBlock(p_150480_, false);
@@ -111,7 +_,17 @@
     }
 
     @Override
-    public void createResult() {
+    public final void createResult() {
+        // Neo: Override the real createResult() method to invoke the vanilla logic, and then call the event hook.
+        // Since the vanilla createResult() has multiple returns, we need wrap it. The choices are this, or pack it in a lambda.
+        this.__createResultImpl();
+        ItemStack leftInput = this.inputSlots.getItem(0);
+        ItemStack rightInput = this.inputSlots.getItem(1);
+        net.neoforged.neoforge.common.CommonHooks.onAnvilUpdate(this, leftInput, rightInput, resultSlots, itemName, this.player);  
+    }
+
+    // Neo: We patch the real createResult() into __createResultImpl() so we can wrap it, and provide the vanilla result in the AnvilUpdateEvent.
+    public void __createResultImpl() {
         ItemStack itemstack = this.inputSlots.getItem(0);
         this.onlyRenaming = false;
         this.cost.set(1);
@@ -176,8 +_,10 @@
                         int j2 = entry.getIntValue();
                         j2 = i2 == j2 ? j2 + 1 : Math.max(j2, i2);
                         Enchantment enchantment = holder.value();
-                        boolean flag1 = enchantment.canEnchant(itemstack);
-                        if (this.player.hasInfiniteMaterials() || itemstack.is(Items.ENCHANTED_BOOK)) {
+                        // Neo: Respect IItemExtension#supportsEnchantment - we also delegate the logic for Enchanted Books to this method.
+                        // Though we still allow creative players to combine any item with any enchantment in the anvil here.
+                        boolean flag1 = itemstack.supportsEnchantment(holder);
+                        if (this.player.getAbilities().instabuild) {
                             flag1 = true;
                         }
 
@@ -301,5 +_,12 @@
 
     public int getCost() {
         return this.cost.get();
+    }
+
+    /**
+     * Neo: Sets the cost. Negative values are clamped to zero.
+     */
+    public void setCost(int value) {
+        this.cost.set(Math.max(0, value));
     }
 }
