--- a/net/minecraft/world/inventory/BrewingStandMenu.java
+++ b/net/minecraft/world/inventory/BrewingStandMenu.java
@@ -153,7 +_,7 @@
 
         @Override
         public boolean mayPlace(ItemStack p_39121_) {
-            return this.potionBrewing.isIngredient(p_39121_);
+            return net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(p_39121_);
         }
     }
 
@@ -176,6 +_,7 @@
         public void onTake(Player p_150499_, ItemStack p_150500_) {
             Optional<Holder<Potion>> optional = p_150500_.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).potion();
             if (optional.isPresent() && p_150499_ instanceof ServerPlayer serverplayer) {
+                net.neoforged.neoforge.event.EventHooks.onPlayerBrewedPotion(p_150499_, p_150500_);
                 CriteriaTriggers.BREWED_POTION.trigger(serverplayer, optional.get());
             }
 
@@ -183,7 +_,7 @@
         }
 
         public static boolean mayPlaceItem(ItemStack p_39134_) {
-            return p_39134_.is(Items.POTION) || p_39134_.is(Items.SPLASH_POTION) || p_39134_.is(Items.LINGERING_POTION) || p_39134_.is(Items.GLASS_BOTTLE);
+            return net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidInput(p_39134_);
         }
     }
 }
