--- a/net/minecraft/world/inventory/BrewingStandMenu.java
+++ b/net/minecraft/world/inventory/BrewingStandMenu.java
@@ -75,7 +_,7 @@
                     if (!this.moveItemStackTo(stack, 3, 4, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (brewingInputs.test(clicked)) {
+                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(player.level().potionBrewing(), clicked)) {
                     if (!this.moveItemStackTo(stack, 0, 3, false)) {
                         return ItemStack.EMPTY;
                     }
@@ -131,13 +_,20 @@
     }
 
     private static class FuelSlot extends Slot {
+        private final PotionBrewing potionBrewing;
+
         public FuelSlot(Container container, int slot, int x, int y) {
+            this(PotionBrewing.EMPTY, container, slot, x, y);
+        }
+
+        public PotionSlot(PotionBrewing potionBrewing, Container container, int slot, int x, int y) {
             super(container, slot, x, y);
+            this.potionBrewing = potionBrewing;
         }
 
         @Override
         public boolean mayPlace(ItemStack itemStack) {
-            return mayPlaceItem(itemStack);
+            return mayPlaceItem(this.potionBrewing, itemStack);
         }
 
         public static boolean mayPlaceItem(ItemStack itemStack) {
