--- a/net/minecraft/world/inventory/BrewingStandMenu.java
+++ b/net/minecraft/world/inventory/BrewingStandMenu.java
@@ -77,7 +_,7 @@
                     if (!this.moveItemStackTo(itemstack1, 3, 4, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(itemstack) && itemstack.getCount() == 1) {
+                } else if (net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidInput(this.brewingStand, itemstack)) {
                     if (!this.moveItemStackTo(itemstack1, 0, 3, false)) {
                         return ItemStack.EMPTY;
                     }
@@ -151,7 +_,7 @@
 
         @Override
         public boolean mayPlace(ItemStack p_39121_) {
-            return PotionBrewing.isIngredient(p_39121_);
+            return net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(this.container, p_39121_);
         }
 
         @Override
@@ -167,7 +_,7 @@
 
         @Override
         public boolean mayPlace(ItemStack p_39132_) {
-            return mayPlaceItem(p_39132_);
+            return net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidInput(this.container, p_39132_);
         }
 
         @Override
@@ -179,6 +_,7 @@
         public void onTake(Player p_150499_, ItemStack p_150500_) {
             Potion potion = PotionUtils.getPotion(p_150500_);
             if (p_150499_ instanceof ServerPlayer) {
+                net.neoforged.neoforge.event.EventHooks.onPlayerBrewedPotion(p_150499_, p_150500_);
                 CriteriaTriggers.BREWED_POTION.trigger((ServerPlayer)p_150499_, potion.builtInRegistryHolder());
             }
 
