--- a/net/minecraft/world/inventory/MenuType.java
+++ b/net/minecraft/world/inventory/MenuType.java
@@ -55,6 +_,14 @@
     }
 
     @Override
+    public T create(int windowId, Inventory playerInv, net.minecraft.network.RegistryFriendlyByteBuf extraData) {
+        if (this.constructor instanceof net.neoforged.neoforge.network.IContainerFactory) {
+            return ((net.neoforged.neoforge.network.IContainerFactory<T>) this.constructor).create(windowId, playerInv, extraData);
+        }
+        return create(windowId, playerInv);
+    }
+
+    @Override
     public FeatureFlagSet requiredFeatures() {
         return this.requiredFeatures;
     }
