--- a/net/minecraft/world/inventory/RecipeBookType.java
+++ b/net/minecraft/world/inventory/RecipeBookType.java
@@ -1,8 +_,13 @@
 package net.minecraft.world.inventory;
 
+@net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.CLIENTBOUND)
 public enum RecipeBookType implements net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     CRAFTING,
     FURNACE,
     BLAST_FURNACE,
     SMOKER;
+
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(RecipeBookType.class);
+    }
 }
