--- a/net/minecraft/world/inventory/RecipeBookType.java
+++ b/net/minecraft/world/inventory/RecipeBookType.java
@@ -2,9 +_,14 @@
 
 import net.neoforged.fml.common.asm.enumextension.IExtensibleEnum;
 
+@net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.CLIENTBOUND)
 public enum RecipeBookType implements IExtensibleEnum {
     CRAFTING,
     FURNACE,
     BLAST_FURNACE,
     SMOKER;
+
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(RecipeBookType.class);
+    }
 }
