--- a/net/minecraft/world/inventory/ResultSlot.java
+++ b/net/minecraft/world/inventory/ResultSlot.java
@@ -47,6 +_,7 @@
     protected void checkTakeAchievements(ItemStack stack) {
         if (this.removeCount > 0) {
             stack.onCraftedBy(this.player.level(), this.player, this.removeCount);
+            net.neoforged.neoforge.event.EventHooks.firePlayerCraftingEvent(this.player, stack, this.craftSlots);
         }
 
         if (this.container instanceof RecipeCraftingHolder recipecraftingholder) {
@@ -63,7 +_,9 @@
         CraftingInput craftinginput = craftinginput$positioned.input();
         int i = craftinginput$positioned.left();
         int j = craftinginput$positioned.top();
+        net.neoforged.neoforge.common.CommonHooks.setCraftingPlayer(player);
         NonNullList<ItemStack> nonnulllist = player.level().getRecipeManager().getRemainingItemsFor(RecipeType.CRAFTING, craftinginput, player.level());
+        net.neoforged.neoforge.common.CommonHooks.setCraftingPlayer(null);
 
         for (int k = 0; k < craftinginput.height(); k++) {
             for (int l = 0; l < craftinginput.width(); l++) {
