--- a/net/minecraft/world/inventory/ResultSlot.java
+++ b/net/minecraft/world/inventory/ResultSlot.java
@@ -7,7 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.CraftingInput;
 import net.minecraft.world.item.crafting.CraftingRecipe;
-import net.minecraft.world.item.crafting.RecipeHolder;
 import net.minecraft.world.item.crafting.RecipeType;
 import net.minecraft.world.level.Level;
 
@@ -51,6 +_,7 @@
     protected void checkTakeAchievements(ItemStack p_40185_) {
         if (this.removeCount > 0) {
             p_40185_.onCraftedBy(this.player, this.removeCount);
+            net.neoforged.neoforge.event.EventHooks.firePlayerCraftedEvent(this.player, p_40185_, this.craftSlots);
         }
 
         if (this.container instanceof RecipeCraftingHolder recipecraftingholder) {
@@ -86,7 +_,9 @@
         CraftingInput craftinginput = craftinginput$positioned.input();
         int i = craftinginput$positioned.left();
         int j = craftinginput$positioned.top();
+        net.neoforged.neoforge.common.CommonHooks.setCraftingPlayer(p_150638_);
         NonNullList<ItemStack> nonnulllist = this.getRemainingItems(craftinginput, p_150638_.level());
+        net.neoforged.neoforge.common.CommonHooks.setCraftingPlayer(null);
 
         for (int k = 0; k < craftinginput.height(); k++) {
             for (int l = 0; l < craftinginput.width(); l++) {
