--- a/net/minecraft/world/inventory/Slot.java
+++ b/net/minecraft/world/inventory/Slot.java
@@ -13,6 +_,8 @@
     public int index;
     public final int x;
     public final int y;
+    @Nullable
+    private Identifier backgroundSprite;
 
     public Slot(Container container, int slot, int x, int y) {
         this.container = container;
@@ -79,7 +_,7 @@
     }
 
     public @Nullable Identifier getNoItemIcon() {
-        return null;
+        return this.backgroundSprite;
     }
 
     public ItemStack remove(int amount) {
@@ -92,6 +_,28 @@
 
     public boolean isActive() {
         return true;
+    }
+
+    /// Neo: Returns the index in the inventory for this slot, this value should typically not
+    /// be used, but can be useful for some occasions.
+    ///
+    /// @return Index in associated inventory for this slot.
+    public int getSlotIndex() {
+        return this.slot;
+    }
+
+    /// Neo: {@return whether this slot belongs to the same inventory as the given slot}
+    public boolean isSameInventory(Slot other) {
+        return this.container == other.container;
+    }
+
+    /// Neo: Set the background atlas and sprite location.
+    ///
+    /// @param sprite The sprite located on that atlas.
+    /// @return this, to allow chaining.
+    public Slot setBackground(Identifier sprite) {
+         this.backgroundSprite = sprite;
+         return this;
     }
 
     public Optional<ItemStack> tryRemove(int amount, int maxAmount, Player player) {
