--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -33,6 +33,13 @@
 import net.minecraft.world.level.gamerules.GameRules;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Map;
+import net.minecraft.util.datafix.fixes.BlockStateData;
+import net.minecraft.util.datafix.fixes.ItemIdFix;
+import net.minecraft.util.datafix.fixes.ItemSpawnEggFix;
+// CraftBukkit end
+
 @SuppressForbidden(reason = "System.out setup")
 public class Bootstrap {
 
@@ -64,6 +71,69 @@
                     wrapStreams();
                     Bootstrap.bootstrapDuration.set(Duration.between(instant, Instant.now()).toMillis());
                 }
+                // CraftBukkit start - easier than fixing the decompile
+                BlockStateData.register(1008, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "0")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "0")));
+                BlockStateData.register(1009, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "1")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "1")));
+                BlockStateData.register(1010, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "2")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "2")));
+                BlockStateData.register(1011, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "3")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "3")));
+                BlockStateData.register(1012, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "4")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "4")));
+                BlockStateData.register(1013, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "5")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "5")));
+                BlockStateData.register(1014, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "6")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "6")));
+                BlockStateData.register(1015, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "7")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "7")));
+                BlockStateData.register(1016, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "8")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "8")));
+                BlockStateData.register(1017, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "9")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "9")));
+                BlockStateData.register(1018, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "10")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "10")));
+                BlockStateData.register(1019, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "11")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "11")));
+                BlockStateData.register(1020, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "12")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "12")));
+                BlockStateData.register(1021, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "13")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "13")));
+                BlockStateData.register(1022, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "14")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "14")));
+                BlockStateData.register(1023, BlockStateData.create("minecraft:oak_sign", Map.of("rotation", "15")), BlockStateData.create("minecraft:standing_sign", Map.of("rotation", "15")));
+                ItemIdFix.ITEM_NAMES.put(323, "minecraft:oak_sign");
+
+                BlockStateData.register(1440, BlockStateData.create("minecraft:portal", Map.of("axis", "x")), BlockStateData.create("minecraft:portal", Map.of("axis", "x")));
+
+                ItemIdFix.ITEM_NAMES.put(409, "minecraft:prismarine_shard");
+                ItemIdFix.ITEM_NAMES.put(410, "minecraft:prismarine_crystals");
+                ItemIdFix.ITEM_NAMES.put(411, "minecraft:rabbit");
+                ItemIdFix.ITEM_NAMES.put(412, "minecraft:cooked_rabbit");
+                ItemIdFix.ITEM_NAMES.put(413, "minecraft:rabbit_stew");
+                ItemIdFix.ITEM_NAMES.put(414, "minecraft:rabbit_foot");
+                ItemIdFix.ITEM_NAMES.put(415, "minecraft:rabbit_hide");
+                ItemIdFix.ITEM_NAMES.put(416, "minecraft:armor_stand");
+
+                ItemIdFix.ITEM_NAMES.put(423, "minecraft:mutton");
+                ItemIdFix.ITEM_NAMES.put(424, "minecraft:cooked_mutton");
+                ItemIdFix.ITEM_NAMES.put(425, "minecraft:banner");
+                ItemIdFix.ITEM_NAMES.put(426, "minecraft:end_crystal");
+                ItemIdFix.ITEM_NAMES.put(427, "minecraft:spruce_door");
+                ItemIdFix.ITEM_NAMES.put(428, "minecraft:birch_door");
+                ItemIdFix.ITEM_NAMES.put(429, "minecraft:jungle_door");
+                ItemIdFix.ITEM_NAMES.put(430, "minecraft:acacia_door");
+                ItemIdFix.ITEM_NAMES.put(431, "minecraft:dark_oak_door");
+                ItemIdFix.ITEM_NAMES.put(432, "minecraft:chorus_fruit");
+                ItemIdFix.ITEM_NAMES.put(433, "minecraft:chorus_fruit_popped");
+                ItemIdFix.ITEM_NAMES.put(434, "minecraft:beetroot");
+                ItemIdFix.ITEM_NAMES.put(435, "minecraft:beetroot_seeds");
+                ItemIdFix.ITEM_NAMES.put(436, "minecraft:beetroot_soup");
+                ItemIdFix.ITEM_NAMES.put(437, "minecraft:dragon_breath");
+                ItemIdFix.ITEM_NAMES.put(438, "minecraft:splash_potion");
+                ItemIdFix.ITEM_NAMES.put(439, "minecraft:spectral_arrow");
+                ItemIdFix.ITEM_NAMES.put(440, "minecraft:tipped_arrow");
+                ItemIdFix.ITEM_NAMES.put(441, "minecraft:lingering_potion");
+                ItemIdFix.ITEM_NAMES.put(442, "minecraft:shield");
+                ItemIdFix.ITEM_NAMES.put(443, "minecraft:elytra");
+                ItemIdFix.ITEM_NAMES.put(444, "minecraft:spruce_boat");
+                ItemIdFix.ITEM_NAMES.put(445, "minecraft:birch_boat");
+                ItemIdFix.ITEM_NAMES.put(446, "minecraft:jungle_boat");
+                ItemIdFix.ITEM_NAMES.put(447, "minecraft:acacia_boat");
+                ItemIdFix.ITEM_NAMES.put(448, "minecraft:dark_oak_boat");
+                ItemIdFix.ITEM_NAMES.put(449, "minecraft:totem_of_undying");
+                ItemIdFix.ITEM_NAMES.put(450, "minecraft:shulker_shell");
+                ItemIdFix.ITEM_NAMES.put(452, "minecraft:iron_nugget");
+                ItemIdFix.ITEM_NAMES.put(453, "minecraft:knowledge_book");
+
+                ItemSpawnEggFix.ID_TO_ENTITY[23] = "Arrow";
+                // CraftBukkit end
             }
         }
     }
