--- a/net/minecraft/world/level/pathfinder/PathfindingContext.java
+++ b/net/minecraft/world/level/pathfinder/PathfindingContext.java
@@ -13,6 +_,7 @@
     private final PathTypeCache cache;
     private final BlockPos mobPosition;
     private final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
+    final Mob mob;
 
     public PathfindingContext(CollisionGetter p_331783_, Mob p_331698_) {
         this.level = p_331783_;
@@ -23,11 +_,12 @@
         }
 
         this.mobPosition = p_331698_.blockPosition();
+        this.mob = p_331698_;
     }
 
     public PathType getPathTypeFromState(int p_331972_, int p_330358_, int p_330334_) {
         BlockPos blockpos = this.mutablePos.set(p_331972_, p_330358_, p_330334_);
-        return this.cache == null ? WalkNodeEvaluator.getPathTypeFromState(this.level, blockpos) : this.cache.getOrCompute(this.level, blockpos);
+        return this.cache == null ? WalkNodeEvaluator.getPathTypeFromState(this.level, blockpos, mob) : this.cache.getOrCompute(this.level, blockpos, mob);
     }
 
     public BlockState getBlockState(BlockPos p_330575_) {
@@ -40,5 +_,9 @@
 
     public BlockPos mobPosition() {
         return this.mobPosition;
+    }
+
+    BlockPos currentEvalPos() {
+        return this.mutablePos;
     }
 }
