--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -494,6 +_,8 @@
 
     protected static PathType getPathTypeFromState(BlockGetter level, BlockPos pos) {
         BlockState blockState = level.getBlockState(pos);
+        PathType type = blockState.getBlockPathType(level, pos, null);
+        if (type != null) return type;
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
@@ -509,6 +_,8 @@
             return PathType.COCOA;
         } else if (!blockState.is(Blocks.WITHER_ROSE) && !blockState.is(Blocks.POINTED_DRIPSTONE)) {
             FluidState fluidState = blockState.getFluidState();
+            PathType nonLoggableFluidPathType = fluidState.getBlockPathType(level, pos, null, false);
+            if (nonLoggableFluidPathType != null) return nonLoggableFluidPathType;
             if (fluidState.is(FluidTags.LAVA)) {
                 return PathType.LAVA;
             } else if (isBurningBlock(blockState)) {
@@ -529,6 +_,8 @@
                 if (!blockState.isPathfindable(PathComputationType.LAND)) {
                     return PathType.BLOCKED;
                 } else {
+                    PathType loggableFluidPathType = fluidState.getBlockPathType(level, pos, null, true);
+                    if (loggableFluidPathType != null) return loggableFluidPathType;
                     return fluidState.is(FluidTags.WATER) ? PathType.WATER : PathType.OPEN;
                 }
             } else {
