--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -836,6 +904,7 @@
         return hoverName;
     }
 
+    @Override // Neo: Add override annotation to ensure our extension method signature always matches vanilla
     public <T extends TooltipProvider> void addToTooltip(
         DataComponentType<T> type, Item.TooltipContext context, TooltipDisplay display, Consumer<Component> consumer, TooltipFlag flag
     ) {
