--- a/net/minecraft/server/packs/repository/Pack.java
+++ b/net/minecraft/server/packs/repository/Pack.java
@@ -27,6 +_,8 @@
    private final Pack.Position f_10406_;
    private final boolean f_10407_;
    private final boolean f_10408_;
+   private final boolean hidden; // Forge: Allow packs to be hidden from the UI entirely
+   private final List<Pack> children; // Forge: Allows packs to specify packs which will always be placed beneath them; must be hidden
    private final PackSource f_10409_;
 
    @Nullable
@@ -40,6 +_,34 @@
    }
 
    private Pack(String p_252218_, boolean p_248829_, Pack.ResourcesSupplier p_249377_, Component p_251718_, Pack.Info p_250162_, PackCompatibility p_250361_, Pack.Position p_251298_, boolean p_249753_, PackSource p_251608_) {
+      this(p_252218_, p_248829_, p_249377_, p_251718_, p_250162_, p_250361_, p_251298_, p_249753_, p_251608_, java.util.Collections.emptyList());
+   }
+
+   /**
+    * @return a new pack with the provided children in place of any children this pack currently has
+    */
+   public Pack withChildren(List<Pack> children) {
+      // data and asset formats passed in pack info here are irrelevant, as they are not used in Pack constructor
+      return new Pack(this.f_10401_, this.f_10407_, this.f_244124_, this.f_10403_, new Pack.Info(this.f_10404_, 0, 0, this.f_244623_, this.hidden), this.f_10405_, this.f_10406_, this.f_10408_, this.f_10409_, children);
+   }
+
+   private Pack(String p_252218_, boolean p_248829_, Pack.ResourcesSupplier p_249377_, Component p_251718_, Pack.Info p_250162_, PackCompatibility p_250361_, Pack.Position p_251298_, boolean p_249753_, PackSource p_251608_, List<Pack> children) {
+      List<Pack> flattenedChildren = new java.util.ArrayList<>();
+      List<Pack> remainingChildren = children;
+      // recursively flatten children
+      while (!remainingChildren.isEmpty()) {
+         List<Pack> oldChildren = remainingChildren;
+         remainingChildren = new java.util.ArrayList<>();
+         for (Pack child : oldChildren) {
+            if (!child.isHidden()) {
+               throw new IllegalArgumentException("Child pack "+ child.m_10446_() +" must be hidden");
+            }
+            flattenedChildren.add(child.withChildren(java.util.Collections.emptyList()));
+            remainingChildren.addAll(child.getChildren());
+         }
+      }
+      this.hidden = p_250162_.hidden();
+      this.children = java.util.Collections.unmodifiableList(flattenedChildren);
       this.f_10401_ = p_252218_;
       this.f_244124_ = p_249377_;
       this.f_10403_ = p_251718_;
@@ -62,7 +_,8 @@
          } else {
             FeatureFlagsMetadataSection featureflagsmetadatasection = packresources.m_5550_(FeatureFlagsMetadataSection.f_244642_);
             FeatureFlagSet featureflagset = featureflagsmetadatasection != null ? featureflagsmetadatasection.f_244197_() : FeatureFlagSet.m_246902_();
-            return new Pack.Info(packmetadatasection.m_10373_(), packmetadatasection.m_10374_(), featureflagset);
+            // Forge: Allow separate pack formats for server data and client resources and setting isHidden
+            return new Pack.Info(packmetadatasection.m_10373_(), packmetadatasection.getPackFormat(PackType.SERVER_DATA), packmetadatasection.getPackFormat(PackType.CLIENT_RESOURCES), featureflagset, packresources.isHidden());
          }
       } catch (Exception exception) {
          f_10399_.warn("Failed to read pack metadata", (Throwable)exception);
@@ -116,6 +_,10 @@
       return this.f_10409_;
    }
 
+   public boolean isHidden() { return hidden; }
+
+   public List<Pack> getChildren() { return children; }
+
    public boolean equals(Object p_10448_) {
       if (this == p_10448_) {
          return true;
@@ -131,9 +_,17 @@
       return this.f_10401_.hashCode();
    }
 
-   public static record Info(Component f_244592_, int f_244194_, FeatureFlagSet f_244041_) {
+   public static record Info(Component f_244592_, int dataFormat, int resourceFormat, FeatureFlagSet f_244041_, boolean hidden) {
+      public Info(Component f_244592_, int f_244194_, FeatureFlagSet f_244041_) {
+         this(f_244592_, f_244194_, f_244194_, f_244041_, false);
+      }
+
+      public int getFormat(PackType type) {
+         return type == PackType.SERVER_DATA ? this.dataFormat : this.resourceFormat;
+      }
+
       public PackCompatibility m_246438_(PackType p_249204_) {
-         return PackCompatibility.m_143882_(this.f_244194_, p_249204_);
+         return PackCompatibility.m_143882_(getFormat(p_249204_), p_249204_);
       }
    }
 
