--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -428,8 +_,8 @@
 
     public Vec3 getRedstoneDirection(BlockPos pos) {
         BlockState state = this.level().getBlockState(pos);
-        if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail() && state.getValue(PoweredRailBlock.POWERED)) {
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), pos, this);
             if (shape == RailShape.EAST_WEST) {
                 if (this.isRedstoneConductor(pos.west())) {
                     return new Vec3(1.0, 0.0, 0.0);
