--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecartContainer.java
@@ -94,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player player, InteractionHand hand, Vec3 location) {
+        InteractionResult ret = super.interact(player, hand, location);
+        if (ret.consumesAction()) return ret;
         return this.interactWithContainerVehicle(player);
     }
 
