--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -62,7 +_,7 @@
             this.minecart.setOnRails(onRails);
             if (onRails) {
                 this.moveAlongTrack(level);
-                if (state.is(Blocks.ACTIVATOR_RAIL)) {
+                if (state.getBlock() instanceof PoweredRailBlock poweredRail && poweredRail.isActivatorRail()) {
                     this.minecart.activateMinecart(level, var11.getX(), var11.getY(), var11.getZ(), state.getValue(PoweredRailBlock.POWERED));
                 }
             } else {
@@ -112,7 +_,7 @@
         y = pos.getY();
         boolean powerTrack = false;
         boolean haltTrack = false;
-        if (state.is(Blocks.POWERED_RAIL)) {
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail()) {
             powerTrack = state.getValue(PoweredRailBlock.POWERED);
             haltTrack = !powerTrack;
         }
@@ -123,7 +_,7 @@
         }
 
         Vec3 movement = this.getDeltaMovement();
-        RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+        RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), pos, this.minecart);;
         switch (shape) {
             case ASCENDING_EAST:
                 this.setDeltaMovement(movement.add(-slideSpeed, 0.0, 0.0));
@@ -279,9 +_,10 @@
             yt--;
         }
 
-        BlockState state = this.level().getBlockState(new BlockPos(xt, yt, zt));
+        var railPos = new BlockPos(xt, yt, zt);
+        BlockState state = this.level().getBlockState(railPos);
         if (BaseRailBlock.isRail(state)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), railPos, this.minecart);
             y = yt;
             if (shape.isSlope()) {
                 y = yt + 1;
@@ -317,9 +_,10 @@
             yt--;
         }
 
-        BlockState state = this.level().getBlockState(new BlockPos(xt, yt, zt));
+        BlockPos railPos = new BlockPos(xt, yt, zt);
+        BlockState state = this.level().getBlockState(railPos);
         if (BaseRailBlock.isRail(state)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), railPos, this.minecart);
             Pair<Vec3i, Vec3i> exits = AbstractMinecart.exits(shape);
             Vec3i exit0 = exits.getFirst();
             Vec3i exit1 = exits.getSecond();
