--- a/net/minecraft/client/gui/font/GlyphRenderTypes.java
+++ b/net/minecraft/client/gui/font/GlyphRenderTypes.java
@@ -9,27 +_,49 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public record GlyphRenderTypes(RenderType normal, RenderType seeThrough, RenderType polygonOffset, RenderPipeline guiPipeline) {
+public record GlyphRenderTypes(RenderType normal, RenderType seeThrough, RenderType polygonOffset, RenderPipeline guiPipeline,
+                               // Neo: Allow linear filtering for in-world text
+                               RenderType normalBlur, RenderType seeThroughBlur, RenderType polygonOffsetBlur) {
+
+    /** @deprecated Neo: Use {@link GlyphRenderTypes(RenderType,RenderType,RenderType,RenderPipeline,RenderType,RenderType,RenderType)} instead */
+    @Deprecated
+    public GlyphRenderTypes(RenderType normal, RenderType seeThrough, RenderType polygonOffset, RenderPipeline guiPipeline) {
+        this(normal, seeThrough, polygonOffset, guiPipeline, normal, seeThrough, polygonOffset);
+    }
+
     public static GlyphRenderTypes createForIntensityTexture(ResourceLocation p_285411_) {
         return new GlyphRenderTypes(
             RenderType.textIntensity(p_285411_),
             RenderType.textIntensitySeeThrough(p_285411_),
             RenderType.textIntensityPolygonOffset(p_285411_),
-            RenderPipelines.GUI_TEXT_INTENSITY
+            RenderPipelines.TEXT_INTENSITY,
+            net.neoforged.neoforge.client.NeoForgeRenderTypes.getTextIntensityFiltered(p_285411_),
+            net.neoforged.neoforge.client.NeoForgeRenderTypes.getTextIntensitySeeThroughFiltered(p_285411_),
+            net.neoforged.neoforge.client.NeoForgeRenderTypes.getTextIntensityPolygonOffsetFiltered(p_285411_)
         );
     }
 
     public static GlyphRenderTypes createForColorTexture(ResourceLocation p_285486_) {
         return new GlyphRenderTypes(
-            RenderType.text(p_285486_), RenderType.textSeeThrough(p_285486_), RenderType.textPolygonOffset(p_285486_), RenderPipelines.GUI_TEXT
+            RenderType.text(p_285486_), RenderType.textSeeThrough(p_285486_), RenderType.textPolygonOffset(p_285486_), RenderPipelines.TEXT,
+            net.neoforged.neoforge.client.NeoForgeRenderTypes.getTextFiltered(p_285486_),
+            net.neoforged.neoforge.client.NeoForgeRenderTypes.getTextSeeThroughFiltered(p_285486_),
+            net.neoforged.neoforge.client.NeoForgeRenderTypes.getTextPolygonOffsetFiltered(p_285486_)
         );
     }
 
     public RenderType select(Font.DisplayMode p_285259_) {
+        return this.select(p_285259_, false);
+    }
+
+    /**
+     * Neo: returns the {@link RenderType} to use for the given {@link Font.DisplayMode} and blur setting
+     */
+    public RenderType select(Font.DisplayMode p_285259_, boolean blur) {
         return switch (p_285259_) {
-            case NORMAL -> this.normal;
-            case SEE_THROUGH -> this.seeThrough;
-            case POLYGON_OFFSET -> this.polygonOffset;
+            case NORMAL -> blur ? this.normalBlur : this.normal;
+            case SEE_THROUGH -> blur ? this.seeThroughBlur : this.seeThrough;
+            case POLYGON_OFFSET -> blur ? this.polygonOffsetBlur : this.polygonOffset;
         };
     }
 }
