--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -186,6 +186,30 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import com.mojang.authlib.GameProfile;
+import java.util.UUID;
+import net.minecraft.resources.Identifier;
+import net.minecraft.server.bossevents.CustomBossEvents;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.RandomSequences;
+import net.minecraft.world.level.biome.BiomeSource;
+import net.minecraft.world.level.gamerules.GameRuleMap;
+import net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator;
+import net.minecraft.world.level.levelgen.FlatLevelSource;
+import net.minecraft.world.level.storage.PrimaryLevelData;
+import net.minecraft.world.level.timers.TimerQueue;
+import org.bukkit.Bukkit;
+import org.bukkit.WeatherType;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.generator.CustomWorldChunkManager;
+import org.bukkit.craftbukkit.util.WorldUUID;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.server.MapInitializeEvent;
+import org.bukkit.event.weather.LightningStrikeEvent;
+import org.bukkit.event.world.TimeSkipEvent;
+// CraftBukkit end
+
 public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGetter {
 
     public static final BlockPos END_SPAWN_POINT = new BlockPos(100, 50, 0);
@@ -200,7 +224,7 @@
     private final List<ServerPlayer> players = Lists.newArrayList();
     private final ServerChunkCache chunkSource;
     private final MinecraftServer server;
-    public final ServerLevelData serverLevelData;
+    public final PrimaryLevelData serverLevelData; // CraftBukkit - type
     private final EntityTickList entityTickList = new EntityTickList();
     private final ServerWaypointManager waypointManager;
     private EnvironmentAttributeSystem environmentAttributes;
@@ -227,13 +251,103 @@
     private final boolean tickTime;
     private final LevelDebugSynchronizers debugSynchronizers = new LevelDebugSynchronizers(this);
 
-    public ServerLevel(MinecraftServer server, Executor executor, LevelStorageSource.LevelStorageAccess levelStorage, ServerLevelData levelData, ResourceKey<Level> dimension, LevelStem levelStem, boolean isDebug, long biomeZoomSeed, List<CustomSpawner> customSpawners, boolean tickTime) {
-        super(levelData, dimension, server.registryAccess(), levelStem.type(), false, isDebug, biomeZoomSeed, server.getMaxChainedNeighborUpdates());
+    // CraftBukkit start
+    public final LevelStorageSource.LevelStorageAccess storageSource;
+    public final UUID uuid;
+    // Moved from MinecraftServer start
+    private final SavedDataStorage savedDataStorage;
+    private final GameRules gameRules;
+    private final WorldGenSettings worldGenSettings;
+    private final CustomBossEvents customBossEvents;
+    private final RandomSequences randomSequences;
+    private final WeatherData weatherData;
+    private final TimerQueue<MinecraftServer> scheduledEvents;
+    private final ServerClockManager clockManager;
+
+    public WorldGenSettings getWorldGenSettings() {
+        return this.worldGenSettings;
+    }
+
+    public CustomBossEvents getCustomBossEvents() {
+        return this.customBossEvents;
+    }
+
+    public RandomSource getRandomSequence(Identifier key) {
+        return this.randomSequences.get(key, this.worldGenSettings.options().seed());
+    }
+
+    public RandomSequences getRandomSequences() {
+        return this.randomSequences;
+    }
+
+    public void setWeatherParameters(int clearTime, int rainTime, boolean raining, boolean thundering) {
+        WeatherData weatherdata = this.getWeatherData();
+
+        weatherdata.setClearWeatherTime(clearTime);
+        weatherdata.setRainTime(rainTime);
+        weatherdata.setThunderTime(rainTime);
+        weatherdata.setRaining(raining);
+        weatherdata.setThundering(thundering);
+    }
+
+    public TimerQueue<MinecraftServer> getScheduledEvents() {
+        return this.scheduledEvents;
+    }
+    // Moved from MinecraftServer end
+
+    public LevelChunk getChunkIfLoaded(int x, int z) {
+        return this.chunkSource.getChunk(x, z, false);
+    }
+
+    @Override
+    public ResourceKey<LevelStem> getTypeKey() {
+        return storageSource.dimensionType;
+    }
+
+    // Add env and gen to constructor, IWorldDataServer -> WorldDataServer
+    public ServerLevel(MinecraftServer server, Executor executor, LevelStorageSource.LevelStorageAccess levelStorage, PrimaryLevelData levelData, ResourceKey<Level> dimension, LevelStem levelStem, boolean isDebug, long biomeZoomSeed, List<CustomSpawner> customSpawners, boolean tickTime, SavedDataStorage savedDataStorage, WorldGenSettings worldGenSettings, org.bukkit.World.Environment env, org.bukkit.generator.ChunkGenerator gen, org.bukkit.generator.BiomeProvider biomeProvider) {
+        super(levelData, dimension, server.registryAccess(), levelStem.type(), false, isDebug, biomeZoomSeed, server.getMaxChainedNeighborUpdates(), gen, biomeProvider, env);
+        storageSource = levelStorage;
+        uuid = WorldUUID.getUUID(levelStorage.levelDirectory.path().toFile());
+        // CraftBukkit end
         this.tickTime = tickTime;
         this.server = server;
         this.customSpawners = customSpawners;
         this.serverLevelData = levelData;
         ChunkGenerator chunkgenerator = levelStem.generator();
+        // CraftBukkit start
+        serverLevelData.setWorld(this);
+
+        // Start moved from MinecraftServer<init>
+        this.savedDataStorage = savedDataStorage;
+
+        this.worldGenSettings = worldGenSettings;
+        this.savedDataStorage.set(WorldGenSettings.TYPE, this.worldGenSettings);
+
+        this.randomSequences = (RandomSequences) this.savedDataStorage.computeIfAbsent(RandomSequences.TYPE);
+        this.weatherData = (WeatherData) this.savedDataStorage.computeIfAbsent(WeatherData.TYPE);
+        this.gameRules = new GameRules(this.serverLevelData.enabledFeatures(), (GameRuleMap) this.savedDataStorage.computeIfAbsent(GameRuleMap.TYPE));
+
+        this.clockManager = (ServerClockManager) this.savedDataStorage.computeIfAbsent(ServerClockManager.TYPE);
+        this.clockManager.init(this);
+        this.customBossEvents = (CustomBossEvents) this.savedDataStorage.computeIfAbsent(CustomBossEvents.TYPE);
+        this.scheduledEvents = (TimerQueue) this.savedDataStorage.computeIfAbsent(TimerQueue.TYPE);
+        // End moved from MinecraftServer<init>
+
+        weatherData.setWorld(this);
+        if (biomeProvider != null) {
+            BiomeSource worldChunkManager = new CustomWorldChunkManager(getWorld(), biomeProvider, server.registryAccess().lookupOrThrow(Registries.BIOME));
+            if (chunkgenerator instanceof NoiseBasedChunkGenerator cga) {
+                chunkgenerator = new NoiseBasedChunkGenerator(worldChunkManager, cga.settings);
+            } else if (chunkgenerator instanceof FlatLevelSource cpf) {
+                chunkgenerator = new FlatLevelSource(cpf.settings(), worldChunkManager);
+            }
+        }
+
+        if (gen != null) {
+            chunkgenerator = new org.bukkit.craftbukkit.generator.CustomChunkGenerator(this, chunkgenerator, gen);
+        }
+        // CraftBukkit end
         boolean flag2 = server.forceSynchronousWrites();
         DataFixer datafixer = server.getFixerUpper();
         EntityPersistentStorage<Entity> entitypersistentstorage = new EntityStorage(new SimpleRegionStorage(new RegionStorageInfo(levelStorage.getLevelId(), dimension, "entities"), levelStorage.getDimensionPath(dimension).resolve("entities"), datafixer, flag2, DataFixTypes.ENTITY_CHUNK), this, server);
@@ -251,7 +365,7 @@
         this.chunkSource.getGeneratorState().ensureStructuresGenerated();
         this.portalForcer = new PortalForcer(this);
         if (this.canHaveWeather()) {
-            this.prepareWeather(server.getWeatherData());
+            this.prepareWeather(this.getWeatherData()); // CraftBukkit
         }
 
         this.raids = (Raids) this.getDataStorage().computeIfAbsent(Raids.TYPE);
@@ -259,7 +373,7 @@
             levelData.setGameType(server.getDefaultGameType());
         }
 
-        WorldGenSettings worldgensettings = server.getWorldGenSettings();
+        WorldGenSettings worldgensettings = this.getWorldGenSettings(); // CraftBukkit
         WorldOptions worldoptions = worldgensettings.options();
         long l = worldoptions.seed();
 
@@ -275,6 +389,7 @@
         this.waypointManager = new ServerWaypointManager();
         this.environmentAttributes = EnvironmentAttributeSystem.builder().addDefaultLayers(this).build();
         this.updateSkyBrightness();
+        this.getCraftServer().addWorld(this.getWorld()); // CraftBukkit
     }
 
     @Override
@@ -306,7 +421,7 @@
 
     @Override
     public ServerClockManager clockManager() {
-        return this.server.clockManager();
+        return this.clockManager; // CraftBukkit
     }
 
     @Override
@@ -345,11 +460,12 @@
         if (this.sleepStatus.areEnoughSleeping(i) && this.sleepStatus.areEnoughDeepSleeping(i, this.players)) {
             Optional<Holder<WorldClock>> optional = this.dimensionType().defaultClock();
 
+            boolean wakeup = true; // CraftBukkit
             if ((Boolean) this.getGameRules().get(GameRules.ADVANCE_TIME) && optional.isPresent()) {
-                this.server.clockManager().moveToTimeMarker((Holder) optional.get(), ClockTimeMarkers.WAKE_UP_FROM_SLEEP);
+                wakeup = this.clockManager().moveToTimeMarker((Holder) optional.get(), ClockTimeMarkers.WAKE_UP_FROM_SLEEP, TimeSkipEvent.SkipReason.NIGHT_SKIP); // CraftBukkit
             }
 
-            this.wakeUpAllPlayers();
+            if (wakeup) this.wakeUpAllPlayers(); // CraftBukkit
             if ((Boolean) this.getGameRules().get(GameRules.ADVANCE_WEATHER) && this.isRaining()) {
                 this.resetWeatherCycle();
             }
@@ -385,7 +501,7 @@
 
         this.handlingTick = false;
         profilerfiller.pop();
-        boolean flag1 = this.chunkSource.hasActiveTickets();
+        boolean flag1 = true || this.chunkSource.hasActiveTickets(); // CraftBukkit - this prevents entity cleanup, other issues on servers with no players
 
         if (flag1) {
             this.resetEmptyTime();
@@ -473,7 +589,7 @@
 
     private void wakeUpAllPlayers() {
         this.sleepStatus.removeAllSleepers();
-        ((List) this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList())).forEach((serverplayer) -> {
+        (this.players.stream().filter(LivingEntity::isSleeping).collect(Collectors.toList())).forEach((serverplayer) -> { // CraftBukkit - decompile error
             serverplayer.stopSleepInBed(false, false);
         });
     }
@@ -550,7 +666,7 @@
                         skeletonhorse.setTrap(true);
                         skeletonhorse.setAge(0);
                         skeletonhorse.setPos((double) blockpos.getX(), (double) blockpos.getY(), (double) blockpos.getZ());
-                        this.addFreshEntity(skeletonhorse);
+                        this.addFreshEntity(skeletonhorse, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
                     }
                 }
 
@@ -559,7 +675,7 @@
                 if (lightningbolt != null) {
                     lightningbolt.snapTo(Vec3.atBottomCenterOf(blockpos));
                     lightningbolt.setVisualOnly(flag1);
-                    this.addFreshEntity(lightningbolt);
+                    this.strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.WEATHER); // CraftBukkit
                 }
             }
         }
@@ -574,7 +690,7 @@
         Biome biome = (Biome) this.getBiome(blockpos1).value();
 
         if (biome.shouldFreeze(this, blockpos2)) {
-            this.setBlockAndUpdate(blockpos2, Blocks.ICE.defaultBlockState());
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, blockpos2, Blocks.ICE.defaultBlockState(), null); // CraftBukkit
         }
 
         if (this.isRaining()) {
@@ -590,10 +706,10 @@
                         BlockState blockstate1 = (BlockState) blockstate.setValue(SnowLayerBlock.LAYERS, j + 1);
 
                         Block.pushEntitiesUp(blockstate, blockstate1, this, blockpos1);
-                        this.setBlockAndUpdate(blockpos1, blockstate1);
+                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, blockpos1, blockstate1, null); // CraftBukkit
                     }
                 } else {
-                    this.setBlockAndUpdate(blockpos1, Blocks.SNOW.defaultBlockState());
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this, blockpos1, Blocks.SNOW.defaultBlockState(), null); // CraftBukkit
                 }
             }
 
@@ -709,6 +825,11 @@
     }
 
     private void prepareWeather(WeatherData weatherData) {
+        // CraftBukkit start - SPIGOT-8029: consistent with WorldServer#advanceWeatherCycle
+        if (!this.dimensionType().hasSkyLight()) {
+            return;
+        }
+        // CraftBukkit end
         if (weatherData.isRaining()) {
             this.rainLevel = 1.0F;
             if (weatherData.isThundering()) {
@@ -786,6 +907,7 @@
             this.rainLevel = Mth.clamp(this.rainLevel, 0.0F, 1.0F);
         }
 
+        /* CraftBukkit start
         if (this.oRainLevel != this.rainLevel) {
             this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, this.rainLevel), this.dimension());
         }
@@ -804,6 +926,27 @@
             this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, this.rainLevel));
             this.server.getPlayerList().broadcastAll(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, this.thunderLevel));
         }
+        // */
+        for (int idx = 0; idx < this.players.size(); ++idx) {
+            if (((ServerPlayer) this.players.get(idx)).level() == this) {
+                ((ServerPlayer) this.players.get(idx)).tickWeather();
+            }
+        }
+
+        if (flag != this.isRaining()) {
+            // Only send weather packets to those affected
+            for (int idx = 0; idx < this.players.size(); ++idx) {
+                if (((ServerPlayer) this.players.get(idx)).level() == this) {
+                    ((ServerPlayer) this.players.get(idx)).setPlayerWeather((!flag ? WeatherType.DOWNFALL : WeatherType.CLEAR), false);
+                }
+            }
+        }
+        for (int idx = 0; idx < this.players.size(); ++idx) {
+            if (((ServerPlayer) this.players.get(idx)).level() == this) {
+                ((ServerPlayer) this.players.get(idx)).updateWeather(this.oRainLevel, this.rainLevel, this.oThunderLevel, this.thunderLevel);
+            }
+        }
+        // CraftBukkit end
 
     }
 
@@ -811,10 +954,22 @@
     public void resetWeatherCycle() {
         WeatherData weatherdata = this.getWeatherData();
 
-        weatherdata.setRainTime(0);
+        // CraftBukkit start
         weatherdata.setRaining(false);
-        weatherdata.setThunderTime(0);
+        // If we stop due to everyone sleeping we should reset the weather duration to some other random value.
+        // Not that everyone ever manages to get the whole server to sleep at the same time....
+        if (!weatherdata.isRaining()) {
+            weatherdata.setRainTime(0);
+        }
+        // CraftBukkit end
         weatherdata.setThundering(false);
+        // CraftBukkit start
+        // If we stop due to everyone sleeping we should reset the weather duration to some other random value.
+        // Not that everyone ever manages to get the whole server to sleep at the same time....
+        if (!weatherdata.isThundering()) {
+            weatherdata.setThunderTime(0);
+        }
+        // CraftBukkit end
     }
 
     public void resetEmptyTime() {
@@ -851,6 +1006,7 @@
         profilerfiller.push(holder::getRegisteredName);
         profilerfiller.incrementCounter("tickNonPassenger");
         entity.tick();
+        entity.postTick(); // CraftBukkit
         profilerfiller.pop();
 
         for (Entity entity1 : entity.getPassengers()) {
@@ -871,6 +1027,7 @@
                 profilerfiller.push(holder::getRegisteredName);
                 profilerfiller.incrementCounter("tickPassenger");
                 entity.rideTick();
+                entity.postTick(); // CraftBukkit
                 profilerfiller.pop();
 
                 for (Entity entity2 : entity.getPassengers()) {
@@ -918,10 +1075,21 @@
         ServerChunkCache serverchunkcache = this.getChunkSource();
 
         if (!noSave) {
+            org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.world.WorldSaveEvent(getWorld())); // CraftBukkit
             if (progressListener != null) {
                 progressListener.progressStartNoAbort(Component.translatable("menu.savingLevel"));
             }
 
+            // CraftBukkit start - moved from MinecraftServer.saveAllChunks
+            GameProfile gameprofile = this.server.getSingleplayerProfile();
+
+            this.storageSource.saveDataTag(this.serverLevelData, gameprofile == null ? null : gameprofile.id());
+            if (flush) {
+                this.savedDataStorage.saveAndJoin();
+            } else {
+                this.savedDataStorage.scheduleSave();
+            }
+            // CraftBukkit end
             this.saveLevelData(flush);
             if (progressListener != null) {
                 progressListener.progressStage(Component.translatable("menu.savingChunks"));
@@ -1003,18 +1171,40 @@
 
     @Override
     public boolean addFreshEntity(Entity entity) {
-        return this.addEntity(entity);
+        // CraftBukkit start
+        return this.addFreshEntity(entity, CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    @Override
+    public boolean addFreshEntity(Entity entity, CreatureSpawnEvent.SpawnReason reason) {
+        return this.addEntity(entity, reason);
+        // CraftBukkit end
     }
 
     public boolean addWithUUID(Entity entity) {
-        return this.addEntity(entity);
+        // CraftBukkit start
+        return this.addWithUUID(entity, CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public boolean addWithUUID(Entity entity, CreatureSpawnEvent.SpawnReason reason) {
+        return this.addEntity(entity, reason);
+        // CraftBukkit end
     }
 
     public void addDuringTeleport(Entity entity) {
+        // CraftBukkit start
+        // SPIGOT-6415: Don't call spawn event for entities which travel trough worlds,
+        // since it is only an implementation detail, that a new entity is created when
+        // they are traveling between worlds.
+        this.addDuringTeleport(entity, null);
+    }
+
+    public void addDuringTeleport(Entity entity, CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
         if (entity instanceof ServerPlayer serverplayer) {
             this.addPlayer(serverplayer);
         } else {
-            this.addEntity(entity);
+            this.addEntity(entity, reason); // CraftBukkit
         }
 
     }
@@ -1039,24 +1229,37 @@
         this.entityManager.addNewEntity(player);
     }
 
-    private boolean addEntity(Entity entity) {
+    // CraftBukkit start
+    private boolean addEntity(Entity entity, CreatureSpawnEvent.SpawnReason spawnReason) {
         if (entity.isRemoved()) {
-            ServerLevel.LOGGER.warn("Tried to add entity {} but it was marked as removed already", entity.typeHolder().getRegisteredName());
+            // ServerLevel.LOGGER.warn("Tried to add entity {} but it was marked as removed already", entity.typeHolder().getRegisteredName()); // CraftBukkit
             return false;
         } else {
+            // SPIGOT-6415: Don't call spawn event when reason is null. For example when an entity teleports to a new world.
+            if (spawnReason != null && !CraftEventFactory.doEntityAddEventCalling(this, entity, spawnReason)) {
+                return false;
+            }
+            // CraftBukkit end
+
             return this.entityManager.addNewEntity(entity);
         }
     }
 
     public boolean tryAddFreshEntityWithPassengers(Entity entity) {
-        Stream stream = entity.getSelfAndPassengers().map(Entity::getUUID);
+        // CraftBukkit start
+        return this.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public boolean tryAddFreshEntityWithPassengers(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
+        Stream<UUID> stream = entity.getSelfAndPassengers().map(Entity::getUUID); // CraftBukkit - decompile error
         PersistentEntitySectionManager persistententitysectionmanager = this.entityManager;
 
         Objects.requireNonNull(this.entityManager);
         if (stream.anyMatch(persistententitysectionmanager::isLoaded)) {
             return false;
         } else {
-            this.addFreshEntityWithPassengers(entity);
+            this.addFreshEntityWithPassengers(entity, reason); // CraftBukkit
             return true;
         }
     }
@@ -1068,17 +1271,45 @@
     }
 
     public void removePlayerImmediately(ServerPlayer player, Entity.RemovalReason reason) {
-        player.remove(reason);
+        player.remove(reason, null); // CraftBukkit - add Bukkit remove cause
+    }
+
+    // CraftBukkit start
+    public boolean strikeLightning(Entity entitylightning) {
+        return this.strikeLightning(entitylightning, LightningStrikeEvent.Cause.UNKNOWN);
+    }
+
+    public boolean strikeLightning(Entity entitylightning, LightningStrikeEvent.Cause cause) {
+        LightningStrikeEvent lightning = CraftEventFactory.callLightningStrikeEvent((org.bukkit.entity.LightningStrike) entitylightning.getBukkitEntity(), cause);
+
+        if (lightning.isCancelled()) {
+            return false;
+        }
+
+        return this.addFreshEntity(entitylightning);
     }
+    // CraftBukkit end
 
     @Override
     public void destroyBlockProgress(int id, BlockPos blockPos, int progress) {
+        // CraftBukkit start
+        Player player = null;
+        Entity entity = this.getEntity(id);
+        if (entity instanceof Player) player = (Player) entity;
+        // CraftBukkit end
+
         for (ServerPlayer serverplayer : this.server.getPlayerList().getPlayers()) {
             if (serverplayer.level() == this && serverplayer.getId() != id) {
                 double d0 = (double) blockPos.getX() - serverplayer.getX();
                 double d1 = (double) blockPos.getY() - serverplayer.getY();
                 double d2 = (double) blockPos.getZ() - serverplayer.getZ();
 
+                // CraftBukkit start
+                if (player != null && !serverplayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
+                    continue;
+                }
+                // CraftBukkit end
+
                 if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0D) {
                     serverplayer.connection.send(new ClientboundBlockDestructionPacket(id, blockPos, progress));
                 }
@@ -1181,8 +1412,21 @@
 
         if (Shapes.joinIsNotEmpty(voxelshape, voxelshape1, BooleanOp.NOT_SAME)) {
             List<PathNavigation> list = new ObjectArrayList();
+            // CraftBukkit start - fix SPIGOT-6362
+            java.util.Iterator<Mob> iterator = this.navigatingMobs.iterator();
 
-            for (Mob mob : this.navigatingMobs) {
+            while (iterator.hasNext()) {
+                Mob mob;
+                try {
+                    mob = iterator.next();
+                } catch (java.util.ConcurrentModificationException ex) {
+                    // This can happen because the pathfinder update below may trigger a chunk load, which in turn may cause more navigators to register
+                    // In this case we just run the update again across all the iterators as the chunk will then be loaded
+                    // As this is a relative edge case it is much faster than copying navigators (on either read or write)
+                    sendBlockUpdated(pos, old, current, updateFlags);
+                    return;
+                }
+                // CraftBukkit end
                 PathNavigation pathnavigation = mob.getNavigation();
 
                 if (pathnavigation.shouldRecomputePath(pos)) {
@@ -1245,6 +1489,12 @@
 
     @Override
     public void explode(@Nullable Entity source, @Nullable DamageSource damageSource, @Nullable ExplosionDamageCalculator damageCalculator, double x, double y, double z, float r, boolean fire, Level.ExplosionInteraction interactionType, ParticleOptions smallExplosionParticles, ParticleOptions largeExplosionParticles, WeightedList<ExplosionParticleInfo> blockParticles, Holder<SoundEvent> explosionSound) {
+        // CraftBukkit start
+        this.explode0(source, damageSource, damageCalculator, x, y, z, r, fire, interactionType, smallExplosionParticles, largeExplosionParticles, blockParticles, explosionSound);
+    }
+
+    public ServerExplosion explode0(@Nullable Entity source, @Nullable DamageSource damageSource, @Nullable ExplosionDamageCalculator damageCalculator, double x, double y, double z, float r, boolean fire, Level.ExplosionInteraction interactionType, ParticleOptions smallExplosionParticles, ParticleOptions largeExplosionParticles, WeightedList<ExplosionParticleInfo> blockParticles, Holder<SoundEvent> explosionSound) {
+        // CraftBukkit end
         Explosion.BlockInteraction explosion_blockinteraction;
 
         switch (interactionType) {
@@ -1263,6 +1513,11 @@
             case TRIGGER:
                 explosion_blockinteraction = Explosion.BlockInteraction.TRIGGER_BLOCK;
                 break;
+            // CraftBukkit start - handle custom explosion type
+            case STANDARD:
+                explosion_blockinteraction = Explosion.BlockInteraction.DESTROY;
+                break;
+            // CraftBukkit end
             default:
                 throw new MatchException((String) null, (Throwable) null);
         }
@@ -1271,6 +1526,11 @@
         Vec3 vec3 = new Vec3(x, y, z);
         ServerExplosion serverexplosion = new ServerExplosion(this, source, damageSource, damageCalculator, vec3, r, fire, explosion_blockinteraction1);
         int i = serverexplosion.explode();
+        // CraftBukkit start
+        if (serverexplosion.wasCanceled) {
+            return serverexplosion;
+        }
+        // CraftBukkit end
         ParticleOptions particleoptions2 = serverexplosion.isSmall() ? smallExplosionParticles : largeExplosionParticles;
 
         for (ServerPlayer serverplayer : this.players) {
@@ -1281,6 +1541,7 @@
             }
         }
 
+        return serverexplosion; // CraftBukkit
     }
 
     private Explosion.BlockInteraction getDestroyType(GameRule<Boolean> gameRule) {
@@ -1340,15 +1601,22 @@
     }
 
     public <T extends ParticleOptions> int sendParticles(T particle, double x, double y, double z, int count, double xDist, double yDist, double zDist, double speed) {
-        return this.sendParticles(particle, false, false, x, y, z, count, xDist, yDist, zDist, speed);
+        return this.sendParticlesSource(null, particle, false, false, x, y, z, count, xDist, yDist, zDist, speed); // CraftBukkit - visibility api support
     }
 
     public <T extends ParticleOptions> int sendParticles(T particle, boolean overrideLimiter, boolean alwaysShow, double x, double y, double z, int count, double xDist, double yDist, double zDist, double speed) {
+        return this.sendParticlesSource(null, particle, overrideLimiter, alwaysShow, x, y, z, count, xDist, yDist, zDist, speed); // CraftBukkit - visibility api support
+    }
+
+    // CraftBukkit start - visibility api support
+    public <T extends ParticleOptions> int sendParticlesSource(ServerPlayer sender, T particle, boolean overrideLimiter, boolean alwaysShow, double x, double y, double z, int count, double xDist, double yDist, double zDist, double speed) {
+        // CraftBukkit end
         ClientboundLevelParticlesPacket clientboundlevelparticlespacket = new ClientboundLevelParticlesPacket(particle, overrideLimiter, alwaysShow, x, y, z, (float) xDist, (float) yDist, (float) zDist, (float) speed, count);
         int j = 0;
 
         for (int k = 0; k < this.players.size(); ++k) {
             ServerPlayer serverplayer = (ServerPlayer) this.players.get(k);
+            if (sender != null && !serverplayer.getBukkitEntity().canSee(sender.getBukkitEntity())) continue; // CraftBukkit
 
             if (this.sendParticles(serverplayer, overrideLimiter, x, y, z, clientboundlevelparticlespacket)) {
                 ++j;
@@ -1462,16 +1730,33 @@
         return this.noSave;
     }
 
+    // CraftBukkit start
+    public SavedDataStorage getWorldContainerDataStorage() {
+        return this.savedDataStorage;
+    }
+    // CraftBukkit end
+
     public SavedDataStorage getDataStorage() {
         return this.getChunkSource().getDataStorage();
     }
 
     @Override
     public @Nullable MapItemSavedData getMapData(MapId id) {
-        return (MapItemSavedData) this.getServer().getDataStorage().get(MapItemSavedData.type(id));
+        // CraftBukkit start
+        MapItemSavedData mapitemsaveddata = (MapItemSavedData) this.getServer().getDataStorage().get(MapItemSavedData.type(id));
+        if (mapitemsaveddata != null) {
+            mapitemsaveddata.id = id;
+        }
+        return mapitemsaveddata;
+        // CraftBukkit end
     }
 
     public void setMapData(MapId id, MapItemSavedData data) {
+        // CraftBukkit start
+        data.id = id;
+        MapInitializeEvent event = new MapInitializeEvent(data.mapView);
+        Bukkit.getServer().getPluginManager().callEvent(event);
+        // CraftBukkit end
         this.getServer().getDataStorage().set(MapItemSavedData.type(id), data);
     }
 
@@ -1481,12 +1766,12 @@
 
     @Override
     public void setRespawnData(LevelData.RespawnData respawnData) {
-        this.getServer().setRespawnData(respawnData);
+        this.getServer().setRespawnData(respawnData, this); // CraftBukkit
     }
 
     @Override
     public LevelData.RespawnData getRespawnData() {
-        return this.getServer().getRespawnData();
+        return this.serverLevelData.getRespawnData(); // CraftBukkit
     }
 
     public LongSet getForceLoadedChunks() {
@@ -1669,12 +1954,12 @@
     }
 
     public boolean isFlat() {
-        return this.server.getWorldData().isFlatWorld();
+        return this.serverLevelData.isFlatWorld(); // CraftBukkit
     }
 
     @Override
     public long getSeed() {
-        return this.server.getWorldGenSettings().options().seed();
+        return this.getWorldGenSettings().options().seed(); // CraftBukkit
     }
 
     public @Nullable EnderDragonFight getDragonFight() {
@@ -1682,7 +1967,7 @@
     }
 
     public WeatherData getWeatherData() {
-        return this.server.getWeatherData();
+        return this.weatherData; // CraftBukkit
     }
 
     @Override
@@ -1770,6 +2055,7 @@
     public void close() throws IOException {
         super.close();
         this.entityManager.close();
+        this.savedDataStorage.close(); // CraftBukkit - from MinecraftServer#stopServer
     }
 
     @Override
@@ -1829,7 +2115,7 @@
     }
 
     public GameRules getGameRules() {
-        return this.server.getGameRules();
+        return this.gameRules; // CraftBukkit
     }
 
     @Override
@@ -1862,7 +2148,7 @@
     }
 
     public boolean isAllowedToEnterPortal(Level toLevel) {
-        return toLevel.dimension() == Level.NETHER ? (Boolean) this.getGameRules().get(GameRules.ALLOW_ENTERING_NETHER_USING_PORTALS) : true;
+        return toLevel.getTypeKey() == LevelStem.NETHER ? (Boolean) this.getGameRules().get(GameRules.ALLOW_ENTERING_NETHER_USING_PORTALS) : true; // CraftBukkit
     }
 
     public boolean isPvpAllowed() {
@@ -1943,6 +2229,7 @@
             }
 
             entity.updateDynamicGameEventListener(DynamicGameEventListener::add);
+            entity.valid = true; // CraftBukkit
         }
 
         public void onTrackingEnd(Entity entity) {
@@ -1971,6 +2258,14 @@
 
             entity.updateDynamicGameEventListener(DynamicGameEventListener::remove);
             ServerLevel.this.debugSynchronizers.dropEntity(entity);
+            // CraftBukkit start
+            entity.valid = false;
+            if (!(entity instanceof ServerPlayer)) {
+                for (ServerPlayer player : players) {
+                    player.getBukkitEntity().onEntityRemove(entity);
+                }
+            }
+            // CraftBukkit end
         }
 
         public void onSectionChange(Entity entity) {
