--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -227,8 +_,10 @@
     }
 
     private void ageUp() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.FROG, (timer) -> {})) return;
         if (this.level() instanceof ServerLevel serverlevel) {
             this.convertTo(EntityType.FROG, ConversionParams.single(this, false, false), p_409339_ -> {
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_409339_);
                 p_409339_.finalizeSpawn(serverlevel, this.level().getCurrentDifficultyAt(p_409339_.blockPosition()), EntitySpawnReason.CONVERSION, null);
                 p_409339_.setPersistenceRequired();
                 p_409339_.fudgePositionAfterSizeChange(this.getDimensions(this.getPose()));
