--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -222,8 +_,10 @@
     }
 
     private void ageUp() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.FROG, (timer) -> {})) return;
         if (this.level() instanceof ServerLevel serverlevel) {
             this.convertTo(EntityType.FROG, ConversionParams.single(this, false, false), p_432547_ -> {
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_432547_);
                 p_432547_.finalizeSpawn(serverlevel, this.level().getCurrentDifficultyAt(p_432547_.blockPosition()), EntitySpawnReason.CONVERSION, null);
                 p_432547_.setPersistenceRequired();
                 p_432547_.fudgePositionAfterSizeChange(this.getDimensions(this.getPose()));
