--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -219,8 +_,10 @@
     }
 
     private void ageUp() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.FROG, (timer) -> {})) return;
         if (this.level() instanceof ServerLevel serverlevel) {
             this.convertTo(EntityType.FROG, ConversionParams.single(this, false, false), p_454577_ -> {
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_454577_);
                 p_454577_.finalizeSpawn(serverlevel, serverlevel.getCurrentDifficultyAt(p_454577_.blockPosition()), EntitySpawnReason.CONVERSION, null);
                 p_454577_.setPersistenceRequired();
                 p_454577_.fudgePositionAfterSizeChange(this.getDimensions(this.getPose()));
