--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -228,8 +_,10 @@
     }
 
     private void ageUp() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.FROG, (timer) -> {})) return;
         if (this.level() instanceof ServerLevel serverlevel) {
             this.convertTo(EntityType.FROG, ConversionParams.single(this, false, false), p_381493_ -> {
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_381493_);
                 p_381493_.finalizeSpawn(serverlevel, this.level().getCurrentDifficultyAt(p_381493_.blockPosition()), EntitySpawnReason.CONVERSION, null);
                 p_381493_.setPersistenceRequired();
                 p_381493_.fudgePositionAfterSizeChange(this.getDimensions(this.getPose()));
