--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -76,6 +79,8 @@
                     for (int i = 0; i < 4; i++) {
                         BlockState testState = level.getBlockState(pos.below(height + 1));
                         if (!testState.is(this.plant)) {
+                            var soilDecision2 = testState.canSustainPlant(level, pos.below(height + 1), Direction.UP, state);
+                            if (!soilDecision2.isDefault()) pillarOnEndStone = soilDecision2.isTrue();
                             if (testState.is(Blocks.END_STONE)) {
                                 pillarOnEndStone = true;
                             }
