++++ Target missing
--- a/net/minecraft/client/resources/model/UnbakedGeometry.java
+++ b/net/minecraft/client/resources/model/UnbakedGeometry.java
@@ -9,5 +9,9 @@
 public interface UnbakedGeometry extends net.neoforged.neoforge.client.extensions.UnbakedGeometryExtension {
     UnbakedGeometry EMPTY = (textureSlots, modelBaker, modelState, name) -> QuadCollection.EMPTY;
 
+    /**
+     * @deprecated Neo: Use {@link #bake(TextureSlots, ModelBaker, ModelState, ModelDebugName, net.minecraft.util.context.ContextMap)}.
+     */
+    @Deprecated
     QuadCollection bake(TextureSlots textureSlots, ModelBaker modelBaker, ModelState modelState, ModelDebugName name);
 }
