--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -127,6 +_,7 @@
         if (this.tradingPlayer instanceof ServerPlayer) {
             CriteriaTriggers.TRADE.trigger((ServerPlayer)this.tradingPlayer, this, p_35274_.getResult());
         }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.TradeWithVillagerEvent(this.tradingPlayer, p_35274_, this));
     }
 
     protected abstract void rewardTradeXp(MerchantOffer p_35299_);
@@ -183,13 +_,6 @@
         }
 
         this.readInventoryFromTag(p_35290_, this.registryAccess());
-    }
-
-    @Nullable
-    @Override
-    public Entity changeDimension(DimensionTransition p_350994_) {
-        this.stopTrading();
-        return super.changeDimension(p_350994_);
     }
 
     protected void stopTrading() {
