--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -127,6 +_,7 @@
         if (this.tradingPlayer instanceof ServerPlayer) {
             CriteriaTriggers.TRADE.trigger((ServerPlayer)this.tradingPlayer, this, offer.getResult());
         }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.TradeWithVillagerEvent(this.tradingPlayer, offer, this));
     }
 
     protected abstract void rewardTradeXp(MerchantOffer offer);
@@ -183,13 +_,6 @@
         }
 
         this.readInventoryFromTag(compound, this.registryAccess());
-    }
-
-    @Nullable
-    @Override
-    public Entity changeDimension(DimensionTransition transition) {
-        this.stopTrading();
-        return super.changeDimension(transition);
     }
 
     protected void stopTrading() {
