--- a/net/minecraft/world/entity/npc/CatSpawner.java
+++ b/net/minecraft/world/entity/npc/CatSpawner.java
@@ -83,8 +_,8 @@
         if (cat == null) {
             return 0;
         } else {
+            cat.moveTo(pos, 0.0F, 0.0F); // Fix MC-147659: Some witch huts spawn the incorrect cat
             cat.finalizeSpawn(serverLevel, serverLevel.getCurrentDifficultyAt(pos), MobSpawnType.NATURAL, null);
-            cat.moveTo(pos, 0.0F, 0.0F);
             serverLevel.addFreshEntityWithPassengers(cat);
             return 1;
         }
