--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -322,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
-        if (itemstack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping()) {
+        if (itemstack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping() || player.isSecondaryUseActive()) {
             return super.mobInteract(player, hand);
         } else if (this.isBaby()) {
             this.setUnhappy();
@@ -742,8 +_,9 @@
 
     @Override
     protected Component getTypeName() {
+        net.minecraft.resources.ResourceLocation profName = BuiltInRegistries.VILLAGER_PROFESSION.getKey(this.getVillagerData().getProfession());
         return Component.translatable(
-            this.getType().getDescriptionId() + "." + BuiltInRegistries.VILLAGER_PROFESSION.getKey(this.getVillagerData().getProfession()).getPath()
+            this.getType().getDescriptionId() + '.' + (!"minecraft".equals(profName.getNamespace()) ? profName.getNamespace() + '.' : "") + profName.getPath()
         );
     }
 
@@ -799,7 +_,7 @@
 
     @Override
     public void thunderHit(ServerLevel level, LightningBolt lightning) {
-        if (level.getDifficulty() != Difficulty.PEACEFUL) {
+        if (level.getDifficulty() != Difficulty.PEACEFUL && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.WITCH, (timer) -> {})) {
             LOGGER.info("Villager {} was struck by lightning {}.", this, lightning);
             Witch witch = EntityType.WITCH.create(level);
             if (witch != null) {
@@ -812,6 +_,7 @@
                 }
 
                 witch.setPersistenceRequired();
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, witch);
                 level.addFreshEntityWithPassengers(witch);
                 this.releaseAllPois();
                 this.discard();
@@ -831,7 +_,7 @@
     @Override
     public boolean wantsToPickUp(ItemStack stack) {
         Item item = stack.getItem();
-        return (WANTED_ITEMS.contains(item) || this.getVillagerData().getProfession().requestedItems().contains(item))
+        return (WANTED_ITEMS.contains(item) || this.getVillagerData().getProfession().requestedItems().contains(item) || (stack.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(this)))
             && this.getInventory().canAddItem(stack);
     }
 
