--- a/net/minecraft/client/renderer/item/ItemStackRenderState.java
+++ b/net/minecraft/client/renderer/item/ItemStackRenderState.java
@@ -140,6 +_,13 @@
         }
 
         public void setRenderType(RenderType p_405446_) {
+            // Neo: Guard against models using chunk render types.
+            if (p_405446_.getChunkLayerId() != -1) {
+                throw new IllegalArgumentException("""
+                        Attempting to render an item BakedModel with an invalid RenderType: %s.
+                        Chunk render types are not supported, and the equivalent render types from the Sheets class should be used.
+                        """.formatted(p_405446_));
+            }
             this.renderType = p_405446_;
         }
 
