--- a/net/minecraft/client/renderer/item/ItemProperties.java
+++ b/net/minecraft/client/renderer/item/ItemProperties.java
@@ -42,6 +_,9 @@
     private static final Map<Item, Map<ResourceLocation, ItemPropertyFunction>> PROPERTIES = Maps.newHashMap();
 
     private static ClampedItemPropertyFunction registerGeneric(ResourceLocation name, ClampedItemPropertyFunction property) {
+        return (ClampedItemPropertyFunction) registerGeneric(name, (ItemPropertyFunction) property);
+    }
+    public static ItemPropertyFunction registerGeneric(ResourceLocation name, ItemPropertyFunction property) {
         GENERIC_PROPERTIES.put(name, property);
         return property;
     }
@@ -51,6 +_,10 @@
     }
 
     private static void register(Item item, ResourceLocation name, ClampedItemPropertyFunction property) {
+        register(item, name, (ItemPropertyFunction) property);
+    }
+
+    public static void register(Item item, ResourceLocation name, ItemPropertyFunction property) {
         PROPERTIES.computeIfAbsent(item, p_117828_ -> Maps.newHashMap()).put(name, property);
     }
 
