--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
@@ -21,10 +21,16 @@
 
 @OnlyIn(Dist.CLIENT)
 public interface BlockStateModel extends net.neoforged.neoforge.client.extensions.BlockStateModelExtension {
+    /// @deprecated Neo: Use [#collectParts(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, net.minecraft.world.level.block.state.BlockState, RandomSource, List)].
+    @Deprecated
     void collectParts(RandomSource random, List<BlockStateModelPart> output);
 
+    /// @deprecated Neo: Use [#particleMaterial(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, net.minecraft.world.level.block.state.BlockState)].
+    @Deprecated
     Material.Baked particleMaterial();
 
+    /// @deprecated Neo: Use [#hasTranslucency(net.minecraft.client.renderer.block.BlockAndTintGetter, net.minecraft.core.BlockPos, BlockState)] instead
+    @Deprecated
     boolean hasTranslucency();
 
     @OnlyIn(Dist.CLIENT)
