--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -51,6 +_,7 @@
    }
 
    protected void m_142310_(BlockState p_152940_, Level p_152941_, BlockPos p_152942_, Fluid p_152943_) {
+      if (!net.minecraftforge.common.ForgeHooks.handleDripInfo(p_152943_, p_152941_, p_152942_))
       if (p_152943_ == Fluids.f_76193_) {
          BlockState blockstate = Blocks.f_152476_.m_49966_();
          p_152941_.m_46597_(p_152942_, blockstate);
