--- a/net/minecraft/world/item/enchantment/EnchantmentHelper.java
+++ b/net/minecraft/world/item/enchantment/EnchantmentHelper.java
@@ -48,8 +_,7 @@
 
 public class EnchantmentHelper {
     public static int getItemEnchantmentLevel(Holder<Enchantment> p_346179_, ItemStack p_44845_) {
-        ItemEnchantments itemenchantments = p_44845_.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
-        return itemenchantments.getLevel(p_346179_);
+        return p_44845_.getEnchantmentLevel(p_346179_);
     }
 
     public static ItemEnchantments updateEnchantments(ItemStack p_331034_, Consumer<ItemEnchantments.Mutable> p_332031_) {
@@ -120,7 +_,7 @@
     }
 
     private static void runIterationOnItem(ItemStack p_345425_, EnchantmentHelper.EnchantmentVisitor p_345023_) {
-        ItemEnchantments itemenchantments = p_345425_.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
+        ItemEnchantments itemenchantments = p_345425_.getAllEnchantments(); // Neo: Allow gameplay enchantments to run too.
 
         for (Entry<Holder<Enchantment>> entry : itemenchantments.entrySet()) {
             p_345023_.accept(entry.getKey(), entry.getIntValue());
@@ -131,7 +_,7 @@
         ItemStack p_44852_, EquipmentSlot p_345566_, LivingEntity p_345792_, EnchantmentHelper.EnchantmentInSlotVisitor p_345683_
     ) {
         if (!p_44852_.isEmpty()) {
-            ItemEnchantments itemenchantments = p_44852_.get(DataComponents.ENCHANTMENTS);
+            ItemEnchantments itemenchantments = p_44852_.getAllEnchantments(); // Neo: Allow gameplay enchantments to run too.
             if (itemenchantments != null && !itemenchantments.isEmpty()) {
                 EnchantedItemInUse enchantediteminuse = new EnchantedItemInUse(p_44852_, p_345566_, p_345792_);
 
@@ -484,7 +_,7 @@
 
     public static int getEnchantmentCost(RandomSource p_220288_, int p_220289_, int p_220290_, ItemStack p_220291_) {
         Item item = p_220291_.getItem();
-        int i = item.getEnchantmentValue();
+        int i = p_220291_.getEnchantmentValue();
         if (i <= 0) {
             return 0;
         } else {
@@ -529,7 +_,7 @@
     public static List<EnchantmentInstance> selectEnchantment(RandomSource p_220298_, ItemStack p_220299_, int p_220300_, Stream<Holder<Enchantment>> p_346061_) {
         List<EnchantmentInstance> list = Lists.newArrayList();
         Item item = p_220299_.getItem();
-        int i = item.getEnchantmentValue();
+        int i = p_220299_.getEnchantmentValue();
         if (i <= 0) {
             return list;
         } else {
