--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -504,6 +_,26 @@
         return new Enchantment.Builder(p_345873_);
     }
 
+    // TODO 1.21: Make IEnchantmentExtension additions data-driven along with these methods:
+
+//    /**
+//     * This applies specifically to applying at the enchanting table. The other method {@link #canEnchant(ItemStack)}
+//     * applies for <i>all possible</i> enchantments.
+//     * @param stack
+//     * @return
+//     */
+//    public boolean canApplyAtEnchantingTable(ItemStack stack) {
+//        return stack.canApplyAtEnchantingTable(this);
+//    }
+//
+//    /**
+//     * Is this enchantment allowed to be enchanted on books via Enchantment Table
+//     * @return false to disable the vanilla feature
+//     */
+//    public boolean isAllowedOnBooks() {
+//        return true;
+//    }
+
     public static class Builder {
         private final Enchantment.EnchantmentDefinition definition;
         private HolderSet<Enchantment> exclusiveSet = HolderSet.direct();
