--- a/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -95,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player player, InteractionHand hand) {
+        InteractionResult ret = super.interact(player, hand);
+        if (ret.consumesAction()) return ret;
         return this.interactWithContainerVehicle(player);
     }
 
