--- a/net/minecraft/world/entity/vehicle/ChestBoat.java
+++ b/net/minecraft/world/entity/vehicle/ChestBoat.java
@@ -122,7 +_,8 @@
             case DARK_OAK -> Items.DARK_OAK_CHEST_BOAT;
             case MANGROVE -> Items.MANGROVE_CHEST_BOAT;
             case BAMBOO -> Items.BAMBOO_CHEST_RAFT;
-            default -> Items.OAK_CHEST_BOAT;
+            case OAK -> Items.OAK_CHEST_BOAT;
+            default -> this.getVariant().chestBoatItem.get();
         };
     }
 
