--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -107,6 +_,9 @@
 
             this.setLootTable(null);
             LootParams.Builder lootparams$builder = new LootParams.Builder((ServerLevel)this.level()).withParameter(LootContextParams.ORIGIN, this.position());
+            // Neo: set the chest to attacking_entity for loot context.
+            if (this instanceof AbstractMinecartContainer entityContainer)
+                lootparams$builder.withParameter(LootContextParams.ATTACKING_ENTITY, entityContainer);
             if (p_219950_ != null) {
                 lootparams$builder.withLuck(p_219950_.getLuck()).withParameter(LootContextParams.THIS_ENTITY, p_219950_);
             }
@@ -173,6 +_,6 @@
     }
 
     default boolean isChestVehicleStillValid(Player p_219955_) {
-        return !this.isRemoved() && p_219955_.canInteractWithEntity(this.getBoundingBox(), 4.0);
+        return !this.isRemoved() && net.neoforged.neoforge.common.CommonHooks.menuRangeValidity(p_219955_, this, 4.0);
     }
 }
