--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -107,6 +_,9 @@
 
             this.setLootTable(null);
             LootParams.Builder lootparams$builder = new LootParams.Builder((ServerLevel)this.level()).withParameter(LootContextParams.ORIGIN, this.position());
+            // Neo: set the chest to attacking_entity for loot context.
+            if (this instanceof AbstractMinecartContainer entityContainer)
+                lootparams$builder.withParameter(LootContextParams.ATTACKING_ENTITY, entityContainer);
             if (player != null) {
                 lootparams$builder.withLuck(player.getLuck()).withParameter(LootContextParams.THIS_ENTITY, player);
             }
