--- a/net/minecraft/world/entity/vehicle/MinecartFurnace.java
+++ b/net/minecraft/world/entity/vehicle/MinecartFurnace.java
@@ -116,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player player, InteractionHand hand) {
+        InteractionResult ret = super.interact(player, hand);
+        if (ret.consumesAction()) return ret;
         ItemStack itemstack = player.getItemInHand(hand);
         if (INGREDIENT.test(itemstack) && this.fuel + 3600 <= 32000) {
             itemstack.consume(1, player);
@@ -128,6 +_,11 @@
         }
 
         return InteractionResult.sidedSuccess(this.level().isClientSide);
+    }
+
+    @Override
+    public float getMaxCartSpeedOnRail() {
+        return 0.2f;
     }
 
     @Override
