--- a/net/minecraft/world/entity/vehicle/Minecart.java
+++ b/net/minecraft/world/entity/vehicle/Minecart.java
@@ -19,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player player, InteractionHand hand) {
+        InteractionResult ret = super.interact(player, hand);
+        if (ret.consumesAction()) return ret;
         if (player.isSecondaryUseActive()) {
             return InteractionResult.PASS;
         } else if (this.isVehicle()) {
