--- a/net/minecraft/world/entity/vehicle/MinecartSpawner.java
+++ b/net/minecraft/world/entity/vehicle/MinecartSpawner.java
@@ -17,6 +_,11 @@
         public void broadcastEvent(Level p_150342_, BlockPos p_150343_, int p_150344_) {
             p_150342_.broadcastEntityEvent(MinecartSpawner.this, (byte)p_150344_);
         }
+
+        @Override
+        public com.mojang.datafixers.util.Either<net.minecraft.world.level.block.entity.BlockEntity, net.minecraft.world.entity.Entity> getOwner() {
+            return com.mojang.datafixers.util.Either.right(MinecartSpawner.this);
+        }
     };
     private final Runnable ticker;
 
