--- a/net/minecraft/data/models/BlockModelGenerators.java
+++ b/net/minecraft/data/models/BlockModelGenerators.java
@@ -82,7 +_,7 @@
 import net.minecraft.world.level.block.state.properties.Tilt;
 import net.minecraft.world.level.block.state.properties.WallSide;
 
-public class BlockModelGenerators {
+public class BlockModelGenerators implements net.neoforged.neoforge.common.extensions.IBlockModelGenerators {
     public final Consumer<BlockStateGenerator> blockStateOutput;
     public final BiConsumer<ResourceLocation, Supplier<JsonElement>> modelOutput;
     public final Consumer<Item> skippedAutoModelsOutput;
@@ -207,12 +_,23 @@
         return createRotatedVariant(p_176180_, p_176181_, resourcelocation).with(createRotatedPillar());
     }
 
+    public final String modId;
+
+    // Neo: Use mod id sensitive version
+    @Deprecated
     public BlockModelGenerators(
         Consumer<BlockStateGenerator> p_124481_, BiConsumer<ResourceLocation, Supplier<JsonElement>> p_124482_, Consumer<Item> p_124483_
     ) {
+        this(p_124481_, p_124482_, p_124483_, ResourceLocation.DEFAULT_NAMESPACE);
+    }
+
+    public BlockModelGenerators(
+        Consumer<BlockStateGenerator> p_124481_, BiConsumer<ResourceLocation, Supplier<JsonElement>> p_124482_, Consumer<Item> p_124483_, String modId
+    ) {
         this.blockStateOutput = p_124481_;
         this.modelOutput = p_124482_;
         this.skippedAutoModelsOutput = p_124483_;
+        this.modId = modId;
     }
 
     public void skipAutoItemBlock(Block p_124525_) {
