--- a/net/minecraft/data/models/ItemModelGenerators.java
+++ b/net/minecraft/data/models/ItemModelGenerators.java
@@ -24,7 +_,7 @@
 import net.minecraft.world.item.equipment.EquipmentModels;
 import net.minecraft.world.item.equipment.Equippable;
 
-public class ItemModelGenerators {
+public class ItemModelGenerators implements net.neoforged.neoforge.common.extensions.IItemModelGenerators {
     public static final ResourceLocation TRIM_TYPE_PREDICATE_ID = ResourceLocation.withDefaultNamespace("trim_type");
     public static final List<ItemModelGenerators.TrimModelData> GENERATED_TRIM_MODELS = List.of(
         new ItemModelGenerators.TrimModelData("quartz", 0.1F, Map.of()),
@@ -39,9 +_,17 @@
         new ItemModelGenerators.TrimModelData("amethyst", 1.0F, Map.of())
     );
     public final BiConsumer<ResourceLocation, Supplier<JsonElement>> output;
+    public final String modId;
 
+    // Neo: Use mod id sensitive version
+    @Deprecated
     public ItemModelGenerators(BiConsumer<ResourceLocation, Supplier<JsonElement>> p_125082_) {
+        this(p_125082_, ResourceLocation.DEFAULT_NAMESPACE);
+    }
+
+    public ItemModelGenerators(BiConsumer<ResourceLocation, Supplier<JsonElement>> p_125082_, String modId) {
         this.output = p_125082_;
+        this.modId = modId;
     }
 
     public void generateFlatItem(Item p_125089_, ModelTemplate p_125090_) {
