--- a/net/minecraft/data/models/ModelProvider.java
+++ b/net/minecraft/data/models/ModelProvider.java
@@ -24,13 +_,26 @@
 import net.minecraft.world.item.Item;
 import net.minecraft.world.level.block.Block;
 
-public class ModelProvider implements DataProvider {
+public class ModelProvider implements DataProvider, net.neoforged.neoforge.common.extensions.IModelProviderExtension {
     private final PackOutput.PathProvider blockStatePathProvider;
     private final PackOutput.PathProvider modelPathProvider;
+    public final String modId;
 
+    // NeoForge: Use the above constructor passing in their mod id
+    @Deprecated
     public ModelProvider(PackOutput p_252226_) {
+        this(p_252226_, ResourceLocation.DEFAULT_NAMESPACE);
+    }
+
+    public ModelProvider(PackOutput p_252226_, String modId) {
         this.blockStatePathProvider = p_252226_.createPathProvider(PackOutput.Target.RESOURCE_PACK, "blockstates");
         this.modelPathProvider = p_252226_.createPathProvider(PackOutput.Target.RESOURCE_PACK, "models");
+        this.modId = modId;
+    }
+
+    protected void registerModels(BlockModelGenerators blockModels, ItemModelGenerators itemModels) {
+        blockModels.run();
+        itemModels.run();
     }
 
     @Override
@@ -52,19 +_,18 @@
             }
         };
         Consumer<Item> consumer1 = set::add;
-        new BlockModelGenerators(consumer, biconsumer, consumer1).run();
-        new ItemModelGenerators(biconsumer).run();
+        registerModels(new BlockModelGenerators(consumer, biconsumer, consumer1, this.modId), new ItemModelGenerators(biconsumer, this.modId));
         List<Block> list = BuiltInRegistries.BLOCK
             .entrySet()
             .stream()
-            .filter(p_300706_ -> true)
+            .filter(p_300706_ -> p_300706_.getKey().location().getNamespace().equals(modId))
             .map(Entry::getValue)
             .filter(p_125117_ -> !map.containsKey(p_125117_))
             .toList();
         if (!list.isEmpty()) {
             throw new IllegalStateException("Missing blockstate definitions for: " + list);
         } else {
-            BuiltInRegistries.BLOCK.forEach(p_125128_ -> {
+            BuiltInRegistries.BLOCK.stream().filter(block -> BuiltInRegistries.BLOCK.getKey(block).getNamespace().equals(modId)).forEach(p_125128_ -> {
                 Item item = Item.BY_BLOCK.get(p_125128_);
                 if (item != null) {
                     if (set.contains(item)) {
