--- a/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/TeleportRandomlyConsumeEffect.java
@@ -54,7 +_,9 @@
             }
 
             Vec3 oldPos = user.position();
-            if (user.randomTeleport(xx, yy, zz, true)) {
+            var event = net.neoforged.neoforge.event.EventHooks.onItemConsumptionTeleport(user, stack, xx, yy, zz);
+            if (event.isCanceled()) return false;
+            if (user.randomTeleport(event.getTargetX(), event.getTargetY(), event.getTargetZ(), true)) {
                 level.gameEvent(GameEvent.TELEPORT, oldPos, GameEvent.Context.of(user));
                 SoundSource soundSource;
                 SoundEvent soundEvent;
