--- a/net/minecraft/client/resources/SplashManager.java
+++ b/net/minecraft/client/resources/SplashManager.java
@@ -41,8 +_,12 @@
 
     protected List<Component> prepare(ResourceManager manager, ProfilerFiller profiler) {
         try {
+            var splashes = Minecraft.getInstance().getResourceManager().getResourceOrThrow(SPLASHES_LOCATION);
+            if (splashes.sourcePackId().equals("vanilla"))
+                return net.neoforged.neoforge.client.resources.NeoForgeSplashHooks.loadSplashes(manager);
+
             List var4;
-            try (BufferedReader reader = Minecraft.getInstance().getResourceManager().openAsReader(SPLASHES_LOCATION)) {
+            try (BufferedReader reader = splashes.openAsReader()) {
                 var4 = reader.lines().map(String::trim).filter(line -> line.hashCode() != 125780783).map(SplashManager::literalSplash).toList();
             }
 
