--- a/net/minecraft/client/resources/SplashManager.java
+++ b/net/minecraft/client/resources/SplashManager.java
@@ -40,7 +_,16 @@
     }
 
     protected List<Component> prepare(ResourceManager manager, ProfilerFiller profiler) {
-        try (BufferedReader reader = Minecraft.getInstance().getResourceManager().openAsReader(SPLASHES_LOCATION)) {
+        net.minecraft.server.packs.resources.Resource splashes;
+        try {
+            splashes = Minecraft.getInstance().getResourceManager().getResourceOrThrow(SPLASHES_LOCATION);
+        } catch (IOException e) {
+            return List.of();
+        }
+        if (splashes.sourcePackId().equals("vanilla") || !net.neoforged.neoforge.client.config.NeoForgeClientConfig.INSTANCE.skipModdedSplashesIfVanillaModified.getAsBoolean())
+            return net.neoforged.neoforge.client.resources.NeoForgeSplashHooks.loadSplashes(manager);
+
+        try (BufferedReader reader = splashes.openAsReader()) {
             return reader.lines().map(String::trim).filter(line -> line.hashCode() != 125780783).map(SplashManager::literalSplash).toList();
         } catch (IOException ignored) {
             return List.of();
