--- a/net/minecraft/core/DefaultedMappedRegistry.java
+++ b/net/minecraft/core/DefaultedMappedRegistry.java
@@ -38,6 +_,12 @@
         return identifier == null ? this.defaultKey : identifier;
     }
 
+    @Nullable
+    @Override
+    public Identifier getKeyOrNull(T element) {
+        return super.getKey(element);
+    }
+
     @Override
     public T getValue(@Nullable Identifier p_468616_) {
         T t = super.getValue(p_468616_);
