--- a/net/minecraft/core/HolderLookup.java
+++ b/net/minecraft/core/HolderLookup.java
@@ -26,7 +_,7 @@
         return this.listTags().map(HolderSet.Named::key);
     }
 
-    public interface Provider {
+    public interface Provider extends net.neoforged.neoforge.common.extensions.IHolderLookupProviderExtension {
         Stream<ResourceKey<? extends Registry<?>>> listRegistries();
 
         <T> Optional<HolderLookup.RegistryLookup<T>> lookup(ResourceKey<? extends Registry<? extends T>> p_256285_);
@@ -96,6 +_,11 @@
             };
         }
 
+        @org.jetbrains.annotations.Nullable
+        default <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> attachment, ResourceKey<T> key) {
+            return null;
+        }
+
         public interface Delegate<T> extends HolderLookup.RegistryLookup<T> {
             HolderLookup.RegistryLookup<T> parent();
 
@@ -127,6 +_,12 @@
             @Override
             default Stream<HolderSet.Named<T>> listTags() {
                 return this.parent().listTags();
+            }
+
+            @Override
+            @org.jetbrains.annotations.Nullable
+            default <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> attachment, ResourceKey<T> key) {
+                return parent().getData(attachment, key);
             }
         }
     }
