--- a/net/minecraft/core/HolderSet.java
+++ b/net/minecraft/core/HolderSet.java
@@ -174,6 +_,9 @@
 
         void bind(List<Holder<T>> contents) {
             this.contents = List.copyOf(contents);
+            for (Runnable runnable : this.invalidationCallbacks) {
+                runnable.run(); // FORGE: invalidate listeners when tags rebind
+            }
         }
 
         public TagKey<T> key() {
