--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -276,6 +_,48 @@
             return "Reference{" + this.key + "=" + this.value + "}";
         }
 
+        // Neo: Add DeferredHolder-compatible hashCode() overrides
+        @Override
+        public int hashCode() {
+            return this.key == null ? super.hashCode() : this.key.hashCode();
+        }
+
+        // Neo: Add DeferredHolder-compatible equals() overrides
+        @Override
+        public boolean equals(@Nullable Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj instanceof Holder.Reference<?> h) {
+                // Return false like vanilla would do if they are intrusive holders without bound keys and not reference-equal
+                return h.keyOrNull() != null && this.keyOrNull() != null && h.key() == this.key();
+            }
+            return false;
+        }
+
+        /// Neo: {@return the registry lookup owning this holder, if available}
+        @Override
+        public HolderLookup.@Nullable RegistryLookup<T> unwrapLookup() {
+            return this.owner instanceof HolderLookup.RegistryLookup<T> rl ? rl : null;
+        }
+
+        /// Neo: {@return the key of this holder if bound, else `null`}
+        @Nullable
+        @Override
+        public ResourceKey<T> keyOrNull() {
+            return this.key;
+        }
+
+        @Nullable
+        @Override
+        public <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type) {
+            HolderLookup.RegistryLookup<T> lookup = unwrapLookup();
+            if (lookup != null) {
+                return lookup.getData(type, key());
+            }
+            return null;
+        }
+
         protected enum Type {
             STAND_ALONE,
             INTRUSIVE;
