--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -11,7 +_,7 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.TagKey;
 
-public interface Holder<T> {
+public interface Holder<T> extends net.neoforged.neoforge.registries.datamaps.IWithData<T> {
     T value();
 
     boolean isBound();
@@ -41,6 +_,15 @@
         return this.unwrapKey().map(p_316542_ -> p_316542_.location().toString()).orElse("[unregistered]");
     }
 
+    /**
+     * {@return the holder that this holder wraps}
+     *
+     * Useful for holders that delegate to another holder.
+     */
+    default Holder<T> getDelegate() {
+        return this;
+    }
+
     static <T> Holder<T> direct(T p_205710_) {
         return new Holder.Direct<>(p_205710_);
     }
@@ -220,6 +_,14 @@
             }
         }
 
+        @org.jetbrains.annotations.Nullable
+        public <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type) {
+            if (owner instanceof HolderLookup.RegistryLookup<T> lookup) {
+                return lookup.getData(type, key());
+            }
+            return null;
+        }
+
         public void bindTags(Collection<TagKey<T>> p_205770_) {
             this.tags = Set.copyOf(p_205770_);
         }
@@ -232,6 +_,18 @@
         @Override
         public String toString() {
             return "Reference{" + this.key + "=" + this.value + "}";
+        }
+
+        // Neo: Add DeferredHolder-compatible hashCode() and equals() overrides
+        @Override
+        public int hashCode() {
+            return key().hashCode();
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) return true;
+            return obj instanceof Holder<?> h && h.kind() == Kind.REFERENCE && h.unwrapKey().orElseThrow() == this.key();
         }
 
         public static enum Type {
