--- a/net/minecraft/core/HolderSet.java
+++ b/net/minecraft/core/HolderSet.java
@@ -174,6 +_,9 @@
 
         void bind(List<Holder<T>> p_205836_) {
             this.contents = List.copyOf(p_205836_);
+            for (Runnable runnable : this.invalidationCallbacks) {
+                runnable.run(); // FORGE: invalidate listeners when tags rebind
+            }
         }
 
         public TagKey<T> key() {
@@ -217,6 +_,11 @@
         @Override
         public boolean canSerializeIn(HolderOwner<T> p_256542_) {
             return this.owner.canSerializeIn(p_256542_);
+        }
+        // FORGE: Keep a list of invalidation callbacks so they can be run when tags rebind
+        private List<Runnable> invalidationCallbacks = new java.util.ArrayList<>();
+        public void addInvalidationListener(Runnable runnable) {
+            invalidationCallbacks.add(runnable);
         }
     }
 }
