--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -47,7 +_,7 @@
     }
 
     private DataResult<Holder.Reference<T>> safeCastToReference(Holder<T> holder) {
-        return holder instanceof Holder.Reference<T> reference
+        return holder.getDelegate() instanceof Holder.Reference reference
             ? DataResult.success(reference)
             : DataResult.error(() -> "Unregistered holder in " + this.key() + ": " + holder);
     }
