--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -22,7 +_,7 @@
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 
-public interface Registry<T> extends Keyable, IdMap<T> {
+public interface Registry<T> extends Keyable, IdMap<T>, net.neoforged.neoforge.registries.IRegistryExtension<T> {
     ResourceKey<? extends Registry<T>> key();
 
     default Codec<T> byNameCodec() {
