--- a/net/minecraft/core/RegistrySetBuilder.java
+++ b/net/minecraft/core/RegistrySetBuilder.java
@@ -72,6 +_,10 @@
         return this.add(key, Lifecycle.stable(), bootstrap);
     }
 
+    public List<? extends ResourceKey<? extends Registry<?>>> getEntryKeys() {
+        return this.entries.stream().map(RegistrySetBuilder.RegistryStub::key).toList();
+    }
+
     private RegistrySetBuilder.BuildState createState(RegistryAccess registryAccess) {
         RegistrySetBuilder.BuildState registrysetbuilder$buildstate = RegistrySetBuilder.BuildState.create(
             registryAccess, this.entries.stream().map(RegistrySetBuilder.RegistryStub::key)
@@ -185,16 +_,22 @@
                 lazyholder.supplier = () -> cloner.clone((T)p_311483_.value(), registryLookupProvider, object.getValue());
                 map.put(resourcekey, lazyholder);
             });
-            HolderLookup.RegistryLookup<T> registrylookup1 = lookupProvider.lookupOrThrow(registryKey);
-            registrylookup1.listElements().forEach(p_311506_ -> {
-                ResourceKey<T> resourcekey = p_311506_.key();
-                map.computeIfAbsent(resourcekey, p_311494_ -> {
-                    RegistrySetBuilder.LazyHolder<T> lazyholder = new RegistrySetBuilder.LazyHolder<>(owner, resourcekey);
-                    lazyholder.supplier = () -> cloner.clone((T)p_311506_.value(), lookupProvider, object.getValue());
-                    return lazyholder;
+            Optional<HolderLookup.RegistryLookup<T>> lookup = lookupProvider.lookup(registryKey);
+            Lifecycle lifecycle;
+            if (lookup.isPresent()) {
+                HolderLookup.RegistryLookup<T> registrylookup1 = lookup.get();
+                registrylookup1.listElements().forEach(p_311506_ -> {
+                    ResourceKey<T> resourcekey = p_311506_.key();
+                    map.computeIfAbsent(resourcekey, p_311494_ -> {
+                        RegistrySetBuilder.LazyHolder<T> lazyholder = new RegistrySetBuilder.LazyHolder<>(owner, resourcekey);
+                        lazyholder.supplier = () -> cloner.clone((T) p_311506_.value(), lookupProvider, object.getValue());
+                        return lazyholder;
+                    });
                 });
-            });
-            Lifecycle lifecycle = registrylookup.registryLifecycle().add(registrylookup1.registryLifecycle());
+                lifecycle = registrylookup.registryLifecycle().add(registrylookup1.registryLifecycle());
+            } else {
+                lifecycle = registrylookup.registryLifecycle();
+            }
             return lookupFromMap(registryKey, lifecycle, owner, map);
         }
     }
@@ -236,7 +_,7 @@
             RegistrySetBuilder.UniversalLookup registrysetbuilder$universallookup = new RegistrySetBuilder.UniversalLookup(registrysetbuilder$universalowner);
             Builder<ResourceLocation, HolderGetter<?>> builder = ImmutableMap.builder();
             registryAccess.registries()
-                .forEach(p_258197_ -> builder.put(p_258197_.key().location(), RegistrySetBuilder.wrapContextLookup(p_258197_.value().asLookup())));
+                .forEach(p_258197_ -> builder.put(p_258197_.key().location(), net.neoforged.neoforge.common.CommonHooks.wrapRegistryLookup(p_258197_.value().asLookup())));
             registries.forEach(p_256603_ -> builder.put(p_256603_.location(), registrysetbuilder$universallookup));
             return new RegistrySetBuilder.BuildState(
                 registrysetbuilder$universalowner, registrysetbuilder$universallookup, builder.build(), new HashMap<>(), list
@@ -260,6 +_,11 @@
                 @Override
                 public <S> HolderGetter<S> lookup(ResourceKey<? extends Registry<? extends S>> p_255961_) {
                     return (HolderGetter<S>)BuildState.this.registries.getOrDefault(p_255961_.location(), BuildState.this.lookup);
+                }
+
+                @Override
+                public <S> Optional<HolderLookup.RegistryLookup<S>> registryLookup(ResourceKey<? extends Registry<? extends S>> registry) {
+                    return Optional.ofNullable((HolderLookup.RegistryLookup<S>) BuildState.this.registries.get(registry.location()));
                 }
             };
         }
