--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -48,7 +_,7 @@
     }
 
     private DataResult<Holder.Reference<T>> safeCastToReference(Holder<T> p_326365_) {
-        return p_326365_ instanceof Holder.Reference<T> reference
+        return p_326365_.getDelegate() instanceof Holder.Reference reference
             ? DataResult.success(reference)
             : DataResult.error(() -> "Unregistered holder in " + this.key() + ": " + p_326365_);
     }
