--- a/net/minecraft/core/Registry.java
+++ b/net/minecraft/core/Registry.java
@@ -49,10 +_,13 @@
     }
 
     private DataResult<Holder.Reference<T>> safeCastToReference(Holder<T> holder) {
-        return holder instanceof Holder.Reference<T> reference
+        return holder.getDelegate() instanceof Holder.Reference reference
             ? DataResult.success(reference)
             : DataResult.error(() -> "Unregistered holder in " + this.key() + ": " + holder);
     }
+
+    @Override//Neo: Force Implementation of the method from IRegistryExtension rather than defaulting to the empty map implementation from RegistryLookup
+    <A> java.util.Map<ResourceKey<T>, A> getDataMap(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type);
 
     @Override
     default <U> Stream<U> keys(DynamicOps<U> ops) {
