--- a/net/minecraft/core/RegistrySetBuilder.java
+++ b/net/minecraft/core/RegistrySetBuilder.java
@@ -70,6 +_,10 @@
         return this.add(key, Lifecycle.stable(), bootstrap);
     }
 
+    public List<? extends ResourceKey<? extends Registry<?>>> getEntryKeys() {
+        return this.entries.stream().map(RegistrySetBuilder.RegistryStub::key).toList();
+    }
+
     private RegistrySetBuilder.BuildState createState(RegistryAccess context) {
         RegistrySetBuilder.BuildState state = RegistrySetBuilder.BuildState.create(context, this.entries.stream().map(RegistrySetBuilder.RegistryStub::key));
         this.entries.forEach(e -> e.apply(state));
@@ -187,16 +_,22 @@
                 holder.supplier = () -> cloner.clone((T)elementHolder.value(), patchProvider, targetProvider.get());
                 entries.put(elementKey, holder);
             });
-            HolderLookup.RegistryLookup<T> fallbackContents = fallbackProvider.lookupOrThrow(registryKey);
-            fallbackContents.listElements().forEach(elementHolder -> {
-                ResourceKey<T> elementKey = elementHolder.key();
-                entries.computeIfAbsent(elementKey, key -> {
-                    RegistrySetBuilder.LazyHolder<T> holder = new RegistrySetBuilder.LazyHolder<>(owner, elementKey);
-                    holder.supplier = () -> cloner.clone((T)elementHolder.value(), fallbackProvider, targetProvider.get());
-                    return holder;
+            Optional<? extends HolderLookup.RegistryLookup<T>> fallbackContents = fallbackProvider.lookup(registryKey);
+            Lifecycle lifecycle;
+            if (fallbackContents.isPresent()) {
+                HolderLookup.RegistryLookup<T> registrylookup1 = fallbackContents.get();
+                registrylookup1.listElements().forEach(elementHolder -> {
+                    ResourceKey<T> elementKey = elementHolder.key();
+                    entries.computeIfAbsent(elementKey, key -> {
+                        RegistrySetBuilder.LazyHolder<T> holder = new RegistrySetBuilder.LazyHolder<>(owner, elementKey);
+                        holder.supplier = () -> cloner.clone((T) elementHolder.value(), fallbackProvider, targetProvider.get());
+                        return holder;
+                    });
                 });
-            });
-            Lifecycle lifecycle = patchContents.registryLifecycle().add(fallbackContents.registryLifecycle());
+                lifecycle = patchContents.registryLifecycle().add(registrylookup1.registryLifecycle());
+            } else {
+                lifecycle = patchContents.registryLifecycle();
+            }
             return lookupFromMap(registryKey, lifecycle, owner, entries);
         }
     }
@@ -242,7 +_,7 @@
             RegistrySetBuilder.UniversalLookup lookup = new RegistrySetBuilder.UniversalLookup(owner);
             Builder<Identifier, HolderGetter<?>> registries = ImmutableMap.builder();
             context.registries()
-                .forEach(contextRegistry -> registries.put(contextRegistry.key().identifier(), RegistrySetBuilder.wrapContextLookup(contextRegistry.value())));
+                .forEach(contextRegistry -> registries.put(contextRegistry.key().identifier(), net.neoforged.neoforge.common.CommonHooks.wrapRegistryLookup(contextRegistry.value())));
             newRegistries.forEach(newRegistry -> registries.put(newRegistry.identifier(), lookup));
             return new RegistrySetBuilder.BuildState(owner, lookup, registries.build(), new HashMap<>(), errors);
         }
@@ -269,6 +_,11 @@
                 public <S> HolderGetter<S> lookup(ResourceKey<? extends Registry<? extends S>> key) {
                     return (HolderGetter<S>)BuildState.this.registries.getOrDefault(key.identifier(), BuildState.this.lookup);
                 }
+
+                @Override
+                public <S> Optional<HolderLookup.RegistryLookup<S>> registryLookup(ResourceKey<? extends Registry<? extends S>> registry) {
+                    return Optional.ofNullable((HolderLookup.RegistryLookup<S>) BuildState.this.registries.get(registry.identifier()));
+                }
             };
         }
 
@@ -419,6 +_,26 @@
         @Override
         public Optional<Holder.Reference<Object>> get(ResourceKey<Object> id) {
             return Optional.of(this.getOrCreate(id));
+        }
+
+        @Override
+        public Stream<Holder.Reference<Object>> listElements() {
+            return holders.values().stream();
+        }
+
+        @Override
+        public Stream<ResourceKey<Object>> listElementIds() {
+            return holders.keySet().stream();
+        }
+
+        @Override
+        public Stream<HolderSet.Named<Object>> listTags() {
+            return Stream.empty();
+        }
+
+        @Override
+        public Stream<TagKey<Object>> listTagIds() {
+            return Stream.empty();
         }
 
         private <T> Holder.Reference<T> getOrCreate(ResourceKey<T> id) {
