--- a/net/minecraft/core/WritableRegistry.java
+++ b/net/minecraft/core/WritableRegistry.java
@@ -6,6 +_,8 @@
 public interface WritableRegistry<T> extends Registry<T> {
    Holder.Reference<T> register(ResourceKey<T> p_256320_, T p_255978_, Lifecycle p_256625_);
 
+   Holder.Reference<T> registerWithId(ResourceKey<T> key, T value, Lifecycle lifecycle, int id);
+
    boolean isEmpty();
 
    HolderGetter<T> createRegistrationLookup();
