--- a/net/minecraft/client/resources/model/sprite/AtlasManager.java
+++ b/net/minecraft/client/resources/model/sprite/AtlasManager.java
@@ -53,6 +_,7 @@
     private int maxMipmapLevels;
 
     public AtlasManager(TextureManager textureManager, int maxMipmapLevels) {
+        List<AtlasManager.AtlasConfig> KNOWN_ATLASES = net.neoforged.neoforge.client.ClientHooks.gatherTextureAtlases(AtlasManager.KNOWN_ATLASES);
         for (AtlasManager.AtlasConfig info : KNOWN_ATLASES) {
             TextureAtlas atlasTexture = new TextureAtlas(info.textureId);
             textureManager.register(info.textureId, atlasTexture);
