--- a/net/minecraft/commands/arguments/coordinates/BlockPosArgument.java
+++ b/net/minecraft/commands/arguments/coordinates/BlockPosArgument.java
@@ -36,9 +_,9 @@
 
     public static BlockPos getLoadedBlockPos(CommandContext<CommandSourceStack> context, ServerLevel level, String name) throws CommandSyntaxException {
         BlockPos blockpos = getBlockPos(context, name);
-        if (!level.hasChunkAt(blockpos)) {
+        if (!context.getSource().getUnsidedLevel().hasChunkAt(blockpos)) {
             throw ERROR_NOT_LOADED.create();
-        } else if (!level.isInWorldBounds(blockpos)) {
+        } else if (!context.getSource().getUnsidedLevel().isInWorldBounds(blockpos)) {
             throw ERROR_OUT_OF_WORLD.create();
         } else {
             return blockpos;
