--- a/net/minecraft/world/item/alchemy/PotionBrewing.java
+++ b/net/minecraft/world/item/alchemy/PotionBrewing.java
@@ -19,15 +_,37 @@
     private final List<Ingredient> containers;
     private final List<PotionBrewing.Mix<Potion>> potionMixes;
     private final List<PotionBrewing.Mix<Item>> containerMixes;
+    private final net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry registry;
 
     PotionBrewing(List<Ingredient> p_340915_, List<PotionBrewing.Mix<Potion>> p_341164_, List<PotionBrewing.Mix<Item>> p_341170_) {
+        this(p_340915_, p_341164_, p_341170_, List.of());
+    }
+
+    PotionBrewing(List<Ingredient> p_340915_, List<PotionBrewing.Mix<Potion>> p_341164_, List<PotionBrewing.Mix<Item>> p_341170_, List<net.neoforged.neoforge.common.brewing.IBrewingRecipe> recipes) {
         this.containers = p_340915_;
         this.potionMixes = p_341164_;
         this.containerMixes = p_341170_;
+        this.registry = new net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry(recipes);
     }
 
     public boolean isIngredient(ItemStack p_43507_) {
-        return this.isContainerIngredient(p_43507_) || this.isPotionIngredient(p_43507_);
+        return this.registry.isValidIngredient(p_43507_) || this.isContainerIngredient(p_43507_) || this.isPotionIngredient(p_43507_);
+    }
+
+    /**
+     * Checks if an item stack is a valid input for brewing,
+     * for use in the lower 3 slots where water bottles would normally go.
+     */
+    public boolean isInput(ItemStack stack) {
+        return this.registry.isValidInput(stack) || isContainer(stack);
+    }
+
+    /**
+     * Retrieves recipes that use the more general interface.
+     * This does NOT include the container and potion mixes.
+     */
+    public List<net.neoforged.neoforge.common.brewing.IBrewingRecipe> getRecipes() {
+        return registry.recipes();
     }
 
     private boolean isContainer(ItemStack p_341168_) {
@@ -71,6 +_,7 @@
     }
 
     public boolean hasMix(ItemStack p_43509_, ItemStack p_43510_) {
+        if (registry.hasOutput(p_43509_, p_43510_)) return true;
         return !this.isContainer(p_43509_) ? false : this.hasContainerMix(p_43509_, p_43510_) || this.hasPotionMix(p_43509_, p_43510_);
     }
 
@@ -103,6 +_,8 @@
         if (p_43531_.isEmpty()) {
             return p_43531_;
         } else {
+            var customMix = registry.getOutput(p_43531_, p_43530_); // Parameters are swapped compared to what vanilla passes!
+            if (!customMix.isEmpty()) return customMix;
             Optional<Holder<Potion>> optional = p_43531_.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).potion();
             if (optional.isEmpty()) {
                 return p_43531_;
@@ -127,6 +_,7 @@
     public static PotionBrewing bootstrap(FeatureFlagSet p_341301_) {
         PotionBrewing.Builder potionbrewing$builder = new PotionBrewing.Builder(p_341301_);
         addVanillaMixes(potionbrewing$builder);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.brewing.RegisterBrewingRecipesEvent(potionbrewing$builder));
         return potionbrewing$builder.build();
     }
 
@@ -194,6 +_,7 @@
         private final List<Ingredient> containers = new ArrayList<>();
         private final List<PotionBrewing.Mix<Potion>> potionMixes = new ArrayList<>();
         private final List<PotionBrewing.Mix<Item>> containerMixes = new ArrayList<>();
+        private final List<net.neoforged.neoforge.common.brewing.IBrewingRecipe> recipes = new ArrayList<>();
         private final FeatureFlagSet enabledFeatures;
 
         public Builder(FeatureFlagSet p_340975_) {
@@ -237,8 +_,26 @@
             }
         }
 
+        /**
+         * Adds a new simple brewing recipe.
+         *
+         * @param input      the ingredient that goes in the same slot as water bottles would
+         * @param ingredient the ingredient that goes in the same slot as nether wart would
+         * @param output     the item stack that will replace the input once brewing is done
+         */
+        public void addRecipe(Ingredient input, Ingredient ingredient, ItemStack output) {
+            addRecipe(new net.neoforged.neoforge.common.brewing.BrewingRecipe(input, ingredient, output));
+        }
+
+        /**
+         * Adds a new brewing recipe with custom logic.
+         */
+        public void addRecipe(net.neoforged.neoforge.common.brewing.IBrewingRecipe recipe) {
+            this.recipes.add(recipe);
+        }
+
         public PotionBrewing build() {
-            return new PotionBrewing(List.copyOf(this.containers), List.copyOf(this.potionMixes), List.copyOf(this.containerMixes));
+            return new PotionBrewing(List.copyOf(this.containers), List.copyOf(this.potionMixes), List.copyOf(this.containerMixes), List.copyOf(this.recipes));
         }
     }
 
