--- a/net/minecraft/world/item/alchemy/Potion.java
+++ b/net/minecraft/world/item/alchemy/Potion.java
@@ -24,6 +_,8 @@
     private final String name;
     private final List<MobEffectInstance> effects;
     private FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
+    private it.unimi.dsi.fastutil.objects.ReferenceSet<net.neoforged.neoforge.flag.Flag> requiredFlags = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
+    private java.util.Set<net.neoforged.neoforge.flag.Flag> requiredFlagsView = it.unimi.dsi.fastutil.objects.ReferenceSets.unmodifiable(requiredFlags);
 
     public Potion(MobEffectInstance... p_43487_) {
         this(null, p_43487_);
@@ -34,6 +_,10 @@
         this.effects = List.of(p_43485_);
     }
 
+    /**
+     * @deprecated Prefer {@linkplain #requiredFlags(net.neoforged.neoforge.flag.Flag...)}
+     */
+    @Deprecated
     public Potion requiredFeatures(FeatureFlag... p_338520_) {
         this.requiredFeatures = FeatureFlags.REGISTRY.subset(p_338520_);
         return this;
@@ -42,6 +_,16 @@
     @Override
     public FeatureFlagSet requiredFeatures() {
         return this.requiredFeatures;
+    }
+
+    public Potion requiredFlags(net.neoforged.neoforge.flag.Flag... requiredFlags) {
+        java.util.Collections.addAll(this.requiredFlags, requiredFlags);
+        return this;
+    }
+
+    @Override
+    public java.util.Set<net.neoforged.neoforge.flag.Flag> requiredFlags() {
+        return requiredFlagsView;
     }
 
     public static String getName(Optional<Holder<Potion>> p_330503_, String p_43493_) {
