--- a/net/minecraft/world/item/alchemy/PotionBrewing.java
+++ b/net/minecraft/world/item/alchemy/PotionBrewing.java
@@ -198,9 +_,9 @@
    }
 
    public static class Mix<T> {
-      final T from;
+      public final T from;
       public final Ingredient ingredient;
-      final T to;
+      public final T to;
 
       public Mix(T p_43536_, Ingredient p_43537_, T p_43538_) {
          this.from = p_43536_;
