--- a/net/minecraft/world/item/alchemy/PotionContents.java
+++ b/net/minecraft/world/item/alchemy/PotionContents.java
@@ -189,6 +_,10 @@
             lines.accept(CommonComponents.EMPTY);
             lines.accept(Component.translatable("potion.whenDrank").withStyle(ChatFormatting.DARK_PURPLE));
 
+            // Neo: Override handling of potion attribute tooltips to support IAttributeExtension
+            net.neoforged.neoforge.common.util.AttributeUtil.addPotionTooltip(modifiers, lines);
+            if (true) return;
+
             for (Pair<Holder<Attribute>, AttributeModifier> entry : modifiers) {
                 AttributeModifier modifier = entry.getSecond();
                 double amount = modifier.amount();
