--- a/net/minecraft/client/gui/components/debug/DebugEntryLookingAtEntity.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryLookingAtEntity.java
@@ -24,6 +_,9 @@
         if (entity != null) {
             result.add(ChatFormatting.UNDERLINE + "Targeted Entity");
             result.add(entity.typeHolder().getRegisteredName());
+
+            // Neo: Add entity type tags for the targeted entity
+            entity.getType().getTags().map(tag -> "#" + tag.location()).forEach(result::add);
         }
 
         displayer.addToGroup(GROUP, result);
