--- a/net/minecraft/client/gui/components/debug/DebugEntryLookingAtEntity.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryLookingAtEntity.java
@@ -25,6 +_,9 @@
         if (entity != null) {
             list.add(ChatFormatting.UNDERLINE + "Targeted Entity");
             list.add(String.valueOf(BuiltInRegistries.ENTITY_TYPE.getKey(entity.getType())));
+
+            // Neo: Add entity type tags for the targeted entity
+            entity.getType().getTags().map(tag -> "#" + tag.location()).forEach(list::add);
         }
 
         p_434750_.addToGroup(GROUP, list);
