--- a/net/minecraft/server/jsonrpc/IncomingRpcMethod.java
+++ b/net/minecraft/server/jsonrpc/IncomingRpcMethod.java
@@ -119,10 +_,14 @@
             return this.method.create(methodinfo, new IncomingRpcMethod.Attributes(this.runOnMainThread, this.discoverable));
         }
 
+        /* @deprecated Neo: Use the RegisterEvent to register instead of this  */
+        @Deprecated
         public T register(Registry<IncomingRpcMethod> p_451139_, String p_451409_) {
             return this.register(p_451139_, ResourceLocation.withDefaultNamespace(p_451409_));
         }
 
+        /* @deprecated Neo: Use the RegisterEvent to register instead of this  */
+        @Deprecated
         private T register(Registry<IncomingRpcMethod> p_443333_, ResourceLocation p_451500_) {
             return Registry.register(p_443333_, p_451500_, this.build());
         }
@@ -161,9 +_,15 @@
                         jsonelement = jsonarray.get(0);
                     }
 
-                    Params params = this.paramsCodec.parse(JsonOps.INSTANCE, jsonelement).getOrThrow(InvalidParameterJsonRpcException::new);
+                    com.mojang.serialization.DynamicOps<JsonElement> ops;
+                    try {
+                        ops = p_442932_.getServer().registryAccess().createSerializationContext(JsonOps.INSTANCE);
+                    } catch (net.minecraft.server.jsonrpc.methods.MethodNotFoundJsonRpcException e) {
+                        ops = JsonOps.INSTANCE;
+                    }
+                    Params params = this.paramsCodec.parse(ops, jsonelement).getOrThrow(InvalidParameterJsonRpcException::new);
                     Result result = this.function.apply(p_442932_, params, p_442918_);
-                    return this.resultCodec.encodeStart(JsonOps.INSTANCE, result).getOrThrow(EncodeJsonRpcException::new);
+                    return this.resultCodec.encodeStart(ops, result).getOrThrow(EncodeJsonRpcException::new);
                 }
             } else {
                 throw new InvalidParameterJsonRpcException("Expected params as array or named");
