--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -84,7 +_,7 @@
             this.timeInOverworld = 0;
         }
 
-        if (this.timeInOverworld > 300) {
+        if (this.timeInOverworld > 300 && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> this.timeInOverworld = timer)) {
             this.playConvertedSound();
             this.finishConversion(p_376586_);
         }
@@ -103,7 +_,10 @@
         this.convertTo(
             EntityType.ZOMBIFIED_PIGLIN,
             ConversionParams.single(this, true, true),
-            p_393202_ -> p_393202_.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0))
+            p_371722_ -> {
+                p_371722_.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0));
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_371722_);
+            }
         );
     }
 
