--- a/net/minecraft/world/entity/monster/piglin/PiglinAi.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinAi.java
@@ -377,7 +_,7 @@
         ItemStack itemstack = piglin.getItemInHand(InteractionHand.OFF_HAND);
         piglin.setItemInHand(InteractionHand.OFF_HAND, ItemStack.EMPTY);
         if (piglin.isAdult()) {
-            boolean flag = isBarterCurrency(itemstack);
+            boolean flag = itemstack.isPiglinCurrency();
             if (shouldBarter && flag) {
                 throwItems(piglin, getBarterResponseItems(piglin));
             } else if (!flag) {
@@ -461,7 +_,7 @@
             return false;
         } else if (isAdmiringDisabled(piglin) && piglin.getBrain().hasMemoryValue(MemoryModuleType.ATTACK_TARGET)) {
             return false;
-        } else if (isBarterCurrency(stack)) {
+        } else if (stack.isPiglinCurrency()) {
             return isNotHoldingLovedItemInOffHand(piglin);
         } else {
             boolean flag = piglin.canAddToInventory(stack);
@@ -551,7 +_,7 @@
     }
 
     protected static boolean canAdmire(Piglin piglin, ItemStack stack) {
-        return !isAdmiringDisabled(piglin) && !isAdmiringItem(piglin) && piglin.isAdult() && isBarterCurrency(stack);
+        return !isAdmiringDisabled(piglin) && !isAdmiringItem(piglin) && piglin.isAdult() && stack.isPiglinCurrency();
     }
 
     protected static void wasHurtBy(Piglin piglin, LivingEntity target) {
@@ -641,7 +_,7 @@
     public static boolean isWearingGold(LivingEntity livingEntity) {
         for (ItemStack itemstack : livingEntity.getArmorAndBodyArmorSlots()) {
             Item item = itemstack.getItem();
-            if (item instanceof ArmorItem && ((ArmorItem)item).getMaterial().is(ArmorMaterials.GOLD)) {
+            if (itemstack.makesPiglinsNeutral(livingEntity)) {
                 return true;
             }
         }
@@ -792,7 +_,7 @@
     }
 
     private static boolean hasCrossbow(LivingEntity piglin) {
-        return piglin.isHolding(Items.CROSSBOW);
+        return piglin.isHolding(is -> is.getItem() instanceof net.minecraft.world.item.CrossbowItem);
     }
 
     private static void admireGoldItem(LivingEntity piglin) {
