--- a/net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
+++ b/net/minecraft/world/entity/monster/piglin/StopHoldingItemIfNoLongerAdmiring.java
@@ -12,7 +_,7 @@
         return BehaviorBuilder.create(
             p_259197_ -> p_259197_.group(p_259197_.absent(MemoryModuleType.ADMIRING_ITEM))
                     .apply(p_259197_, p_259512_ -> (p_333638_, p_333639_, p_333640_) -> {
-                            if (!p_333639_.getOffhandItem().isEmpty() && !p_333639_.getOffhandItem().is(Items.SHIELD)) {
+                            if (!p_333639_.getOffhandItem().isEmpty() && !p_333639_.getOffhandItem().canPerformAction(net.neoforged.neoforge.common.ToolActions.SHIELD_BLOCK)) {
                                 PiglinAi.stopHoldingOffHandItem(p_333639_, true);
                                 return true;
                             } else {
