--- a/net/minecraft/data/recipes/packs/BundleRecipeProvider.java
+++ b/net/minecraft/data/recipes/packs/BundleRecipeProvider.java
@@ -8,8 +_,8 @@
 import net.minecraft.world.item.Items;
 
 public class BundleRecipeProvider extends RecipeProvider {
-    public BundleRecipeProvider(PackOutput p_248813_) {
-        super(p_248813_);
+    public BundleRecipeProvider(PackOutput p_248813_, java.util.concurrent.CompletableFuture<net.minecraft.core.HolderLookup.Provider> lookupProvider) {
+        super(p_248813_, lookupProvider);
     }
 
     @Override
