--- a/net/minecraft/data/recipes/packs/VanillaRecipeProvider.java
+++ b/net/minecraft/data/recipes/packs/VanillaRecipeProvider.java
@@ -48,8 +48,8 @@
    public static final ImmutableList<ItemLike> REDSTONE_SMELTABLES = ImmutableList.of(Items.REDSTONE_ORE, Items.DEEPSLATE_REDSTONE_ORE);
    public static final ImmutableList<ItemLike> EMERALD_SMELTABLES = ImmutableList.of(Items.EMERALD_ORE, Items.DEEPSLATE_EMERALD_ORE);
 
-   public VanillaRecipeProvider(PackOutput p_250820_) {
-      super(p_250820_);
+   public VanillaRecipeProvider(PackOutput p_250820_, CompletableFuture<net.minecraft.core.HolderLookup.Provider> lookupProvider) {
+      super(p_250820_, lookupProvider);
    }
 
    @Override
