--- a/net/minecraft/data/recipes/packs/VanillaRecipeProvider.java
+++ b/net/minecraft/data/recipes/packs/VanillaRecipeProvider.java
@@ -81,7 +_,7 @@
     }
 
     @Override
-    protected void buildRecipes(RecipeOutput p_301191_) {
+    protected void buildRecipes(RecipeOutput p_301191_, HolderLookup.Provider holderLookup) {
         generateForEnabledBlockFamilies(p_301191_, FeatureFlagSet.of(FeatureFlags.VANILLA));
         planksFromLog(p_301191_, Blocks.ACACIA_PLANKS, ItemTags.ACACIA_LOGS, 4);
         planksFromLogs(p_301191_, Blocks.BIRCH_PLANKS, ItemTags.BIRCH_LOGS, 4);
