--- a/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -137,6 +_,9 @@
         }
 
         this.renderers.add(new ChunkCullingDebugRenderer(minecraft));
+
+        // Neo: Allow registering custom debug renderers
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.RegisterDebugRenderersEvent(factory -> renderers.add(factory.apply(minecraft))));
     }
 
     public void emitGizmos(Frustum frustum, double camX, double camY, double camZ, float partialTicks) {
