--- a/net/minecraft/client/renderer/debug/DebugRenderer.java
+++ b/net/minecraft/client/renderer/debug/DebugRenderer.java
@@ -137,6 +_,8 @@
         }
 
         this.renderers.add(new ChunkCullingDebugRenderer(minecraft));
+        // Neo: Allow registering custom debug renderers
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.RegisterDebugRenderersEvent(factory -> renderers.add(factory.apply(minecraft))));
     }
 
     public void emitGizmos(Frustum p_456065_, double p_455789_, double p_454848_, double p_456243_, float p_455292_) {
