--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -137,8 +_,10 @@
         return jpanel;
     }
 
+    private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
     public void start() {
         this.logAppenderThread.start();
+        latch.countDown();
     }
 
     public void close() {
@@ -152,6 +_,9 @@
     }
 
     public void print(JTextArea textArea, JScrollPane scrollPane, String line) {
+        try {
+            latch.await();
+        } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
         if (!SwingUtilities.isEventDispatchThread()) {
             SwingUtilities.invokeLater(() -> this.print(textArea, scrollPane, line));
         } else {
