--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -39,6 +_,7 @@
     private Thread logAppenderThread;
     private final Collection<Runnable> finalizers = Lists.newArrayList();
     private final AtomicBoolean isClosing = new AtomicBoolean();
+    private final java.util.concurrent.CountDownLatch startupLatch = new java.util.concurrent.CountDownLatch(1);
 
     public static MinecraftServerGui showFrameFor(DedicatedServer server) {
         try {
@@ -138,6 +_,7 @@
 
     public void start() {
         this.logAppenderThread.start();
+        this.startupLatch.countDown();
     }
 
     public void close() {
@@ -151,6 +_,11 @@
     }
 
     public void print(JTextArea console, JScrollPane scrollPane, String line) {
+        try {
+            // Neo: Prevent logging until after constructor has ended.
+            this.startupLatch.await();
+        } catch (InterruptedException ignored) {}
+
         if (!SwingUtilities.isEventDispatchThread()) {
             SwingUtilities.invokeLater(() -> this.print(console, scrollPane, line));
         } else {
