--- a/net/minecraft/util/context/ContextMap.java
+++ b/net/minecraft/util/context/ContextMap.java
@@ -68,7 +_,7 @@
 
         public ContextMap buildAndValidate(ContextKeySet paramSet) {
             Set<ContextKey<?>> notAllowed = Sets.difference(this.params.keySet(), paramSet.allowed());
-            if (!notAllowed.isEmpty()) {
+            if (false && !notAllowed.isEmpty()) {
                 throw new IllegalArgumentException("Parameters not allowed in this parameter set: " + notAllowed);
             } else {
                 Set<ContextKey<?>> missingRequired = Sets.difference(paramSet.required(), this.params.keySet());
