--- a/net/minecraft/server/jsonrpc/internalapi/MinecraftApi.java
+++ b/net/minecraft/server/jsonrpc/internalapi/MinecraftApi.java
@@ -16,6 +_,24 @@
     private final MinecraftServerSettingsService minecraftServerSettingsService;
     private final MinecraftServerStateService minecraftServerStateService;
     private final MinecraftExecutorService executorService;
+    private final @org.jetbrains.annotations.Nullable DedicatedServer server;
+    private final @org.jetbrains.annotations.Nullable JsonRpcLogger logger;
+
+    // Neo: Use the constructor with the server and logger instead
+    @Deprecated
+    public MinecraftApi(
+            NotificationManager p_449655_,
+            MinecraftAllowListService p_443286_,
+            MinecraftBanListService p_443467_,
+            MinecraftPlayerListService p_442735_,
+            MinecraftGameRuleService p_442736_,
+            MinecraftOperatorListService p_443566_,
+            MinecraftServerSettingsService p_443225_,
+            MinecraftServerStateService p_443354_,
+            MinecraftExecutorService p_449378_
+    ) {
+        this(p_449655_, p_443286_, p_443467_, p_442735_, p_442736_, p_443566_, p_443225_, p_443354_, p_449378_, null, null);
+    }
 
     public MinecraftApi(
         NotificationManager p_449655_,
@@ -26,7 +_,9 @@
         MinecraftOperatorListService p_443566_,
         MinecraftServerSettingsService p_443225_,
         MinecraftServerStateService p_443354_,
-        MinecraftExecutorService p_449378_
+        MinecraftExecutorService p_449378_,
+        @org.jetbrains.annotations.Nullable DedicatedServer server,
+        @org.jetbrains.annotations.Nullable JsonRpcLogger logger
     ) {
         this.notificationManager = p_449655_;
         this.allowListService = p_443286_;
@@ -37,6 +_,8 @@
         this.minecraftServerSettingsService = p_443225_;
         this.minecraftServerStateService = p_443354_;
         this.executorService = p_449378_;
+        this.server = server;
+        this.logger = logger;
     }
 
     public <V> CompletableFuture<V> submit(Supplier<V> p_442544_) {
@@ -79,6 +_,14 @@
         return this.notificationManager;
     }
 
+    public @org.jetbrains.annotations.Nullable DedicatedServer getServer() {
+        return this.server;
+    }
+
+    public @org.jetbrains.annotations.Nullable JsonRpcLogger getLogger() {
+        return this.logger;
+    }
+
     public static MinecraftApi of(DedicatedServer p_449765_) {
         JsonRpcLogger jsonrpclogger = new JsonRpcLogger();
         MinecraftAllowListServiceImpl minecraftallowlistserviceimpl = new MinecraftAllowListServiceImpl(p_449765_, jsonrpclogger);
@@ -98,7 +_,9 @@
             minecraftoperatorlistserviceimpl,
             minecraftserversettingsserviceimpl,
             minecraftserverstateserviceimpl,
-            minecraftexecutorservice
+            minecraftexecutorservice,
+            p_449765_,
+            jsonrpclogger
         );
     }
 }
