--- a/net/minecraft/server/jsonrpc/internalapi/MinecraftApi.java
+++ b/net/minecraft/server/jsonrpc/internalapi/MinecraftApi.java
@@ -17,7 +_,10 @@
     private final MinecraftServerSettingsService minecraftServerSettingsService;
     private final MinecraftServerStateService minecraftServerStateService;
     private final MinecraftExecutorService executorService;
+    private final @org.jspecify.annotations.Nullable JsonRpcLogger logger;
 
+    /// @deprecated Neo: Use the constructor with the server and logger instead
+    @Deprecated
     public MinecraftApi(
         NotificationManager notificationManager,
         MinecraftAllowListService allowListService,
@@ -29,6 +_,21 @@
         MinecraftServerStateService minecraftServerStateService,
         MinecraftExecutorService executorService
     ) {
+        this(notificationManager, allowListService, banListService, minecraftPlayerListService, gameRuleService, minecraftOperatorListService, minecraftServerSettingsService, minecraftServerStateService, executorService, null);
+    }
+
+    public MinecraftApi(
+            NotificationManager notificationManager,
+            MinecraftAllowListService allowListService,
+            MinecraftBanListService banListService,
+            MinecraftPlayerListService minecraftPlayerListService,
+            MinecraftGameRuleService gameRuleService,
+            MinecraftOperatorListService minecraftOperatorListService,
+            MinecraftServerSettingsService minecraftServerSettingsService,
+            MinecraftServerStateService minecraftServerStateService,
+            MinecraftExecutorService executorService,
+            @org.jspecify.annotations.Nullable JsonRpcLogger logger
+    ) {
         this.notificationManager = notificationManager;
         this.allowListService = allowListService;
         this.banListService = banListService;
@@ -38,6 +_,7 @@
         this.minecraftServerSettingsService = minecraftServerSettingsService;
         this.minecraftServerStateService = minecraftServerStateService;
         this.executorService = executorService;
+        this.logger = logger;
     }
 
     public <V> CompletableFuture<V> submit(Supplier<V> supplier) {
@@ -80,6 +_,10 @@
         return this.notificationManager;
     }
 
+    public JsonRpcLogger logger() {
+        return java.util.Objects.requireNonNull(this.logger);
+    }
+
     public static MinecraftApi of(NotificationManager notificationManager) {
         JsonRpcLogger jsonrpcLogger = new JsonRpcLogger();
         MinecraftAllowListServiceImpl allowListService = new MinecraftAllowListServiceImpl(notificationManager, jsonrpcLogger);
@@ -99,7 +_,8 @@
             operatorListService,
             serverSettingsService,
             serverStateService,
-            executorService
+            executorService,
+            jsonrpcLogger
         );
     }
 }
