--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2473,10 +2594,13 @@
         boolean isFalling = this.getDeltaMovement().y <= 0.0;
         double oldY = this.getY();
         double baseGravity = this.getEffectiveGravity();
+        if (this.getFluidInteraction().isInFluidMatching(this, (entity, type, _) -> !type.isVanilla() && entity.moveInFluid(type, input, baseGravity))) {
+            return;
+        }
         if (this.isInWater()) {
             this.travelInWater(input, baseGravity, isFalling, oldY);
             this.floatInWaterWhileRidden();
-        } else {
+        } else if (this.isInLava()) { // Neo: only apply lava movement if the player is actually in lava and not another fluid
             this.travelInLava(input, baseGravity, isFalling, oldY);
         }
     }
@@ -2541,7 +2666,7 @@
 
     private void floatInWaterWhileRidden() {
         boolean canEntityFloatInWater = this.is(EntityTypeTags.CAN_FLOAT_WHILE_RIDDEN);
-        if (canEntityFloatInWater && this.isVehicle() && this.getFluidHeight(FluidTags.WATER) > this.getFluidJumpThreshold()) {
+        if (canEntityFloatInWater && this.isVehicle() && this.getFluidInteraction().getMaxFluidHeightMatching(this, (_, type, _) -> type.getIsWaterLike()) > this.getFluidJumpThreshold()) {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.04F, 0.0));
         }
     }
