--- a/com/mojang/blaze3d/systems/CommandEncoder.java
+++ b/com/mojang/blaze3d/systems/CommandEncoder.java
@@ -25,6 +_,8 @@
 
     void clearDepthTexture(GpuTexture p_410276_, double p_410184_);
 
+    void clearStencilTexture(GpuTexture texture, int value);
+
     void writeToBuffer(GpuBuffer p_410195_, ByteBuffer p_410481_, int p_410055_);
 
     GpuBuffer.ReadView readBuffer(GpuBuffer p_410459_);
