--- a/com/mojang/blaze3d/systems/CommandEncoder.java
+++ b/com/mojang/blaze3d/systems/CommandEncoder.java
@@ -249,6 +_,15 @@
         this.backend.clearDepthTexture(depthTexture, clearDepth);
     }
 
+    public void clearStencilTexture(GpuTexture texture, int value) {
+        if (this.isInRenderPass) {
+            throw new IllegalStateException("Close the existing render pass before creating a new one!");
+        }
+
+        this.verifyStencilTexture(texture);
+        this.backend.clearStencilTexture(texture, value);
+    }
+
     public void writeToBuffer(GpuBufferSlice destination, ByteBuffer data) {
         if (this.isInRenderPass) {
             throw new IllegalStateException("Close the existing render pass before performing additional commands");
@@ -696,6 +_,25 @@
                 throw new UnsupportedOperationException("Textures with multiple depths or layers are not yet supported for copying");
             }
 
+            final var sourceFormat = source.getFormat();
+            final var destFormat = destination.getFormat();
+            if (sourceFormat.hasDepthAspect() || sourceFormat.hasStencilAspect()) {
+                if (sourceFormat != destFormat) {
+                    throw new IllegalArgumentException(
+                            "When copying depth or stencil data, the source and destination texture formats must be identical. Source: "
+                                    + sourceFormat + ", Destination: " + destFormat
+                    );
+                }
+            }
+
+            if (sourceFormat.hasColorAspect() != destFormat.hasColorAspect()) {
+                throw new IllegalArgumentException(
+                        "Source and destination texture formats must have consistent color aspects. Source: "
+                                + sourceFormat + ", Destination: " + destFormat
+                );
+            }
+
+
             this.backend.copyTextureToTexture(source, destination, mipLevel, destX, destY, sourceX, sourceY, width, height);
         } else {
             throw new IllegalArgumentException(
@@ -761,6 +_,24 @@
 
         if (depthTexture.getDepthOrLayers() > 1) {
             throw new UnsupportedOperationException("Clearing a texture with multiple layers or depths is not yet supported");
+        }
+    }
+
+    private void verifyStencilTexture(GpuTexture stencilTexture) {
+        if (!stencilTexture.getFormat().hasStencilAspect()) {
+            throw new IllegalStateException("Trying to clear stencil in a texture that has no stencil component!");
+        }
+
+        if (stencilTexture.isClosed()) {
+            throw new IllegalStateException("Stencil texture is closed");
+        }
+
+        if ((stencilTexture.usage() & GpuTexture.USAGE_RENDER_ATTACHMENT) == 0) {
+            throw new IllegalStateException("Stencil texture must have USAGE_RENDER_ATTACHMENT");
+        }
+
+        if ((stencilTexture.usage() & GpuTexture.USAGE_COPY_DST) == 0) {
+            throw new IllegalStateException("Stencil texture must have USAGE_COPY_DST");
         }
     }
 
