--- a/com/mojang/blaze3d/systems/RenderPass.java
+++ b/com/mojang/blaze3d/systems/RenderPass.java
@@ -116,6 +_,16 @@
         this.backend.setUniform(name, value);
     }
 
+    public void setViewport(int x, int y, int width, int height) {
+        if (width < 0 || height < 0) {
+            throw new IllegalArgumentException("Viewport size must be >0, was " + width + "x" + height);
+        }
+        if (x < this.renderArea.x() || y < this.renderArea.y() || x + width > this.renderArea.x() + this.renderArea.width() || y + height > this.renderArea.height()) {
+            throw new IllegalArgumentException("Viewport at " + x + ", " + y + " with size " + width + "x" + height + " is out of bounds for render area " + this.renderArea);
+        }
+        this.backend.setViewport(x, y, width, height);
+    }
+
     public void enableScissor(int x, int y, int width, int height) {
         if (width > 0 && height > 0) {
             if (x >= this.renderArea.x()
