--- a/com/mojang/blaze3d/systems/RenderSystem.java
+++ b/com/mojang/blaze3d/systems/RenderSystem.java
@@ -74,6 +_,7 @@
     private static @Nullable DynamicUniforms dynamicUniforms;
     private static final ScissorState scissorStateForRenderTypeDraws = new ScissorState();
     private static final SamplerCache samplerCache = new SamplerCache();
+    private static final net.neoforged.neoforge.client.pipeline.PipelineModifierStack PIPELINE_MODIFIERS = new net.neoforged.neoforge.client.pipeline.PipelineModifierStack();
 
     public static SamplerCache getSamplerCache() {
         return samplerCache;
@@ -121,6 +_,7 @@
 
         dynamicUniforms.reset();
         Minecraft.getInstance().levelRenderer.endFrame();
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.FlipFrameEvent());
     }
 
     public static void setShaderFog(GpuBufferSlice fog) {
@@ -379,4 +_,38 @@
     @OnlyIn(Dist.CLIENT)
     record GpuAsyncTask(Runnable callback, GpuFence fence) {
     }
+
+    /**
+     * Neo: Push the provided {@link net.neoforged.neoforge.client.pipeline.PipelineModifier PipelineModifier} to be applied to subsequent rendering.
+     * <p>
+     * Must be paired with a corresponding {@link #popPipelineModifier()} call after flushing the used buffers
+     */
+    public static void pushPipelineModifier(net.minecraft.resources.ResourceKey<net.neoforged.neoforge.client.pipeline.PipelineModifier> modifier) {
+        PIPELINE_MODIFIERS.push(modifier);
+    }
+
+    /**
+     * Neo: Pop the last {@link net.neoforged.neoforge.client.pipeline.PipelineModifier PipelineModifier} off the modifier stack
+     */
+    public static void popPipelineModifier() {
+        PIPELINE_MODIFIERS.pop();
+    }
+
+    /**
+     * Neo: Run the provided {@link Runnable} with the provided {@link net.neoforged.neoforge.client.pipeline.PipelineModifier PipelineModifier} applied
+     * to the pipelines used by the rendering done in the {@link Runnable}
+     */
+    public static void renderWithPipelineModifier(net.minecraft.resources.ResourceKey<net.neoforged.neoforge.client.pipeline.PipelineModifier> modifier, Runnable renderTask) {
+        PIPELINE_MODIFIERS.renderWithModifier(modifier, renderTask);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static com.mojang.blaze3d.pipeline.RenderPipeline applyPipelineModifiers(com.mojang.blaze3d.pipeline.RenderPipeline pipeline) {
+        return PIPELINE_MODIFIERS.apply(pipeline);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static void ensurePipelineModifiersEmpty() {
+        PIPELINE_MODIFIERS.ensureEmpty();
+     }
 }
