--- a/com/mojang/blaze3d/systems/RenderPass.java
+++ b/com/mojang/blaze3d/systems/RenderPass.java
@@ -68,6 +_,10 @@
         }
     }
 
+    public void setViewport(int x, int y, int width, int height) {
+        this.backend.setViewport(x, y, width, height);
+    }
+
     public void enableScissor(int x, int y, int width, int height) {
         this.backend.enableScissor(x, y, width, height);
     }
