--- a/com/mojang/blaze3d/systems/RenderSystem.java
+++ b/com/mojang/blaze3d/systems/RenderSystem.java
@@ -83,6 +_,7 @@
     private static final SamplerCache samplerCache = new SamplerCache();
     private static @Nullable PipelineCache fallbackPipelineCache;
     private static @Nullable PipelineCache currentPipelineCache;
+    private static final net.neoforged.neoforge.client.pipeline.PipelineModifierStack PIPELINE_MODIFIERS = new net.neoforged.neoforge.client.pipeline.PipelineModifierStack();
 
     public static SamplerCache getSamplerCache() {
         return samplerCache;
@@ -103,6 +_,8 @@
     }
 
     public static @Nullable CompiledRenderPipeline getCompiledPipelineNullable(RenderPipeline pipeline) {
+        pipeline = PIPELINE_MODIFIERS.apply(pipeline);
+
         if (currentPipelineCache != null) {
             CompiledRenderPipeline cachedPipeline = currentPipelineCache.get(pipeline);
             if (cachedPipeline != null) {
@@ -491,5 +_,28 @@
     }
 
     private record GpuAsyncTask(Runnable callback, GpuFence fence) {
+    }
+
+    /// Neo: Push the provided [net.neoforged.neoforge.client.pipeline.PipelineModifier] to be applied to subsequent rendering.
+    ///
+    /// Must be paired with a corresponding [#popPipelineModifier()] call after flushing the used buffers
+    public static void pushPipelineModifier(net.minecraft.resources.ResourceKey<net.neoforged.neoforge.client.pipeline.PipelineModifier> modifier) {
+        PIPELINE_MODIFIERS.push(modifier);
+    }
+
+    /// Neo: Pop the last [net.neoforged.neoforge.client.pipeline.PipelineModifier] off the modifier stack
+    public static void popPipelineModifier() {
+        PIPELINE_MODIFIERS.pop();
+    }
+
+    /// Neo: Run the provided [Runnable] with the provided [net.neoforged.neoforge.client.pipeline.PipelineModifier] applied
+    /// to the pipelines used by the rendering done in the [Runnable]
+    public static void renderWithPipelineModifier(net.minecraft.resources.ResourceKey<net.neoforged.neoforge.client.pipeline.PipelineModifier> modifier, Runnable renderTask) {
+        PIPELINE_MODIFIERS.renderWithModifier(modifier, renderTask);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static void ensurePipelineModifiersEmpty() {
+        PIPELINE_MODIFIERS.ensureEmpty();
     }
 }
