--- a/net/minecraft/world/level/chunk/ImposterProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ImposterProtoChunk.java
@@ -219,6 +_,10 @@
     }
 
     @Override
+    public void findBlocks(java.util.function.BiPredicate<BlockState, BlockPos> p_285343_, BiConsumer<BlockPos, BlockState> p_285030_) {
+        this.wrapped.findBlocks(p_285343_, p_285030_);
+    }
+
     public TickContainerAccess<Block> getBlockTicks() {
         return this.allowWrites ? this.wrapped.getBlockTicks() : BlackholeTickAccess.emptyContainer();
     }
@@ -291,5 +_,10 @@
     @Override
     public ChunkSkyLightSources getSkyLightSources() {
         return this.wrapped.getSkyLightSources();
+    }
+
+    @Override
+    public net.neoforged.neoforge.attachment.AttachmentHolder.AsField getAttachmentHolder() {
+        return wrapped.getAttachmentHolder();
     }
 }
