--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -65,7 +_,7 @@
 
         FluidState fluidstate = blockstate.getFluidState();
         FluidState fluidstate1 = p_62995_.getFluidState();
-        if (!blockstate.isAir()) {
+        if (!blockstate.isEmpty()) { // NEO: Patched to make modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections
             --this.nonEmptyBlockCount;
             if (blockstate.isRandomlyTicking()) {
                 --this.tickingBlockCount;
@@ -76,7 +_,7 @@
             --this.tickingFluidCount;
         }
 
-        if (!p_62995_.isAir()) {
+        if (!p_62995_.isEmpty()) { // NEO: Patched to make modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections
             ++this.nonEmptyBlockCount;
             if (p_62995_.isRandomlyTicking()) {
                 ++this.tickingBlockCount;
