--- a/net/minecraft/client/sounds/MusicManager.java
+++ b/net/minecraft/client/sounds/MusicManager.java
@@ -24,7 +_,10 @@
     }
 
     public void tick() {
-        Music music = this.minecraft.getSituationalMusic();
+        Music music = net.neoforged.neoforge.client.ClientHooks.selectMusic(this.minecraft.getSituationalMusic());
+        if (music == null) {
+            return;
+        }
         if (this.currentMusic != null) {
             if (!music.getEvent().value().getLocation().equals(this.currentMusic.getLocation()) && music.replaceCurrentMusic()) {
                 this.minecraft.getSoundManager().stop(this.currentMusic);
