--- a/net/minecraft/client/sounds/MusicManager.java
+++ b/net/minecraft/client/sounds/MusicManager.java
@@ -25,7 +_,15 @@
     }
 
     public void tick() {
-        MusicInfo musicinfo = this.minecraft.getSituationalMusic();
+        MusicInfo musicinfo = net.neoforged.neoforge.client.ClientHooks.selectMusic(this.minecraft.getSituationalMusic(), this.currentMusic);
+        if (musicinfo == null) {
+            if (this.currentMusic != null) {
+                this.stopPlaying();
+            }
+            this.nextSongDelay = 0;
+            return;
+        }
+
         float f = musicinfo.volume();
         if (this.currentMusic != null && this.currentGain != f) {
             boolean flag = this.fadePlaying(f);
