--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -83,6 +_,7 @@
         this.options = options;
         this.soundBuffers = new SoundBufferLibrary(resourceProvider);
         this.lastSeenDevices = this.deviceTracker.currentDevices();
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     public void reload() {
@@ -100,6 +_,7 @@
 
         this.destroy();
         this.loadLibrary();
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     private synchronized void loadLibrary() {
@@ -368,7 +_,8 @@
             return SoundEngine.PlayResult.NOT_STARTED;
         }
 
-        if (!instance.canPlaySound()) {
+        instance = net.neoforged.neoforge.client.ClientHooks.playSound(this, instance);
+        if (instance == null || !instance.canPlaySound()) {
             return SoundEngine.PlayResult.NOT_STARTED;
         }
 
@@ -434,10 +_,11 @@
                 this.instanceToChannel.put(instance, handle);
                 this.instanceBySource.put(soundSource, instance);
                 float pitch = this.calculatePitch(instance);
+                SoundInstance finalInstance = instance;
                 handle.execute(channel -> {
                     channel.setPitch(pitch);
                     channel.setVolume(volume);
-                    if (instance.getAttenuation() == SoundInstance.Attenuation.LINEAR) {
+                    if (finalInstance.getAttenuation() == SoundInstance.Attenuation.LINEAR) {
                         channel.linearAttenuation(attenuationDistance);
                     } else {
                         channel.disableAttenuation();
@@ -445,17 +_,20 @@
 
                     channel.setLooping(isLooping && !isStreaming);
                     channel.setSelfPosition(position);
-                    channel.setRelative(instance.isRelative());
+                    channel.setRelative(finalInstance.isRelative());
                 });
+                final SoundInstance soundinstance = instance;
                 if (!isStreaming) {
                     this.soundBuffers.getCompleteBuffer(sound.getPath()).thenAccept(soundBuffer -> handle.execute(channel -> {
                         channel.attachStaticBuffer(soundBuffer);
                         channel.play();
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.sound.PlaySoundSourceEvent(this, soundinstance, channel));
                     }));
                 } else {
-                    this.soundBuffers.getStream(sound.getPath(), isLooping).thenAccept(stream -> handle.execute(channel -> {
+                    soundinstance.getStream(this.soundBuffers, sound, isLooping).thenAccept(stream -> handle.execute(channel -> {
                         channel.attachBufferStream(stream);
                         channel.play();
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.sound.PlayStreamingSourceEvent(this, soundinstance, channel));
                     }));
                 }
 
