--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -83,6 +_,7 @@
         this.options = options;
         this.soundBuffers = new SoundBufferLibrary(resourceProvider);
         this.lastSeenDevices = this.deviceTracker.currentDevices();
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     public void reload() {
@@ -100,6 +_,7 @@
 
         this.destroy();
         this.loadLibrary();
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.sound.SoundEngineLoadEvent(this));
     }
 
     private synchronized void loadLibrary() {
@@ -368,6 +_,15 @@
             return SoundEngine.PlayResult.NOT_STARTED;
         }
 
+        instance = net.neoforged.neoforge.client.ClientHooks.playSound(this, instance);
+        if (instance == null) {
+            return PlayResult.NOT_STARTED;
+        }
+        return playWithoutEvent(instance);
+    }
+
+    /// Neo: split off of [#play(SoundInstance)] to avoid issues with lambdas capturing the re-assigned sound instance
+    private SoundEngine.PlayResult playWithoutEvent(SoundInstance instance) {
         if (!instance.canPlaySound()) {
             return SoundEngine.PlayResult.NOT_STARTED;
         }
@@ -451,11 +_,13 @@
                     this.soundBuffers.getCompleteBuffer(sound.getPath()).thenAccept(soundBuffer -> handle.execute(channel -> {
                         channel.attachStaticBuffer(soundBuffer);
                         channel.play();
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.sound.PlaySoundSourceEvent(this, instance, channel));
                     }));
                 } else {
-                    this.soundBuffers.getStream(sound.getPath(), isLooping).thenAccept(stream -> handle.execute(channel -> {
+                    instance.getStream(this.soundBuffers, sound, isLooping).thenAccept(stream -> handle.execute(channel -> {
                         channel.attachBufferStream(stream);
                         channel.play();
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.sound.PlayStreamingSourceEvent(this, instance, channel));
                     }));
                 }
 
