--- a/net/minecraft/client/renderer/blockentity/ChestRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/ChestRenderer.java
@@ -88,7 +90,7 @@
         float open = state.open;
         open = 1.0F - open;
         open = 1.0F - open * open * open;
-        SpriteId spriteId = Sheets.chooseSprite(state.material, state.type);
+        SpriteId spriteId = state.customSprite != null ? state.customSprite : Sheets.chooseSprite(state.material, state.type);
         ChestModel model = this.models.select(state.type);
         submitNodeCollector.submitModel(
             model, open, poseStack, state.lightCoords, OverlayTexture.NO_OVERLAY, -1, spriteId, this.sprites, 0, state.breakProgress
