--- a/net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
+++ b/net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
@@ -81,7 +81,7 @@
         for (int slot = 0; slot < result.size(); slot++) {
             ItemStack itemStack = input.getItem(slot);
             if (!itemStack.isEmpty()) {
-                ItemStack remainder = itemStack.getItem().getCraftingRemainder();
+                ItemStack remainder = itemStack.getCraftingRemainder();
                 if (!remainder.isEmpty()) {
                     result.set(slot, remainder);
                 } else if (!itemStack.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY).layers().isEmpty()) {
