--- a/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -87,7 +87,7 @@
 
         for (int slot = 0; slot < result.size(); slot++) {
             ItemStack itemStack = input.getItem(slot);
-            ItemStack remainder = itemStack.getItem().getCraftingRemainder();
+            ItemStack remainder = itemStack.getCraftingRemainder();
             if (!remainder.isEmpty()) {
                 result.set(slot, remainder);
             } else if (itemStack.has(DataComponents.WRITTEN_BOOK_CONTENT)) {
