--- a/net/minecraft/core/component/BlockTransformer.java
+++ b/net/minecraft/core/component/BlockTransformer.java
@@ -53,6 +_,10 @@
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<BlockTransformer>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.BLOCK_TRANSFORMER);
 
     public InteractionResult transformBlock(UseOnContext context) {
+        return transformBlock(context, context.getItemInHand().get(DataComponents.BLOCK_TRANSFORMER));
+    }
+
+    public InteractionResult transformBlock(UseOnContext context, @Nullable Holder<BlockTransformer> transformer) {
         if (playerHasBlockingItemUseIntent(context)) {
             return InteractionResult.PASS;
         }
@@ -60,7 +_,7 @@
         BlockPos pos = context.getClickedPos();
         Level level = context.getLevel();
 
-        for (BlockTransformer.BlockTransformData transformData : this.transforms) {
+        for (BlockTransformer.BlockTransformData transformData : net.neoforged.neoforge.common.DataMapHooks.appendDatamapTransformers(transformer, this.transforms)) {
             Direction clickedFace = context.getClickedFace();
             if (!transformData.disallowedFaces().contains(clickedFace)) {
                 BlockState newBlockState = transformData.blockStateProvider.value().getOptionalState(level, level.getRandom(), pos);
