--- a/net/minecraft/core/component/DataComponentHolder.java
+++ b/net/minecraft/core/component/DataComponentHolder.java
@@ -3,7 +_,7 @@
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 
-public interface DataComponentHolder {
+public interface DataComponentHolder extends net.neoforged.neoforge.common.extensions.IDataComponentHolderExtension {
     DataComponentMap getComponents();
 
     @Nullable
