--- a/net/minecraft/core/component/DataComponentPatch.java
+++ b/net/minecraft/core/component/DataComponentPatch.java
@@ -239,6 +_,7 @@
         }
 
         public <T> DataComponentPatch.Builder set(DataComponentType<T> p_332004_, T p_331566_) {
+            net.neoforged.neoforge.common.CommonHooks.validateComponent(p_331566_);
             this.map.put(p_332004_, Optional.of(p_331566_));
             return this;
         }
@@ -250,6 +_,12 @@
 
         public <T> DataComponentPatch.Builder set(TypedDataComponent<T> p_331321_) {
             return this.set(p_331321_.type(), p_331321_.value());
+        }
+
+        // Neo allow for easier merging of patches
+        public DataComponentPatch.Builder putAll(DataComponentPatch patch) {
+            map.putAll(patch.map);
+            return this;
         }
 
         public DataComponentPatch build() {
