--- a/net/minecraft/core/component/DataComponentPatch.java
+++ b/net/minecraft/core/component/DataComponentPatch.java
@@ -156,6 +_,14 @@
         return new DataComponentPatch.Builder();
     }
 
+    /**
+     * {@return the patch stored for the given component type, null if no patch is stored}
+     */
+    @SuppressWarnings("unchecked")
+    public <T> @Nullable Optional<T> getPatch(DataComponentType<? extends T> type) {
+        return (Optional<T>) this.map.get(type);
+    }
+
     public <T> @Nullable T get(DataComponentGetter prototype, DataComponentType<? extends T> type) {
         return getFromPatchAndPrototype(this.map, prototype, type);
     }
@@ -254,7 +_,44 @@
         private Builder() {
         }
 
+        // CraftBukkit start
+        public void copy(DataComponentPatch orig) {
+            this.map.putAll(orig.map);
+        }
+
+        public void clear(DataComponentType<?> type) {
+            this.map.remove(type);
+        }
+
+        public boolean isSet(DataComponentType<?> type) {
+            return map.containsKey(type);
+        }
+
+        public boolean isEmpty() {
+            return this.map.isEmpty();
+        }
+
+        @Override
+        public boolean equals(Object object) {
+            if (this == object) {
+                return true;
+            }
+
+            if (object instanceof DataComponentPatch.Builder patch) {
+                return this.map.equals(patch.map);
+            }
+
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return this.map.hashCode();
+        }
+        // CraftBukkit end
+
         public <T> DataComponentPatch.Builder set(DataComponentType<T> type, T value) {
+            net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
             this.map.put(type, Optional.of(value));
             return this;
         }
