--- a/net/minecraft/core/component/DataComponentMap.java
+++ b/net/minecraft/core/component/DataComponentMap.java
@@ -126,7 +_,7 @@
         };
     }
 
-    public static class Builder {
+    public static class Builder implements net.neoforged.neoforge.common.extensions.IDataComponentMapBuilderExtensions {
         private final Reference2ObjectMap<DataComponentType<?>, Object> map = new Reference2ObjectArrayMap<>();
 
         Builder() {
