--- a/net/minecraft/core/component/DataComponentInitializers.java
+++ b/net/minecraft/core/component/DataComponentInitializers.java
@@ -68,6 +_,21 @@
         Set<Holder.Reference<T>> elementsWithComponents = Sets.newIdentityHashSet();
         elementBuilders.builders.forEach((elementKey, elementBuilder) -> {
             Holder.Reference<T> element = registry.getOrThrow((ResourceKey<T>)elementKey);
+
+            // Neo: Support programmatic customization of the default components for items
+            if (elementKey.isFor(net.minecraft.core.registries.Registries.ITEM)) {
+                var item = (net.minecraft.world.item.Item) element.value();
+                var modifier = net.neoforged.neoforge.internal.RegistrationEvents.componentModifiersByItem.get(item);
+                if (modifier != null) {
+                    modifier.accept(elementBuilder);
+                }
+                for (var pair : net.neoforged.neoforge.internal.RegistrationEvents.componentModifiersByPredicate) {
+                    if (pair.getFirst().test(item)) {
+                        pair.getSecond().accept(elementBuilder);
+                    }
+                }
+            }
+
             DataComponentMap components = elementBuilder.build();
             entries.add(new DataComponentInitializers.BakedEntry<>(element, components));
             elementsWithComponents.add(element);
@@ -115,6 +_,7 @@
         }
 
         default <C> DataComponentInitializers.Initializer<T> add(DataComponentType<C> type, C value) {
+            net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
             return this.andThen((components, context, key) -> components.set(type, value));
         }
     }
@@ -141,7 +_,11 @@
         C create(HolderLookup.Provider context);
 
         default <T> DataComponentInitializers.Initializer<T> asInitializer(DataComponentType<C> type) {
-            return (components, context, key) -> components.set(type, this.create(context));
+            return (components, context, key) -> {
+                var value = this.create(context);
+                net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
+                components.set(type, value);
+            };
         }
     }
 }
