--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -65,6 +_,7 @@
     }
 
     public <T> @Nullable T set(DataComponentType<T> type, @Nullable T value) {
+        net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
         this.ensureMapOwnership();
         T defaultValue = this.prototype.get(type);
         Optional<T> lastValue;
@@ -203,6 +_,10 @@
         return size;
     }
 
+    public boolean isPatchEmpty() {
+        return this.patch.isEmpty();
+    }
+
     public DataComponentPatch asPatch() {
         if (this.patch.isEmpty()) {
             return DataComponentPatch.EMPTY;
@@ -221,6 +_,13 @@
         return (DataComponentMap)(this.patch.isEmpty() ? this.prototype : this.copy());
     }
 
+    /**
+     * {@return true if the contained patch equals the given patch.}
+     */
+    public boolean patchEquals(DataComponentPatch patch) {
+        return this.patch.equals(patch.map);
+    }
+
     @Override
     public boolean equals(Object obj) {
         return this == obj
@@ -230,7 +_,24 @@
 
     @Override
     public int hashCode() {
-        return this.prototype.hashCode() + this.patch.hashCode() * 31;
+        return this.prototype.hashCode() + hashPatch(this.patch) * 31;
+    }
+
+    // Neo: Change implementation of hashCode to reduce collisions.
+    // For a map, hashCode is specified as the sum of the hash codes of its entries.
+    // We do that, but change the entry hash code to 8191^<key hash> * <value hash>,
+    // where <key hash> is the lower bits of the identity hash code of the key.
+    private static int hashPatch(Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch) {
+        int h = 0, n = patch.size();
+        var iterator = it.unimi.dsi.fastutil.objects.Reference2ObjectMaps.fastIterator(patch);
+        while (n-- != 0) {
+            var entry = iterator.next();
+            int exponent = System.identityHashCode(entry.getKey()) & 0xff;
+            // Use 8191 instead of the usual 31, as 31 can produce many collisions with typical integer component ranges (0-255) if the exponent difference is only 1
+            int entryHash = com.google.common.math.IntMath.pow(8191, exponent) * entry.getValue().hashCode();
+            h += entryHash;
+        }
+        return h;
     }
 
     @Override
