--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -64,6 +_,7 @@
 
     @Nullable
     public <T> T set(DataComponentType<? super T> component, @Nullable T value) {
+        net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
         this.ensureMapOwnership();
         T t = this.prototype.get((DataComponentType<? extends T>)component);
         Optional<T> optional;
@@ -192,6 +_,10 @@
         }
 
         return i;
+    }
+
+    public boolean isPatchEmpty() {
+        return this.patch.isEmpty();
     }
 
     public DataComponentPatch asPatch() {
