--- a/net/minecraft/client/gui/screens/ConnectScreen.java
+++ b/net/minecraft/client/gui/screens/ConnectScreen.java
@@ -104,6 +_,8 @@
                     }
 
                     if (resolvedAddress.isEmpty()) {
+                        ConnectScreen.LOGGER.error("Couldn't connect to server: Unknown host \"{}\"", hostAndPort.getHost());
+                        net.neoforged.neoforge.network.DualStackUtils.logInitialPreferences();
                         minecraft.execute(
                             () -> minecraft.gui
                                 .setScreen(
