--- a/net/minecraft/client/gui/screens/MenuScreens.java
+++ b/net/minecraft/client/gui/screens/MenuScreens.java
@@ -41,6 +_,10 @@
     private static final Map<MenuType<?>, MenuScreens.ScreenConstructor<?, ?>> SCREENS = Maps.newHashMap();
 
     public static <T extends AbstractContainerMenu> void create(@Nullable MenuType<T> p_96202_, Minecraft p_96203_, int p_96204_, Component p_96205_) {
+        getScreenFactory(p_96202_, p_96203_, p_96204_, p_96205_).ifPresent(f -> f.fromPacket(p_96205_, p_96202_, p_96203_, p_96204_));
+    }
+
+    public static <T extends AbstractContainerMenu> java.util.Optional<ScreenConstructor<T, ?>> getScreenFactory(@Nullable MenuType<T> p_96202_, Minecraft p_96203_, int p_96204_, Component p_96205_) {
         if (p_96202_ == null) {
             LOGGER.warn("Trying to open invalid screen with name: {}", p_96205_.getString());
         } else {
@@ -48,9 +_,10 @@
             if (screenconstructor == null) {
                 LOGGER.warn("Failed to create screen for menu type: {}", BuiltInRegistries.MENU.getKey(p_96202_));
             } else {
-                screenconstructor.fromPacket(p_96205_, p_96202_, p_96203_, p_96204_);
+                return java.util.Optional.of(screenconstructor);
             }
         }
+        return java.util.Optional.empty();
     }
 
     @Nullable
@@ -58,6 +_,10 @@
         return (MenuScreens.ScreenConstructor<T, ?>)SCREENS.get(p_96200_);
     }
 
+    /**
+     * @deprecated Use {@link net.neoforged.neoforge.client.event.RegisterMenuScreensEvent} to register menu screens
+     */
+    @Deprecated
     public static <M extends AbstractContainerMenu, U extends Screen & MenuAccess<M>> void register(
         MenuType<? extends M> p_96207_, MenuScreens.ScreenConstructor<M, U> p_96208_
     ) {
@@ -65,6 +_,14 @@
         if (screenconstructor != null) {
             throw new IllegalStateException("Duplicate registration for " + BuiltInRegistries.MENU.getKey(p_96207_));
         }
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static void init() {
+        var map = new java.util.HashMap<MenuType<?>, ScreenConstructor<?, ?>>();
+        var event = new net.neoforged.neoforge.client.event.RegisterMenuScreensEvent(map);
+        net.neoforged.fml.ModLoader.get().postEvent(event);
+        SCREENS.putAll(map);
     }
 
     public static boolean selfTest() {
