--- a/net/minecraft/client/gui/screens/OptionsScreen.java
+++ b/net/minecraft/client/gui/screens/OptionsScreen.java
@@ -97,7 +_,10 @@
 
     @Override
     public void onClose() {
-        this.minecraft.setScreen(this.lastScreen);
+        // We need to consider 2 potential parent screens here:
+        // 1. From the main menu, in which case display the main menu
+        // 2. From the pause menu, in which case exit back to game
+        this.minecraft.setScreen(this.lastScreen instanceof PauseScreen ? null : this.lastScreen);
     }
 
     private void applyPacks(PackRepository p_275714_) {
