--- a/net/minecraft/client/gui/screens/PauseScreen.java
+++ b/net/minecraft/client/gui/screens/PauseScreen.java
@@ -166,6 +_,7 @@
             );
         }
 
+        helper.addChild(Button.builder(Component.translatable("fml.menu.mods"), _ -> this.minecraft.gui.setScreen(new net.neoforged.neoforge.client.gui.ModListScreen(this))).width(BUTTON_WIDTH_FULL).build(), 2);
         this.disconnectButton = helper.addChild(
             Button.builder(
                     CommonComponents.disconnectButtonLabel(this.minecraft.isLocalServer()),
