--- a/net/minecraft/client/gui/screens/Screen.java
+++ b/net/minecraft/client/gui/screens/Screen.java
@@ -116,6 +_,7 @@
         }
     }
 
+    /** Neo: mixins targeting this method won't fire for container screens as {@link net.minecraft.client.gui.screens.inventory.AbstractContainerScreen#render} replicates this method in place of a super call to insert an event */
     @Override
     public void render(GuiGraphics guiGraphics, int mouseX, int mouseY, float partialTick) {
         this.renderBackground(guiGraphics, mouseX, mouseY, partialTick);
@@ -201,7 +_,7 @@
     }
 
     public void onClose() {
-        this.minecraft.setScreen(null);
+        this.minecraft.popGuiLayer();
     }
 
     protected <T extends GuiEventListener & Renderable & NarratableEntry> T addRenderableWidget(T widget) {
@@ -312,8 +_,11 @@
         this.width = width;
         this.height = height;
         if (!this.initialized) {
+            if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Pre(this, this.children, this::addEventWidget, this::removeWidget)).isCanceled()) {
             this.init();
             this.setInitialFocus();
+            }
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Post(this, this.children, this::addEventWidget, this::removeWidget));
         } else {
             this.repositionElements();
         }
@@ -326,8 +_,11 @@
     protected void rebuildWidgets() {
         this.clearWidgets();
         this.clearFocus();
+        if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Pre(this, this.children, this::addEventWidget, this::removeWidget)).isCanceled()) {
         this.init();
         this.setInitialFocus();
+        }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Post(this, this.children, this::addEventWidget, this::removeWidget));
     }
 
     @Override
@@ -354,6 +_,7 @@
 
         this.renderBlurredBackground(partialTick);
         this.renderMenuBackground(guiGraphics);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.BackgroundRendered(this, guiGraphics));
     }
 
     protected void renderBlurredBackground(float partialTick) {
@@ -467,6 +_,10 @@
     public void onFilesDrop(List<Path> packs) {
     }
 
+    public Minecraft getMinecraft() {
+        return this.minecraft;
+    }
+
     private void scheduleNarration(long delay, boolean stopSuppression) {
         this.nextNarrationTime = Util.getMillis() + delay;
         if (stopSuppression) {
@@ -643,5 +_,13 @@
             this.index = index;
             this.priority = priority;
         }
+    }
+
+    private void addEventWidget(GuiEventListener b) {
+        if (b instanceof Renderable r)
+            this.renderables.add(r);
+        if (b instanceof NarratableEntry ne)
+            this.narratables.add(ne);
+        children.add(b);
     }
 }
