--- a/net/minecraft/client/gui/screens/Screen.java
+++ b/net/minecraft/client/gui/screens/Screen.java
@@ -117,6 +_,7 @@
         }
     }
 
+    /** Neo: mixins targeting this method won't fire for container screens as {@link net.minecraft.client.gui.screens.inventory.AbstractContainerScreen#render} replicates this method in place of a super call to insert an event */
     @Override
     public void render(GuiGraphics p_281549_, int p_281550_, int p_282878_, float p_282465_) {
         this.renderBackground(p_281549_, p_281550_, p_282878_, p_282465_);
@@ -202,7 +_,7 @@
     }
 
     public void onClose() {
-        this.minecraft.setScreen(null);
+        this.minecraft.popGuiLayer();
     }
 
     protected <T extends GuiEventListener & Renderable & NarratableEntry> T addRenderableWidget(T p_169406_) {
@@ -314,8 +_,11 @@
         this.width = p_96608_;
         this.height = p_96609_;
         if (!this.initialized) {
+            if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Pre(this, this.children, this::addEventWidget, this::removeWidget)).isCanceled()) {
             this.init();
             this.setInitialFocus();
+            }
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Post(this, this.children, this::addEventWidget, this::removeWidget));
         } else {
             this.repositionElements();
         }
@@ -328,8 +_,11 @@
     protected void rebuildWidgets() {
         this.clearWidgets();
         this.clearFocus();
+        if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Pre(this, this.children, this::addEventWidget, this::removeWidget)).isCanceled()) {
         this.init();
         this.setInitialFocus();
+        }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Post(this, this.children, this::addEventWidget, this::removeWidget));
     }
 
     @Override
@@ -356,6 +_,7 @@
 
         this.renderBlurredBackground(p_294317_);
         this.renderMenuBackground(p_283688_);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.BackgroundRendered(this, p_283688_));
     }
 
     protected void renderBlurredBackground(float p_330683_) {
@@ -484,6 +_,10 @@
     public void onFilesDrop(List<Path> p_96591_) {
     }
 
+    public Minecraft getMinecraft() {
+        return this.minecraft;
+    }
+
     private void scheduleNarration(long p_169381_, boolean p_169382_) {
         this.nextNarrationTime = Util.getMillis() + p_169381_;
         if (p_169382_) {
@@ -650,5 +_,13 @@
             this.index = p_169425_;
             this.priority = p_169426_;
         }
+    }
+
+    private void addEventWidget(GuiEventListener b) {
+        if (b instanceof Renderable r)
+            this.renderables.add(r);
+        if (b instanceof NarratableEntry ne)
+            this.narratables.add(ne);
+        children.add(b);
     }
 }
