--- a/net/minecraft/client/gui/screens/Screen.java
+++ b/net/minecraft/client/gui/screens/Screen.java
@@ -104,6 +_,11 @@
     public final void renderWithTooltipAndSubtitles(GuiGraphics p_434220_, int p_435721_, int p_435145_, float p_435742_) {
         p_434220_.nextStratum();
         this.renderBackground(p_434220_, p_435721_, p_435145_, p_435742_);
+        // Neo: We cannot post this event in the renderBackground method, since many screens override it and don't call super.
+        if (this instanceof net.minecraft.client.gui.screens.inventory.AbstractContainerScreen<?> containerScreen) {
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ContainerScreenEvent.Render.Background(containerScreen, p_434220_, p_435721_, p_435145_));
+        }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Render.Background(this, p_434220_, p_435721_, p_435145_, p_435742_));
         p_434220_.nextStratum();
         this.render(p_434220_, p_435721_, p_435145_, p_435742_);
         p_434220_.renderDeferredElements();
@@ -191,7 +_,7 @@
     }
 
     public void onClose() {
-        this.minecraft.setScreen(null);
+        this.minecraft.popGuiLayer();
     }
 
     protected <T extends GuiEventListener & Renderable & NarratableEntry> T addRenderableWidget(T p_169406_) {
@@ -236,7 +_,7 @@
         return p_282833_.getTooltipLines(
             Item.TooltipContext.of(p_281881_.level),
             p_281881_.player,
-            p_281881_.options.advancedItemTooltips ? TooltipFlag.Default.ADVANCED : TooltipFlag.Default.NORMAL
+            net.neoforged.neoforge.client.ClientTooltipFlag.of(p_281881_.options.advancedItemTooltips ? TooltipFlag.Default.ADVANCED : TooltipFlag.Default.NORMAL)
         );
     }
 
@@ -343,8 +_,11 @@
         this.width = p_96608_;
         this.height = p_96609_;
         if (!this.initialized) {
+            if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Pre(this, this.children, this::addEventWidget, this::removeWidget)).isCanceled()) {
             this.init();
             this.setInitialFocus();
+            }
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Post(this, this.children, this::addEventWidget, this::removeWidget));
         } else {
             this.repositionElements();
         }
@@ -361,8 +_,11 @@
     protected void rebuildWidgets() {
         this.clearWidgets();
         this.clearFocus();
+        if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Pre(this, this.children, this::addEventWidget, this::removeWidget)).isCanceled()) {
         this.init();
         this.setInitialFocus();
+        }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Init.Post(this, this.children, this::addEventWidget, this::removeWidget));
     }
 
     protected void fadeWidgets(float p_421625_) {
@@ -488,6 +_,10 @@
     public void onFilesDrop(List<Path> p_96591_) {
     }
 
+    public Minecraft getMinecraft() {
+        return this.minecraft;
+    }
+
     private void scheduleNarration(long p_169381_, boolean p_169382_) {
         this.nextNarrationTime = Util.getMillis() + p_169381_;
         if (p_169382_) {
@@ -642,6 +_,14 @@
     @Nullable
     public Music getBackgroundMusic() {
         return null;
+    }
+
+    private void addEventWidget(GuiEventListener b) {
+        if (b instanceof Renderable r)
+            this.renderables.add(r);
+        if (b instanceof NarratableEntry ne)
+            this.narratables.add(ne);
+        children.add(b);
     }
 
     @OnlyIn(Dist.CLIENT)
