--- a/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/net/minecraft/client/gui/screens/TitleScreen.java
@@ -113,15 +_,16 @@
         int copyrightWidth = this.font.width(COPYRIGHT_TEXT);
         int copyrightX = this.width - copyrightWidth - 2;
         int spacing = 24;
-        int topPos = this.height / 4 + 48;
+        int topPos = this.height / 4 + 32;
+        int numberOfButtons = 4;
+        int currentButton = 0;
         if (this.minecraft.isDemo()) {
             topPos = this.createDemoMenuOptions(topPos, 24);
         } else {
             topPos = this.createNormalMenuOptions(topPos, 24);
+            this.addRenderableWidget(net.neoforged.neoforge.client.gui.widget.ModsButton.create(this)).setPosition(getHorizontalPosition(++currentButton, numberOfButtons, Button.DEFAULT_HEIGHT), topPos + spacing);
         }
 
-        int numberOfButtons = 3;
-        int currentButton = 0;
         topPos += 24;
         this.friends = this.addRenderableWidget(
             CommonButtons.friends(
@@ -130,17 +_,17 @@
                 !this.minecraft.isDemo()
             )
         );
-        this.friends.setPosition(this.getHorizontalPosition(++currentButton, 3, 20), topPos);
+        this.friends.setPosition(this.getHorizontalPosition(++currentButton, numberOfButtons, 20), topPos);
         SpriteIconButton language = this.addRenderableWidget(
             CommonButtons.language(
                 20, var1x -> this.minecraft.gui.setScreen(new LanguageSelectScreen(this, this.minecraft.options, this.minecraft.getLanguageManager())), true
             )
         );
-        language.setPosition(this.getHorizontalPosition(++currentButton, 3, 20), topPos);
+        language.setPosition(this.getHorizontalPosition(++currentButton, numberOfButtons, 20), topPos);
         SpriteIconButton accessibility = this.addRenderableWidget(
             CommonButtons.accessibility(20, var1x -> this.minecraft.gui.setScreen(new AccessibilityOptionsScreen(this, this.minecraft.options)), true)
         );
-        accessibility.setPosition(this.getHorizontalPosition(++currentButton, 3, 20), topPos);
+        accessibility.setPosition(this.getHorizontalPosition(++currentButton, numberOfButtons, 20), topPos);
         Button.Builder var10001 = Button.builder(
             Component.translatable("menu.options"), var1x -> this.minecraft.gui.setScreen(new OptionsScreen(this, this.minecraft.options, false))
         );
@@ -310,6 +_,7 @@
         this.extractPanorama(graphics, a);
         super.extractRenderState(graphics, mouseX, mouseY, a);
         this.logoRenderer.extractRenderState(graphics, this.width, this.logoRenderer.keepLogoThroughFade() ? 1.0F : widgetFade);
+        net.neoforged.neoforge.client.ClientHooks.renderMainMenu(this, graphics, this.font, this.width, this.height, ARGB.as8BitChannel(widgetFade));
         if (this.splash != null && !this.minecraft.options.hideSplashTexts().get()) {
             this.splash.extractRenderState(graphics, this.width, this.font, widgetFade);
         }
@@ -323,7 +_,13 @@
             versionString = versionString + I18n.get("menu.modded");
         }
 
-        graphics.text(this.font, versionString, 2, this.height - 10, ARGB.white(widgetFade));
+        int brandingColor = ARGB.color(widgetFade, -1);
+        net.neoforged.neoforge.internal.BrandingControl.forEachLine(true, true, (brdline, brd) ->
+                graphics.text(this.font, brd, 2, this.height - (10 + brdline * (this.font.lineHeight + 1)), brandingColor)
+        );
+        net.neoforged.neoforge.internal.BrandingControl.forEachAboveCopyrightLine((brdline, brd) ->
+                graphics.text(this.font, brd, this.width - font.width(brd), this.height - (10 + (brdline + 1) * (this.font.lineHeight + 1)), brandingColor)
+        );
         if (this.realmsNotificationsEnabled() && widgetFade >= 1.0F) {
             this.realmsNotificationsScreen.extractRenderState(graphics, mouseX, mouseY, a);
         }
