--- a/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/net/minecraft/client/gui/screens/TitleScreen.java
@@ -63,6 +_,7 @@
     @Nullable
     private TitleScreen.WarningLabel warningLabel;
     private final LogoRenderer logoRenderer;
+    private net.neoforged.neoforge.client.gui.TitleScreenModUpdateIndicator modUpdateNotification;
 
     public TitleScreen() {
         this(false);
@@ -120,30 +_,34 @@
         int j = this.width - i - 2;
         int k = 24;
         int l = this.height / 4 + 48;
+        Button modButton = null;
         if (this.minecraft.isDemo()) {
             this.createDemoMenuOptions(l, 24);
         } else {
             this.createNormalMenuOptions(l, 24);
+            modButton = this.addRenderableWidget(Button.builder(Component.translatable("fml.menu.mods"), button -> this.minecraft.setScreen(new net.neoforged.neoforge.client.gui.ModListScreen(this)))
+                .pos(this.width / 2 - 100, l + 24 * 3).size(200, 20).build());
         }
+        modUpdateNotification = net.neoforged.neoforge.client.gui.TitleScreenModUpdateIndicator.init(this, modButton);
 
         SpriteIconButton spriteiconbutton = this.addRenderableWidget(
             CommonButtons.language(
                 20, p_280830_ -> this.minecraft.setScreen(new LanguageSelectScreen(this, this.minecraft.options, this.minecraft.getLanguageManager())), true
             )
         );
-        spriteiconbutton.setPosition(this.width / 2 - 124, l + 72 + 12);
+        spriteiconbutton.setPosition(this.width / 2 - 124, l + 96 + 12);
         this.addRenderableWidget(
             Button.builder(Component.translatable("menu.options"), p_280838_ -> this.minecraft.setScreen(new OptionsScreen(this, this.minecraft.options)))
-                .bounds(this.width / 2 - 100, l + 72 + 12, 98, 20)
+                .bounds(this.width / 2 - 100, l + 96 + 12, 98, 20)
                 .build()
         );
         this.addRenderableWidget(
-            Button.builder(Component.translatable("menu.quit"), p_280831_ -> this.minecraft.stop()).bounds(this.width / 2 + 2, l + 72 + 12, 98, 20).build()
+            Button.builder(Component.translatable("menu.quit"), p_280831_ -> this.minecraft.stop()).bounds(this.width / 2 + 2, l + 96 + 12, 98, 20).build()
         );
         SpriteIconButton spriteiconbutton1 = this.addRenderableWidget(
             CommonButtons.accessibility(20, p_280835_ -> this.minecraft.setScreen(new AccessibilityOptionsScreen(this, this.minecraft.options)), true)
         );
-        spriteiconbutton1.setPosition(this.width / 2 + 104, l + 72 + 12);
+        spriteiconbutton1.setPosition(this.width / 2 + 104, l + 96 + 12);
         this.addRenderableWidget(
             new PlainTextButton(
                 j, this.height - 10, i, 10, COPYRIGHT_TEXT, p_280834_ -> this.minecraft.setScreen(new CreditsAndAttributionScreen(this)), this.font
@@ -294,6 +_,7 @@
                 this.warningLabel.render(p_282860_, i);
             }
 
+            net.neoforged.neoforge.client.ClientHooks.renderMainMenu(this, p_282860_, this.font, this.width, this.height, i);
             if (this.splash != null && !this.minecraft.options.hideSplashTexts().get()) {
                 this.splash.render(p_282860_, this.width, this.font, i);
             }
@@ -309,7 +_,13 @@
                 s = s + I18n.get("menu.modded");
             }
 
-            p_282860_.drawString(this.font, s, 2, this.height - 10, 16777215 | i);
+            net.neoforged.neoforge.internal.BrandingControl.forEachLine(true, true, (brdline, brd) ->
+                p_282860_.drawString(this.font, brd, 2, this.height - ( 10 + brdline * (this.font.lineHeight + 1)), 16777215 | i)
+            );
+
+            net.neoforged.neoforge.internal.BrandingControl.forEachAboveCopyrightLine((brdline, brd) ->
+                p_282860_.drawString(this.font, brd, this.width - font.width(brd), this.height - (10 + (brdline + 1) * ( this.font.lineHeight + 1)), 16777215 | i)
+            );
 
             for(GuiEventListener guieventlistener : this.children()) {
                 if (guieventlistener instanceof AbstractWidget) {
@@ -322,6 +_,8 @@
                 RenderSystem.enableDepthTest();
                 this.realmsNotificationsScreen.render(p_282860_, p_281753_, p_283539_, p_282628_);
             }
+            if (f1 >= 1.0f) modUpdateNotification.render(p_282860_, p_281753_, p_283539_, p_282628_);
+
         }
     }
 
