--- a/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/net/minecraft/client/gui/screens/TitleScreen.java
@@ -114,11 +_,14 @@
         int i = this.font.width(COPYRIGHT_TEXT);
         int j = this.width - i - 2;
         int k = 24;
-        int l = this.height / 4 + 48;
+        int l = this.height / 4 + 42;
         if (this.minecraft.isDemo()) {
             this.createDemoMenuOptions(l, 24);
         } else {
             this.createNormalMenuOptions(l, 24);
+            this.addRenderableWidget(new net.neoforged.neoforge.client.gui.widget.ModsButton(Button.builder(Component.translatable("fml.menu.mods", net.neoforged.fml.ModList.get().size()), button -> this.minecraft.setScreen(new net.neoforged.neoforge.client.gui.ModListScreen(this)))
+                    .pos(this.width / 2 - 100, l + 24 * 3).size(200, 20)));
+            l += 12; // Move down Options, Quit, Language, and Accessibility buttons to make room for mods button
         }
 
         SpriteIconButton spriteiconbutton = this.addRenderableWidget(
@@ -285,6 +_,7 @@
         if ((i & -67108864) != 0) {
             super.render(p_282860_, p_281753_, p_283539_, p_282628_);
             this.logoRenderer.renderLogo(p_282860_, this.width, f);
+            net.neoforged.neoforge.client.ClientHooks.renderMainMenu(this, p_282860_, this.font, this.width, this.height, i);
             if (this.splash != null && !this.minecraft.options.hideSplashTexts().get()) {
                 this.splash.render(p_282860_, this.width, this.font, i);
             }
@@ -300,7 +_,13 @@
                 s = s + I18n.get("menu.modded");
             }
 
-            p_282860_.drawString(this.font, s, 2, this.height - 10, 16777215 | i);
+            net.neoforged.neoforge.internal.BrandingControl.forEachLine(true, true, (brdline, brd) ->
+                    p_282860_.drawString(this.font, brd, 2, this.height - (10 + brdline * (this.font.lineHeight + 1)), 16777215 | i)
+            );
+
+            net.neoforged.neoforge.internal.BrandingControl.forEachAboveCopyrightLine((brdline, brd) ->
+                    p_282860_.drawString(this.font, brd, this.width - font.width(brd), this.height - (10 + (brdline + 1) * (this.font.lineHeight + 1)), 16777215 | i)
+            );
             if (this.realmsNotificationsEnabled() && f >= 1.0F) {
                 RenderSystem.enableDepthTest();
                 this.realmsNotificationsScreen.render(p_282860_, p_281753_, p_283539_, p_282628_);
