--- a/net/minecraft/client/renderer/extract/LevelExtractor.java
+++ b/net/minecraft/client/renderer/extract/LevelExtractor.java
@@ -234,7 +242,7 @@
         for (Entity entity : this.level.entitiesForRendering()) {
             if (this.isEntityVisible(entity, frustum, camX, camY, camZ)
                 && (entity != camera.entity() || camera.isDetached() || camera.entity() instanceof LivingEntity && ((LivingEntity)camera.entity()).isSleeping())
-                && (!(entity instanceof LocalPlayer) || camera.entity() == entity)) {
+                && (!(entity instanceof LocalPlayer) || camera.entity() == entity || (entity == minecraft.player && !minecraft.player.isSpectator()))) { // Neo: render local player entity when it is not the camera entity
                 if (entity.tickCount == 0) {
                     entity.xOld = entity.getX();
                     entity.yOld = entity.getY();
