--- a/net/minecraft/data/DataGenerator.java
+++ b/net/minecraft/data/DataGenerator.java
@@ -19,6 +_,7 @@
     protected final PackOutput vanillaPackOutput;
     protected final Set<String> allProviderIds = new HashSet<>();
     protected final Map<String, DataProvider> providersToRun = new LinkedHashMap<>();
+    private final Map<String, DataProvider> providersView = java.util.Collections.unmodifiableMap(this.providersToRun);
 
     public DataGenerator(Path output) {
         this.vanillaPackOutput = new PackOutput(output);
@@ -35,6 +_,55 @@
         return new DataGenerator.PackGenerator(toRun, packId, new PackOutput(packOutputDir));
     }
 
+    public PackGenerator getBuiltinDatapack(boolean run, String namespace, String path) {
+        var packPath = vanillaPackOutput.getOutputFolder(PackOutput.Target.DATA_PACK).resolve(namespace).resolve("datapacks").resolve(path);
+        return new PackGenerator(run, namespace + '_' + path, new PackOutput(packPath));
+    }
+
+    public Map<String, DataProvider> getProvidersView() {
+         return this.providersView;
+    }
+
+    public PackOutput getPackOutput() {
+        return this.vanillaPackOutput;
+    }
+
+    public PackOutput getPackOutput(String path) {
+        return new PackOutput(vanillaPackOutput.getOutputFolder().resolve(path));
+    }
+
+    public PackGenerator getPackGenerator(boolean run, String providerPrefix, String path) {
+        return new PackGenerator(run, providerPrefix, getPackOutput(path));
+    }
+
+    public <T extends DataProvider> T addProvider(boolean run, DataProvider.Factory<T> factory) {
+        return addProvider(run, factory.create(this.vanillaPackOutput));
+    }
+
+    public <T extends DataProvider> T addProvider(boolean run, T provider) {
+        String id = provider.getName();
+
+        if (!DataGenerator.this.allProviderIds.add(id))
+            throw new IllegalStateException("Duplicate provider: " + id);
+
+        if (run)
+            DataGenerator.this.providersToRun.put(id, provider);
+
+        return provider;
+    }
+
+    public void merge(DataGenerator other) {
+        other.providersToRun.forEach((id, provider) -> {
+            if(!allProviderIds.add(id))
+                throw new IllegalStateException("Duplicate provider: " + id);
+
+            providersToRun.put(id, provider);
+        });
+
+        other.providersToRun.clear();
+        other.allProviderIds.clear();
+    }
+
     static {
         Bootstrap.bootStrap();
     }
@@ -61,6 +_,7 @@
                     DataGenerator.LOGGER.debug("Generator {} already run for version {}", providerId, this.version.name());
                 } else {
                     DataGenerator.LOGGER.info("Starting provider: {}", providerId);
+                    net.neoforged.fml.loading.progress.StartupNotificationManager.addModMessage("Generating: " + providerId);
                     stopwatch.start();
                     cache.applyUpdate(cache.generateUpdate(providerId, provider::run).join());
                     stopwatch.stop();
@@ -112,6 +_,7 @@
             Stopwatch stopwatch = Stopwatch.createUnstarted();
             this.providersToRun.forEach((providerId, provider) -> {
                 DataGenerator.LOGGER.info("Starting uncached provider: {}", providerId);
+                net.neoforged.fml.loading.progress.StartupNotificationManager.addModMessage("Generating: " + providerId);
                 stopwatch.start();
                 provider.run(CachedOutput.NO_CACHE).join();
                 stopwatch.stop();
