--- a/net/minecraft/data/Main.java
+++ b/net/minecraft/data/Main.java
@@ -73,10 +_,13 @@
         OptionSpec<Void> serverOption = parser.accepts("server", "Include server generators");
         OptionSpec<Void> devOption = parser.accepts("dev", "Include development tools");
         OptionSpec<Void> reportsOption = parser.accepts("reports", "Include data reports");
-        parser.accepts("validate", "Validate inputs");
+        OptionSpec<Void> validateSpec = parser.accepts("validate", "Validate inputs");
         OptionSpec<Void> allOption = parser.accepts("all", "Include all generators");
         OptionSpec<String> outputOption = parser.accepts("output", "Output folder").withRequiredArg().defaultsTo("generated");
         OptionSpec<String> inputOption = parser.accepts("input", "Input folder").withRequiredArg();
+        OptionSpec<String> existing = parser.accepts("existing", "Existing resource packs that generated resources can reference").withRequiredArg();
+        OptionSpec<String> mod = parser.accepts("mod", "A modid to dump").withRequiredArg().withValuesSeparatedBy(",");
+        OptionSpec<Void> flat = parser.accepts("flat", "Do not append modid prefix to output directory when generating for multiple mods");
         OptionSet optionSet = parser.parse(args);
         if (!optionSet.has(helpOption) && optionSet.hasOptions()) {
             Path output = Paths.get(outputOption.value(optionSet));
@@ -85,11 +_,18 @@
             boolean dev = allOptions || optionSet.has(devOption);
             boolean reports = allOptions || optionSet.has(reportsOption);
             Collection<Path> input = optionSet.valuesOf(inputOption).stream().map(x$0 -> Paths.get(x$0)).toList();
-            DataGenerator generator = new DataGenerator.Cached(output, SharedConstants.getCurrentVersion(), true);
+            java.util.Set<String> mods = new java.util.HashSet<>(optionSet.valuesOf(mod));
+            boolean isFlat = mods.isEmpty() || optionSet.has(flat);
+            boolean validate = optionSet.has(validateSpec);
+            DataGenerator generator = new DataGenerator.Cached(isFlat ? output : output.resolve("minecraft"), SharedConstants.getCurrentVersion(), true);
+            java.util.Collection<Path> existingPacks = optionSet.valuesOf(existing).stream().map(Paths::get).toList();
+            if (mods.contains("minecraft") || mods.isEmpty()) {
             addServerDefinitionProviders(generator, server, reports);
             addServerConverters(generator, input, server, dev);
-            generator.run();
             Util.shutdownExecutors();
+            }
+            net.neoforged.neoforge.data.loading.DatagenModLoader.begin(mods, output, input, existingPacks, dev, reports, validate, isFlat, () -> {},
+                    net.neoforged.neoforge.data.event.GatherDataEvent.Server::new, generator, packConsumer -> {});
         } else {
             parser.printHelpOn(System.out);
         }
