--- a/net/minecraft/world/clock/ServerClockManager.java
+++ b/net/minecraft/world/clock/ServerClockManager.java
@@ -116,9 +134,22 @@
         this.setDirty();
     }
 
+    /**
+     * Returns the current ratio between game ticks and clock ticks. This value cannot be 0 or negative.
+     * The default is 1.
+     */
+    public float getRate(Holder<WorldClock> definition) {
+        return getInstance(definition).rate;
+    }
+
     @Override
     public long getTotalTicks(Holder<WorldClock> definition) {
         return this.getInstance(definition).totalTicks;
+    }
+
+    @Override
+    public float getPartialTick(Holder<WorldClock> definition) {
+        return this.getInstance(definition).partialTick;
     }
 
     public ClientboundSetTimePacket createFullSyncPacket() {
