--- a/net/minecraft/world/level/block/AbstractCauldronBlock.java
+++ b/net/minecraft/world/level/block/AbstractCauldronBlock.java
@@ -99,4 +_,22 @@
 
     protected void receiveStalactiteDrip(BlockState p_151975_, Level p_151976_, BlockPos p_151977_, Fluid p_151978_) {
     }
+
+    @Override
+    public void onPlace(BlockState p_51978_, Level p_51979_, BlockPos p_51980_, BlockState p_51981_, boolean p_51982_) {
+        super.onPlace(p_51978_, p_51979_, p_51980_, p_51981_, p_51982_);
+        // Neo: Invalidate cauldron capabilities when a cauldron is added
+        if (net.neoforged.neoforge.fluids.CauldronFluidContent.getForBlock(p_51981_.getBlock()) == null) {
+            p_51979_.invalidateCapabilities(p_51980_);
+        }
+    }
+
+    @Override
+    protected void affectNeighborsAfterRemoval(BlockState p_394424_, ServerLevel p_394241_, BlockPos p_393520_, boolean p_394545_) {
+        super.affectNeighborsAfterRemoval(p_394424_, p_394241_, p_393520_, p_394545_);
+        // Neo: Invalidate cauldron capabilities when a cauldron is removed
+        if (net.neoforged.neoforge.fluids.CauldronFluidContent.getForBlock(p_394241_.getBlockState(p_393520_).getBlock()) == null) {
+            p_394241_.invalidateCapabilities(p_393520_);
+        }
+    }
 }
