--- a/net/minecraft/world/level/block/AbstractCauldronBlock.java
+++ b/net/minecraft/world/level/block/AbstractCauldronBlock.java
@@ -101,4 +_,22 @@
 
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
     }
+
+    @Override
+    public void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean isMoving) {
+        super.onPlace(state, level, pos, oldState, isMoving);
+        // Neo: Invalidate cauldron capabilities when a cauldron is added
+        if (net.neoforged.neoforge.fluids.CauldronFluidContent.getForBlock(oldState.getBlock()) == null) {
+            level.invalidateCapabilities(pos);
+        }
+    }
+
+    @Override
+    public void onRemove(BlockState state, Level level, BlockPos pos, BlockState newState, boolean movedByPiston) {
+        super.onRemove(state, level, pos, newState, movedByPiston);
+        // Neo: Invalidate cauldron capabilities when a cauldron is removed
+        if (net.neoforged.neoforge.fluids.CauldronFluidContent.getForBlock(newState.getBlock()) == null) {
+            level.invalidateCapabilities(pos);
+        }
+    }
 }
