--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -86,7 +_,8 @@
             return null;
         } else {
             BlockState belowState = context.getLevel().getBlockState(context.getClickedPos().below());
-            if (belowState.is(BlockTags.SUPPORTS_BAMBOO)) {
+            var soilDecision = belowState.canSustainPlant(context.getLevel(), context.getClickedPos().below(), net.minecraft.core.Direction.UP, this.defaultBlockState());
+            if (soilDecision.isDefault() ? belowState.is(BlockTags.SUPPORTS_BAMBOO) : soilDecision.isTrue()) {
                 if (belowState.is(Blocks.BAMBOO_SAPLING)) {
                     return this.defaultBlockState().setValue(AGE, 0);
                 } else if (belowState.is(Blocks.BAMBOO)) {
@@ -119,10 +_,11 @@
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (state.getValue(STAGE) == 0) {
-            if (random.nextInt(3) == 0 && level.isEmptyBlock(pos.above()) && level.getRawBrightness(pos.above(), 0) >= 9) {
+            if (level.isEmptyBlock(pos.above()) && level.getRawBrightness(pos.above(), 0) >= 9) {
                 int height = this.getHeightBelowUpToMax(level, pos) + 1;
-                if (height < 16) {
+                if (height < 16 && net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, random.nextInt(3) == 0)) {
                     this.growBamboo(state, level, pos, random, height);
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
                 }
             }
         }
@@ -130,6 +_,8 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+        var soilDecision = level.getBlockState(pos.below()).canSustainPlant(level, pos.below(), Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return level.getBlockState(pos.below()).is(BlockTags.SUPPORTS_BAMBOO);
     }
 
