--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -143,6 +_,7 @@
         if (!oldState.is(state.getBlock())) {
             if (inPortalDimension(level)) {
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(level, pos, Direction.Axis.X);
+                optional = net.neoforged.neoforge.event.EventHooks.onTrySpawnPortal(level, pos, optional);
                 if (optional.isPresent()) {
                     optional.get().createPortalBlocks();
                     return;
@@ -185,7 +_,7 @@
             boolean flag = false;
 
             for (Direction direction : Direction.values()) {
-                if (level.getBlockState(blockpos$mutableblockpos.set(pos).move(direction)).is(Blocks.OBSIDIAN)) {
+                if (level.getBlockState(blockpos$mutableblockpos.set(pos).move(direction)).isPortalFrame(level, blockpos$mutableblockpos)) {
                     flag = true;
                     break;
                 }
