--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -334,4 +_,11 @@
     private static int[][] bedAboveStandUpOffsets(Direction forward) {
         return new int[][]{{0, 0}, {-forward.getStepX(), -forward.getStepZ()}};
     }
+
+    // Neoforge: beds are relocatable if and only if both halves are being relocated
+    @Override
+    public boolean isRelocatable(LevelReader level, BlockPos thisPos, BlockState thisState, java.util.function.Predicate<BlockPos> containsPosition) {
+        return super.isRelocatable(level, thisPos, thisState, containsPosition)
+            && containsPosition.test(thisPos.relative(getConnectedDirection(thisState))); // getConnectedDirection points to the other half
+    }
 }
