--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -334,4 +_,20 @@
     private static int[][] bedAboveStandUpOffsets(Direction forward) {
         return new int[][]{{0, 0}, {-forward.getStepX(), -forward.getStepZ()}};
     }
+
+    // Neoforge: beds are relocatable if and only if both halves are being relocated
+    @Override
+    public net.neoforged.neoforge.common.util.BlockRelocability getRelocability(LevelReader level, BlockPos thisPos, BlockState thisState) {
+        if (thisState.is(net.neoforged.neoforge.common.Tags.Blocks.RELOCATION_NOT_SUPPORTED)) {
+            return net.neoforged.neoforge.common.util.BlockRelocability.Never.INSTANCE;
+        }
+        // verify the other half exists; if not, don't deny relocation
+        BedPart thisPart = thisState.getValue(PART);
+        BlockPos neighborPos = thisPos.relative(getConnectedDirection(thisState)); // getConnectedDirection points to the other half
+        BlockState neighborState = level.getBlockState(neighborPos);
+        BlockState expectedState = thisState.setValue(PART, thisPart == BedPart.HEAD ? BedPart.FOOT : BedPart.HEAD);
+        if (neighborState != expectedState)
+            return net.neoforged.neoforge.common.util.BlockRelocability.Always.INSTANCE;
+        return new net.neoforged.neoforge.common.util.BlockRelocability.CuboidMultiblock(net.minecraft.world.level.levelgen.structure.BoundingBox.fromCorners(thisPos, neighborPos));
+    }
 }
