--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -126,7 +_,7 @@
         boolean flag = false;
         if (i >= 5) {
             Item item = stack.getItem();
-            if (stack.is(Items.SHEARS)) {
+            if (stack.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_HARVEST)) {
                 level.playSound(player, player.getX(), player.getY(), player.getZ(), SoundEvents.BEEHIVE_SHEAR, SoundSource.BLOCKS, 1.0F, 1.0F);
                 dropHoneycomb(level, pos);
                 stack.hurtAndBreak(1, player, LivingEntity.getSlotForHand(hand));
