--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -126,7 +_,7 @@
         boolean flag = false;
         if (i >= 5) {
             Item item = p_316844_.getItem();
-            if (p_316844_.is(Items.SHEARS)) {
+            if (p_316844_.canPerformAction(net.neoforged.neoforge.common.ToolActions.SHEARS_HARVEST)) {
                 p_316306_.playSound(p_316824_, p_316824_.getX(), p_316824_.getY(), p_316824_.getZ(), SoundEvents.BEEHIVE_SHEAR, SoundSource.BLOCKS, 1.0F, 1.0F);
                 dropHoneycomb(p_316306_, p_316497_);
                 p_316844_.hurtAndBreak(1, p_316824_, LivingEntity.getSlotForHand(p_316436_));
