--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -675,7 +_,7 @@
     public static final Block RED_BED = registerBed("red_bed", DyeColor.RED);
     public static final Block BLACK_BED = registerBed("black_bed", DyeColor.BLACK);
     public static final Block POWERED_RAIL = register(
-        "powered_rail", PoweredRailBlock::new, BlockBehaviour.Properties.of().noCollission().strength(0.7F).sound(SoundType.METAL)
+        "powered_rail", p_55218_ -> new PoweredRailBlock(p_55218_, true), BlockBehaviour.Properties.of().noCollission().strength(0.7F).sound(SoundType.METAL)
     );
     public static final Block DETECTOR_RAIL = register(
         "detector_rail", DetectorRailBlock::new, BlockBehaviour.Properties.of().noCollission().strength(0.7F).sound(SoundType.METAL)
@@ -6935,7 +_,8 @@
     static {
         for (Block block : BuiltInRegistries.BLOCK) {
             for (BlockState blockstate : block.getStateDefinition().getPossibleStates()) {
-                Block.BLOCK_STATE_REGISTRY.add(blockstate);
+                // Neo: comment out, it's done in NeoForgeRegistryCallbacks
+                //Block.BLOCK_STATE_REGISTRY.add(blockstate);
                 blockstate.initCache();
             }
         }
