--- a/net/minecraft/world/level/block/BubbleColumnBlock.java
+++ b/net/minecraft/world/level/block/BubbleColumnBlock.java
@@ -121,10 +_,12 @@
     private static BlockState getColumnState(BlockState blockState) {
         if (blockState.is(Blocks.BUBBLE_COLUMN)) {
             return blockState;
-        } else if (blockState.is(Blocks.SOUL_SAND)) {
+        }
+        net.neoforged.neoforge.common.enums.BubbleColumnDirection bubbleColumnDirection = blockState.getBubbleColumnDirection(); // Neo: PR #931
+        if (bubbleColumnDirection == net.neoforged.neoforge.common.enums.BubbleColumnDirection.UPWARD) {
             return Blocks.BUBBLE_COLUMN.defaultBlockState().setValue(DRAG_DOWN, Boolean.valueOf(false));
         } else {
-            return blockState.is(Blocks.MAGMA_BLOCK)
+            return bubbleColumnDirection == net.neoforged.neoforge.common.enums.BubbleColumnDirection.DOWNWARD
                 ? Blocks.BUBBLE_COLUMN.defaultBlockState().setValue(DRAG_DOWN, Boolean.valueOf(true))
                 : Blocks.WATER.defaultBlockState();
         }
@@ -190,7 +_,7 @@
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
         BlockState blockstate = level.getBlockState(pos.below());
-        return blockstate.is(Blocks.BUBBLE_COLUMN) || blockstate.is(Blocks.MAGMA_BLOCK) || blockstate.is(Blocks.SOUL_SAND);
+        return blockstate.is(Blocks.BUBBLE_COLUMN) || blockstate.getBubbleColumnDirection() != net.neoforged.neoforge.common.enums.BubbleColumnDirection.NONE; // Neo: PR #931
     }
 
     @Override
