--- a/net/minecraft/world/level/block/BucketPickup.java
+++ b/net/minecraft/world/level/block/BucketPickup.java
@@ -12,5 +_,6 @@
 public interface BucketPickup extends net.neoforged.neoforge.common.extensions.IBucketPickupExtension {
     ItemStack pickupBlock(@Nullable LivingEntity user, LevelAccessor level, BlockPos pos, BlockState state);
 
+    @Deprecated//Forge: Use state sensitive variant instead
     Optional<SoundEvent> getPickupSound();
 }
