--- a/net/minecraft/world/level/block/BucketPickup.java
+++ b/net/minecraft/world/level/block/BucketPickup.java
@@ -14,5 +_,9 @@
 public interface BucketPickup extends IBucketPickupExtension {
     ItemStack pickupBlock(@Nullable LivingEntity user, LevelAccessor level, BlockPos pos, BlockState state);
 
+    /**
+     * Neo: use state sensitive variant instead.
+     */
+    @Deprecated
     Optional<SoundEvent> getPickupSound();
 }
