--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -41,6 +_,7 @@
 
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (!level.isAreaLoaded(pos, 1)) return; // Forge: prevent growing cactus from loading unloaded chunks with block update
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
         }
@@ -58,6 +_,7 @@
 
             if (i < 3) {
                 int j = state.getValue(AGE);
+                if(net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, blockpos, state, true)) {
                 if (j == 15) {
                     level.setBlockAndUpdate(blockpos, this.defaultBlockState());
                     BlockState blockstate = state.setValue(AGE, Integer.valueOf(0));
@@ -66,6 +_,8 @@
                 } else {
                     level.setBlock(pos, state.setValue(AGE, Integer.valueOf(j + 1)), 4);
                 }
+                net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
+                }
             }
         }
     }
@@ -99,6 +_,8 @@
         }
 
         BlockState blockstate1 = level.getBlockState(pos.below());
+        net.neoforged.neoforge.common.util.TriState soilDecision = blockstate1.canSustainPlant(level, pos.below(), Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return (blockstate1.is(Blocks.CACTUS) || blockstate1.is(BlockTags.SAND)) && !level.getBlockState(pos.above()).liquid();
     }
 
