--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -45,6 +_,7 @@
 
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (!level.isAreaLoaded(pos, 1)) return; // Forge: prevent growing cactus from loading unloaded chunks with block update
         if (!state.canSurvive(level, pos)) {
             level.destroyBlock(pos, true);
         }
@@ -57,6 +_,7 @@
             int height = 1;
             int age = state.getValue(AGE);
 
+            if(net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, above, state, true)) {
             while (level.getBlockState(pos.below(height)).is(this)) {
                 if (++height == 3 && age == 15) {
                     return;
@@ -78,6 +_,8 @@
             if (age < 15) {
                 level.setBlock(pos, state.setValue(AGE, age + 1), 260);
             }
+            net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
+            }
         }
     }
 
@@ -119,6 +_,8 @@
         }
 
         BlockState belowState = level.getBlockState(pos.below());
+        var soilDecision = belowState.canSustainPlant(level, pos.below(), Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return (belowState.is(this) || belowState.is(BlockTags.SUPPORTS_CACTUS)) && !level.getBlockState(pos.above()).liquid();
     }
 
