--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -58,6 +_,7 @@
 
     @Override
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
+        if (fluid.getFluidType().handleCauldronDrip(fluid, level, pos)) return;
         if (fluid == Fluids.WATER) {
             BlockState blockstate = Blocks.WATER_CAULDRON.defaultBlockState();
             level.setBlockAndUpdate(pos, blockstate);
