--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -53,6 +53,7 @@
 
    @Override
    protected void receiveStalactiteDrip(BlockState p_152940_, Level p_152941_, BlockPos p_152942_, Fluid p_152943_) {
+      if (!net.neoforged.neoforge.common.CommonHooks.handleDripInfo(p_152943_, p_152941_, p_152942_))
       if (p_152943_ == Fluids.WATER) {
          BlockState blockstate = Blocks.WATER_CAULDRON.defaultBlockState();
          p_152941_.setBlockAndUpdate(p_152942_, blockstate);
