--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -50,10 +_,11 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.getRandom().nextInt(5) == 0) {
+        if (true) {
             int age = state.getValue(AGE);
-            if (age < 2) {
+            if (age < 2 && net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, level.getRandom().nextInt(5) == 0)) {
                 level.setBlock(pos, state.setValue(AGE, age + 1), 2);
+                net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
             }
         }
     }
@@ -61,6 +_,8 @@
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
         BlockState relativeState = level.getBlockState(pos.relative(state.getValue(FACING)));
+        var soilDecision = relativeState.canSustainPlant(level, pos.relative(state.getValue(FACING)), state.getValue(FACING).getOpposite(), state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return relativeState.is(BlockTags.SUPPORTS_COCOA);
     }
 
