--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -64,10 +_,11 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        if (level.random.nextInt(5) == 0) {
+        if (true) {
             int i = state.getValue(AGE);
-            if (i < 2) {
+            if (i < 2 && net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, level.random.nextInt(5) == 0)) {
                 level.setBlock(pos, state.setValue(AGE, Integer.valueOf(i + 1)), 2);
+                net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
             }
         }
     }
@@ -75,6 +_,8 @@
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
         BlockState blockstate = level.getBlockState(pos.relative(state.getValue(FACING)));
+        net.neoforged.neoforge.common.util.TriState soilDecision = blockstate.canSustainPlant(level, pos.relative(state.getValue(FACING)), state.getValue(FACING).getOpposite(), state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return blockstate.is(BlockTags.JUNGLE_LOGS);
     }
 
