--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -64,17 +_,21 @@
 
     @Override
     protected void randomTick(BlockState p_221000_, ServerLevel p_221001_, BlockPos p_221002_, RandomSource p_221003_) {
-        if (p_221001_.random.nextInt(5) == 0) {
+        if (true) {
             int i = p_221000_.getValue(AGE);
-            if (i < 2) {
+            if (i < 2 && net.neoforged.neoforge.common.CommonHooks.canCropGrow(p_221001_, p_221002_, p_221000_, p_221001_.random.nextInt(5) == 0)) {
                 p_221001_.setBlock(p_221002_, p_221000_.setValue(AGE, Integer.valueOf(i + 1)), 2);
+                net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(p_221001_, p_221002_, p_221000_);
             }
         }
     }
 
     @Override
     protected boolean canSurvive(BlockState p_51767_, LevelReader p_51768_, BlockPos p_51769_) {
-        BlockState blockstate = p_51768_.getBlockState(p_51769_.relative(p_51767_.getValue(FACING)));
+        BlockPos attachmentPos = p_51769_.relative(p_51767_.getValue(FACING));
+        BlockState blockstate = p_51768_.getBlockState(attachmentPos).getGroundStateForPlantChecks(p_51768_, attachmentPos);
+        net.neoforged.neoforge.common.util.TriState soilDecision = blockstate.canSustainPlant(p_51768_, attachmentPos, p_51767_.getValue(FACING).getOpposite(), p_51767_);
+        if (soilDecision != net.neoforged.neoforge.common.util.TriState.DEFAULT) return soilDecision == net.neoforged.neoforge.common.util.TriState.TRUE;
         return blockstate.is(BlockTags.JUNGLE_LOGS);
     }
 
