--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -111,6 +_,15 @@
         if (state.getValue(LEVEL) == 7) {
             level.scheduleTick(pos, state.getBlock(), 20);
         }
+        // Neo: Invalidate composter capabilities when a composter is added
+        if (!oldState.is(this)) level.invalidateCapabilities(pos);
+    }
+
+    @Override
+    protected void affectNeighborsAfterRemoval(BlockState state, ServerLevel level, BlockPos pos, boolean movedByPiston) {
+        super.affectNeighborsAfterRemoval(state, level, pos, movedByPiston);
+        // Neo: Invalidate composter capabilities when a composter is removed
+        if (!level.getBlockState(pos).is(state.getBlock())) level.invalidateCapabilities(pos);
     }
 
     @Override
@@ -286,7 +_,7 @@
 
         @Override
         public boolean canPlaceItemThroughFace(int slot, ItemStack itemStack, @Nullable Direction direction) {
-            return !this.changed && direction == Direction.UP && itemStack.has(DataComponents.COMPOSTABLE);
+            return !this.changed && direction == Direction.UP && getValue(itemStack) > 0f;
         }
 
         @Override
@@ -345,5 +_,11 @@
             ComposterBlock.empty(null, this.state, this.level, this.pos);
             this.changed = true;
         }
+    }
+
+    public static float getValue(ItemStack item) {
+        var value = item.getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.COMPOSTABLES);
+        if (value != null) return value.chance();
+        return -1f;
     }
 }
