--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -47,6 +_,8 @@
     public static final int MIN_LEVEL = 0;
     public static final int MAX_LEVEL = 7;
     public static final IntegerProperty LEVEL = BlockStateProperties.LEVEL_COMPOSTER;
+    /** @deprecated Neo: Use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#COMPOSTABLES compostable} data map instead */
+    @Deprecated
     public static final Object2FloatMap<ItemLike> COMPOSTABLES = new Object2FloatOpenHashMap<>();
     private static final int HOLE_WIDTH = 12;
     private static final VoxelShape[] SHAPES = Util.make(
@@ -242,6 +_,15 @@
         if (state.getValue(LEVEL) == 7) {
             level.scheduleTick(pos, state.getBlock(), 20);
         }
+        // Neo: Invalidate composter capabilities when a composter is added
+        if (!oldState.is(this)) level.invalidateCapabilities(pos);
+    }
+
+    @Override
+    protected void affectNeighborsAfterRemoval(BlockState state, ServerLevel level, BlockPos pos, boolean movedByPiston) {
+        super.affectNeighborsAfterRemoval(state, level, pos, movedByPiston);
+        // Neo: Invalidate composter capabilities when a composter is removed
+        if (!level.getBlockState(pos).is(state.getBlock())) level.invalidateCapabilities(pos);
     }
 
     @Override
@@ -249,7 +_,7 @@
         ItemStack itemStack, BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult
     ) {
         int fillLevel = state.getValue(LEVEL);
-        if (fillLevel < 8 && COMPOSTABLES.containsKey(itemStack.getItem())) {
+        if (fillLevel < 8 && getValue(itemStack) > 0) {
             if (fillLevel < 7 && !level.isClientSide()) {
                 BlockState newState = addItem(player, state, level, pos, itemStack);
                 level.levelEvent(1500, pos, state != newState ? 1 : 0);
@@ -276,7 +_,7 @@
 
     public static BlockState insertItem(Entity sourceEntity, BlockState state, ServerLevel level, ItemStack itemStack, BlockPos pos) {
         int fillLevel = state.getValue(LEVEL);
-        if (fillLevel < 7 && COMPOSTABLES.containsKey(itemStack.getItem())) {
+        if (fillLevel < 7 && getValue(itemStack) > 0) {
             BlockState newState = addItem(sourceEntity, state, level, pos, itemStack);
             itemStack.shrink(1);
             return newState;
@@ -307,7 +_,7 @@
 
     private static BlockState addItem(@Nullable Entity sourceEntity, BlockState state, LevelAccessor level, BlockPos pos, ItemStack itemStack) {
         int fillLevel = state.getValue(LEVEL);
-        float chance = COMPOSTABLES.getFloat(itemStack.getItem());
+        float chance = getValue(itemStack);
         if ((fillLevel != 0 || !(chance > 0.0F)) && !(level.getRandom().nextDouble() < chance)) {
             return state;
         }
@@ -407,7 +_,7 @@
 
         @Override
         public boolean canPlaceItemThroughFace(int slot, ItemStack itemStack, @Nullable Direction direction) {
-            return !this.changed && direction == Direction.UP && ComposterBlock.COMPOSTABLES.containsKey(itemStack.getItem());
+            return !this.changed && direction == Direction.UP && getValue(itemStack) > 0f;
         }
 
         @Override
@@ -465,5 +_,11 @@
             ComposterBlock.empty(null, this.state, this.level, this.pos);
             this.changed = true;
         }
+    }
+
+    public static float getValue(ItemStack item) {
+        var value = item.getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.COMPOSTABLES);
+        if (value != null) return value.chance();
+        return -1f;
     }
 }
