--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -35,7 +_,7 @@
 
     @Override
     public void onLand(Level level, BlockPos pos, BlockState state, BlockState replacedBlock, FallingBlockEntity entity) {
-        if (shouldSolidify(level, pos, replacedBlock)) {
+        if (shouldSolidify(level, pos, state, replacedBlock.getFluidState())) { // Forge: Use block of falling entity instead of block at replaced position, and check if shouldSolidify with the FluidState of the replaced block
             level.setBlock(pos, this.concrete.defaultBlockState(), 3);
         }
     }
@@ -48,20 +_,24 @@
         return shouldSolidify(level, pos, replacedBlock) ? this.concrete.defaultBlockState() : super.getStateForPlacement(context);
     }
 
+    private static boolean shouldSolidify(BlockGetter level, BlockPos pos, BlockState replacedBlock, net.minecraft.world.level.material.FluidState fluidState) {
+        return replacedBlock.canBeHydrated(level, pos, fluidState, pos) || touchesLiquid(level, pos, replacedBlock);
+    }
+
     private static boolean shouldSolidify(BlockGetter level, BlockPos pos, BlockState replacedBlock) {
-        return canSolidify(replacedBlock) || touchesLiquid(level, pos);
+        return shouldSolidify(level, pos, replacedBlock, level.getFluidState(pos));
     }
 
-    private static boolean touchesLiquid(BlockGetter level, BlockPos pos) {
+    private static boolean touchesLiquid(BlockGetter level, BlockPos pos, BlockState state) {
         boolean touchesLiquid = false;
         BlockPos.MutableBlockPos testPos = pos.mutable();
 
         for (Direction direction : Direction.values()) {
             BlockState blockState = level.getBlockState(testPos);
-            if (direction != Direction.DOWN || canSolidify(blockState)) {
+            if (direction != Direction.DOWN || state.canBeHydrated(level, pos, blockState.getFluidState(), testPos)) {
                 testPos.setWithOffset(pos, direction);
                 blockState = level.getBlockState(testPos);
-                if (canSolidify(blockState) && !blockState.isFaceSturdy(level, pos, direction.getOpposite())) {
+                if (state.canBeHydrated(level, pos, blockState.getFluidState(), testPos) && !blockState.isFaceSturdy(level, pos, direction.getOpposite())) {
                     touchesLiquid = true;
                     break;
                 }
@@ -86,7 +_,7 @@
         BlockState neighbourState,
         RandomSource random
     ) {
-        return touchesLiquid(level, pos)
+        return touchesLiquid(level, pos, state)
             ? this.concrete.defaultBlockState()
             : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
     }
