--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -60,9 +_,10 @@
     }
 
     protected boolean scanForWater(BlockGetter level, BlockPos blockPos) {
+        BlockState state = level.getBlockState(blockPos);
         for (Direction direction : Direction.values()) {
             FluidState fluidState = level.getFluidState(blockPos.relative(direction));
-            if (fluidState.is(FluidTags.WATER)) {
+            if (state.canBeHydrated(level, blockPos, fluidState, blockPos.relative(direction))) {
                 return true;
             }
         }
