--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -51,9 +_,10 @@
     }
 
     protected boolean scanForWater(BlockGetter level, BlockPos pos) {
+        BlockState state = level.getBlockState(pos);
         for (Direction direction : Direction.values()) {
             FluidState fluidstate = level.getFluidState(pos.relative(direction));
-            if (fluidstate.is(FluidTags.WATER)) {
+            if (state.canBeHydrated(level, pos, fluidstate, pos.relative(direction))) {
                 return true;
             }
         }
