--- a/net/minecraft/world/level/block/CrafterBlock.java
+++ b/net/minecraft/world/level/block/CrafterBlock.java
@@ -183,7 +_,8 @@
 
     private void dispenseItem(ServerLevel level, BlockPos pos, CrafterBlockEntity blockEntity, ItemStack results, BlockState blockState, RecipeHolder<?> recipe) {
         Direction direction = blockState.getValue(ORIENTATION).front();
-        Container into = HopperBlockEntity.getContainerAt(level, pos.relative(direction));
+        var containerOrHandler = HopperBlockEntity.getContainerOrHandlerAt(level, pos.relative(direction), direction.getOpposite());
+        Container into = containerOrHandler.container();
         ItemStack remaining = results.copy();
         if (into != null && (into instanceof CrafterBlockEntity || results.getCount() > into.getMaxStackSize(results))) {
             while (!remaining.isEmpty()) {
@@ -195,10 +_,14 @@
 
                 remaining.shrink(1);
             }
-        } else if (into != null) {
+        } else if (!containerOrHandler.isEmpty()) {
             while (!remaining.isEmpty()) {
                 int oldSize = remaining.getCount();
-                remaining = HopperBlockEntity.addItem(blockEntity, into, remaining, direction.getOpposite());
+                if (into != null) {
+                    remaining = HopperBlockEntity.addItem(blockEntity, into, remaining, direction.getOpposite());
+                } else {
+                    remaining = net.neoforged.neoforge.transfer.item.ItemUtil.insertItemReturnRemaining(containerOrHandler.itemHandler(), remaining, false, null);
+                }
                 if (oldSize == remaining.getCount()) {
                     break;
                 }
