--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -77,12 +_,14 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (!level.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
         if (level.getRawBrightness(pos, 0) >= 9) {
             int age = this.getAge(state);
             if (age < this.getMaxAge()) {
-                float growthSpeed = getGrowthSpeed(this, level, pos);
-                if (random.nextInt((int)(25.0F / growthSpeed) + 1) == 0) {
+                float f = getGrowthSpeed(state, level, pos);
+                if (net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, random.nextInt((int)(25.0F / f) + 1) == 0)) {
                     level.setBlock(pos, this.getStateForAge(age + 1), 2);
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
                 }
             }
         }
@@ -97,7 +_,8 @@
         return Mth.nextInt(level.getRandom(), 2, 5);
     }
 
-    protected static float getGrowthSpeed(Block type, BlockGetter level, BlockPos pos) {
+    protected static float getGrowthSpeed(BlockState cropBlockState, BlockGetter level, BlockPos pos) {
+        Block type = cropBlockState.getBlock();
         float speed = 1.0F;
         BlockPos below = pos.below();
 
@@ -105,9 +_,10 @@
             for (int zz = -1; zz <= 1; zz++) {
                 float blockSpeed = 0.0F;
                 BlockState blockState = level.getBlockState(below.offset(xx, 0, zz));
-                if (blockState.is(BlockTags.GROWS_CROPS)) {
+                var soilDecision = blockState.canSustainPlant(level, below.offset(xx, 0, zz), net.minecraft.core.Direction.UP, blockState);
+                if (soilDecision.isDefault() ? blockState.is(BlockTags.GROWS_CROPS) : soilDecision.isTrue()) {
                     blockSpeed = 1.0F;
-                    if (blockState.getValueOrElse(FarmlandBlock.MOISTURE, 0) > 0) {
+                    if (blockState.isFertile(level, pos.offset(xx, 0, zz))) {
                         blockSpeed = 3.0F;
                     }
                 }
@@ -143,6 +_,8 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+        var soilDecision = level.getBlockState(pos.below()).canSustainPlant(level, pos.below(), net.minecraft.core.Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return hasSufficientLight(level, pos) && super.canSurvive(state, level, pos);
     }
 
@@ -152,7 +_,7 @@
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
-        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverLevel, entity)) {
             serverLevel.destroyBlock(pos, true, entity);
         }
 
