--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -24,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class CropBlock extends VegetationBlock implements BonemealableBlock {
+public class CropBlock extends VegetationBlock implements BonemealableBlock, net.neoforged.neoforge.common.extensions.IHarvestableExtension {
     public static final MapCodec<CropBlock> CODEC = simpleCodec(CropBlock::new);
     public static final int MAX_AGE = 7;
     public static final IntegerProperty AGE = BlockStateProperties.AGE_7;
@@ -62,10 +_,18 @@
         return state.getValue(this.getAgeProperty());
     }
 
+    /**
+     * @deprecated Neo: use overload with level context instead
+     * */
+    @Deprecated
     public BlockState getStateForAge(int age) {
         return this.defaultBlockState().setValue(this.getAgeProperty(), age);
     }
 
+    public BlockState getStateForAge(LevelReader level, BlockPos pos, BlockState state, int age){
+        return getStateForAge(age);
+    }
+
     public final boolean isMaxAge(BlockState state) {
         return this.getAge(state) >= this.getMaxAge();
     }
@@ -77,12 +_,14 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (!level.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
         if (level.getRawBrightness(pos, 0) >= 9) {
             int age = this.getAge(state);
             if (age < this.getMaxAge()) {
-                float growthSpeed = getGrowthSpeed(this, level, pos);
-                if (random.nextInt((int)(25.0F / growthSpeed) + 1) == 0) {
-                    level.setBlock(pos, this.getStateForAge(age + 1), 2);
+                float growthSpeed = getGrowthSpeed(state, level, pos);
+                if (net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, random.nextInt((int)(25.0F / growthSpeed) + 1) == 0)) {
+                    level.setBlock(pos, this.getStateForAge(level, pos, state, age + 1), 2);
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
                 }
             }
         }
@@ -90,14 +_,15 @@
 
     public void growCrops(Level level, BlockPos pos, BlockState state) {
         int age = Math.min(this.getMaxAge(), this.getAge(state) + this.getBonemealAgeIncrease(level));
-        level.setBlock(pos, this.getStateForAge(age), 2);
+        level.setBlock(pos, this.getStateForAge(level, pos, state, age), 2);
     }
 
     protected int getBonemealAgeIncrease(Level level) {
         return Mth.nextInt(level.getRandom(), 2, 5);
     }
 
-    protected static float getGrowthSpeed(Block type, BlockGetter level, BlockPos pos) {
+    protected static float getGrowthSpeed(BlockState cropBlockState, BlockGetter level, BlockPos pos) {
+        Block type = cropBlockState.getBlock();
         float speed = 1.0F;
         BlockPos below = pos.below();
 
@@ -105,9 +_,10 @@
             for (int zz = -1; zz <= 1; zz++) {
                 float blockSpeed = 0.0F;
                 BlockState blockState = level.getBlockState(below.offset(xx, 0, zz));
-                if (blockState.is(BlockTags.GROWS_CROPS)) {
+                var soilDecision = blockState.canSustainPlant(level, below.offset(xx, 0, zz), net.minecraft.core.Direction.UP, blockState);
+                if (soilDecision.isDefault() ? blockState.is(BlockTags.GROWS_CROPS) : soilDecision.isTrue()) {
                     blockSpeed = 1.0F;
-                    if (blockState.getValueOrElse(FarmlandBlock.MOISTURE, 0) > 0) {
+                    if (blockState.isFertile(level, pos.offset(xx, 0, zz))) {
                         blockSpeed = 3.0F;
                     }
                 }
@@ -143,6 +_,8 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+        var soilDecision = level.getBlockState(pos.below()).canSustainPlant(level, pos.below(), net.minecraft.core.Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return hasSufficientLight(level, pos) && super.canSurvive(state, level, pos);
     }
 
@@ -152,7 +_,7 @@
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
-        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+        if (level instanceof ServerLevel serverLevel && entity instanceof Ravager && net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverLevel, entity)) {
             serverLevel.destroyBlock(pos, true, entity);
         }
 
@@ -187,4 +_,5 @@
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
     }
+
 }
