--- a/net/minecraft/world/level/block/DoorBlock.java
+++ b/net/minecraft/world/level/block/DoorBlock.java
@@ -119,7 +_,7 @@
 
     @Override
     public BlockState playerWillDestroy(Level p_52755_, BlockPos p_52756_, BlockState p_52757_, Player p_52758_) {
-        if (!p_52755_.isClientSide && (p_52758_.isCreative() || !p_52758_.hasCorrectToolForDrops(p_52757_))) {
+        if (!p_52755_.isClientSide && (p_52758_.isCreative() || !p_52758_.hasCorrectToolForDrops(p_52757_, p_52755_, p_52756_))) {
             DoublePlantBlock.preventDropFromBottomPart(p_52755_, p_52756_, p_52757_, p_52758_);
         }
 
