--- a/net/minecraft/world/level/block/DoublePlantBlock.java
+++ b/net/minecraft/world/level/block/DoublePlantBlock.java
@@ -79,6 +_,7 @@
             return super.canSurvive(state, level, pos);
         } else {
             BlockState belowState = level.getBlockState(pos.below());
+            if (state.getBlock() != this) return super.canSurvive(state, level, pos); //Forge: This function is called during world gen and placement, before this block is set, so if we are not 'here' then assume it's the pre-check.
             return belowState.is(this) && belowState.getValue(HALF) == DoubleBlockHalf.LOWER;
         }
     }
@@ -132,5 +_,20 @@
     @Override
     protected long getSeed(BlockState state, BlockPos pos) {
         return Mth.getSeed(pos.getX(), pos.below(state.getValue(HALF) == DoubleBlockHalf.LOWER ? 0 : 1).getY(), pos.getZ());
+    }
+
+    // Neoforge: Relocatable if and only if both halves are being relocated
+    @Override
+    public boolean isRelocatable(LevelReader level, BlockPos thisPos, BlockState thisState, java.util.function.Predicate<BlockPos> containsPosition) {
+        if (!super.isRelocatable(level, thisPos, thisState, containsPosition))
+            return false;
+        // verify the other half exists; if not, don't deny relocation
+        DoubleBlockHalf thisHalf = thisState.getValue(HALF);
+        BlockPos neighborPos = thisPos.relative(thisHalf.getDirectionToOther());
+        BlockState neighborState = level.getBlockState(neighborPos);
+        BlockState expectedState = thisState.setValue(HALF, thisHalf.getOtherHalf());
+        if (neighborState != expectedState)
+            return true;
+        return containsPosition.test(thisPos.relative(thisState.getValue(HALF).getDirectionToOther()));
     }
 }
