--- a/net/minecraft/world/level/block/DoublePlantBlock.java
+++ b/net/minecraft/world/level/block/DoublePlantBlock.java
@@ -80,6 +_,7 @@
         }
 
         BlockState belowState = level.getBlockState(pos.below());
+        if (state.getBlock() != this) return super.canSurvive(state, level, pos); // Neo: This function is called during world gen and placement, before this block is set, so if we are not 'here' then assume it's the pre-check.
         return belowState.is(this) && belowState.getValue(HALF) == DoubleBlockHalf.LOWER;
     }
 
@@ -132,5 +_,21 @@
     @Override
     protected long getSeed(BlockState state, BlockPos pos) {
         return Mth.getSeed(pos.getX(), pos.below(state.getValue(HALF) == DoubleBlockHalf.LOWER ? 0 : 1).getY(), pos.getZ());
+    }
+
+    // Neo: Relocatable if and only if both halves are being relocated
+    @Override
+    public net.neoforged.neoforge.common.util.BlockRelocability getRelocability(LevelReader level, BlockPos pos, BlockState state) {
+        if (state.is(net.neoforged.neoforge.common.Tags.Blocks.RELOCATION_NOT_SUPPORTED)) {
+            return net.neoforged.neoforge.common.util.BlockRelocability.No.INSTANCE;
+        }
+        // verify the other half exists; if not, don't deny relocation
+        DoubleBlockHalf thisHalf = state.getValue(HALF);
+        BlockPos neighborPos = pos.relative(thisHalf.getDirectionToOther());
+        BlockState neighborState = level.getBlockState(neighborPos);
+        BlockState expectedState = state.setValue(HALF, thisHalf.getOtherHalf());
+        if (neighborState != expectedState)
+            return net.neoforged.neoforge.common.util.BlockRelocability.Yes.INSTANCE;
+        return new net.neoforged.neoforge.common.util.BlockRelocability.Multiblock(java.util.Set.of(pos, neighborPos));
     }
 }
