--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -61,7 +_,13 @@
                     Container container = HopperBlockEntity.getContainerAt(p_52944_, p_52945_.relative(direction));
                     ItemStack itemstack1;
                     if (container == null) {
-                        itemstack1 = DISPENSE_BEHAVIOUR.dispense(blocksource, itemstack);
+                        var leftover = net.neoforged.neoforge.items.VanillaInventoryCodeHooks.insertNeighbor(p_52944_, p_52945_, direction, itemstack.copyWithCount(1));
+                        if (leftover.isEmpty()) {
+                            itemstack1 = itemstack.copy();
+                            itemstack1.shrink(1);
+                        } else {
+                            itemstack1 = DISPENSE_BEHAVIOUR.dispense(blocksource, itemstack);
+                        }
                     } else {
                         itemstack1 = HopperBlockEntity.addItem(dispenserblockentity, container, itemstack.copyWithCount(1), direction.getOpposite());
                         if (itemstack1.isEmpty()) {
