--- a/net/minecraft/world/level/block/Fallable.java
+++ b/net/minecraft/world/level/block/Fallable.java
@@ -7,7 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockState;
 
-public interface Fallable {
+public interface Fallable extends net.neoforged.neoforge.common.extensions.IFallableExtension {
     default void onLand(Level p_153220_, BlockPos p_153221_, BlockState p_153222_, BlockState p_153223_, FallingBlockEntity p_153224_) {
     }
 
