--- a/net/minecraft/world/level/block/FarmlandBlock.java
+++ b/net/minecraft/world/level/block/FarmlandBlock.java
@@ -102,11 +_,8 @@
 
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
-        if (level instanceof ServerLevel serverLevel && level.getRandom().nextFloat() < fallDistance - 0.5 && entity instanceof LivingEntity) {
-            boolean canTrample = entity instanceof Player player
-                ? !level.getServer().isUnderSpawnProtection(serverLevel, pos, player)
-                : serverLevel.getGameRules().get(GameRules.MOB_GRIEFING);
-            if (canTrample && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
+        if (level instanceof ServerLevel serverLevel) {
+            if (net.neoforged.neoforge.common.CommonHooks.onFarmlandTrample(serverLevel, pos, state, fallDistance, entity)) {
                 this.turnToBaseBlock(entity, state, level, pos);
             }
         }
