--- a/net/minecraft/world/level/block/FarmlandBlock.java
+++ b/net/minecraft/world/level/block/FarmlandBlock.java
@@ -102,11 +_,8 @@
 
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
-        if (level instanceof ServerLevel serverLevel && level.getRandom().nextFloat() < fallDistance - 0.5 && entity instanceof LivingEntity) {
-            boolean canTrample = entity instanceof Player player
-                ? !level.getServer().isUnderSpawnProtection(serverLevel, pos, player)
-                : serverLevel.getGameRules().get(GameRules.MOB_GRIEFING);
-            if (canTrample && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) {
+        if (level instanceof ServerLevel serverLevel) {
+            if (net.neoforged.neoforge.common.CommonHooks.onFarmlandTrample(serverLevel, pos, state, fallDistance, entity)) {
                 this.turnToBaseBlock(entity, state, level, pos);
             }
         }
@@ -125,7 +_,9 @@
     }
 
     private static boolean isNearWater(LevelReader level, BlockPos pos) {
-        return level.findBlocksIn(pos.offset(-4, 0, -4), pos.offset(4, 1, 4)).filterState(state -> state.getFluidState().is(FluidTags.WATER)).anyMatched();
+        BlockState farmland = level.getBlockState(pos);
+        return level.findBlocksIn(pos.offset(-4, 0, -4), pos.offset(4, 1, 4)).filterState(state -> farmland.canBeHydrated(level, pos, state.getFluidState())).anyMatched()
+                || net.neoforged.neoforge.common.FarmlandWaterManager.hasBlockWaterTicket(level, pos);
     }
 
     @Override
