--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -146,7 +_,7 @@
 
         boolean opens = state.getValue(OPEN);
         level.playSound(
-            player, pos, opens ? this.type.fenceGateOpen() : this.type.fenceGateClose(), SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
+            player, pos, opens ? openSound : closeSound, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
         );
         level.gameEvent(player, opens ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
         return InteractionResult.SUCCESS;
@@ -158,7 +_,7 @@
             boolean open = state.getValue(OPEN);
             level.setBlockAndUpdate(pos, state.setValue(OPEN, !open));
             level.playSound(
-                null, pos, open ? this.type.fenceGateClose() : this.type.fenceGateOpen(), SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
+                null, pos, open ? closeSound : openSound, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
             );
             level.gameEvent(open ? GameEvent.BLOCK_CLOSE : GameEvent.BLOCK_OPEN, pos, GameEvent.Context.of(state));
         }
@@ -176,7 +_,7 @@
                     level.playSound(
                         null,
                         pos,
-                        hasPower ? this.type.fenceGateOpen() : this.type.fenceGateClose(),
+                        hasPower ? openSound : closeSound,
                         SoundSource.BLOCKS,
                         1.0F,
                         level.getRandom().nextFloat() * 0.1F + 0.9F
