--- a/net/minecraft/world/level/block/FenceGateBlock.java
+++ b/net/minecraft/world/level/block/FenceGateBlock.java
@@ -39,7 +_,10 @@
 
 public class FenceGateBlock extends HorizontalDirectionalBlock {
     public static final MapCodec<FenceGateBlock> CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(WoodType.CODEC.fieldOf("wood_type").forGetter(b -> b.type), propertiesCodec()).apply(i, FenceGateBlock::new)
+        i -> i.group(WoodType.CODEC.optionalFieldOf("wood_type").forGetter(b -> java.util.Optional.ofNullable(b.type)), propertiesCodec(),
+                        net.minecraft.sounds.SoundEvent.DIRECT_CODEC.optionalFieldOf("open_sound").forGetter(fence -> java.util.Optional.of(fence.openSound).filter(s -> fence.type == null || s != fence.type.fenceGateOpen())),
+                        net.minecraft.sounds.SoundEvent.DIRECT_CODEC.optionalFieldOf("close_sound").forGetter(fence -> java.util.Optional.of(fence.closeSound).filter(s -> fence.type == null || s != fence.type.fenceGateClose())))
+                .apply(i, FenceGateBlock::new)
     );
     public static final BooleanProperty OPEN = BlockStateProperties.OPEN;
     public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
@@ -56,6 +_,8 @@
     private static final Map<Direction.Axis, VoxelShape> SHAPE_OCCLUSION_WALL = Maps.newEnumMap(
         Util.mapValues(SHAPE_OCCLUSION, v -> v.move(0.0, -0.1875, 0.0).optimize())
     );
+    public final net.minecraft.sounds.SoundEvent openSound, closeSound;
+    @org.jspecify.annotations.Nullable
     private final WoodType type;
 
     @Override
@@ -64,8 +_,19 @@
     }
 
     public FenceGateBlock(WoodType type, BlockBehaviour.Properties properties) {
-        super(properties.sound(type.soundType()));
-        this.type = type;
+        this(java.util.Optional.of(type), properties.sound(type.soundType()), java.util.Optional.of(type.fenceGateOpen()), java.util.Optional.of(type.fenceGateClose()));
+    }
+
+    public FenceGateBlock(BlockBehaviour.Properties properties, net.minecraft.sounds.SoundEvent openSound, net.minecraft.sounds.SoundEvent closeSound) {
+        this(java.util.Optional.empty(), properties, java.util.Optional.of(openSound), java.util.Optional.of(closeSound));
+    }
+
+    public FenceGateBlock(java.util.Optional<WoodType> type, BlockBehaviour.Properties properties, java.util.Optional<net.minecraft.sounds.SoundEvent> openSound, java.util.Optional<net.minecraft.sounds.SoundEvent> closeSound) {
+        super(properties);
+        com.google.common.base.Preconditions.checkArgument(type.isPresent() || (openSound.isPresent() && closeSound.isPresent()), "Fence gates must have sounds set");
+        this.type = type.orElse(null);
+        this.openSound = openSound.orElseGet(() -> this.type.fenceGateOpen()); // Type may be null, so we cannot do a method ref
+        this.closeSound = closeSound.orElseGet(() -> this.type.fenceGateClose());
         this.registerDefaultState(this.stateDefinition.any().setValue(OPEN, false).setValue(POWERED, false).setValue(IN_WALL, false));
     }
 
@@ -156,7 +_,7 @@
 
         boolean opens = state.getValue(OPEN);
         level.playSound(
-            player, pos, opens ? this.type.fenceGateOpen() : this.type.fenceGateClose(), SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
+            player, pos, opens ? openSound : closeSound, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
         );
         level.gameEvent(player, opens ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
         return InteractionResult.SUCCESS;
@@ -168,7 +_,7 @@
             boolean open = state.getValue(OPEN);
             level.setBlockAndUpdate(pos, state.setValue(OPEN, !open));
             level.playSound(
-                null, pos, open ? this.type.fenceGateClose() : this.type.fenceGateOpen(), SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
+                null, pos, open ? closeSound : openSound, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.1F + 0.9F
             );
             level.gameEvent(open ? GameEvent.BLOCK_CLOSE : GameEvent.BLOCK_OPEN, pos, GameEvent.Context.of(state));
         }
@@ -186,7 +_,7 @@
                     level.playSound(
                         null,
                         pos,
-                        hasPower ? this.type.fenceGateOpen() : this.type.fenceGateClose(),
+                        hasPower ? openSound : closeSound,
                         SoundSource.BLOCKS,
                         1.0F,
                         level.getRandom().nextFloat() * 0.1F + 0.9F
