--- a/net/minecraft/world/level/block/FenceBlock.java
+++ b/net/minecraft/world/level/block/FenceBlock.java
@@ -50,6 +_,11 @@
     }
 
     public boolean connectsTo(BlockState state, boolean faceSolid, Direction direction) {
+        // Neo: allow the adjacent block to force or opt out of being connected to (incoming direction is the neighbor's face)
+        var adjResult = state.canBeConnectedTo(direction, this);
+        if (!adjResult.isDefault()) {
+            return adjResult.isTrue();
+        }
         Block block = state.getBlock();
         boolean sameFence = this.isSameFence(state);
         boolean gate = block instanceof FenceGateBlock && FenceGateBlock.connectsToDirection(state, direction);
