--- a/net/minecraft/world/level/block/FungusBlock.java
+++ b/net/minecraft/world/level/block/FungusBlock.java
@@ -73,6 +_,9 @@
 
     @Override
     public void performBonemeal(ServerLevel p_221243_, RandomSource p_221244_, BlockPos p_221245_, BlockState p_221246_) {
-        this.getFeature(p_221243_).ifPresent(p_256352_ -> p_256352_.value().place(p_221243_, p_221243_.getChunkSource().getGenerator(), p_221244_, p_221245_));
+        // Neo: Fire the BlockGrowFeatureEvent and change the ifPresent call to use the event's result.
+        var event = net.neoforged.neoforge.event.EventHooks.fireBlockGrowFeature(p_221243_, p_221244_, p_221245_, this.getFeature(p_221243_).orElse(null));
+        if (event.isCanceled()) return;
+        Optional.ofNullable(event.getFeature()).ifPresent(p_256352_ -> p_256352_.value().place(p_221243_, p_221243_.getChunkSource().getGenerator(), p_221244_, p_221245_));
     }
 }
