--- a/net/minecraft/world/level/block/MangrovePropaguleBlock.java
+++ b/net/minecraft/world/level/block/MangrovePropaguleBlock.java
@@ -93,7 +_,12 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
-        return isHanging(state) ? level.getBlockState(pos.above()).is(Blocks.MANGROVE_LEAVES) : super.canSurvive(state, level, pos);
+        if (isHanging(state)) {
+            net.neoforged.neoforge.common.util.TriState soilDecision = level.getBlockState(pos.above()).canSustainPlant(level, pos.above(), Direction.DOWN, state);
+            if (!soilDecision.isDefault()) return soilDecision.isTrue();
+            return level.getBlockState(pos.above()).is(Blocks.MANGROVE_LEAVES);
+        }
+        return super.canSurvive(state, level, pos);
     }
 
     @Override
