--- a/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -86,12 +_,13 @@
 
     @Override
     public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
-        return isLower(state) && !sufficientLight(level, pos) ? false : super.canSurvive(state, level, pos);
+        net.neoforged.neoforge.common.util.TriState soilDecision = level.getBlockState(pos.below()).canSustainPlant(level, pos.below(), Direction.UP, state);
+        return isLower(state) && !sufficientLight(level, pos) ? soilDecision.isTrue() : super.canSurvive(state, level, pos);
     }
 
     @Override
     protected boolean mayPlaceOn(BlockState state, BlockGetter level, BlockPos pos) {
-        return state.is(Blocks.FARMLAND);
+        return state.getBlock() instanceof net.minecraft.world.level.block.FarmBlock;
     }
 
     @Override
@@ -125,7 +_,7 @@
 
     @Override
     public void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        float f = CropBlock.getGrowthSpeed(this, level, pos);
+        float f = CropBlock.getGrowthSpeed(state, level, pos);
         boolean flag = random.nextInt((int)(25.0F / f) + 1) == 0;
         if (flag) {
             this.grow(level, state, pos, 1);
