--- a/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -102,7 +_,8 @@
 
     @Override
     public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
-        return isLower(state) && !sufficientLight(level, pos) ? false : super.canSurvive(state, level, pos);
+        var soilDecision = level.getBlockState(pos.below()).canSustainPlant(level, pos.below(), Direction.UP, state);
+        return isLower(state) && !sufficientLight(level, pos) ? soilDecision.isTrue() : super.canSurvive(state, level, pos);
     }
 
     @Override
@@ -139,7 +_,7 @@
 
     @Override
     public void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        float growthSpeed = CropBlock.getGrowthSpeed(this, level, pos);
+        float growthSpeed = CropBlock.getGrowthSpeed(state, level, pos);
         boolean shouldProgressGrowth = random.nextInt((int)(25.0F / growthSpeed) + 1) == 0;
         if (shouldProgressGrowth) {
             this.grow(level, state, pos, 1);
