--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -184,7 +_,7 @@
 
     @VisibleForTesting
     public static void maybeTransferFluid(BlockState state, ServerLevel level, BlockPos pos, float randChance) {
-        if (!(randChance > 0.17578125F) || !(randChance > 0.05859375F)) {
+        if (true) { //Neo: remove the water and lava drip chance checks to allow modded fluids to drip into cauldrons
             if (isStalactiteStartPos(state, level, pos)) {
                 Optional<PointedDripstoneBlock.FluidInfo> optional = getFluidAboveStalactite(level, pos, state);
                 if (!optional.isEmpty()) {
@@ -193,14 +_,12 @@
                     if (fluid == Fluids.WATER) {
                         f = 0.17578125F;
                     } else {
-                        if (fluid != Fluids.LAVA) {
-                            return;
-                        }
 
                         f = 0.05859375F;
                     }
 
-                    if (!(randChance >= f)) {
+                    net.neoforged.neoforge.fluids.FluidType.DripstoneDripInfo dripInfo = fluid.getFluidType().getDripInfo();
+                    if (dripInfo != null && !(randChance >= dripInfo.chance())) {
                         BlockPos blockpos = findTip(state, level, pos, 11, false);
                         if (blockpos != null) {
                             if (optional.get().sourceState.is(Blocks.MUD) && fluid == Fluids.WATER) {
@@ -411,7 +_,8 @@
         double d2 = (double)((float)(pos.getY() + 1) - 0.6875F) - 0.0625;
         double d3 = (double)pos.getZ() + 0.5 + vec3.z;
         Fluid fluid = getDripFluid(level, p_fluid);
-        ParticleOptions particleoptions = fluid.is(FluidTags.LAVA) ? ParticleTypes.DRIPPING_DRIPSTONE_LAVA : ParticleTypes.DRIPPING_DRIPSTONE_WATER;
+        ParticleOptions particleoptions = fluid.getFluidType().getDripInfo() != null ? fluid.getFluidType().getDripInfo().dripParticle() : ParticleTypes.DRIPPING_DRIPSTONE_WATER;
+        if (particleoptions != null)
         level.addParticle(particleoptions, d1, d2, d3, 0.0, 0.0, 0.0);
     }
 
@@ -563,7 +_,7 @@
     }
 
     private static boolean canFillCauldron(Fluid fluid) {
-        return fluid == Fluids.LAVA || fluid == Fluids.WATER;
+        return fluid.getFluidType().getDripInfo() != null;
     }
 
     private static boolean canGrow(BlockState dripstoneState, BlockState state) {
