--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -89,7 +_,7 @@
 
     @VisibleForTesting
     public static void maybeTransferFluid(BlockState state, ServerLevel level, BlockPos pos, float randomValue) {
-        if (!(randomValue > 0.17578125F) || !(randomValue > 0.05859375F)) {
+        if (true) { //Neo: remove the water and lava drip chance checks to allow modded fluids to drip into cauldrons
             if (isStalactiteStartPos(state, level, pos)) {
                 Optional<PointedDripstoneBlock.FluidInfo> fluidInfo = getFluidAboveStalactite(level, pos, state);
                 if (!fluidInfo.isEmpty()) {
@@ -98,14 +_,12 @@
                     if (fluid == Fluids.WATER) {
                         transferProbability = 0.17578125F;
                     } else {
-                        if (fluid != Fluids.LAVA) {
-                            return;
-                        }
 
                         transferProbability = 0.05859375F;
                     }
 
-                    if (!(randomValue >= transferProbability)) {
+                    net.neoforged.neoforge.fluids.FluidType.DripstoneDripInfo dripInfo = fluid.getFluidType().getDripInfo();
+                    if (dripInfo != null && !(randomValue >= dripInfo.chance())) {
                         BlockPos stalactiteTipPos = findTip(state, level, pos, 11, false);
                         if (stalactiteTipPos != null) {
                             if (fluidInfo.get().sourceState.is(Blocks.MUD) && fluid == Fluids.WATER) {
@@ -190,7 +_,7 @@
     }
 
     private static boolean canFillCauldron(Fluid fluidAbove) {
-        return fluidAbove == Fluids.LAVA || fluidAbove == Fluids.WATER;
+        return fluidAbove.getFluidType().getDripInfo() != null;
     }
 
     @Override
@@ -203,7 +_,9 @@
         if (fluidAbove.isSame(Fluids.EMPTY)) {
             return level.environmentAttributes().getValue(EnvironmentAttributes.DEFAULT_DRIPSTONE_PARTICLE, posAbove);
         } else {
-            return fluidAbove.is(FluidTags.LAVA) ? ParticleTypes.DRIPPING_DRIPSTONE_LAVA : ParticleTypes.DRIPPING_DRIPSTONE_WATER;
+            ParticleOptions options = fluidAbove.getFluidType().getDripInfo() != null ? fluidAbove.getFluidType().getDripInfo().dripParticle() : ParticleTypes.DRIPPING_DRIPSTONE_WATER;
+            if (options == null) options = level.environmentAttributes().getValue(EnvironmentAttributes.DEFAULT_DRIPSTONE_PARTICLE, posAbove);
+            return options;
         }
     }
 
