--- a/net/minecraft/world/level/block/PumpkinBlock.java
+++ b/net/minecraft/world/level/block/PumpkinBlock.java
@@ -35,7 +_,7 @@
     protected ItemInteractionResult useItemOn(
         ItemStack p_316383_, BlockState p_316676_, Level p_316272_, BlockPos p_316484_, Player p_316367_, InteractionHand p_316216_, BlockHitResult p_316827_
     ) {
-        if (!p_316383_.is(Items.SHEARS)) {
+        if (!p_316383_.canPerformAction(net.neoforged.neoforge.common.ToolActions.SHEARS_CARVE)) {
             return super.useItemOn(p_316383_, p_316676_, p_316272_, p_316484_, p_316367_, p_316216_, p_316827_);
         } else if (p_316272_.isClientSide) {
             return ItemInteractionResult.sidedSuccess(p_316272_.isClientSide);
