--- a/net/minecraft/world/level/block/RailBlock.java
+++ b/net/minecraft/world/level/block/RailBlock.java
@@ -163,6 +_,9 @@
                         return p_55402_.setValue(SHAPE, RailShape.NORTH_EAST);
                     case NORTH_EAST:
                         return p_55402_.setValue(SHAPE, RailShape.NORTH_WEST);
+                    case NORTH_SOUTH: //Forge fix: MC-196102
+                    case EAST_WEST:
+                        return p_55402_; // TODO 1.20.3 PORTING: reevaluate
                 }
         }
 
