--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -253,7 +_,7 @@
         BlockState blockstate = level.getBlockState(blockpos);
         if (nonNormalCubeAbove) {
             boolean flag = blockstate.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, blockpos, blockstate);
-            if (flag && shouldConnectTo(level.getBlockState(blockpos.above()))) {
+            if (flag && level.getBlockState(blockpos.above()).canRedstoneConnectTo(level, blockpos.above(), null)) {
                 if (blockstate.isFaceSturdy(level, blockpos, direction.getOpposite())) {
                     return RedstoneSide.UP;
                 }
@@ -262,10 +_,14 @@
             }
         }
 
-        return !shouldConnectTo(blockstate, direction)
-                && (blockstate.isRedstoneConductor(level, blockpos) || !shouldConnectTo(level.getBlockState(blockpos.below())))
-            ? RedstoneSide.NONE
-            : RedstoneSide.SIDE;
+        if (blockstate.canRedstoneConnectTo(level, blockpos, direction)) {
+             return RedstoneSide.SIDE;
+        } else if (blockstate.isRedstoneConductor(level, blockpos)) {
+             return RedstoneSide.NONE;
+        } else {
+             BlockPos blockPosBelow = blockpos.below();
+             return level.getBlockState(blockPosBelow).canRedstoneConnectTo(level, blockPosBelow, null) ? RedstoneSide.SIDE : RedstoneSide.NONE;
+        }
     }
 
     @Override
