--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -240,7 +_,7 @@
         BlockState relativeState = level.getBlockState(relativePos);
         if (canConnectUp) {
             boolean isPlaceableAbove = relativeState.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, relativePos, relativeState);
-            if (isPlaceableAbove && shouldConnectTo(level.getBlockState(relativePos.above()))) {
+            if (isPlaceableAbove && level.getBlockState(relativePos.above()).canRedstoneConnectTo(level, relativePos.above(), null)) {
                 if (relativeState.isFaceSturdy(level, relativePos, direction.getOpposite())) {
                     return RedstoneSide.UP;
                 }
@@ -249,10 +_,14 @@
             }
         }
 
-        return !shouldConnectTo(relativeState, direction)
-                && (relativeState.isRedstoneConductor(level, relativePos) || !shouldConnectTo(level.getBlockState(relativePos.below())))
-            ? RedstoneSide.NONE
-            : RedstoneSide.SIDE;
+        if (relativeState.canRedstoneConnectTo(level, relativePos, direction)) {
+             return RedstoneSide.SIDE;
+        } else if (relativeState.isRedstoneConductor(level, relativePos)) {
+             return RedstoneSide.NONE;
+        } else {
+             BlockPos blockPosBelow = relativePos.below();
+             return level.getBlockState(blockPosBelow).canRedstoneConnectTo(level, blockPosBelow, null) ? RedstoneSide.SIDE : RedstoneSide.NONE;
+        }
     }
 
     @Override
