--- a/net/minecraft/world/level/block/SeagrassBlock.java
+++ b/net/minecraft/world/level/block/SeagrassBlock.java
@@ -21,7 +21,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SeagrassBlock extends BushBlock implements BonemealableBlock, LiquidBlockContainer {
+public class SeagrassBlock extends BushBlock implements BonemealableBlock, LiquidBlockContainer, net.neoforged.neoforge.common.IForgeShearable {
    protected static final float AABB_OFFSET = 6.0F;
    protected static final VoxelShape SHAPE = Block.box(2.0, 0.0, 2.0, 14.0, 12.0, 14.0);
 
