--- a/net/minecraft/world/level/block/SmallDripleafBlock.java
+++ b/net/minecraft/world/level/block/SmallDripleafBlock.java
@@ -86,6 +_,8 @@
         } else {
             BlockPos belowPos = pos.below();
             BlockState belowState = level.getBlockState(belowPos);
+            var soilDecision = belowState.canSustainPlant(level, belowPos, Direction.UP, state);
+            if (!soilDecision.isDefault()) return soilDecision.isTrue();
             return this.mayPlaceOn(belowState, level, belowPos);
         }
     }
