--- a/net/minecraft/world/level/block/SoundType.java
+++ b/net/minecraft/world/level/block/SoundType.java
@@ -879,6 +_,8 @@
     private final SoundEvent hitSound;
     private final SoundEvent fallSound;
 
+    /// @deprecated Neo: Use [net.neoforged.neoforge.common.util.DeferredSoundType] instead for suppliers
+    @Deprecated
     public SoundType(float volume, float pitch, SoundEvent breakSound, SoundEvent stepSound, SoundEvent placeSound, SoundEvent hitSound, SoundEvent fallSound) {
         this.volume = volume;
         this.pitch = pitch;
