--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -50,6 +_,7 @@
     }
 
     private boolean removeWaterBreadthFirstSearch(Level level, BlockPos startPos) {
+        BlockState spongeState = level.getBlockState(startPos);
         return BlockPos.breadthFirstTraversal(startPos, 6, 65, (pos, consumer) -> {
             for (Direction direction : ALL_DIRECTIONS) {
                 consumer.accept(pos.relative(direction));
@@ -60,7 +_,7 @@
             } else {
                 BlockState state = level.getBlockState(pos);
                 FluidState fluidState = level.getFluidState(pos);
-                if (!fluidState.is(FluidTags.WATER)) {
+                if (!spongeState.canBeHydrated(level, startPos, fluidState, pos)) {
                     return BlockPos.TraversalNodeStatus.SKIP;
                 } else if (state.getBlock() instanceof BucketPickup bucketPickup && !bucketPickup.pickupBlock(null, level, pos, state).isEmpty()) {
                     return BlockPos.TraversalNodeStatus.ACCEPT;
