--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -49,6 +_,7 @@
     }
 
     private boolean removeWaterBreadthFirstSearch(Level level, BlockPos pos) {
+        BlockState spongeState = level.getBlockState(pos);
         return BlockPos.breadthFirstTraversal(
                 pos,
                 6,
@@ -64,7 +_,7 @@
                     } else {
                         BlockState blockstate = level.getBlockState(p_294069_);
                         FluidState fluidstate = level.getFluidState(p_294069_);
-                        if (!fluidstate.is(FluidTags.WATER)) {
+                        if (!spongeState.canBeHydrated(level, pos, fluidstate, p_294069_)) {
                             return false;
                         } else {
                             if (blockstate.getBlock() instanceof BucketPickup bucketpickup
