--- a/net/minecraft/world/level/block/SpreadingSnowyBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyBlock.java
@@ -48,8 +_,16 @@
         Optional<Block> baseBlock = blocks.getOptional(this.baseBlock);
         if (!baseBlock.isEmpty()) {
             if (!canStayAlive(state, level, pos)) {
+                // Neo: prevent loading unloaded chunks when checking neighbor's light and spreading
+                if (!level.isAreaLoaded(pos, 1)) {
+                    return;
+                }
                 level.setBlockAndUpdate(pos, baseBlock.get().defaultBlockState());
             } else {
+                // Neo: prevent loading unloaded chunks when checking neighbor's light and spreading
+                if (!level.isAreaLoaded(pos, 3)) {
+                    return;
+                }
                 if (level.getMaxLocalRawBrightness(pos.above()) >= 9) {
                     BlockState defaultBlockState = this.defaultBlockState();
 
