--- a/net/minecraft/world/level/block/StairBlock.java
+++ b/net/minecraft/world/level/block/StairBlock.java
@@ -94,8 +_,8 @@
                 .setValue(SHAPE, StairsShape.STRAIGHT)
                 .setValue(WATERLOGGED, Boolean.valueOf(false))
         );
-        this.base = p_56862_.getBlock();
-        this.baseState = p_56862_;
+        this.base = Blocks.AIR;
+        this.baseState = Blocks.AIR.defaultBlockState();
     }
 
     @Override
