--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -58,12 +_,15 @@
 
             if (i < 3) {
                 int j = state.getValue(AGE);
+                if (net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, true)) {
                 if (j == 15) {
                     level.setBlockAndUpdate(pos.above(), this.defaultBlockState());
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos.above(), this.defaultBlockState());
                     level.setBlock(pos, state.setValue(AGE, Integer.valueOf(0)), 4);
                 } else {
                     level.setBlock(pos, state.setValue(AGE, Integer.valueOf(j + 1)), 4);
                 }
+                }
             }
         }
     }
@@ -83,13 +_,15 @@
         if (blockstate.is(this)) {
             return true;
         } else {
+            net.neoforged.neoforge.common.util.TriState soilDecision = blockstate.canSustainPlant(level, pos.below(), Direction.UP, state);
+            if (!soilDecision.isDefault()) return soilDecision.isTrue();
             if (blockstate.is(BlockTags.DIRT) || blockstate.is(BlockTags.SAND)) {
                 BlockPos blockpos = pos.below();
 
                 for (Direction direction : Direction.Plane.HORIZONTAL) {
                     BlockState blockstate1 = level.getBlockState(blockpos.relative(direction));
                     FluidState fluidstate = level.getFluidState(blockpos.relative(direction));
-                    if (fluidstate.is(FluidTags.WATER) || blockstate1.is(Blocks.FROSTED_ICE)) {
+                    if (state.canBeHydrated(level, pos, fluidstate, blockpos.relative(direction)) || blockstate1.is(Blocks.FROSTED_ICE)) {
                         return true;
                     }
                 }
