--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -57,11 +_,14 @@
 
             if (height < 3) {
                 int age = state.getValue(AGE);
+                if (net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, true)) {
                 if (age == 15) {
                     level.setBlockAndUpdate(pos.above(), this.defaultBlockState());
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos.above(), this.defaultBlockState());
                     level.setBlock(pos, state.setValue(AGE, 0), 260);
                 } else {
                     level.setBlock(pos, state.setValue(AGE, age + 1), 260);
+                }
                 }
             }
         }
