--- a/net/minecraft/world/level/block/TntBlock.java
+++ b/net/minecraft/world/level/block/TntBlock.java
@@ -43,7 +_,7 @@
     @Override
     protected void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston) {
         if (!oldState.is(state.getBlock())) {
-            if (level.hasNeighborSignal(pos) && prime(level, pos)) {
+            if (level.hasNeighborSignal(pos) && onCaughtFire(state, level, pos, null, null)) {
                 level.removeBlock(pos, false);
             }
         }
@@ -51,7 +_,7 @@
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, @Nullable Orientation orientation, boolean movedByPiston) {
-        if (level.hasNeighborSignal(pos) && prime(level, pos)) {
+        if (level.hasNeighborSignal(pos) && onCaughtFire(state, level, pos, null, null)) {
             level.removeBlock(pos, false);
         }
     }
@@ -59,7 +_,7 @@
     @Override
     public BlockState playerWillDestroy(Level level, BlockPos pos, BlockState state, Player player) {
         if (!level.isClientSide() && !player.getAbilities().instabuild && state.getValue(UNSTABLE)) {
-            prime(level, pos);
+            onCaughtFire(state, level, pos, null, null);
         }
 
         return super.playerWillDestroy(level, pos, state, player);
@@ -102,7 +_,7 @@
             return super.useItemOn(itemStack, state, level, pos, player, hand, hitResult);
         }
 
-        if (prime(level, pos, player, itemStack)) {
+        if (onCaughtFire(state, level, pos, hitResult.getDirection(), player)) {
             level.setBlock(pos, Blocks.AIR.defaultBlockState(), 11);
             Item item = itemStack.getItem();
             if (itemStack.is(Items.FLINT_AND_STEEL)) {
@@ -146,5 +_,10 @@
     @Override
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(UNSTABLE);
+    }
+
+    @Override
+    public boolean onCaughtFire(BlockState state, Level world, BlockPos pos, net.minecraft.core.@Nullable Direction face, @Nullable LivingEntity igniter) {
+        return prime(world, pos, igniter);
     }
 }
