--- a/net/minecraft/world/level/block/TripWireBlock.java
+++ b/net/minecraft/world/level/block/TripWireBlock.java
@@ -104,7 +_,7 @@
 
     @Override
     public BlockState playerWillDestroy(Level level, BlockPos pos, BlockState state, Player player) {
-        if (!level.isClientSide && !player.getMainHandItem().isEmpty() && player.getMainHandItem().is(Items.SHEARS)) {
+        if (!level.isClientSide && !player.getMainHandItem().isEmpty() && player.getMainHandItem().canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_DISARM)) {
             level.setBlock(pos, state.setValue(DISARMED, Boolean.valueOf(true)), 4);
             level.gameEvent(player, GameEvent.SHEAR, pos);
         }
