--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -166,7 +_,7 @@
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (level.getGameRules().get(GameRules.SPREAD_VINES)) {
-            if (random.nextInt(4) == 0) {
+            if (level.getRandom().nextInt(4) == 0 && level.isAreaLoaded(pos, 4)) { // Forge: check area to prevent loading unloaded chunks
                 Direction testDirection = Direction.getRandom(random);
                 BlockPos abovePos = pos.above();
                 if (testDirection.getAxis().isHorizontal() && !state.getValue(getPropertyForFace(testDirection))) {
