--- a/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
+++ b/net/minecraft/world/level/block/WeatheringCopperStairBlock.java
@@ -13,7 +_,7 @@
     public static final MapCodec<WeatheringCopperStairBlock> CODEC = RecordCodecBuilder.mapCodec(
         p_308852_ -> p_308852_.group(
                     WeatheringCopper.WeatherState.CODEC.fieldOf("weathering_state").forGetter(ChangeOverTimeBlock::getAge),
-                    BlockState.CODEC.fieldOf("base_state").forGetter(p_304556_ -> p_304556_.baseState),
+                    BlockState.CODEC.fieldOf("base_state").forGetter(StairBlock::getModelState),
                     propertiesCodec()
                 )
                 .apply(p_308852_, WeatheringCopperStairBlock::new)
