--- a/net/minecraft/world/level/block/WeatheringCopper.java
+++ b/net/minecraft/world/level/block/WeatheringCopper.java
@@ -18,6 +_,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public interface WeatheringCopper extends ChangeOverTimeBlock<WeatheringCopper.WeatherState> {
+    /**
+     * @deprecated Neo: Use {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#OXIDIZABLES the data map}, this field will be ignored in a future version
+     */
+    @Deprecated
     Supplier<BiMap<Block, Block>> NEXT_BY_BLOCK = Suppliers.memoize(
         () -> {
             Builder<Block, Block> builder = ImmutableBiMap.builder();
@@ -42,16 +_,20 @@
             return builder.build();
         }
     );
+    /**
+     * @deprecated Neo: Use the {@link net.neoforged.neoforge.common.DataMapHooks#INVERSE_OXIDIZABLES_DATAMAP inverse map} generated from the data map, this field will be ignored in a future version
+     */
+    @Deprecated
     Supplier<BiMap<Block, Block>> PREVIOUS_BY_BLOCK = Suppliers.memoize(() -> NEXT_BY_BLOCK.get().inverse());
 
     static Optional<Block> getPrevious(Block block) {
-        return Optional.ofNullable(PREVIOUS_BY_BLOCK.get().get(block));
+        return Optional.ofNullable(net.neoforged.neoforge.common.DataMapHooks.getPreviousOxidizedStage(block));
     }
 
     static Block getFirst(Block block) {
         Block candiate = block;
 
-        for (Block previous = PREVIOUS_BY_BLOCK.get().get(candiate); previous != null; previous = PREVIOUS_BY_BLOCK.get().get(candiate)) {
+        for (Block previous = net.neoforged.neoforge.common.DataMapHooks.getPreviousOxidizedStage(block); previous != null; previous = net.neoforged.neoforge.common.DataMapHooks.getPreviousOxidizedStage(previous)) {
             candiate = previous;
         }
 
@@ -63,7 +_,7 @@
     }
 
     static Optional<Block> getNext(Block block) {
-        return Optional.ofNullable(NEXT_BY_BLOCK.get().get(block));
+        return Optional.ofNullable(net.neoforged.neoforge.common.DataMapHooks.getNextOxidizedStage(block));
     }
 
     static BlockState getFirst(BlockState state) {
