--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -497,8 +499,8 @@
                         windowCandidate.handle(),
                         new GpuDebugOptions(
                             this.options.glDebugVerbosity,
-                            SharedConstants.DEBUG_SYNCHRONOUS_GL_LOGS,
-                            gameConfig.game.renderDebugLabels,
+                            SharedConstants.DEBUG_SYNCHRONOUS_GL_LOGS || enableGlDebug,
+                            gameConfig.game.renderDebugLabels || enableGlDebug,
                             gameConfig.game.vulkanValidation
                         )
                     );
@@ -648,6 +655,7 @@
         this.resourceManager.registerReloadListener(particleResources);
         this.particleEngine = new ParticleEngine(this.level, particleResources);
         particleResources.onReload(this.particleEngine::clearParticles);
+        net.neoforged.neoforge.client.ClientHooks.onRegisterParticleProviders(particleResources);
         this.gameRenderer = new GameRenderer(this, this.entityRenderDispatcher.getItemInHandRenderer(), this.modelManager);
         this.resourceManager.registerReloadListener(this.gameRenderer.createReloadListener());
         WindowRenderState windowRenderState = this.gameRenderer.gameRenderState().windowRenderState;
@@ -696,6 +704,7 @@
             MessageBox.error(message.toString());
         }
 
+        net.neoforged.neoforge.client.ClientHooks.initClientHooks(this, this.resourceManager);
         this.window.setAllowCursorChanges(this.options.allowCursorChanges().get());
         this.window.setQuitShortcuts(this.options.quitShortcuts().get());
         this.resizeGui();
