--- a/net/minecraft/client/MouseHandler.java
+++ b/net/minecraft/client/MouseHandler.java
@@ -76,6 +76,8 @@
                 this.activeButton = null;
             }
 
+            if (net.neoforged.neoforge.client.ClientHooks.onMouseButtonPre(rawButtonInfo, action)) return;
+            boolean screenHandled = false;
             if (this.minecraft.gui.overlay() == null) {
                 if (pressed
                     && this.minecraft.handleGlobalKeyPress(InputConstants.Type.MOUSE.getOrCreate(buttonInfo.button()), buttonInfo.hasControlDownWithQuirk())) {
