--- a/net/minecraft/client/gui/screens/inventory/tooltip/TooltipRenderUtil.java
+++ b/net/minecraft/client/gui/screens/inventory/tooltip/TooltipRenderUtil.java
@@ -17,16 +_,22 @@
     private static final int BORDER_COLOR_BOTTOM = 1344798847;
 
     public static void renderTooltipBackground(GuiGraphics guiGraphics, int x, int y, int width, int height, int z) {
+        renderTooltipBackground(guiGraphics, x, y, width, height, z, BACKGROUND_COLOR, BACKGROUND_COLOR, BORDER_COLOR_TOP, BORDER_COLOR_BOTTOM);
+    }
+
+    // Forge: Allow specifying colors for the inner border gradient and a gradient instead of a single color for the background and outer border
+    public static void renderTooltipBackground(GuiGraphics guiGraphics, int x, int y, int width, int height, int z, int backgroundTop, int backgroundBottom, int borderTop, int borderBottom)
+    {
         int i = x - 3;
         int j = y - 3;
         int k = width + 3 + 3;
         int l = height + 3 + 3;
-        renderHorizontalLine(guiGraphics, i, j - 1, k, z, -267386864);
-        renderHorizontalLine(guiGraphics, i, j + l, k, z, -267386864);
-        renderRectangle(guiGraphics, i, j, k, l, z, -267386864);
-        renderVerticalLine(guiGraphics, i - 1, j, l, z, -267386864);
-        renderVerticalLine(guiGraphics, i + k, j, l, z, -267386864);
-        renderFrameGradient(guiGraphics, i, j + 1, k, l, z, 1347420415, 1344798847);
+        renderHorizontalLine(guiGraphics, i, j - 1, k, z, backgroundTop);
+        renderHorizontalLine(guiGraphics, i, j + l, k, z, backgroundBottom);
+        renderRectangle(guiGraphics, i, j, k, l, z, backgroundTop, backgroundBottom);
+        renderVerticalLineGradient(guiGraphics, i - 1, j, l, z, backgroundTop, backgroundBottom);
+        renderVerticalLineGradient(guiGraphics, i + k, j, l, z, backgroundTop, backgroundBottom);
+        renderFrameGradient(guiGraphics, i, j + 1, k, l, z, borderTop, borderBottom);
     }
 
     private static void renderFrameGradient(
@@ -52,7 +_,16 @@
         guiGraphics.fill(x, y, x + length, y + 1, z, color);
     }
 
+    /**
+    * @deprecated Forge: Use gradient overload instead
+    */
+    @Deprecated
     private static void renderRectangle(GuiGraphics guiGraphics, int x, int y, int width, int height, int z, int color) {
-        guiGraphics.fill(x, y, x + width, y + height, z, color);
+        renderRectangle(guiGraphics, x, y, width, height, z, color, color);
+    }
+
+    // Forge: Allow specifying a gradient instead of a single color for the background
+    private static void renderRectangle(GuiGraphics guiGraphics, int x, int y, int width, int height, int z, int color, int colorTo) {
+        guiGraphics.fillGradient(x, y, x + width, y + height, z, color, colorTo);
     }
 }
