--- a/net/minecraft/server/network/CommonListenerCookie.java
+++ b/net/minecraft/server/network/CommonListenerCookie.java
@@ -3,8 +_,8 @@
 import com.mojang.authlib.GameProfile;
 import net.minecraft.server.level.ClientInformation;
 
-public record CommonListenerCookie(GameProfile gameProfile, int latency, ClientInformation clientInformation) {
+public record CommonListenerCookie(GameProfile gameProfile, int latency, ClientInformation clientInformation, boolean isModdedConnection) {
     public static CommonListenerCookie createInitial(GameProfile p_302024_) {
-        return new CommonListenerCookie(p_302024_, 0, ClientInformation.createDefault());
+        return new CommonListenerCookie(p_302024_, 0, ClientInformation.createDefault(), false);
     }
 }
