--- a/net/minecraft/server/network/PlayerChunkSender.java
+++ b/net/minecraft/server/network/PlayerChunkSender.java
@@ -74,9 +_,12 @@
     }
 
     private static void sendChunk(ServerGamePacketListenerImpl p_295237_, ServerLevel p_294963_, LevelChunk p_295144_) {
-        p_295237_.send(new ClientboundLevelChunkWithLightPacket(p_295144_, p_294963_.getLightEngine(), null, null));
+        p_295237_.send(p_295144_.getAuxLightManager(p_295144_.getPos()).sendLightDataTo(
+                new ClientboundLevelChunkWithLightPacket(p_295144_, p_294963_.getLightEngine(), null, null)
+        ));
         ChunkPos chunkpos = p_295144_.getPos();
         DebugPackets.sendPoiPacketsForChunk(p_294963_, chunkpos);
+        net.neoforged.neoforge.common.CommonHooks.fireChunkSent(p_295237_.player, p_295144_, p_294963_);
     }
 
     private List<LevelChunk> collectChunksToSend(ChunkMap p_296053_, ChunkPos p_295659_) {
