--- a/net/minecraft/server/network/PlayerChunkSender.java
+++ b/net/minecraft/server/network/PlayerChunkSender.java
@@ -74,9 +_,12 @@
     }
 
     private static void sendChunk(ServerGamePacketListenerImpl packetListener, ServerLevel level, LevelChunk chunk) {
-        packetListener.send(new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null));
+        packetListener.send(chunk.getAuxLightManager(chunk.getPos()).sendLightDataTo(
+                new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null)
+        ));
         ChunkPos chunkpos = chunk.getPos();
         DebugPackets.sendPoiPacketsForChunk(level, chunkpos);
+        net.neoforged.neoforge.event.EventHooks.fireChunkSent(packetListener.player, chunk, level);
     }
 
     private List<LevelChunk> collectChunksToSend(ChunkMap chunkMap, ChunkPos chunkPos) {
