--- a/net/minecraft/server/network/PlayerChunkSender.java
+++ b/net/minecraft/server/network/PlayerChunkSender.java
@@ -74,13 +_,16 @@
     }
 
     private static void sendChunk(ServerGamePacketListenerImpl connection, ServerLevel level, LevelChunk chunk) {
-        connection.send(new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null));
+        connection.send(chunk.getAuxLightManager(chunk.getPos()).sendLightDataTo(
+                new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null)
+        ));
         ChunkPos pos = chunk.getPos();
         if (SharedConstants.DEBUG_VERBOSE_SERVER_EVENTS) {
             LOGGER.debug("SEN {}", pos);
         }
 
         level.debugSynchronizers().startTrackingChunk(connection.player, chunk.getPos());
+        net.neoforged.neoforge.event.EventHooks.fireChunkSent(connection.player, chunk, level);
     }
 
     private List<LevelChunk> collectChunksToSend(ChunkMap chunkMap, ChunkPos playerPos) {
