--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -105,11 +_,22 @@
         this.connection.flushChannel();
     }
 
+    @Override
     public void send(Packet<?> p_294278_) {
         this.send(p_294278_, null);
     }
+    
+    @Override
+    public void send(net.minecraft.network.protocol.common.custom.CustomPacketPayload packetPayload) {
+        this.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(packetPayload));
+    }
 
+    @Override
     public void send(Packet<?> p_295099_, @Nullable PacketSendListener p_296321_) {
+        if (!net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().canSendPacket(p_295099_, this)) {
+            return;
+        }
+        
         boolean flag = !this.suspendFlushingOnServerThread || !this.server.isSameThread();
 
         try {
@@ -121,6 +_,11 @@
             throw new ReportedException(crashreport);
         }
     }
+    
+    @Override
+    public void send(net.minecraft.network.protocol.common.custom.CustomPacketPayload packetPayload, @Nullable PacketSendListener listener) {
+        this.send(new net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket(packetPayload), listener);
+    }
 
     public void disconnect(Component p_294116_) {
         this.connection.send(new ClientboundDisconnectPacket(p_294116_), PacketSendListener.thenRun(() -> this.connection.disconnect(p_294116_)));
@@ -142,8 +_,46 @@
     public int latency() {
         return this.latency;
     }
-
+    
+    /**
+     * Creates a new cookie for this connection.
+     *
+     * @param p_301973_ The client information.
+     * @return The cookie.
+     * @deprecated Use {@link #createCookie(ClientInformation, boolean)} instead, keeping the connection type information available.
+     */
+    @Deprecated
     protected CommonListenerCookie createCookie(ClientInformation p_301973_) {
-        return new CommonListenerCookie(this.playerProfile(), this.latency, p_301973_);
+        return new CommonListenerCookie(this.playerProfile(), this.latency, p_301973_, false);
+    }
+    
+    /**
+     * Creates a new cookie for this connection.
+     *
+     * @param p_301973_ The client information.
+     * @param isModdedConnection Whether the connection is modded.
+     * @return The cookie.
+     */
+    protected CommonListenerCookie createCookie(ClientInformation p_301973_, boolean isModdedConnection) {
+        return new CommonListenerCookie(this.playerProfile(), this.latency, p_301973_, isModdedConnection);
+    }
+    
+    @Override
+    public Connection getConnection() {
+        return connection;
+    }
+    @Override
+    public net.minecraft.util.thread.ReentrantBlockableEventLoop<?> getMainThreadEventLoop() {
+        return server;
+    }
+    
+    @Override
+    public boolean isVanillaConnection() {
+        return net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().isVanillaConnection(getConnection());
+    }
+    
+    @Override
+    public boolean isConnected(net.minecraft.resources.ResourceLocation payloadId) {
+        return net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().isConnected(this, payloadId);
     }
 }
