--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -38,6 +_,7 @@
 
 public class ServerConnectionListener {
     private static final Logger LOGGER = LogUtils.getLogger();
+    private static final int READ_TIMEOUT = Integer.parseInt(System.getProperty("neoforge.readTimeout", "30"));
     final MinecraftServer server;
     public volatile boolean running;
     private final List<ChannelFuture> channels = Collections.synchronizedList(Lists.newArrayList());
@@ -49,6 +_,8 @@
     }
 
     public void startTcpServerListener(@Nullable InetAddress p_9712_, int p_9713_) throws IOException {
+        if (p_9712_ == null) p_9712_ = new java.net.InetSocketAddress(p_9713_).getAddress();
+        net.neoforged.neoforge.network.DualStackUtils.checkIPv6(p_9712_);
         synchronized (this.channels) {
             EventLoopGroupHolder eventloopgroupholder = EventLoopGroupHolder.remote(this.server.useNativeTransport());
             this.channels.add(new ServerBootstrap().channel(eventloopgroupholder.serverChannelCls()).childHandler(new ChannelInitializer<Channel>() {
@@ -59,7 +_,7 @@
                     } catch (ChannelException channelexception) {
                     }
 
-                    ChannelPipeline channelpipeline = p_9729_.pipeline().addLast("timeout", new ReadTimeoutHandler(30));
+                    ChannelPipeline channelpipeline = p_9729_.pipeline().addLast("timeout", new ReadTimeoutHandler(READ_TIMEOUT));
                     if (ServerConnectionListener.this.server.repliesToStatus()) {
                         channelpipeline.addLast("legacy_query", new LegacyQueryHandler(ServerConnectionListener.this.getServer()));
                     }
@@ -71,7 +_,7 @@
                     connection.configurePacketHandler(channelpipeline);
                     connection.setListenerForServerboundHandshake(new ServerHandshakePacketListenerImpl(ServerConnectionListener.this.server, connection));
                 }
-            }).group(eventloopgroupholder.eventLoopGroup()).localAddress(p_9712_, p_9713_).bind().syncUninterruptibly());
+            }).group(eventloopgroupholder.serverEventLoopGroup()).localAddress(p_9712_, p_9713_).bind().syncUninterruptibly());
         }
     }
 
@@ -102,7 +_,7 @@
                         }
                     }
                 )
-                .group(EventLoopGroupHolder.local().eventLoopGroup())
+                .group(EventLoopGroupHolder.local().serverEventLoopGroup())
                 .localAddress(LocalAddress.ANY)
                 .bind()
                 .syncUninterruptibly();
