--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -490,18 +_,21 @@
 
                 if (fail && level.noCollision(vehicle, oldAABB) || this.isEntityCollidingWithAnythingNew(level, vehicle, oldAABB, targetX, targetY, targetZ)) {
                     vehicle.absSnapTo(oldX, oldY, oldZ, targetYRot, targetXRot);
+                    resyncPlayerWithVehicle(vehicle); // Neo - Resync player position on vehicle moving
                     this.send(ClientboundMoveVehiclePacket.fromEntity(vehicle));
                     vehicle.removeLatestMovementRecording();
                     return;
                 }
 
                 vehicle.absSnapTo(targetX, targetY, targetZ, targetYRot, targetXRot);
+                resyncPlayerWithVehicle(vehicle); // Neo - Resync player position on vehicle moving
                 this.player.level().getChunkSource().move(this.player);
                 Vec3 clientDeltaMovement = new Vec3(vehicle.getX() - oldX, vehicle.getY() - oldY, vehicle.getZ() - oldZ);
                 this.handlePlayerKnownMovement(clientDeltaMovement);
                 vehicle.setOnGroundWithMovement(packet.onGround(), clientDeltaMovement);
                 vehicle.doCheckFallDamage(clientDeltaMovement.x, clientDeltaMovement.y, clientDeltaMovement.z, packet.onGround());
                 this.player.checkMovementStatistics(clientDeltaMovement.x, clientDeltaMovement.y, clientDeltaMovement.z);
+                this.player.checkRidingStatistics(clientDeltaMovement.x, clientDeltaMovement.y, clientDeltaMovement.z); // Neo: check riding stats too as vanilla checks them in rideTick based on the assumption that Entity#rideTick will move the entity, which we break
                 this.clientVehicleIsFloating = oyDist >= -0.03125
                     && !vehicleRestsOnSomething
                     && !this.server.allowFlight()
@@ -515,6 +_,23 @@
         }
     }
 
+    private void resyncPlayerWithVehicle(Entity vehicle) {
+        Vec3 oldPos = this.player.position();
+        float yRot = this.player.getYRot();
+        float xRot = this.player.getXRot();
+        float yHeadRot = this.player.getYHeadRot();
+
+        vehicle.positionRider(this.player);
+
+        // preserve old rotation and store old position in xo/yo/zo
+        this.player.setYRot(yRot);
+        this.player.setXRot(xRot);
+        this.player.setYHeadRot(yHeadRot);
+        this.player.xo = oldPos.x;
+        this.player.yo = oldPos.y;
+        this.player.zo = oldPos.z;
+    }
+
     private boolean noBlocksAround(Entity entity) {
         return entity.level()
             .getBlockStates(entity.getBoundingBox().inflate(0.0625).expandTowards(0.0, -0.55, 0.0))
@@ -698,7 +_,7 @@
             if (level.isLoaded(pos)) {
                 BlockState blockState = level.getBlockState(pos);
                 boolean includeData = this.player.hasInfiniteMaterials() && packet.includeData();
-                ItemStack itemStack = blockState.getCloneItemStack(level, pos, includeData);
+                ItemStack itemStack = blockState.getCloneItemStack(pos, level, includeData, player);
                 if (!itemStack.isEmpty()) {
                     if (includeData) {
                         addBlockDataToItem(blockState, level, pos, itemStack);
@@ -1151,7 +_,7 @@
                                         && !playerStandsOnSomething
                                         && !this.player.isSpectator()
                                         && !this.server.allowFlight()
-                                        && !this.player.getAbilities().mayfly
+                                        && !this.player.mayFly()
                                         && !this.player.hasEffect(MobEffects.LEVITATION)
                                         && !isFallFlying
                                         && !isAutoSpinAttack
@@ -1280,9 +_,10 @@
                     }
                 case SWAP_ITEM_WITH_OFFHAND:
                     if (!this.player.isSpectator()) {
-                        ItemStack swap = this.player.getItemInHand(InteractionHand.OFF_HAND);
-                        this.player.setItemInHand(InteractionHand.OFF_HAND, this.player.getItemInHand(InteractionHand.MAIN_HAND));
-                        this.player.setItemInHand(InteractionHand.MAIN_HAND, swap);
+                        var event = net.neoforged.neoforge.common.CommonHooks.onLivingSwapHandItems(this.player);
+                        if (event.isCanceled()) return;
+                        this.player.setItemInHand(InteractionHand.OFF_HAND, event.getItemSwappedToOffHand());
+                        this.player.setItemInHand(InteractionHand.MAIN_HAND, event.getItemSwappedToMainHand());
                         this.player.stopUsingItem();
                     }
 
@@ -1306,7 +_,7 @@
                 case ABORT_DESTROY_BLOCK:
                 case STOP_DESTROY_BLOCK:
                     this.player.gameMode.handleBlockBreakAction(pos, action, packet.getDirection(), this.player.level().getMaxY(), packet.getSequence());
-                    this.ackBlockChangesUpTo(packet.getSequence());
+                    this.player.connection.ackBlockChangesUpTo = packet.getSequence();
                     return;
                 default:
                     throw new IllegalArgumentException("Invalid player action");
@@ -1328,7 +_,7 @@
     public void handleUseItemOn(ServerboundUseItemOnPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         if (this.hasClientLoaded()) {
-            this.ackBlockChangesUpTo(packet.getSequence());
+            this.player.connection.ackBlockChangesUpTo = packet.getSequence();
             ServerLevel level = this.player.level();
             InteractionHand hand = packet.getHand();
             ItemStack itemStack = this.player.getItemInHand(hand);
@@ -1480,8 +_,9 @@
                 }
 
                 CompletableFuture<FilteredText> filteredFuture = this.filterTextPacket(signedMessage.signedContent());
-                Component decorated = this.server.getChatDecorator().decorate(this.player, signedMessage.decoratedContent());
+                Component decorated = net.neoforged.neoforge.common.CommonHooks.getServerChatSubmittedDecorator().decorate(this.player, signedMessage.decoratedContent());
                 this.chatMessageChain.append(filteredFuture, filtered -> {
+                    if (decorated == null) return; // Forge: ServerChatEvent was canceled if this is null.
                     PlayerChatMessage filteredMessage = signedMessage.withUnsignedContent(decorated).filter(filtered.mask());
                     this.broadcastChatMessage(filteredMessage);
                 });
@@ -1812,7 +_,11 @@
                     ItemStack tool = this.player.getItemInHand(hand);
                     if (tool.isItemEnabled(level.enabledFeatures())) {
                         ItemStack usedItemStack = tool.copy();
-                        if (this.player.interactOn(target, hand, location) instanceof InteractionResult.Success success) {
+                        var interactionResult = net.neoforged.neoforge.common.CommonHooks.onInteractEntityAt(player, target, location, hand);
+                        if (interactionResult == null) {
+                            interactionResult = this.player.interactOn(target, hand, location);
+                        }
+                        if (interactionResult instanceof InteractionResult.Success success) {
                             ItemStack awardedForStack = success.wasItemInteraction() ? usedItemStack : ItemStack.EMPTY;
                             CriteriaTriggers.PLAYER_INTERACTED_WITH_ENTITY.trigger(this.player, awardedForStack, target);
                             if (success.swingSource() == InteractionResult.SwingSource.SERVER) {
@@ -2033,14 +_,16 @@
     @Override
     public void handlePlayerAbilities(ServerboundPlayerAbilitiesPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        this.player.getAbilities().flying = packet.isFlying() && this.player.getAbilities().mayfly;
+        this.player.getAbilities().flying = packet.isFlying() && this.player.mayFly();
     }
 
     @Override
     public void handleClientInformation(ServerboundClientInformationPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         boolean wasHatShown = this.player.isModelPartShown(PlayerModelPart.HAT);
+        net.minecraft.server.level.ClientInformation oldInfo = this.player.clientInformation();
         this.player.updateOptions(packet.information());
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.ClientInformationUpdatedEvent(this.player, oldInfo, packet.information()));
         if (this.player.isModelPartShown(PlayerModelPart.HAT) != wasHatShown) {
             this.server.getPlayerList().broadcastAll(new ClientboundPlayerInfoUpdatePacket(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_HAT, this.player));
         }
@@ -2116,7 +_,7 @@
             this.connection
                 .setupInboundProtocol(
                     ConfigurationProtocols.SERVERBOUND,
-                    new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation()))
+                    new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation(), this.connectionType))
                 );
         }
     }
@@ -2151,6 +_,7 @@
 
     @Override
     public void handleCustomPayload(ServerboundCustomPayloadPacket packet) {
+        super.handleCustomPayload(packet); // Neo: Call super to invoke modded payload handling.
     }
 
     @Override
