--- a/net/minecraft/server/network/ServerHibernateConfigPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerHibernateConfigPacketListenerImpl.java
@@ -78,4 +_,24 @@
     public ServerConfigurationPacketListenerImpl unfreeze(TheGame p_420243_) {
         return new ServerConfigurationPacketListenerImpl(p_420243_, this.connection, this.createCookie(this.clientInformation));
     }
+
+    @Override
+    public Connection getConnection() {
+        return null;
+    }
+
+    @Override
+    public net.minecraft.util.thread.ReentrantBlockableEventLoop<?> getMainThreadEventLoop() {
+        return null;
+    }
+
+    @Override
+    public net.neoforged.neoforge.network.connection.ConnectionType getConnectionType() {
+        return net.neoforged.neoforge.network.connection.ConnectionType.NEOFORGE;
+    }
+
+    @Override
+    public void finishCurrentTask(ConfigurationTask.Type task) {
+
+    }
 }
