--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -41,6 +_,8 @@
     @Nullable
     private ConfigurationTask currentTask;
     private ClientInformation clientInformation;
+    private boolean isModdedConnection = false;
+    private boolean isHandlingModdedConfigurationPhase = false;
 
     public ServerConfigurationPacketListenerImpl(MinecraftServer p_294645_, Connection p_295787_, CommonListenerCookie p_302003_) {
         super(p_294645_, p_295787_, p_302003_);
@@ -65,6 +_,11 @@
     }
 
     public void startConfiguration() {
+        this.send(new net.neoforged.neoforge.network.payload.ModdedNetworkQueryPayload());
+        this.send(new net.minecraft.network.protocol.common.ClientboundPingPacket(0));
+    }
+
+    private void runConfiguration() {
         this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
         LayeredRegistryAccess<RegistryLayer> layeredregistryaccess = this.server.registries();
         this.send(new ClientboundUpdateEnabledFeaturesPacket(FeatureFlags.REGISTRY.toNames(this.server.getWorldData().enabledFeatures())));
@@ -86,8 +_,45 @@
 
     private void addOptionalTasks() {
         this.server.getServerResourcePack().ifPresent(p_296496_ -> this.configurationTasks.add(new ServerResourcePackConfigurationTask(p_296496_)));
-    }
-
+        
+        this.configurationTasks.add(new net.neoforged.neoforge.network.configuration.ModdedConfigurationPhaseStarted(this));
+        this.configurationTasks.addAll(net.neoforged.fml.ModLoader.get().postEventWithReturn(new net.neoforged.neoforge.network.event.OnGameConfigurationEvent(this)).getConfigurationTasks());
+        this.configurationTasks.add(new net.neoforged.neoforge.network.configuration.ModdedConfigurationPhaseCompleted(this));
+    }
+    
+    @Override
+    public void handleCustomPayload(net.minecraft.network.protocol.common.ServerboundCustomPayloadPacket p_294276_) {
+        if (p_294276_.payload() instanceof net.neoforged.neoforge.network.payload.ModdedNetworkQueryPayload moddedEnvironmentPayload) {
+            this.isModdedConnection = true;
+            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance()
+                    .onModdedConnectionDetectedAtServer(
+                            this,
+                            moddedEnvironmentPayload.configuration(),
+                            moddedEnvironmentPayload.play()
+                    );
+            return;
+        }
+        
+        if (!isHandlingModdedConfigurationPhase) {
+            super.handleCustomPayload(p_294276_);
+            return;
+        }
+
+        net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onModdedPacketAtServer(this, p_294276_);
+    }
+
+    @Override
+    public void handlePong(net.minecraft.network.protocol.common.ServerboundPongPacket p_295142_) {
+        super.handlePong(p_295142_);
+        if (p_295142_.getId() == 0) {
+            if (!this.isModdedConnection && !net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onVanillaConnectionDetectedAtServer(this)) {
+                return;
+            }
+            
+            this.runConfiguration();
+        }
+    }
+    
     @Override
     public void handleClientInformation(ServerboundClientInformationPacket p_302032_) {
         this.clientInformation = p_302032_.information();
@@ -121,7 +_,7 @@
             }
 
             ServerPlayer serverplayer = playerlist.getPlayerForLogin(this.gameProfile, this.clientInformation);
-            playerlist.placeNewPlayer(this.connection, serverplayer, this.createCookie(this.clientInformation));
+            playerlist.placeNewPlayer(this.connection, serverplayer, this.createCookie(this.clientInformation, this.isModdedConnection));
             this.connection.resumeInboundAfterProtocolChange();
         } catch (Exception exception) {
             LOGGER.error("Couldn't place player in world", (Throwable)exception);
@@ -146,8 +_,9 @@
             }
         }
     }
-
-    private void finishCurrentTask(ConfigurationTask.Type p_294853_) {
+    
+    @Override
+    public void finishCurrentTask(ConfigurationTask.Type p_294853_) {
         ConfigurationTask.Type configurationtask$type = this.currentTask != null ? this.currentTask.type() : null;
         if (!p_294853_.equals(configurationtask$type)) {
             throw new IllegalStateException("Unexpected request for task finish, current task: " + configurationtask$type + ", requested: " + p_294853_);
@@ -155,5 +_,20 @@
             this.currentTask = null;
             this.startNextTask();
         }
+    }
+    
+    public void onModdedConfigurationPhaseStarted() {
+        isHandlingModdedConfigurationPhase = true;
+        finishCurrentTask(net.neoforged.neoforge.network.configuration.ModdedConfigurationPhaseStarted.TYPE);
+    }
+    
+    public void onModdedConfigurationPhaseEnded() {
+        isHandlingModdedConfigurationPhase = false;
+        finishCurrentTask(net.neoforged.neoforge.network.configuration.ModdedConfigurationPhaseCompleted.TYPE);
+    }
+    
+    @Override
+    public boolean isVanillaConnection() {
+        return net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().isVanillaConnection(getConnection());
     }
 }
