--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -42,7 +42,7 @@
     public static CompletableFuture<ChunkAccess> generateStructureStarts(WorldGenContext context, ChunkStep step, StaticCache2D<GenerationChunkHolder> chunks, ChunkAccess chunk) {
         ServerLevel serverlevel = context.level();
 
-        if (serverlevel.getServer().getWorldGenSettings().options().generateStructures()) {
+        if (serverlevel.getWorldGenSettings().options().generateStructures()) { // CraftBukkit
             context.generator().createStructures(serverlevel.registryAccess(), serverlevel.getChunkSource().getGeneratorState(), serverlevel.structureManager(), chunk, context.structureManager(), serverlevel.dimension());
         }
 
