--- a/net/minecraft/core/dispenser/FlintAndSteelDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/FlintAndSteelDispenseItemBehavior.java
@@ -19,6 +19,12 @@
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.block.CraftBlock;
+
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
 
 public class FlintAndSteelDispenseItemBehavior extends OptionalDispenseItemBehavior {
 
@@ -28,6 +34,30 @@
     protected ItemStack execute(BlockSource source, ItemStack dispensed) {
         ServerLevel serverlevel = source.level();
 
+        // CraftBukkit start
+        org.bukkit.block.Block bukkitBlock = CraftBlock.at(serverlevel, source.pos());
+        CraftItemStack craftItem = CraftItemStack.asCraftMirror(dispensed);
+
+        BlockDispenseEvent event = new BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+        if (!DispenserBlock.eventFired) {
+            serverlevel.getCraftServer().getPluginManager().callEvent(event);
+        }
+
+        if (event.isCancelled()) {
+            return dispensed;
+        }
+
+        if (!event.getItem().equals(craftItem)) {
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseitembehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+            if (dispenseitembehavior != DispenseItemBehavior.NOOP && dispenseitembehavior != this) {
+                dispenseitembehavior.dispense(source, eventStack);
+                return dispensed;
+            }
+        }
+        // CraftBukkit end
+
         this.setSuccess(true);
         Direction direction = (Direction) source.state().getValue(DispenserBlock.FACING);
         BlockPos blockpos = source.pos().relative(direction);
@@ -35,10 +65,14 @@
 
         if (!tryIgniteExplosiveEntities(serverlevel, blockpos)) {
             if (BaseFireBlock.canBePlacedAt(serverlevel, blockpos, direction)) {
-                serverlevel.setBlockAndUpdate(blockpos, BaseFireBlock.getState(serverlevel, blockpos));
-                serverlevel.gameEvent((Entity) null, (Holder) GameEvent.BLOCK_PLACE, blockpos);
+                // CraftBukkit start - Ignition by dispensing flint and steel
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(serverlevel, blockpos, source.pos()).isCancelled()) {
+                    serverlevel.setBlockAndUpdate(blockpos, BaseFireBlock.getState(serverlevel, blockpos));
+                    serverlevel.gameEvent((Entity) null, (Holder) GameEvent.BLOCK_PLACE, blockpos);
+                }
+                // CraftBukkit end
             } else if (!CampfireBlock.canLight(blockstate) && !CandleBlock.canLight(blockstate) && !CandleCakeBlock.canLight(blockstate)) {
-                if (blockstate.getBlock() instanceof TntBlock) {
+                if (blockstate.getBlock() instanceof TntBlock && org.bukkit.craftbukkit.event.CraftEventFactory.callTNTPrimeEvent(serverlevel, blockpos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.DISPENSER, null, source.pos())) { // CraftBukkit - TNTPrimeEvent
                     if (TntBlock.prime(serverlevel, blockpos)) {
                         serverlevel.removeBlock(blockpos, false);
                     } else {
