--- a/net/minecraft/core/dispenser/MinecartDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/MinecartDispenseItemBehavior.java
@@ -16,6 +16,12 @@
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.phys.Vec3;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
+// CraftBukkit end
+
 public class MinecartDispenseItemBehavior extends DefaultDispenseItemBehavior {
 
     private final DefaultDispenseItemBehavior defaultDispenseItemBehavior = new DefaultDispenseItemBehavior();
@@ -62,11 +68,38 @@
         }
 
         Vec3 vec31 = new Vec3(d0, d1 + d3, d2);
-        AbstractMinecart abstractminecart = AbstractMinecart.createMinecart(serverlevel, vec31.x, vec31.y, vec31.z, this.entityType, EntitySpawnReason.DISPENSER, dispensed, (Player) null);
+        // CraftBukkit start
+        // EntityMinecartAbstract entityminecartabstract = EntityMinecartAbstract.createMinecart(worldserver, vec3d1.x, vec3d1.y, vec3d1.z, this.entityType, EntitySpawnReason.DISPENSER, itemstack, (EntityHuman) null);
+        ItemStack itemstack1 = dispensed.split(1);
+        org.bukkit.block.Block block2 = CraftBlock.at(serverlevel, source.pos());
+        CraftItemStack craftItem = CraftItemStack.asCraftMirror(itemstack1);
+
+        BlockDispenseEvent event = new BlockDispenseEvent(block2, craftItem.clone(), new org.bukkit.util.Vector(vec31.x, vec31.y, vec31.z));
+        if (!DispenserBlock.eventFired) {
+            serverlevel.getCraftServer().getPluginManager().callEvent(event);
+        }
+
+        if (event.isCancelled()) {
+            dispensed.grow(1);
+            return dispensed;
+        }
+
+        if (!event.getItem().equals(craftItem)) {
+            dispensed.grow(1);
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior dispenseitembehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+            if (dispenseitembehavior != DispenseItemBehavior.NOOP && dispenseitembehavior != this) {
+                dispenseitembehavior.dispense(source, eventStack);
+                return dispensed;
+            }
+        }
+        AbstractMinecart abstractminecart = AbstractMinecart.createMinecart(serverlevel, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), this.entityType, EntitySpawnReason.DISPENSER, itemstack1, (Player) null);
 
         if (abstractminecart != null) {
-            serverlevel.addFreshEntity(abstractminecart);
-            dispensed.shrink(1);
+            if (!serverlevel.addFreshEntity(abstractminecart)) dispensed.grow(1);
+            // itemstack.shrink(1); // CraftBukkit - handled during event processing
+            // CraftBukkit end
         }
 
         return dispensed;
