--- a/net/minecraft/client/gui/screens/options/controls/KeyBindsScreen.java
+++ b/net/minecraft/client/gui/screens/options/controls/KeyBindsScreen.java
@@ -25,6 +_,11 @@
     public long lastKeySelection;
     private KeyBindsList keyBindsList;
     private Button resetButton;
+    // Neo: These are to hold the last key and modifier pressed so they can be checked in keyReleased
+    private InputConstants.Key lastPressedKey = InputConstants.UNKNOWN;
+    private InputConstants.Key lastPressedModifier = InputConstants.UNKNOWN;
+    private boolean isLastKeyHeldDown = false;
+    private boolean isLastModifierHeldDown = false;
 
     public KeyBindsScreen(Screen p_344996_, Options p_344771_) {
         super(p_344996_, p_344771_, TITLE);
@@ -43,7 +_,7 @@
     protected void addFooter() {
         this.resetButton = Button.builder(Component.translatable("controls.resetAll"), p_346345_ -> {
             for (KeyMapping keymapping : this.options.keyMappings) {
-                keymapping.setKey(keymapping.getDefaultKey());
+                keymapping.setToDefault();
             }
 
             this.keyBindsList.resetMappingAndUpdateButtons();
@@ -74,18 +_,67 @@
     @Override
     public boolean keyPressed(KeyEvent p_446842_) {
         if (this.selectedKey != null) {
-            if (p_446842_.isEscape()) {
+            var key = InputConstants.getKey(p_446842_);
+            if (this.lastPressedModifier == InputConstants.UNKNOWN && net.neoforged.neoforge.client.settings.KeyModifier.isKeyCodeModifier(key)) {
+                this.lastPressedModifier = key;
+                this.isLastModifierHeldDown = true;
+            } else {
+                this.lastPressedKey = key;
+                this.isLastKeyHeldDown = true;
+            }
+            return true;
+        } else {
+            return super.keyPressed(p_446842_);
+        }
+    }
+
+    @Override
+    public boolean keyReleased(KeyEvent event) {
+        // We ignore events from keys with the scan code 63 as they're emitted
+        // (only as RELEASE, not PRESS) by Mac systems to indicate that "Fn" is being pressed
+        // See https://github.com/neoforged/NeoForge/issues/1683
+        if (this.selectedKey != null && (!net.minecraft.client.input.InputQuirks.ON_OSX || event.scancode() != 63)) {
+            if (event.isEscape()) {
+                this.selectedKey.setKeyModifierAndCode(net.neoforged.neoforge.client.settings.KeyModifier.NONE, InputConstants.UNKNOWN);
                 this.selectedKey.setKey(InputConstants.UNKNOWN);
+                this.lastPressedKey = InputConstants.UNKNOWN;
+                this.lastPressedModifier = InputConstants.UNKNOWN;
+                this.isLastKeyHeldDown = false;
+                this.isLastModifierHeldDown = false;
             } else {
-                this.selectedKey.setKey(InputConstants.getKey(p_446842_));
-            }
+                var key = InputConstants.getKey(event);
+                if (this.lastPressedKey.equals(key)) {
+                    this.isLastKeyHeldDown = false;
+                } else if (this.lastPressedModifier.equals(key)) {
+                    this.isLastModifierHeldDown = false;
+                }
 
+                if (!this.isLastKeyHeldDown && !this.isLastModifierHeldDown) {
+                    if (!this.lastPressedKey.equals(InputConstants.UNKNOWN)) {
+                        this.selectedKey.setKeyModifierAndCode(
+                                net.neoforged.neoforge.client.settings.KeyModifier.getKeyModifier(this.lastPressedModifier),
+                                this.lastPressedKey
+                        );
+                        this.selectedKey.setKey(this.lastPressedKey);
+                    } else {
+                        this.selectedKey.setKeyModifierAndCode(
+                                net.neoforged.neoforge.client.settings.KeyModifier.NONE,
+                                this.lastPressedModifier
+                        );
+                        this.selectedKey.setKey(this.lastPressedModifier);
+                    }
+                    this.lastPressedKey = InputConstants.UNKNOWN;
+                    this.lastPressedModifier = InputConstants.UNKNOWN;
+                } else {
+                    return true;
+                }
+            }
             this.selectedKey = null;
             this.lastKeySelection = Util.getMillis();
             this.keyBindsList.resetMappingAndUpdateButtons();
             return true;
         } else {
-            return super.keyPressed(p_446842_);
+            return super.keyReleased(event);
         }
     }
 
