--- a/net/minecraft/world/damagesource/DeathMessageType.java
+++ b/net/minecraft/world/damagesource/DeathMessageType.java
@@ -3,20 +_,41 @@
 import com.mojang.serialization.Codec;
 import net.minecraft.util.StringRepresentable;
 
-public enum DeathMessageType implements StringRepresentable {
+@net.neoforged.fml.common.asm.enumextension.NamedEnum
+@net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.CLIENTBOUND)
+public enum DeathMessageType implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     DEFAULT("default"),
     FALL_VARIANTS("fall_variants"),
     INTENTIONAL_GAME_DESIGN("intentional_game_design");
 
     public static final Codec<DeathMessageType> CODEC = StringRepresentable.fromEnum(DeathMessageType::values);
     private final String id;
+    private final net.neoforged.neoforge.common.damagesource.IDeathMessageProvider msgFunction;
 
+    @net.neoforged.fml.common.asm.enumextension.ReservedConstructor
     private DeathMessageType(String p_270201_) {
-        this.id = p_270201_;
+        this(p_270201_, net.neoforged.neoforge.common.damagesource.IDeathMessageProvider.DEFAULT);
+    }
+
+    private DeathMessageType(String id, net.neoforged.neoforge.common.damagesource.IDeathMessageProvider msgFunction) {
+        this.id = id;
+        this.msgFunction = msgFunction;
     }
 
     @Override
     public String getSerializedName() {
         return this.id;
+    }
+
+    /**
+     * The death message function is used when an entity dies to a damage type with this death message type.
+     * @return The {@link net.neoforged.neoforge.common.damagesource.IDeathMessageProvider} associated with this death message type.
+     */
+    public net.neoforged.neoforge.common.damagesource.IDeathMessageProvider getMessageFunction() {
+        return this.msgFunction;
+    }
+
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(DeathMessageType.class);
     }
 }
