--- a/net/minecraft/data/registries/RegistriesDatapackGenerator.java
+++ b/net/minecraft/data/registries/RegistriesDatapackGenerator.java
@@ -59,20 +_,22 @@
         CachedOutput cache, HolderLookup.Provider registries, DynamicOps<JsonElement> writeOps, RegistryDataLoader.RegistryData<T> v
     ) {
         ResourceKey<? extends Registry<T>> registryKey = v.key();
+        var conditionalCodec = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodecWithConditions(v.elementCodec());
         return registries.lookup(registryKey)
             .map(
                 registry -> {
                     PackOutput.PathProvider pathProvider = this.output.createRegistryElementsPathProvider(registryKey);
                     return CompletableFuture.allOf(
                         registry.listElements()
-                            .map(e -> dumpValue(pathProvider.json(e.key().identifier()), cache, writeOps, v.elementCodec(), e.value()))
+                            .filter(holder -> this.namespacePredicate.test(holder.key().identifier().getNamespace()))
+                            .map(e -> (CompletableFuture<Object>) dumpValue(pathProvider.json(e.key().identifier()), cache, writeOps, conditionalCodec, Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(conditions.getOrDefault(e.key(), java.util.List.of()), e.value()))))
                             .toArray(CompletableFuture[]::new)
                     );
                 }
             );
     }
 
-    private static <E> CompletableFuture<?> dumpValue(Path path, CachedOutput cache, DynamicOps<JsonElement> ops, Encoder<E> codec, E value) {
+    private static <E> CompletableFuture<?> dumpValue(Path path, CachedOutput cache, DynamicOps<JsonElement> ops, Encoder<java.util.Optional<net.neoforged.neoforge.common.conditions.WithConditions<E>>> codec, java.util.Optional<net.neoforged.neoforge.common.conditions.WithConditions<E>> value) {
         return codec.encodeStart(ops, value)
             .mapOrElse(
                 result -> DataProvider.saveStable(cache, result, path),
