--- a/net/minecraft/data/registries/RegistryPatchGenerator.java
+++ b/net/minecraft/data/registries/RegistryPatchGenerator.java
@@ -20,7 +_,7 @@
             p_311522_ -> {
                 RegistryAccess.Frozen registryaccess$frozen = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
                 Cloner.Factory cloner$factory = new Cloner.Factory();
-                RegistryDataLoader.WORLDGEN_REGISTRIES.forEach(p_311524_ -> p_311524_.runWithArguments(cloner$factory::addCodec));
+                net.neoforged.neoforge.registries.DataPackRegistriesHooks.getDataPackRegistriesWithDimensions().forEach(p_311524_ -> p_311524_.runWithArguments(cloner$factory::addCodec));
                 RegistrySetBuilder.PatchedRegistries registrysetbuilder$patchedregistries = p_309174_.buildPatch(
                     registryaccess$frozen, p_311522_, cloner$factory
                 );
