--- a/net/minecraft/data/registries/VanillaRegistries.java
+++ b/net/minecraft/data/registries/VanillaRegistries.java
@@ -101,6 +_,9 @@
         .add(Registries.VILLAGER_TRADE, VillagerTrades::bootstrap)
         .add(Registries.TRADE_SET, TradeSets::bootstrap);
 
+    // Neo: Keep a set of all actual vanilla registries to allow filtering out registries that are added by mods for Vanilla clients.
+    public static final List<? extends net.minecraft.resources.ResourceKey<? extends net.minecraft.core.Registry<?>>> DATAPACK_REGISTRY_KEYS = BUILDER.getEntryKeys();
+
     private static void validateThatAllBiomeFeaturesHaveBiomeFilter(HolderLookup.Provider provider) {
         validateThatAllBiomeFeaturesHaveBiomeFilter(provider.lookupOrThrow(Registries.PLACED_FEATURE), provider.lookupOrThrow(Registries.BIOME));
     }
