--- a/net/minecraft/data/recipes/RecipeOutput.java
+++ b/net/minecraft/data/recipes/RecipeOutput.java
@@ -10,7 +_,9 @@
 import net.neoforged.neoforge.common.extensions.IRecipeOutputExtension;
 
 public interface RecipeOutput extends BootstrapContextAccess, IRecipeOutputExtension {
-    void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancement);
+    default void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancement) {
+        accept(id, recipe, advancement, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+    }
 
     Advancement.Builder advancement();
 }
