--- a/net/minecraft/data/recipes/RecipeOutput.java
+++ b/net/minecraft/data/recipes/RecipeOutput.java
@@ -6,8 +_,10 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.world.item.crafting.Recipe;
 
-public interface RecipeOutput {
-    void accept(ResourceLocation location, Recipe<?> recipe, @Nullable AdvancementHolder advancement);
+public interface RecipeOutput extends net.neoforged.neoforge.common.extensions.IRecipeOutputExtension {
+    default void accept(ResourceLocation location, Recipe<?> recipe, @Nullable AdvancementHolder advancement) {
+        accept(location, recipe, advancement, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+    }
 
     Advancement.Builder advancement();
 }
