--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -43,11 +_,11 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 
-public abstract class RecipeProvider implements DataProvider {
+public abstract class RecipeProvider implements DataProvider, net.neoforged.neoforge.common.conditions.IConditionBuilder {
     protected final PackOutput.PathProvider recipePathProvider;
     protected final PackOutput.PathProvider advancementPathProvider;
+    protected final CompletableFuture<net.minecraft.core.HolderLookup.Provider> lookupProvider;
     private static final Map<BlockFamily.Variant, BiFunction<ItemLike, ItemLike, RecipeBuilder>> SHAPE_BUILDERS = ImmutableMap.<BlockFamily.Variant, BiFunction<ItemLike, ItemLike, RecipeBuilder>>builder(
-            
         )
         .put(BlockFamily.Variant.BUTTON, (p_176733_, p_176734_) -> buttonBuilder(p_176733_, Ingredient.of(p_176734_)))
         .put(BlockFamily.Variant.CHISELED, (p_248037_, p_248038_) -> chiseledBuilder(RecipeCategory.BUILDING_BLOCKS, p_248037_, Ingredient.of(p_248038_)))
@@ -66,27 +_,29 @@
         .put(BlockFamily.Variant.WALL, (p_248024_, p_248025_) -> wallBuilder(RecipeCategory.DECORATIONS, p_248024_, Ingredient.of(p_248025_)))
         .build();
 
-    public RecipeProvider(PackOutput p_248933_) {
+    public RecipeProvider(PackOutput p_248933_, CompletableFuture<net.minecraft.core.HolderLookup.Provider> lookupProvider) {
         this.recipePathProvider = p_248933_.createPathProvider(PackOutput.Target.DATA_PACK, "recipes");
         this.advancementPathProvider = p_248933_.createPathProvider(PackOutput.Target.DATA_PACK, "advancements");
+        this.lookupProvider = lookupProvider;
     }
 
     @Override
     public CompletableFuture<?> run(final CachedOutput p_254020_) {
+        return this.lookupProvider.thenCompose(provider -> {
         final Set<ResourceLocation> set = Sets.newHashSet();
         final List<CompletableFuture<?>> list = new ArrayList<>();
         this.buildRecipes(
             new RecipeOutput() {
                 @Override
-                public void accept(ResourceLocation p_312039_, Recipe<?> p_312254_, @Nullable AdvancementHolder p_311794_) {
+                public void accept(ResourceLocation p_312039_, Recipe<?> p_312254_, @org.jetbrains.annotations.Nullable AdvancementHolder p_311794_, List<net.neoforged.neoforge.common.conditions.ICondition> conditions, List<net.neoforged.neoforge.common.conditions.WithConditions<com.mojang.datafixers.util.Pair<Recipe<?>, Advancement>>> alternatives) {
                     if (!set.add(p_312039_)) {
                         throw new IllegalStateException("Duplicate recipe " + p_312039_);
                     } else {
-                        list.add(DataProvider.saveStable(p_254020_, Recipe.CODEC, p_312254_, RecipeProvider.this.recipePathProvider.json(p_312039_)));
+                        list.add(DataProvider.saveStable(p_254020_, net.neoforged.neoforge.common.util.NeoForgeExtraCodecs.CONDITIONAL_RECIPE_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.ConditionalObject<>(conditions, p_312254_, alternatives.stream().<net.neoforged.neoforge.common.conditions.WithConditions<Recipe<?>>>map(w -> new net.neoforged.neoforge.common.conditions.WithConditions<>(w.conditions(), w.carrier().getFirst())).toList())), RecipeProvider.this.recipePathProvider.json(p_312039_)));
                         if (p_311794_ != null) {
                             list.add(
                                 DataProvider.saveStable(
-                                    p_254020_, Advancement.CODEC, p_311794_.value(), RecipeProvider.this.advancementPathProvider.json(p_311794_.id())
+                                    p_254020_, net.neoforged.neoforge.common.util.NeoForgeExtraCodecs.CONDITIONAL_ADVANCEMENT_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.ConditionalObject<>(conditions, p_311794_.value(), alternatives.stream().filter(w -> w.carrier().getSecond() != null).map(w -> new net.neoforged.neoforge.common.conditions.WithConditions<>(w.conditions(), w.carrier().getSecond())).toList())), RecipeProvider.this.advancementPathProvider.json(p_311794_.id())
                                 )
                             );
                         }
@@ -99,16 +_,20 @@
                 }
             }
         );
-        return CompletableFuture.allOf(list.toArray(p_253414_ -> new CompletableFuture[p_253414_]));
+        return CompletableFuture.allOf(list.toArray(p_253414_ -> new CompletableFuture[p_253414_])); });
     }
 
     protected CompletableFuture<?> buildAdvancement(CachedOutput p_253674_, AdvancementHolder p_301116_) {
-        return DataProvider.saveStable(p_253674_, Advancement.CODEC, p_301116_.value(), this.advancementPathProvider.json(p_301116_.id()));
+        return buildAdvancement(p_253674_, p_301116_, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+    }
+
+    protected CompletableFuture<?> buildAdvancement(CachedOutput p_253674_, AdvancementHolder p_301116_, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
+        return DataProvider.saveStable(p_253674_, net.neoforged.neoforge.common.util.NeoForgeExtraCodecs.CONDITIONAL_ADVANCEMENT_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(p_301116_.value(), conditions)), this.advancementPathProvider.json(p_301116_.id()));
     }
 
     protected abstract void buildRecipes(RecipeOutput p_301172_);
 
-    protected static void generateForEnabledBlockFamilies(RecipeOutput p_301146_, FeatureFlagSet p_251836_) {
+    protected void generateForEnabledBlockFamilies(RecipeOutput p_301146_, FeatureFlagSet p_251836_) {
         BlockFamilies.getAllFamilies().filter(BlockFamily::shouldGenerateRecipe).forEach(p_313461_ -> generateRecipes(p_301146_, p_313461_, p_251836_));
     }
 
