--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -83,18 +_,18 @@
         this.buildRecipes(
             new RecipeOutput() {
                 @Override
-                public void accept(ResourceLocation p_312039_, Recipe<?> p_312254_, @Nullable AdvancementHolder p_311794_) {
+                public void accept(ResourceLocation p_312039_, Recipe<?> p_312254_, @Nullable AdvancementHolder p_311794_, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
                     if (!set.add(p_312039_)) {
                         throw new IllegalStateException("Duplicate recipe " + p_312039_);
                     } else {
-                        list.add(DataProvider.saveStable(p_324494_, p_324248_, Recipe.CODEC, p_312254_, RecipeProvider.this.recipePathProvider.json(p_312039_)));
+                        list.add(DataProvider.saveStable(p_324494_, p_324248_, Recipe.CONDITIONAL_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(p_312254_, conditions)), RecipeProvider.this.recipePathProvider.json(p_312039_)));
                         if (p_311794_ != null) {
                             list.add(
                                 DataProvider.saveStable(
                                     p_324494_,
                                     p_324248_,
-                                    Advancement.CODEC,
-                                    p_311794_.value(),
+                                    Advancement.CONDITIONAL_CODEC,
+                                    Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(p_311794_.value(), conditions)),
                                     RecipeProvider.this.advancementPathProvider.json(p_311794_.id())
                                 )
                             );
@@ -112,12 +_,16 @@
     }
 
     protected CompletableFuture<?> buildAdvancement(CachedOutput p_253674_, HolderLookup.Provider p_323646_, AdvancementHolder p_301116_) {
-        return DataProvider.saveStable(p_253674_, p_323646_, Advancement.CODEC, p_301116_.value(), this.advancementPathProvider.json(p_301116_.id()));
+        return buildAdvancement(p_253674_, p_323646_, p_301116_, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+    }
+
+    protected CompletableFuture<?> buildAdvancement(CachedOutput p_253674_, HolderLookup.Provider p_323646_, AdvancementHolder p_301116_, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
+        return DataProvider.saveStable(p_253674_, p_323646_, Advancement.CONDITIONAL_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(p_301116_.value(), conditions)), this.advancementPathProvider.json(p_301116_.id()));
     }
 
     protected abstract void buildRecipes(RecipeOutput p_301172_);
 
-    protected static void generateForEnabledBlockFamilies(RecipeOutput p_301146_, FeatureFlagSet p_251836_) {
+    protected void generateForEnabledBlockFamilies(RecipeOutput p_301146_, FeatureFlagSet p_251836_) {
         BlockFamilies.getAllFamilies().filter(BlockFamily::shouldGenerateRecipe).forEach(p_313461_ -> generateRecipes(p_301146_, p_313461_, p_251836_));
     }
 
@@ -598,17 +_,18 @@
     }
 
     protected static void waxRecipes(RecipeOutput p_301254_, FeatureFlagSet p_313879_) {
-        HoneycombItem.WAXABLES
-            .get()
+        BuiltInRegistries.BLOCK.getDataMap(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.WAXABLES)
             .forEach(
                 (p_337490_, p_337491_) -> {
-                    if (p_337491_.requiredFeatures().isSubsetOf(p_313879_)) {
-                        ShapelessRecipeBuilder.shapeless(RecipeCategory.BUILDING_BLOCKS, p_337491_)
-                            .requires(p_337490_)
+                    final var block = BuiltInRegistries.BLOCK.get(p_337490_);
+                    final var waxed = block.builtInRegistryHolder().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.WAXABLES).waxable();
+                    if (block.requiredFeatures().isSubsetOf(p_313879_)) {
+                        ShapelessRecipeBuilder.shapeless(RecipeCategory.BUILDING_BLOCKS, waxed)
+                            .requires(block)
                             .requires(Items.HONEYCOMB)
-                            .group(getItemName(p_337491_))
-                            .unlockedBy(getHasName(p_337490_), has(p_337490_))
-                            .save(p_301254_, getConversionRecipeName(p_337491_, Items.HONEYCOMB));
+                            .group(getItemName(waxed))
+                            .unlockedBy(getHasName(block), has(block))
+                            .save(p_301254_, getConversionRecipeName(waxed, Items.HONEYCOMB));
                     }
                 }
             );
