--- a/net/minecraft/data/recipes/RecipeProvider.java
+++ b/net/minecraft/data/recipes/RecipeProvider.java
@@ -84,18 +_,18 @@
         this.buildRecipes(
             new RecipeOutput() {
                 @Override
-                public void accept(ResourceLocation p_312039_, Recipe<?> p_312254_, @Nullable AdvancementHolder p_311794_) {
+                public void accept(ResourceLocation p_312039_, Recipe<?> p_312254_, @Nullable AdvancementHolder p_311794_, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
                     if (!set.add(p_312039_)) {
                         throw new IllegalStateException("Duplicate recipe " + p_312039_);
                     } else {
-                        list.add(DataProvider.saveStable(output, registries, Recipe.CODEC, p_312254_, RecipeProvider.this.recipePathProvider.json(p_312039_)));
+                        list.add(DataProvider.saveStable(output, registries, Recipe.CONDITIONAL_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(p_312254_, conditions)), RecipeProvider.this.recipePathProvider.json(p_312039_)));
                         if (p_311794_ != null) {
                             list.add(
                                 DataProvider.saveStable(
                                     output,
                                     registries,
-                                    Advancement.CODEC,
-                                    p_311794_.value(),
+                                    Advancement.CONDITIONAL_CODEC,
+                                    Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(p_311794_.value(), conditions)),
                                     RecipeProvider.this.advancementPathProvider.json(p_311794_.id())
                                 )
                             );
@@ -113,12 +_,16 @@
     }
 
     protected CompletableFuture<?> buildAdvancement(CachedOutput output, HolderLookup.Provider registries, AdvancementHolder advancement) {
-        return DataProvider.saveStable(output, registries, Advancement.CODEC, advancement.value(), this.advancementPathProvider.json(advancement.id()));
+        return buildAdvancement(output, registries, advancement, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+    }
+
+    protected CompletableFuture<?> buildAdvancement(CachedOutput output, HolderLookup.Provider registries, AdvancementHolder advancement, net.neoforged.neoforge.common.conditions.ICondition... conditions) {
+        return DataProvider.saveStable(output, registries, Advancement.CONDITIONAL_CODEC, Optional.of(new net.neoforged.neoforge.common.conditions.WithConditions<>(advancement.value(), conditions)), this.advancementPathProvider.json(advancement.id()));
     }
 
     protected abstract void buildRecipes(RecipeOutput recipeOutput);
 
-    protected static void generateForEnabledBlockFamilies(RecipeOutput recipeOutput, FeatureFlagSet enabledFeatures) {
+    protected void generateForEnabledBlockFamilies(RecipeOutput recipeOutput, FeatureFlagSet enabledFeatures) {
         BlockFamilies.getAllFamilies().filter(BlockFamily::shouldGenerateRecipe).forEach(p_313461_ -> generateRecipes(recipeOutput, p_313461_, enabledFeatures));
     }
 
