--- a/net/minecraft/data/recipes/ShapedRecipeBuilder.java
+++ b/net/minecraft/data/recipes/ShapedRecipeBuilder.java
@@ -51,6 +_,10 @@
         return new ShapedRecipeBuilder(items, category, item, count);
     }
 
+    public static ShapedRecipeBuilder shaped(HolderGetter<Item> items, RecipeCategory category, ItemStack result) {
+        return new ShapedRecipeBuilder(items, category, result);
+    }
+
     public ShapedRecipeBuilder define(Character symbol, TagKey<Item> tag) {
         return this.define(symbol, Ingredient.of(this.items.getOrThrow(tag)));
     }
