--- a/net/minecraft/data/recipes/ShapedRecipeBuilder.java
+++ b/net/minecraft/data/recipes/ShapedRecipeBuilder.java
@@ -25,6 +_,7 @@
     private final RecipeCategory category;
     private final Item result;
     private final int count;
+    private final ItemStack resultStack; // Neo: add stack result support
     private final List<String> rows = Lists.newArrayList();
     private final Map<Character, Ingredient> key = Maps.newLinkedHashMap();
     private final Map<String, Criterion<?>> criteria = new LinkedHashMap<>();
@@ -33,9 +_,14 @@
     private boolean showNotification = true;
 
     public ShapedRecipeBuilder(RecipeCategory category, ItemLike result, int count) {
+        this(category, new ItemStack(result, count));
+    }
+
+    public ShapedRecipeBuilder(RecipeCategory category, ItemStack result) {
         this.category = category;
-        this.result = result.asItem();
-        this.count = count;
+        this.result = result.getItem();
+        this.count = result.getCount();
+        this.resultStack = result;
     }
 
     public static ShapedRecipeBuilder shaped(RecipeCategory category, ItemLike result) {
@@ -46,6 +_,10 @@
         return new ShapedRecipeBuilder(category, result, count);
     }
 
+    public static ShapedRecipeBuilder shaped(RecipeCategory category, ItemStack result) {
+        return new ShapedRecipeBuilder(category, result);
+    }
+
     public ShapedRecipeBuilder define(Character symbol, TagKey<Item> tag) {
         return this.define(symbol, Ingredient.of(tag));
     }
@@ -106,7 +_,7 @@
             Objects.requireNonNullElse(this.group, ""),
             RecipeBuilder.determineBookCategory(this.category),
             shapedrecipepattern,
-            new ItemStack(this.result, this.count),
+            this.resultStack,
             this.showNotification
         );
         recipeOutput.accept(id, shapedrecipe, advancement$builder.build(id.withPrefix("recipes/" + this.category.getFolderName() + "/")));
