--- a/net/minecraft/data/recipes/ShapelessRecipeBuilder.java
+++ b/net/minecraft/data/recipes/ShapelessRecipeBuilder.java
@@ -21,23 +_,24 @@
 import net.minecraft.world.item.crafting.ShapelessRecipe;
 import net.minecraft.world.level.ItemLike;
 
-public class ShapelessRecipeBuilder implements RecipeBuilder {
+public class ShapelessRecipeBuilder implements RecipeBuilder, net.neoforged.neoforge.common.extensions.IShapelessRecipeBuilderExtension {
     private final HolderGetter<Item> items;
     private final RecipeCategory category;
-    private final ItemStack result;
+    private final net.neoforged.neoforge.common.crafting.outgredient.Outgredient<ItemStack> result;
     private final List<Ingredient> ingredients = new ArrayList<>();
     private final Map<String, Criterion<?>> criteria = new LinkedHashMap<>();
     @Nullable
     private String group;
 
-    private ShapelessRecipeBuilder(HolderGetter<Item> p_363417_, RecipeCategory p_250837_, ItemStack p_363612_) {
+    @Deprecated // Neo: use the various #shapeless methods instead
+    public ShapelessRecipeBuilder(HolderGetter<Item> p_363417_, RecipeCategory p_250837_, net.neoforged.neoforge.common.crafting.outgredient.Outgredient<ItemStack> p_363612_) {
         this.items = p_363417_;
         this.category = p_250837_;
         this.result = p_363612_;
     }
 
     public static ShapelessRecipeBuilder shapeless(HolderGetter<Item> p_364294_, RecipeCategory p_361887_, ItemStack p_364359_) {
-        return new ShapelessRecipeBuilder(p_364294_, p_361887_, p_364359_);
+        return net.neoforged.neoforge.common.extensions.IShapelessRecipeBuilderExtension.shapeless(p_364294_, p_361887_, net.neoforged.neoforge.common.crafting.outgredient.Outgredient.ofItem(p_364359_));
     }
 
     public static ShapelessRecipeBuilder shapeless(HolderGetter<Item> p_362315_, RecipeCategory p_250714_, ItemLike p_249659_) {
@@ -45,7 +_,7 @@
     }
 
     public static ShapelessRecipeBuilder shapeless(HolderGetter<Item> p_360448_, RecipeCategory p_252339_, ItemLike p_250836_, int p_249928_) {
-        return new ShapelessRecipeBuilder(p_360448_, p_252339_, p_250836_.asItem().getDefaultInstance().copyWithCount(p_249928_));
+        return shapeless(p_360448_, p_252339_, p_250836_.asItem().getDefaultInstance().copyWithCount(p_249928_));
     }
 
     public ShapelessRecipeBuilder requires(TagKey<Item> p_206420_) {
@@ -88,7 +_,7 @@
 
     @Override
     public Item getResult() {
-        return this.result.getItem();
+        return this.result.resolve().getItem();
     }
 
     @Override
