--- a/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
+++ b/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
@@ -27,13 +_,18 @@
     private final RecipeCategory category;
     private final CookingBookCategory bookCategory;
     private final Item result;
+    private final ItemStack stackResult; // Neo: add stack result support
     private final Ingredient ingredient;
     private final float experience;
     private final int cookingTime;
     private final Map<String, Criterion<?>> criteria = new LinkedHashMap<>();
     @Nullable
     private String group;
+    @Nullable
+    private String override;
     private final AbstractCookingRecipe.Factory<?> factory;
+    @Nullable
+    private final AbstractCookingRecipe.NeoForgeFactory<?> neoForgeFactory;
 
     private SimpleCookingRecipeBuilder(
         RecipeCategory p_251345_,
@@ -44,13 +_,59 @@
         int p_250189_,
         AbstractCookingRecipe.Factory<?> p_311960_
     ) {
-        this.category = p_251345_;
-        this.bookCategory = p_251607_;
-        this.result = p_252112_.asItem();
-        this.ingredient = p_250362_;
-        this.experience = p_251204_;
-        this.cookingTime = p_250189_;
-        this.factory = p_311960_;
+        this(p_251345_, p_251607_, new ItemStack(p_252112_), p_250362_, p_251204_, p_250189_, p_311960_);
+    }
+
+    private SimpleCookingRecipeBuilder(
+            RecipeCategory p_251345_,
+            CookingBookCategory p_251607_,
+            ItemStack result,
+            Ingredient p_250362_,
+            float p_251204_,
+            int p_250189_,
+            AbstractCookingRecipe.Factory<?> p_311960_
+    ) {
+        this.category = p_251345_;
+        this.bookCategory = p_251607_;
+        this.result = result.getItem();
+        this.stackResult = result;
+        this.ingredient = p_250362_;
+        this.experience = p_251204_;
+        this.cookingTime = p_250189_;
+        this.factory = p_311960_;
+        this.neoForgeFactory = null;
+    }
+
+    private SimpleCookingRecipeBuilder(
+            RecipeCategory p_251345_,
+            CookingBookCategory p_251607_,
+            ItemLike p_252112_,
+            Ingredient p_250362_,
+            float p_251204_,
+            int p_250189_,
+            AbstractCookingRecipe.NeoForgeFactory<?> p_311960_
+    ) {
+        this(p_251345_, p_251607_, new ItemStack(p_252112_), p_250362_, p_251204_, p_250189_, p_311960_);
+    }
+
+    private SimpleCookingRecipeBuilder(
+            RecipeCategory p_251345_,
+            CookingBookCategory p_251607_,
+            ItemStack result,
+            Ingredient p_250362_,
+            float p_251204_,
+            int p_250189_,
+            AbstractCookingRecipe.NeoForgeFactory<?> p_311960_
+    ) {
+        this.category = p_251345_;
+        this.bookCategory = p_251607_;
+        this.result = result.getItem();
+        this.stackResult = result;
+        this.ingredient = p_250362_;
+        this.experience = p_251204_;
+        this.cookingTime = p_250189_;
+        this.factory = p_311960_;
+        this.neoForgeFactory = p_311960_;
     }
 
     public static <T extends AbstractCookingRecipe> SimpleCookingRecipeBuilder generic(
@@ -85,11 +_,48 @@
         return new SimpleCookingRecipeBuilder(p_250319_, CookingBookCategory.FOOD, p_250377_, p_248930_, p_252329_, p_250482_, SmokingRecipe::new);
     }
 
+    public static <T extends AbstractCookingRecipe> SimpleCookingRecipeBuilder generic(
+            Ingredient p_250999_,
+            RecipeCategory p_248815_,
+            ItemStack result,
+            float p_251320_,
+            int p_248693_,
+            RecipeSerializer<T> p_250921_,
+            AbstractCookingRecipe.Factory<T> p_312657_
+    ) {
+        return new SimpleCookingRecipeBuilder(p_248815_, determineSmeltingRecipeCategory(result.getItem()), result, p_250999_, p_251320_, p_248693_, p_312657_);
+    }
+
+    public static SimpleCookingRecipeBuilder campfireCooking(Ingredient p_249393_, RecipeCategory p_249372_, ItemStack result, float p_252321_, int p_251916_) {
+        return new SimpleCookingRecipeBuilder(p_249372_, CookingBookCategory.FOOD, result, p_249393_, p_252321_, p_251916_, CampfireCookingRecipe::new);
+    }
+
+    public static SimpleCookingRecipeBuilder blasting(Ingredient p_252115_, RecipeCategory p_249421_, ItemStack result, float p_250383_, int p_250476_) {
+        return new SimpleCookingRecipeBuilder(
+                p_249421_, determineBlastingRecipeCategory(result.getItem()), result, p_252115_, p_250383_, p_250476_, BlastingRecipe::new
+        );
+    }
+
+    public static SimpleCookingRecipeBuilder smelting(Ingredient p_249223_, RecipeCategory p_251240_, ItemStack result, float p_249452_, int p_250496_) {
+        return new SimpleCookingRecipeBuilder(
+                p_251240_, determineSmeltingRecipeCategory(result.getItem()), result, p_249223_, p_249452_, p_250496_, SmeltingRecipe::new
+        );
+    }
+
+    public static SimpleCookingRecipeBuilder smoking(Ingredient p_248930_, RecipeCategory p_250319_, ItemStack result, float p_252329_, int p_250482_) {
+        return new SimpleCookingRecipeBuilder(p_250319_, CookingBookCategory.FOOD, result, p_248930_, p_252329_, p_250482_, SmokingRecipe::new);
+    }
+
     public SimpleCookingRecipeBuilder unlockedBy(String p_176792_, Criterion<?> p_300970_) {
         this.criteria.put(p_176792_, p_300970_);
         return this;
     }
 
+    public SimpleCookingRecipeBuilder override(@Nullable String override) {
+        this.override = override;
+        return this;
+    }
+
     public SimpleCookingRecipeBuilder group(@Nullable String p_176795_) {
         this.group = p_176795_;
         return this;
@@ -108,10 +_,20 @@
             .rewards(AdvancementRewards.Builder.recipe(p_126264_))
             .requirements(AdvancementRequirements.Strategy.OR);
         this.criteria.forEach(advancement$builder::addCriterion);
-        AbstractCookingRecipe abstractcookingrecipe = this.factory
-            .create(
-                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, new ItemStack(this.result), this.experience, this.cookingTime
-            );
+        AbstractCookingRecipe abstractcookingrecipe;
+        if (this.neoForgeFactory != null) {
+            abstractcookingrecipe = this.neoForgeFactory
+                    .create(
+                            Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, this.stackResult, this.experience, this.cookingTime, Objects.requireNonNullElse(this.override, "")
+                    );
+        } else {
+            abstractcookingrecipe = this.factory
+                    .create(
+                            Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, this.stackResult, this.experience, this.cookingTime
+                    );
+        }
+
+
         p_301266_.accept(p_126264_, abstractcookingrecipe, advancement$builder.build(p_126264_.withPrefix("recipes/" + this.category.getFolderName() + "/")));
     }
 
