--- a/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
+++ b/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
@@ -26,7 +_,7 @@
 public class SimpleCookingRecipeBuilder implements RecipeBuilder {
     private final RecipeCategory category;
     private final CookingBookCategory bookCategory;
-    private final Item result;
+    private final ItemStack result; // Neo: patched to ItemStack
     private final Ingredient ingredient;
     private final float experience;
     private final int cookingTime;
@@ -44,9 +_,21 @@
         int p_250189_,
         AbstractCookingRecipe.Factory<?> p_311960_
     ) {
+        this(p_251345_, p_251607_, new ItemStack(p_252112_), p_250362_, p_251204_, p_250189_, p_311960_);
+    }
+
+    private SimpleCookingRecipeBuilder(
+            RecipeCategory p_251345_,
+            CookingBookCategory p_251607_,
+            ItemStack result,
+            Ingredient p_250362_,
+            float p_251204_,
+            int p_250189_,
+            AbstractCookingRecipe.Factory<?> p_311960_
+    ) {
         this.category = p_251345_;
         this.bookCategory = p_251607_;
-        this.result = p_252112_.asItem();
+        this.result = result;
         this.ingredient = p_250362_;
         this.experience = p_251204_;
         this.cookingTime = p_250189_;
@@ -85,6 +_,38 @@
         return new SimpleCookingRecipeBuilder(p_250319_, CookingBookCategory.FOOD, p_250377_, p_248930_, p_252329_, p_250482_, SmokingRecipe::new);
     }
 
+    public static <T extends AbstractCookingRecipe> SimpleCookingRecipeBuilder generic(
+            Ingredient p_250999_,
+            RecipeCategory p_248815_,
+            ItemStack result,
+            float p_251320_,
+            int p_248693_,
+            RecipeSerializer<T> p_250921_,
+            AbstractCookingRecipe.Factory<T> p_312657_
+    ) {
+        return new SimpleCookingRecipeBuilder(p_248815_, determineSmeltingRecipeCategory(result.getItem()), result, p_250999_, p_251320_, p_248693_, p_312657_);
+    }
+
+    public static SimpleCookingRecipeBuilder campfireCooking(Ingredient p_249393_, RecipeCategory p_249372_, ItemStack result, float p_252321_, int p_251916_) {
+        return new SimpleCookingRecipeBuilder(p_249372_, CookingBookCategory.FOOD, result, p_249393_, p_252321_, p_251916_, CampfireCookingRecipe::new);
+    }
+
+    public static SimpleCookingRecipeBuilder blasting(Ingredient p_252115_, RecipeCategory p_249421_, ItemStack result, float p_250383_, int p_250476_) {
+        return new SimpleCookingRecipeBuilder(
+                p_249421_, determineBlastingRecipeCategory(result.getItem()), result, p_252115_, p_250383_, p_250476_, BlastingRecipe::new
+        );
+    }
+
+    public static SimpleCookingRecipeBuilder smelting(Ingredient p_249223_, RecipeCategory p_251240_, ItemStack result, float p_249452_, int p_250496_) {
+        return new SimpleCookingRecipeBuilder(
+                p_251240_, determineSmeltingRecipeCategory(result.getItem()), result, p_249223_, p_249452_, p_250496_, SmeltingRecipe::new
+        );
+    }
+
+    public static SimpleCookingRecipeBuilder smoking(Ingredient p_248930_, RecipeCategory p_250319_, ItemStack result, float p_252329_, int p_250482_) {
+        return new SimpleCookingRecipeBuilder(p_250319_, CookingBookCategory.FOOD, result, p_248930_, p_252329_, p_250482_, SmokingRecipe::new);
+    }
+
     public SimpleCookingRecipeBuilder unlockedBy(String p_176792_, Criterion<?> p_300970_) {
         this.criteria.put(p_176792_, p_300970_);
         return this;
@@ -97,7 +_,7 @@
 
     @Override
     public Item getResult() {
-        return this.result;
+        return this.result.getItem();
     }
 
     @Override
@@ -110,7 +_,7 @@
         this.criteria.forEach(advancement$builder::addCriterion);
         AbstractCookingRecipe abstractcookingrecipe = this.factory
             .create(
-                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, new ItemStack(this.result), this.experience, this.cookingTime
+                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, this.result, this.experience, this.cookingTime
             );
         p_301266_.accept(p_126264_, abstractcookingrecipe, advancement$builder.build(p_126264_.withPrefix("recipes/" + this.category.getFolderName() + "/")));
     }
