--- a/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
+++ b/net/minecraft/data/recipes/SimpleCookingRecipeBuilder.java
@@ -25,10 +_,11 @@
 import net.minecraft.world.item.crafting.SmokingRecipe;
 import net.minecraft.world.level.ItemLike;
 
-public class SimpleCookingRecipeBuilder implements RecipeBuilder {
+public class SimpleCookingRecipeBuilder implements RecipeBuilder, net.neoforged.neoforge.common.extensions.ISimpleCookingRecipeBuilderExtension {
     private final RecipeCategory category;
     private final CookingBookCategory bookCategory;
     private final Item result;
+    private final net.neoforged.neoforge.common.crafting.outgredient.Outgredient<ItemStack> resultStack; // Neo: add result support
     private final Ingredient ingredient;
     private final float experience;
     private final int cookingTime;
@@ -37,18 +_,20 @@
     private String group;
     private final AbstractCookingRecipe.Factory<?> factory;
 
-    private SimpleCookingRecipeBuilder(
-        RecipeCategory p_251345_,
-        CookingBookCategory p_251607_,
-        ItemLike p_252112_,
-        Ingredient p_250362_,
-        float p_251204_,
-        int p_250189_,
-        AbstractCookingRecipe.Factory<?> p_311960_
+    @Deprecated // Neo: use the various static methods instead
+    public SimpleCookingRecipeBuilder(
+            RecipeCategory p_251345_,
+            CookingBookCategory p_251607_,
+            net.neoforged.neoforge.common.crafting.outgredient.Outgredient<ItemStack> result,
+            Ingredient p_250362_,
+            float p_251204_,
+            int p_250189_,
+            AbstractCookingRecipe.Factory<?> p_311960_
     ) {
         this.category = p_251345_;
         this.bookCategory = p_251607_;
-        this.result = p_252112_.asItem();
+        this.result = net.minecraft.world.item.Items.AIR;
+        this.resultStack = result;
         this.ingredient = p_250362_;
         this.experience = p_251204_;
         this.cookingTime = p_250189_;
@@ -64,27 +_,23 @@
         RecipeSerializer<T> p_250921_,
         AbstractCookingRecipe.Factory<T> p_312657_
     ) {
-        return new SimpleCookingRecipeBuilder(p_248815_, determineRecipeCategory(p_250921_, p_249766_), p_249766_, p_250999_, p_251320_, p_248693_, p_312657_);
+        return net.neoforged.neoforge.common.extensions.ISimpleCookingRecipeBuilderExtension.generic(p_250999_, p_248815_, p_249766_.asItem().getDefaultInstance(), p_251320_, p_248693_, p_250921_, p_312657_);
     }
 
     public static SimpleCookingRecipeBuilder campfireCooking(Ingredient p_249393_, RecipeCategory p_249372_, ItemLike p_251516_, float p_252321_, int p_251916_) {
-        return new SimpleCookingRecipeBuilder(p_249372_, CookingBookCategory.FOOD, p_251516_, p_249393_, p_252321_, p_251916_, CampfireCookingRecipe::new);
+        return net.neoforged.neoforge.common.extensions.ISimpleCookingRecipeBuilderExtension.campfireCooking(p_249393_, p_249372_, p_251516_.asItem().getDefaultInstance(), p_252321_, p_251916_);
     }
 
     public static SimpleCookingRecipeBuilder blasting(Ingredient p_252115_, RecipeCategory p_249421_, ItemLike p_251247_, float p_250383_, int p_250476_) {
-        return new SimpleCookingRecipeBuilder(
-            p_249421_, determineBlastingRecipeCategory(p_251247_), p_251247_, p_252115_, p_250383_, p_250476_, BlastingRecipe::new
-        );
+        return net.neoforged.neoforge.common.extensions.ISimpleCookingRecipeBuilderExtension.blasting(p_252115_, p_249421_, p_251247_.asItem().getDefaultInstance(), p_250383_, p_250476_);
     }
 
     public static SimpleCookingRecipeBuilder smelting(Ingredient p_249223_, RecipeCategory p_251240_, ItemLike p_249551_, float p_249452_, int p_250496_) {
-        return new SimpleCookingRecipeBuilder(
-            p_251240_, determineSmeltingRecipeCategory(p_249551_), p_249551_, p_249223_, p_249452_, p_250496_, SmeltingRecipe::new
-        );
+        return net.neoforged.neoforge.common.extensions.ISimpleCookingRecipeBuilderExtension.blasting(p_249223_, p_251240_, p_249551_.asItem().getDefaultInstance(), p_249452_, p_250496_);
     }
 
     public static SimpleCookingRecipeBuilder smoking(Ingredient p_248930_, RecipeCategory p_250319_, ItemLike p_250377_, float p_252329_, int p_250482_) {
-        return new SimpleCookingRecipeBuilder(p_250319_, CookingBookCategory.FOOD, p_250377_, p_248930_, p_252329_, p_250482_, SmokingRecipe::new);
+        return net.neoforged.neoforge.common.extensions.ISimpleCookingRecipeBuilderExtension.smoking(p_248930_, p_250319_, p_250377_.asItem().getDefaultInstance(), p_252329_, p_250482_);
     }
 
     public SimpleCookingRecipeBuilder unlockedBy(String p_176792_, Criterion<?> p_300970_) {
@@ -99,7 +_,7 @@
 
     @Override
     public Item getResult() {
-        return this.result;
+        return this.resultStack.resolve().getItem();
     }
 
     @Override
@@ -112,7 +_,7 @@
         this.criteria.forEach(advancement$builder::addCriterion);
         AbstractCookingRecipe abstractcookingrecipe = this.factory
             .create(
-                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, new ItemStack(this.result), this.experience, this.cookingTime
+                Objects.requireNonNullElse(this.group, ""), this.bookCategory, this.ingredient, this.resultStack, this.experience, this.cookingTime
             );
         p_301266_.accept(
             p_380371_, abstractcookingrecipe, advancement$builder.build(p_380371_.location().withPrefix("recipes/" + this.category.getFolderName() + "/"))
