--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -20,6 +20,14 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+// CraftBukkit end
+
 public class TemptingSensor extends Sensor<PathfinderMob> {
 
     private static final TargetingConditions TEMPT_TARGETING = TargetingConditions.forNonCombat().ignoreLineOfSight();
@@ -48,7 +56,7 @@
     protected void doTick(ServerLevel level, PathfinderMob body) {
         Brain<?> brain = body.getBrain();
         TargetingConditions targetingconditions = TemptingSensor.TEMPT_TARGETING.copy().range((double) ((float) body.getAttributeValue(Attributes.TEMPT_RANGE)));
-        Stream stream = level.players().stream().filter(EntitySelector.NO_SPECTATORS).filter((serverplayer) -> {
+        Stream<net.minecraft.server.level.ServerPlayer> stream = level.players().stream().filter(EntitySelector.NO_SPECTATORS).filter((serverplayer) -> { // CraftBukkit - decompile error
             return targetingconditions.test(level, body, serverplayer);
         }).filter((serverplayer) -> {
             return this.playerHoldingTemptation(body, serverplayer);
@@ -62,7 +70,17 @@
         if (!list.isEmpty()) {
             Player player = (Player) list.get(0);
 
-            brain.setMemory(MemoryModuleType.TEMPTING_PLAYER, player);
+            // CraftBukkit start
+            EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(body, player, EntityTargetEvent.TargetReason.TEMPT);
+            if (event.isCancelled()) {
+                return;
+            }
+            if (event.getTarget() instanceof HumanEntity) {
+                brain.setMemory(MemoryModuleType.TEMPTING_PLAYER, ((CraftHumanEntity) event.getTarget()).getHandle());
+            } else {
+                brain.eraseMemory(MemoryModuleType.TEMPTING_PLAYER);
+            }
+            // CraftBukkit end
         } else {
             brain.eraseMemory(MemoryModuleType.TEMPTING_PLAYER);
         }
