--- a/net/minecraft/client/server/IntegratedServer.java
+++ b/net/minecraft/client/server/IntegratedServer.java
@@ -96,10 +_,12 @@
         LOGGER.info("Starting integrated minecraft server version {}", SharedConstants.getCurrentVersion().name());
         this.setUsesAuthentication(true);
         this.initializeKeyPair();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
         this.loadLevel();
         GameProfile host = this.getSingleplayerProfile();
         String levelName = this.getWorldData().getLevelName();
         this.setMotd(host != null ? host.name() + " - " + levelName : levelName);
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
 
@@ -255,6 +_,7 @@
 
     @Override
     public void halt(boolean wait) {
+        if (isRunning())
         this.executeBlocking(() -> {
             for (ServerPlayer player : Lists.newArrayList(this.getPlayerList().getPlayers())) {
                 if (!player.getUUID().equals(this.uuid)) {
