--- a/net/minecraft/network/protocol/login/ClientboundCustomQueryPacket.java
+++ b/net/minecraft/network/protocol/login/ClientboundCustomQueryPacket.java
@@ -8,15 +_,15 @@
 
 public record ClientboundCustomQueryPacket(int transactionId, CustomQueryPayload payload) implements Packet<ClientLoginPacketListener> {
     private static final int MAX_PAYLOAD_SIZE = 1048576;
-
+    
     public ClientboundCustomQueryPacket(FriendlyByteBuf p_179810_) {
         this(p_179810_.readVarInt(), readPayload(p_179810_.readResourceLocation(), p_179810_));
     }
-
+    
     private static CustomQueryPayload readPayload(ResourceLocation p_295267_, FriendlyByteBuf p_295117_) {
         return readUnknownPayload(p_295267_, p_295117_);
     }
-
+    
     private static DiscardedQueryPayload readUnknownPayload(ResourceLocation p_294837_, FriendlyByteBuf p_296380_) {
         int i = p_296380_.readableBytes();
         if (i >= 0 && i <= 1048576) {
@@ -26,14 +_,14 @@
             throw new IllegalArgumentException("Payload may not be larger than 1048576 bytes");
         }
     }
-
+    
     @Override
     public void write(FriendlyByteBuf p_134757_) {
         p_134757_.writeVarInt(this.transactionId);
         p_134757_.writeResourceLocation(this.payload.id());
         this.payload.write(p_134757_);
     }
-
+    
     public void handle(ClientLoginPacketListener p_134754_) {
         p_134754_.handleCustomQuery(this);
     }
