--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -733,7 +775,7 @@
 
     @Override
     protected void renderBg(GuiGraphics graphics, float a, int xm, int ym) {
-        for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
+        for (CreativeModeTab tab : currentPage.getVisibleTabs()) {
             if (tab != selectedTab) {
                 this.renderTabButton(graphics, xm, ym, tab);
             }
