--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -237,11 +252,11 @@
     }
 
     public void defaultErrorCallback(int errorCode, long description) {
-        if (!RenderSystem.isOnRenderThread()) {
-            LOGGER.error("Error callback from invalid thread ({})", Thread.currentThread().getName());
-        }
-
         String errorString = MemoryUtil.memUTF8(description);
+        if (!RenderSystem.isOnRenderThread()) {
+            LOGGER.error("Error callback from invalid thread ({}): {}: {}", Thread.currentThread().getName(), errorCode, errorString);
+        }
+
         LOGGER.error("########## GL ERROR ##########");
         LOGGER.error("@ {}", this.errorSection);
         LOGGER.error("{}: {}", errorCode, errorString);
