--- a/net/minecraft/client/gui/screens/debug/DebugOptionsScreen.java
+++ b/net/minecraft/client/gui/screens/debug/DebugOptionsScreen.java
@@ -192,6 +_,10 @@
             DebugScreenEntry entry = DebugScreenEntries.getEntry(location);
             this.isAllowed = entry != null && entry.isAllowed(DebugOptionsScreen.this.minecraft.showOnlyReducedInfo());
             String name = location.getPath();
+            // Neo: Display the full RL for modded entries
+            if (!location.getNamespace().equals(Identifier.DEFAULT_NAMESPACE)) {
+                name = location.toString();
+            }
             if (this.isAllowed) {
                 this.name = name;
             } else {
@@ -252,8 +_,11 @@
         public void extractContent(GuiGraphicsExtractor graphics, int mouseX, int mouseY, boolean hovered, float a) {
             int x = this.getContentX();
             int y = this.getContentY();
-            graphics.text(DebugOptionsScreen.this.minecraft.font, this.name, x, y + 5, this.isAllowed ? -1 : -8355712);
+            // Neo: Moved 'k' up for 'maxX' calculation for scrolling string render
             int buttonsStartX = x + this.getContentWidth() - this.never.getWidth() - this.overlay.getWidth() - this.always.getWidth();
+            // Neo: Scroll entry names that go out of bounds
+
+            graphics.drawScrollingString(graphics.textRenderer(), DebugOptionsScreen.this.minecraft.font, Component.literal(this.name).withColor(this.isAllowed ? -1 : -8355712), x, buttonsStartX, y + 5);
             if (!this.isAllowed && hovered && mouseX < buttonsStartX) {
                 graphics.setTooltipForNextFrame(DebugOptionsScreen.NOT_ALLOWED_TOOLTIP, mouseX, mouseY);
             }
@@ -317,6 +_,9 @@
 
         public void updateSearch(String value) {
             this.clearEntries();
+            // Neo: Custom entry appending logic to better handle searching, categories and sorting
+            net.neoforged.neoforge.client.ClientHooks.updateDebugScreenEntriesForSearch(value, category -> addEntry(new CategoryEntry(category.label())), id -> addEntry(new OptionEntry(id)));
+            if(false){
             List<Map.Entry<Identifier, DebugScreenEntry>> all = new ArrayList<>(DebugScreenEntries.allEntries().entrySet());
             all.sort(COMPARATOR);
             DebugEntryCategory currentCategory = null;
@@ -331,6 +_,7 @@
 
                     this.addEntry(DebugOptionsScreen.this.new OptionEntry(entry.getKey()));
                 }
+            }
             }
 
             this.notifyListUpdated();
