--- a/net/minecraft/stats/ServerRecipeBook.java
+++ b/net/minecraft/stats/ServerRecipeBook.java
@@ -25,6 +25,8 @@
 import net.minecraft.world.item.crafting.display.RecipeDisplayId;
 import org.slf4j.Logger;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class ServerRecipeBook extends RecipeBook {
 
     public static final String RECIPE_BOOK_TAG = "recipeBook";
@@ -66,7 +68,7 @@
         for (RecipeHolder<?> recipeholder : recipes) {
             ResourceKey<Recipe<?>> resourcekey = recipeholder.id();
 
-            if (!this.known.contains(resourcekey) && !recipeholder.value().isSpecial()) {
+            if (!this.known.contains(resourcekey) && !recipeholder.value().isSpecial() && CraftEventFactory.handlePlayerRecipeListUpdateEvent(player, resourcekey.identifier())) { // CraftBukkit
                 this.add(resourcekey);
                 this.addHighlight(resourcekey);
                 this.displayResolver.displaysForRecipe(resourcekey, (recipedisplayentry) -> {
@@ -76,7 +78,7 @@
             }
         }
 
-        if (!list.isEmpty()) {
+        if (!list.isEmpty() && player.connection != null) { // SPIGOT-4478 during PlayerLoginEvent
             player.connection.send(new ClientboundRecipeBookAddPacket(list, false));
         }
 
@@ -97,7 +99,7 @@
             }
         }
 
-        if (!list.isEmpty()) {
+        if (!list.isEmpty() && player.connection != null) { // SPIGOT-4478 during PlayerLoginEvent
             player.connection.send(new ClientboundRecipeBookRemovePacket(list));
         }
 
