--- a/net/minecraft/stats/ServerStatsCounter.java
+++ b/net/minecraft/stats/ServerStatsCounter.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.stats;
 
 import com.google.common.collect.Sets;
@@ -59,11 +60,11 @@
     private final Path file;
     private final Set<Stat<?>> dirty = Sets.newHashSet();
 
-    private static <T> Codec<Map<Stat<?>, Integer>> createTypedStatsCodec(StatType<T> type) {
+    private static <T> Codec<Map<Stat<T>, Integer>> createTypedStatsCodec(StatType<T> type) { // CraftBukkit - decompile error
         Codec<T> codec = type.getRegistry().byNameCodec();
 
         Objects.requireNonNull(type);
-        Codec<Stat<?>> codec1 = codec.flatComapMap(type::get, (stat) -> {
+        Codec<Stat<T>> codec1 = codec.flatComapMap(type::get, (stat) -> { // CraftBukkit - decompile error
             return stat.getType() == type ? DataResult.success(stat.getValue()) : DataResult.error(() -> {
                 String s = String.valueOf(type);
 
