--- a/net/minecraft/world/item/crafting/AbstractCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/AbstractCookingRecipe.java
@@ -74,7 +_,7 @@
                         Codec.STRING.optionalFieldOf("group", "").forGetter(SingleItemRecipe::group),
                         CookingBookCategory.CODEC.fieldOf("category").orElse(CookingBookCategory.MISC).forGetter(AbstractCookingRecipe::category),
                         Ingredient.CODEC.fieldOf("ingredient").forGetter(SingleItemRecipe::input),
-                        ItemStackTemplate.CODEC.fieldOf("result").forGetter(SingleItemRecipe::result),
+                        ItemStack.CODEC.fieldOf("result").forGetter(SingleItemRecipe::result),
                         Codec.FLOAT.fieldOf("experience").orElse(0.0F).forGetter(AbstractCookingRecipe::experience),
                         Codec.INT.fieldOf("cookingtime").orElse(defaultCookingTime).forGetter(AbstractCookingRecipe::cookingTime)
                     )
