--- a/net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
+++ b/net/minecraft/world/item/crafting/BannerDuplicateRecipe.java
@@ -81,7 +_,7 @@
         for (int i = 0; i < nonnulllist.size(); i++) {
             ItemStack itemstack = p_345377_.getItem(i);
             if (!itemstack.isEmpty()) {
-                ItemStack itemstack1 = itemstack.getItem().getCraftingRemainder();
+                ItemStack itemstack1 = itemstack.getCraftingRemainder();
                 if (!itemstack1.isEmpty()) {
                     nonnulllist.set(i, itemstack1);
                 } else if (!itemstack.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY).layers().isEmpty()) {
