--- a/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -87,7 +_,7 @@
 
         for (int slot = 0; slot < result.size(); slot++) {
             ItemStack itemStack = input.getItem(slot);
-            ItemStackTemplate remainder = itemStack.getItem().getCraftingRemainder();
+            ItemStackTemplate remainder = itemStack.getCraftingRemainder();
             if (remainder != null) {
                 result.set(slot, remainder.create());
             } else if (itemStack.has(DataComponents.WRITTEN_BOOK_CONTENT)) {
