--- a/net/minecraft/world/item/crafting/BookCloningRecipe.java
+++ b/net/minecraft/world/item/crafting/BookCloningRecipe.java
@@ -68,6 +_,7 @@
             CompoundTag compoundtag = itemstack.getTag().copy();
             compoundtag.putInt("generation", WrittenBookItem.getGeneration(itemstack) + 1);
             itemstack2.setTag(compoundtag);
+            net.neoforged.neoforge.attachment.AttachmentInternals.copyStackAttachments(itemstack, itemstack2);
             return itemstack2;
         } else {
             return ItemStack.EMPTY;
@@ -79,8 +_,8 @@
 
         for(int i = 0; i < nonnulllist.size(); ++i) {
             ItemStack itemstack = p_43820_.getItem(i);
-            if (itemstack.getItem().hasCraftingRemainingItem()) {
-                nonnulllist.set(i, new ItemStack(itemstack.getItem().getCraftingRemainingItem()));
+            if (itemstack.hasCraftingRemainingItem()) {
+                nonnulllist.set(i, itemstack.getCraftingRemainingItem());
             } else if (itemstack.getItem() instanceof WrittenBookItem) {
                 nonnulllist.set(i, itemstack.copyWithCount(1));
                 break;
