--- a/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/CampfireCookingRecipe.java
@@ -8,6 +_,10 @@
         super(RecipeType.CAMPFIRE_COOKING, p_250140_, p_251808_, p_249826_, p_251839_, p_251432_, p_251471_);
     }
 
+    public CampfireCookingRecipe(String p_250140_, CookingBookCategory p_251808_, Ingredient p_249826_, ItemStack p_251839_, float p_251432_, int p_251471_, String override) {
+        super(RecipeType.CAMPFIRE_COOKING, p_250140_, p_251808_, p_249826_, p_251839_, p_251432_, p_251471_, override);
+    }
+
     @Override
     public ItemStack getToastSymbol() {
         return new ItemStack(Blocks.CAMPFIRE);
