--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -23,16 +_,20 @@
 import net.minecraft.world.level.ItemLike;
 
 public final class Ingredient implements StackedContents.IngredientInfo<Holder<Item>>, Predicate<ItemStack> {
-    public static final StreamCodec<RegistryFriendlyByteBuf, Ingredient> CONTENTS_STREAM_CODEC = ByteBufCodecs.holderSet(Registries.ITEM)
-        .map(Ingredient::new, p_360055_ -> p_360055_.values);
-    public static final StreamCodec<RegistryFriendlyByteBuf, Optional<Ingredient>> OPTIONAL_CONTENTS_STREAM_CODEC = ByteBufCodecs.holderSet(Registries.ITEM)
+    public static final StreamCodec<RegistryFriendlyByteBuf, Ingredient> CONTENTS_STREAM_CODEC = net.neoforged.neoforge.common.crafting.IngredientCodecs.streamCodec(ByteBufCodecs.holderSet(Registries.ITEM)
+        .map(Ingredient::new, p_360055_ -> p_360055_.getValuesForSync()));
+    public static final StreamCodec<RegistryFriendlyByteBuf, Optional<Ingredient>> OPTIONAL_CONTENTS_STREAM_CODEC = net.neoforged.neoforge.common.crafting.IngredientCodecs.optionalStreamCodec(ByteBufCodecs.holderSet(Registries.ITEM)
         .map(
             p_360058_ -> p_360058_.size() == 0 ? Optional.empty() : Optional.of(new Ingredient((HolderSet<Item>)p_360058_)),
-            p_360056_ -> p_360056_.map(p_360062_ -> p_360062_.values).orElse(HolderSet.direct())
-        );
+            p_360056_ -> p_360056_.map(p_360062_ -> p_360062_.getValuesForSync()).orElse(HolderSet.direct())
+        ));
     public static final Codec<HolderSet<Item>> NON_AIR_HOLDER_SET_CODEC = HolderSetCodec.create(Registries.ITEM, Item.CODEC, false);
-    public static final Codec<Ingredient> CODEC = ExtraCodecs.nonEmptyHolderSet(NON_AIR_HOLDER_SET_CODEC).xmap(Ingredient::new, p_360061_ -> p_360061_.values);
+    public static final Codec<Ingredient> CODEC = net.neoforged.neoforge.common.crafting.IngredientCodecs.codec(ExtraCodecs.nonEmptyHolderSet(NON_AIR_HOLDER_SET_CODEC).xmap(Ingredient::new, p_360061_ -> p_360061_.values));
     private final HolderSet<Item> values;
+    @org.jetbrains.annotations.Nullable
+    private net.neoforged.neoforge.common.crafting.ICustomIngredient customIngredient = null;
+    @org.jetbrains.annotations.Nullable
+    private List<Holder<Item>> customIngredientValues;
 
     private Ingredient(HolderSet<Item> p_365027_) {
         p_365027_.unwrap().ifRight(p_360057_ -> {
@@ -45,12 +_,23 @@
         this.values = p_365027_;
     }
 
+    public Ingredient(net.neoforged.neoforge.common.crafting.ICustomIngredient customIngredient) {
+        this.values = HolderSet.empty();
+        this.customIngredient = customIngredient;
+    }
+
     public static boolean testOptionalIngredient(Optional<Ingredient> p_362504_, ItemStack p_363604_) {
         return p_362504_.<Boolean>map(p_360060_ -> p_360060_.test(p_363604_)).orElseGet(p_363604_::isEmpty);
     }
 
     @Deprecated
     public Stream<Holder<Item>> items() {
+        if (this.customIngredient != null) {
+            if (this.customIngredientValues == null) {
+                this.customIngredientValues = this.customIngredient.items().toList();
+            }
+            return this.customIngredientValues.stream();
+        }
         return this.values.stream();
     }
 
@@ -59,16 +_,67 @@
     }
 
     public boolean test(ItemStack p_43914_) {
+        if (this.customIngredient != null) {
+            return this.customIngredient.test(p_43914_);
+        }
         return p_43914_.is(this.values);
     }
 
     public boolean acceptsItem(Holder<Item> p_389400_) {
+        if (this.customIngredient != null) {
+            if (this.customIngredientValues == null) {
+                this.customIngredientValues = this.customIngredient.items().toList();
+            }
+            return this.customIngredientValues.contains(p_389400_);
+        }
         return this.values.contains(p_389400_);
     }
 
     @Override
     public boolean equals(Object p_301003_) {
-        return p_301003_ instanceof Ingredient ingredient ? Objects.equals(this.values, ingredient.values) : false;
+        return p_301003_ instanceof Ingredient ingredient ? java.util.Objects.equals(this.customIngredient, ingredient.customIngredient) && Objects.equals(this.values, ingredient.values) : false;
+    }
+
+    @Override
+    public int hashCode() {
+        if (this.customIngredient != null) {
+            return this.customIngredient.hashCode();
+        }
+        return this.values.hashCode();
+    }
+
+    /**
+      * Retrieves the underlying values of this ingredient.
+      * If this is a {@linkplain #isCustom custom ingredient}, an exception is thrown.
+      */
+    public HolderSet<Item> getValues() {
+        if (isCustom()) {
+            throw new IllegalStateException("Cannot retrieve values from custom ingredient!");
+        }
+        return this.values;
+    }
+
+    /**
+     * Retrieves the holder set to use for syncing {@linkplain #isSimple() simple} ingredients
+     */
+    private HolderSet<Item> getValuesForSync() {
+        if (isCustom()) {
+            return HolderSet.direct(this.items().toList());
+        }
+        return this.values;
+    }
+
+    public boolean isSimple() {
+        return this.customIngredient == null || this.customIngredient.isSimple();
+    }
+
+    @org.jetbrains.annotations.Nullable
+    public net.neoforged.neoforge.common.crafting.ICustomIngredient getCustomIngredient() {
+        return this.customIngredient;
+    }
+
+    public boolean isCustom() {
+        return this.customIngredient != null;
     }
 
     public static Ingredient of(ItemLike p_364285_) {
@@ -88,6 +_,9 @@
     }
 
     public SlotDisplay display() {
+        if (this.customIngredient != null) {
+            return this.customIngredient.display();
+        }
         return (SlotDisplay)this.values
             .unwrap()
             .map(SlotDisplay.TagSlotDisplay::new, p_380837_ -> new SlotDisplay.Composite(p_380837_.stream().map(Ingredient::displayForSingleItem).toList()));
