--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -60,6 +_,7 @@
     }
 
     public void finalizeRecipeLoading(FeatureFlagSet enabledFlags) {
+        this.recipes.order(this.recipePriorities); // Neo: order recipes by priority.
         List<SelectableRecipe.SingleInputEntry<StonecutterRecipe>> stonecutterRecipes = new ArrayList<>();
         List<RecipeManager.IngredientCollector> propertySetCollectors = RECIPE_PROPERTY_SETS.entrySet()
             .stream()
@@ -222,6 +_,16 @@
         return recipe -> recipe.getType() == type && recipe instanceof SingleItemRecipe singleItemRecipe
             ? Optional.of(singleItemRecipe.input())
             : Optional.empty();
+    }
+
+    // Neo: expose recipe map
+    public RecipeMap recipeMap() {
+        return this.recipes;
+    }
+
+    // Neo: set recipe priority map
+    public void setPriorityMap(Object2IntMap<ResourceKey<net.minecraft.world.item.crafting.Recipe<?>>> recipePriorities) {
+        this.recipePriorities = recipePriorities;
     }
 
     public interface CachedCheck<I extends RecipeInput, T extends Recipe<I>> {
