--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -53,17 +_,23 @@
         this.hasErrors = false;
         Map<RecipeType<?>, Builder<ResourceLocation, RecipeHolder<?>>> map = Maps.newHashMap();
         Builder<ResourceLocation, RecipeHolder<?>> builder = ImmutableMap.builder();
-        RegistryOps<JsonElement> registryops = this.registries.createSerializationContext(JsonOps.INSTANCE);
+        RegistryOps<JsonElement> registryops = this.makeConditionalOps(); // Neo: add condition context
 
         for(Entry<ResourceLocation, JsonElement> entry : p_44037_.entrySet()) {
             ResourceLocation resourcelocation = entry.getKey();
+            if (resourcelocation.getPath().startsWith("_")) continue; //Forge: filter anything beginning with "_" as it's used for metadata.
 
             try {
                 JsonObject jsonobject = GsonHelper.convertToJsonObject(entry.getValue(), "top element");
-                Recipe<?> recipe = Util.getOrThrow(Recipe.CODEC.parse(registryops, jsonobject), JsonParseException::new);
+                var decoded = Util.getOrThrow(Recipe.CONDITIONAL_CODEC.parse(registryops, jsonobject), JsonParseException::new);
+                decoded.ifPresentOrElse(r -> {
+                Recipe<?> recipe = r.carrier();
                 RecipeHolder<?> recipeholder = new RecipeHolder<>(resourcelocation, recipe);
                 map.computeIfAbsent(recipe.getType(), p_44075_ -> ImmutableMap.builder()).put(resourcelocation, recipeholder);
                 builder.put(resourcelocation, recipeholder);
+                }, () -> {
+                    LOGGER.debug("Skipping loading recipe {} as its conditions were not met", resourcelocation);
+                });
             } catch (IllegalArgumentException | JsonParseException jsonparseexception) {
                 LOGGER.error("Parsing error loading recipe {}", resourcelocation, jsonparseexception);
             }
