--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -61,6 +_,7 @@
     private SelectableRecipe.SingleInputSet<StonecutterRecipe> stonecutterRecipes = SelectableRecipe.SingleInputSet.empty();
     private List<RecipeManager.ServerDisplayInfo> allDisplays = List.of();
     private Map<ResourceKey<Recipe<?>>, List<RecipeManager.ServerDisplayInfo>> recipeToDisplay = Map.of();
+    private Object2IntMap<ResourceKey<net.minecraft.world.item.crafting.Recipe<?>>> recipePriorities = it.unimi.dsi.fastutil.objects.Object2IntMaps.emptyMap(); // Neo: recipe priorities set from RecipePriorityManager.
 
     public RecipeManager(HolderLookup.Provider registries) {
         this.registries = registries;
@@ -69,7 +_,7 @@
     protected RecipeMap prepare(ResourceManager manager, ProfilerFiller profiler) {
         SortedMap<Identifier, Recipe<?>> recipes = new TreeMap<>();
         SimpleJsonResourceReloadListener.scanDirectory(
-            manager, RECIPE_LISTER, this.registries.createSerializationContext(JsonOps.INSTANCE), Recipe.CODEC, recipes
+            manager, RECIPE_LISTER, new net.neoforged.neoforge.common.conditions.ConditionalOps<>(this.registries.createSerializationContext(JsonOps.INSTANCE), getContext()), Recipe.CODEC, recipes // Neo: add condition context
         );
         List<RecipeHolder<?>> recipeHolders = new ArrayList<>(recipes.size());
         recipes.forEach((id, recipe) -> {
@@ -86,6 +_,7 @@
     }
 
     public void finalizeRecipeLoading(FeatureFlagSet enabledFlags) {
+        this.recipes.order(this.recipePriorities); // Neo: order recipes by priority.
         List<SelectableRecipe.SingleInputEntry<StonecutterRecipe>> stonecutterRecipes = new ArrayList<>();
         List<RecipeManager.IngredientCollector> propertySetCollectors = RECIPE_PROPERTY_SETS.entrySet()
             .stream()
@@ -250,6 +_,16 @@
         return recipe -> recipe.getType() == type && recipe instanceof SingleItemRecipe singleItemRecipe
             ? Optional.of(singleItemRecipe.input())
             : Optional.empty();
+    }
+
+    // Neo: expose recipe map
+    public RecipeMap recipeMap() {
+        return this.recipes;
+    }
+
+    // Neo: set recipe priority map
+    public void setPriorityMap(Object2IntMap<ResourceKey<net.minecraft.world.item.crafting.Recipe<?>>> recipePriorities) {
+        this.recipePriorities = recipePriorities;
     }
 
     public interface CachedCheck<I extends RecipeInput, T extends Recipe<I>> {
