--- a/net/minecraft/world/item/crafting/RecipeSerializer.java
+++ b/net/minecraft/world/item/crafting/RecipeSerializer.java
@@ -46,8 +46,10 @@
    RecipeSerializer<SmithingTrimRecipe> SMITHING_TRIM = register("smithing_trim", new SmithingTrimRecipe.Serializer());
    RecipeSerializer<DecoratedPotRecipe> DECORATED_POT_RECIPE = register("crafting_decorated_pot", new SimpleCraftingRecipeSerializer<>(DecoratedPotRecipe::new));
 
+   // Forge: use fromJson with IContext if you need the context
    Codec<T> codec();
 
+   @org.jetbrains.annotations.Nullable
    T fromNetwork(FriendlyByteBuf p_44106_);
 
    void toNetwork(FriendlyByteBuf p_44101_, T p_44102_);
