--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -58,11 +_,24 @@
     public boolean matches(CraftingInput input, Level level) {
         if (input.ingredientCount() != this.ingredients.size()) {
             return false;
+        } else if (!isSimple) {
+            var nonEmptyItems = new java.util.ArrayList<ItemStack>(input.ingredientCount());
+            for (var item : input.items())
+                if (!item.isEmpty())
+                    nonEmptyItems.add(item);
+            return net.neoforged.neoforge.common.util.RecipeMatcher.findMatches(nonEmptyItems, this.ingredients) != null;
         } else {
             return input.size() == 1 && this.ingredients.size() == 1
                 ? this.ingredients.getFirst().test(input.getItem(0))
                 : input.stackedContents().canCraft(this, null);
         }
+    }
+
+    /**
+     * {@return the result of this shapeless recipe or null if it is not static and needs ot be obtained by assembling it}
+     */
+    public @Nullable ItemStackTemplate result() {
+        return result;
     }
 
     public ItemStack assemble(CraftingInput input) {
