--- a/net/minecraft/world/item/crafting/ShapedRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapedRecipe.java
@@ -81,10 +_,20 @@
         return this.pattern.height();
     }
 
+    // Neo: Expose a getter for pattern field.
+    public ShapedRecipePattern getPattern() {
+        return this.pattern;
+    }
+
+    // Neo: Expose a getter for result field without needing a provider param.
+    public ItemStack getResultItem() {
+        return this.result;
+    }
+
     @Override
     public boolean isIncomplete() {
         NonNullList<Ingredient> nonnulllist = this.getIngredients();
-        return nonnulllist.isEmpty() || nonnulllist.stream().filter(p_151277_ -> !p_151277_.isEmpty()).anyMatch(p_151273_ -> p_151273_.getItems().length == 0);
+        return nonnulllist.isEmpty() || nonnulllist.stream().filter(p_151277_ -> !p_151277_.isEmpty()).anyMatch(Ingredient::hasNoItems);
     }
 
     public static class Serializer implements RecipeSerializer<ShapedRecipe> {
