--- a/net/minecraft/world/item/crafting/ShapedRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapedRecipe.java
@@ -20,7 +_,7 @@
 
 public class ShapedRecipe implements CraftingRecipe {
     public final ShapedRecipePattern pattern;
-    final ItemStack result;
+    final net.neoforged.neoforge.common.crafting.result.Result<ItemStack> result;
     final String group;
     final CraftingBookCategory category;
     final boolean showNotification;
@@ -28,6 +_,10 @@
     private PlacementInfo placementInfo;
 
     public ShapedRecipe(String p_272759_, CraftingBookCategory p_273506_, ShapedRecipePattern p_312827_, ItemStack p_272852_, boolean p_312010_) {
+        this(p_272759_, p_273506_, p_312827_, net.neoforged.neoforge.common.crafting.result.ResultWrapper.item(p_272852_), p_312010_);
+    }
+
+    public ShapedRecipe(String p_272759_, CraftingBookCategory p_273506_, ShapedRecipePattern p_312827_, net.neoforged.neoforge.common.crafting.result.Result<ItemStack> p_272852_, boolean p_312010_) {
         this.group = p_272759_;
         this.category = p_273506_;
         this.pattern = p_312827_;
@@ -78,7 +_,7 @@
     }
 
     public ItemStack assemble(CraftingInput p_345201_, HolderLookup.Provider p_335688_) {
-        return this.result.copy();
+        return this.result.resolve();
     }
 
     public int getWidth() {
@@ -96,7 +_,7 @@
                 this.pattern.width(),
                 this.pattern.height(),
                 this.pattern.ingredients().stream().map(p_380107_ -> p_380107_.map(Ingredient::display).orElse(SlotDisplay.Empty.INSTANCE)).toList(),
-                new SlotDisplay.ItemStackSlotDisplay(this.result),
+                this.result.display(),
                 new SlotDisplay.ItemSlotDisplay(Items.CRAFTING_TABLE)
             )
         );
@@ -108,7 +_,7 @@
                     Codec.STRING.optionalFieldOf("group", "").forGetter(p_311729_ -> p_311729_.group),
                     CraftingBookCategory.CODEC.fieldOf("category").orElse(CraftingBookCategory.MISC).forGetter(p_311732_ -> p_311732_.category),
                     ShapedRecipePattern.MAP_CODEC.forGetter(p_311733_ -> p_311733_.pattern),
-                    ItemStack.STRICT_CODEC.fieldOf("result").forGetter(p_311730_ -> p_311730_.result),
+                    net.neoforged.neoforge.common.crafting.result.ResultCodecs.ITEM_STACK_RESULT_CODEC.fieldOf("result").forGetter(p_311730_ -> p_311730_.result),
                     Codec.BOOL.optionalFieldOf("show_notification", true).forGetter(p_311731_ -> p_311731_.showNotification)
                 )
                 .apply(p_340778_, ShapedRecipe::new)
@@ -131,7 +_,7 @@
             String s = p_319998_.readUtf();
             CraftingBookCategory craftingbookcategory = p_319998_.readEnum(CraftingBookCategory.class);
             ShapedRecipePattern shapedrecipepattern = ShapedRecipePattern.STREAM_CODEC.decode(p_319998_);
-            ItemStack itemstack = ItemStack.STREAM_CODEC.decode(p_319998_);
+            net.neoforged.neoforge.common.crafting.result.Result<ItemStack> itemstack = net.neoforged.neoforge.common.crafting.result.ResultCodecs.ITEM_STACK_RESULT_STREAM_CODEC.decode(p_319998_);
             boolean flag = p_319998_.readBoolean();
             return new ShapedRecipe(s, craftingbookcategory, shapedrecipepattern, itemstack, flag);
         }
@@ -140,7 +_,7 @@
             p_320738_.writeUtf(p_320586_.group);
             p_320738_.writeEnum(p_320586_.category);
             ShapedRecipePattern.STREAM_CODEC.encode(p_320738_, p_320586_.pattern);
-            ItemStack.STREAM_CODEC.encode(p_320738_, p_320586_.result);
+            net.neoforged.neoforge.common.crafting.result.ResultCodecs.ITEM_STACK_RESULT_STREAM_CODEC.encode(p_320738_, p_320586_.result);
             p_320738_.writeBoolean(p_320586_.showNotification);
         }
     }
