--- a/net/minecraft/world/item/crafting/SmeltingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmeltingRecipe.java
@@ -8,6 +_,10 @@
         super(RecipeType.SMELTING, p_250200_, p_251114_, p_250340_, p_250306_, p_249577_, p_250030_);
     }
 
+    public SmeltingRecipe(String p_250200_, CookingBookCategory p_251114_, Ingredient p_250340_, ItemStack p_250306_, float p_249577_, int p_250030_, String override) {
+        super(RecipeType.SMELTING, p_250200_, p_251114_, p_250340_, p_250306_, p_249577_, p_250030_, override);
+    }
+
     @Override
     public ItemStack getToastSymbol() {
         return new ItemStack(Blocks.FURNACE);
