--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -22,12 +22,14 @@
     private final ItemStackTemplate result;
     private final List<Ingredient> ingredients;
     private @Nullable PlacementInfo placementInfo;
+    private final boolean isSimple;
 
     public ShapelessRecipe(String group, CraftingBookCategory category, ItemStackTemplate result, List<Ingredient> ingredients) {
         this.group = group;
         this.category = category;
         this.result = result;
         this.ingredients = ingredients;
+        this.isSimple = ingredients.stream().allMatch(Ingredient::isSimple);
     }
 
     @Override
@@ -85,7 +100,7 @@
                     Codec.STRING.optionalFieldOf("group", "").forGetter(o -> o.group),
                     CraftingBookCategory.CODEC.fieldOf("category").orElse(CraftingBookCategory.MISC).forGetter(o -> o.category),
                     ItemStackTemplate.CODEC.fieldOf("result").forGetter(o -> o.result),
-                    Ingredient.CODEC.listOf(1, 9).fieldOf("ingredients").forGetter(o -> o.ingredients)
+                    Codec.lazyInitialized(() -> Ingredient.CODEC.listOf(1, ShapedRecipePattern.maxHeight * ShapedRecipePattern.maxWidth)).fieldOf("ingredients").forGetter(o -> o.ingredients)
                 )
                 .apply(r, ShapelessRecipe::new)
         );
