--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -358,6 +362,7 @@
                 this.targets.main = pass.readsAndWrites(this.targets.main);
                 pass.executes(
                     () -> {
+                        if (this.levelRenderState.customSkyboxRenderer == null || !this.levelRenderState.customSkyboxRenderer.renderSky(levelRenderState, state, cameraState.viewRotationMatrix, () -> RenderSystem.setShaderFog(skyFog))) {
                         RenderSystem.setShaderFog(skyFog);
                         if (state.skybox == DimensionType.Skybox.END) {
                             this.skyRenderer.renderEndSky();
@@ -377,6 +382,10 @@
                                 this.skyRenderer.renderDarkDisc();
                             }
                         }
+                        } // Neo: End of skip if custom sky is in effect
+                        Profiler.get().push("neoforge_render_after_sky");
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterSky(this, this.levelRenderState, null, cameraState.viewRotationMatrix, this.visibleSections));
+                        Profiler.get().pop();
                     }
                 );
             }
