--- a/net/minecraft/client/renderer/entity/state/EntityRenderState.java
+++ b/net/minecraft/client/renderer/entity/state/EntityRenderState.java
@@ -13,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class EntityRenderState {
+public class EntityRenderState extends net.neoforged.neoforge.client.renderstate.BaseRenderState {
     public static final int NO_OUTLINE = 0;
     public EntityType<?> entityType;
     public double x;
@@ -43,6 +_,7 @@
     public ServerHitboxesRenderState serverHitboxesRenderState;
     public float shadowRadius;
     public final List<EntityRenderState.ShadowPiece> shadowPieces = new ArrayList<>();
+    public float partialTick;
 
     public boolean appearsGlowing() {
         return this.outlineColor != 0;
