--- a/net/minecraft/client/renderer/entity/state/ItemClusterRenderState.java
+++ b/net/minecraft/client/renderer/entity/state/ItemClusterRenderState.java
@@ -14,6 +_,7 @@
     public final ItemStackRenderState item = new ItemStackRenderState();
     public int count;
     public int seed;
+    public boolean shouldSpread;
 
     public void extractItemGroupRenderState(Entity entity, ItemStack stack, ItemModelResolver itemModelResolver) {
         itemModelResolver.updateForNonLiving(this.item, stack, ItemDisplayContext.GROUND, entity);
