--- a/net/minecraft/client/particle/BreakingItemParticle.java
+++ b/net/minecraft/client/particle/BreakingItemParticle.java
@@ -33,7 +_,8 @@
 
     protected BreakingItemParticle(ClientLevel level, double x, double y, double z, ItemStack stack) {
         super(level, x, y, z, 0.0, 0.0, 0.0);
-        this.setSprite(Minecraft.getInstance().getItemRenderer().getModel(stack, level, null, 0).getParticleIcon());
+        var model = Minecraft.getInstance().getItemRenderer().getModel(stack, level, null, 0);
+        this.setSprite(model.getOverrides().resolve(model, stack, level, null, 0).getParticleIcon(net.neoforged.neoforge.client.model.data.ModelData.EMPTY));
         this.gravity = 1.0F;
         this.quadSize /= 2.0F;
         this.uo = this.random.nextFloat() * 3.0F;
