--- a/net/minecraft/client/particle/ParticleEngine.java
+++ b/net/minecraft/client/particle/ParticleEngine.java
@@ -28,7 +_,7 @@
         ParticleRenderType.SINGLE_QUADS, ParticleRenderType.ITEM_PICKUP, ParticleRenderType.ELDER_GUARDIANS
     );
     protected ClientLevel level;
-    private final Map<ParticleRenderType, ParticleGroup<?>> particles = Maps.newIdentityHashMap();
+    private final Map<ParticleRenderType, ParticleGroup<?>> particles = Maps.newTreeMap(net.neoforged.neoforge.client.ClientHooks.makeParticleRenderTypeComparator(RENDER_ORDER));
     private final Queue<TrackingEmitter> trackingEmitters = Queues.newArrayDeque();
     private final Queue<Particle> particlesToAdd = Queues.newArrayDeque();
     private final Object2IntOpenHashMap<ParticleLimit> trackedParticleCounts = new Object2IntOpenHashMap<>();
@@ -67,7 +_,7 @@
     ) {
         ParticleProvider<T> particleprovider = (ParticleProvider<T>)this.resourceManager
             .getProviders()
-            .get(BuiltInRegistries.PARTICLE_TYPE.getId(p_107396_.getType()));
+            .get(BuiltInRegistries.PARTICLE_TYPE.getKey(p_107396_.getType()));
         return particleprovider == null
             ? null
             : particleprovider.createParticle(p_107396_, this.level, p_107397_, p_107398_, p_107399_, p_107400_, p_107401_, p_107402_, this.random);
