--- a/net/minecraft/client/particle/ParticleRenderType.java
+++ b/net/minecraft/client/particle/ParticleRenderType.java
@@ -7,14 +_,18 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public record ParticleRenderType(String name, @Nullable RenderType renderType) {
+public record ParticleRenderType(String name, @Nullable RenderType renderType, boolean translucent) {
     public static final ParticleRenderType TERRAIN_SHEET = new ParticleRenderType("TERRAIN_SHEET", RenderType.translucentParticle(TextureAtlas.LOCATION_BLOCKS));
     public static final ParticleRenderType PARTICLE_SHEET_OPAQUE = new ParticleRenderType(
-        "PARTICLE_SHEET_OPAQUE", RenderType.opaqueParticle(TextureAtlas.LOCATION_PARTICLES)
+        "PARTICLE_SHEET_OPAQUE", RenderType.opaqueParticle(TextureAtlas.LOCATION_PARTICLES), false
     );
     public static final ParticleRenderType PARTICLE_SHEET_TRANSLUCENT = new ParticleRenderType(
         "PARTICLE_SHEET_TRANSLUCENT", RenderType.translucentParticle(TextureAtlas.LOCATION_PARTICLES)
     );
     public static final ParticleRenderType CUSTOM = new ParticleRenderType("CUSTOM", null);
     public static final ParticleRenderType NO_RENDER = new ParticleRenderType("NO_RENDER", null);
+
+    public ParticleRenderType(String name, @Nullable RenderType renderType) {
+        this(name, renderType, true);
+    }
 }
