--- a/net/minecraft/client/particle/Particle.java
+++ b/net/minecraft/client/particle/Particle.java
@@ -19,6 +_,7 @@
 @OnlyIn(Dist.CLIENT)
 public abstract class Particle {
     private static final AABB INITIAL_AABB = new AABB(0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+    public static final AABB INFINITE_EXTENT_AABB = net.neoforged.neoforge.client.extensions.IBlockEntityRendererExtension.INFINITE_EXTENT_AABB;
     private static final double MAXIMUM_COLLISION_VELOCITY_SQUARED = Mth.square(100.0);
     protected final ClientLevel level;
     protected double xo;
@@ -243,6 +_,18 @@
 
     public Optional<ParticleGroup> getParticleGroup() {
         return Optional.empty();
+    }
+
+    /**
+     * Returns the bounding box that should be used for particle culling. {@link Particle#INFINITE_EXTENT_AABB} can be
+     * returned for particles that should not be called.
+     */
+    public AABB getRenderBoundingBox(float partialTicks) {
+        return getBoundingBox().inflate(1.0);
+    }
+
+    public Vec3 getPos() {
+        return new Vec3(this.x, this.y, this.z);
     }
 
     @OnlyIn(Dist.CLIENT)
