--- a/net/minecraft/client/particle/PortalParticle.java
+++ b/net/minecraft/client/particle/PortalParticle.java
@@ -91,6 +_,7 @@
             this.x = this.xStart + this.xd * f2;
             this.y = this.yStart + this.yd * f2 + (1.0F - f);
             this.z = this.zStart + this.zd * f2;
+            this.setPos(this.x, this.y, this.z); // FORGE: update the particle's bounding box
         }
     }
 
