--- a/net/minecraft/client/particle/SingleQuadParticle.java
+++ b/net/minecraft/client/particle/SingleQuadParticle.java
@@ -81,6 +_,12 @@
             .endVertex();
     }
 
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(float partialTicks) {
+        float size = getQuadSize(partialTicks);
+        return new net.minecraft.world.phys.AABB(this.x - size, this.y - size, this.z - size, this.x + size, this.y + size, this.z + size);
+    }
+
     public float getQuadSize(float p_107681_) {
         return this.quadSize;
     }
