--- a/net/minecraft/client/particle/TerrainParticle.java
+++ b/net/minecraft/client/particle/TerrainParticle.java
@@ -41,7 +_,7 @@
         this.rCol = 0.6F;
         this.gCol = 0.6F;
         this.bCol = 0.6F;
-        if (!state.is(Blocks.GRASS_BLOCK)) {
+        if (net.neoforged.neoforge.client.extensions.common.IClientBlockExtensions.of(state).areBreakingParticlesTinted(state, level, pos)) {
             int i = Minecraft.getInstance().getBlockColors().getColor(state, level, pos, 0);
             this.rCol *= (float)(i >> 16 & 0xFF) / 255.0F;
             this.gCol *= (float)(i >> 8 & 0xFF) / 255.0F;
@@ -97,8 +_,14 @@
     ) {
         BlockState blockstate = type.getState();
         return !blockstate.isAir() && !blockstate.is(Blocks.MOVING_PISTON) && blockstate.shouldSpawnTerrainParticles()
-            ? new TerrainParticle(level, x, y, z, xSpeed, ySpeed, zSpeed, blockstate)
+            ? new TerrainParticle(level, x, y, z, xSpeed, ySpeed, zSpeed, blockstate).updateSprite(blockstate, type.getPos())
             : null;
+    }
+
+    public TerrainParticle updateSprite(BlockState state, BlockPos pos) { //FORGE: we cannot assume that the x y z of the particles match the block pos of the block.
+        if (pos != null) // There are cases where we are not able to obtain the correct source pos, and need to fallback to the non-model data version
+            this.setSprite(Minecraft.getInstance().getBlockRenderer().getBlockModelShaper().getTexture(state, level, pos));
+        return this;
     }
 
     @OnlyIn(Dist.CLIENT)
