--- a/net/minecraft/client/particle/VibrationSignalParticle.java
+++ b/net/minecraft/client/particle/VibrationSignalParticle.java
@@ -79,6 +_,7 @@
                 this.x = Mth.lerp(alpha, this.x, destination.x());
                 this.y = Mth.lerp(alpha, this.y, destination.y());
                 this.z = Mth.lerp(alpha, this.z, destination.z());
+                this.setPos(this.x, this.y, this.z); // FORGE: Update the particle's bounding box
                 double dx = this.x - destination.x();
                 double dy = this.y - destination.y();
                 double dz = this.z - destination.z();
