--- a/net/minecraft/network/protocol/status/ServerStatus.java
+++ b/net/minecraft/network/protocol/status/ServerStatus.java
@@ -21,7 +_,8 @@
     Optional<ServerStatus.Players> players,
     Optional<ServerStatus.Version> version,
     Optional<ServerStatus.Favicon> favicon,
-    boolean enforcesSecureChat
+    boolean enforcesSecureChat,
+    boolean isModded
 ) {
     public static final Codec<ServerStatus> CODEC = RecordCodecBuilder.create(
         p_337519_ -> p_337519_.group(
@@ -29,10 +_,21 @@
                     ServerStatus.Players.CODEC.lenientOptionalFieldOf("players").forGetter(ServerStatus::players),
                     ServerStatus.Version.CODEC.lenientOptionalFieldOf("version").forGetter(ServerStatus::version),
                     ServerStatus.Favicon.CODEC.lenientOptionalFieldOf("favicon").forGetter(ServerStatus::favicon),
-                    Codec.BOOL.lenientOptionalFieldOf("enforcesSecureChat", Boolean.valueOf(false)).forGetter(ServerStatus::enforcesSecureChat)
+                    Codec.BOOL.optionalFieldOf("enforcesSecureChat", Boolean.valueOf(false)).forGetter(ServerStatus::enforcesSecureChat),
+                    Codec.BOOL.optionalFieldOf("isModded", Boolean.FALSE).forGetter(ServerStatus::isModded)
                 )
                 .apply(p_337519_, ServerStatus::new)
     );
+
+    public ServerStatus(
+        Component description,
+        Optional<ServerStatus.Players> players,
+        Optional<ServerStatus.Version> version,
+        Optional<ServerStatus.Favicon> favicon,
+        boolean enforcesSecureChat
+    ) {
+        this(description, players, version, favicon, enforcesSecureChat, false);
+    }
 
     public static record Favicon(byte[] iconBytes) {
         private static final String PREFIX = "data:image/png;base64,";
