--- a/net/minecraft/network/protocol/status/ServerStatus.java
+++ b/net/minecraft/network/protocol/status/ServerStatus.java
@@ -21,7 +_,8 @@
    Optional<ServerStatus.Players> players,
    Optional<ServerStatus.Version> version,
    Optional<ServerStatus.Favicon> favicon,
-   boolean enforcesSecureChat
+   boolean enforcesSecureChat,
+   Optional<net.neoforged.neoforge.network.ServerStatusPing> neoForgeData
 ) {
    public static final Codec<ServerStatus> CODEC = RecordCodecBuilder.create(
       p_273270_ -> p_273270_.group(
@@ -29,7 +_,8 @@
                ServerStatus.Players.CODEC.optionalFieldOf("players").forGetter(ServerStatus::players),
                ServerStatus.Version.CODEC.optionalFieldOf("version").forGetter(ServerStatus::version),
                ServerStatus.Favicon.CODEC.optionalFieldOf("favicon").forGetter(ServerStatus::favicon),
-               Codec.BOOL.optionalFieldOf("enforcesSecureChat", Boolean.valueOf(false)).forGetter(ServerStatus::enforcesSecureChat)
+               Codec.BOOL.optionalFieldOf("enforcesSecureChat", Boolean.valueOf(false)).forGetter(ServerStatus::enforcesSecureChat),
+               net.neoforged.neoforge.network.ServerStatusPing.CODEC.optionalFieldOf("neoForgeData").forGetter(ServerStatus::neoForgeData)
             )
             .apply(p_273270_, ServerStatus::new)
    );
