--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -403,6 +403,7 @@
 
       this.minecraft.debugRenderer.clear();
       this.minecraft.player.resetPos();
+      net.neoforged.neoforge.client.ForgeHooksClient.firePlayerLogin(this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().connection);
       this.minecraft.player.setId(p_105030_.playerId());
       this.level.addEntity(this.minecraft.player);
       this.minecraft.player.input = new KeyboardInput(this.minecraft.options);
@@ -416,6 +417,7 @@
       this.minecraft.player.setPortalCooldown(commonplayerspawninfo.portalCooldown());
       this.minecraft.gameMode.setLocalMode(commonplayerspawninfo.gameType(), commonplayerspawninfo.previousGameType());
       this.minecraft.options.setServerRenderDistance(p_105030_.chunkRadius());
+      net.neoforged.neoforge.network.NetworkHooks.sendMCRegistryPackets(connection, net.neoforged.neoforge.network.PlayNetworkDirection.PLAY_TO_SERVER);
       this.chatSession = null;
       this.lastSeenMessages = new LastSeenMessagesTracker(20);
       this.messageSignatureCache = MessageSignatureCache.createDefault();
@@ -1110,7 +1112,9 @@
          localplayer1.getAttributes().assignValues(localplayer.getAttributes());
       }
 
+      localplayer1.updateSyncFields(localplayer); // Forge: fix MC-10657
       localplayer1.resetPos();
+      net.neoforged.neoforge.client.ForgeHooksClient.firePlayerRespawn(this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.connection);
       this.level.addEntity(localplayer1);
       localplayer1.setYRot(-180.0F);
       localplayer1.input = new KeyboardInput(this.minecraft.options);
@@ -1230,10 +1234,7 @@
       PacketUtils.ensureRunningOnSameThread(p_104976_, this, this.minecraft);
       BlockPos blockpos = p_104976_.getPos();
       this.minecraft.level.getBlockEntity(blockpos, p_104976_.getType()).ifPresent(p_205557_ -> {
-         CompoundTag compoundtag = p_104976_.getTag();
-         if (compoundtag != null) {
-            p_205557_.load(compoundtag);
-         }
+         p_205557_.onDataPacket(connection, p_104976_);
 
          if (p_205557_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
             ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
@@ -1395,7 +1396,9 @@
    @Override
    public void handleCommands(ClientboundCommandsPacket p_104990_) {
       PacketUtils.ensureRunningOnSameThread(p_104990_, this, this.minecraft);
-      this.commands = new CommandDispatcher<>(p_104990_.getRoot(CommandBuildContext.simple(this.registryAccess, this.enabledFeatures)));
+      var context = CommandBuildContext.simple(this.registryAccess, this.enabledFeatures);
+      this.commands = new CommandDispatcher<>(p_104990_.getRoot(context));
+      this.commands = net.neoforged.neoforge.client.ClientCommandHandler.mergeServerCommands(this.commands, context);
    }
 
    @Override
@@ -1417,6 +1420,7 @@
       ClientRecipeBook clientrecipebook = this.minecraft.player.getRecipeBook();
       clientrecipebook.setupCollections(this.recipeManager.getRecipes(), this.minecraft.level.registryAccess());
       this.minecraft.populateSearchTree(SearchRegistry.RECIPE_COLLECTIONS, clientrecipebook.getCollections());
+      net.neoforged.neoforge.client.ForgeHooksClient.onRecipesUpdated(this.recipeManager);
    }
 
    @Override
@@ -1523,7 +1527,8 @@
          Blocks.rebuildCache();
       }
 
-      CreativeModeTabs.searchTab().rebuildSearchTree();
+      CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(net.minecraft.world.item.CreativeModeTab::rebuildSearchTree);
+      net.neoforged.neoforge.common.MinecraftForge.EVENT_BUS.post(new net.neoforged.neoforge.event.TagsUpdatedEvent(this.registryAccess, true, connection.isMemoryConnection()));
    }
 
    @Override
@@ -1855,7 +1860,7 @@
    }
 
    @Override
-   public void handleCustomPayload(CustomPacketPayload p_295851_) {
+   public void handleCustomPayload(net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket p_295727_, CustomPacketPayload p_295851_) {
       if (p_295851_ instanceof PathfindingDebugPayload pathfindingdebugpayload) {
          this.minecraft
             .debugRenderer
@@ -1922,11 +1927,15 @@
             .gameEventListenerRenderer
             .trackListener(gameeventlistenerdebugpayload.listenerPos(), gameeventlistenerdebugpayload.listenerRange());
       } else {
-         this.handleUnknownCustomPayload(p_295851_);
+         this.handleUnknownCustomPayload(p_295727_, p_295851_);
       }
    }
 
-   private void handleUnknownCustomPayload(CustomPacketPayload p_294389_) {
+   private void handleUnknownCustomPayload(net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket p_295727_, CustomPacketPayload p_294389_) {
+      if (p_294389_ instanceof net.neoforged.neoforge.network.custom.payload.SimplePayload simplePayload && net.neoforged.neoforge.network.NetworkHooks.onCustomPayload(p_295727_, simplePayload, connection)) {
+         return;
+      }
+
       LOGGER.warn("Unknown custom packet payload: {}", p_294389_.id());
    }
 
@@ -2302,6 +2311,8 @@
    }
 
    public void sendChat(String p_249888_) {
+      p_249888_ = net.neoforged.neoforge.client.ForgeHooksClient.onClientSendMessage(p_249888_);
+      if (p_249888_.isEmpty()) return;
       Instant instant = Instant.now();
       long i = Crypt.SaltSupplier.getLong();
       LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
@@ -2311,6 +2322,7 @@
    }
 
    public void sendCommand(String p_250092_) {
+      if (net.neoforged.neoforge.client.ClientCommandHandler.runCommand(p_250092_)) return;
       Instant instant = Instant.now();
       long i = Crypt.SaltSupplier.getLong();
       LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
