--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -368,6 +_,7 @@
     private final ChunkBatchSizeCalculator chunkBatchSizeCalculator = new ChunkBatchSizeCalculator();
     private final PingDebugMonitor pingDebugMonitor;
     private final DebugSampleSubscriber debugSampleSubscriber;
+    private net.neoforged.neoforge.network.connection.ConnectionType connectionType;
     @Nullable
     private LevelLoadStatusManager levelLoadStatusManager;
     private boolean serverEnforcesSecureChat;
@@ -389,7 +_,8 @@
             p_253924_.gui.getChat().restoreState(p_295121_.chatState());
         }
 
-        this.potionBrewing = PotionBrewing.bootstrap(this.enabledFeatures);
+        this.connectionType = p_295121_.connectionType();
+        this.potionBrewing = PotionBrewing.bootstrap(this.enabledFeatures, this.registryAccess);
         this.fuelValues = FuelValues.vanillaBurnTimes(p_295121_.receivedRegistries(), this.enabledFeatures);
     }
 
@@ -452,12 +_,13 @@
 
         this.minecraft.debugRenderer.clear();
         this.minecraft.player.resetPos();
+        net.neoforged.neoforge.client.ClientHooks.firePlayerLogin(this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().connection);
         this.minecraft.player.setId(p_105030_.playerId());
         this.level.addEntity(this.minecraft.player);
         this.minecraft.player.input = new KeyboardInput(this.minecraft.options);
         this.minecraft.gameMode.adjustPlayer(this.minecraft.player);
         this.minecraft.cameraEntity = this.minecraft.player;
-        this.startWaitingForNewLevel(this.minecraft.player, this.level, ReceivingLevelScreen.Reason.OTHER);
+        this.startWaitingForNewLevel(this.minecraft.player, this.level, ReceivingLevelScreen.Reason.OTHER, null, null);
         this.minecraft.player.setReducedDebugInfo(p_105030_.reducedDebugInfo());
         this.minecraft.player.setShowDeathScreen(p_105030_.showDeathScreen());
         this.minecraft.player.setDoLimitedCrafting(p_105030_.doLimitedCrafting());
@@ -902,7 +_,8 @@
                         this.serverCookies,
                         chatcomponent$state,
                         this.customReportDetails,
-                        this.serverLinks
+                        this.serverLinks,
+                        this.connectionType
                     )
                 )
             );
@@ -1217,7 +_,7 @@
             localplayer1 = this.minecraft.gameMode.createPlayer(this.level, localplayer.getStats(), localplayer.getRecipeBook());
         }
 
-        this.startWaitingForNewLevel(localplayer1, this.level, receivinglevelscreen$reason);
+        this.startWaitingForNewLevel(localplayer1, this.level, receivinglevelscreen$reason, localplayer.isDeadOrDying() ? null : resourcekey, localplayer.isDeadOrDying() ? null : resourcekey1);
         localplayer1.setId(localplayer.getId());
         this.minecraft.player = localplayer1;
         if (flag) {
@@ -1245,6 +_,7 @@
             localplayer1.getAttributes().assignBaseValues(localplayer.getAttributes());
         }
 
+        net.neoforged.neoforge.client.ClientHooks.firePlayerRespawn(this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.connection);
         this.level.addEntity(localplayer1);
         localplayer1.input = new KeyboardInput(this.minecraft.options);
         this.minecraft.gameMode.adjustPlayer(localplayer1);
@@ -1391,10 +_,7 @@
         PacketUtils.ensureRunningOnSameThread(p_104976_, this, this.minecraft);
         BlockPos blockpos = p_104976_.getPos();
         this.minecraft.level.getBlockEntity(blockpos, p_104976_.getType()).ifPresent(p_337415_ -> {
-            CompoundTag compoundtag = p_104976_.getTag();
-            if (!compoundtag.isEmpty()) {
-                p_337415_.loadWithComponents(compoundtag, this.registryAccess);
-            }
+            p_337415_.onDataPacket(connection, p_104976_, this.registryAccess);
 
             if (p_337415_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
                 ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
@@ -1505,9 +_,15 @@
         }
     }
 
+    /** @deprecated Neo: use {@link #startWaitingForNewLevel(LocalPlayer, ClientLevel, ReceivingLevelScreen.Reason, ResourceKey, ResourceKey)} instead. */
+    @Deprecated
     private void startWaitingForNewLevel(LocalPlayer p_304688_, ClientLevel p_304528_, ReceivingLevelScreen.Reason p_341690_) {
+        this.startWaitingForNewLevel(p_304688_, p_304528_, p_341690_, null, null);
+    }
+
+    private void startWaitingForNewLevel(LocalPlayer p_304688_, ClientLevel p_304528_, ReceivingLevelScreen.Reason p_341690_, @Nullable ResourceKey<Level> toDimension, @Nullable ResourceKey<Level> fromDimension) {
         this.levelLoadStatusManager = new LevelLoadStatusManager(p_304688_, p_304528_, this.minecraft.levelRenderer);
-        this.minecraft.setScreen(new ReceivingLevelScreen(this.levelLoadStatusManager::levelReady, p_341690_));
+        this.minecraft.setScreen(net.neoforged.neoforge.client.DimensionTransitionScreenManager.getScreen(toDimension, fromDimension).create(this.levelLoadStatusManager::levelReady, p_341690_));
     }
 
     @Override
@@ -1555,7 +_,9 @@
     @Override
     public void handleCommands(ClientboundCommandsPacket p_104990_) {
         PacketUtils.ensureRunningOnSameThread(p_104990_, this, this.minecraft);
-        this.commands = new CommandDispatcher<>(p_104990_.getRoot(CommandBuildContext.simple(this.registryAccess, this.enabledFeatures)));
+        var context = CommandBuildContext.simple(this.registryAccess, this.enabledFeatures);
+        this.commands = new CommandDispatcher<>(p_104990_.getRoot(context));
+        this.commands = net.neoforged.neoforge.client.ClientCommandHandler.mergeServerCommands(this.commands, context);
     }
 
     @Override
@@ -1574,6 +_,11 @@
     public void handleUpdateRecipes(ClientboundUpdateRecipesPacket p_105132_) {
         PacketUtils.ensureRunningOnSameThread(p_105132_, this, this.minecraft);
         this.recipes = new ClientRecipeContainer(p_105132_.itemSets(), p_105132_.stonecutterRecipes());
+
+        // Neo: abuse recipe sync to overwrite fuel values with datamap values after their sync (tag update doesn't fire on initial sync and the constructor is too early)
+        if (this.connectionType.isNeoForge()) {
+            this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registryAccess, this.enabledFeatures);
+        }
     }
 
     @Override
@@ -1698,8 +_,11 @@
         });
         list.forEach(Registry.PendingTags::apply);
         this.fuelValues = FuelValues.vanillaBurnTimes(this.registryAccess, this.enabledFeatures);
-        List<ItemStack> list1 = List.copyOf(CreativeModeTabs.searchTab().getDisplayItems());
-        this.searchTrees.updateCreativeTags(list1);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.TagsUpdatedEvent(this.registryAccess, true, flag));
+        CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(tab -> {
+            List<ItemStack> stacks = List.copyOf(tab.getDisplayItems());
+            this.searchTrees.updateCreativeTags(stacks, net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getTagSearchKey(tab));
+        });
     }
 
     @Override
@@ -2540,6 +_,8 @@
     }
 
     public void sendChat(String p_249888_) {
+        p_249888_ = net.neoforged.neoforge.client.ClientHooks.onClientSendMessage(p_249888_);
+        if (p_249888_.isEmpty()) return;
         Instant instant = Instant.now();
         long i = Crypt.SaltSupplier.getLong();
         LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
@@ -2549,6 +_,7 @@
     }
 
     public void sendCommand(String p_250092_) {
+        if (net.neoforged.neoforge.client.ClientCommandHandler.runCommand(p_250092_)) return;
         SignableCommand<SharedSuggestionProvider> signablecommand = SignableCommand.of(this.parseCommand(p_250092_));
         if (signablecommand.arguments().isEmpty()) {
             this.send(new ServerboundChatCommandPacket(p_250092_));
@@ -2640,6 +_,10 @@
 
     public Scoreboard scoreboard() {
         return this.scoreboard;
+    }
+
+    public net.neoforged.neoforge.network.connection.ConnectionType getConnectionType() {
+        return this.connectionType;
     }
 
     public PotionBrewing potionBrewing() {
