--- a/net/minecraft/client/multiplayer/ClientHandshakePacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientHandshakePacketListenerImpl.java
@@ -168,7 +_,8 @@
                         FeatureFlags.DEFAULT_FLAGS,
                         null,
                         this.serverData,
-                        this.parent
+                        this.parent,
+                        false
                     )
                 )
             );
