--- a/net/minecraft/client/multiplayer/ClientDebugSubscriber.java
+++ b/net/minecraft/client/multiplayer/ClientDebugSubscriber.java
@@ -62,7 +_,8 @@
             addFlag(set, DebugSubscriptions.STRUCTURES, SharedConstants.DEBUG_STRUCTURES);
             addFlag(set, DebugSubscriptions.VILLAGE_SECTIONS, SharedConstants.DEBUG_VILLAGE_SECTIONS);
         }
-
+        // Neo: Allow requesting custom subscriptions
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.AddDebugSubscriptionFlagsEvent((subscription, flag) -> addFlag(set, subscription, flag)));
         return set;
     }
 
