--- a/net/minecraft/client/multiplayer/ClientHandshakePacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientHandshakePacketListenerImpl.java
@@ -166,9 +_,7 @@
                         this.minecraft.getTelemetryManager().createWorldSessionManager(this.newWorld, this.worldLoadDuration, this.minigameName),
                         ClientRegistryLayer.createRegistryAccess().compositeAccess(),
                         FeatureFlags.DEFAULT_FLAGS,
-                        null,
-                        this.serverData,
-                        this.parent
+                            null, this.serverData, this.parent, false
                     )
                 )
             );
