--- a/net/minecraft/client/multiplayer/ClientLevel.java
+++ b/net/minecraft/client/multiplayer/ClientLevel.java
@@ -138,7 +_,7 @@
     private final @Nullable EndFlashState endFlashState;
     private final Minecraft minecraft = Minecraft.getInstance();
     private final List<AbstractClientPlayer> players = Lists.newArrayList();
-    private final List<EnderDragonPart> dragonParts = Lists.newArrayList();
+    private final List<net.neoforged.neoforge.entity.PartEntity<?>> dragonParts = Lists.newArrayList();
     private final Map<MapId, MapItemSavedData> mapData = Maps.newHashMap();
     private int skyFlashTime;
     private final Object2ObjectArrayMap<ColorResolver, BlockTintCache> tintCaches = Util.make(new Object2ObjectArrayMap<>(3), cache -> {
@@ -146,6 +_,7 @@
         cache.put(BiomeColors.FOLIAGE_COLOR_RESOLVER, new BlockTintCache(pos -> this.calculateBlockTint(pos, BiomeColors.FOLIAGE_COLOR_RESOLVER)));
         cache.put(BiomeColors.DRY_FOLIAGE_COLOR_RESOLVER, new BlockTintCache(pos -> this.calculateBlockTint(pos, BiomeColors.DRY_FOLIAGE_COLOR_RESOLVER)));
         cache.put(BiomeColors.WATER_COLOR_RESOLVER, new BlockTintCache(pos -> this.calculateBlockTint(pos, BiomeColors.WATER_COLOR_RESOLVER)));
+        net.neoforged.neoforge.client.ColorResolverManager.registerBlockTintCaches(ClientLevel.this, cache);
     });
     private final ClientChunkCache chunkSource;
     private final Deque<Runnable> lightUpdateQueue = Queues.newArrayDeque();
@@ -157,6 +_,7 @@
     private final EnvironmentAttributeSystem environmentAttributes;
     private final int seaLevel;
     private static final Set<Item> MARKER_PARTICLE_ITEMS = Set.of(Items.BARRIER, Items.LIGHT);
+    private final net.neoforged.neoforge.model.data.ModelDataManager modelDataManager = new net.neoforged.neoforge.model.data.ModelDataManager(this);
 
     public void handleBlockChangedAck(int sequence) {
         if (SharedConstants.DEBUG_BLOCK_BREAK) {
@@ -202,10 +_,15 @@
     @Override
     public boolean setBlock(BlockPos pos, BlockState blockState, @Block.UpdateFlags int updateFlags, int updateLimit) {
         if (this.blockStatePredictionHandler.isPredicting()) {
+            // Neo: Record and store a snapshot in the prediction so that BE data can be restored if the break is denied.
+            // Fixes MC-36093 and permits correct server-side only cancellation of block changes.
+            var snapshot = net.neoforged.neoforge.common.util.BlockSnapshot.create(this.dimension(), this, pos, updateFlags);
+
             BlockState oldState = this.getBlockState(pos);
             boolean success = super.setBlock(pos, blockState, updateFlags, updateLimit);
             if (success) {
                 this.blockStatePredictionHandler.retainKnownServerState(pos, oldState, this.minecraft.player);
+                this.blockStatePredictionHandler.retainSnapshot(pos, snapshot);
             }
 
             return success;
@@ -239,6 +_,7 @@
         this.serverSimulationDistance = serverSimulationDistance;
         this.environmentAttributes = this.addEnvironmentAttributeLayers(EnvironmentAttributeSystem.builder()).build();
         this.updateSkyBrightness();
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Load(this));
     }
 
     private EnvironmentAttributeSystem.Builder addEnvironmentAttributeLayers(EnvironmentAttributeSystem.Builder environmentAttributes) {
@@ -348,7 +_,11 @@
         entity.setOldPosAndRot();
         entity.tickCount++;
         Profiler.get().push(entity.typeHolder()::getRegisteredName);
-        entity.tick();
+        // Neo: Permit cancellation of Entity#tick via EntityTickEvent.Pre
+        if (!net.neoforged.neoforge.event.EventHooks.fireEntityTickPre(entity).isCanceled()) {
+            entity.tick();
+            net.neoforged.neoforge.event.EventHooks.fireEntityTickPost(entity);
+        }
         Profiler.get().pop();
 
         for (Entity passenger : entity.getPassengers()) {
@@ -399,8 +_,10 @@
     }
 
     public void addEntity(Entity entity) {
+        if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityJoinLevelEvent(entity, this)).isCanceled()) return;
         this.removeEntity(entity.getId(), Entity.RemovalReason.DISCARDED);
         this.entityStorage.addEntity(entity);
+        entity.onAddedToLevel();
     }
 
     public void removeEntity(int id, Entity.RemovalReason reason) {
@@ -547,6 +_,13 @@
     public void playSeededSound(
         @Nullable Entity except, double x, double y, double z, Holder<SoundEvent> sound, SoundSource source, float volume, float pitch, long seed
     ) {
+        net.neoforged.neoforge.event.PlayLevelSoundEvent.AtPosition event = net.neoforged.neoforge.event.EventHooks.onPlaySoundAtPosition(this, x, y, z, sound, source, volume, pitch);
+        if (event.isCanceled() || event.getSound() == null) return;
+        sound = event.getSound();
+        source = event.getSource();
+        volume = event.getNewVolume();
+        pitch = event.getNewPitch();
+
         if (except == this.minecraft.player) {
             this.playSound(x, y, z, sound.value(), source, volume, pitch, false, seed);
         }
@@ -556,6 +_,12 @@
     public void playSeededSound(
         @Nullable Entity except, Entity sourceEntity, Holder<SoundEvent> sound, SoundSource source, float volume, float pitch, long seed
     ) {
+        net.neoforged.neoforge.event.PlayLevelSoundEvent.AtEntity event = net.neoforged.neoforge.event.EventHooks.onPlaySoundAtEntity(sourceEntity, sound, source, volume, pitch);
+        if (event.isCanceled() || event.getSound() == null) return;
+        sound = event.getSound();
+        source = event.getSource();
+        volume = event.getNewVolume();
+        pitch = event.getNewPitch();
         if (except == this.minecraft.player) {
             this.minecraft.getSoundManager().play(new EntityBoundSoundInstance(sound.value(), source, volume, pitch, sourceEntity, seed));
         }
@@ -765,7 +_,7 @@
         return this.players;
     }
 
-    public List<EnderDragonPart> dragonParts() {
+    public List<net.neoforged.neoforge.entity.PartEntity<?>> dragonParts() {
         return this.dragonParts;
     }
 
@@ -876,7 +_,7 @@
 
     @Override
     public void addDestroyBlockEffect(BlockPos pos, BlockState blockState) {
-        if (!blockState.isAir() && blockState.shouldSpawnTerrainParticles()) {
+        if (!blockState.isAir() && !net.neoforged.neoforge.client.extensions.common.IClientBlockExtensions.of(blockState).addDestroyEffects(blockState, this, pos, this.minecraft.particleEngine)) {
             VoxelShape shape = blockState.getShape(this, pos);
             double density = 0.25;
             shape.forAllBoxes(
@@ -902,7 +_,7 @@
                                     .add(
                                         new TerrainParticle(
                                             this, pos.getX() + x, pos.getY() + y, pos.getZ() + z, relX - 0.5, relY - 0.5, relZ - 0.5, blockState, pos
-                                        )
+                                        ).updateSprite(blockState, pos)
                                     );
                             }
                         }
@@ -912,9 +_,16 @@
         }
     }
 
+    /**
+     * @deprecated Neo: use {@link #addBreakingBlockEffect(BlockPos, Direction, net.minecraft.world.phys.HitResult)} instead
+     */
     public void addBreakingBlockEffect(BlockPos pos, Direction direction) {
+        this.addBreakingBlockEffect(pos, direction, null);
+    }
+
+    public void addBreakingBlockEffect(BlockPos pos, Direction direction, net.minecraft.world.phys.@Nullable HitResult hitResult) {
         BlockState blockState = this.getBlockState(pos);
-        if (blockState.getRenderShape() != RenderShape.INVISIBLE && blockState.shouldSpawnTerrainParticles()) {
+        if (blockState.getRenderShape() != RenderShape.INVISIBLE && !net.neoforged.neoforge.client.extensions.common.IClientBlockExtensions.of(blockState).addHitEffects(blockState, this, hitResult, this.minecraft.particleEngine)) {
             int x = pos.getX();
             int y = pos.getY();
             int z = pos.getZ();
@@ -947,7 +_,7 @@
                 xp = x + shape.maxX + 0.1F;
             }
 
-            this.minecraft.particleEngine.add(new TerrainParticle(this, xp, yp, zp, 0.0, 0.0, 0.0, blockState, pos).setPower(0.2F).scale(0.6F));
+            this.minecraft.particleEngine.add(new TerrainParticle(this, xp, yp, zp, 0.0, 0.0, 0.0, blockState, pos).updateSprite(blockState, pos).setPower(0.2F).scale(0.6F));
         }
     }
 
@@ -1066,6 +_,7 @@
         }
 
         public void setDifficulty(Difficulty difficulty) {
+            net.neoforged.neoforge.common.CommonHooks.onDifficultyChange(difficulty, this.difficulty);
             this.difficulty = difficulty;
         }
 
@@ -1112,6 +_,9 @@
                     ClientLevel.this.dragonParts.addAll(Arrays.asList(dragon.getSubEntities()));
                     break;
                 default:
+                    if (entity.isMultipartEntity()) {
+                        ClientLevel.this.dragonParts.addAll(Arrays.asList(entity.getParts()));
+                    }
             }
         }
 
@@ -1125,10 +_,38 @@
                     ClientLevel.this.dragonParts.removeAll(Arrays.asList(dragon.getSubEntities()));
                     break;
                 default:
+                    if (entity.isMultipartEntity()) {
+                        ClientLevel.this.dragonParts.removeAll(Arrays.asList(entity.getParts()));
+                    }
             }
+
+            entity.onRemovedFromLevel();
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityLeaveLevelEvent(entity, ClientLevel.this));
         }
 
         public void onSectionChange(Entity entity) {
         }
+    }
+
+    @Override
+    public net.neoforged.neoforge.model.data.ModelDataManager getModelDataManager() {
+        return modelDataManager;
+    }
+
+    @Override
+    public net.neoforged.neoforge.model.data.ModelData getModelData(BlockPos pos) {
+        return modelDataManager.getAt(pos);
+    }
+
+    @Override
+    public float getShade(float normalX, float normalY, float normalZ, boolean shade) {
+        // TODO 1.21.11: Make this extensible / use our dimension special effects replacement.
+        var cardinalLightType = this.dimensionType().cardinalLightType();
+        boolean constantAmbientLight = cardinalLightType == DimensionType.CardinalLightType.NETHER;
+
+        if (!shade)
+            return constantAmbientLight ? 0.9F : 1.0F;
+        float yFactor = constantAmbientLight ? 0.9F : normalY < 0 ? 0.5F : 1.0F;
+        return Math.min(normalX * normalX * 0.6F + normalY * normalY * yFactor + normalZ * normalZ * 0.8F, 1.0F);
     }
 }
