--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -343,6 +_,7 @@
     private final ChunkBatchSizeCalculator chunkBatchSizeCalculator = new ChunkBatchSizeCalculator();
     private final PingDebugMonitor pingDebugMonitor;
     private final DebugSampleSubscriber debugSampleSubscriber;
+    private net.neoforged.neoforge.network.connection.ConnectionType connectionType;
     @Nullable
     private LevelLoadStatusManager levelLoadStatusManager;
     private boolean serverEnforcesSecureChat;
@@ -364,6 +_,7 @@
             p_253924_.gui.getChat().restoreState(p_295121_.chatState());
         }
 
+        this.connectionType = p_295121_.connectionType();
         this.potionBrewing = PotionBrewing.bootstrap(this.enabledFeatures);
     }
 
@@ -425,6 +_,7 @@
 
         this.minecraft.debugRenderer.clear();
         this.minecraft.player.resetPos();
+        net.neoforged.neoforge.client.ClientHooks.firePlayerLogin(this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().connection);
         this.minecraft.player.setId(p_105030_.playerId());
         this.level.addEntity(this.minecraft.player);
         this.minecraft.player.input = new KeyboardInput(this.minecraft.options);
@@ -823,7 +_,8 @@
                         this.serverData,
                         this.postDisconnectScreen,
                         this.serverCookies,
-                        chatcomponent$state
+                        chatcomponent$state,
+                        this.connectionType
                     )
                 )
             );
@@ -1155,6 +_,7 @@
         }
 
         localplayer1.resetPos();
+        net.neoforged.neoforge.client.ClientHooks.firePlayerRespawn(this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.connection);
         this.level.addEntity(localplayer1);
         localplayer1.setYRot(-180.0F);
         localplayer1.input = new KeyboardInput(this.minecraft.options);
@@ -1283,10 +_,7 @@
         PacketUtils.ensureRunningOnSameThread(p_104976_, this, this.minecraft);
         BlockPos blockpos = p_104976_.getPos();
         this.minecraft.level.getBlockEntity(blockpos, p_104976_.getType()).ifPresent(p_337415_ -> {
-            CompoundTag compoundtag = p_104976_.getTag();
-            if (!compoundtag.isEmpty()) {
-                p_337415_.loadWithComponents(compoundtag, this.registryAccess);
-            }
+            p_337415_.onDataPacket(connection, p_104976_, this.registryAccess);
 
             if (p_337415_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
                 ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
@@ -1453,7 +_,9 @@
     @Override
     public void handleCommands(ClientboundCommandsPacket p_104990_) {
         PacketUtils.ensureRunningOnSameThread(p_104990_, this, this.minecraft);
-        this.commands = new CommandDispatcher<>(p_104990_.getRoot(CommandBuildContext.simple(this.registryAccess, this.enabledFeatures)));
+        var context = CommandBuildContext.simple(this.registryAccess, this.enabledFeatures);
+        this.commands = new CommandDispatcher<>(p_104990_.getRoot(context));
+        this.commands = net.neoforged.neoforge.client.ClientCommandHandler.mergeServerCommands(this.commands, context);
     }
 
     @Override
@@ -1475,6 +_,7 @@
         ClientRecipeBook clientrecipebook = this.minecraft.player.getRecipeBook();
         clientrecipebook.setupCollections(this.recipeManager.getRecipes(), this.minecraft.level.registryAccess());
         this.minecraft.populateSearchTree(SearchRegistry.RECIPE_COLLECTIONS, clientrecipebook.getCollections());
+        net.neoforged.neoforge.client.ClientHooks.onRecipesUpdated(this.recipeManager);
     }
 
     @Override
@@ -1906,7 +_,21 @@
     }
 
     @Override
-    public void handleCustomPayload(CustomPacketPayload p_295851_) {
+    public void handleCustomPayload(net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket p_295727_, CustomPacketPayload p_295851_) {
+        if (p_295851_ instanceof net.neoforged.neoforge.network.payload.MinecraftRegisterPayload minecraftRegisterPayload) {
+            this.connectionType = this.connectionType.withMinecraftRegisterPayload();
+            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onMinecraftRegister(this, minecraftRegisterPayload.newChannels());
+            return;
+        }
+        if (p_295851_  instanceof net.neoforged.neoforge.network.payload.MinecraftUnregisterPayload minecraftUnregisterPayload) {
+            this.connectionType = this.connectionType.withMinecraftRegisterPayload();
+            net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onMinecraftUnregister(this, minecraftUnregisterPayload.forgottenChannels());
+            return;
+        }
+        if (this.connectionType.isNotVanilla() && net.neoforged.neoforge.network.registration.NetworkRegistry.getInstance().onModdedPacketAtClient(this, p_295727_)) {
+            return;
+        }
+        PacketUtils.ensureRunningOnSameThread(p_295727_, this, this.minecraft); //Neo: We move this here to ensure that only vanilla packets are handled on the main thread.
         if (p_295851_ instanceof PathfindingDebugPayload pathfindingdebugpayload) {
             this.minecraft
                 .debugRenderer
@@ -2389,6 +_,8 @@
     }
 
     public void sendChat(String p_249888_) {
+        p_249888_ = net.neoforged.neoforge.client.ClientHooks.onClientSendMessage(p_249888_);
+        if (p_249888_.isEmpty()) return;
         Instant instant = Instant.now();
         long i = Crypt.SaltSupplier.getLong();
         LastSeenMessagesTracker.Update lastseenmessagestracker$update = this.lastSeenMessages.generateAndApplyUpdate();
@@ -2398,6 +_,7 @@
     }
 
     public void sendCommand(String p_250092_) {
+        if (net.neoforged.neoforge.client.ClientCommandHandler.runCommand(p_250092_)) return;
         SignableCommand<SharedSuggestionProvider> signablecommand = SignableCommand.of(this.parseCommand(p_250092_));
         if (signablecommand.arguments().isEmpty()) {
             this.send(new ServerboundChatCommandPacket(p_250092_));
@@ -2472,6 +_,10 @@
 
     public Scoreboard scoreboard() {
         return this.scoreboard;
+    }
+
+    public net.neoforged.neoforge.network.connection.ConnectionType getConnectionType() {
+        return this.connectionType;
     }
 
     public PotionBrewing potionBrewing() {
