--- a/net/minecraft/client/multiplayer/CommonListenerCookie.java
+++ b/net/minecraft/client/multiplayer/CommonListenerCookie.java
@@ -15,8 +_,17 @@
     WorldSessionTelemetryManager telemetryManager,
     RegistryAccess.Frozen receivedRegistries,
     FeatureFlagSet enabledFeatures,
+    boolean isModdedConnection,
     @Nullable String serverBrand,
     @Nullable ServerData serverData,
     @Nullable Screen postDisconnectScreen
 ) {
+
+    /**
+     * @deprecated Use {@link #CommonListenerCookie(GameProfile, WorldSessionTelemetryManager, RegistryAccess.Frozen, FeatureFlagSet, boolean, String, ServerData, Screen)} instead, to indicate whether the connection is modded.
+     */
+    @Deprecated
+    public CommonListenerCookie(GameProfile localGameProfile, WorldSessionTelemetryManager telemetryManager, RegistryAccess.Frozen receivedRegistries, FeatureFlagSet enabledFeatures, @Nullable String serverBrand, @Nullable ServerData serverData, @Nullable Screen postDisconnectScreen) {
+        this(localGameProfile, telemetryManager, receivedRegistries, enabledFeatures, false, serverBrand, serverData, postDisconnectScreen);
+    }
 }
