--- a/net/minecraft/client/multiplayer/CommonListenerCookie.java
+++ b/net/minecraft/client/multiplayer/CommonListenerCookie.java
@@ -22,6 +_,25 @@
     @Nullable ServerData serverData,
     @Nullable Screen postDisconnectScreen,
     Map<ResourceLocation, byte[]> serverCookies,
-    @Nullable ChatComponent.State chatState
+    @Nullable ChatComponent.State chatState,
+    net.neoforged.neoforge.network.connection.ConnectionType connectionType
 ) {
+    /**
+     * @deprecated Use {@link #CommonListenerCookie(GameProfile, WorldSessionTelemetryManager, RegistryAccess.Frozen, FeatureFlagSet, String, ServerData, Screen, Map, ChatComponent.State, net.neoforged.neoforge.network.connection.ConnectionType)}
+     * instead,to indicate whether the connection is modded.
+     */
+    @Deprecated
+    public CommonListenerCookie(
+            GameProfile localGameProfile,
+            WorldSessionTelemetryManager telemetryManager,
+            RegistryAccess.Frozen receivedRegistries,
+            FeatureFlagSet enabledFeatures,
+            @Nullable String serverBrand,
+            @Nullable ServerData serverData,
+            @Nullable Screen postDisconnectScreen,
+            Map<ResourceLocation, byte[]> serverCookies,
+            @Nullable ChatComponent.State chatState
+    ) {
+        this(localGameProfile, telemetryManager, receivedRegistries, enabledFeatures, serverBrand, serverData, postDisconnectScreen, serverCookies, chatState, net.neoforged.neoforge.network.connection.ConnectionType.VANILLA);
+    }
 }
