--- a/net/minecraft/client/multiplayer/KnownPacksManager.java
+++ b/net/minecraft/client/multiplayer/KnownPacksManager.java
@@ -43,6 +_,12 @@
             }
         }
 
+        if (response.size() > 1024) {
+            response = response.subList(0, 1024);
+            selectedPacks = selectedPacks.subList(0, 1024);
+            com.mojang.logging.LogUtils.getLogger().warn("NeoForge: too many KnownPacks requested; only the first 1024 will be sent via KnownPack, the rest will be synced normally");
+        }
+
         this.repository.setSelected(selectedPacks);
         return response;
     }
