--- a/net/minecraft/client/multiplayer/ServerData.java
+++ b/net/minecraft/client/multiplayer/ServerData.java
@@ -34,6 +_,7 @@
     private byte[] iconBytes;
     private ServerData.Type type;
     private ServerData.State state = ServerData.State.INITIAL;
+    public net.neoforged.neoforge.client.ExtendedServerListData neoForgeData = null;
 
     public ServerData(String name, String ip, ServerData.Type type) {
         this.name = name;
