--- a/net/minecraft/client/multiplayer/TagCollector.java
+++ b/net/minecraft/client/multiplayer/TagCollector.java
@@ -44,5 +_,6 @@
             this.applyTags(registryAccess, p_326446_ -> true);
             refreshBuiltInTagDependentData();
         }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.TagsUpdatedEvent(registryAccess, true, isMemoryConnection));
     }
 }
