--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -27,8 +_,13 @@
         this.add(p_38708_, FoodConstants.saturationByModifier(p_38708_, p_38709_));
     }
 
+    // Use the LivingEntity sensitive version in favour of this.
+    @Deprecated
     public void eat(ItemStack p_38714_) {
-        FoodProperties foodproperties = p_38714_.get(DataComponents.FOOD);
+        this.eat(p_38714_, null);
+    }
+    public void eat(ItemStack p_38714_, @org.jetbrains.annotations.Nullable net.minecraft.world.entity.LivingEntity entity) {
+        FoodProperties foodproperties = p_38714_.getFoodProperties(entity);
         if (foodproperties != null) {
             this.add(foodproperties.nutrition(), foodproperties.saturation());
         }
