--- a/net/minecraft/client/resources/model/geometry/BakedQuad.java
+++ b/net/minecraft/client/resources/model/geometry/BakedQuad.java
@@ -28,11 +_,17 @@
     long packedUV2,
     long packedUV3,
     Direction direction,
-    BakedQuad.MaterialInfo materialInfo
+    BakedQuad.MaterialInfo materialInfo,
+    net.neoforged.neoforge.client.model.quad.BakedNormals bakedNormals,
+    net.neoforged.neoforge.client.model.quad.BakedColors bakedColors
 ) {
     public static final int VERTEX_COUNT = 4;
     public static final int FLAG_TRANSLUCENT = 1;
     public static final int FLAG_ANIMATED = 2;
+
+    public BakedQuad(Vector3fc position0, Vector3fc position1, Vector3fc position2, Vector3fc position3, long packedUV0, long packedUV1, long packedUV2, long packedUV3, Direction direction, BakedQuad.MaterialInfo materialInfo) {
+        this(position0, position1, position2, position3, packedUV0, packedUV1, packedUV2, packedUV3, direction, materialInfo, net.neoforged.neoforge.client.model.quad.BakedNormals.UNSPECIFIED, net.neoforged.neoforge.client.model.quad.BakedColors.DEFAULT);
+    }
 
     public Vector3fc position(int vertex) {
         return switch (vertex) {
