--- a/net/minecraft/client/resources/model/geometry/UnbakedGeometry.java
+++ b/net/minecraft/client/resources/model/geometry/UnbakedGeometry.java
@@ -14,5 +_,7 @@
 public interface UnbakedGeometry extends UnbakedGeometryExtension {
     UnbakedGeometry EMPTY = (var0, var1, var2, var3) -> QuadCollection.EMPTY;
 
+    /// @deprecated Neo: Use [#bake(TextureSlots, ModelBaker, ModelState, ModelDebugName, net.minecraft.util.context.ContextMap)].
+    @Deprecated
     QuadCollection bake(TextureSlots textureSlots, ModelBaker modelBaker, ModelState modelState, ModelDebugName name);
 }
