--- a/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
+++ b/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
@@ -65,27 +_,32 @@
         int outlineColor,
         int order
     ) {
+        net.neoforged.neoforge.client.extensions.common.IClientItemExtensions extensions = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(itemStack);
+        model = extensions.getGenericArmorModel(itemStack, layerType, model);
         EquipmentClientInfo equipmentInfo = this.equipmentAssets.get(equipmentAssetId);
         List<EquipmentClientInfo.Layer> layers = equipmentInfo.getLayers(layerType);
         if (!layers.isEmpty()) {
-            int dyeColor = DyedItemColor.getOrDefault(itemStack, 0);
+            int dyeColor = extensions.getDefaultDyeColor(itemStack);
             boolean hasFoil = itemStack.hasFoil();
             ArmorTrim trim = itemStack.get(DataComponents.TRIM);
             boolean hasTrim = trim != null && layerType != EquipmentClientInfo.LayerType.HUMANOID_BABY;
             boolean renderShaderGlint = hasFoil && !hasTrim;
             int nextOrder = order;
 
+            int idx = 0;
             for (EquipmentClientInfo.Layer layer : layers) {
-                int color = getColorForLayer(layer, dyeColor);
+                int color = extensions.getArmorLayerTintColor(itemStack, layer, idx, dyeColor);
                 if (color != 0) {
                     Identifier layerTexture = layer.usePlayerTexture() && playerTextureOverride != null
                         ? playerTextureOverride
                         : this.layerTextureLookup.apply(new EquipmentLayerRenderer.LayerTextureKey(layerType, layer));
+                    layerTexture = net.neoforged.neoforge.client.ClientHooks.getArmorTexture(itemStack, layerType, layer, layerTexture);
                     RenderType renderType = renderShaderGlint ? RenderTypes.armorCutoutNoCullGlint(layerTexture) : RenderTypes.armorCutoutNoCull(layerTexture);
                     submitNodeCollector.order(nextOrder++)
                         .submitModel(model, state, poseStack, renderType, lightCoords, OverlayTexture.NO_OVERLAY, color, null, outlineColor);
                     renderShaderGlint = false;
                 }
+                idx++;
             }
 
             if (hasTrim) {
