--- a/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
+++ b/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
@@ -66,18 +_,22 @@
         int p_435821_,
         int p_436591_
     ) {
+        net.neoforged.neoforge.client.extensions.common.IClientItemExtensions extensions = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(p_371670_);
+        p_371731_ = extensions.getGenericArmorModel(p_371670_, p_387484_, p_371731_);
         List<EquipmentClientInfo.Layer> list = this.equipmentAssets.get(p_387603_).getLayers(p_387484_);
         if (!list.isEmpty()) {
-            int i = DyedItemColor.getOrDefault(p_371670_, 0);
+            int i = extensions.getDefaultDyeColor(p_371670_);
             boolean flag = p_371670_.hasFoil();
             int j = p_436591_;
 
+            int idx = 0;
             for (EquipmentClientInfo.Layer equipmentclientinfo$layer : list) {
-                int k = getColorForLayer(equipmentclientinfo$layer, i);
+                int k = extensions.getArmorLayerTintColor(p_371670_, equipmentclientinfo$layer, idx, i);
                 if (k != 0) {
                     ResourceLocation resourcelocation = equipmentclientinfo$layer.usePlayerTexture() && p_371639_ != null
                         ? p_371639_
                         : this.layerTextureLookup.apply(new EquipmentLayerRenderer.LayerTextureKey(p_387484_, equipmentclientinfo$layer));
+                    resourcelocation = net.neoforged.neoforge.client.ClientHooks.getArmorTexture(p_371670_, p_387484_, equipmentclientinfo$layer, resourcelocation);
                     p_435795_.order(j++)
                         .submitModel(
                             p_371731_,
@@ -100,6 +_,7 @@
 
                     flag = false;
                 }
+                idx++;
             }
 
             ArmorTrim armortrim = p_371670_.get(DataComponents.TRIM);
