--- a/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
+++ b/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
@@ -67,18 +_,22 @@
         int outlineColor,
         int order
     ) {
+        net.neoforged.neoforge.client.extensions.common.IClientItemExtensions extensions = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(itemStack);
+        model = extensions.getGenericArmorModel(itemStack, layerType, model);
         List<EquipmentClientInfo.Layer> layers = this.equipmentAssets.get(equipmentAssetId).getLayers(layerType);
         if (!layers.isEmpty()) {
-            int dyeColor = DyedItemColor.getOrDefault(itemStack, 0);
+            int dyeColor = extensions.getDefaultDyeColor(itemStack);
             boolean renderFoil = itemStack.hasFoil();
             int nextOrder = order;
 
+            int idx = 0;
             for (EquipmentClientInfo.Layer layer : layers) {
-                int color = getColorForLayer(layer, dyeColor);
+                int color = extensions.getArmorLayerTintColor(itemStack, layer, idx, dyeColor);
                 if (color != 0) {
                     Identifier layerTexture = layer.usePlayerTexture() && playerTextureOverride != null
                         ? playerTextureOverride
                         : this.layerTextureLookup.apply(new EquipmentLayerRenderer.LayerTextureKey(layerType, layer));
+                    layerTexture = net.neoforged.neoforge.client.ClientHooks.getArmorTexture(itemStack, layerType, layer, layerTexture);
                     submitNodeCollector.order(nextOrder++)
                         .submitModel(
                             model,
@@ -110,6 +_,7 @@
 
                     renderFoil = false;
                 }
+                idx++;
             }
 
             ArmorTrim trim = itemStack.get(DataComponents.TRIM);
