--- a/net/minecraft/client/renderer/entity/layers/SnowGolemHeadLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/SnowGolemHeadLayer.java
@@ -43,6 +_,9 @@
                 RenderType renderType = state.appearsGlowing() && state.isInvisible
                     ? RenderTypes.outline(TextureAtlas.LOCATION_BLOCKS)
                     : ItemBlockRenderTypes.getRenderType(pumpkinBlockState);
+                if (!state.appearsGlowing() || !state.isInvisible) {
+                    submitNodeCollector.submitBlockModel(poseStack, pumpkinBlockState, model, 0.0F, 0.0F, 0.0F, lightCoords, overlayCoords, state.outlineColor);
+                } else
                 submitNodeCollector.submitBlockModel(poseStack, renderType, model, 0.0F, 0.0F, 0.0F, lightCoords, overlayCoords, state.outlineColor);
                 poseStack.popPose();
             }
