--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -338,6 +_,10 @@
         });
     }
 
+    public static java.util.Set<ResourceLocation> getVanillaRegistrationOrder() {
+        return java.util.Collections.unmodifiableSet(LOADERS.keySet());
+    }
+
     @FunctionalInterface
     interface RegistryBootstrap<T> {
         Object run(Registry<T> p_260128_);
