--- a/net/minecraft/core/registries/Registries.java
+++ b/net/minecraft/core/registries/Registries.java
@@ -309,10 +_,10 @@
     }
 
     public static String elementsDirPath(ResourceKey<? extends Registry<?>> p_350960_) {
-        return p_350960_.identifier().getPath();
+        return net.neoforged.neoforge.common.CommonHooks.prefixNamespace(p_350960_.identifier());
     }
 
     public static String tagsDirPath(ResourceKey<? extends Registry<?>> p_350980_) {
-        return "tags/" + p_350980_.identifier().getPath();
+        return "tags/" + net.neoforged.neoforge.common.CommonHooks.prefixNamespace(p_350980_.identifier());
     }
 }
