--- a/net/minecraft/core/registries/Registries.java
+++ b/net/minecraft/core/registries/Registries.java
@@ -249,10 +_,10 @@
     }
 
     public static String elementsDirPath(ResourceKey<? extends Registry<?>> registryKey) {
-        return registryKey.location().getPath();
+        return net.neoforged.neoforge.common.CommonHooks.prefixNamespace(registryKey.location());
     }
 
     public static String tagsDirPath(ResourceKey<? extends Registry<?>> registryKey) {
-        return "tags/" + registryKey.location().getPath();
+        return "tags/" + net.neoforged.neoforge.common.CommonHooks.prefixNamespace(registryKey.location());
     }
 }
