--- a/net/minecraft/world/level/storage/loot/LootPool.java
+++ b/net/minecraft/world/level/storage/loot/LootPool.java
@@ -119,6 +173,8 @@
         private final ImmutableList.Builder<Holder<LootItemFunction>> functions = ImmutableList.builder();
         private Holder<NumberProvider> rolls = ConstantValue.exactly(1.0F);
         private Holder<NumberProvider> bonusRolls = ConstantValue.exactly(0.0F);
+        @org.jspecify.annotations.Nullable
+        private String name;
 
         public LootPool.Builder setRolls(Holder<NumberProvider> rolls) {
             this.rolls = rolls;
