--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -342,9 +_,9 @@
     protected void playStepSound(BlockPos pos, BlockState blockState) {
         if (!blockState.liquid()) {
             BlockState aboveState = this.level().getBlockState(pos.above());
-            SoundType soundType = blockState.getSoundType();
+            SoundType soundType = blockState.getSoundType(level(), pos, this);
             if (aboveState.is(Blocks.SNOW)) {
-                soundType = aboveState.getSoundType();
+                soundType = aboveState.getSoundType(level(), pos, this);
             }
 
             if (this.isVehicle() && this.canGallop) {
@@ -773,6 +_,7 @@
         Vec3 movement = this.getDeltaMovement();
         this.setDeltaMovement(movement.x, impulse, movement.z);
         this.needsSync = true;
+        net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
         if (input.z > 0.0) {
             float sin = Mth.sin(this.getYRot() * (float) (Math.PI / 180.0));
             float cos = Mth.cos(this.getYRot() * (float) (Math.PI / 180.0));
@@ -1022,6 +_,11 @@
 
         this.randomizeAttributes(level.getRandom());
         return super.finalizeSpawn(level, difficulty, spawnReason, groupData);
+    }
+
+    // Neo: Inventory getter
+    public net.minecraft.world.Container getInventory() {
+        return this.inventory;
     }
 
     public boolean hasInventoryChanged(Container oldInventory) {
