--- a/net/minecraft/client/renderer/blockentity/state/ChestRenderState.java
+++ b/net/minecraft/client/renderer/blockentity/state/ChestRenderState.java
@@ -10,6 +_,7 @@
     public float open;
     public float angle;
     public ChestRenderState.ChestMaterialType material = ChestRenderState.ChestMaterialType.REGULAR;
+    public net.minecraft.client.resources.model.@org.jspecify.annotations.Nullable Material customMaterial;
 
     @OnlyIn(Dist.CLIENT)
     public static enum ChestMaterialType {
