--- a/net/minecraft/world/level/storage/loot/LootContext.java
+++ b/net/minecraft/world/level/storage/loot/LootContext.java
@@ -26,6 +_,14 @@
     private final RandomSource random;
     private final HolderGetter.Provider lootDataResolver;
     private final Set<LootContext.VisitedEntry<?>> visitedElements = Sets.newLinkedHashSet();
+    // Neo: Keep track of the original loot table ID through modifications
+    @Nullable
+    private Identifier queriedLootTableId;
+
+    private LootContext(LootParams params, RandomSource random, HolderGetter.Provider lootDataResolver, @Nullable Identifier queriedLootTableId) {
+        this(params, random, lootDataResolver);
+        this.queriedLootTableId = queriedLootTableId;
+    }
 
     private LootContext(LootParams params, RandomSource random, HolderGetter.Provider lootDataResolver) {
         this.params = params;
@@ -73,6 +_,16 @@
         return this.params.getLevel();
     }
 
+    public void setQueriedLootTableId(@org.jspecify.annotations.Nullable Identifier queriedLootTableId) {
+        if (this.queriedLootTableId == null && queriedLootTableId != null) {
+            this.queriedLootTableId = queriedLootTableId;
+        }
+    }
+
+    public Identifier getQueriedLootTableId() {
+        return java.util.Objects.requireNonNullElse(this.queriedLootTableId, net.neoforged.neoforge.common.loot.LootTableIdCondition.UNKNOWN_LOOT_TABLE);
+    }
+
     public static LootContext.VisitedEntry<LootTable> createVisitedEntry(LootTable table) {
         return new LootContext.VisitedEntry<>(LootDataType.TABLE, table);
     }
@@ -114,6 +_,14 @@
     public static class Builder {
         private final LootParams params;
         private @Nullable RandomSource random;
+        @Nullable
+        private Identifier queriedLootTableId; // Neo: correctly pass around loot table ID with copy constructor
+
+        public Builder(LootContext context) {
+            this.params = context.params;
+            this.random = context.random;
+            this.queriedLootTableId = context.queriedLootTableId;
+        }
 
         public Builder(LootParams params) {
             this.params = params;
@@ -132,6 +_,11 @@
             return this;
         }
 
+        public LootContext.Builder withQueriedLootTableId(Identifier queriedLootTableId) {
+            this.queriedLootTableId = queriedLootTableId;
+            return this;
+        }
+
         public ServerLevel getLevel() {
             return this.params.getLevel();
         }
@@ -140,7 +_,7 @@
             ServerLevel level = this.getLevel();
             MinecraftServer server = level.getServer();
             RandomSource random = Optional.ofNullable(this.random).or(() -> randomSequenceKey.map(server::getRandomSequence)).orElseGet(level::getRandom);
-            return new LootContext(this.params, random, server.reloadableRegistries().lookup());
+            return new LootContext(this.params, random, server.reloadableRegistries().lookup(), this.queriedLootTableId);
         }
     }
 
@@ -164,6 +_,12 @@
         @Override
         public ContextKey<? extends Entity> contextParam() {
             return this.param;
+        }
+
+        // Neo: This method is patched in to expose the same name used in getByName so that ContextNbtProvider#forEntity serializes it properly
+        // TODO 1.21.11: To be consistent with getParam() being renamed to contextParam(), consider renaming to name()
+        public String getName() {
+            return this.name;
         }
 
         public static LootContext.EntityTarget getByName(String name) {
