--- a/net/minecraft/data/Main.java
+++ b/net/minecraft/data/Main.java
@@ -85,10 +88,17 @@
             boolean dev = allOptions || optionSet.has(devOption);
             boolean reports = allOptions || optionSet.has(reportsOption);
             Collection<Path> input = optionSet.valuesOf(inputOption).stream().map(x$0 -> Paths.get(x$0)).toList();
-            DataGenerator generator = new DataGenerator(output, SharedConstants.getCurrentVersion(), true);
+            java.util.Set<String> mods = new java.util.HashSet<>(optionSet.valuesOf(mod));
+            boolean isFlat = mods.isEmpty() || optionSet.has(flat);
+            boolean validate = optionSet.has(validateSpec);
+            DataGenerator generator = new DataGenerator(isFlat ? output : output.resolve("minecraft"), SharedConstants.getCurrentVersion(), true);
+            java.util.Collection<Path> existingPacks = optionSet.valuesOf(existing).stream().map(Paths::get).toList();
+            if (mods.contains("minecraft") || mods.isEmpty()) {
             addServerProviders(generator, input, server, dev, reports);
-            generator.run();
             Util.shutdownExecutors();
+            }
+            net.neoforged.neoforge.data.loading.DatagenModLoader.begin(mods, output, input, existingPacks, dev, reports, validate, isFlat, () -> {},
+                    net.neoforged.neoforge.data.event.GatherDataEvent.Server::new, generator, packConsumer -> {});
         } else {
             parser.printHelpOn(System.out);
         }
