--- a/net/minecraft/resources/HolderSetCodec.java
+++ b/net/minecraft/resources/HolderSetCodec.java
@@ -22,6 +_,8 @@
     private final Codec<Holder<E>> elementCodec;
     private final Codec<List<Holder<E>>> homogenousListCodec;
     private final Codec<Either<TagKey<E>, List<Holder<E>>>> registryAwareCodec;
+    private final Codec<net.neoforged.neoforge.registries.holdersets.ICustomHolderSet<E>> forgeDispatchCodec;
+    private final Codec<Either<net.neoforged.neoforge.registries.holdersets.ICustomHolderSet<E>, Either<TagKey<E>, List<Holder<E>>>>> combinedCodec;
 
     private static <E> Codec<List<Holder<E>>> homogenousList(Codec<Holder<E>> p_206668_, boolean p_206669_) {
         Codec<List<Holder<E>>> codec = ExtraCodecs.validate(p_206668_.listOf(), ExtraCodecs.ensureHomogenous(Holder::kind));
@@ -43,6 +_,10 @@
         this.elementCodec = p_206661_;
         this.homogenousListCodec = homogenousList(p_206661_, p_206662_);
         this.registryAwareCodec = Codec.either(TagKey.hashedCodec(p_206660_), this.homogenousListCodec);
+        // FORGE: make registry-specific dispatch codec and make forge-or-vanilla either codec
+        this.forgeDispatchCodec = net.neoforged.neoforge.registries.NeoForgeRegistries.HOLDER_SET_TYPES.byNameCodec()
+             .dispatch(net.neoforged.neoforge.registries.holdersets.ICustomHolderSet::type, type -> type.makeCodec(p_206660_, p_206661_, p_206662_));
+        this.combinedCodec = new ExtraCodecs.EitherCodec<>(this.forgeDispatchCodec, this.registryAwareCodec);
     }
 
     @Override
@@ -51,12 +_,12 @@
             Optional<HolderGetter<E>> optional = registryops.getter(this.registryKey);
             if (optional.isPresent()) {
                 HolderGetter<E> holdergetter = optional.get();
-                return this.registryAwareCodec
-                    .decode(p_206696_, p_206697_)
+                // NEO: use the wrapped codec to decode custom/tag/list instead of just tag/list
+                return this.combinedCodec.decode(p_206696_, p_206697_)
                     .flatMap(
                         p_332561_ -> {
                             DataResult<HolderSet<E>> dataresult = p_332561_.getFirst()
-                                .map(p_332559_ -> lookupTag(holdergetter, p_332559_), p_332564_ -> DataResult.success(HolderSet.direct(p_332564_)));
+                                    .map(DataResult::success, tagOrList -> tagOrList.map(p_332559_ -> lookupTag(holdergetter, p_332559_), p_332564_ -> DataResult.success(HolderSet.direct(p_332564_))));
                             return dataresult.map(p_332563_ -> Pair.of(p_332563_, (T)p_332561_.getSecond()));
                         }
                     );
@@ -80,6 +_,9 @@
                     return DataResult.error(() -> "HolderSet " + p_206674_ + " is not valid in current registry set");
                 }
 
+                // FORGE: use the dispatch codec to encode custom holdersets, otherwise fall back to vanilla tag/list
+                if (p_206674_ instanceof net.neoforged.neoforge.registries.holdersets.ICustomHolderSet<E> customHolderSet)
+                     return this.forgeDispatchCodec.encode(customHolderSet, p_206675_, p_206676_);
                 return this.registryAwareCodec.encode(p_206674_.unwrap().mapRight(List::copyOf), p_206675_, p_206676_);
             }
         }
