--- a/net/minecraft/resources/Identifier.java
+++ b/net/minecraft/resources/Identifier.java
@@ -150,6 +_,12 @@
         return result;
     }
 
+    // Normal compare sorts by path first, this compares namespace first.
+    public int compareNamespaced(Identifier o) {
+        int ret = this.namespace.compareTo(o.namespace);
+        return ret != 0 ? ret : this.path.compareTo(o.path);
+    }
+
     public Path resolveAgainst(Path root) {
         return root.resolve(this.getNamespace(), new String[]{this.getPath()});
     }
