--- a/net/minecraft/resources/RegistryDataLoader.java
+++ b/net/minecraft/resources/RegistryDataLoader.java
@@ -130,7 +_,7 @@
     public static final List<RegistryDataLoader.RegistryData<?>> DIMENSION_REGISTRIES = List.of(
         new RegistryDataLoader.RegistryData<>(Registries.LEVEL_STEM, LevelStem.CODEC)
     );
-    public static final List<RegistryDataLoader.RegistryData<?>> SYNCHRONIZED_REGISTRIES = List.of(
+    public static final List<RegistryDataLoader.RegistryData<?>> SYNCHRONIZED_REGISTRIES = net.neoforged.neoforge.registries.DataPackRegistriesHooks.grabNetworkableRegistries(List.of(
         new RegistryDataLoader.RegistryData<>(Registries.BIOME, Biome.NETWORK_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.CHAT_TYPE, ChatType.DIRECT_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.TRIM_PATTERN, TrimPattern.DIRECT_CODEC),
@@ -159,7 +_,7 @@
         new RegistryDataLoader.RegistryData<>(Registries.DIALOG, Dialog.DIRECT_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.WORLD_CLOCK, WorldClock.DIRECT_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.TIMELINE, Timeline.NETWORK_CODEC)
-    );
+    )); // Neo: Keep the list so custom registries can be added later
 
     public static CompletableFuture<RegistryAccess.Frozen> load(
         ResourceManager resourceManager,
@@ -173,7 +_,7 @@
                 return new ResourceManagerRegistryLoadTask<>(data, Lifecycle.stable(), loadingErrors, resourceManager);
             }
         };
-        return load(loaderFactory, contextRegistries, registriesToLoad, executor);
+        return load(loaderFactory, contextRegistries, registriesToLoad, executor, true);
     }
 
     public static CompletableFuture<RegistryAccess.Frozen> load(
@@ -189,14 +_,16 @@
                 return new NetworkRegistryLoadTask<>(data, Lifecycle.stable(), loadingErrors, entries, knownDataSource);
             }
         };
-        return load(loaderFactory, contextRegistries, registriesToLoad, executor);
+        return load(loaderFactory, contextRegistries, registriesToLoad, executor, false);
     }
 
+    // Neo: Added parameter to know we are loading server-side from resources
     private static CompletableFuture<RegistryAccess.Frozen> load(
         RegistryDataLoader.LoaderFactory loaderFactory,
         List<HolderLookup.RegistryLookup<?>> contextRegistries,
         List<RegistryDataLoader.RegistryData<?>> registriesToLoad,
-        Executor executor
+        Executor executor,
+        boolean fromResources
     ) {
         return CompletableFuture.<CompletableFuture<RegistryAccess.Frozen>>supplyAsync(
                 () -> {
@@ -215,6 +_,15 @@
                     return CompletableFuture.allOf(loadCompletions)
                         .thenApplyAsync(
                             ignored -> {
+                                if (fromResources && net.neoforged.neoforge.gametest.GameTestHooks.isGametestEnabled()) {
+                                    var byKeyMap = loadTasks.stream().collect(java.util.stream.Collectors.toMap(d -> d.registryKey(), d -> d.readOnlyRegistry()));
+                                    var envRegistry = (net.minecraft.core.WritableRegistry) byKeyMap.get(Registries.TEST_ENVIRONMENT);
+                                    var testsRegistry = (net.minecraft.core.WritableRegistry) byKeyMap.get(Registries.TEST_INSTANCE);
+                                    if (envRegistry != null && testsRegistry != null) {
+                                        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.event.RegisterGameTestsEvent(envRegistry, testsRegistry));
+                                    }
+                                }
+
                                 List<RegistryLoadTask<?>> frozenRegistries = loadTasks.stream().filter(task -> task.freezeRegistry(loadingErrors)).toList();
                                 if (!loadingErrors.isEmpty()) {
                                     throw logErrors(loadingErrors);
@@ -309,7 +_,11 @@
     public record NetworkedRegistryData(List<RegistrySynchronization.PackedRegistryEntry> elements, TagNetworkSerialization.NetworkPayload tags) {
     }
 
-    public record RegistryData<T>(ResourceKey<? extends Registry<T>> key, Codec<T> elementCodec, RegistryValidator<T> validator) {
+    public record RegistryData<T>(ResourceKey<? extends Registry<T>> key, Codec<T> elementCodec, RegistryValidator<T> validator, java.util.function.Consumer<net.neoforged.neoforge.registries.RegistryBuilder<T>> registryBuilderConsumer) {
+        public RegistryData(ResourceKey<? extends Registry<T>> key, Codec<T> elementCodec, RegistryValidator<T> validator) {
+            this(key, elementCodec, validator, _ -> {});
+        }
+
         private RegistryData(ResourceKey<? extends Registry<T>> key, Codec<T> elementCodec) {
             this(key, elementCodec, RegistryValidator.none());
         }
