--- a/net/minecraft/resources/RegistryOps.java
+++ b/net/minecraft/resources/RegistryOps.java
@@ -15,7 +_,7 @@
 import net.minecraft.core.Registry;
 import net.minecraft.util.ExtraCodecs;
 
-public class RegistryOps<T> extends DelegatingOps<T> {
+public class RegistryOps<T> extends net.neoforged.neoforge.common.util.dfu.ContextOps<T> {
     public final RegistryOps.RegistryInfoLookup lookupProvider;
 
     public static <T> RegistryOps<T> create(DynamicOps<T> p_256342_, HolderLookup.Provider p_255950_) {
@@ -35,8 +_,29 @@
         this.lookupProvider = p_255799_;
     }
 
+    protected RegistryOps(RegistryOps<T> other, Map<net.neoforged.neoforge.common.util.dfu.ContextOps.Key<?>, Object> additionalContext) {
+        super(other, additionalContext);
+        this.lookupProvider = other.lookupProvider;
+    }
+
+    protected RegistryOps(RegistryOps<T> other) {
+        this(other, Map.of());
+    }
+
+    /** {@inheritDoc} **/
+    @Override
+    public <Z> RegistryOps<T> withContext(net.neoforged.neoforge.common.util.dfu.ContextOps.Key<Z> key, Z value) {
+        return withContext(Map.of(key, value));
+    }
+
+    /** {@inheritDoc} **/
+    @Override
+    public RegistryOps<T> withContext(Map<net.neoforged.neoforge.common.util.dfu.ContextOps.Key<?>, Object> contextMap) {
+        return contextMap.isEmpty() ? this : new RegistryOps<>(this, contextMap);
+    }
+
     public <U> RegistryOps<U> withParent(DynamicOps<U> p_330654_) {
-        return (RegistryOps<U>)(p_330654_ == this.delegate ? this : new RegistryOps<>(p_330654_, this.lookupProvider));
+        return (RegistryOps<U>)(p_330654_ == this.delegate ? this : new RegistryOps<>(p_330654_, this.lookupProvider).withContext(this.contextObjects));
     }
 
     public <E> Optional<HolderOwner<E>> owner(ResourceKey<? extends Registry<? extends E>> p_255757_) {
@@ -74,6 +_,20 @@
                     : DataResult.error(() -> "Not a registry ops")
             )
             .forGetter(p_255526_ -> null);
+    }
+
+    public static <E> com.mojang.serialization.MapCodec<HolderLookup.RegistryLookup<E>> retrieveRegistryLookup(ResourceKey<? extends Registry<? extends E>> resourceKey) {
+        return ExtraCodecs.retrieveContext(ops -> {
+            if (!(ops instanceof RegistryOps<?> registryOps))
+                return DataResult.error(() -> "Not a registry ops");
+
+            return registryOps.lookupProvider.lookup(resourceKey).map(registryInfo -> {
+                if (!(registryInfo.owner() instanceof HolderLookup.RegistryLookup<E> registryLookup))
+                    return DataResult.<HolderLookup.RegistryLookup<E>>error(() -> "Found holder getter but was not a registry lookup for " + resourceKey);
+
+                return DataResult.success(registryLookup, registryInfo.elementsLifecycle());
+            }).orElseGet(() -> DataResult.error(() -> "Unknown registry: " + resourceKey));
+        });
     }
 
     public static <E, O> RecordCodecBuilder<O, Holder.Reference<E>> retrieveElement(ResourceKey<E> p_256347_) {
