--- a/net/minecraft/resources/RegistryDataLoader.java
+++ b/net/minecraft/resources/RegistryDataLoader.java
@@ -122,7 +_,7 @@
     public static final List<RegistryDataLoader.RegistryData<?>> DIMENSION_REGISTRIES = List.of(
         new RegistryDataLoader.RegistryData<>(Registries.LEVEL_STEM, LevelStem.CODEC)
     );
-    public static final List<RegistryDataLoader.RegistryData<?>> SYNCHRONIZED_REGISTRIES = List.of(
+    public static final List<RegistryDataLoader.RegistryData<?>> SYNCHRONIZED_REGISTRIES = net.neoforged.neoforge.registries.DataPackRegistriesHooks.grabNetworkableRegistries(List.of(
         new RegistryDataLoader.RegistryData<>(Registries.BIOME, Biome.NETWORK_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.CHAT_TYPE, ChatType.DIRECT_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.TRIM_PATTERN, TrimPattern.DIRECT_CODEC),
@@ -147,7 +_,7 @@
         new RegistryDataLoader.RegistryData<>(Registries.DIALOG, Dialog.DIRECT_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.WORLD_CLOCK, WorldClock.DIRECT_CODEC),
         new RegistryDataLoader.RegistryData<>(Registries.TIMELINE, Timeline.NETWORK_CODEC)
-    );
+    )); // Neo: Keep the list so custom registries can be added later
 
     public static CompletableFuture<RegistryAccess.Frozen> load(
         ResourceManager resourceManager,
@@ -161,7 +_,7 @@
                 return new ResourceManagerRegistryLoadTask<>(data, Lifecycle.stable(), loadingErrors, resourceManager);
             }
         };
-        return load(loaderFactory, contextRegistries, registriesToLoad, executor);
+        return load(loaderFactory, contextRegistries, registriesToLoad, executor, true);
     }
 
     public static CompletableFuture<RegistryAccess.Frozen> load(
@@ -177,14 +_,16 @@
                 return new NetworkRegistryLoadTask<>(data, Lifecycle.stable(), loadingErrors, entries, knownDataSource);
             }
         };
-        return load(loaderFactory, contextRegistries, registriesToLoad, executor);
+        return load(loaderFactory, contextRegistries, registriesToLoad, executor, false);
     }
 
+    // Neo: Added parameter to know we are loading server-side from resources
     private static CompletableFuture<RegistryAccess.Frozen> load(
         RegistryDataLoader.LoaderFactory loaderFactory,
         List<HolderLookup.RegistryLookup<?>> contextRegistries,
         List<RegistryDataLoader.RegistryData<?>> registriesToLoad,
-        Executor executor
+        Executor executor,
+        boolean fromResources
     ) {
         return CompletableFuture.<CompletableFuture<RegistryAccess.Frozen>>supplyAsync(
                 () -> {
