--- a/net/minecraft/resources/ResourceKey.java
+++ b/net/minecraft/resources/ResourceKey.java
@@ -75,6 +_,15 @@
         return createRegistryKey(this.registryName);
     }
 
+    @Override
+    public int compareTo(ResourceKey<?> o) {
+        int ret = this.registry().compareTo(o.registry());
+        if (ret != 0) {
+            return ret;
+        }
+        return this.identifier().compareTo(o.identifier());
+    }
+
     private record InternKey(Identifier registry, Identifier identifier) {
     }
 }
