--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -140,8 +_,14 @@
       if (i == 0) {
          i = this.namespace.compareTo(p_135826_.namespace);
       }
-
+      
       return i;
+   }
+
+   // Normal compare sorts by path first, this compares namespace first.
+   public int compareNamespaced(ResourceLocation o) {
+      int ret = this.namespace.compareTo(o.namespace);
+      return ret != 0 ? ret : this.path.compareTo(o.path);
    }
 
    public String toDebugFileName() {
