--- a/net/minecraft/resources/ResourceKey.java
+++ b/net/minecraft/resources/ResourceKey.java
@@ -66,6 +_,13 @@
         return createRegistryKey(this.registryName);
     }
 
+    @Override
+    public int compareTo(ResourceKey<?> o) {
+        int ret = this.registry().compareTo(o.registry());
+        if (ret == 0) ret = this.location().compareTo(o.location());
+        return ret;
+    }
+
     record InternKey(ResourceLocation registry, ResourceLocation location) {
     }
 }
