--- a/net/minecraft/network/chat/ClickEvent.java
+++ b/net/minecraft/network/chat/ClickEvent.java
@@ -82,7 +_,7 @@
         }
 
         public static DataResult<ClickEvent.Action> filterForSerialization(ClickEvent.Action action) {
-            return !action.isAllowedFromServer()
+            return !action.isAllowedFromServer() && (net.neoforged.neoforge.server.ServerLifecycleHooks.getCurrentServer() == null || net.neoforged.neoforge.server.ServerLifecycleHooks.getCurrentServer().isDedicatedServer()) // Neo: Allow open file commands to work on integrated servers. PR #915
                 ? DataResult.error(() -> "Action not allowed: " + action)
                 : DataResult.success(action, Lifecycle.stable());
         }
