--- a/net/minecraft/network/chat/ComponentSerialization.java
+++ b/net/minecraft/network/chat/ComponentSerialization.java
@@ -123,6 +_,7 @@
         p_442635_.put("selector", SelectorContents.MAP_CODEC);
         p_442635_.put("nbt", NbtContents.MAP_CODEC);
         p_442635_.put("object", ObjectContents.MAP_CODEC);
+        p_442635_.put("neoforge:inserting", net.neoforged.neoforge.common.util.InsertingContents.MAP_CODEC);
     }
 
     static class FuzzyCodec<T> extends MapCodec<T> {
