--- a/net/minecraft/network/chat/SignableCommand.java
+++ b/net/minecraft/network/chat/SignableCommand.java
@@ -24,7 +_,8 @@
         List<SignableCommand.Argument<S>> list;
         CommandContextBuilder<S> commandcontextbuilder2;
         for(list = collectArguments(s, commandcontextbuilder);
-            (commandcontextbuilder2 = commandcontextbuilder1.getChild()) != null && commandcontextbuilder2.getRootNode() != commandcontextbuilder.getRootNode();
+            // NEO: Check if the command node is a RootCommandNode, instead of simply being the original root node; fixes #186
+            (commandcontextbuilder2 = commandcontextbuilder1.getChild()) != null && !(commandcontextbuilder2.getRootNode() instanceof com.mojang.brigadier.tree.RootCommandNode);
             commandcontextbuilder1 = commandcontextbuilder2
         ) {
             list.addAll(collectArguments(s, commandcontextbuilder2));
