--- a/com/mojang/blaze3d/opengl/DirectStateAccess.java
+++ b/com/mojang/blaze3d/opengl/DirectStateAccess.java
@@ -42,7 +_,11 @@
 
     abstract int createFrameBufferObject();
 
-    abstract void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, int bindSlot);
+    abstract void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, int bindSlot, boolean useStencil);
+
+    public void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, int bindSlot) {
+        bindFrameBufferTextures(fbo, color0, depth, mipLevel, bindSlot, false);
+    }
 
     abstract void blitFrameBuffers(
         int source, int dest, int srcX0, int srcY0, int srcX1, int srcY1, int dstX0, int dstY0, int dstX1, int dstY1, int mask, int filter
@@ -101,9 +_,14 @@
         }
 
         @Override
-        public void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, @GpuBuffer.Usage int bindSlot) {
+        public void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, @GpuBuffer.Usage int bindSlot, boolean useStencil) {
             ARBDirectStateAccess.glNamedFramebufferTexture(fbo, 36064, color0, mipLevel);
             ARBDirectStateAccess.glNamedFramebufferTexture(fbo, 36096, depth, mipLevel);
+            if (useStencil) {
+                ARBDirectStateAccess.glNamedFramebufferTexture(fbo, org.lwjgl.opengl.GL32C.GL_STENCIL_ATTACHMENT, depth, mipLevel);
+            } else {
+                ARBDirectStateAccess.glNamedFramebufferTexture(fbo, org.lwjgl.opengl.GL32C.GL_STENCIL_ATTACHMENT, 0, 0);
+            }
             if (bindSlot != 0) {
                 GlStateManager._glBindFramebuffer(bindSlot, fbo);
             }
@@ -224,12 +_,17 @@
         }
 
         @Override
-        public void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, int bindSlot) {
+        public void bindFrameBufferTextures(int fbo, int color0, int depth, int mipLevel, int bindSlot, boolean useStencil) {
             int tempBindSlot = bindSlot == 0 ? '\u8ca9' : bindSlot;
             int oldFbo = GlStateManager.getFrameBuffer(tempBindSlot);
             GlStateManager._glBindFramebuffer(tempBindSlot, fbo);
             GlStateManager._glFramebufferTexture2D(tempBindSlot, 36064, 3553, color0, mipLevel);
             GlStateManager._glFramebufferTexture2D(tempBindSlot, 36096, 3553, depth, mipLevel);
+            if (useStencil) {
+                GlStateManager._glFramebufferTexture2D(tempBindSlot, org.lwjgl.opengl.GL32C.GL_STENCIL_ATTACHMENT, 3553, depth, mipLevel);
+            } else {
+                GlStateManager._glFramebufferTexture2D(tempBindSlot, org.lwjgl.opengl.GL32C.GL_STENCIL_ATTACHMENT, 3553, 0, 0);
+            }
             if (bindSlot == 0) {
                 GlStateManager._glBindFramebuffer(tempBindSlot, oldFbo);
             }
