--- a/com/mojang/blaze3d/opengl/GlBackend.java
+++ b/com/mojang/blaze3d/opengl/GlBackend.java
@@ -26,6 +_,13 @@
     public WindowAndDevice createDeviceWithWindow(
         int width, int height, String title, long monitor, ShaderSource defaultShaderSource, GpuDebugOptions debugOptions
     ) throws BackendCreationException {
+        var earlyLoadingScreen = net.neoforged.fml.loading.EarlyLoadingScreenController.current();
+        if (earlyLoadingScreen != null) {
+            var window = earlyLoadingScreen.takeOverGlfwWindow();
+            GLFW.glfwSetWindowTitle(window, title);
+            return new WindowAndDevice(window, net.neoforged.neoforge.client.ClientHooks.createGpuDevice(window, defaultShaderSource, debugOptions));
+        }
+
         GLFWErrorCapture glfwErrors = new GLFWErrorCapture();
 
         WindowAndDevice error;
@@ -55,7 +_,7 @@
                 throw new BackendCreationException("Failed to create window with OpenGL context");
             }
 
-            error = new WindowAndDevice(window, new GpuDevice(new GlDevice(window, defaultShaderSource, debugOptions)));
+            error = new WindowAndDevice(window, net.neoforged.neoforge.client.ClientHooks.createGpuDevice(window, defaultShaderSource, debugOptions));
         }
 
         return error;
