--- a/com/mojang/blaze3d/opengl/GlCommandEncoder.java
+++ b/com/mojang/blaze3d/opengl/GlCommandEncoder.java
@@ -603,6 +_,26 @@
             GlStateManager._disableScissorTest();
         }
 
+        var stencilTestOpt = renderPass.pipeline.info().getStencilTest();
+        if (stencilTestOpt.isPresent()) {
+            var stencilTest = stencilTestOpt.get();
+            GlStateManager._enableStencilTest();
+            var front = stencilTest.front();
+            var back = stencilTest.back();
+            if (front.equals(back)) {
+                GlStateManager._stencilFunc(GlConst.toGl(front.compare()), stencilTest.referenceValue(), stencilTest.readMask());
+                GlStateManager._stencilOp(GlConst.toGl(front.fail()), GlConst.toGl(front.depthFail()), GlConst.toGl(front.pass()));
+            } else {
+                GlStateManager._stencilFuncFront(GlConst.toGl(front.compare()), stencilTest.referenceValue(), stencilTest.readMask());
+                GlStateManager._stencilFuncBack(GlConst.toGl(back.compare()), stencilTest.referenceValue(), stencilTest.readMask());
+                GlStateManager._stencilOpFront(GlConst.toGl(front.fail()), GlConst.toGl(front.depthFail()), GlConst.toGl(front.pass()));
+                GlStateManager._stencilOpBack(GlConst.toGl(back.fail()), GlConst.toGl(back.depthFail()), GlConst.toGl(back.pass()));
+            }
+            GlStateManager._stencilMask(stencilTest.writeMask());
+        } else {
+            GlStateManager._disableStencilTest();
+        }
+
         return true;
     }
 
