--- a/com/mojang/blaze3d/opengl/GlConst.java
+++ b/com/mojang/blaze3d/opengl/GlConst.java
@@ -147,14 +_,18 @@
             case EQUAL_DEPTH_TEST -> 514;
             case LESS_DEPTH_TEST -> 513;
             case GREATER_DEPTH_TEST -> 516;
-            default -> 515;
+            case LEQUAL_DEPTH_TEST -> 515;
+            case NOT_EQUAL_DEPTH_TEST -> org.lwjgl.opengl.GL11C.GL_NOTEQUAL;
+            case GEQUAL_DEPTH_TEST -> org.lwjgl.opengl.GL11C.GL_GEQUAL;
+            case NEVER_DEPTH_TEST -> org.lwjgl.opengl.GL11C.GL_NEVER;
         };
     }
 
     public static int toGl(PolygonMode p_412150_) {
         return switch (p_412150_) {
             case WIREFRAME -> 6913;
-            default -> 6914;
+            case FILL -> 6914;
+            case POINT -> org.lwjgl.opengl.GL11C.GL_POINT;
         };
     }
 
@@ -174,6 +_,11 @@
             case SRC_ALPHA -> 770;
             case SRC_COLOR -> 768;
             case ZERO -> 0;
+            case SRC_ALPHA_SATURATE -> org.lwjgl.opengl.GL11C.GL_SRC_ALPHA_SATURATE;
+            case SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_SRC1_COLOR;
+            case ONE_MINUS_SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_COLOR;
+            case SRC1_ALPHA -> org.lwjgl.opengl.GL15C.GL_SRC1_ALPHA;
+            case ONE_MINUS_SRC1_ALPHA -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_ALPHA;
         };
     }
 
@@ -194,6 +_,10 @@
             case SRC_ALPHA_SATURATE -> 776;
             case SRC_COLOR -> 768;
             case ZERO -> 0;
+            case SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_SRC1_COLOR;
+            case ONE_MINUS_SRC1_COLOR -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_COLOR;
+            case SRC1_ALPHA -> org.lwjgl.opengl.GL15C.GL_SRC1_ALPHA;
+            case ONE_MINUS_SRC1_ALPHA -> org.lwjgl.opengl.GL33C.GL_ONE_MINUS_SRC1_ALPHA;
         };
     }
 
@@ -229,7 +_,9 @@
     public static int toGl(AddressMode p_412266_) {
         return switch (p_412266_) {
             case REPEAT -> 10497;
+            case MIRRORED_REPEAT -> org.lwjgl.opengl.GL14C.GL_MIRRORED_REPEAT;
             case CLAMP_TO_EDGE -> 33071;
+            case CLAMP_TO_BORDER -> org.lwjgl.opengl.GL13C.GL_CLAMP_TO_BORDER;
         };
     }
 
@@ -247,35 +_,294 @@
 
     public static int toGlInternalId(TextureFormat p_412409_) {
         return switch (p_412409_) {
-            case RGBA8 -> 32856;
-            case RED8 -> 33321;
-            case RED8I -> 33329;
-            case DEPTH32 -> 33191;
+            case R4G4_UNORM_PACK8 -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R4G4B4A4_UNORM_PACK16, A4B4G4R4_UNORM_PACK16, B4G4R4A4_UNORM_PACK16, A4R4G4B4_UNORM_PACK16 -> org.lwjgl.opengl.GL11C.GL_RGBA4;
+            case R5G6B5_UNORM_PACK16, B5G6R5_UNORM_PACK16 -> org.lwjgl.opengl.GL41C.GL_RGB565;
+            case R5G5B5A1_UNORM_PACK16, A1B5G5R5_UNORM_PACK16, B5G5R5A1_UNORM_PACK16, A1R5G5B5_UNORM_PACK16 -> org.lwjgl.opengl.GL11C.GL_RGB5_A1;
+
+            case A2R10G10B10_UNORM_PACK32 -> org.lwjgl.opengl.GL11C.GL_RGB10_A2;
+            case A2R10G10B10_UINT_PACK32 -> org.lwjgl.opengl.GL33C.GL_RGB10_A2UI;
+            case A2B10G10R10_UNORM_PACK32 -> org.lwjgl.opengl.GL11C.GL_RGB10_A2;
+            case A2B10G10R10_UINT_PACK32 -> org.lwjgl.opengl.GL33C.GL_RGB10_A2UI;
+
+            case E5B9G9R9_UFLOAT_PACK32 -> org.lwjgl.opengl.GL30C.GL_RGB9_E5;
+
+            case R8_UNORM -> org.lwjgl.opengl.GL30C.GL_R8;
+            case R8_SNORM -> org.lwjgl.opengl.GL31C.GL_R8_SNORM;
+            case R8_UINT -> org.lwjgl.opengl.GL30C.GL_R8UI;
+            case R8_SINT -> org.lwjgl.opengl.GL30C.GL_R8I;
+            case R8_SRGB -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R8G8_UNORM -> org.lwjgl.opengl.GL30C.GL_RG8;
+            case R8G8_SNORM -> org.lwjgl.opengl.GL31C.GL_RG8_SNORM;
+            case R8G8_UINT -> org.lwjgl.opengl.GL30C.GL_RG8UI;
+            case R8G8_SINT -> org.lwjgl.opengl.GL30C.GL_RG8I;
+            case R8G8_SRGB -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R8G8B8_UNORM -> org.lwjgl.opengl.GL30C.GL_RGB8;
+            case R8G8B8_SNORM -> org.lwjgl.opengl.GL31C.GL_RGB8_SNORM;
+            case R8G8B8_UINT -> org.lwjgl.opengl.GL30C.GL_RGB8UI;
+            case R8G8B8_SINT -> org.lwjgl.opengl.GL30C.GL_RGB8I;
+            case R8G8B8_SRGB -> org.lwjgl.opengl.GL21C.GL_SRGB8;
+
+            case B8G8R8_UNORM -> org.lwjgl.opengl.GL30C.GL_RGB8;
+            case B8G8R8_SNORM -> org.lwjgl.opengl.GL31C.GL_RGB8_SNORM;
+            case B8G8R8_UINT -> org.lwjgl.opengl.GL30C.GL_RGB8UI;
+            case B8G8R8_SINT -> org.lwjgl.opengl.GL30C.GL_RGB8I;
+            case B8G8R8_SRGB -> org.lwjgl.opengl.GL21C.GL_SRGB8;
+
+            case R8G8B8A8_UNORM -> org.lwjgl.opengl.GL30C.GL_RGBA8;
+            case R8G8B8A8_SNORM -> org.lwjgl.opengl.GL31C.GL_RGBA8_SNORM;
+            case R8G8B8A8_UINT -> org.lwjgl.opengl.GL30C.GL_RGBA8UI;
+            case R8G8B8A8_SINT -> org.lwjgl.opengl.GL30C.GL_RGBA8I;
+            case R8G8B8A8_SRGB -> org.lwjgl.opengl.GL21C.GL_SRGB8_ALPHA8;
+
+            case B8G8R8A8_UNORM -> org.lwjgl.opengl.GL30C.GL_RGBA8;
+            case B8G8R8A8_SNORM -> org.lwjgl.opengl.GL31C.GL_RGBA8_SNORM;
+            case B8G8R8A8_UINT -> org.lwjgl.opengl.GL30C.GL_RGBA8UI;
+            case B8G8R8A8_SINT -> org.lwjgl.opengl.GL30C.GL_RGBA8I;
+            case B8G8R8A8_SRGB -> org.lwjgl.opengl.GL21C.GL_SRGB8_ALPHA8;
+
+            case R16_UNORM -> org.lwjgl.opengl.GL30C.GL_R16;
+            case R16_SNORM -> org.lwjgl.opengl.GL31C.GL_R16_SNORM;
+            case R16_UINT -> org.lwjgl.opengl.GL30C.GL_R16UI;
+            case R16_SINT -> org.lwjgl.opengl.GL30C.GL_R16I;
+            case R16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_R16F;
+
+            case R16G16_UNORM -> org.lwjgl.opengl.GL30C.GL_RG16;
+            case R16G16_SNORM -> org.lwjgl.opengl.GL31C.GL_RG16_SNORM;
+            case R16G16_UINT -> org.lwjgl.opengl.GL30C.GL_RG16UI;
+            case R16G16_SINT -> org.lwjgl.opengl.GL30C.GL_RG16I;
+            case R16G16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RG16F;
+
+            case R16G16B16_UNORM -> org.lwjgl.opengl.GL30C.GL_RGB16;
+            case R16G16B16_SNORM -> org.lwjgl.opengl.GL31C.GL_RGB16_SNORM;
+            case R16G16B16_UINT -> org.lwjgl.opengl.GL30C.GL_RGB16UI;
+            case R16G16B16_SINT -> org.lwjgl.opengl.GL30C.GL_RGB16I;
+            case R16G16B16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RGB16F;
+
+            case R16G16B16A16_UNORM -> org.lwjgl.opengl.GL30C.GL_RGBA16;
+            case R16G16B16A16_SNORM -> org.lwjgl.opengl.GL31C.GL_RGBA16_SNORM;
+            case R16G16B16A16_UINT -> org.lwjgl.opengl.GL30C.GL_RGBA16UI;
+            case R16G16B16A16_SINT -> org.lwjgl.opengl.GL30C.GL_RGBA16I;
+            case R16G16B16A16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RGBA16F;
+
+            case R32_UINT -> org.lwjgl.opengl.GL30C.GL_R32UI;
+            case R32_SINT -> org.lwjgl.opengl.GL30C.GL_R32I;
+            case R32_SFLOAT -> org.lwjgl.opengl.GL30C.GL_R32F;
+
+            case R32G32_UINT -> org.lwjgl.opengl.GL30C.GL_RG32UI;
+            case R32G32_SINT -> org.lwjgl.opengl.GL30C.GL_RG32I;
+            case R32G32_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RG32F;
+
+            case R32G32B32_UINT -> org.lwjgl.opengl.GL30C.GL_RGB32UI;
+            case R32G32B32_SINT -> org.lwjgl.opengl.GL30C.GL_RGB32I;
+            case R32G32B32_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RGB32F;
+
+            case R32G32B32A32_UINT -> org.lwjgl.opengl.GL30C.GL_RGBA32UI;
+            case R32G32B32A32_SINT -> org.lwjgl.opengl.GL30C.GL_RGBA32I;
+            case R32G32B32A32_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RGBA32F;
+
+            case D16_UNORM -> org.lwjgl.opengl.GL30C.GL_DEPTH_COMPONENT16;
+            case X8_D24_UNORM_PACK32 -> org.lwjgl.opengl.GL30C.GL_DEPTH_COMPONENT24;
+            case D32_SFLOAT -> org.lwjgl.opengl.GL30C.GL_DEPTH_COMPONENT32F;
+
+            case S8_UINT -> org.lwjgl.opengl.GL41C.GL_STENCIL_INDEX8;
+
+            case D24_UNORM_S8_UINT -> org.lwjgl.opengl.GL30C.GL_DEPTH24_STENCIL8;
+            case D32_SFLOAT_S8_UINT -> org.lwjgl.opengl.GL30C.GL_DEPTH32F_STENCIL8;
+
+            case BC1_RGB_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGB_S3TC_DXT1_EXT;
+            case BC1_RGB_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_S3TC_DXT1_EXT;
+            case BC1_RGBA_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT1_EXT;
+            case BC1_RGBA_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT;
+            case BC2_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT3_EXT;
+            case BC2_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT;
+            case BC3_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT5_EXT;
+            case BC3_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT;
         };
     }
 
     public static int toGlExternalId(TextureFormat p_412322_) {
         return switch (p_412322_) {
-            case RGBA8 -> 6408;
-            case RED8 -> 6403;
-            case RED8I -> 6403;
-            case DEPTH32 -> 6402;
+            case R4G4_UNORM_PACK8 -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R4G4B4A4_UNORM_PACK16, A4B4G4R4_UNORM_PACK16 -> org.lwjgl.opengl.GL11C.GL_RGBA;
+            case B4G4R4A4_UNORM_PACK16, A4R4G4B4_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_BGRA;
+
+            case R5G6B5_UNORM_PACK16, B5G6R5_UNORM_PACK16 -> org.lwjgl.opengl.GL11C.GL_RGB;
+
+            case R5G5B5A1_UNORM_PACK16, A1B5G5R5_UNORM_PACK16 -> org.lwjgl.opengl.GL11C.GL_RGBA;
+            case B5G5R5A1_UNORM_PACK16, A1R5G5B5_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_BGRA;
+
+            case A2R10G10B10_UNORM_PACK32 -> org.lwjgl.opengl.GL12C.GL_BGRA;
+            case A2R10G10B10_UINT_PACK32 -> org.lwjgl.opengl.GL12C.GL_BGRA;
+            case A2B10G10R10_UNORM_PACK32 -> org.lwjgl.opengl.GL11C.GL_RGBA;
+            case A2B10G10R10_UINT_PACK32 -> org.lwjgl.opengl.GL11C.GL_RGBA;
+
+            case E5B9G9R9_UFLOAT_PACK32 -> org.lwjgl.opengl.GL11C.GL_RGB;
+
+            case R8_UNORM, R8_SNORM, R8_UINT, R8_SINT -> org.lwjgl.opengl.GL11C.GL_RED;
+            case R8_SRGB -> org.lwjgl.opengl.GL11C.GL_NONE;
+            case R8G8_UNORM, R8G8_SNORM, R8G8_UINT, R8G8_SINT -> org.lwjgl.opengl.GL30C.GL_RG;
+            case R8G8_SRGB -> org.lwjgl.opengl.GL11C.GL_NONE;
+            case R8G8B8_UNORM, R8G8B8_SNORM, R8G8B8_UINT, R8G8B8_SINT, R8G8B8_SRGB -> org.lwjgl.opengl.GL11C.GL_RGB;
+            case B8G8R8_UNORM, B8G8R8_SNORM, B8G8R8_UINT, B8G8R8_SINT, B8G8R8_SRGB -> org.lwjgl.opengl.GL12C.GL_BGR;
+            case R8G8B8A8_UNORM, R8G8B8A8_SNORM, R8G8B8A8_UINT, R8G8B8A8_SINT, R8G8B8A8_SRGB -> org.lwjgl.opengl.GL11C.GL_RGBA;
+            case B8G8R8A8_UNORM, B8G8R8A8_SNORM, B8G8R8A8_UINT, B8G8R8A8_SINT, B8G8R8A8_SRGB -> org.lwjgl.opengl.GL12C.GL_BGRA;
+
+            case R16_UNORM, R16_SNORM, R16_UINT, R16_SINT, R16_SFLOAT -> org.lwjgl.opengl.GL11C.GL_RED;
+            case R16G16_UNORM, R16G16_SNORM, R16G16_UINT, R16G16_SINT, R16G16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RG;
+            case R16G16B16_UNORM, R16G16B16_SNORM, R16G16B16_UINT, R16G16B16_SINT, R16G16B16_SFLOAT -> org.lwjgl.opengl.GL11C.GL_RGB;
+            case R16G16B16A16_UNORM, R16G16B16A16_SNORM, R16G16B16A16_UINT, R16G16B16A16_SINT, R16G16B16A16_SFLOAT -> org.lwjgl.opengl.GL11C.GL_RGBA;
+
+            case R32_UINT, R32_SINT, R32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_RED;
+            case R32G32_UINT, R32G32_SINT, R32G32_SFLOAT -> org.lwjgl.opengl.GL30C.GL_RG;
+            case R32G32B32_UINT, R32G32B32_SINT, R32G32B32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_RGB;
+            case R32G32B32A32_UINT, R32G32B32A32_SINT, R32G32B32A32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_RGBA;
+
+            case D16_UNORM, X8_D24_UNORM_PACK32, D32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_DEPTH_COMPONENT;
+
+            case S8_UINT -> org.lwjgl.opengl.GL11C.GL_STENCIL_INDEX;
+
+            case D24_UNORM_S8_UINT, D32_SFLOAT_S8_UINT -> org.lwjgl.opengl.GL30C.GL_DEPTH_STENCIL;
+
+            case BC1_RGB_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGB_S3TC_DXT1_EXT;
+            case BC1_RGB_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_S3TC_DXT1_EXT;
+            case BC1_RGBA_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT1_EXT;
+            case BC1_RGBA_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT;
+            case BC2_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT3_EXT;
+            case BC2_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT;
+            case BC3_UNORM_BLOCK -> org.lwjgl.opengl.EXTTextureCompressionS3TC.GL_COMPRESSED_RGBA_S3TC_DXT5_EXT;
+            case BC3_SRGB_BLOCK -> org.lwjgl.opengl.EXTTextureSRGB.GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT;
         };
     }
 
     public static int toGlType(TextureFormat p_412407_) {
         return switch (p_412407_) {
-            case RGBA8 -> 5121;
-            case RED8 -> 5121;
-            case RED8I -> 5121;
-            case DEPTH32 -> 5126;
+            case R4G4_UNORM_PACK8 -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R4G4B4A4_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_4_4_4_4;
+            case A4B4G4R4_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_4_4_4_4_REV;
+            case B4G4R4A4_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_4_4_4_4;
+            case A4R4G4B4_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_4_4_4_4_REV;
+
+            case R5G6B5_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_5_6_5;
+            case B5G6R5_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_5_6_5_REV;
+
+            case R5G5B5A1_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_5_5_5_1;
+            case A1B5G5R5_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_1_5_5_5_REV;
+            case B5G5R5A1_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_5_5_5_1;
+            case A1R5G5B5_UNORM_PACK16 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_SHORT_1_5_5_5_REV;
+
+            case A2R10G10B10_UNORM_PACK32 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_INT_2_10_10_10_REV;
+            case A2R10G10B10_UINT_PACK32 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_INT_2_10_10_10_REV;
+            case A2B10G10R10_UNORM_PACK32 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_INT_2_10_10_10_REV;
+            case A2B10G10R10_UINT_PACK32 -> org.lwjgl.opengl.GL12C.GL_UNSIGNED_INT_2_10_10_10_REV;
+
+            case E5B9G9R9_UFLOAT_PACK32 -> org.lwjgl.opengl.GL30C.GL_UNSIGNED_INT_5_9_9_9_REV;
+
+            case R8_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8_SNORM -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8_SINT -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8_SRGB -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R8G8_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8G8_SNORM -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8G8_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8G8_SINT -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8G8_SRGB -> org.lwjgl.opengl.GL11C.GL_NONE;
+
+            case R8G8B8_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8G8B8_SNORM -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8G8B8_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8G8B8_SINT -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8G8B8_SRGB -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+
+            case B8G8R8_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case B8G8R8_SNORM -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case B8G8R8_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case B8G8R8_SINT -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case B8G8R8_SRGB -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+
+            case R8G8B8A8_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8G8B8A8_SNORM -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8G8B8A8_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case R8G8B8A8_SINT -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case R8G8B8A8_SRGB -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+
+            case B8G8R8A8_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case B8G8R8A8_SNORM -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case B8G8R8A8_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+            case B8G8R8A8_SINT -> org.lwjgl.opengl.GL11C.GL_BYTE;
+            case B8G8R8A8_SRGB -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_BYTE;
+
+            case R16_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16_SNORM -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16_SINT -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_HALF_FLOAT;
+
+            case R16G16_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16G16_SNORM -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16G16_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16G16_SINT -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16G16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_HALF_FLOAT;
+
+            case R16G16B16_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16G16B16_SNORM -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16G16B16_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16G16B16_SINT -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16G16B16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_HALF_FLOAT;
+
+            case R16G16B16A16_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16G16B16A16_SNORM -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16G16B16A16_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case R16G16B16A16_SINT -> org.lwjgl.opengl.GL11C.GL_SHORT;
+            case R16G16B16A16_SFLOAT -> org.lwjgl.opengl.GL30C.GL_HALF_FLOAT;
+
+            case R32_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_INT;
+            case R32_SINT -> org.lwjgl.opengl.GL11C.GL_INT;
+            case R32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_FLOAT;
+
+            case R32G32_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_INT;
+            case R32G32_SINT -> org.lwjgl.opengl.GL11C.GL_INT;
+            case R32G32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_FLOAT;
+
+            case R32G32B32_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_INT;
+            case R32G32B32_SINT -> org.lwjgl.opengl.GL11C.GL_INT;
+            case R32G32B32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_FLOAT;
+
+            case R32G32B32A32_UINT -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_INT;
+            case R32G32B32A32_SINT -> org.lwjgl.opengl.GL11C.GL_INT;
+            case R32G32B32A32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_FLOAT;
+
+            case D16_UNORM -> org.lwjgl.opengl.GL11C.GL_UNSIGNED_SHORT;
+            case X8_D24_UNORM_PACK32 -> org.lwjgl.opengl.GL30C.GL_UNSIGNED_INT_24_8;
+            case D32_SFLOAT -> org.lwjgl.opengl.GL11C.GL_FLOAT;
+
+            case S8_UINT ->  GL_UNSIGNED_BYTE;
+
+            case D24_UNORM_S8_UINT -> org.lwjgl.opengl.GL30C.GL_UNSIGNED_INT_24_8;
+            case D32_SFLOAT_S8_UINT -> org.lwjgl.opengl.GL30C.GL_FLOAT_32_UNSIGNED_INT_24_8_REV;
+
+            // these are supported by OpenGL, but the type isn't used
+            case BC1_RGB_UNORM_BLOCK, BC1_RGB_SRGB_BLOCK, BC1_RGBA_UNORM_BLOCK, BC1_RGBA_SRGB_BLOCK, BC2_UNORM_BLOCK, BC2_SRGB_BLOCK, BC3_UNORM_BLOCK, BC3_SRGB_BLOCK -> org.lwjgl.opengl.GL11C.GL_NONE;
         };
     }
 
     public static int toGl(ShaderType p_412490_) {
         return switch (p_412490_) {
             case VERTEX -> 35633;
+            case TESSELLATION_CONTROL -> org.lwjgl.opengl.GL40C.GL_TESS_CONTROL_SHADER;
+            case TESSELLATION_EVALUATION -> org.lwjgl.opengl.GL40C.GL_TESS_EVALUATION_SHADER;
+            case GEOMETRY -> org.lwjgl.opengl.GL32C.GL_GEOMETRY_SHADER;
             case FRAGMENT -> 35632;
+            case COMPUTE -> org.lwjgl.opengl.GL43C.GL_COMPUTE_SHADER;
+            case TASK -> org.lwjgl.opengl.NVMeshShader.GL_TASK_SHADER_NV;
+            case MESH -> org.lwjgl.opengl.NVMeshShader.GL_MESH_SHADER_NV;
         };
     }
 
@@ -309,5 +_,31 @@
         } else {
             return 35044;
         }
+    }
+
+    public static int toGl(net.neoforged.neoforge.client.stencil.StencilFunction func) {
+        return switch (func) {
+            case NEVER -> org.lwjgl.opengl.GL32.GL_NEVER;
+            case ALWAYS -> org.lwjgl.opengl.GL32.GL_ALWAYS;
+            case LESS -> org.lwjgl.opengl.GL32.GL_LESS;
+            case LEQUAL -> org.lwjgl.opengl.GL32.GL_LEQUAL;
+            case EQUAL -> org.lwjgl.opengl.GL32.GL_EQUAL;
+            case GEQUAL -> org.lwjgl.opengl.GL32.GL_GEQUAL;
+            case GREATER -> org.lwjgl.opengl.GL32.GL_GREATER;
+            case NOTEQUAL -> org.lwjgl.opengl.GL32.GL_NOTEQUAL;
+        };
+    }
+
+    public static int toGl(net.neoforged.neoforge.client.stencil.StencilOperation op) {
+        return switch (op) {
+            case KEEP -> org.lwjgl.opengl.GL32.GL_KEEP;
+            case ZERO -> org.lwjgl.opengl.GL32.GL_ZERO;
+            case REPLACE -> org.lwjgl.opengl.GL32.GL_REPLACE;
+            case INCR -> org.lwjgl.opengl.GL32.GL_INCR;
+            case DECR -> org.lwjgl.opengl.GL32.GL_DECR;
+            case INVERT -> org.lwjgl.opengl.GL32.GL_INVERT;
+            case INCR_WRAP -> org.lwjgl.opengl.GL32.GL_INCR_WRAP;
+            case DECR_WRAP -> org.lwjgl.opengl.GL32.GL_DECR_WRAP;
+        };
     }
 }
