--- a/com/mojang/blaze3d/opengl/GlRenderPass.java
+++ b/com/mojang/blaze3d/opengl/GlRenderPass.java
@@ -58,6 +_,7 @@
 
     @Override
     public void setPipeline(RenderPipeline pipeline) {
+        pipeline = com.mojang.blaze3d.systems.RenderSystem.applyPipelineModifiers(pipeline);
         if (this.pipeline == null || this.pipeline.info() != pipeline) {
             this.dirtyUniforms.addAll(this.uniforms.keySet());
             this.dirtyUniforms.addAll(this.samplers.keySet());
