--- a/net/minecraft/client/renderer/FogRenderer.java
+++ b/net/minecraft/client/renderer/FogRenderer.java
@@ -197,6 +_,12 @@
             fogBlue = fogBlue * (1.0F - f7) + fogBlue * f9 * f7;
         }
 
+        Vector3f fogColor = net.neoforged.neoforge.client.ClientHooks.getFogColor(activeRenderInfo, partialTicks, level, renderDistanceChunks, bossColorModifier, fogRed, fogGreen, fogBlue);
+
+        fogRed = fogColor.x();
+        fogGreen = fogColor.y();
+        fogBlue = fogColor.z();
+
         RenderSystem.clearColor(fogRed, fogGreen, fogBlue, 0.0F);
     }
 
@@ -273,6 +_,7 @@
         RenderSystem.setShaderFogStart(fogrenderer$fogdata.start);
         RenderSystem.setShaderFogEnd(fogrenderer$fogdata.end);
         RenderSystem.setShaderFogShape(fogrenderer$fogdata.shape);
+        net.neoforged.neoforge.client.ClientHooks.onFogRender(fogMode, fogtype, camera, partialTick, farPlaneDistance, fogrenderer$fogdata.start, fogrenderer$fogdata.end, fogrenderer$fogdata.shape);
     }
 
     public static void levelFogColor() {
