--- a/net/minecraft/client/renderer/FogRenderer.java
+++ b/net/minecraft/client/renderer/FogRenderer.java
@@ -196,7 +_,7 @@
             f2 = f2 * (1.0F - f10) + f2 * f12 * f10;
         }
 
-        return new Vector4f(f, f1, f2, 1.0F);
+        return net.neoforged.neoforge.client.ClientHooks.getFogColor(p_363451_, p_364931_, p_363033_, p_363851_, p_363957_, f, f1, f2);
     }
 
     public static boolean toggleFog() {
@@ -274,9 +_,9 @@
                 fogrenderer$fogdata.shape = FogShape.CYLINDER;
             }
 
-            return new FogParameters(
+            return net.neoforged.neoforge.client.ClientHooks.onFogRender(p_234174_, fogtype, p_234173_, p_234177_, p_234175_, new FogParameters(
                 fogrenderer$fogdata.start, fogrenderer$fogdata.end, fogrenderer$fogdata.shape, p_365467_.x, p_365467_.y, p_365467_.z, p_365467_.w
-            );
+            ));
         }
     }
 
