--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -166,14 +_,14 @@
             bufferSource,
             this.submitNodeStorage,
             this.featureRenderDispatcher,
-            List.of(
-                new GuiEntityRenderer(bufferSource, minecraft.getEntityRenderDispatcher()),
-                new GuiSkinRenderer(bufferSource),
-                new GuiBookModelRenderer(bufferSource),
-                new GuiBannerResultRenderer(bufferSource, atlasManager),
-                new GuiSignRenderer(bufferSource, atlasManager),
-                new GuiProfilerChartRenderer(bufferSource)
-            )
+            net.neoforged.neoforge.client.ClientHooks.gatherPictureInPictureRenderers(List.of(
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiEntityRenderState.class, buffers -> new GuiEntityRenderer(buffers, minecraft.getEntityRenderDispatcher())),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiSkinRenderState.class, GuiSkinRenderer::new),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiBookModelRenderState.class, GuiBookModelRenderer::new),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiBannerResultRenderState.class, buffers -> new GuiBannerResultRenderer(buffers, atlasManager)),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiSignRenderState.class, buffers -> new GuiSignRenderer(buffers, atlasManager)),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiProfilerChartRenderState.class, GuiProfilerChartRenderer::new)
+            ))
         );
         this.screenEffectRenderer = new ScreenEffectRenderer(minecraft, atlasManager, bufferSource);
     }
@@ -232,6 +_,7 @@
                 break;
             case null:
             default:
+                if (!net.neoforged.neoforge.client.ClientHooks.loadEntityShader(cameraEntity, this))
                 this.clearPostEffect();
         }
     }
@@ -342,6 +_,9 @@
                 return;
             }
 
+            // Neo: Prevent screen shake if the damage type is marked as "forge:no_flinch"
+            if (cameraState.entityRenderState.preventFlinch) return;
+
             hurt /= cameraState.entityRenderState.hurtDuration;
             hurt = Mth.sin(hurt * hurt * hurt * hurt * (float) Math.PI);
             float rr = cameraState.entityRenderState.hurtDir;
@@ -473,7 +_,8 @@
                 }
             } else if (resourcesLoaded && this.minecraft.screen != null) {
                 try {
-                    this.minecraft.screen.renderWithTooltipAndSubtitles(graphics, xMouse, yMouse, deltaTracker.getGameTimeDeltaTicks());
+                    // Neo: Wrap Screen#render to allow for GUI Layers and ScreenEvent.Render.[Pre/Post]
+                    net.neoforged.neoforge.client.ClientHooks.drawScreen(this.minecraft.screen, graphics, xMouse, yMouse, deltaTracker.getGameTimeDeltaTicks());
                 } catch (Throwable var14) {
                     CrashReport report = CrashReport.forThrowable(var14, "Rendering screen");
                     CrashReportCategory category = report.addCategory("Screen render details");
@@ -710,6 +_,9 @@
                 !shouldCreateBossFog,
                 chunkSectionsToRender
             );
+        profiler.popPush("neoforge_render_after_level");
+        LevelRenderer levelRenderer = this.minecraft.levelRenderer;
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterLevel(levelRenderer, levelRenderer.levelRenderState, null, modelViewMatrix, levelRenderer.getRenderableSections()));
         profiler.popPush("hand");
         boolean isSleeping = cameraState.entityRenderState.isSleeping;
         this.hudProjection.setupPerspective(0.05F, 100.0F, cameraState.hudFov, this.minecraft.getWindow().getWidth(), this.minecraft.getWindow().getHeight());
