--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -150,17 +_,17 @@
         this.guiRenderer = new GuiRenderer(
             this.gameRenderState.guiRenderState,
             this.featureRenderDispatcher,
-            List.of(
-                new GuiEntityRenderer(minecraft.getEntityRenderDispatcher()),
-                new GuiSkinRenderer(),
-                new GuiBookModelRenderer(),
-                new GuiBannerResultRenderer(atlasManager),
-                new GuiProfilerChartRenderer()
-            )
+            net.neoforged.neoforge.client.ClientHooks.gatherPictureInPictureRenderers(List.of(
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.renderer.state.gui.pip.GuiEntityRenderState.class, () -> new GuiEntityRenderer(minecraft.getEntityRenderDispatcher())),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.renderer.state.gui.pip.GuiSkinRenderState.class, GuiSkinRenderer::new),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.renderer.state.gui.pip.GuiBookModelRenderState.class, GuiBookModelRenderer::new),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.renderer.state.gui.pip.GuiBannerResultRenderState.class, () -> new GuiBannerResultRenderer(atlasManager)),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.renderer.state.gui.pip.GuiProfilerChartRenderState.class, GuiProfilerChartRenderer::new)
+            ))
         );
         this.screenEffectRenderer = new ScreenEffectRenderer(minecraft, atlasManager);
         this.debugCrosshairRenderer = new DebugCrosshairRenderer();
-        this.mainRenderTarget = new MainTarget(minecraft.getWindow().getWidth(), minecraft.getWindow().getHeight());
+        this.mainRenderTarget = net.neoforged.neoforge.client.ClientHooks.instantiateMainTarget(minecraft.getWindow().getWidth(), minecraft.getWindow().getHeight());
     }
 
     @Override
@@ -219,6 +_,7 @@
                 break;
             case null:
             default:
+                if (!net.neoforged.neoforge.client.ClientHooks.loadEntityShader(cameraEntity, this))
                 this.clearPostEffect();
         }
     }
@@ -309,6 +_,9 @@
                 return;
             }
 
+            // Neo: Prevent screen shake if the damage type is marked as "forge:no_flinch"
+            if (cameraState.entityRenderState.preventFlinch) return;
+
             hurt /= cameraState.entityRenderState.hurtDuration;
             hurt = Mth.sin(hurt * hurt * hurt * hurt * (float) Math.PI);
             float rr = cameraState.entityRenderState.hurtDir;
@@ -563,6 +_,9 @@
         this.minecraft
             .levelRenderer
             .render(this.resourcePool, deltaTracker, renderOutline, cameraState, modelViewMatrix, terrainFog, cameraState.fogData.color, !shouldCreateBossFog);
+        profiler.popPush("neoforge_render_after_level");
+        LevelRenderer levelRenderer = this.minecraft.levelRenderer;
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterLevel(levelRenderer, levelRenderer.levelRenderState, null, modelViewMatrix, levelRenderer.visibleSections()));
         profiler.popPush("hand");
         boolean isSleeping = cameraState.entityRenderState.isSleeping;
         this.hudProjection
