--- a/net/minecraft/client/renderer/GameRenderer.java
+++ b/net/minecraft/client/renderer/GameRenderer.java
@@ -164,14 +_,14 @@
             bufferSource,
             this.submitNodeStorage,
             this.featureRenderDispatcher,
-            List.of(
-                new GuiEntityRenderer(bufferSource, minecraft.getEntityRenderDispatcher()),
-                new GuiSkinRenderer(bufferSource),
-                new GuiBookModelRenderer(bufferSource),
-                new GuiBannerResultRenderer(bufferSource, atlasManager),
-                new GuiSignRenderer(bufferSource, atlasManager),
-                new GuiProfilerChartRenderer(bufferSource)
-            )
+            net.neoforged.neoforge.client.ClientHooks.gatherPictureInPictureRenderers(List.of(
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiEntityRenderState.class, buffers -> new GuiEntityRenderer(buffers, minecraft.getEntityRenderDispatcher())),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiSkinRenderState.class, GuiSkinRenderer::new),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiBookModelRenderState.class, GuiBookModelRenderer::new),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiBannerResultRenderState.class, buffers -> new GuiBannerResultRenderer(buffers, atlasManager)),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiSignRenderState.class, buffers -> new GuiSignRenderer(buffers, atlasManager)),
+                new net.neoforged.neoforge.client.gui.PictureInPictureRendererRegistration<>(net.minecraft.client.gui.render.state.pip.GuiProfilerChartRenderState.class, GuiProfilerChartRenderer::new)
+            ))
         );
         this.screenEffectRenderer = new ScreenEffectRenderer(minecraft, atlasManager, bufferSource);
     }
@@ -229,6 +_,7 @@
                 break;
             case null:
             default:
+                if (!net.neoforged.neoforge.client.ClientHooks.loadEntityShader(cameraEntity, this))
                 this.clearPostEffect();
         }
     }
@@ -328,6 +_,9 @@
             if (hurt < 0.0F) {
                 return;
             }
+
+            // Neo: Prevent screen shake if the damage type is marked as "forge:no_flinch"
+            if (cameraState.entityRenderState.preventFlinch) return;
 
             hurt /= cameraState.entityRenderState.hurtDuration;
             hurt = Mth.sin(hurt * hurt * hurt * hurt * (float) Math.PI);
