--- a/net/minecraft/client/renderer/ItemBlockRenderTypes.java
+++ b/net/minecraft/client/renderer/ItemBlockRenderTypes.java
@@ -38,8 +_,13 @@
         };
     }
 
+    /// @deprecated Neo: use [#getBlockModelRenderType(BlockStateModel, net.minecraft.world.level.BlockAndTintGetter, net.minecraft.core.BlockPos, BlockState)] instead
     public static RenderType getBlockModelRenderType(BlockStateModel model) {
-        return model.hasTranslucency() ? Sheets.translucentBlockSheet() : Sheets.cutoutBlockSheet();
+        return getBlockModelRenderType(model, net.minecraft.world.level.EmptyBlockAndTintGetter.INSTANCE, net.minecraft.core.BlockPos.ZERO, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState());
+    }
+
+    public static RenderType getBlockModelRenderType(BlockStateModel model, net.minecraft.world.level.BlockAndTintGetter level, net.minecraft.core.BlockPos pos, BlockState state) {
+        return model.hasTranslucency(level, pos, state) ? Sheets.translucentBlockSheet() : Sheets.cutoutBlockSheet();
     }
 
     public static ChunkSectionLayer getRenderLayer(FluidState state) {
@@ -53,5 +_,19 @@
 
     public static void setCutoutLeaves(boolean cutoutLeaves) {
         ItemBlockRenderTypes.cutoutLeaves = cutoutLeaves;
+    }
+
+    /**
+     * Helper to set the RenderType for Fluids
+     */
+    public static synchronized void setRenderLayer(Fluid fluid, ChunkSectionLayer layer) {
+        checkClientLoading();
+        LAYER_BY_FLUID.put(fluid, layer);
+    }
+
+    private static void checkClientLoading() {
+        com.google.common.base.Preconditions.checkState(net.neoforged.neoforge.client.loading.ClientModLoader.isLoading(),
+                  "Render layers can only be set during client loading! This should ideally be done from `FMLClientSetupEvent`."
+        );
     }
 }
