--- a/net/minecraft/client/renderer/ItemBlockRenderTypes.java
+++ b/net/minecraft/client/renderer/ItemBlockRenderTypes.java
@@ -54,4 +_,29 @@
     public static void setCutoutLeaves(boolean cutoutLeaves) {
         ItemBlockRenderTypes.cutoutLeaves = cutoutLeaves;
     }
+
+    // Neo: RenderType for block setters injected
+    /**
+     * Helper to set the RenderType for Blocks
+     * @deprecated Set your render type in your block model's JSON (e.g. {@code "render_type": "cutout"}) or override {@link net.minecraft.client.renderer.block.model.BlockModelPart#getRenderType(BlockState)}
+     * */
+    @Deprecated(since = "1.19")
+    public static synchronized void setRenderLayer(Block block, ChunkSectionLayer layer) {
+        checkClientLoading();
+        TYPE_BY_BLOCK.put(block, layer);
+    }
+
+    /**
+     * Helper to set the RenderType for Fluids
+     */
+    public static synchronized void setRenderLayer(Fluid fluid, ChunkSectionLayer layer) {
+        checkClientLoading();
+        LAYER_BY_FLUID.put(fluid, layer);
+    }
+
+    private static void checkClientLoading() {
+        com.google.common.base.Preconditions.checkState(net.neoforged.neoforge.client.loading.ClientModLoader.isLoading(),
+                  "Render layers can only be set during client loading! This should ideally be done from `FMLClientSetupEvent`."
+        );
+    }
 }
