--- a/net/minecraft/client/renderer/ItemBlockRenderTypes.java
+++ b/net/minecraft/client/renderer/ItemBlockRenderTypes.java
@@ -54,4 +_,18 @@
     public static void setCutoutLeaves(boolean cutoutLeaves) {
         ItemBlockRenderTypes.cutoutLeaves = cutoutLeaves;
     }
+
+    /**
+     * Helper to set the RenderType for Fluids
+     */
+    public static synchronized void setRenderLayer(Fluid fluid, ChunkSectionLayer layer) {
+        checkClientLoading();
+        LAYER_BY_FLUID.put(fluid, layer);
+    }
+
+    private static void checkClientLoading() {
+        com.google.common.base.Preconditions.checkState(net.neoforged.neoforge.client.loading.ClientModLoader.isLoading(),
+                  "Render layers can only be set during client loading! This should ideally be done from `FMLClientSetupEvent`."
+        );
+    }
 }
