--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -314,6 +_,8 @@
             this.submitBlockOutline(poseStack, this.submitNodeStorage, levelRenderState);
         }
 
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.SubmitCustomGeometryEvent(levelRenderState, this.submitNodeStorage, poseStack, this.visibleSections));
+
         this.finalizeGizmoCollection();
         this.finalizedGizmos.standardPrimitives().submit(submitNodeCollector, levelRenderState.cameraRenderState, false);
         this.finalizedGizmos.alwaysOnTopPrimitives().submit(submitNodeCollector, levelRenderState.cameraRenderState, true);
@@ -336,7 +_,13 @@
         this.sectionRenderDispatcher.setCameraPosition(cameraPos);
     }
 
+    /// @deprecated Neo: use [#addSkyPass(FrameGraphBuilder, CameraRenderState, GpuBufferSlice, Matrix4fc)] instead
+    @Deprecated
     private void addSkyPass(FrameGraphBuilder frame, CameraRenderState cameraState, GpuBufferSlice skyFog) {
+        this.addSkyPass(frame, cameraState, skyFog, RenderSystem.getModelViewStack());
+    }
+
+    private void addSkyPass(FrameGraphBuilder frame, CameraRenderState cameraState, GpuBufferSlice skyFog, Matrix4fc modelViewMatrix) {
         FogType fogType = cameraState.fogType;
         if (fogType != FogType.POWDER_SNOW && fogType != FogType.LAVA && !cameraState.entityRenderState.doesMobEffectBlockSky) {
             if (this.levelRenderState.shouldResetSkyRenderer || this.skyRenderer == null) {
@@ -353,6 +_,7 @@
                 this.targets.main = pass.readsAndWrites(this.targets.main);
                 pass.executes(
                     () -> {
+                        if (this.levelRenderState.customSkyboxRenderer == null || !this.levelRenderState.customSkyboxRenderer.renderSky(levelRenderState, state, modelViewMatrix, () -> RenderSystem.setShaderFog(skyFog))) {
                         RenderSystem.setShaderFog(skyFog);
                         if (state.skybox == DimensionType.Skybox.END) {
                             this.skyRenderer.renderEndSky();
@@ -372,12 +_,18 @@
                                 this.skyRenderer.renderDarkDisc();
                             }
                         }
+                        } // Neo: End of skip if custom sky is in effect
+                        Profiler.get().push("neoforge_render_after_sky");
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterSky(this, this.levelRenderState, null, modelViewMatrix, this.visibleSections));
+                        Profiler.get().pop();
                     }
                 );
             }
         }
     }
 
+    /// @deprecated Neo: use [#addMainPass(FrameGraphBuilder, FeatureRenderDispatcher.PreparedFrame, GpuBufferSlice, LevelRenderState, ProfilerFiller, ChunkSectionsToRender, Matrix4fc)] instead
+    @Deprecated
     private void addMainPass(
         FrameGraphBuilder frame,
         FeatureRenderDispatcher.PreparedFrame featureFrame,
@@ -916,10 +_,10 @@
             section.setWasPreviouslyEmpty(false);
             if (rebuildSync) {
                 profiler.push("compileSectionSynchronously");
-                section.compileSync(state.region());
+                section.compileSync(state.region(), state.additionalRenderers());
                 profiler.pop();
             } else {
-                section.compileAsync(state.region());
+                section.compileAsync(state.region(), state.additionalRenderers());
             }
         }
 
@@ -977,7 +_,7 @@
                     poseStack.translate(state.blockState().getOffset(pos));
                     BlockStateModel model = this.modelManager.getBlockStateModelSet().get(state.blockState());
                     random.setSeed(state.blockState().getSeed(pos));
-                    model.collectParts(random, parts);
+                    model.collectParts(net.minecraft.client.renderer.block.BlockAndTintGetter.EMPTY, BlockPos.ZERO, state.blockState(), random, parts);
                     submitNodeCollector.submitBreakingBlockModel(poseStack, List.copyOf(parts), state.progress(), model.hasMaterialFlag(1));
                     parts.clear();
                     poseStack.popPose();
@@ -989,6 +_,13 @@
     private void submitBlockOutline(PoseStack poseStack, SubmitNodeCollector submitNodeCollector, LevelRenderState levelRenderState) {
         BlockOutlineRenderState state = levelRenderState.blockOutlineRenderState;
         if (state != null) {
+            boolean cancel = false;
+            for (net.neoforged.neoforge.client.CustomBlockOutlineRenderer customRenderer : state.customRenderers()) {
+                cancel |= customRenderer.render(state, submitNodeCollector, poseStack, levelRenderState);
+            }
+            if (cancel) {
+                return;
+            }
             Vec3 cameraPos = levelRenderState.cameraRenderState.pos;
             BlockPos pos = state.pos();
             poseStack.pushPose();
