--- a/net/minecraft/client/renderer/LevelEventHandler.java
+++ b/net/minecraft/client/renderer/LevelEventHandler.java
@@ -301,8 +_,8 @@
                 break;
             case 2001:
                 BlockState blockState = Block.stateById(data);
-                if (!blockState.isAir()) {
-                    SoundType soundType = blockState.getSoundType();
+                if (!blockState.isAir() && !net.neoforged.neoforge.client.extensions.common.IClientBlockExtensions.of(blockState).playBreakSound(blockState, this.level, pos)) {
+                    SoundType soundType = blockState.getSoundType(this.level, pos, null);
                     this.level
                         .playLocalSound(
                             pos, soundType.getBreakSound(), SoundSource.BLOCKS, (soundType.getVolume() + 1.0F) / 2.0F, soundType.getPitch() * 0.8F, false
