--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -496,7 +_,7 @@
         profiler.push("entities");
         this.extractVisibleEntities(camera, frustum, deltaTracker, this.levelRenderState);
         profiler.popPush("blockEntities");
-        this.extractVisibleBlockEntities(camera, deltaPartialTick, this.levelRenderState);
+        this.extractVisibleBlockEntities(camera, deltaPartialTick, this.levelRenderState, frustum);
         profiler.popPush("blockOutline");
         this.extractBlockOutline(camera, this.levelRenderState);
         profiler.popPush("blockBreaking");
@@ -516,6 +_,13 @@
             );
         profiler.popPush("particles");
         this.minecraft.particleEngine.extract(this.particlesRenderState, new Frustum(frustum).offset(-3.0F), camera, deltaPartialTick);
+        profiler.popPush("neoforge_custom");
+        this.levelRenderState.customWeatherEffectRenderer = net.neoforged.neoforge.client.CustomEnvironmentEffectsRendererManager.getCustomWeatherEffectRenderer(this.level, camera.position());
+        this.levelRenderState.customSkyboxRenderer = net.neoforged.neoforge.client.CustomEnvironmentEffectsRendererManager.getCustomSkyboxRenderer(this.level, camera.position());
+        this.levelRenderState.customCloudsRenderer = net.neoforged.neoforge.client.CustomEnvironmentEffectsRendererManager.getCustomCloudsRenderer(this.level, camera.position());
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ExtractLevelRenderStateEvent(
+                this, this.levelRenderState, this.level, camera, frustum, deltaTracker, this.ticks
+        ));
         profiler.pop();
         profiler.popPush("debug");
         this.debugRenderer.emitGizmos(frustum, cameraPos.x, cameraPos.y, cameraPos.z, deltaTracker.getGameTimeDeltaPartialTick(false));
@@ -528,7 +_,7 @@
         this.targets.main = frame.importExternal("main", this.minecraft.getMainRenderTarget());
         int screenWidth = this.minecraft.getMainRenderTarget().width;
         int screenHeight = this.minecraft.getMainRenderTarget().height;
-        RenderTargetDescriptor screenSizeTargetDescriptor = new RenderTargetDescriptor(screenWidth, screenHeight, true, 0);
+        RenderTargetDescriptor screenSizeTargetDescriptor = new RenderTargetDescriptor(screenWidth, screenHeight, true, 0, this.minecraft.getMainRenderTarget().useStencil);
         PostChain transparencyChain = this.getTransparencyChain();
         if (transparencyChain != null) {
             this.targets.translucent = frame.createInternal("translucent", screenSizeTargetDescriptor);
@@ -542,6 +_,9 @@
             this.targets.entityOutline = frame.importExternal("entity_outline", this.entityOutlineTarget);
         }
 
+        var setupEvent = net.neoforged.neoforge.client.ClientHooks.fireFrameGraphSetup(frame, this.targets, screenSizeTargetDescriptor, frustum, camera, modelViewMatrix, projectionMatrix, deltaTracker, profiler);
+        this.levelRenderState.haveGlowingEntities |= setupEvent.isOutlineProcessingEnabled();
+
         FramePass clearPass = frame.addPass("clear");
         this.targets.main = clearPass.readsAndWrites(this.targets.main);
         clearPass.executes(
@@ -558,7 +_,7 @@
             }
         );
         if (shouldRenderSky) {
-            this.addSkyPass(frame, camera, terrainFog);
+            this.addSkyPass(frame, camera, terrainFog, modelViewMatrix);
         }
 
         this.addMainPass(frame, frustum, modelViewMatrix, terrainFog, renderOutline, this.levelRenderState, deltaTracker, profiler);
@@ -573,12 +_,12 @@
             if (ARGB.alpha(cloudColor) > 0) {
                 float cloudHeight = camera.attributeProbe().getValue(EnvironmentAttributes.CLOUD_HEIGHT, deltaPartialTick);
                 this.addCloudsPass(
-                    frame, cloudsType, this.levelRenderState.cameraRenderState.pos, this.levelRenderState.gameTime, deltaPartialTick, cloudColor, cloudHeight
+                    frame, cloudsType, this.levelRenderState.cameraRenderState.pos, this.levelRenderState.gameTime, deltaPartialTick, cloudColor, cloudHeight, modelViewMatrix
                 );
             }
         }
 
-        this.addWeatherPass(frame, terrainFog);
+        this.addWeatherPass(frame, terrainFog, modelViewMatrix);
         if (transparencyChain != null) {
             transparencyChain.addToFrame(frame, screenWidth, screenHeight, this.targets);
         }
@@ -663,6 +_,9 @@
 
                 ChunkSectionsToRender chunkSectionsToRender = this.prepareChunkRenders(modelViewMatrix, camX, camY, camZ);
                 chunkSectionsToRender.renderGroup(ChunkSectionLayerGroup.OPAQUE, this.chunkLayerSampler);
+                profiler.push("neoforge_render_after_opaque_blocks");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterOpaqueBlocks(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
+                profiler.pop();
                 this.minecraft.gameRenderer.getLighting().setupFor(Lighting.Entry.LEVEL);
                 if (this.shouldShowEntityOutlines() && entityOutlineTarget != null) {
                     RenderTarget outlineTarget = entityOutlineTarget.get();
@@ -680,6 +_,9 @@
                 this.submitBlockEntities(poseStack, levelRenderState, this.submitNodeStorage);
                 profiler.popPush("submitParticles");
                 this.particlesRenderState.submit(this.submitNodeStorage, levelRenderState.cameraRenderState);
+                Profiler.get().push("neoforge_render_after_particles");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterParticles(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
+                Profiler.get().pop();
                 profiler.popPush("solidFeatures");
                 this.featureRenderDispatcher.renderSolidFeatures();
                 bufferSource.endBatch();
@@ -705,6 +_,8 @@
                 crumblingBufferSource.endBatch();
                 profiler.pop();
                 this.renderBuffers.outlineBufferSource().endOutlineBatch();
+                profiler.popPush("neoforge_render_after_entities");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterEntities(this, this.levelRenderState, poseStack, modelViewMatrix, this.getRenderableSections()));
                 if (renderOutline) {
                     this.renderBlockOutline(bufferSource, poseStack, false, levelRenderState);
                     bufferSource.endBatch();
@@ -716,6 +_,8 @@
                 this.checkPoseStack(poseStack);
                 profiler.push("translucentTerrain");
                 chunkSectionsToRender.renderGroup(ChunkSectionLayerGroup.TRANSLUCENT, this.chunkLayerSampler);
+                profiler.popPush("neoforge_render_after_translucent_blocks");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterTranslucentBlocks(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
                 if (renderOutline) {
                     this.renderBlockOutline(bufferSource, poseStack, true, levelRenderState);
                 }
@@ -730,9 +_,18 @@
         );
     }
 
+    /**
+     * @deprecated Neo: use {@link #addCloudsPass(FrameGraphBuilder, CloudStatus, Vec3, long, float, int, float, Matrix4f)} instead
+     */
+    @Deprecated
     private void addCloudsPass(
         FrameGraphBuilder frame, CloudStatus cloudsType, Vec3 cameraPosition, long gameTime, float partialTicks, int cloudColor, float cloudHeight
     ) {
+        addCloudsPass(frame, cloudsType, cameraPosition, gameTime, partialTicks, cloudColor, cloudHeight, RenderSystem.getModelViewMatrix());
+    }
+
+    // Neo: Add modelViewMatrix for use in our event
+    private void addCloudsPass(FrameGraphBuilder frame, CloudStatus cloudsType, Vec3 cameraPosition, long gameTime, float partialTicks, int cloudColor, float cloudHeight, Matrix4f modelViewMatrix) {
         FramePass pass = frame.addPass("clouds");
         if (this.targets.clouds != null) {
             this.targets.clouds = pass.readsAndWrites(this.targets.clouds);
@@ -740,10 +_,22 @@
             this.targets.main = pass.readsAndWrites(this.targets.main);
         }
 
-        pass.executes(() -> this.cloudRenderer.render(cloudColor, cloudsType, cloudHeight, cameraPosition, gameTime, partialTicks));
+        pass.executes(() -> {
+            if (this.levelRenderState.customCloudsRenderer == null || !this.levelRenderState.customCloudsRenderer.renderClouds(this.levelRenderState, cameraPosition, cloudsType, cloudColor, cloudHeight, modelViewMatrix)) {
+                this.cloudRenderer.render(cloudColor, cloudsType, cloudHeight, cameraPosition, gameTime, partialTicks);
+            }
+        });
     }
 
+    /**
+     * @deprecated Neo: use {@link #addWeatherPass(FrameGraphBuilder, GpuBufferSlice, Matrix4f)} instead
+     */
+    @Deprecated
     private void addWeatherPass(FrameGraphBuilder frame, GpuBufferSlice fog) {
+        addWeatherPass(frame, fog, RenderSystem.getModelViewMatrix());
+    }
+
+    private void addWeatherPass(FrameGraphBuilder frame, GpuBufferSlice fog, Matrix4f modelViewMatrix) {
         int renderDistance = this.minecraft.options.getEffectiveRenderDistance() * 16;
         float depthFar = this.minecraft.gameRenderer.getDepthFar();
         FramePass pass = frame.addPass("weather");
@@ -757,6 +_,9 @@
             RenderSystem.setShaderFog(fog);
             CameraRenderState cameraState = this.levelRenderState.cameraRenderState;
             this.weatherEffectRenderer.render(cameraState.pos, this.levelRenderState.weatherRenderState);
+            Profiler.get().push("neoforge_render_after_weather");
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterWeather(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
+            Profiler.get().pop();
             this.worldBorderRenderer.render(this.levelRenderState.worldBorderRenderState, cameraState.pos, renderDistance, depthFar);
         });
     }
@@ -808,7 +_,7 @@
                             || camera.isDetached()
                             || camera.entity() instanceof LivingEntity && ((LivingEntity)camera.entity()).isSleeping()
                     )
-                    && (!(entity instanceof LocalPlayer) || camera.entity() == entity)) {
+                    && (!(entity instanceof LocalPlayer) || camera.entity() == entity || (entity == minecraft.player && !minecraft.player.isSpectator()))) { // Neo: render local player entity when it is not the camera entity
                     if (entity.tickCount == 0) {
                         entity.xOld = entity.getX();
                         entity.yOld = entity.getY();
@@ -821,6 +_,9 @@
                     if (state.appearsGlowing() && shouldShowEntityOutlines) {
                         output.haveGlowingEntities = true;
                     }
+                    else if (shouldShowEntityOutlines && entity.hasCustomOutlineRendering(this.minecraft.player)) { // FORGE: allow custom outline rendering
+                        output.haveGlowingEntities = true;
+                    }
                 }
             }
         }
@@ -843,12 +_,20 @@
         }
     }
 
+    /**
+     * @deprecated Neo: use {@link #extractVisibleBlockEntities(Camera, float, LevelRenderState, Frustum)} instead
+     */
+    @Deprecated
     private void extractVisibleBlockEntities(Camera camera, float deltaPartialTick, LevelRenderState levelRenderState) {
+        extractVisibleBlockEntities(camera, deltaPartialTick, levelRenderState, null);
+    }
+    private void extractVisibleBlockEntities(Camera camera, float deltaPartialTick, LevelRenderState levelRenderState, @Nullable Frustum frustum) {
         Vec3 cameraPos = camera.position();
         double camX = cameraPos.x();
         double camY = cameraPos.y();
         double camZ = cameraPos.z();
         PoseStack poseStack = new PoseStack();
+        boolean shouldShowEntityOutlines = this.shouldShowEntityOutlines();
 
         for (SectionRenderDispatcher.RenderSection section : this.visibleSections) {
             List<BlockEntity> renderableBlockEntities = section.getSectionMesh().getRenderableBlockEntities();
@@ -866,9 +_,12 @@
                         breakProgress = null;
                     }
 
-                    BlockEntityRenderState state = this.blockEntityRenderDispatcher.tryExtractRenderState(blockEntity, deltaPartialTick, breakProgress);
+                    BlockEntityRenderState state = this.blockEntityRenderDispatcher.tryExtractRenderState(blockEntity, deltaPartialTick, breakProgress, frustum);
                     if (state != null) {
                         levelRenderState.blockEntityRenderStates.add(state);
+                        if (shouldShowEntityOutlines && blockEntity.hasCustomOutlineRendering(this.minecraft.player)) { // Neo: allow custom outline rendering
+                            levelRenderState.haveGlowingEntities = true;
+                        }
                     }
                 }
             }
@@ -881,9 +_,12 @@
             if (blockEntity.isRemoved()) {
                 iterator.remove();
             } else {
-                BlockEntityRenderState state = this.blockEntityRenderDispatcher.tryExtractRenderState(blockEntity, deltaPartialTick, null);
+                BlockEntityRenderState state = this.blockEntityRenderDispatcher.tryExtractRenderState(blockEntity, deltaPartialTick, null, frustum);
                 if (state != null) {
                     levelRenderState.blockEntityRenderStates.add(state);
+                    if (shouldShowEntityOutlines && blockEntity.hasCustomOutlineRendering(this.minecraft.player)) { // Neo: allow custom outline rendering
+                        levelRenderState.haveGlowingEntities = true;
+                    }
                 }
             }
         }
@@ -947,19 +_,23 @@
                 BlockPos pos = blockHitResult.getBlockPos();
                 BlockState state = this.level.getBlockState(pos);
                 if (!state.isAir() && this.level.getWorldBorder().isWithinBounds(pos)) {
-                    boolean isBlockTranslucent = ItemBlockRenderTypes.getChunkRenderType(state).sortOnUpload();
+                    boolean isBlockTranslucent = net.neoforged.neoforge.client.ClientHooks.isInTranslucentBlockOutlinePass(this.level, pos, state);
                     boolean highContrast = this.minecraft.options.highContrastBlockOutline().get();
                     CollisionContext context = CollisionContext.of(camera.entity());
+                    var event = new net.neoforged.neoforge.client.event.ExtractBlockOutlineRenderStateEvent(this, this.level, pos, state, blockHitResult, context, isBlockTranslucent, highContrast, camera, levelRenderState);
+                    if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled()) {
+                        return;
+                    }
                     VoxelShape shape = state.getShape(this.level, pos, context);
                     if (SharedConstants.DEBUG_SHAPES) {
                         VoxelShape collisionShape = state.getCollisionShape(this.level, pos, context);
                         VoxelShape occlusionShape = state.getOcclusionShape();
                         VoxelShape interactionShape = state.getInteractionShape(this.level, pos);
                         levelRenderState.blockOutlineRenderState = new BlockOutlineRenderState(
-                            pos, isBlockTranslucent, highContrast, shape, collisionShape, occlusionShape, interactionShape
+                            pos, isBlockTranslucent, highContrast, shape, collisionShape, occlusionShape, interactionShape, event.getCustomRenderers()
                         );
                     } else {
-                        levelRenderState.blockOutlineRenderState = new BlockOutlineRenderState(pos, isBlockTranslucent, highContrast, shape);
+                        levelRenderState.blockOutlineRenderState = new BlockOutlineRenderState(pos, isBlockTranslucent, highContrast, shape, event.getCustomRenderers());
                     }
                 }
             }
@@ -971,6 +_,13 @@
     ) {
         BlockOutlineRenderState state = levelRenderState.blockOutlineRenderState;
         if (state != null) {
+            boolean cancel = false;
+            for (net.neoforged.neoforge.client.CustomBlockOutlineRenderer customRenderer : state.customRenderers()) {
+                cancel |= customRenderer.render(state, bufferSource, poseStack, onlyTranslucentBlocks, levelRenderState);
+            }
+            if (cancel) {
+                return;
+            }
             if (state.isTranslucent() == onlyTranslucentBlocks) {
                 Vec3 cameraPos = levelRenderState.cameraRenderState.pos;
                 if (state.highContrast()) {
@@ -1152,7 +_,15 @@
         }
     }
 
+    /**
+     * @deprecated Neo: use {@link #addSkyPass(FrameGraphBuilder, Camera, GpuBufferSlice, Matrix4f)} instead
+     */
+    @Deprecated
     private void addSkyPass(FrameGraphBuilder frame, Camera camera, GpuBufferSlice skyFog) {
+        this.addSkyPass(frame, camera, skyFog, RenderSystem.getModelViewMatrix());
+    }
+
+    private void addSkyPass(FrameGraphBuilder frame, Camera camera, GpuBufferSlice skyFog, Matrix4f modelViewMatrix) {
         FogType fogType = camera.getFluidInCamera();
         if (fogType != FogType.POWDER_SNOW && fogType != FogType.LAVA && !this.doesMobEffectBlockSky(camera)) {
             SkyRenderState state = this.levelRenderState.skyRenderState;
@@ -1163,6 +_,7 @@
                     this.targets.main = pass.readsAndWrites(this.targets.main);
                     pass.executes(
                         () -> {
+                            if (this.levelRenderState.customSkyboxRenderer == null || !this.levelRenderState.customSkyboxRenderer.renderSky(levelRenderState, state, modelViewMatrix, () -> RenderSystem.setShaderFog(skyFog))) {
                             RenderSystem.setShaderFog(skyFog);
                             if (state.skybox == DimensionType.Skybox.END) {
                                 skyRenderer.renderEndSky();
@@ -1181,6 +_,10 @@
                                     skyRenderer.renderDarkDisc();
                                 }
                             }
+                            } // Neo: End of skip if custom sky is in effect
+                            Profiler.get().push("neoforge_render_after_sky");
+                            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterSky(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
+                            Profiler.get().pop();
                         }
                     );
                 }
@@ -1403,7 +_,7 @@
         } else {
             int packedBrightness = brightnessGetter.packedBrightness(level, pos);
             int block = LightCoordsUtil.block(packedBrightness);
-            int blockSelfEmission = state.getLightEmission();
+            int blockSelfEmission = state.getLightEmission(level, pos);
             return block < blockSelfEmission ? LightCoordsUtil.withBlock(packedBrightness, blockSelfEmission) : packedBrightness;
         }
     }
@@ -1455,6 +_,22 @@
 
     public CloudRenderer getCloudRenderer() {
         return this.cloudRenderer;
+    }
+
+    public int getTicks() {
+        return this.ticks;
+    }
+
+    public void iterateVisibleBlockEntities(java.util.function.Consumer<BlockEntity> blockEntityConsumer) {
+        for (var chunkInfo : this.visibleSections) {
+            chunkInfo.getSectionMesh().getRenderableBlockEntities().forEach(blockEntityConsumer);
+        }
+        this.level.getGloballyRenderedBlockEntities().forEach(blockEntityConsumer);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public Iterable<? extends net.neoforged.neoforge.client.IRenderableSection> getRenderableSections() {
+        return this.visibleSections;
     }
 
     public Gizmos.TemporaryCollection collectPerFrameGizmos() {
