--- a/net/minecraft/client/renderer/LightTexture.java
+++ b/net/minecraft/client/renderer/LightTexture.java
@@ -205,4 +_,20 @@
             return pack(j, i);
         }
     }
+
+    // Neo: block and sky only return the integer part. So we add these methods that work with fractional parts too.
+    // A full block or sky value is encoded as `0xij` where i is the integer part and j is the fractional part.
+    // Shifting by 4 then masking with 15 (= 0xF) only returns the integer part.
+
+    public static int packWithFraction(int block, int sky) {
+        return block | sky << 16;
+    }
+
+    public static int blockWithFraction(int light) {
+        return light & 0xFF;
+    }
+
+    public static int skyWithFraction(int light) {
+        return light >>> 16 & 0xFF;
+    }
 }
