--- a/net/minecraft/client/renderer/MapRenderer.java
+++ b/net/minecraft/client/renderer/MapRenderer.java
@@ -43,6 +_,10 @@
 
         for (MapRenderState.MapDecorationRenderState maprenderstate$mapdecorationrenderstate : p_360365_.decorations) {
             if (!p_362262_ || maprenderstate$mapdecorationrenderstate.renderOnFrame) {
+                if (net.neoforged.neoforge.client.gui.map.MapDecorationRendererManager.render(maprenderstate$mapdecorationrenderstate, p_362483_, p_434609_, p_360365_, MapRenderer.this.decorationSprites, p_362262_, p_364872_, i)) {
+                    i++;
+                    continue;
+                }
                 p_362483_.pushPose();
                 p_362483_.translate(maprenderstate$mapdecorationrenderstate.x / 2.0F + 64.0F, maprenderstate$mapdecorationrenderstate.y / 2.0F + 64.0F, -0.02F);
                 p_362483_.mulPose(Axis.ZP.rotationDegrees(maprenderstate$mapdecorationrenderstate.rot * 360 / 16.0F));
@@ -113,13 +_,15 @@
         p_364922_.texture = this.mapTextureManager.prepareMapTexture(p_361383_, p_363500_);
         p_364922_.decorations.clear();
 
+        net.neoforged.neoforge.client.renderstate.RenderStateExtensions.onUpdateMapRenderState(p_363500_, p_364922_);
         for (MapDecoration mapdecoration : p_363500_.getDecorations()) {
-            p_364922_.decorations.add(this.extractDecorationRenderState(mapdecoration));
+            p_364922_.decorations.add(net.neoforged.neoforge.client.renderstate.RenderStateExtensions.onUpdateMapDecorationRenderState(mapdecoration.type(), p_363500_, p_364922_, this.extractDecorationRenderState(mapdecoration)));
         }
     }
 
     private MapRenderState.MapDecorationRenderState extractDecorationRenderState(MapDecoration p_364175_) {
         MapRenderState.MapDecorationRenderState maprenderstate$mapdecorationrenderstate = new MapRenderState.MapDecorationRenderState();
+        maprenderstate$mapdecorationrenderstate.type = p_364175_.type();
         maprenderstate$mapdecorationrenderstate.atlasSprite = this.decorationSprites.getSprite(p_364175_.getSpriteLocation());
         maprenderstate$mapdecorationrenderstate.x = p_364175_.x();
         maprenderstate$mapdecorationrenderstate.y = p_364175_.y();
