--- a/net/minecraft/client/renderer/MapRenderer.java
+++ b/net/minecraft/client/renderer/MapRenderer.java
@@ -43,6 +_,10 @@
 
         for (MapRenderState.MapDecorationRenderState decoration : mapRenderState.decorations) {
             if (!showOnlyFrame || decoration.renderOnFrame) {
+                if (net.neoforged.neoforge.client.gui.map.MapDecorationRendererManager.render(decoration, poseStack, submitNodeCollector, mapRenderState, MapRenderer.this.decorationSprites, showOnlyFrame, lightCoords, count)) {
+                    count++;
+                    continue;
+                }
                 poseStack.pushPose();
                 poseStack.translate(decoration.x / 2.0F + 64.0F, decoration.y / 2.0F + 64.0F, -0.02F);
                 poseStack.mulPose(Axis.ZP.rotationDegrees(decoration.rot * 360 / 16.0F));
@@ -84,13 +_,15 @@
         mapRenderState.texture = this.mapTextureManager.prepareMapTexture(mapId, mapData);
         mapRenderState.decorations.clear();
 
+        net.neoforged.neoforge.client.renderstate.RenderStateExtensions.onUpdateMapRenderState(mapData, mapRenderState);
         for (MapDecoration decoration : mapData.getDecorations()) {
-            mapRenderState.decorations.add(this.extractDecorationRenderState(decoration));
+            mapRenderState.decorations.add(net.neoforged.neoforge.client.renderstate.RenderStateExtensions.onUpdateMapDecorationRenderState(decoration.type(), mapData, mapRenderState, this.extractDecorationRenderState(decoration)));
         }
     }
 
     private MapRenderState.MapDecorationRenderState extractDecorationRenderState(MapDecoration decoration) {
         MapRenderState.MapDecorationRenderState state = new MapRenderState.MapDecorationRenderState();
+        state.type = decoration.type();
         state.atlasSprite = this.decorationSprites.getSprite(decoration.getSpriteLocation());
         state.x = decoration.x();
         state.y = decoration.y();
