--- a/net/minecraft/client/renderer/OrderedSubmitNodeCollector.java
+++ b/net/minecraft/client/renderer/OrderedSubmitNodeCollector.java
@@ -132,6 +_,8 @@
 
     void submitMovingBlock(PoseStack poseStack, MovingBlockRenderState movingBlockRenderState);
 
+    /** @deprecated Neo: use {@link #submitBlockModel(PoseStack, RenderType, BlockStateModel, float, float, float, int, int, int, BlockState)} instead */
+    @Deprecated
     void submitBlockModel(
         PoseStack poseStack, RenderType renderType, BlockStateModel model, float r, float g, float b, int lightCoords, int overlayCoords, int outlineColor
     );
