--- a/net/minecraft/client/renderer/PanoramaRenderer.java
+++ b/net/minecraft/client/renderer/PanoramaRenderer.java
@@ -10,7 +_,7 @@
 @OnlyIn(Dist.CLIENT)
 public class PanoramaRenderer {
     public static final ResourceLocation PANORAMA_OVERLAY = ResourceLocation.withDefaultNamespace("textures/gui/title/background/panorama_overlay.png");
-    private final Minecraft minecraft;
+    private Minecraft minecraft;
     private final CubeMap cubeMap;
     private float spin;
 
