--- a/net/minecraft/client/renderer/PostChain.java
+++ b/net/minecraft/client/renderer/PostChain.java
@@ -103,9 +_,9 @@
             ResourceLocation resourcelocation1 = entry.getKey();
             RenderTargetDescriptor rendertargetdescriptor = switch (entry.getValue()) {
                 case PostChainConfig.FixedSizedTarget(int i, int j) -> {
-                    yield new RenderTargetDescriptor(i, j, true);
+                    yield new RenderTargetDescriptor(i, j, true, net.neoforged.neoforge.client.ClientHooks.isStencilEnabled());
                 }
-                case PostChainConfig.FullScreenTarget postchainconfig$fullscreentarget -> new RenderTargetDescriptor(p_361423_, p_362735_, true);
+                case PostChainConfig.FullScreenTarget postchainconfig$fullscreentarget -> new RenderTargetDescriptor(p_361423_, p_362735_, true, net.neoforged.neoforge.client.ClientHooks.isStencilEnabled());
                 default -> throw new MatchException(null, null);
             };
             map.put(resourcelocation1, p_362523_.createInternal(resourcelocation1.toString(), rendertargetdescriptor));
