--- a/net/minecraft/client/renderer/PostPass.java
+++ b/net/minecraft/client/renderer/PostPass.java
@@ -142,7 +_,7 @@
             ResourceHandle<RenderTarget> resourcehandle = this.getHandle(p_361239_);
             RenderTarget rendertarget = resourcehandle.get();
             rendertarget.setFilterMode(this.bilinear ? 9729 : 9728);
-            p_366564_.bindSampler(this.samplerName + "Sampler", this.depthBuffer ? rendertarget.getDepthTextureId() : rendertarget.getColorTextureId());
+            p_366564_.bindSampler(this.samplerName + "Sampler", this.depthBuffer() ? rendertarget.getDepthTextureId() : rendertarget.getColorTextureId());
             p_366564_.safeGetUniform(this.samplerName + "Size").set((float)rendertarget.width, (float)rendertarget.height);
         }
 
