--- a/net/minecraft/client/renderer/PostChain.java
+++ b/net/minecraft/client/renderer/PostChain.java
@@ -156,7 +_,8 @@
                                 throw new ChainedJsonException("Render target '" + s4 + "' can't be used as depth buffer");
                             }
 
-                            ResourceLocation resourcelocation = ResourceLocation.withDefaultNamespace("textures/effect/" + s4 + ".png");
+                            ResourceLocation rl = ResourceLocation.tryParse(s4);
+                            ResourceLocation resourcelocation = ResourceLocation.fromNamespaceAndPath(rl.getNamespace(), "textures/effect/" + rl.getPath() + ".png");
                             this.resourceProvider
                                 .getResource(resourcelocation)
                                 .orElseThrow(() -> new ChainedJsonException("Render target or texture '" + s4 + "' does not exist"));
@@ -257,6 +_,7 @@
     public void addTempTarget(String name, int width, int height) {
         RenderTarget rendertarget = new TextureTarget(width, height, true, Minecraft.ON_OSX);
         rendertarget.setClearColor(0.0F, 0.0F, 0.0F, 0.0F);
+        if (screenTarget.isStencilEnabled()) { rendertarget.enableStencil(); }
         this.customRenderTargets.put(name, rendertarget);
         if (width == this.screenWidth && height == this.screenHeight) {
             this.fullSizedTargets.add(rendertarget);
