--- a/net/minecraft/client/renderer/RenderPipelines.java
+++ b/net/minecraft/client/renderer/RenderPipelines.java
@@ -744,4 +_,14 @@
     public static List<RenderPipeline> getStaticPipelines() {
         return PIPELINES_BY_LOCATION.values().stream().toList();
     }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static void registerCustomPipelines() {
+        var event = new net.neoforged.neoforge.client.event.RegisterRenderPipelinesEvent(pipeline -> {
+            if (PIPELINES_BY_LOCATION.putIfAbsent(pipeline.getLocation(), pipeline) != null) {
+                throw new IllegalStateException("Duplicate RenderPipeline registration for location: " + pipeline.getLocation());
+            }
+        });
+        net.neoforged.fml.ModLoader.postEvent(event);
+    }
 }
