--- a/net/minecraft/client/renderer/RenderBuffers.java
+++ b/net/minecraft/client/renderer/RenderBuffers.java
@@ -32,6 +_,7 @@
             put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map, RenderTypes.entityGlint());
             put((Object2ObjectLinkedOpenHashMap<RenderType, ByteBufferBuilder>)map, RenderTypes.waterMask());
         });
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.RegisterRenderBuffersEvent(fixedBuffers));
         this.bufferSource = MultiBufferSource.immediateWithBuffers(fixedBuffers, new ByteBufferBuilder(786432));
         this.outlineBufferSource = new OutlineBufferSource();
         SequencedMap<RenderType, ByteBufferBuilder> crumblingBuffers = Util.make(
