--- a/net/minecraft/client/renderer/RenderType.java
+++ b/net/minecraft/client/renderer/RenderType.java
@@ -1049,6 +_,9 @@
                     if (scissorstate.enabled()) {
                         renderpass.enableScissor(scissorstate.x(), scissorstate.y(), scissorstate.width(), scissorstate.height());
                     }
+                    if (RenderSystem.STENCIL_TEST != null) {
+                        renderpass.enableStencilTest(RenderSystem.STENCIL_TEST);
+                    }
 
                     RenderSystem.bindDefaultUniforms(renderpass);
                     renderpass.setUniform("DynamicTransforms", gpubufferslice);
