--- a/net/minecraft/client/renderer/ShaderManager.java
+++ b/net/minecraft/client/renderer/ShaderManager.java
@@ -169,7 +_,7 @@
         ShaderManager.CompilationCache shadermanager$compilationcache = new ShaderManager.CompilationCache(p_366597_);
         Map<ShaderProgram, ShaderManager.CompilationException> map = new HashMap<>();
 
-        for (ShaderProgram shaderprogram : CoreShaders.getProgramsToPreload()) {
+        for (ShaderProgram shaderprogram : net.neoforged.neoforge.client.CoreShaderManager.getProgramsToPreload()) {
             try {
                 shadermanager$compilationcache.programs.put(shaderprogram, Optional.of(shadermanager$compilationcache.compileProgram(shaderprogram)));
             } catch (ShaderManager.CompilationException shadermanager$compilationexception) {
