--- a/net/minecraft/client/renderer/Sheets.java
+++ b/net/minecraft/client/renderer/Sheets.java
@@ -171,12 +_,14 @@
         p_110781_.accept(ENDER_CHEST_LOCATION);
     }
 
-    public static Material createSignMaterial(WoodType p_173386_) {
-        return new Material(SIGN_SHEET, new ResourceLocation("entity/signs/" + p_173386_.name()));
+    private static Material createSignMaterial(WoodType p_173386_) {
+        ResourceLocation location = new ResourceLocation(p_173386_.name());
+        return new Material(SIGN_SHEET, new ResourceLocation(location.getNamespace(), "entity/signs/" + location.getPath()));
     }
 
-    public static Material createHangingSignMaterial(WoodType p_251735_) {
-        return new Material(SIGN_SHEET, new ResourceLocation("entity/signs/hanging/" + p_251735_.name()));
+    private static Material createHangingSignMaterial(WoodType p_251735_) {
+        ResourceLocation location = new ResourceLocation(p_251735_.name());
+        return new Material(SIGN_SHEET, new ResourceLocation(location.getNamespace(), "entity/signs/hanging/" + location.getPath()));
     }
 
     public static Material getSignMaterial(WoodType p_173382_) {
@@ -237,6 +_,30 @@
             case SINGLE:
             default:
                 return p_110773_;
+        }
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static void init() {
+        var event = new net.neoforged.neoforge.client.event.RegisterClientWoodTypeMaterialsEvent(SIGN_MATERIALS, HANGING_SIGN_MATERIALS);
+        net.neoforged.fml.ModLoader.get().postEvent(event);
+    }
+
+    /**
+     * @deprecated Use {@link net.neoforged.neoforge.client.event.RegisterClientWoodTypeMaterialsEvent} instead
+     */
+    @Deprecated(forRemoval = true)
+    public static void addWoodType(WoodType woodType) {
+        SIGN_MATERIALS.put(woodType, createSignMaterial(woodType));
+        HANGING_SIGN_MATERIALS.put(woodType, createHangingSignMaterial(woodType));
+    }
+
+    static {
+        if (net.neoforged.fml.ModLoader.isLoadingStateValid() && !net.neoforged.fml.ModLoader.get().hasCompletedState("LOAD_REGISTRIES")) {
+            com.mojang.logging.LogUtils.getLogger().error(
+                      "net.minecraft.client.renderer.Sheets loaded too early, modded registry-based materials may not work correctly",
+                      new IllegalStateException("net.minecraft.client.renderer.Sheets loaded too early")
+            );
         }
     }
 }
