--- a/net/minecraft/client/renderer/Sheets.java
+++ b/net/minecraft/client/renderer/Sheets.java
@@ -146,11 +_,13 @@
     }
 
     private static Material createSignMaterial(WoodType woodType) {
-        return new Material(SIGN_SHEET, ResourceLocation.withDefaultNamespace("entity/signs/" + woodType.name()));
+        ResourceLocation location = ResourceLocation.parse(woodType.name());
+        return new Material(SIGN_SHEET, ResourceLocation.fromNamespaceAndPath(location.getNamespace(), "entity/signs/" + location.getPath()));
     }
 
     private static Material createHangingSignMaterial(WoodType woodType) {
-        return new Material(SIGN_SHEET, ResourceLocation.withDefaultNamespace("entity/signs/hanging/" + woodType.name()));
+        ResourceLocation location = ResourceLocation.parse(woodType.name());
+        return new Material(SIGN_SHEET, ResourceLocation.fromNamespaceAndPath(location.getNamespace(), "entity/signs/hanging/" + location.getPath()));
     }
 
     public static Material getSignMaterial(WoodType woodType) {
@@ -209,6 +_,23 @@
             case SINGLE:
             default:
                 return doubleMaterial;
+        }
+    }
+
+    /**
+     * Not threadsafe. Enqueue it in client setup.
+     */
+    public static void addWoodType(WoodType woodType) {
+        SIGN_MATERIALS.put(woodType, createSignMaterial(woodType));
+        HANGING_SIGN_MATERIALS.put(woodType, createHangingSignMaterial(woodType));
+    }
+
+    static {
+        if (!net.neoforged.fml.ModLoader.hasErrors() && !net.neoforged.neoforge.internal.CommonModLoader.areRegistriesLoaded()) {
+            com.mojang.logging.LogUtils.getLogger().error(
+                      "net.minecraft.client.renderer.Sheets loaded too early, modded registry-based materials may not work correctly",
+                      new IllegalStateException("net.minecraft.client.renderer.Sheets loaded too early")
+            );
         }
     }
 }
