--- a/net/minecraft/client/renderer/Sheets.java
+++ b/net/minecraft/client/renderer/Sheets.java
@@ -146,11 +_,11 @@
     }
 
     private static SpriteId createSignSprite(WoodType type) {
-        return SIGN_MAPPER.defaultNamespaceApply(type.name());
+        return SIGN_MAPPER.apply(Identifier.parse(type.name()));
     }
 
     private static SpriteId createHangingSignSprite(WoodType type) {
-        return HANGING_SIGN_MAPPER.defaultNamespaceApply(type.name());
+        return HANGING_SIGN_MAPPER.apply(Identifier.parse(type.name()));
     }
 
     public static SpriteId getSignSprite(WoodType type) {
