--- a/net/minecraft/client/renderer/SpriteCoordinateExpander.java
+++ b/net/minecraft/client/renderer/SpriteCoordinateExpander.java
@@ -17,12 +_,14 @@
 
     @Override
     public VertexConsumer addVertex(float x, float y, float z) {
-        return this.delegate.addVertex(x, y, z);
+        this.delegate.addVertex(x, y, z);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setColor(int r, int g, int b, int a) {
-        return this.delegate.setColor(r, g, b, a);
+        this.delegate.setColor(r, g, b, a);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
@@ -32,22 +_,26 @@
 
     @Override
     public VertexConsumer setUv(float u, float v) {
-        return this.delegate.setUv(this.sprite.getU(u), this.sprite.getV(v));
+        this.delegate.setUv(this.sprite.getU(u), this.sprite.getV(v));
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv1(int u, int v) {
-        return this.delegate.setUv1(u, v);
+        this.delegate.setUv1(u, v);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv2(int u, int v) {
-        return this.delegate.setUv2(u, v);
+        this.delegate.setUv2(u, v);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setNormal(float x, float y, float z) {
-        return this.delegate.setNormal(x, y, z);
+        this.delegate.setNormal(x, y, z);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
