--- a/net/minecraft/client/renderer/SpriteCoordinateExpander.java
+++ b/net/minecraft/client/renderer/SpriteCoordinateExpander.java
@@ -17,32 +_,38 @@
 
     @Override
     public VertexConsumer addVertex(float x, float y, float z) {
-        return this.delegate.addVertex(x, y, z);
+        this.delegate.addVertex(x, y, z);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setColor(int red, int green, int blue, int alpha) {
-        return this.delegate.setColor(red, green, blue, alpha);
+        this.delegate.setColor(red, green, blue, alpha);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv(float u, float v) {
-        return this.delegate.setUv(this.sprite.getU(u), this.sprite.getV(v));
+        this.delegate.setUv(this.sprite.getU(u), this.sprite.getV(v));
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv1(int u, int v) {
-        return this.delegate.setUv1(u, v);
+        this.delegate.setUv1(u, v);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setUv2(int u, int v) {
-        return this.delegate.setUv2(u, v);
+        this.delegate.setUv2(u, v);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
     public VertexConsumer setNormal(float normalX, float normalY, float normalZ) {
-        return this.delegate.setNormal(normalX, normalY, normalZ);
+        this.delegate.setNormal(normalX, normalY, normalZ);
+        return this; //Neo: Fix MC-263524 not working with chained methods
     }
 
     @Override
