--- a/net/minecraft/client/renderer/SubmitNodeStorage.java
+++ b/net/minecraft/client/renderer/SubmitNodeStorage.java
@@ -173,8 +_,13 @@
 
     @OnlyIn(Dist.CLIENT)
     public record BlockModelSubmit(
-        PoseStack.Pose pose, RenderType renderType, BlockStateModel model, float r, float g, float b, int lightCoords, int overlayCoords, int outlineColor
+        PoseStack.Pose pose, RenderType renderType, BlockStateModel model, float r, float g, float b, int lightCoords, int overlayCoords, int outlineColor, @Nullable BlockState state
     ) {
+        /** @deprecated Neo: use {@link #BlockModelSubmit(PoseStack.Pose,RenderType,BlockStateModel,float,float,float,int,int,int,BlockState)} instead */
+        @Deprecated
+        public BlockModelSubmit(PoseStack.Pose pose, RenderType renderType, BlockStateModel model, float r, float g, float b, int lightCoords, int overlayCoords, int outlineColor) {
+            this(pose, renderType, model, r, g, b, lightCoords, overlayCoords, outlineColor, null);
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
