--- a/net/minecraft/client/renderer/WeatherEffectRenderer.java
+++ b/net/minecraft/client/renderer/WeatherEffectRenderer.java
@@ -220,6 +_,10 @@
     }
 
     public void tickRainParticles(ClientLevel level, Camera camera, int ticks, ParticleStatus particleStatus, int weatherRadius) {
+        var customWeatherEffectRenderer = net.neoforged.neoforge.client.CustomEnvironmentEffectsRendererManager.getCustomWeatherEffectRenderer(level, camera.position());
+        if (customWeatherEffectRenderer != null && customWeatherEffectRenderer.tickRain(level, ticks, camera))
+            return;
+
         float rainLevel = level.getRainLevel(1.0F);
         if (!(rainLevel <= 0.0F)) {
             RandomSource random = RandomSource.create(ticks * 312987231L);
