--- a/net/minecraft/client/renderer/WeatherEffectRenderer.java
+++ b/net/minecraft/client/renderer/WeatherEffectRenderer.java
@@ -117,7 +_,17 @@
         return this.vertexBuffer;
     }
 
+    /// @deprecated Neo: use [#render(Vec3, WeatherRenderState, net.minecraft.client.renderer.state.level.LevelRenderState)] instead
+    @Deprecated
     public void render(Vec3 cameraPos, WeatherRenderState renderState) {
+        this.render(cameraPos, renderState, null);
+    }
+
+    public void render(Vec3 cameraPos, WeatherRenderState renderState, net.minecraft.client.renderer.state.level.@org.jspecify.annotations.Nullable LevelRenderState levelRenderState) {
+        if (levelRenderState != null && levelRenderState.customWeatherEffectRenderer != null && levelRenderState.customWeatherEffectRenderer.renderSnowAndRain(levelRenderState, renderState, cameraPos)) {
+            return;
+        }
+
         int columnCount = renderState.rainColumns.size() + renderState.snowColumns.size();
         if (columnCount != 0) {
             TextureManager textureManager = Minecraft.getInstance().getTextureManager();
