--- a/net/minecraft/client/renderer/WeatherEffectRenderer.java
+++ b/net/minecraft/client/renderer/WeatherEffectRenderer.java
@@ -156,6 +_,8 @@
     }
 
     public void tickRainParticles(ClientLevel p_361823_, Camera p_364990_, int p_361788_, ParticleStatus p_363302_) {
+        if (p_361823_.effects().tickRain(p_361823_, p_361788_, p_364990_))
+            return;
         float f = p_361823_.getRainLevel(1.0F) / (Minecraft.useFancyGraphics() ? 1.0F : 2.0F);
         if (!(f <= 0.0F)) {
             RandomSource randomsource = RandomSource.create(p_361788_ * 312987231L);
