--- a/com/mojang/blaze3d/resource/RenderTargetDescriptor.java
+++ b/com/mojang/blaze3d/resource/RenderTargetDescriptor.java
@@ -9,10 +_,14 @@
 import org.joml.Vector4fc;
 
 @OnlyIn(Dist.CLIENT)
-public record RenderTargetDescriptor(int width, int height, boolean useDepth, Vector4fc clearColor, GpuFormat format)
+public record RenderTargetDescriptor(int width, int height, boolean useDepth, boolean useStencil, Vector4fc clearColor, GpuFormat format)
     implements ResourceDescriptor<RenderTarget> {
+    public RenderTargetDescriptor(int width, int height, boolean useDepth, Vector4fc clearColor, GpuFormat format) {
+        this(width, height, useDepth, false, clearColor, format);
+    }
+
     public RenderTarget allocate() {
-        return new TextureTarget(null, this.width, this.height, this.useDepth, this.format);
+        return new TextureTarget(null, this.width, this.height, this.useDepth, this.useStencil, this.format);
     }
 
     public void prepare(RenderTarget resource) {
@@ -23,6 +_,9 @@
         } else {
             RenderSystem.getDevice().createCommandEncoder().clearColorTexture(resource.getColorTexture(), this.clearColor);
         }
+        if (this.useStencil) {
+            RenderSystem.getDevice().createCommandEncoder().clearStencilTexture(resource.getDepthTexture(), 0);
+        }
     }
 
     public void free(RenderTarget resource) {
@@ -33,6 +_,6 @@
     public boolean canUsePhysicalResource(ResourceDescriptor<?> other) {
         return !(other instanceof RenderTargetDescriptor descriptor)
             ? false
-            : this.width == descriptor.width && this.height == descriptor.height && this.useDepth == descriptor.useDepth && this.format == descriptor.format;
+            : this.width == descriptor.width && this.height == descriptor.height && this.useDepth == descriptor.useDepth && this.format == descriptor.format && this.useStencil == descriptor.useStencil;
     }
 }
