--- a/net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
@@ -37,7 +37,7 @@
             if (recipe.isPresent()) {
                 ItemStack result = recipe.get().value().assemble(input);
                 if (!result.isEmpty()) {
-                    return result.copyWithCount(itemStack.getCount());
+                    return result.copyWithCount(itemStack.getCount() * result.getCount()); // Forge: Support smelting returning multiple
                 }
             }
 
