--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -180,7 +180,7 @@
         int screenWidth = this.gameRenderer.mainRenderTarget().width;
         int screenHeight = this.gameRenderer.mainRenderTarget().height;
         RenderTargetDescriptor screenSizeTargetDescriptor = new RenderTargetDescriptor(
-            screenWidth, screenHeight, true, SCREEN_SIZE_TARGET_CLEAR_COLOR, GpuFormat.RGBA8_UNORM
+            screenWidth, screenHeight, true, this.gameRenderer.mainRenderTarget().useStencil, SCREEN_SIZE_TARGET_CLEAR_COLOR, GpuFormat.RGBA8_UNORM
         );
         PostChain transparencyChain = this.getTransparencyChain();
         if (transparencyChain != null) {
@@ -225,11 +227,12 @@
                 deltaPartialTick,
                 this.levelRenderState.cloudColor,
                 this.levelRenderState.cloudHeight,
-                this.optionsRenderState.cloudRange
+                this.optionsRenderState.cloudRange,
+                modelViewMatrix
             );
         }
 
-        this.addWeatherPass(frame, terrainFog);
+        this.addWeatherPass(frame, terrainFog, modelViewMatrix);
         if (transparencyChain != null) {
             transparencyChain.addToFrame(frame, screenWidth, screenHeight, this.targets);
         }
@@ -459,10 +517,20 @@
             this.targets.main = pass.readsAndWrites(this.targets.main);
         }
 
-        pass.executes(() -> this.cloudRenderer.render(cloudColor, cloudStatus, cloudHeight, cloudRange, cameraPosition, gameTime, partialTicks));
+        pass.executes(() -> {
+            if (this.levelRenderState.customCloudsRenderer == null || !this.levelRenderState.customCloudsRenderer.renderClouds(this.levelRenderState, cameraPosition, cloudStatus, cloudColor, cloudHeight, cloudRange, modelViewMatrix)) {
+                this.cloudRenderer.render(cloudColor, cloudStatus, cloudHeight, cloudRange, cameraPosition, gameTime, partialTicks);
+            }
+        });
     }
 
@@ -475,7 +543,10 @@
             () -> {
                 RenderSystem.setShaderFog(fog);
                 CameraRenderState cameraState = this.levelRenderState.cameraRenderState;
-                this.weatherEffectRenderer.render(cameraState.pos, this.levelRenderState.weatherRenderState);
+                this.weatherEffectRenderer.render(cameraState.pos, this.levelRenderState.weatherRenderState, this.levelRenderState);
                 Profiler.get().push("neoforge_render_after_weather");
                 net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterWeather(this, this.levelRenderState, null, modelViewMatrix, this.visibleSections));
                 Profiler.get().pop();
                 this.worldBorderRenderer
                     .render(this.levelRenderState.worldBorderRenderState, cameraState.pos, renderDistance, this.levelRenderState.cameraRenderState.depthFar);
             }
