--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -180,7 +180,7 @@
         int screenWidth = this.gameRenderer.mainRenderTarget().width;
         int screenHeight = this.gameRenderer.mainRenderTarget().height;
         RenderTargetDescriptor screenSizeTargetDescriptor = new RenderTargetDescriptor(
-            screenWidth, screenHeight, true, SCREEN_SIZE_TARGET_CLEAR_COLOR, GpuFormat.RGBA8_UNORM
+            screenWidth, screenHeight, true, this.gameRenderer.mainRenderTarget().useStencil, SCREEN_SIZE_TARGET_CLEAR_COLOR, GpuFormat.RGBA8_UNORM
         );
         PostChain transparencyChain = this.getTransparencyChain();
         if (transparencyChain != null) {
