--- a/net/minecraft/client/renderer/ScreenEffectRenderer.java
+++ b/net/minecraft/client/renderer/ScreenEffectRenderer.java
@@ -122,6 +129,10 @@
     }
 
     private static @Nullable BlockState getViewBlockingState(Player player) {
+        return net.minecraft.Optionull.map(getViewBlockingStateAndPos(player), org.apache.commons.lang3.tuple.Pair::getLeft);
+    }
+
+    private static org.apache.commons.lang3.tuple.@Nullable Pair<BlockState, BlockPos> getViewBlockingStateAndPos(Player player) {
         if (player.noPhysics) {
             return null;
         }
@@ -136,7 +147,7 @@
             );
             BlockState blockState = player.level().getBlockState(testPos);
             if (blockState.getRenderShape() != RenderShape.INVISIBLE && blockState.isViewBlocking(player.level(), testPos)) {
-                return blockState;
+                return org.apache.commons.lang3.tuple.Pair.of(blockState, testPos.immutable());
             }
         }
 
