--- a/net/minecraft/commands/arguments/selector/EntitySelector.java
+++ b/net/minecraft/commands/arguments/selector/EntitySelector.java
@@ -104,7 +_,7 @@
     }
 
     private void checkPermissions(CommandSourceStack p_121169_) throws CommandSyntaxException {
-        if (this.usesSelector && !p_121169_.allowsSelectors()) {
+        if (this.usesSelector && !net.neoforged.neoforge.common.CommonHooks.canUseEntitySelectors(p_121169_)) {
             throw EntityArgument.ERROR_SELECTORS_NOT_ALLOWED.create();
         }
     }
@@ -150,8 +_,9 @@
             } else {
                 Predicate<Entity> predicate = this.getPredicate(vec3, aabb, p_121161_.enabledFeatures());
                 List<Entity> list = new ObjectArrayList<>();
+                // TODO: remove all this messy temp stuff
                 if (this.isWorldLimited()) {
-                    this.addEntities(list, p_121161_.getLevel(), aabb, predicate);
+                    this.addEntities(list, p_121161_.getUnsidedLevel(), aabb, predicate);
                 } else {
                     for (ServerLevel serverlevel : p_121161_.getServer().getAllLevels()) {
                         this.addEntities(list, serverlevel, aabb, predicate);
@@ -163,13 +_,21 @@
         }
     }
 
-    private void addEntities(List<Entity> p_121155_, ServerLevel p_121156_, @Nullable AABB p_352947_, Predicate<Entity> p_121158_) {
+    private void addEntities(List<Entity> p_121155_, net.minecraft.world.level.Level p_121156_, @Nullable AABB p_352947_, Predicate<Entity> p_121158_) {
         int i = this.getResultLimit();
         if (p_121155_.size() < i) {
             if (p_352947_ != null) {
                 p_121156_.getEntities(this.type, p_352947_, p_121158_, p_121155_, i);
             } else {
-                p_121156_.getEntities(this.type, p_121158_, p_121155_, i);
+                if (p_121156_ instanceof ServerLevel serverLevel) {
+                    serverLevel.getEntities(this.type, p_121158_, p_121155_, i);
+                } else {
+                    ((net.minecraft.client.multiplayer.ClientLevel) p_121156_).entitiesForRendering().forEach(entity -> {
+                        if (p_121158_.test(entity) && p_121155_.size() < i) {
+                            p_121155_.add(entity);
+                        }
+                    });
+                }
             }
         }
     }
