--- a/com/mojang/blaze3d/framegraph/FrameGraphBuilder.java
+++ b/com/mojang/blaze3d/framegraph/FrameGraphBuilder.java
@@ -211,6 +_,11 @@
         public String toString() {
             return this.createdBy != null ? this.holder + "#" + this.version + " (from " + this.createdBy + ")" : this.holder + "#" + this.version;
         }
+
+        @Nullable
+        public ResourceDescriptor<T> getDescriptor() {
+            return this.holder.getDescriptor();
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
@@ -265,6 +_,11 @@
                 this.physicalResource = null;
             }
         }
+
+        @Override
+        public ResourceDescriptor<T> getDescriptor() {
+            return descriptor;
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
@@ -363,6 +_,11 @@
         @Override
         public String toString() {
             return this.name;
+        }
+
+        @Nullable
+        public ResourceDescriptor<T> getDescriptor() {
+            return null;
         }
     }
 }
