--- a/net/minecraft/world/level/storage/loot/functions/CopyComponentsFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/CopyComponentsFunction.java
@@ -75,7 +_,7 @@
         return new CopyComponentsFunction.Builder(p_331133_);
     }
 
-    public static class Builder extends LootItemConditionalFunction.Builder<CopyComponentsFunction.Builder> {
+    public static class Builder extends LootItemConditionalFunction.Builder<CopyComponentsFunction.Builder> implements net.neoforged.neoforge.common.extensions.ICopyComponentsFunctionBuilderExtensions {
         private final CopyComponentsFunction.Source source;
         private Optional<ImmutableList.Builder<DataComponentType<?>>> include = Optional.empty();
         private Optional<ImmutableList.Builder<DataComponentType<?>>> exclude = Optional.empty();
