--- a/net/minecraft/world/level/storage/loot/functions/EnchantRandomlyFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/EnchantRandomlyFunction.java
@@ -72,7 +_,7 @@
         Stream<Holder<Enchantment>> compatibleEnchantmentsStream = this.options
             .map(HolderSet::stream)
             .orElseGet(() -> context.getLevel().registryAccess().lookupOrThrow(Registries.ENCHANTMENT).listElements().map(Function.identity()))
-            .filter(candidate -> !shouldCheckCompatibility || candidate.value().canEnchant(itemStack));
+            .filter(candidate -> !shouldCheckCompatibility || itemStack.supportsEnchantment(candidate)); // Neo: Respect IItemExtension#supportsEnchantment
         List<Holder<Enchantment>> compatibleEnchantments = compatibleEnchantmentsStream.toList();
         Optional<Holder<Enchantment>> enchantment = Util.getRandomSafe(compatibleEnchantments, random);
         if (enchantment.isEmpty()) {
