--- a/net/minecraft/world/level/storage/loot/functions/SetComponentsFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetComponentsFunction.java
@@ -37,4 +_,9 @@
     public static <T> LootItemConditionalFunction.Builder<?> setComponent(DataComponentType<T> p_331454_, T p_331398_) {
         return simpleBuilder(p_331753_ -> new SetComponentsFunction(p_331753_, DataComponentPatch.builder().set(p_331454_, p_331398_).build()));
     }
+
+    // Neo: Utility for modded component types, to remove the need to invoke '.value()'
+    public static <T> LootItemConditionalFunction.Builder<?> setComponent(java.util.function.Supplier<? extends DataComponentType<T>> componentType, T value) {
+        return setComponent(componentType.get(), value);
+    }
 }
