--- a/net/minecraft/world/level/storage/loot/functions/ToggleTooltips.java
+++ b/net/minecraft/world/level/storage/loot/functions/ToggleTooltips.java
@@ -56,7 +_,7 @@
     );
     private final Map<ToggleTooltips.ComponentToggle<?>, Boolean> values;
 
-    public ToggleTooltips(List<LootItemCondition> p_335479_, Map<ToggleTooltips.ComponentToggle<?>, Boolean> p_336164_) {
+    private ToggleTooltips(List<LootItemCondition> p_335479_, Map<ToggleTooltips.ComponentToggle<?>, Boolean> p_336164_) {
         super(p_335479_);
         this.values = p_336164_;
     }
