--- /dev/null
+++ b/net/minecraft/world/level/storage/loot/functions/ToggleTooltips.java.rej
@@ -1,0 +_,10 @@
+diff a/projects/neoforge/src/main/java/net/minecraft/world/level/storage/loot/functions/ToggleTooltips.java b/projects/neoforge/src/main/java/net/minecraft/world/level/storage/loot/functions/ToggleTooltips.java	(rejected hunks)
+@@ -22,7 +22,7 @@ public class ToggleTooltips extends LootItemConditionalFunction {
+     );
+     private final Map<DataComponentType<?>, Boolean> values;
+ 
+-    private ToggleTooltips(List<LootItemCondition> p_335479_, Map<DataComponentType<?>, Boolean> p_336164_) {
++    public ToggleTooltips(List<LootItemCondition> p_335479_, Map<DataComponentType<?>, Boolean> p_336164_) {
+         super(p_335479_);
+         this.values = p_336164_;
+     }
