--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -11,7 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
 
-public interface Container extends Clearable, Iterable<ItemStack> {
+public interface Container extends Clearable, Iterable<ItemStack>, net.neoforged.neoforge.common.extensions.IContainerExtension {
     float DEFAULT_DISTANCE_BUFFER = 4.0F;
 
     int getContainerSize();
