--- a/net/minecraft/world/SimpleContainer.java
+++ b/net/minecraft/world/SimpleContainer.java
@@ -127,8 +_,14 @@
 
     @Override
     public void setItem(int p_19162_, ItemStack p_19163_) {
+        setItem(p_19162_, p_19163_, false);
+    }
+
+    @Override
+    public void setItem(int p_19162_, ItemStack p_19163_, boolean insideTransaction) {
         this.items.set(p_19162_, p_19163_);
         p_19163_.limitSize(this.getMaxStackSize(p_19163_));
+        if (!insideTransaction)
         this.setChanged();
     }
 
