--- a/com/mojang/blaze3d/opengl/GlCommandEncoder.java
+++ b/com/mojang/blaze3d/opengl/GlCommandEncoder.java
@@ -203,7 +203,8 @@
             throw new IllegalStateException("Close the existing render pass before creating a new one!");
         } else {
             this.verifyDepthTexture(depthTexture);
-            this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, 0, ((GlTexture)depthTexture).id, 0, 36160);
+            boolean hasStencil = depthTexture.getFormat().hasStencilAspect();
+            this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, 0, ((GlTexture)depthTexture).id, 0, 36160, hasStencil);
             GL11.glDrawBuffer(0);
             GL11.glClearDepth(clearDepth);
             GlStateManager._depthMask(true);
@@ -215,6 +216,23 @@
         }
     }
 
+    @Override
+    public void clearStencilTexture(GpuTexture texture, int value) {
+        if (this.inRenderPass) {
+            throw new IllegalStateException("Close the existing render pass before creating a new one!");
+        } else if (!texture.getFormat().hasStencilAspect()) {
+            throw new IllegalStateException("Trying to clear stencil in a texture that has no stencil component!");
+        } else {
+            this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, 0, ((GlTexture)texture).id, 0, GlConst.GL_FRAMEBUFFER, true);
+            GL11.glDrawBuffer(GlConst.GL_NONE);
+            GL11.glClearStencil(value);
+            GlStateManager._depthMask(true);
+            GlStateManager._clear(GL11.GL_STENCIL_BUFFER_BIT);
+            GL11.glDrawBuffer(GlConst.GL_COLOR_ATTACHMENT0);
+            GlStateManager._glBindFramebuffer(GlConst.GL_FRAMEBUFFER, 0);
+        }
+    }
+
     private void verifyColorTexture(GpuTexture colorTexture) {
         if (!colorTexture.getFormat().hasColorAspect()) {
             throw new IllegalStateException("Trying to clear a non-color texture as color");
@@ -636,16 +654,46 @@
             } else if (destination.getDepthOrLayers() > 1) {
                 throw new UnsupportedOperationException("Textures with multiple depths or layers are not yet supported for copying");
             } else {
+                var sourceFormat = source.getFormat();
+                var destFormat = destination.getFormat();
+
+                if (sourceFormat.hasDepthAspect() || sourceFormat.hasStencilAspect()) {
+                    if (sourceFormat != destFormat) {
+                        throw new IllegalArgumentException(
+                                "When copying depth or stencil data, the source and destination texture formats must be identical. Source: "
+                                        + sourceFormat + ", Destination: " + destFormat
+                        );
+                    }
+                }
+
+                if (sourceFormat.hasColorAspect() != destFormat.hasColorAspect()) {
+                    throw new IllegalArgumentException(
+                            "Source and destination texture formats must have consistent color aspects. Source: "
+                                    + sourceFormat + ", Destination: " + destFormat
+                    );
+                }
+
                 GlStateManager.clearGlErrors();
                 GlStateManager._disableScissorTest();
                 boolean isDepth = source.getFormat().hasDepthAspect();
                 int sourceId = ((GlTexture)source).glId();
                 int destId = ((GlTexture)destination).glId();
-                this.device.directStateAccess().bindFrameBufferTextures(this.readFbo, isDepth ? 0 : sourceId, isDepth ? sourceId : 0, 0, 0);
-                this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, isDepth ? 0 : destId, isDepth ? destId : 0, 0, 0);
+                var hasStencil = source.getFormat().hasStencilAspect();
+                this.device.directStateAccess().bindFrameBufferTextures(this.readFbo, isDepth ? 0 : sourceId, isDepth ? sourceId : 0, 0, 0, hasStencil);
+                this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, isDepth ? 0 : destId, isDepth ? destId : 0, 0, 0, hasStencil);
+                var bufferMask = 0;
+                if (source.getFormat().hasColorAspect()) {
+                    bufferMask |= GlConst.GL_COLOR_BUFFER_BIT;
+                }
+                if (source.getFormat().hasDepthAspect()) {
+                    bufferMask |= GlConst.GL_DEPTH_BUFFER_BIT;
+                }
+                if (source.getFormat().hasStencilAspect()) {
+                    bufferMask |= GL11.GL_STENCIL_BUFFER_BIT;
+                }
                 this.device
                     .directStateAccess()
-                    .blitFrameBuffers(this.readFbo, this.drawFbo, sourceX, sourceY, width, height, destX, destY, width, height, isDepth ? 256 : 16384, 9728);
+                    .blitFrameBuffers(this.readFbo, this.drawFbo, sourceX, sourceY, width, height, destX, destY, width, height, bufferMask, 9728);
                 int error = GlStateManager._getError();
                 if (error != 0) {
                     throw new IllegalStateException(
