--- a/net/minecraft/tags/TagBuilder.java
+++ b/net/minecraft/tags/TagBuilder.java
@@ -5,8 +_,17 @@
 import net.minecraft.resources.Identifier;
 
 public class TagBuilder implements net.neoforged.neoforge.common.extensions.ITagBuilderExtension {
-    private final List<TagEntry> entries = new ArrayList<>();
+    // FORGE: Remove entries are used for datagen.
+    private final List<TagEntry> removeEntries = new ArrayList<>();
+    public java.util.stream.Stream<TagEntry> getRemoveEntries() { return this.removeEntries.stream(); }
+    // FORGE: Add an entry to be removed from this tag in datagen.
+    public TagBuilder remove(final TagEntry entry) {
+        this.removeEntries.add(entry);
+        return this;
+    }
+    // FORGE: is this tag set to replace or not?
     private boolean replace = false;
+    private final List<TagEntry> entries = new ArrayList<>();
 
     public static TagBuilder create() {
         return new TagBuilder();
@@ -44,5 +_,21 @@
 
     public TagBuilder addOptionalTag(Identifier id) {
         return this.add(TagEntry.optionalTag(id));
+    }
+
+    // FORGE: Set the replace property of this tag.
+    public TagBuilder replace(boolean value) {
+        this.replace = value;
+        return this;
+    }
+
+    // FORGE: Shorthand version of replace(true)
+    public TagBuilder replace() {
+        return replace(true);
+    }
+
+    // FORGE: Is this tag set to replace or not?
+    public boolean isReplace() {
+        return this.replace;
     }
 }
