--- a/net/minecraft/tags/TagBuilder.java
+++ b/net/minecraft/tags/TagBuilder.java
@@ -7,6 +_,8 @@
 public class TagBuilder implements net.neoforged.neoforge.common.extensions.ITagBuilderExtension {
     private final List<TagEntry> entries = new ArrayList<>();
     private boolean replace = false;
+    /// Neo: Entries to be removed from this tag
+    private final List<TagEntry> removeEntries = new ArrayList<>();
 
     public static TagBuilder create() {
         return new TagBuilder();
@@ -44,5 +_,23 @@
 
     public TagBuilder addOptionalTag(Identifier id) {
         return this.add(TagEntry.optionalTag(id));
+    }
+
+    /// Neo: Add an entry to be removed from this tag in datagen.
+    ///
+    /// @param entry The entry to be removed
+    public TagBuilder remove(TagEntry entry) {
+        this.removeEntries.add(entry);
+        return this;
+    }
+
+    /// Neo: Shorthand for `setReplace(true)`.
+    public TagBuilder replace() {
+        return setReplace(true);
+    }
+
+    /// Neo: Returns the entries to be removed from this tag.
+    public java.util.stream.Stream<TagEntry> getRemoveEntries() {
+        return this.removeEntries.stream();
     }
 }
