--- a/net/minecraft/tags/TagEntry.java
+++ b/net/minecraft/tags/TagEntry.java
@@ -40,6 +_,15 @@
         return new ExtraCodecs.TagOrElementLocation(this.id, this.tag);
     }
 
+    /**
+     * {@return a copy of this entry with the required flag set to the given parameter}
+     *
+     * @param required whether the new entry is required
+     */
+    TagEntry withRequired(boolean required) {
+        return new TagEntry(this.id, this.tag, required);
+    }
+
     public static TagEntry element(Identifier id) {
         return new TagEntry(id, false, true);
     }
@@ -105,6 +_,19 @@
         }
 
         return result.toString();
+    }
+
+    // Neo: Used in datagen
+    public Identifier getId() {
+        return id;
+    }
+
+    public boolean isRequired() {
+        return required;
+    }
+
+    public boolean isTag() {
+        return tag;
     }
 
     public interface Lookup<T> {
