--- a/net/minecraft/tags/TagEntry.java
+++ b/net/minecraft/tags/TagEntry.java
@@ -111,11 +111,21 @@
       return stringbuilder.toString();
    }
 
+   public ResourceLocation getId() {
+      return id;
+   }
+
+   public boolean isRequired() {
+      return required;
+   }
+
+   public boolean isTag() {
+      return tag;
+   }
+
    public interface Lookup<T> {
       @Nullable
       T element(ResourceLocation p_215956_);
-
-      @Nullable
       Collection<T> tag(ResourceLocation p_215957_);
    }
 }
