--- a/net/minecraft/data/advancements/AdvancementProvider.java
+++ b/net/minecraft/data/advancements/AdvancementProvider.java
@@ -15,6 +_,11 @@
 import net.minecraft.data.PackOutput;
 import net.minecraft.resources.ResourceLocation;
 
+/**
+ * @deprecated NeoForge: Use {@link net.neoforged.neoforge.common.data.AdvancementProvider} instead,
+ *                 provides ease of access for the {@link net.neoforged.neoforge.common.data.ExistingFileHelper} in the generator
+ */
+@Deprecated
 public class AdvancementProvider implements DataProvider {
     private final PackOutput.PathProvider pathProvider;
     private final List<AdvancementSubProvider> subProviders;
@@ -36,7 +_,7 @@
                     throw new IllegalStateException("Duplicate advancement " + p_323113_.id());
                 } else {
                     Path path = this.pathProvider.json(p_323113_.id());
-                    list.add(DataProvider.saveStable(p_254268_, p_323115_, Advancement.CODEC, p_323113_.value(), path));
+                    list.add(DataProvider.saveStable(p_254268_, p_323115_, Advancement.CODEC, p_323113_.value(), path));// TODO: make conditional
                 }
             };
 
