--- a/net/minecraft/data/advancements/AdvancementProvider.java
+++ b/net/minecraft/data/advancements/AdvancementProvider.java
@@ -14,6 +14,11 @@
 import net.minecraft.data.PackOutput;
 import net.minecraft.resources.ResourceLocation;
 
+/**
+ * @deprecated NeoForge: Use {@link net.neoforged.neoforge.common.data.AdvancementProvider} instead,
+ *             provides ease of access for the {@link net.neoforged.neoforge.common.data.ExistingFileHelper} in the generator
+ */
+@Deprecated
 public class AdvancementProvider implements DataProvider {
    private final PackOutput.PathProvider pathProvider;
    private final List<AdvancementSubProvider> subProviders;
