--- a/net/minecraft/client/data/models/blockstates/MultiVariantGenerator.java
+++ b/net/minecraft/client/data/models/blockstates/MultiVariantGenerator.java
@@ -53,6 +_,18 @@
         return new MultiVariantGenerator(this.block, list, this.seenProperties);
     }
 
+    // Neo: copy of the apply(...VariantMutator...) methods to support UnbakedMutators. Must be renamed due to type erasure for PropertyDispatch
+    public MultiVariantGenerator withUnbaked(PropertyDispatch<net.neoforged.neoforge.client.model.generators.blockstate.UnbakedMutator> dispatch) {
+        Set<Property<?>> set = validateAndExpandProperties(this.seenProperties, this.block, dispatch);
+        List<MultiVariantGenerator.Entry> list = this.entries.stream().flatMap(e -> e.applyToUnbaked(dispatch)).toList();
+        return new MultiVariantGenerator(this.block, list, set);
+    }
+
+    public MultiVariantGenerator withUnbaked(net.neoforged.neoforge.client.model.generators.blockstate.UnbakedMutator dispatch) {
+        List<MultiVariantGenerator.Entry> list = this.entries.stream().flatMap(e -> e.applyToUnbaked(dispatch)).toList();
+        return new MultiVariantGenerator(this.block, list, this.seenProperties);
+    }
+
     @Override
     public BlockModelDefinition create() {
         Map<String, BlockStateModel.Unbaked> map = new HashMap<>();
@@ -107,6 +_,18 @@
         }
 
         public Stream<MultiVariantGenerator.Entry> apply(VariantMutator p_404774_) {
+            return Stream.of(new MultiVariantGenerator.Entry(this.properties, this.variant.with(p_404774_)));
+        }
+
+        public Stream<MultiVariantGenerator.Entry> applyToUnbaked(PropertyDispatch<net.neoforged.neoforge.client.model.generators.blockstate.UnbakedMutator> p_404760_) {
+            return p_404760_.getEntries().entrySet().stream().map(p_409007_ -> {
+                PropertyValueList propertyvaluelist = this.properties.extend(p_409007_.getKey());
+                MultiVariant multivariant = this.variant.with(p_409007_.getValue());
+                return new MultiVariantGenerator.Entry(propertyvaluelist, multivariant);
+            });
+        }
+
+        public Stream<MultiVariantGenerator.Entry> applyToUnbaked(net.neoforged.neoforge.client.model.generators.blockstate.UnbakedMutator p_404774_) {
             return Stream.of(new MultiVariantGenerator.Entry(this.properties, this.variant.with(p_404774_)));
         }
     }
