--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -49,6 +_,10 @@
     @Nullable
     private UUID target;
     public final float bobOffs;
+    /**
+     * The maximum age of this EntityItem.  The item is expired once this is reached.
+     */
+    public int lifespan = ItemEntity.LIFETIME;
 
     public ItemEntity(EntityType<? extends ItemEntity> entityType, Level level) {
         super(entityType, level);
@@ -67,6 +_,7 @@
         this.setPos(posX, posY, posZ);
         this.setDeltaMovement(deltaX, deltaY, deltaZ);
         this.setItem(itemStack);
+        this.lifespan = (itemStack.getItem() == null ? ItemEntity.LIFETIME : itemStack.getEntityLifespan(level));
     }
 
     private ItemEntity(ItemEntity other) {
@@ -75,6 +_,7 @@
         this.copyPosition(other);
         this.age = other.age;
         this.bobOffs = other.bobOffs;
+        this.lifespan = other.lifespan;
     }
 
     @Override
@@ -120,6 +_,7 @@
 
     @Override
     public void tick() {
+        if (getItem().onEntityItemUpdate(this)) return;
         if (this.getItem().isEmpty()) {
             this.discard();
         } else {
@@ -132,6 +_,9 @@
             this.yo = this.getY();
             this.zo = this.getZ();
             Vec3 vec3 = this.getDeltaMovement();
+            net.neoforged.neoforge.fluids.FluidType fluidType = this.getMaxHeightFluidType();
+            if (!fluidType.isAir() && !fluidType.isVanilla() && this.getFluidTypeHeight(fluidType) > 0.1F) fluidType.setItemMovement(this);
+            else
             if (this.isInWater() && this.getFluidHeight(FluidTags.WATER) > 0.1F) {
                 this.setUnderwaterMovement();
             } else if (this.isInLava() && this.getFluidHeight(FluidTags.LAVA) > 0.1F) {
@@ -153,7 +_,8 @@
                 this.move(MoverType.SELF, this.getDeltaMovement());
                 float f = 0.98F;
                 if (this.onGround()) {
-                    f = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                    BlockPos groundPos = getBlockPosBelowThatAffectsMyMovement();
+                    f = this.level().getBlockState(groundPos).getFriction(level(), groundPos, this) * 0.98F;
                 }
 
                 this.setDeltaMovement(this.getDeltaMovement().multiply((double)f, 0.98, (double)f));
@@ -185,7 +_,16 @@
                 }
             }
 
-            if (!this.level().isClientSide && this.age >= 6000) {
+            ItemStack item = this.getItem();
+            if (!this.level().isClientSide && this.age >= lifespan) {
+                 int hook = net.neoforged.neoforge.event.EventHooks.onItemExpire(this, item);
+                 if (hook < 0) {
+                     this.discard();
+                 } else {
+                     this.lifespan += hook;
+                 }
+            }
+            if (item.isEmpty() && !this.isRemoved()) {
                 this.discard();
             }
         }
@@ -282,7 +_,7 @@
             this.health = (int)((float)this.health - amount);
             this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
             if (this.health <= 0) {
-                this.getItem().onDestroyed(this);
+                this.getItem().onDestroyed(this, source);
                 this.discard();
             }
 
@@ -295,6 +_,7 @@
         compound.putShort("Health", (short)this.health);
         compound.putShort("Age", (short)this.age);
         compound.putShort("PickupDelay", (short)this.pickupDelay);
+        compound.putInt("Lifespan", this.lifespan);
         if (this.thrower != null) {
             compound.putUUID("Thrower", this.thrower);
         }
@@ -315,6 +_,9 @@
         if (compound.contains("PickupDelay")) {
             this.pickupDelay = compound.getShort("PickupDelay");
         }
+        if (compound.contains("Lifespan")) {
+            this.lifespan = compound.getInt("Lifespan");
+        }
 
         if (compound.hasUUID("Owner")) {
             this.target = compound.getUUID("Owner");
@@ -343,7 +_,22 @@
             ItemStack itemstack = this.getItem();
             Item item = itemstack.getItem();
             int i = itemstack.getCount();
-            if (this.pickupDelay == 0 && (this.target == null || this.target.equals(entity.getUUID())) && entity.getInventory().add(itemstack)) {
+
+            // Neo: Fire item pickup pre/post and adjust handling logic to adhere to the event result.
+            var result = net.neoforged.neoforge.event.EventHooks.fireItemPickupPre(this, entity).canPickup();
+            if (result.isFalse()) {
+                return;
+            }
+
+            // Make a copy of the original stack for use in ItemEntityPickupEvent.Post
+            ItemStack originalCopy = itemstack.copy();
+            // Subvert the vanilla conditions (pickup delay and target check) if the result is true.
+            if ((result.isTrue() || this.pickupDelay == 0 && (this.target == null || this.target.equals(entity.getUUID()))) && entity.getInventory().add(itemstack)) {
+                // Fire ItemEntityPickupEvent.Post
+                net.neoforged.neoforge.event.EventHooks.fireItemPickupPost(this, entity, originalCopy);
+                // Update `i` to reflect the actual pickup amount. Vanilla is wrong here and always reports the whole stack.
+                i = originalCopy.getCount() - itemstack.getCount();
+
                 entity.take(this, i);
                 if (itemstack.isEmpty()) {
                     this.discard();
@@ -398,6 +_,15 @@
         this.target = target;
     }
 
+    /**
+     * Neo: Add getter for ItemEntity's {@link ItemEntity#target target}.
+     * @return The {@link ItemEntity#target target} that can pick up this item entity (if {@code null}, anyone can pick it up)
+     */
+    @Nullable
+    public UUID getTarget() {
+        return this.target;
+    }
+
     public void setThrower(Entity thrower) {
         this.thrower = thrower.getUUID();
         this.cachedThrower = thrower;
@@ -437,7 +_,7 @@
 
     public void makeFakeItem() {
         this.setNeverPickUp();
-        this.age = 5999;
+        this.age = getItem().getEntityLifespan(this.level()) - 1;
     }
 
     public float getSpin(float partialTicks) {
