--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -51,6 +_,10 @@
     private @Nullable EntityReference<Entity> thrower;
     private @Nullable UUID target;
     public final float bobOffs = this.random.nextFloat() * (float) Math.PI * 2.0F;
+    /**
+     * The maximum age of this EntityItem. The item is expired once this is reached.
+     */
+    public int lifespan = ItemEntity.LIFETIME;
 
     public ItemEntity(EntityType<? extends ItemEntity> type, Level level) {
         super(type, level);
@@ -66,6 +_,7 @@
         this.setPos(x, y, z);
         this.setDeltaMovement(deltaX, deltaY, deltaZ);
         this.setItem(itemStack);
+        this.lifespan = (itemStack.getItem() == null ? ItemEntity.LIFETIME : itemStack.getEntityLifespan(level));
     }
 
     @Override
@@ -103,6 +_,7 @@
 
     @Override
     public void tick() {
+        if (getItem().onEntityItemUpdate(this)) return;
         if (this.getItem().isEmpty()) {
             this.discard();
         } else {
@@ -115,6 +_,9 @@
             this.yo = this.getY();
             this.zo = this.getZ();
             Vec3 oldMovement = this.getDeltaMovement();
+            net.neoforged.neoforge.fluids.FluidType fluidType = this.getMaxHeightFluidType();
+            if (!fluidType.isAir() && !fluidType.isVanilla() && this.getFluidTypeHeight(fluidType) > 0.1F) fluidType.setItemMovement(this);
+            else
             if (this.isInWater() && this.getFluidHeight(FluidTags.WATER) > 0.1F) {
                 this.setUnderwaterMovement();
             } else if (this.isInLava() && this.getFluidHeight(FluidTags.LAVA) > 0.1F) {
@@ -137,7 +_,8 @@
                 this.applyEffectsFromBlocks();
                 float friction = 0.98F;
                 if (this.onGround()) {
-                    friction = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+                    BlockPos groundPos = getBlockPosBelowThatAffectsMyMovement();
+                    friction = this.level().getBlockState(groundPos).getFriction(level(), groundPos, this) * 0.98F;
                 }
 
                 this.setDeltaMovement(this.getDeltaMovement().multiply(friction, 0.98, friction));
@@ -169,7 +_,15 @@
                 }
             }
 
-            if (!this.level().isClientSide() && this.age >= 6000) {
+            ItemStack item = this.getItem();
+            if (!this.level().isClientSide() && this.age >= lifespan) {
+            	  // Clamping to MAX_VALUE -1 as age is a Short and going above that would produce an infinite lifespan implicitly (accidentally)
+                this.lifespan = Mth.clamp(lifespan + net.neoforged.neoforge.event.EventHooks.onItemExpire(this), 0, Short.MAX_VALUE - 1);
+                if (this.age >= lifespan) {
+                    this.discard();
+                }
+            }
+            if (item.isEmpty() && !this.isRemoved()) {
                 this.discard();
             }
         }
@@ -279,7 +_,7 @@
             this.health = (int)(this.health - damage);
             this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
             if (this.health <= 0) {
-                this.getItem().onDestroyed(this);
+                this.getItem().onDestroyed(this, source);
                 this.discard();
             }
 
@@ -297,6 +_,7 @@
         output.putShort("Health", (short)this.health);
         output.putShort("Age", (short)this.age);
         output.putShort("PickupDelay", (short)this.pickupDelay);
+        output.putInt("Lifespan", this.lifespan);
         EntityReference.store(this.thrower, output, "Thrower");
         output.storeNullable("Owner", UUIDUtil.CODEC, this.target);
         if (!this.getItem().isEmpty()) {
@@ -309,6 +_,7 @@
         this.health = input.getShortOr("Health", (short)5);
         this.age = input.getShortOr("Age", (short)0);
         this.pickupDelay = input.getShortOr("PickupDelay", (short)0);
+        this.lifespan = input.getIntOr("Lifespan", ItemEntity.LIFETIME);
         this.target = input.read("Owner", UUIDUtil.CODEC).orElse(null);
         this.thrower = EntityReference.read(input, "Thrower");
         this.setItem(input.read("Item", ItemStack.CODEC).orElse(ItemStack.EMPTY));
@@ -323,7 +_,22 @@
             ItemStack itemStack = this.getItem();
             Item item = itemStack.getItem();
             int orgCount = itemStack.getCount();
-            if (this.pickupDelay == 0 && (this.target == null || this.target.equals(player.getUUID())) && player.getInventory().add(itemStack)) {
+
+            // Neo: Fire item pickup pre/post and adjust handling logic to adhere to the event result.
+            var result = net.neoforged.neoforge.event.EventHooks.fireItemPickupPre(this, player).canPickup();
+            if (result.isFalse()) {
+                return;
+            }
+
+            // Make a copy of the original stack for use in ItemEntityPickupEvent.Post
+            ItemStack originalCopy = itemStack.copy();
+            // Subvert the vanilla conditions (pickup delay and target check) if the result is true.
+            if ((result.isTrue() || this.pickupDelay == 0 && (this.target == null || this.target.equals(player.getUUID()))) && player.getInventory().add(itemStack)) {
+                // Fire ItemEntityPickupEvent.Post
+                net.neoforged.neoforge.event.EventHooks.fireItemPickupPost(this, player, originalCopy);
+                // Update `orgCount` to reflect the actual pickup amount. Vanilla is wrong here and always reports the whole stack.
+                orgCount = originalCopy.getCount() - itemStack.getCount();
+
                 player.take(this, orgCount);
                 if (itemStack.isEmpty()) {
                     this.discard();
@@ -369,6 +_,15 @@
         this.target = target;
     }
 
+    /**
+     * Neo: Add getter for ItemEntity's {@link ItemEntity#target target}.
+     * @return The {@link ItemEntity#target target} that can pick up this item entity (if {@code null}, anyone can pick it up)
+     */
+    @Nullable
+    public UUID getTarget() {
+        return this.target;
+    }
+
     public void setThrower(Entity thrower) {
         this.thrower = EntityReference.of(thrower);
     }
@@ -407,7 +_,7 @@
 
     public void makeFakeItem() {
         this.setNeverPickUp();
-        this.age = 5999;
+        this.age = getItem().getEntityLifespan(this.level()) - 1;
     }
 
     public static float getSpin(float ageInTicks, float bobOffset) {
