--- a/net/minecraft/network/protocol/game/ServerGamePacketListener.java
+++ b/net/minecraft/network/protocol/game/ServerGamePacketListener.java
@@ -4,7 +_,7 @@
 import net.minecraft.network.protocol.common.ServerCommonPacketListener;
 import net.minecraft.network.protocol.ping.ServerPingPacketListener;
 
-public interface ServerGamePacketListener extends ServerPingPacketListener, ServerCommonPacketListener {
+public interface ServerGamePacketListener extends ServerPingPacketListener, ServerCommonPacketListener, net.neoforged.neoforge.common.extensions.IServerGamePacketListenerExtension {
     @Override
     default ConnectionProtocol protocol() {
         return ConnectionProtocol.PLAY;
