--- a/net/minecraft/commands/CommandSourceStack.java
+++ b/net/minecraft/commands/CommandSourceStack.java
@@ -48,6 +_,12 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import com.mojang.brigadier.tree.CommandNode;
+import net.minecraft.server.permissions.Permission;
+import net.minecraft.server.permissions.Permissions;
+// CraftBukkit end
+
 import net.neoforged.neoforge.common.extensions.ICommandSourceStackExtension;
 
 public class CommandSourceStack implements SharedSuggestionProvider, ExecutionCommandSource<CommandSourceStack>, ICommandSourceStackExtension {
@@ -67,6 +_,10 @@
     private final Vec2 rotation;
     private final CommandSigningContext signingContext;
     private final TaskChainer chatMessageChainer;
+    // CraftBukkit start
+    private final PermissionSet bukkitPermissions;
+    public volatile CommandNode currentCommand;
+    // CraftBukkit end
 
     public CommandSourceStack(
         CommandSource source,
@@ -127,7 +_,40 @@
         this.rotation = rotation;
         this.signingContext = signingContext;
         this.chatMessageChainer = chatMessageChainer;
-    }
+        // CraftBukkit start
+        this.bukkitPermissions = new DelegatePermissionSet(permissions);
+    }
+
+    private class DelegatePermissionSet implements PermissionSet {
+
+        private final PermissionSet handle;
+
+        public DelegatePermissionSet(PermissionSet handle) {
+            this.handle = handle;
+        }
+
+        @Override
+        public boolean hasPermission(Permission permission) {
+            boolean hasPermission = handle.hasPermission(permission);
+
+            CommandNode currentCommand = CommandSourceStack.this.currentCommand;
+            if (currentCommand != null) {
+                return hasPermission(hasPermission, org.bukkit.craftbukkit.command.VanillaCommandWrapper.getPermission(currentCommand));
+            }
+
+            if (permission.equals(Permissions.COMMANDS_ENTITY_SELECTORS)) {
+                return hasPermission(hasPermission, "minecraft.command.selector");
+            }
+
+            return hasPermission;
+        }
+
+        public boolean hasPermission(boolean hasPermission, String bukkitPermission) {
+            // World is null when loading functions
+            return ((getLevel() == null || !getLevel().getCraftServer().ignoreVanillaPermissions) && hasPermission) || getBukkitSender().hasPermission(bukkitPermission);
+        }
+    }
+    // CraftBukkit end
 
     public CommandSourceStack withSource(CommandSource source) {
         return this.source == source
@@ -377,7 +_,7 @@
 
     @Override
     public PermissionSet permissions() {
-        return this.permissions;
+        return this.bukkitPermissions; // CraftBukkit
     }
 
     public Vec3 getPosition() {
@@ -582,4 +_,10 @@
     public boolean isSilent() {
         return this.silent;
     }
+
+    // CraftBukkit start
+    public org.bukkit.command.CommandSender getBukkitSender() {
+        return source.getBukkitSender(this);
+    }
+    // CraftBukkit end
 }
