--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -2,6 +_,10 @@
 
 import net.minecraft.network.chat.Component;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.command.NullCommandSender;
+// CraftBukkit end
+
 import net.neoforged.neoforge.common.extensions.ICommandSourceExtension;
 
 public interface CommandSource extends ICommandSourceExtension {
@@ -24,6 +_,13 @@
         public boolean shouldInformAdmins() {
             return false;
         }
+
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return NullCommandSender.INSTANCE;
+        }
+        // CraftBukkit end
     };
 
     void sendSystemMessage(Component message);
@@ -37,4 +_,6 @@
     default boolean alwaysAccepts() {
         return false;
     }
+
+    org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper); // CraftBukkit
 }
