--- a/net/minecraft/client/model/geom/LayerDefinitions.java
+++ b/net/minecraft/client/model/geom/LayerDefinitions.java
@@ -561,6 +_,7 @@
         result.put(ModelLayers.MANGROVE_CHEST_BOAT, chestBoatModel);
         result.put(ModelLayers.POPLAR_BOAT, boatModel);
         result.put(ModelLayers.POPLAR_CHEST_BOAT, chestBoatModel);
+        net.neoforged.neoforge.client.ClientHooks.loadLayerDefinitions(result);
         ImmutableMap<ModelLayerLocation, LayerDefinition> definitions = result.build();
         List<ModelLayerLocation> missingDefinitions = ModelLayers.getKnownLocations().filter(l -> !definitions.containsKey(l)).toList();
         if (!missingDefinitions.isEmpty()) {
