--- a/net/minecraft/client/model/geom/ModelLayers.java
+++ b/net/minecraft/client/model/geom/ModelLayers.java
@@ -327,15 +_,18 @@
     }
 
     public static ModelLayerLocation createStandingSignModelName(WoodType type) {
-        return createLocation("sign/standing/" + type.name(), "main");
+        Identifier location = Identifier.parse(type.name());
+        return new ModelLayerLocation(location.withPrefix("sign/standing/"), "main");
     }
 
     public static ModelLayerLocation createWallSignModelName(WoodType type) {
-        return createLocation("sign/wall/" + type.name(), "main");
+        Identifier location = Identifier.parse(type.name());
+        return new ModelLayerLocation(location.withPrefix("sign/wall/"), "main");
     }
 
     public static ModelLayerLocation createHangingSignModelName(WoodType type, HangingSignRenderer.AttachmentType attachmentType) {
-        return createLocation("hanging_sign/" + type.name() + "/" + attachmentType.getSerializedName(), "main");
+        Identifier location = Identifier.parse(type.name());
+        return new ModelLayerLocation(location.withPrefix("hanging_sign/").withSuffix("/" + attachmentType.getSerializedName()), "main");
     }
 
     public static Stream<ModelLayerLocation> getKnownLocations() {
