--- a/net/minecraft/world/level/saveddata/SavedData.java
+++ b/net/minecraft/world/level/saveddata/SavedData.java
@@ -45,6 +_,10 @@
         }
     }
 
-    public static record Factory<T extends SavedData>(Supplier<T> constructor, Function<CompoundTag, T> deserializer, DataFixTypes type) {
+    public static record Factory<T extends SavedData>(Supplier<T> constructor, Function<CompoundTag, T> deserializer, @javax.annotation.Nullable DataFixTypes type) { // Neo: We do not have update logic compatible with DFU, several downstream patches from this record are made to support a nullable type.
+
+        public Factory(Supplier<T> constructor, Function<CompoundTag, T> deserializer) {
+            this(constructor, deserializer, null);
+        }
     }
 }
