--- a/net/minecraft/data/loot/LootTableProvider.java
+++ b/net/minecraft/data/loot/LootTableProvider.java
@@ -62,33 +_,24 @@
     private CompletableFuture<?> run(CachedOutput p_324447_, HolderLookup.Provider p_323978_) {
         WritableRegistry<LootTable> writableregistry = new MappedRegistry<>(Registries.LOOT_TABLE, Lifecycle.experimental());
         Map<RandomSupport.Seed128bit, ResourceLocation> map = new Object2ObjectOpenHashMap<>();
-        this.subProviders.forEach(p_329847_ -> p_329847_.provider().get().generate(p_323978_, (p_339366_, p_339367_) -> {
-                ResourceLocation resourcelocation = sequenceIdForLootTable(p_339366_);
+        getTables().forEach(p_329847_ -> p_329847_.provider().get().generate(p_323978_, (p_335199_, p_335200_) -> {
+                ResourceLocation resourcelocation = sequenceIdForLootTable(p_335199_);
                 ResourceLocation resourcelocation1 = map.put(RandomSequence.seedForKey(resourcelocation), resourcelocation);
                 if (resourcelocation1 != null) {
-                    Util.logAndPauseIfInIde("Loot table random sequence seed collision on " + resourcelocation1 + " and " + p_339366_.location());
+                    Util.logAndPauseIfInIde("Loot table random sequence seed collision on " + resourcelocation1 + " and " + p_335199_.location());
                 }
 
-                p_339367_.setRandomSequence(resourcelocation);
-                LootTable loottable = p_339367_.setParamSet(p_329847_.paramSet).build();
-                writableregistry.register(p_339366_, loottable, RegistrationInfo.BUILT_IN);
+                p_335200_.setRandomSequence(resourcelocation);
+                LootTable loottable = p_335200_.setParamSet(p_329847_.paramSet).build();
+                writableregistry.register(p_335199_, loottable, RegistrationInfo.BUILT_IN);
             }));
         writableregistry.freeze();
         ProblemReporter.Collector problemreporter$collector = new ProblemReporter.Collector();
         HolderGetter.Provider holdergetter$provider = new RegistryAccess.ImmutableRegistryAccess(List.of(writableregistry)).freeze().asGetterLookup();
         ValidationContext validationcontext = new ValidationContext(problemreporter$collector, LootContextParamSets.ALL_PARAMS, holdergetter$provider);
 
-        for (ResourceKey<LootTable> resourcekey : Sets.difference(this.requiredTables, writableregistry.registryKeySet())) {
-            problemreporter$collector.report("Missing built-in table: " + resourcekey.location());
-        }
+        validate(writableregistry, validationcontext, problemreporter$collector);
 
-        writableregistry.holders()
-            .forEach(
-                p_339369_ -> p_339369_.value()
-                        .validate(
-                            validationcontext.setParams(p_339369_.value().getParamSet()).enterElement("{" + p_339369_.key().location() + "}", p_339369_.key())
-                        )
-            );
         Multimap<String, String> multimap = problemreporter$collector.get();
         if (!multimap.isEmpty()) {
             multimap.forEach((p_124446_, p_124447_) -> LOGGER.warn("Found validation problem in {}: {}", p_124446_, p_124447_));
@@ -101,6 +_,20 @@
                 return DataProvider.saveStable(p_324447_, p_323978_, LootTable.DIRECT_CODEC, loottable, path);
             }).toArray(CompletableFuture[]::new));
         }
+    }
+
+    public List<LootTableProvider.SubProviderEntry> getTables() {
+        return this.subProviders;
+    }
+
+    protected void validate(WritableRegistry<LootTable> writableregistry, ValidationContext validationcontext, ProblemReporter.Collector problemreporter$collector) {
+        for(ResourceKey<LootTable> resourcekey : Sets.difference(this.requiredTables, writableregistry.registryKeySet())) {
+            problemreporter$collector.report("Missing built-in table: " + resourcekey.location());
+        }
+
+        writableregistry.holders().forEach(p_335195_ -> {
+            p_335195_.value().validate(validationcontext.setParams(p_335195_.value().getParamSet()).enterElement("{" + p_335195_.key().location() + "}", p_335195_.key()));
+        });
     }
 
     private static ResourceLocation sequenceIdForLootTable(ResourceKey<LootTable> p_336172_) {
