--- a/net/minecraft/data/loot/LootTableProvider.java
+++ b/net/minecraft/data/loot/LootTableProvider.java
@@ -61,7 +_,7 @@
     private CompletableFuture<?> run(CachedOutput p_324447_, HolderLookup.Provider p_323978_) {
         WritableRegistry<LootTable> writableregistry = new MappedRegistry<>(Registries.LOOT_TABLE, Lifecycle.experimental());
         Map<RandomSupport.Seed128bit, ResourceLocation> map = new Object2ObjectOpenHashMap<>();
-        this.subProviders.forEach(p_344197_ -> p_344197_.provider().apply(p_323978_).generate((p_380827_, p_380828_) -> {
+        getTables().forEach(p_344197_ -> p_344197_.provider().apply(p_323978_).generate((p_380827_, p_380828_) -> {
             ResourceLocation resourcelocation = sequenceIdForLootTable(p_380827_);
             ResourceLocation resourcelocation1 = map.put(RandomSequence.seedForKey(resourcelocation), resourcelocation);
             if (resourcelocation1 != null) {
@@ -77,18 +_,8 @@
         HolderGetter.Provider holdergetter$provider = new RegistryAccess.ImmutableRegistryAccess(List.of(writableregistry)).freeze();
         ValidationContext validationcontext = new ValidationContext(problemreporter$collector, LootContextParamSets.ALL_PARAMS, holdergetter$provider);
 
-        for (ResourceKey<LootTable> resourcekey : Sets.difference(this.requiredTables, writableregistry.registryKeySet())) {
-            problemreporter$collector.report(new LootTableProvider.MissingTableProblem(resourcekey));
-        }
+        validate(writableregistry, validationcontext, problemreporter$collector);
 
-        writableregistry.listElements()
-            .forEach(
-                p_421302_ -> p_421302_.value()
-                    .validate(
-                        validationcontext.setContextKeySet(p_421302_.value().getParamSet())
-                            .enterElement(new ProblemReporter.RootElementPathElement(p_421302_.key()), p_421302_.key())
-                    )
-            );
         if (!problemreporter$collector.isEmpty()) {
             problemreporter$collector.forEach((p_421299_, p_421300_) -> LOGGER.warn("Found validation problem in {}: {}", p_421299_, p_421300_.description()));
             throw new IllegalStateException("Failed to validate loot tables, see logs");
@@ -100,6 +_,25 @@
                 return DataProvider.saveStable(p_324447_, p_323978_, LootTable.DIRECT_CODEC, loottable, path);
             }).toArray(CompletableFuture[]::new));
         }
+    }
+
+    public List<LootTableProvider.SubProviderEntry> getTables() {
+        return this.subProviders;
+    }
+
+    protected void validate(WritableRegistry<LootTable> writableregistry, ValidationContext validationcontext, ProblemReporter.Collector problemreporter$collector) {
+        for (ResourceKey<LootTable> resourcekey : Sets.difference(this.requiredTables, writableregistry.registryKeySet())) {
+            problemreporter$collector.report(new LootTableProvider.MissingTableProblem(resourcekey));
+        }
+
+        writableregistry.listElements()
+                .forEach(
+                        p_380823_ -> p_380823_.value()
+                                .validate(
+                                        validationcontext.setContextKeySet(p_380823_.value().getParamSet())
+                                                .enterElement(new ProblemReporter.RootElementPathElement(p_380823_.key()), p_380823_.key())
+                                )
+                );
     }
 
     private static ResourceLocation sequenceIdForLootTable(ResourceKey<LootTable> p_336172_) {
