--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -51,7 +_,11 @@
 
                 for (StructurePiece piece : start.getPieces()) {
                     if (piece.isCloseToChunk(chunkPos, 12)) {
-                        if (piece instanceof PoolElementStructurePiece poolPiece) {
+                        if (piece instanceof net.neoforged.neoforge.common.world.PieceBeardifierModifier pieceBeardifierModifier) {
+                            if (pieceBeardifierModifier.getTerrainAdjustment() != TerrainAdjustment.NONE) {
+                                rigids.add(new Beardifier.Rigid(pieceBeardifierModifier.getBeardifierBox(), pieceBeardifierModifier.getTerrainAdjustment(), pieceBeardifierModifier.getGroundLevelDelta()));
+                            }
+                        } else if (piece instanceof PoolElementStructurePiece poolPiece) {
                             StructureTemplatePool.Projection projection = poolPiece.getElement().getProjection();
                             if (projection == StructureTemplatePool.Projection.RIGID) {
                                 rigids.add(new Beardifier.Rigid(poolPiece.getBoundingBox(), terrainAdjustment, poolPiece.getGroundLevelDelta()));
