--- a/net/minecraft/world/level/levelgen/FlatLevelSource.java
+++ b/net/minecraft/world/level/levelgen/FlatLevelSource.java
@@ -30,9 +_,11 @@
         i -> i.group(FlatLevelGeneratorSettings.CODEC.fieldOf("settings").forGetter(FlatLevelSource::settings)).apply(i, i.stable(FlatLevelSource::new))
     );
     private final FlatLevelGeneratorSettings settings;
+    // Neo: Store the memoized generation settings getter, so the cache can be cleared later
+    private final net.neoforged.neoforge.common.util.MemoizedFunction<Holder<net.minecraft.world.level.biome.Biome>, net.minecraft.world.level.biome.BiomeGenerationSettings> generationSettingsGetter;
 
     public FlatLevelSource(FlatLevelGeneratorSettings generatorSettings) {
-        super(new FixedBiomeSource(generatorSettings.getBiome()), Util.memoize(generatorSettings::adjustGenerationSettings));
+        super(new FixedBiomeSource(generatorSettings.getBiome()), this.generationSettingsGetter = net.neoforged.neoforge.common.util.MemoizedFunction.of(generatorSettings::adjustGenerationSettings));
         this.settings = generatorSettings;
     }
 
@@ -142,5 +_,12 @@
     @Override
     public int getSeaLevel() {
         return -63;
+    }
+
+    // Neo: clear the cache for the generation settings getter on refresh
+    @Override
+    public void refreshFeaturesPerStep() {
+        super.refreshFeaturesPerStep();
+        this.generationSettingsGetter.clear();
     }
 }
