--- a/net/minecraft/client/data/models/model/ModelLocationUtils.java
+++ b/net/minecraft/client/data/models/model/ModelLocationUtils.java
@@ -11,11 +_,13 @@
 public class ModelLocationUtils {
     @Deprecated
     public static Identifier decorateBlockModelLocation(String id) {
-        return Identifier.withDefaultNamespace("block/" + id);
+        // Neo: Use Identifier.parse to support modded paths
+        return Identifier.parse(id).withPrefix("block/");
     }
 
     public static Identifier decorateItemModelLocation(String id) {
-        return Identifier.withDefaultNamespace("item/" + id);
+        // Neo: Use Identifier.parse to support modded paths
+        return Identifier.parse(id).withPrefix("item/");
     }
 
     public static Identifier getModelLocation(Block block, String suffix) {
