--- a/net/minecraft/client/data/models/model/ModelTemplates.java
+++ b/net/minecraft/client/data/models/model/ModelTemplates.java
@@ -249,18 +_,18 @@
     }
 
     public static ModelTemplate create(String id, TextureSlot... slots) {
-        return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("block/" + id)), Optional.empty(), slots);
+        return new ModelTemplate(Optional.of(ModelLocationUtils.decorateBlockModelLocation(id)), Optional.empty(), slots);
     }
 
     public static ModelTemplate createItem(String id, TextureSlot... slots) {
-        return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("item/" + id)), Optional.empty(), slots);
+        return new ModelTemplate(Optional.of(ModelLocationUtils.decorateItemModelLocation(id)), Optional.empty(), slots);
     }
 
     public static ModelTemplate createItem(String id, String suffix, TextureSlot... slots) {
-        return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("item/" + id)), Optional.of(suffix), slots);
+        return new ModelTemplate(Optional.of(ModelLocationUtils.decorateItemModelLocation(id)), Optional.of(suffix), slots);
     }
 
     public static ModelTemplate create(String id, String suffix, TextureSlot... slots) {
-        return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("block/" + id)), Optional.of(suffix), slots);
+        return new ModelTemplate(Optional.of(ModelLocationUtils.decorateBlockModelLocation(id)), Optional.of(suffix), slots);
     }
 }
