--- a/net/minecraft/client/data/models/model/TextureMapping.java
+++ b/net/minecraft/client/data/models/model/TextureMapping.java
@@ -453,4 +_,12 @@
         Identifier id = BuiltInRegistries.ITEM.getKey(item);
         return new Material(id.withPath(path -> "item/" + path + suffix));
     }
+
+    // Neo: Added to allow easier texture map copying
+    public TextureMapping copy() {
+        TextureMapping texturemapping = new TextureMapping();
+        texturemapping.slots.putAll(this.slots);
+        texturemapping.forcedSlots.addAll(this.forcedSlots);
+        return texturemapping;
+    }
 }
