--- a/com/mojang/renderpearl/api/commands/RenderPass.java
+++ b/com/mojang/renderpearl/api/commands/RenderPass.java
@@ -39,6 +_,14 @@
 
     void pushConstants(ByteBuffer value);
 
+    /// Neo: set the viewport to render into.
+    ///
+    /// @param x      The X position of the viewport
+    /// @param y      The Y position of the viewport
+    /// @param width  The width of the viewport
+    /// @param height The height of the viewport
+    void setViewport(int x, int y, int width, int height);
+
     void enableScissor(int x, int y, int width, int height);
 
     void disableScissor();
