--- a/net/minecraft/world/level/chunk/status/ChunkStatus.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatus.java
@@ -53,6 +_,13 @@
 
     @VisibleForTesting
     protected ChunkStatus(@Nullable ChunkStatus p_330316_, EnumSet<Heightmap.Types> p_331442_, ChunkType p_331412_) {
+
+        EnumSet<Heightmap.Types> chunkSaveHeightmaps = EnumSet.copyOf(this.heightmapsAfter);
+        if (this.chunkType != ChunkType.LEVELCHUNK) {
+            chunkSaveHeightmaps.add(Heightmap.Types.WORLD_SURFACE_WG);
+            chunkSaveHeightmaps.add(Heightmap.Types.OCEAN_FLOOR_WG);
+        }
+        this.chunkSaveHeightmaps = chunkSaveHeightmaps;
         this.parent = p_330316_ == null ? this : p_330316_;
         this.chunkType = p_331412_;
         this.heightmapsAfter = p_331442_;
@@ -107,4 +_,16 @@
     public String getName() {
         return BuiltInRegistries.CHUNK_STATUS.getKey(this).toString();
     }
+
+    /**
+     * Neo: Internal use only.
+     * Patch to fix [MC-308222](https://report.bugs.mojang.com/servicedesk/customer/portal/2/MC-308222)
+     * intended for loading a non-fully generated chunk from disk to keep worldgen heightmap data for
+     * structures and features to heightmap snap properly again.
+     * @return A set of the chunk status's heightmaps alongside _WG heightmaps for worldgen-only chunk statuses
+     */
+    public EnumSet<Heightmap.Types> getChunkSaveHeightmaps() {
+        return this.chunkSaveHeightmaps;
+    }
+    private final EnumSet<Heightmap.Types> chunkSaveHeightmaps;
 }
