--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -195,11 +_,22 @@
             }
 
             levelChunk.setFullStatus(holder::getFullStatus);
+            try {
+                holder.currentlyLoading = levelChunk; // Neo: bypass the future chain when getChunk is called, this prevents deadlocks.
             levelChunk.runPostLoad();
+            } finally {
+                holder.currentlyLoading = null; // Neo: Stop bypassing the future chain.
+            }
             levelChunk.setLoaded(true);
+            try {
+                holder.currentlyLoading = levelChunk; // Neo: bypass the future chain when getChunk is called, this prevents deadlocks.
             levelChunk.registerAllBlockEntitiesAfterLevelLoad();
             levelChunk.registerTickContainerInLevel(level);
             levelChunk.setUnsavedListener(context.unsavedListener());
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkEvent.Load(levelChunk, !(protoChunk instanceof ImposterProtoChunk)));
+            } finally {
+                holder.currentlyLoading = null; // Neo: Stop bypassing the future chain.
+            }
             return levelChunk;
         }, context.mainThreadExecutor());
     }
