--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -186,6 +_,7 @@
         }
 
         if (id == 0) {
+            if (net.neoforged.neoforge.event.EventHooks.onPistonMovePre(level, pos, direction, true)) return false;
             if (!this.moveBlocks(level, pos, direction, true)) {
                 return false;
             }
@@ -194,6 +_,7 @@
             level.playSound(null, pos, SoundEvents.PISTON_EXTEND, SoundSource.BLOCKS, 0.5F, level.random.nextFloat() * 0.25F + 0.6F);
             level.gameEvent(GameEvent.BLOCK_ACTIVATE, pos, GameEvent.Context.of(blockstate));
         } else if (id == 1 || id == 2) {
+            if (net.neoforged.neoforge.event.EventHooks.onPistonMovePre(level, pos, direction, false)) return false;
             BlockEntity blockentity = level.getBlockEntity(pos.relative(direction));
             if (blockentity instanceof PistonMovingBlockEntity) {
                 ((PistonMovingBlockEntity)blockentity).finalTick();
@@ -243,6 +_,7 @@
             level.gameEvent(GameEvent.BLOCK_DEACTIVATE, pos, GameEvent.Context.of(blockstate1));
         }
 
+        net.neoforged.neoforge.event.EventHooks.onPistonMovePost(level, pos, direction, (id == 0));
         return true;
     }
 
@@ -314,8 +_,7 @@
                 BlockState blockstate1 = level.getBlockState(blockpos2);
                 BlockEntity blockentity = blockstate1.hasBlockEntity() ? level.getBlockEntity(blockpos2) : null;
                 dropResources(blockstate1, level, blockpos2, blockentity);
-                level.setBlock(blockpos2, Blocks.AIR.defaultBlockState(), 18);
-                level.gameEvent(GameEvent.BLOCK_DESTROY, blockpos2, GameEvent.Context.of(blockstate1));
+                blockstate1.onDestroyedByPushReaction(level, blockpos2, direction, level.getFluidState(blockpos2));
                 if (!blockstate1.is(BlockTags.FIRE)) {
                     level.addDestroyBlockEffect(blockpos2, blockstate1);
                 }
@@ -387,6 +_,10 @@
     @Override
     protected BlockState rotate(BlockState state, Rotation rot) {
         return state.setValue(FACING, rot.rotate(state.getValue(FACING)));
+    }
+
+    public BlockState rotate(BlockState state, net.minecraft.world.level.LevelAccessor world, BlockPos pos, Rotation direction) {
+         return state.getValue(EXTENDED) ? state : super.rotate(state, world, pos, direction);
     }
 
     @Override
