--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -187,6 +_,7 @@
         }
 
         if (p_60195_ == 0) {
+            if (net.neoforged.neoforge.event.EventHooks.onPistonMovePre(p_60193_, p_60194_, direction, true)) return false;
             if (!this.moveBlocks(p_60193_, p_60194_, direction, true)) {
                 return false;
             }
@@ -195,6 +_,7 @@
             p_60193_.playSound(null, p_60194_, SoundEvents.PISTON_EXTEND, SoundSource.BLOCKS, 0.5F, p_60193_.random.nextFloat() * 0.25F + 0.6F);
             p_60193_.gameEvent(GameEvent.BLOCK_ACTIVATE, p_60194_, GameEvent.Context.of(blockstate));
         } else if (p_60195_ == 1 || p_60195_ == 2) {
+            if (net.neoforged.neoforge.event.EventHooks.onPistonMovePre(p_60193_, p_60194_, direction, false)) return false;
             BlockEntity blockentity1 = p_60193_.getBlockEntity(p_60194_.relative(direction));
             if (blockentity1 instanceof PistonMovingBlockEntity) {
                 ((PistonMovingBlockEntity)blockentity1).finalTick();
@@ -246,6 +_,7 @@
             p_60193_.gameEvent(GameEvent.BLOCK_DEACTIVATE, p_60194_, GameEvent.Context.of(blockstate1));
         }
 
+        net.neoforged.neoforge.event.EventHooks.onPistonMovePost(p_60193_, p_60194_, direction, (p_60195_ == 0));
         return true;
     }
 
@@ -317,8 +_,7 @@
                 BlockState blockstate1 = p_60182_.getBlockState(blockpos2);
                 BlockEntity blockentity = blockstate1.hasBlockEntity() ? p_60182_.getBlockEntity(blockpos2) : null;
                 dropResources(blockstate1, p_60182_, blockpos2, blockentity);
-                p_60182_.setBlock(blockpos2, Blocks.AIR.defaultBlockState(), 18);
-                p_60182_.gameEvent(GameEvent.BLOCK_DESTROY, blockpos2, GameEvent.Context.of(blockstate1));
+                blockstate1.onDestroyedByPushReaction(blockstate1, p_60182_, blockpos2, direction, p_60182_.getFluidState(blockpos2));
                 if (!blockstate1.is(BlockTags.FIRE)) {
                     p_60182_.addDestroyBlockEffect(blockpos2, blockstate1);
                 }
@@ -390,6 +_,10 @@
     @Override
     public BlockState rotate(BlockState p_60215_, Rotation p_60216_) {
         return p_60215_.setValue(FACING, p_60216_.rotate(p_60215_.getValue(FACING)));
+    }
+
+    public BlockState rotate(BlockState state, net.minecraft.world.level.LevelAccessor world, BlockPos pos, Rotation direction) {
+         return state.getValue(EXTENDED) ? state : super.rotate(state, world, pos, direction);
     }
 
     @Override
