--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -116,6 +_,8 @@
     private final Set<TagKey<Item>> visibleTags = new HashSet<>();
     private final boolean displayOperatorCreativeTab;
     private final EffectsInInventory effects;
+    private final List<net.neoforged.neoforge.client.gui.CreativeTabsScreenPage> pages = new java.util.ArrayList<>();
+    private net.neoforged.neoforge.client.gui.CreativeTabsScreenPage currentPage = new net.neoforged.neoforge.client.gui.CreativeTabsScreenPage(new java.util.ArrayList<>());
 
     public CreativeModeInventoryScreen(LocalPlayer player, FeatureFlagSet enabledFeatures, boolean displayOperatorCreativeTab) {
         super(new CreativeModeInventoryScreen.ItemPickerMenu(player), player.getInventory(), CommonComponents.EMPTY, 195, 136);
@@ -152,9 +_,11 @@
             return false;
         } else {
             if (searchTrees != null) {
-                List<ItemStack> creativeSearchItems = List.copyOf(CreativeModeTabs.searchTab().getDisplayItems());
-                searchTrees.updateCreativeTooltips(holders, creativeSearchItems);
-                searchTrees.updateCreativeTags(creativeSearchItems);
+                CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(tab -> {
+                    List<ItemStack> list = List.copyOf(tab.getDisplayItems());
+                    searchTrees.updateCreativeTooltips(holders, list, net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getNameSearchKey(tab));
+                    searchTrees.updateCreativeTags(list, net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getTagSearchKey(tab));
+                });
             }
 
             return true;
@@ -164,7 +_,7 @@
     private void refreshCurrentTabContents(Collection<ItemStack> displayList) {
         int oldRowIndex = this.menu.getRowIndexForScroll(this.scrollOffs);
         this.menu.items.clear();
-        if (selectedTab.getType() == CreativeModeTab.Type.SEARCH) {
+        if (selectedTab.hasSearchBar()) {
             this.refreshSearchResults();
         } else {
             this.menu.items.addAll(displayList);
@@ -330,6 +_,34 @@
     protected void init() {
         if (this.minecraft.player.hasInfiniteMaterials()) {
             super.init();
+            this.pages.clear();
+            int tabIndex = 0;
+            List<CreativeModeTab> currentPage = new java.util.ArrayList<>();
+            for (CreativeModeTab sortedCreativeModeTab : net.neoforged.neoforge.common.CreativeModeTabRegistry.getSortedCreativeModeTabs().stream().filter(CreativeModeTab::hasAnyItems).toList()) {
+                currentPage.add(sortedCreativeModeTab);
+                tabIndex++;
+                if (tabIndex == 10) {
+                    this.pages.add(new net.neoforged.neoforge.client.gui.CreativeTabsScreenPage(currentPage));
+                    currentPage = new java.util.ArrayList<>();
+                    tabIndex = 0;
+                }
+            }
+            if (tabIndex != 0) {
+                this.pages.add(new net.neoforged.neoforge.client.gui.CreativeTabsScreenPage(currentPage));
+            }
+            if (this.pages.isEmpty()) {
+                this.currentPage = new net.neoforged.neoforge.client.gui.CreativeTabsScreenPage(new java.util.ArrayList<>());
+            } else {
+                this.currentPage = this.pages.get(0);
+            }
+            if (this.pages.size() > 1) {
+                addRenderableWidget(net.minecraft.client.gui.components.Button.builder(Component.literal("<"), b -> setCurrentPage(this.pages.get(Math.max(this.pages.indexOf(this.currentPage) - 1, 0)))).pos(leftPos,  topPos - 50).size(20, 20).build());
+                addRenderableWidget(net.minecraft.client.gui.components.Button.builder(Component.literal(">"), b -> setCurrentPage(this.pages.get(Math.min(this.pages.indexOf(this.currentPage) + 1, this.pages.size() - 1)))).pos(leftPos + imageWidth - 20, topPos - 50).size(20, 20).build());
+            }
+            this.currentPage = this.pages.stream().filter(page -> page.getVisibleTabs().contains(selectedTab)).findFirst().orElse(this.currentPage);
+            if (!this.currentPage.getVisibleTabs().contains(selectedTab)) {
+                selectedTab = this.currentPage.getVisibleTabs().get(0);
+            }
             this.searchBox = new EditBox(this.font, this.leftPos + 82, this.topPos + 6, 80, 9, Component.translatable("itemGroup.search"));
             this.searchBox.setMaxLength(50);
             this.searchBox.setBordered(false);
@@ -377,7 +_,7 @@
     public boolean charTyped(CharacterEvent event) {
         if (this.ignoreTextInput) {
             return false;
-        } else if (selectedTab.getType() != CreativeModeTab.Type.SEARCH) {
+        } else if (!selectedTab.hasSearchBar()) {
             return false;
         } else {
             String oldContents = this.searchBox.getValue();
@@ -405,7 +_,7 @@
     @Override
     public boolean keyPressed(KeyEvent event) {
         this.ignoreTextInput = false;
-        if (selectedTab.getType() != CreativeModeTab.Type.SEARCH) {
+        if (!selectedTab.hasSearchBar()) {
             if (this.minecraft.options.keyChat.matches(event)) {
                 this.ignoreTextInput = true;
                 this.selectTab(CreativeModeTabs.searchTab());
@@ -441,6 +_,7 @@
     }
 
     private void refreshSearchResults() {
+        if (!selectedTab.hasSearchBar()) return;
         this.menu.items.clear();
         this.visibleTags.clear();
         String searchTerm = this.searchBox.getValue();
@@ -453,10 +_,10 @@
                 SearchTree<ItemStack> tree;
                 if (searchTerm.startsWith("#")) {
                     searchTerm = searchTerm.substring(1);
-                    tree = searchTrees.creativeTagSearch();
+                    tree = searchTrees.creativeTagSearch(net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getTagSearchKey(selectedTab));
                     this.updateVisibleTags(searchTerm);
                 } else {
-                    tree = searchTrees.creativeNameSearch();
+                    tree = searchTrees.creativeNameSearch(net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getNameSearchKey(selectedTab));
                 }
 
                 this.menu.items.addAll(tree.search(searchTerm.toLowerCase(Locale.ROOT)));
@@ -484,7 +_,7 @@
     @Override
     protected void extractLabels(GuiGraphicsExtractor graphics, int xm, int ym) {
         if (selectedTab.showTitle()) {
-            graphics.text(this.font, selectedTab.getDisplayName(), 8, 6, -12566464, false);
+            graphics.text(this.font, selectedTab.getDisplayName(), 8, 6, selectedTab.getLabelColor(), false);
         }
     }
 
@@ -494,7 +_,7 @@
             double xm = event.x() - this.leftPos;
             double ym = event.y() - this.topPos;
 
-            for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
+            for (CreativeModeTab tab : currentPage.getVisibleTabs()) {
                 if (this.checkTabClicked(tab, xm, ym)) {
                     return true;
                 }
@@ -516,7 +_,7 @@
             double ym = event.y() - this.topPos;
             this.scrolling = false;
 
-            for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
+            for (CreativeModeTab tab : currentPage.getVisibleTabs()) {
                 if (this.checkTabClicked(tab, xm, ym)) {
                     this.selectTab(tab);
                     return true;
@@ -534,6 +_,7 @@
     private void selectTab(CreativeModeTab tab) {
         CreativeModeTab oldTab = selectedTab;
         selectedTab = tab;
+        slotColor = tab.getSlotColor();
         this.quickCraftSlots.clear();
         this.menu.items.clear();
         this.clearDraggingState();
@@ -610,13 +_,15 @@
             this.originalSlots = null;
         }
 
-        if (selectedTab.getType() == CreativeModeTab.Type.SEARCH) {
+        if (selectedTab.hasSearchBar()) {
             this.searchBox.setVisible(true);
             this.searchBox.setCanLoseFocus(false);
             this.searchBox.setFocused(true);
             if (oldTab != tab) {
                 this.searchBox.setValue("");
             }
+            this.searchBox.setWidth(selectedTab.getSearchBarWidth());
+            this.searchBox.setX(this.leftPos + (82 /*default left*/ + 89 /*default width*/) - this.searchBox.getWidth());
 
             this.refreshSearchResults();
         } else {
@@ -679,8 +_,14 @@
         this.effects.extractRenderState(graphics, mouseX, mouseY);
         super.extractRenderState(graphics, mouseX, mouseY, a);
 
-        for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
-            if (this.checkTabHovering(graphics, tab, mouseX, mouseY)) {
+        if (this.pages.size() != 1) {
+            Component page = Component.literal(String.format("%d / %d", this.pages.indexOf(this.currentPage) + 1, this.pages.size()));
+            // TODO 1.21.6: This previously bumped Z by 300, investigate if this needs nextStratum
+            graphics.text(font, page.getVisualOrderText(), leftPos + (imageWidth / 2) - (font.width(page) / 2), topPos - 44, -1);
+        }
+
+        for (CreativeModeTab creativemodetab : currentPage.getVisibleTabs()) {
+            if (this.checkTabHovering(graphics, creativemodetab, mouseX, mouseY)) {
                 break;
             }
         }
@@ -701,10 +_,10 @@
     public List<Component> getTooltipFromContainerItem(ItemStack itemStack) {
         boolean isCreativeSlot = this.hoveredSlot != null && this.hoveredSlot instanceof CreativeModeInventoryScreen.CustomCreativeSlot;
         boolean isSingleCategoryTab = selectedTab.getType() == CreativeModeTab.Type.CATEGORY;
-        boolean isSearchTab = selectedTab.getType() == CreativeModeTab.Type.SEARCH;
+        boolean isSearchTab = selectedTab.hasSearchBar();
         TooltipFlag.Default originalTooltipStyle = this.minecraft.options.advancedItemTooltips ? TooltipFlag.Default.ADVANCED : TooltipFlag.Default.NORMAL;
         TooltipFlag tooltipStyle = isCreativeSlot ? originalTooltipStyle.asCreative() : originalTooltipStyle;
-        List<Component> originalLines = itemStack.getTooltipLines(Item.TooltipContext.of(this.minecraft.level), this.minecraft.player, tooltipStyle);
+        List<Component> originalLines = itemStack.getTooltipLines(Item.TooltipContext.of(this.minecraft.level, this.minecraft.player), this.minecraft.player, net.neoforged.neoforge.client.ClientTooltipFlag.of(tooltipStyle));
         if (originalLines.isEmpty()) {
             return originalLines;
         } else if (isSingleCategoryTab && isCreativeSlot) {
@@ -722,7 +_,7 @@
             int i = 1;
 
             for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
-                if (tab.getType() != CreativeModeTab.Type.SEARCH && tab.contains(itemStack)) {
+                if (!tab.hasSearchBar() && tab.contains(itemStack)) {
                     linesToDisplay.add(i++, tab.getDisplayName().copy().withStyle(ChatFormatting.BLUE));
                 }
             }
@@ -735,7 +_,7 @@
     public void extractBackground(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float a) {
         super.extractBackground(graphics, mouseX, mouseY, a);
 
-        for (CreativeModeTab tab : CreativeModeTabs.tabs()) {
+        for (CreativeModeTab tab : currentPage.getVisibleTabs()) {
             if (tab != selectedTab) {
                 this.extractTabButton(graphics, mouseX, mouseY, tab);
             }
@@ -766,10 +_,11 @@
         int yscr = this.topPos + 18;
         int yscr2 = yscr + 112;
         if (selectedTab.canScroll()) {
-            Identifier sprite = this.canScroll() ? SCROLLER_SPRITE : SCROLLER_DISABLED_SPRITE;
-            graphics.blitSprite(RenderPipelines.GUI_TEXTURED, sprite, xscr, yscr + (int)((yscr2 - yscr - 17) * this.scrollOffs), 12, 15);
+            Identifier identifier = selectedTab.getScrollerSprite(); // this.canScroll() ? SCROLLER_SPRITE : SCROLLER_DISABLED_SPRITE;
+            graphics.blitSprite(RenderPipelines.GUI_TEXTURED, identifier, xscr, yscr + (int)((yscr2 - yscr - 17) * this.scrollOffs), 12, 15);
         }
 
+        if (currentPage.getVisibleTabs().contains(selectedTab)) //Forge: only display tab selection when the selected tab is on the current page
         this.extractTabButton(graphics, mouseX, mouseY, selectedTab);
         if (selectedTab.getType() == CreativeModeTab.Type.INVENTORY) {
             InventoryScreen.extractEntityInInventoryFollowsMouse(
@@ -779,7 +_,7 @@
     }
 
     private int getTabX(CreativeModeTab tab) {
-        int pos = tab.column();
+        int pos = currentPage.getColumn(tab);
         int spacing = 27;
         int x = 27 * pos;
         if (tab.isAlignedRight()) {
@@ -791,7 +_,7 @@
 
     private int getTabY(CreativeModeTab tab) {
         int y = 0;
-        if (tab.row() == CreativeModeTab.Row.TOP) {
+        if (currentPage.isTop(tab)) {
             y -= 32;
         } else {
             y += this.imageHeight;
@@ -819,8 +_,8 @@
 
     protected void extractTabButton(GuiGraphicsExtractor graphics, int mouseX, int mouseY, CreativeModeTab tab) {
         boolean selected = tab == selectedTab;
-        boolean isTop = tab.row() == CreativeModeTab.Row.TOP;
-        int pos = tab.column();
+        boolean isTop = currentPage.isTop(tab);
+        int pos = currentPage.getColumn(tab);
         int x = this.leftPos + this.getTabX(tab);
         int y = this.topPos - (isTop ? 28 : -(this.imageHeight - 4));
         Identifier[] sprites;
@@ -834,6 +_,7 @@
             graphics.requestCursor(CursorTypes.POINTING_HAND);
         }
 
+        // TODO 1.20.2: Deal with custom tab backgrounds, and deal with transparency.
         graphics.blitSprite(RenderPipelines.GUI_TEXTURED, sprites[Mth.clamp(pos, 0, sprites.length)], x, y, 26, 32);
         int iconX = x + 13 - 8;
         int iconY = y + 16 - 8 + (isTop ? 1 : -1);
@@ -870,6 +_,14 @@
         }
     }
 
+    public net.neoforged.neoforge.client.gui.CreativeTabsScreenPage getCurrentPage() {
+        return currentPage;
+    }
+
+    public void setCurrentPage(net.neoforged.neoforge.client.gui.CreativeTabsScreenPage currentPage) {
+        this.currentPage = currentPage;
+    }
+
     @OnlyIn(Dist.CLIENT)
     private static class CustomCreativeSlot extends Slot {
         public CustomCreativeSlot(Container container, int slot, int x, int y) {
@@ -1050,6 +_,22 @@
         @Override
         public boolean mayPickup(Player player) {
             return this.target.mayPickup(player);
+        }
+
+        @Override
+        public int getSlotIndex() {
+            return this.target.getSlotIndex();
+        }
+
+        @Override
+        public boolean isSameInventory(Slot other) {
+            return this.target.isSameInventory(other);
+        }
+
+        @Override
+        public Slot setBackground(Identifier sprite) {
+            this.target.setBackground(sprite);
+            return this;
         }
     }
 }
