--- a/net/minecraft/client/gui/screens/inventory/InventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/InventoryScreen.java
@@ -118,9 +_,29 @@
     ) {
         float f = (float)(x1 + x2) / 2.0F;
         float f1 = (float)(y1 + y2) / 2.0F;
-        guiGraphics.enableScissor(x1, y1, x2, y2);
         float f2 = (float)Math.atan((double)((f - mouseX) / 40.0F));
         float f3 = (float)Math.atan((double)((f1 - mouseY) / 40.0F));
+        // Forge: Allow passing in direct angle components instead of mouse position
+        renderEntityInInventoryFollowsAngle(guiGraphics, x1, y1, x2, y2, scale, yOffset, f2, f3, entity);
+    }
+
+    public static void renderEntityInInventoryFollowsAngle(
+        GuiGraphics guiGraphics,
+        int x1,
+        int y1,
+        int x2,
+        int y2,
+        int scale,
+        float yOffset,
+        float angleXComponent,
+        float angleYComponent,
+        LivingEntity entity
+    ) {
+        float f = (float)(x1 + x2) / 2.0F;
+        float f1 = (float)(y1 + y2) / 2.0F;
+        guiGraphics.enableScissor(x1, y1, x2, y2);
+        float f2 = angleXComponent;
+        float f3 = angleYComponent;
         Quaternionf quaternionf = new Quaternionf().rotateZ((float) Math.PI);
         Quaternionf quaternionf1 = new Quaternionf().rotateX(f3 * 20.0F * (float) (Math.PI / 180.0));
         quaternionf.mul(quaternionf1);
