--- a/net/minecraft/client/gui/screens/inventory/InventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/InventoryScreen.java
@@ -111,6 +_,13 @@
         float centerY = (y0 + y1) / 2.0F;
         float xAngle = (float)Math.atan((centerX - mouseX) / 40.0F);
         float yAngle = (float)Math.atan((centerY - mouseY) / 40.0F);
+        // Forge: Allow passing in direct angle components instead of mouse position
+        renderEntityInInventoryFollowsAngle(graphics, x0, y0, x1, y1, size, offsetY, xAngle, yAngle, entity);
+    }
+
+    public static void renderEntityInInventoryFollowsAngle(
+        GuiGraphicsExtractor graphics, int x0, int y0, int x1, int y1, int size, float offsetY, float xAngle, float yAngle, LivingEntity entity
+    ) {
         Quaternionf rotation = new Quaternionf().rotateZ((float) Math.PI);
         Quaternionf xRotation = new Quaternionf().rotateX(yAngle * 20.0F * (float) (Math.PI / 180.0));
         rotation.mul(xRotation);
