--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -336,9 +_,9 @@
     public boolean wantsToPickUp(ServerLevel level, ItemStack itemStack) {
         ItemStack itemInHand = this.getItemInHand(InteractionHand.MAIN_HAND);
         return !itemInHand.isEmpty()
-            && level.getGameRules().get(GameRules.MOB_GRIEFING)
             && this.inventory.canAddItem(itemStack)
-            && this.allayConsidersItemEqual(itemInHand, itemStack);
+            && this.allayConsidersItemEqual(itemInHand, itemStack)
+            && net.neoforged.neoforge.event.EventHooks.canEntityGrief(level, this);
     }
 
     private boolean allayConsidersItemEqual(ItemStack item1, ItemStack item2) {
