--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -368,9 +_,9 @@
     public boolean wantsToPickUp(ItemStack stack) {
         ItemStack itemstack = this.getItemInHand(InteractionHand.MAIN_HAND);
         return !itemstack.isEmpty()
-            && this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
             && this.inventory.canAddItem(stack)
-            && this.allayConsidersItemEqual(itemstack, stack);
+            && this.allayConsidersItemEqual(itemstack, stack)
+            && net.neoforged.neoforge.event.EventHooks.canEntityGrief(this.level(), this);
     }
 
     private boolean allayConsidersItemEqual(ItemStack first, ItemStack second) {
