--- a/net/minecraft/world/item/ArmorItem.java
+++ b/net/minecraft/world/item/ArmorItem.java
@@ -49,6 +_,8 @@
         } else {
             LivingEntity livingentity = list.get(0);
             EquipmentSlot equipmentslot = livingentity.getEquipmentSlotForItem(p_40400_);
+            // Neo: Respect IItemExtension#canEquip in dispenseArmor
+            if (!p_40400_.canEquip(equipmentslot, livingentity)) return false;
             ItemStack itemstack = p_40400_.split(1);
             livingentity.setItemSlot(equipmentslot, itemstack);
             if (livingentity instanceof Mob) {
