--- a/net/minecraft/world/item/ArrowItem.java
+++ b/net/minecraft/world/item/ArrowItem.java
@@ -24,4 +_,10 @@
         arrow.pickup = AbstractArrow.Pickup.ALLOWED;
         return arrow;
     }
+
+    public boolean isInfinite(ItemStack stack, ItemStack bow, net.minecraft.world.entity.LivingEntity livingEntity) {
+        // TODO 1.21 - probably needs to decide based on tags/other datapack logic
+        int enchant = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(livingEntity.registryAccess().registryOrThrow(net.minecraft.core.registries.Registries.ENCHANTMENT).getHolderOrThrow(net.minecraft.world.item.enchantment.Enchantments.INFINITY), bow);
+        return enchant > 0 && this.getClass() == net.minecraft.world.item.ArrowItem.class;
+    }
 }
