--- a/net/minecraft/world/item/ArrowItem.java
+++ b/net/minecraft/world/item/ArrowItem.java
@@ -13,4 +_,9 @@
     public AbstractArrow createArrow(Level p_40513_, ItemStack p_40514_, LivingEntity p_40515_) {
         return new Arrow(p_40513_, p_40515_, p_40514_.copyWithCount(1));
     }
+
+    public boolean isInfinite(ItemStack stack, ItemStack bow, net.minecraft.world.entity.LivingEntity livingEntity) {
+        int enchant = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.INFINITY, bow);
+        return enchant > 0 && this.getClass() == net.minecraft.world.item.ArrowItem.class;
+    }
 }
