--- a/net/minecraft/world/item/BoneMealItem.java
+++ b/net/minecraft/world/item/BoneMealItem.java
@@ -37,7 +_,7 @@
         Level level = context.getLevel();
         BlockPos blockpos = context.getClickedPos();
         BlockPos blockpos1 = blockpos.relative(context.getClickedFace());
-        if (growCrop(context.getItemInHand(), level, blockpos)) {
+        if (applyBonemeal(context.getItemInHand(), level, blockpos, context.getPlayer())) {
             if (!level.isClientSide) {
                 context.getPlayer().gameEvent(GameEvent.ITEM_INTERACT_FINISH);
                 level.levelEvent(1505, blockpos, 15);
@@ -60,8 +_,17 @@
         }
     }
 
+    @Deprecated //Forge: Use Player/Hand version
     public static boolean growCrop(ItemStack stack, Level level, BlockPos pos) {
+        if (level instanceof net.minecraft.server.level.ServerLevel)
+            return applyBonemeal(stack, level, pos, null);
+        return false;
+    }
+
+    public static boolean applyBonemeal(ItemStack stack, Level level, BlockPos pos, @Nullable net.minecraft.world.entity.player.Player player) {
         BlockState blockstate = level.getBlockState(pos);
+        var event = net.neoforged.neoforge.event.EventHooks.fireBonemealEvent(player, level, pos, blockstate, stack);
+        if (event.isCanceled()) return event.isSuccessful();
         if (blockstate.getBlock() instanceof BonemealableBlock bonemealableblock && bonemealableblock.isValidBonemealTarget(level, pos, blockstate)) {
             if (level instanceof ServerLevel) {
                 if (bonemealableblock.isBonemealSuccess(level, level.random, pos, blockstate)) {
