--- a/net/minecraft/world/item/BoneMealItem.java
+++ b/net/minecraft/world/item/BoneMealItem.java
@@ -37,7 +_,7 @@
         BlockPos pos = context.getClickedPos();
         BlockPos relative = pos.relative(context.getClickedFace());
         ItemStack boneMealStack = context.getItemInHand();
-        if (growCrop(boneMealStack, level, pos)) {
+        if (applyBonemeal(boneMealStack, level, pos, context.getPlayer())) {
             if (!level.isClientSide()) {
                 boneMealStack.causeUseVibration(context.getPlayer(), GameEvent.ITEM_INTERACT_FINISH);
                 level.levelEvent(1505, pos, 15);
@@ -61,8 +_,17 @@
         }
     }
 
+    @Deprecated //Forge: Use Player/Hand version
     public static boolean growCrop(ItemStack itemStack, Level level, BlockPos pos) {
+        if (level instanceof net.minecraft.server.level.ServerLevel)
+            return applyBonemeal(itemStack, level, pos, null);
+        return false;
+    }
+
+    public static boolean applyBonemeal(ItemStack itemStack, Level level, BlockPos pos, net.minecraft.world.entity.player.@Nullable Player player) {
         BlockState state = level.getBlockState(pos);
+        var event = net.neoforged.neoforge.event.EventHooks.fireBonemealEvent(player, level, pos, state, itemStack);
+        if (event.isCanceled()) return event.isSuccessful();
         if (state.getBlock() instanceof BonemealableBlock block && block.isValidBonemealTarget(level, pos, state)) {
             if (level instanceof ServerLevel) {
                 if (block.isBonemealSuccess(level, level.getRandom(), pos, state)) {
