--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -31,7 +_,7 @@
                 float f = getPowerForTime(i);
                 if (!((double)f < 0.1)) {
                     List<ItemStack> list = draw(p_40667_, itemstack, player);
-                    if (p_40668_ instanceof ServerLevel serverlevel && !list.isEmpty()) {
+                    if (p_40668_ instanceof ServerLevel serverlevel && !list.isEmpty() && net.neoforged.neoforge.event.EventHooks.onArrowLoose(itemstack, serverlevel, player, i, true) > 0) {
                         this.shoot(serverlevel, player, player.getUsedItemHand(), p_40667_, list, f * 3.0F, 1.0F, f == 1.0F, null);
                     }
 
@@ -82,6 +_,10 @@
     public InteractionResultHolder<ItemStack> use(Level p_40672_, Player p_40673_, InteractionHand p_40674_) {
         ItemStack itemstack = p_40673_.getItemInHand(p_40674_);
         boolean flag = !p_40673_.getProjectile(itemstack).isEmpty();
+
+        InteractionResultHolder<ItemStack> ret = net.neoforged.neoforge.event.EventHooks.onArrowNock(itemstack, p_40672_, p_40673_, p_40674_, flag);
+        if (ret != null) return ret;
+
         if (!p_40673_.hasInfiniteMaterials() && !flag) {
             return InteractionResultHolder.fail(itemstack);
         } else {
