--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -28,6 +_,8 @@
             ItemStack itemstack = player.getProjectile(stack);
             if (!itemstack.isEmpty()) {
                 int i = this.getUseDuration(stack, entityLiving) - timeLeft;
+                i = net.neoforged.neoforge.event.EventHooks.onArrowLoose(itemstack, level, player, i, !itemstack.isEmpty());
+                if (i < 0) return;
                 float f = getPowerForTime(i);
                 if (!((double)f < 0.1)) {
                     List<ItemStack> list = draw(stack, itemstack, player);
@@ -82,6 +_,10 @@
     public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
         boolean flag = !player.getProjectile(itemstack).isEmpty();
+
+        InteractionResultHolder<ItemStack> ret = net.neoforged.neoforge.event.EventHooks.onArrowNock(itemstack, level, player, hand, flag);
+        if (ret != null) return ret;
+
         if (!player.hasInfiniteMaterials() && !flag) {
             return InteractionResultHolder.fail(itemstack);
         } else {
