--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -25,7 +_,7 @@
     public void releaseUsing(ItemStack p_40667_, Level p_40668_, LivingEntity p_40669_, int p_40670_) {
         if (p_40669_ instanceof Player player) {
             ItemStack itemstack = player.getProjectile(p_40667_);
-            if (!itemstack.isEmpty()) {
+            if (!itemstack.isEmpty()) { // Porting 1.20.5 redo EventHooks.onArrowLoose
                 int i = this.getUseDuration(p_40667_) - p_40670_;
                 float f = getPowerForTime(i);
                 if (!((double)f < 0.1)) {
@@ -81,6 +_,10 @@
     public InteractionResultHolder<ItemStack> use(Level p_40672_, Player p_40673_, InteractionHand p_40674_) {
         ItemStack itemstack = p_40673_.getItemInHand(p_40674_);
         boolean flag = !p_40673_.getProjectile(itemstack).isEmpty();
+
+        InteractionResultHolder<ItemStack> ret = net.neoforged.neoforge.common.CommonHooks.onArrowNock(itemstack, p_40672_, p_40673_, p_40674_, flag);
+        if (ret != null) return ret;
+
         if (!p_40673_.hasInfiniteMaterials() && !flag) {
             return InteractionResultHolder.fail(itemstack);
         } else {
