--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -59,7 +_,7 @@
                     ItemStack itemstack3 = bucketpickup.pickupBlock(p_40704_, p_40703_, blockpos, blockstate1);
                     if (!itemstack3.isEmpty()) {
                         p_40704_.awardStat(Stats.ITEM_USED.get(this));
-                        bucketpickup.getPickupSound().ifPresent(p_150709_ -> p_40704_.playSound(p_150709_, 1.0F, 1.0F));
+                        bucketpickup.getPickupSound(blockstate1).ifPresent(p_150709_ -> p_40704_.playSound(p_150709_, 1.0F, 1.0F));
                         p_40703_.gameEvent(p_40704_, GameEvent.FLUID_PICKUP, blockpos);
                         ItemStack itemstack2 = ItemUtils.createFilledResult(itemstack, p_40704_, itemstack3);
                         if (!p_40703_.isClientSide()) {
@@ -73,8 +_,8 @@
                 return InteractionResult.FAIL;
             } else {
                 BlockState blockstate = p_40703_.getBlockState(blockpos);
-                BlockPos blockpos2 = blockstate.getBlock() instanceof LiquidBlockContainer && this.content == Fluids.WATER ? blockpos : blockpos1;
-                if (this.emptyContents(p_40704_, p_40703_, blockpos2, blockhitresult)) {
+                BlockPos blockpos2 = canBlockContainFluid(p_40704_, p_40703_, blockpos, blockstate) ? blockpos : blockpos1;
+                if (this.emptyContents(p_40704_, p_40703_, blockpos2, blockhitresult, itemstack)) {
                     this.checkExtraContent(p_40704_, p_40703_, itemstack, blockpos2);
                     if (p_40704_ instanceof ServerPlayer) {
                         CriteriaTriggers.PLACED_BLOCK.trigger((ServerPlayer)p_40704_, blockpos2, itemstack);
@@ -99,7 +_,12 @@
     }
 
     @Override
+    @Deprecated // Neo: use the ItemStack sensitive version
     public boolean emptyContents(@Nullable LivingEntity p_394627_, Level p_150717_, BlockPos p_150718_, @Nullable BlockHitResult p_150719_) {
+        return this.emptyContents(p_394627_, p_150717_, p_150718_, p_150719_, null);
+    }
+
+    public boolean emptyContents(@Nullable LivingEntity p_394627_, Level p_150717_, BlockPos p_150718_, @Nullable BlockHitResult p_150719_, @Nullable ItemStack container) {
         if (!(this.content instanceof FlowingFluid flowingfluid)) {
             return false;
         } else {
@@ -110,9 +_,14 @@
             boolean flag2 = $$8
                 || $$7 instanceof LiquidBlockContainer liquidblockcontainer
                     && liquidblockcontainer.canPlaceLiquid(p_394627_, p_150717_, p_150718_, blockstate, this.content);
+            // TODO: The stack is only used as a parameter in the vaporization check. What if the fluidType is different? And should the rest of the method be made stack-aware as well?
+            var containedFluidStack = container != null ? net.neoforged.neoforge.transfer.fluid.FluidUtil.getFirstStackContained(container) : net.neoforged.neoforge.fluids.FluidStack.EMPTY;
             boolean flag3 = blockstate.isAir() || flag2 && (!flag1 || p_150719_ == null);
             if (!flag3) {
-                return p_150719_ != null && this.emptyContents(p_394627_, p_150717_, p_150719_.getBlockPos().relative(p_150719_.getDirection()), null);
+                return p_150719_ != null && this.emptyContents(p_394627_, p_150717_, p_150719_.getBlockPos().relative(p_150719_.getDirection()), null, container);
+            } else if (!containedFluidStack.isEmpty() && this.content.getFluidType().isVaporizedOnPlacement(p_150717_, p_150718_, containedFluidStack)) {
+                this.content.getFluidType().onVaporize(p_394627_, p_150717_, p_150718_, containedFluidStack);
+                return true;
             } else if (p_150717_.dimensionType().ultraWarm() && this.content.is(FluidTags.WATER)) {
                 int l = p_150718_.getX();
                 int i = p_150718_.getY();
@@ -131,7 +_,7 @@
                 }
 
                 return true;
-            } else if ($$7 instanceof LiquidBlockContainer liquidblockcontainer1 && this.content == Fluids.WATER) {
+            } else if ($$7 instanceof LiquidBlockContainer liquidblockcontainer1 && liquidblockcontainer1.canPlaceLiquid(p_394627_, p_150717_, p_150718_, blockstate, content)) {
                 liquidblockcontainer1.placeLiquid(p_150717_, p_150718_, blockstate, flowingfluid.getSource(false));
                 this.playEmptySound(p_394627_, p_150717_, p_150718_);
                 return true;
@@ -151,9 +_,15 @@
     }
 
     protected void playEmptySound(@Nullable LivingEntity p_394061_, LevelAccessor p_40697_, BlockPos p_40698_) {
-        SoundEvent soundevent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
+        SoundEvent soundevent = this.content.getFluidType().getSound(p_394061_, p_40697_, p_40698_, net.neoforged.neoforge.common.SoundActions.BUCKET_EMPTY);
+        if(soundevent == null) soundevent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
         p_40697_.playSound(p_394061_, p_40698_, soundevent, SoundSource.BLOCKS, 1.0F, 1.0F);
         p_40697_.gameEvent(p_394061_, GameEvent.FLUID_PLACE, p_40698_);
+    }
+
+    protected boolean canBlockContainFluid(@Nullable Player player, Level worldIn, BlockPos posIn, BlockState blockstate)
+    {
+        return blockstate.getBlock() instanceof LiquidBlockContainer && ((LiquidBlockContainer)blockstate.getBlock()).canPlaceLiquid(player, worldIn, posIn, blockstate, this.content);
     }
 
     public Fluid getContent() {
