--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -59,7 +_,7 @@
                     ItemStack taken = bucketPickupBlock.pickupBlock(player, level, pos, blockState);
                     if (!taken.isEmpty()) {
                         player.awardStat(Stats.ITEM_USED.get(this));
-                        bucketPickupBlock.getPickupSound().ifPresent(soundEvent -> player.playSound(soundEvent, 1.0F, 1.0F));
+                        bucketPickupBlock.getPickupSound(blockState).ifPresent(soundEvent -> player.playSound(soundEvent, 1.0F, 1.0F));
                         level.gameEvent(player, GameEvent.FLUID_PICKUP, pos);
                         ItemStack result = ItemUtils.createFilledResult(itemStack, player, taken);
                         if (!level.isClientSide()) {
@@ -73,8 +_,8 @@
                 return InteractionResult.FAIL;
             } else {
                 BlockState clicked = level.getBlockState(pos);
-                BlockPos placePos = clicked.getBlock() instanceof LiquidBlockContainer && this.content == Fluids.WATER ? pos : directionOffsetPos;
-                if (this.emptyContents(player, level, placePos, hitResult)) {
+                BlockPos placePos = canBlockContainFluid(player, level, pos, clicked) && this.content == Fluids.WATER ? pos : directionOffsetPos;
+                if (this.emptyContents(player, level, placePos, hitResult, itemStack)) {
                     this.checkExtraContent(player, level, itemStack, placePos);
                     if (player instanceof ServerPlayer) {
                         CriteriaTriggers.PLACED_BLOCK.trigger((ServerPlayer)player, placePos, itemStack);
@@ -99,7 +_,12 @@
     }
 
     @Override
+    @Deprecated // Neo: use the ItemStack sensitive version
     public boolean emptyContents(@Nullable LivingEntity user, Level level, BlockPos pos, @Nullable BlockHitResult hitResult) {
+        return this.emptyContents(user, level, pos, hitResult, null);
+    }
+
+    public boolean emptyContents(@Nullable LivingEntity user, Level level, BlockPos pos, @Nullable BlockHitResult hitResult, @Nullable ItemStack containerItem) {
         if (!(this.content instanceof FlowingFluid flowingFluid)) {
             return false;
         } else {
@@ -109,9 +_,14 @@
             boolean shiftKeyDown = user != null && user.isShiftKeyDown();
             boolean placeLiquid = mayReplace
                 || block instanceof LiquidBlockContainer container && container.canPlaceLiquid(user, level, pos, blockState, this.content);
+            // TODO: The stack is only used as a parameter in the vaporization check. What if the fluidType is different? And should the rest of the method be made stack-aware as well?
+            var containedFluidStack = containerItem != null ? net.neoforged.neoforge.transfer.fluid.FluidUtil.getFirstStackContained(containerItem) : net.neoforged.neoforge.fluids.FluidStack.EMPTY;
             boolean canPlaceFluidInsideBlock = blockState.isAir() || placeLiquid && (!shiftKeyDown || hitResult == null);
             if (!canPlaceFluidInsideBlock) {
-                return hitResult != null && this.emptyContents(user, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null);
+                return hitResult != null && this.emptyContents(user, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null, containerItem);
+            } else if (!containedFluidStack.isEmpty() && this.content.getFluidType().isVaporizedOnPlacement(level, pos, containedFluidStack)) {
+                this.content.getFluidType().onVaporize(user, level, pos, containedFluidStack);
+                return true;
             } else if (level.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, pos) && this.content.is(FluidTags.WATER)) {
                 int x = pos.getX();
                 int y = pos.getY();
@@ -124,8 +_,8 @@
                 }
 
                 return true;
-            } else if (block instanceof LiquidBlockContainer containerx && this.content == Fluids.WATER) {
-                containerx.placeLiquid(level, pos, blockState, flowingFluid.getSource(false));
+            } else if (block instanceof LiquidBlockContainer container && container.canPlaceLiquid(user, level, pos, blockState, content)) {
+                container.placeLiquid(level, pos, blockState, flowingFluid.getSource(false));
                 this.playEmptySound(user, level, pos);
                 return true;
             } else {
@@ -144,9 +_,15 @@
     }
 
     protected void playEmptySound(@Nullable LivingEntity user, LevelAccessor level, BlockPos pos) {
-        SoundEvent soundEvent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
+        SoundEvent soundEvent = this.content.getFluidType().getSound(user, level, pos, net.neoforged.neoforge.common.SoundActions.BUCKET_EMPTY);
+        if(soundEvent == null) soundEvent = this.content.is(FluidTags.LAVA) ? SoundEvents.BUCKET_EMPTY_LAVA : SoundEvents.BUCKET_EMPTY;
         level.playSound(user, pos, soundEvent, SoundSource.BLOCKS, 1.0F, 1.0F);
         level.gameEvent(user, GameEvent.FLUID_PLACE, pos);
+    }
+
+    protected boolean canBlockContainFluid(@Nullable Player player, Level worldIn, BlockPos posIn, BlockState blockstate)
+    {
+        return blockstate.getBlock() instanceof LiquidBlockContainer && ((LiquidBlockContainer)blockstate.getBlock()).canPlaceLiquid(player, worldIn, posIn, blockstate, this.content);
     }
 
     public Fluid getContent() {
