--- a/net/minecraft/world/item/BundleItem.java
+++ b/net/minecraft/world/item/BundleItem.java
@@ -38,7 +_,7 @@
 
     @Override
     public boolean overrideStackedOnOther(ItemStack stack, Slot slot, ClickAction action, Player player) {
-        if (action != ClickAction.SECONDARY) {
+        if (stack.getCount() != 1 || action != ClickAction.SECONDARY) {
             return false;
         } else {
             BundleContents bundlecontents = stack.get(DataComponents.BUNDLE_CONTENTS);
@@ -71,6 +_,7 @@
     public boolean overrideOtherStackedOnMe(
         ItemStack stack, ItemStack other, Slot slot, ClickAction action, Player player, SlotAccess access
     ) {
+        if (stack.getCount() != 1) return false;
         if (action == ClickAction.SECONDARY && slot.allowModification(player)) {
             BundleContents bundlecontents = stack.get(DataComponents.BUNDLE_CONTENTS);
             if (bundlecontents == null) {
