--- a/net/minecraft/world/item/ChorusFruitItem.java
+++ b/net/minecraft/world/item/ChorusFruitItem.java
@@ -34,7 +_,9 @@
                 }
 
                 Vec3 vec3 = entityLiving.position();
-                if (entityLiving.randomTeleport(d0, d1, d2, true)) {
+                net.neoforged.neoforge.event.entity.EntityTeleportEvent.ChorusFruit event = net.neoforged.neoforge.event.EventHooks.onChorusFruitTeleport(entityLiving, d0, d1, d2);
+                if (event.isCanceled()) return itemstack;
+                if (entityLiving.randomTeleport(event.getTargetX(), event.getTargetY(), event.getTargetZ(), true)) {
                     level.gameEvent(GameEvent.TELEPORT, vec3, GameEvent.Context.of(entityLiving));
                     SoundSource soundsource;
                     SoundEvent soundevent;
