--- a/net/minecraft/world/item/CompassItem.java
+++ b/net/minecraft/world/item/CompassItem.java
@@ -103,6 +_,7 @@
                 ItemStack itemstack1 = new ItemStack(Items.COMPASS, 1);
                 CompoundTag compoundtag = itemstack.hasTag() ? itemstack.getTag().copy() : new CompoundTag();
                 itemstack1.setTag(compoundtag);
+                net.neoforged.neoforge.attachment.AttachmentInternals.copyStackAttachments(itemstack, itemstack1);
                 if (!player.getAbilities().instabuild) {
                     itemstack.shrink(1);
                 }
