--- a/net/minecraft/world/item/CreativeModeTab.java
+++ b/net/minecraft/world/item/CreativeModeTab.java
@@ -13,6 +_,8 @@
 import net.minecraft.world.level.ItemLike;
 
 public class CreativeModeTab {
+    private static final net.minecraft.resources.ResourceLocation SCROLLER_SPRITE = net.minecraft.resources.ResourceLocation.withDefaultNamespace("container/creative_inventory/scroller");
+    private static final net.minecraft.resources.ResourceLocation SCROLLER_DISABLED_SPRITE = net.minecraft.resources.ResourceLocation.withDefaultNamespace("container/creative_inventory/scroller_disabled");
     static final ResourceLocation DEFAULT_BACKGROUND = createTextureLocation("items");
     private final Component displayName;
     ResourceLocation backgroundTexture = DEFAULT_BACKGROUND;
@@ -28,6 +_,15 @@
     private Set<ItemStack> displayItemsSearchTab = ItemStackLinkedSet.createTypeAndComponentsSet();
     private final Supplier<ItemStack> iconGenerator;
     private final CreativeModeTab.DisplayItemsGenerator displayItemsGenerator;
+    @Nullable
+    private final net.minecraft.resources.ResourceLocation scrollerSpriteLocation;
+    private final boolean hasSearchBar;
+    private final int searchBarWidth;
+    private final net.minecraft.resources.ResourceLocation tabsImage;
+    private final int labelColor;
+    private final int slotColor;
+    public final java.util.List<net.minecraft.resources.ResourceLocation> tabsBefore;
+    public final java.util.List<net.minecraft.resources.ResourceLocation> tabsAfter;
 
     CreativeModeTab(
         CreativeModeTab.Row p_260217_,
@@ -35,7 +_,15 @@
         CreativeModeTab.Type p_260176_,
         Component p_260100_,
         Supplier<ItemStack> p_259543_,
-        CreativeModeTab.DisplayItemsGenerator p_259085_
+        CreativeModeTab.DisplayItemsGenerator p_259085_,
+        net.minecraft.resources.ResourceLocation scrollerSpriteLocation,
+        boolean hasSearchBar,
+        int searchBarWidth,
+        net.minecraft.resources.ResourceLocation tabsImage,
+        int labelColor,
+        int slotColor,
+        java.util.List<net.minecraft.resources.ResourceLocation> tabsBefore,
+        java.util.List<net.minecraft.resources.ResourceLocation> tabsAfter
     ) {
         this.row = p_260217_;
         this.column = p_259557_;
@@ -43,12 +_,29 @@
         this.iconGenerator = p_259543_;
         this.displayItemsGenerator = p_259085_;
         this.type = p_260176_;
+        this.scrollerSpriteLocation = scrollerSpriteLocation;
+        this.hasSearchBar = hasSearchBar;
+        this.searchBarWidth = searchBarWidth;
+        this.tabsImage = tabsImage;
+        this.labelColor = labelColor;
+        this.slotColor = slotColor;
+        this.tabsBefore = java.util.List.copyOf(tabsBefore);
+        this.tabsAfter = java.util.List.copyOf(tabsAfter);
+    }
+
+    protected CreativeModeTab(CreativeModeTab.Builder builder) {
+        this(builder.row, builder.column, builder.type, builder.displayName, builder.iconGenerator, builder.displayItemsGenerator, builder.spriteScrollerLocation, builder.hasSearchBar, builder.searchBarWidth, builder.tabsImage, builder.labelColor, builder.slotColor, builder.tabsBefore, builder.tabsAfter);
+    }
+
+    public static CreativeModeTab.Builder builder() {
+        return new CreativeModeTab.Builder(Row.TOP, 0);
     }
 
     public static ResourceLocation createTextureLocation(String p_350882_) {
         return ResourceLocation.withDefaultNamespace("textures/gui/container/creative_inventory/tab_" + p_350882_ + ".png");
     }
 
+    /** @deprecated Forge: use {@link #builder()} **/ @Deprecated
     public static CreativeModeTab.Builder builder(CreativeModeTab.Row p_259342_, int p_260312_) {
         return new CreativeModeTab.Builder(p_259342_, p_260312_);
     }
@@ -106,7 +_,7 @@
         ResourceKey<CreativeModeTab> resourcekey = BuiltInRegistries.CREATIVE_MODE_TAB
             .getResourceKey(this)
             .orElseThrow(() -> new IllegalStateException("Unregistered creative tab: " + this));
-        this.displayItemsGenerator.accept(p_270156_, creativemodetab$itemdisplaybuilder);
+        net.neoforged.neoforge.event.EventHooks.onCreativeModeTabBuildContents(this, resourcekey, this.displayItemsGenerator, p_270156_, creativemodetab$itemdisplaybuilder);
         this.displayItems = creativemodetab$itemdisplaybuilder.tabContents;
         this.displayItemsSearchTab = creativemodetab$itemdisplaybuilder.searchTabContents;
     }
@@ -123,9 +_,37 @@
         return this.displayItemsSearchTab.contains(p_259317_);
     }
 
+    public boolean hasSearchBar() {
+        return this.hasSearchBar;
+    }
+
+    public int getSearchBarWidth() {
+        return searchBarWidth;
+    }
+
+    public net.minecraft.resources.ResourceLocation getTabsImage() {
+        return tabsImage;
+    }
+
+    public int getLabelColor() {
+        return labelColor;
+    }
+
+    public int getSlotColor() {
+        return slotColor;
+    }
+
+    public net.minecraft.resources.ResourceLocation getScrollerSprite() {
+         if (scrollerSpriteLocation == null)
+              return this.canScroll() ? SCROLLER_SPRITE : SCROLLER_DISABLED_SPRITE;
+         return scrollerSpriteLocation;
+    }
+
     public static class Builder {
         private static final CreativeModeTab.DisplayItemsGenerator EMPTY_GENERATOR = (p_270422_, p_259433_) -> {
         };
+        private static final net.minecraft.resources.ResourceLocation CREATIVE_INVENTORY_TABS_IMAGE = net.minecraft.resources.ResourceLocation.withDefaultNamespace("textures/gui/container/creative_inventory/tabs.png");
+        private static final net.minecraft.resources.ResourceLocation CREATIVE_ITEM_SEARCH_BACKGROUND = createTextureLocation("item_search");
         private final CreativeModeTab.Row row;
         private final int column;
         private Component displayName = Component.empty();
@@ -136,6 +_,16 @@
         private boolean alignedRight = false;
         private CreativeModeTab.Type type = CreativeModeTab.Type.CATEGORY;
         private ResourceLocation backgroundTexture = CreativeModeTab.DEFAULT_BACKGROUND;
+        @org.jetbrains.annotations.Nullable
+        private net.minecraft.resources.ResourceLocation spriteScrollerLocation;
+        private boolean hasSearchBar = false;
+        private int searchBarWidth = 89;
+        private net.minecraft.resources.ResourceLocation tabsImage = CREATIVE_INVENTORY_TABS_IMAGE;
+        private int labelColor = 4210752;
+        private int slotColor = -2130706433;
+        private java.util.function.Function<CreativeModeTab.Builder, CreativeModeTab> tabFactory = CreativeModeTab::new;
+        private final java.util.List<net.minecraft.resources.ResourceLocation> tabsBefore = new java.util.ArrayList<>();
+        private final java.util.List<net.minecraft.resources.ResourceLocation> tabsAfter = new java.util.ArrayList<>();
 
         public Builder(CreativeModeTab.Row p_259171_, int p_259661_) {
             this.row = p_259171_;
@@ -174,6 +_,8 @@
 
         protected CreativeModeTab.Builder type(CreativeModeTab.Type p_259283_) {
             this.type = p_259283_;
+            if (p_259283_ == Type.SEARCH)
+                return this.withSearchBar();
             return this;
         }
 
@@ -182,13 +_,94 @@
             return this;
         }
 
+        /**
+         * Gives this tab a search bar.
+         * <p>Note that, if using a custom {@link #withBackgroundLocation(net.minecraft.resources.ResourceLocation) background image}, you will need to make sure that your image contains the input box and the scroll bar.</p>
+         */
+        public CreativeModeTab.Builder withSearchBar() {
+            this.hasSearchBar = true;
+            if (this.backgroundTexture == null)
+                return this.backgroundTexture(CREATIVE_ITEM_SEARCH_BACKGROUND);
+            return this;
+        }
+
+        /**
+         * Gives this tab a search bar, with a specific width.
+         * @param searchBarWidth the width of the search bar
+         */
+        public CreativeModeTab.Builder withSearchBar(int searchBarWidth) {
+            this.searchBarWidth = searchBarWidth;
+            return withSearchBar();
+        }
+
+        /**
+         * Sets the location of the scroll bar background.
+         */
+        public CreativeModeTab.Builder withScrollBarSpriteLocation(net.minecraft.resources.ResourceLocation scrollBarSpriteLocation) {
+             this.spriteScrollerLocation = scrollBarSpriteLocation;
+             return this;
+        }
+
+        /**
+         * Sets the image of the tab to a custom resource location, instead of an item's texture.
+         */
+        public CreativeModeTab.Builder withTabsImage(net.minecraft.resources.ResourceLocation tabsImage) {
+            this.tabsImage = tabsImage;
+            return this;
+        }
+
+        /**
+         * Sets the color of the tab label.
+         */
+        public CreativeModeTab.Builder withLabelColor(int labelColor) {
+             this.labelColor = labelColor;
+             return this;
+        }
+
+        /**
+         * Sets the color of tab's slots.
+         */
+        public CreativeModeTab.Builder withSlotColor(int slotColor) {
+             this.slotColor = slotColor;
+             return this;
+        }
+
+        public CreativeModeTab.Builder withTabFactory(java.util.function.Function<CreativeModeTab.Builder, CreativeModeTab> tabFactory) {
+            this.tabFactory = tabFactory;
+            return this;
+        }
+
+        /** Define tabs that should come <i>before</i> this tab. This tab will be placed <strong>after</strong> the {@code tabs}. **/
+        public CreativeModeTab.Builder withTabsBefore(net.minecraft.resources.ResourceLocation... tabs) {
+            this.tabsBefore.addAll(java.util.List.of(tabs));
+            return this;
+        }
+
+        /** Define tabs that should come <i>after</i> this tab. This tab will be placed <strong>before</strong> the {@code tabs}.**/
+        public CreativeModeTab.Builder withTabsAfter(net.minecraft.resources.ResourceLocation... tabs) {
+            this.tabsAfter.addAll(java.util.List.of(tabs));
+            return this;
+        }
+
+        /** Define tabs that should come <i>before</i> this tab. This tab will be placed <strong>after</strong> the {@code tabs}. **/
+        @SafeVarargs
+        public final CreativeModeTab.Builder withTabsBefore(net.minecraft.resources.ResourceKey<CreativeModeTab>... tabs) {
+            java.util.stream.Stream.of(tabs).map(net.minecraft.resources.ResourceKey::location).forEach(this.tabsBefore::add);
+            return this;
+        }
+
+        /** Define tabs that should come <i>after</i> this tab. This tab will be placed <strong>before</strong> the {@code tabs}.**/
+        @SafeVarargs
+        public final CreativeModeTab.Builder withTabsAfter(net.minecraft.resources.ResourceKey<CreativeModeTab>... tabs) {
+            java.util.stream.Stream.of(tabs).map(net.minecraft.resources.ResourceKey::location).forEach(this.tabsAfter::add);
+            return this;
+        }
+
         public CreativeModeTab build() {
             if ((this.type == CreativeModeTab.Type.HOTBAR || this.type == CreativeModeTab.Type.INVENTORY) && this.displayItemsGenerator != EMPTY_GENERATOR) {
                 throw new IllegalStateException("Special tabs can't have display items");
             } else {
-                CreativeModeTab creativemodetab = new CreativeModeTab(
-                    this.row, this.column, this.type, this.displayName, this.iconGenerator, this.displayItemsGenerator
-                );
+                CreativeModeTab creativemodetab = tabFactory.apply(this);
                 creativemodetab.alignedRight = this.alignedRight;
                 creativemodetab.showTitle = this.showTitle;
                 creativemodetab.canScroll = this.canScroll;
