--- a/net/minecraft/world/item/DispensibleContainerItem.java
+++ b/net/minecraft/world/item/DispensibleContainerItem.java
@@ -6,9 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.BlockHitResult;
 
-public interface DispensibleContainerItem {
+public interface DispensibleContainerItem extends net.neoforged.neoforge.common.extensions.IDispensibleContainerItemExtension {
     default void checkExtraContent(@Nullable Player player, Level level, ItemStack containerStack, BlockPos pos) {
     }
 
+    @Deprecated //Forge: use the ItemStack sensitive version
     boolean emptyContents(@Nullable Player player, Level level, BlockPos pos, @Nullable BlockHitResult result);
 }
