--- a/net/minecraft/world/item/FireChargeItem.java
+++ b/net/minecraft/world/item/FireChargeItem.java
@@ -32,7 +_,8 @@
         BlockPos pos = context.getClickedPos();
         BlockState blockState = level.getBlockState(pos);
         boolean used = false;
-        if (!CampfireBlock.canLight(blockState) && !CandleBlock.canLight(blockState) && !CandleCakeBlock.canLight(blockState)) {
+        BlockState blockstate2 = blockState.getToolModifiedState(context, net.neoforged.neoforge.common.ItemAbilities.FIRESTARTER_LIGHT, false);
+        if (blockstate2 == null) {
             pos = pos.relative(context.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, pos, context.getHorizontalDirection())) {
                 this.playSound(level, pos);
@@ -42,7 +_,7 @@
             }
         } else {
             this.playSound(level, pos);
-            level.setBlockAndUpdate(pos, blockState.setValue(BlockStateProperties.LIT, true));
+            level.setBlockAndUpdate(pos, blockstate2);
             level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, pos);
             used = true;
         }
@@ -84,5 +_,10 @@
             .power(1.0F)
             .overrideDispenseEvent(1018)
             .build();
+    }
+
+    @Override
+    public boolean canPerformAction(ItemInstance stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_FIRECHARGE_ACTIONS.contains(itemAbility);
     }
 }
