--- a/net/minecraft/world/item/FishingRodItem.java
+++ b/net/minecraft/world/item/FishingRodItem.java
@@ -24,7 +_,11 @@
         if (player.fishing != null) {
             if (!level.isClientSide()) {
                 int dmg = player.fishing.retrieve(itemStack);
+                ItemStack original = itemStack.copy();
                 itemStack.hurtAndBreak(dmg, player, hand.asEquipmentSlot());
+                if(itemStack.isEmpty()) {
+                    net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem(player, original, hand);
+                }
             }
 
             level.playSound(
@@ -60,5 +_,10 @@
         }
 
         return InteractionResult.SUCCESS;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemInstance stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_FISHING_ROD_ACTIONS.contains(itemAbility);
     }
 }
