--- a/net/minecraft/world/item/FishingRodItem.java
+++ b/net/minecraft/world/item/FishingRodItem.java
@@ -24,7 +_,11 @@
         if (player.fishing != null) {
             if (!level.isClientSide()) {
                 int dmg = player.fishing.retrieve(itemStack);
+                ItemStack original = itemStack.copy();
                 itemStack.hurtAndBreak(dmg, player, hand.asEquipmentSlot());
+                if(itemStack.isEmpty()) {
+                    net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem(player, original, hand);
+                }
             }
 
             level.playSound(
@@ -52,7 +_,8 @@
             if (level instanceof ServerLevel serverLevel) {
                 int lureSpeed = (int)(EnchantmentHelper.getFishingTimeReduction(serverLevel, itemStack, player) * 20.0F);
                 int luck = EnchantmentHelper.getFishingLuckBonus(serverLevel, itemStack, player);
-                Projectile.spawnProjectile(new FishingHook(player, level, luck, lureSpeed), serverLevel, itemStack);
+                int fishingRodSlot = hand == InteractionHand.OFF_HAND ? -1 : player.getInventory().getSelectedSlot();
+                Projectile.spawnProjectile(new FishingHook(player, level, luck, lureSpeed, fishingRodSlot), serverLevel, itemStack);
             }
 
             player.awardStat(Stats.ITEM_USED.get(this));
@@ -60,5 +_,10 @@
         }
 
         return InteractionResult.SUCCESS;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemInstance stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_FISHING_ROD_ACTIONS.contains(itemAbility);
     }
 }
