--- a/net/minecraft/world/item/FishingRodItem.java
+++ b/net/minecraft/world/item/FishingRodItem.java
@@ -24,7 +_,11 @@
         if (p_41291_.fishing != null) {
             if (!p_41290_.isClientSide()) {
                 int i = p_41291_.fishing.retrieve(itemstack);
+                ItemStack original = itemstack.copy();
                 itemstack.hurtAndBreak(i, p_41291_, p_41292_.asEquipmentSlot());
+                if(itemstack.isEmpty()) {
+                    net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem(p_41291_, original, p_41292_);
+                }
             }
 
             p_41290_.playSound(
@@ -60,5 +_,10 @@
         }
 
         return InteractionResult.SUCCESS;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_FISHING_ROD_ACTIONS.contains(itemAbility);
     }
 }
