--- a/net/minecraft/world/item/FlintAndSteelItem.java
+++ b/net/minecraft/world/item/FlintAndSteelItem.java
@@ -28,7 +_,8 @@
         Level level = context.getLevel();
         BlockPos pos = context.getClickedPos();
         BlockState state = level.getBlockState(pos);
-        if (!CampfireBlock.canLight(state) && !CandleBlock.canLight(state) && !CandleCakeBlock.canLight(state)) {
+        BlockState blockstate2 = state.getToolModifiedState(context, net.neoforged.neoforge.common.ItemAbilities.FIRESTARTER_LIGHT, false);
+        if (blockstate2 == null) {
             BlockPos relativePos = pos.relative(context.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, relativePos, context.getHorizontalDirection())) {
                 level.playSound(player, relativePos, SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
@@ -47,7 +_,7 @@
             }
         } else {
             level.playSound(player, pos, SoundEvents.FLINTANDSTEEL_USE, SoundSource.BLOCKS, 1.0F, level.getRandom().nextFloat() * 0.4F + 0.8F);
-            level.setBlock(pos, state.setValue(BlockStateProperties.LIT, true), 11);
+            level.setBlock(pos, blockstate2, 11);
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
             if (player != null) {
                 context.getItemInHand().hurtAndBreak(1, player, context.getHand().asEquipmentSlot());
@@ -55,5 +_,10 @@
 
             return InteractionResult.SUCCESS;
         }
+    }
+
+    @Override
+    public boolean canPerformAction(ItemInstance stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_FLINT_ACTIONS.contains(itemAbility);
     }
 }
