--- a/net/minecraft/world/item/HoneycombItem.java
+++ b/net/minecraft/world/item/HoneycombItem.java
@@ -3,6 +_,7 @@
 import com.google.common.base.Suppliers;
 import com.google.common.collect.BiMap;
 import com.google.common.collect.ImmutableBiMap;
+
 import java.util.Optional;
 import java.util.function.Supplier;
 import net.minecraft.advancements.CriteriaTriggers;
@@ -21,7 +_,7 @@
 
 public class HoneycombItem extends Item implements SignApplicator {
     public static final Supplier<BiMap<Block, Block>> WAXABLES = Suppliers.memoize(
-        () -> ImmutableBiMap.<Block, Block>builder()
+        () -> com.google.common.collect.HashBiMap.create(ImmutableBiMap.<Block, Block>builder()
                 .put(Blocks.COPPER_BLOCK, Blocks.WAXED_COPPER_BLOCK)
                 .put(Blocks.EXPOSED_COPPER, Blocks.WAXED_EXPOSED_COPPER)
                 .put(Blocks.WEATHERED_COPPER, Blocks.WAXED_WEATHERED_COPPER)
@@ -58,7 +_,7 @@
                 .put(Blocks.EXPOSED_COPPER_BULB, Blocks.WAXED_EXPOSED_COPPER_BULB)
                 .put(Blocks.WEATHERED_COPPER_BULB, Blocks.WAXED_WEATHERED_COPPER_BULB)
                 .put(Blocks.OXIDIZED_COPPER_BULB, Blocks.WAXED_OXIDIZED_COPPER_BULB)
-                .build()
+                .build())
     );
     public static final Supplier<BiMap<Block, Block>> WAX_OFF_BY_BLOCK = Suppliers.memoize(() -> WAXABLES.get().inverse());
 
@@ -103,5 +_,20 @@
     @Override
     public boolean canApplyToSign(SignText p_277550_, Player p_277640_) {
         return true;
+    }
+
+    /**
+     * Registers a before and after blocks that can be waxed and unwaxed
+     * <p>
+     * Important: This should be done during {@link net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent} as the map behind this is NOT thread safe.
+     *
+     * @param before the unwaxed block
+     * @param after  the waxed block
+     */
+    public static void registerWaxableBlock(Block before, Block after) {
+        java.util.Objects.requireNonNull(before, "Waxable before before must not be null");
+        java.util.Objects.requireNonNull(after, "Waxable Block after must not be null");
+
+        HoneycombItem.WAXABLES.get().put(before, after);
     }
 }
