--- a/net/minecraft/world/item/Items.java
+++ b/net/minecraft/world/item/Items.java
@@ -2194,11 +_,19 @@
     }
 
     public static Item registerBlock(Block p_252092_, Block... p_248886_) {
-        Item item = registerBlock(p_252092_);
-
-        for (Block block : p_248886_) {
-            Item.BY_BLOCK.put(block, item);
-        }
+        Item item = registerItem(
+                blockIdToItemId(p_252092_.builtInRegistryHolder().key()),
+                p -> new BlockItem(p_252092_, p) {
+                    @Override
+                    public void registerBlocks(java.util.Map<Block, Item> map, Item self) {
+                        super.registerBlocks(map, self);
+                        for (Block block : p_248886_) {
+                            map.put(block, self);
+                        }
+                    }
+                },
+                new Item.Properties().useBlockDescriptionPrefix()
+        );
 
         return item;
     }
