--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -50,9 +_,9 @@
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
 
-public class Item implements FeatureElement, ItemLike {
+public class Item implements FeatureElement, ItemLike, net.neoforged.neoforge.common.extensions.IItemExtension {
     private static final Logger LOGGER = LogUtils.getLogger();
-    public static final Map<Block, Item> BY_BLOCK = Maps.newHashMap();
+    public static final Map<Block, Item> BY_BLOCK = net.neoforged.neoforge.registries.GameData.getBlockItemMap();
     protected static final UUID BASE_ATTACK_DAMAGE_UUID = UUID.fromString("CB3F55D3-645C-4F38-A497-9C13A33DB5CF");
     protected static final UUID BASE_ATTACK_SPEED_UUID = UUID.fromString("FA233E1C-4180-4865-B01B-BCCE9785ACA3");
     public static final int MAX_STACK_SIZE = 64;
@@ -70,6 +_,7 @@
     @Nullable
     private final FoodProperties foodProperties;
     private final FeatureFlagSet requiredFeatures;
+    private final java.util.function.BooleanSupplier isFeatureEnabled;
 
     public static int getId(Item p_41394_) {
         return p_41394_ == null ? 0 : BuiltInRegistries.ITEM.getId(p_41394_);
@@ -92,12 +_,15 @@
         this.foodProperties = p_41383_.foodProperties;
         this.isFireResistant = p_41383_.isFireResistant;
         this.requiredFeatures = p_41383_.requiredFeatures;
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        this.isFeatureEnabled = p_41383_.isFeatureEnabled;
+        if (SharedConstants.IS_RUNNING_IN_IDE && false) {
             String s = this.getClass().getSimpleName();
             if (!s.endsWith("Item")) {
                 LOGGER.error("Item classes should end with Item and {} doesn't.", s);
             }
         }
+        this.canRepair = p_41383_.canRepair;
+        initClient();
     }
 
     @Deprecated
@@ -108,6 +_,8 @@
     public void onUseTick(Level p_41428_, LivingEntity p_41429_, ItemStack p_41430_, int p_41431_) {
     }
 
+    /** @deprecated Forge: {@link IItemExtension#onDestroyed(ItemEntity, DamageSource) Use damage source sensitive version} */
+    @Deprecated
     public void onDestroyed(ItemEntity p_150887_) {
     }
 
@@ -132,9 +_,9 @@
     }
 
     public InteractionResultHolder<ItemStack> use(Level p_41432_, Player p_41433_, InteractionHand p_41434_) {
-        if (this.isEdible()) {
-            ItemStack itemstack = p_41433_.getItemInHand(p_41434_);
-            if (p_41433_.canEat(this.getFoodProperties().canAlwaysEat())) {
+        ItemStack itemstack = p_41433_.getItemInHand(p_41434_);
+        if (itemstack.isEdible()) {
+            if (p_41433_.canEat(itemstack.getFoodProperties(p_41433_).canAlwaysEat())) {
                 p_41433_.startUsingItem(p_41434_);
                 return InteractionResultHolder.consume(itemstack);
             } else {
@@ -149,10 +_,12 @@
         return this.isEdible() ? p_41411_.eat(p_41410_, p_41409_) : p_41409_;
     }
 
+    @Deprecated // Use ItemStack sensitive version.
     public final int getMaxStackSize() {
         return this.maxStackSize;
     }
 
+    @Deprecated // Use ItemStack sensitive version.
     public final int getMaxDamage() {
         return this.maxDamage;
     }
@@ -166,11 +_,12 @@
     }
 
     public int getBarWidth(ItemStack p_150900_) {
-        return Math.round(13.0F - (float)p_150900_.getDamageValue() * 13.0F / (float)this.maxDamage);
+        return Math.round(13.0F - (float)p_150900_.getDamageValue() * 13.0F / (float)this.getMaxDamage(p_150900_));
     }
 
     public int getBarColor(ItemStack p_150901_) {
-        float f = Math.max(0.0F, ((float)this.maxDamage - (float)p_150901_.getDamageValue()) / (float)this.maxDamage);
+        float stackMaxDamage = this.getMaxDamage(p_150901_);
+        float f = Math.max(0.0F, (stackMaxDamage - (float)p_150901_.getDamageValue()) / stackMaxDamage);
         return Mth.hsvToRgb(f / 3.0F, 1.0F, 1.0F);
     }
 
@@ -206,7 +_,7 @@
 
     @Override
     public String toString() {
-        return BuiltInRegistries.ITEM.getKey(this).getPath();
+        return BuiltInRegistries.ITEM.getKey(this).toString();
     }
 
     protected String getOrCreateDescriptionId() {
@@ -230,10 +_,12 @@
     }
 
     @Nullable
+    @Deprecated // Use ItemStack sensitive version.
     public final Item getCraftingRemainingItem() {
         return this.craftingRemainingItem;
     }
 
+    @Deprecated // Use ItemStack sensitive version.
     public boolean hasCraftingRemainingItem() {
         return this.craftingRemainingItem != null;
     }
@@ -258,7 +_,7 @@
 
     public int getUseDuration(ItemStack p_41454_) {
         if (p_41454_.getItem().isEdible()) {
-            return this.getFoodProperties().isFastFood() ? 16 : 32;
+            return p_41454_.getFoodProperties(null).isFastFood() ? 16 : 32;
         } else {
             return 0;
         }
@@ -300,7 +_,7 @@
     }
 
     public boolean isEnchantable(ItemStack p_41456_) {
-        return this.getMaxStackSize() == 1 && this.canBeDepleted();
+        return this.getMaxStackSize(p_41456_) == 1 && this.isDamageable(p_41456_);
     }
 
     protected static BlockHitResult getPlayerPOVHitResult(Level p_41436_, Player p_41437_, ClipContext.Fluid p_41438_) {
@@ -313,11 +_,14 @@
         float f5 = Mth.sin(-f * (float) (Math.PI / 180.0));
         float f6 = f3 * f4;
         float f7 = f2 * f4;
-        double d0 = 5.0;
-        Vec3 vec31 = vec3.add((double)f6 * 5.0, (double)f5 * 5.0, (double)f7 * 5.0);
+        double d0 = p_41437_.getBlockReach();
+        if (!p_41437_.isCreative() && d0 != 0) d0 += 0.5D; // The vanilla constant here was 5.0, but the default survival block reach is 4.5. Creative default is already 5.0.
+        Vec3 vec31 = vec3.add((double)f6 * d0, (double)f5 * d0, (double)f7 * d0);
         return p_41436_.clip(new ClipContext(vec3, vec31, ClipContext.Block.OUTLINE, p_41438_, p_41437_));
     }
 
+    /** Forge: Use ItemStack sensitive version. */
+    @Deprecated
     public int getEnchantmentValue() {
         return 0;
     }
@@ -326,12 +_,20 @@
         return false;
     }
 
+    @Deprecated // Use ItemStack sensitive version.
     public Multimap<Attribute, AttributeModifier> getDefaultAttributeModifiers(EquipmentSlot p_41388_) {
         return ImmutableMultimap.of();
     }
 
+    protected final boolean canRepair;
+
+    @Override
+    public boolean isRepairable(ItemStack stack) {
+        return canRepair && isDamageable(stack);
+    }
+
     public boolean useOnRelease(ItemStack p_41464_) {
-        return false;
+        return p_41464_.getItem() == Items.CROSSBOW;
     }
 
     public ItemStack getDefaultInstance() {
@@ -342,6 +_,8 @@
         return this.foodProperties != null;
     }
 
+    // Use IForgeItem#getFoodProperties(ItemStack, LivingEntity) in favour of this.
+    @Deprecated
     @Nullable
     public FoodProperties getFoodProperties() {
         return this.foodProperties;
@@ -372,6 +_,37 @@
         return this.requiredFeatures;
     }
 
+    @Override
+    public boolean isFeatureEnabled() {
+        return isFeatureEnabled.getAsBoolean();
+    }
+
+    // FORGE START
+    private Object renderProperties;
+
+    /*
+        DO NOT CALL, IT WILL DISAPPEAR IN THE FUTURE
+        Call RenderProperties.get instead
+     */
+    public Object getRenderPropertiesInternal() {
+        return renderProperties;
+    }
+
+    private void initClient() {
+        // Minecraft instance isn't available in datagen, so don't call initializeClient if in datagen
+        if (net.neoforged.fml.loading.FMLEnvironment.dist == net.neoforged.api.distmarker.Dist.CLIENT && !net.neoforged.fml.loading.FMLLoader.getLaunchHandler().isData()) {
+            initializeClient(properties -> {
+                if (properties == this)
+                    throw new IllegalStateException("Don't extend IItemRenderProperties in your item, use an anonymous class instead.");
+                this.renderProperties = properties;
+            });
+        }
+    }
+
+    public void initializeClient(java.util.function.Consumer<net.neoforged.neoforge.client.extensions.common.IClientItemExtensions> consumer) {
+    }
+    // END FORGE
+
     public static class Properties {
         int maxStackSize = 64;
         int maxDamage;
@@ -382,6 +_,8 @@
         FoodProperties foodProperties;
         boolean isFireResistant;
         FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
+        private boolean canRepair = true;
+        private java.util.function.BooleanSupplier isFeatureEnabled = net.neoforged.neoforge.common.extensions.IFeatureElementExtension::always;
 
         public Item.Properties food(FoodProperties p_41490_) {
             this.foodProperties = p_41490_;
@@ -422,8 +_,20 @@
             return this;
         }
 
+        public Item.Properties setNoRepair() {
+            canRepair = false;
+            return this;
+        }
+
+        /** @deprecated use {@linkplain #isFeatureEnabled(java.util.function.BooleanSupplier)} */
+        @java.lang.Deprecated
         public Item.Properties requiredFeatures(FeatureFlag... p_250948_) {
             this.requiredFeatures = FeatureFlags.REGISTRY.subset(p_250948_);
+            return this;
+        }
+
+        public Item.Properties isFeatureEnabled(java.util.function.BooleanSupplier isFeatureEnabled) {
+            this.isFeatureEnabled = isFeatureEnabled;
             return this;
         }
     }
