--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -81,12 +_,13 @@
 import net.minecraft.world.level.block.state.pattern.BlockInWorld;
 import org.slf4j.Logger;
 
-public final class ItemStack {
+public final class ItemStack extends net.neoforged.neoforge.attachment.AttachmentHolder implements net.neoforged.neoforge.common.extensions.IItemStackExtension {
     public static final Codec<ItemStack> CODEC = RecordCodecBuilder.create(
         p_311716_ -> p_311716_.group(
                     BuiltInRegistries.ITEM.holderByNameCodec().fieldOf("id").forGetter(ItemStack::getItemHolder),
                     Codec.INT.fieldOf("Count").forGetter(ItemStack::getCount),
-                    CompoundTag.CODEC.optionalFieldOf("tag").forGetter(p_281115_ -> Optional.ofNullable(p_281115_.getTag()))
+                    CompoundTag.CODEC.optionalFieldOf("tag").forGetter(p_281115_ -> Optional.ofNullable(p_281115_.getTag())),
+                    ExtraCodecs.strictOptionalField(net.neoforged.neoforge.common.crafting.CraftingHelper.TAG_CODEC, ATTACHMENTS_NBT_KEY).forGetter(s -> Optional.ofNullable(s.serializeAttachments()))
                 )
                 .apply(p_311716_, ItemStack::new)
     );
@@ -104,7 +_,9 @@
     public static final Codec<ItemStack> ITEM_WITH_COUNT_CODEC = RecordCodecBuilder.create(
         p_311720_ -> p_311720_.group(
                     ITEM_NON_AIR_CODEC.fieldOf("item").forGetter(ItemStack::getItem),
-                    ExtraCodecs.strictOptionalField(ExtraCodecs.POSITIVE_INT, "count", 1).forGetter(ItemStack::getCount)
+                    ExtraCodecs.strictOptionalField(ExtraCodecs.POSITIVE_INT, "count", 1).forGetter(ItemStack::getCount),
+                    ExtraCodecs.strictOptionalField(net.neoforged.neoforge.common.crafting.CraftingHelper.TAG_CODEC, "nbt").forGetter(stack -> java.util.Optional.ofNullable(net.neoforged.neoforge.common.crafting.CraftingHelper.getTagForWriting(stack))),
+                    ExtraCodecs.strictOptionalField(net.neoforged.neoforge.common.crafting.CraftingHelper.TAG_CODEC, ATTACHMENTS_NBT_KEY).forGetter(s -> Optional.ofNullable(s.serializeAttachments()))
                 )
                 .apply(p_311720_, ItemStack::new)
     );
@@ -162,18 +_,30 @@
     }
 
     public ItemStack(Holder<Item> p_312081_, int p_41605_, Optional<CompoundTag> p_41606_) {
+        this(p_312081_.value(), p_41605_, p_41606_);
+    }
+    public ItemStack(ItemLike p_312081_, int p_41605_, Optional<CompoundTag> p_41606_) {
         this(p_312081_, p_41605_);
         p_41606_.ifPresent(this::setTag);
     }
+    private ItemStack(Holder<Item> p_312081_, int p_41605_, Optional<CompoundTag> p_41606_, Optional<CompoundTag> attachmentsNbt) {
+        this(p_312081_.value(), p_41605_, p_41606_, attachmentsNbt);
+    }
+    private ItemStack(ItemLike p_312081_, int p_41605_, Optional<CompoundTag> p_41606_, Optional<CompoundTag> attachmentsNbt) {
+        this(p_312081_, p_41605_, attachmentsNbt.orElse(null));
+        p_41606_.ifPresent(this::setTag);
+    }
 
     public ItemStack(Holder<Item> p_220155_, int p_220156_) {
         this(p_220155_.value(), p_220156_);
     }
 
-    public ItemStack(ItemLike p_41601_, int p_41602_) {
+    public ItemStack(ItemLike p_41601_, int p_41602_) { this(p_41601_, p_41602_, (CompoundTag) null); }
+    public ItemStack(ItemLike p_41601_, int p_41602_, @Nullable CompoundTag attachmentsNbt) {
         this.item = p_41601_.asItem();
         this.count = p_41602_;
-        if (this.item.canBeDepleted()) {
+        if (attachmentsNbt != null) deserializeAttachments(attachmentsNbt);
+        if (this.item.isDamageable(this)) {
             this.setDamageValue(this.getDamageValue());
         }
     }
@@ -187,10 +_,15 @@
         this.count = p_41608_.getByte("Count");
         if (p_41608_.contains("tag", 10)) {
             this.tag = p_41608_.getCompound("tag").copy();
+            if (this.tag.contains(ATTACHMENTS_NBT_KEY, Tag.TAG_COMPOUND)) { // Neo: Read contained attachments
+                deserializeAttachments(this.tag.getCompound(ATTACHMENTS_NBT_KEY));
+                this.tag = net.neoforged.neoforge.attachment.AttachmentInternals.cleanTag(this.tag);
+            }
+            if (this.tag != null)
             this.getItem().verifyTagAfterLoad(this.tag);
         }
 
-        if (this.getItem().canBeDepleted()) {
+        if (this.getItem().isDamageable(this)) {
             this.setDamageValue(this.getDamageValue());
         }
     }
@@ -262,6 +_,19 @@
     }
 
     public InteractionResult useOn(UseOnContext p_41662_) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(p_41662_, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_AFTER_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        if (!p_41662_.getLevel().isClientSide) return net.neoforged.neoforge.common.CommonHooks.onPlaceItemIntoWorld(p_41662_);
+        return onItemUse(p_41662_, (c) -> getItem().useOn(p_41662_));
+    }
+
+    public InteractionResult onItemUseFirst(UseOnContext p_41662_) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(p_41662_, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_BEFORE_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        return onItemUse(p_41662_, (c) -> getItem().onItemUseFirst(this, p_41662_));
+    }
+
+    private InteractionResult onItemUse(UseOnContext p_41662_, java.util.function.Function<UseOnContext, InteractionResult> callback) {
         Player player = p_41662_.getPlayer();
         BlockPos blockpos = p_41662_.getClickedPos();
         BlockInWorld blockinworld = new BlockInWorld(p_41662_.getLevel(), blockpos, false);
@@ -271,7 +_,7 @@
             return InteractionResult.PASS;
         } else {
             Item item = this.getItem();
-            InteractionResult interactionresult = item.useOn(p_41662_);
+            InteractionResult interactionresult = callback.apply(p_41662_);
             if (player != null && interactionresult.shouldAwardStats()) {
                 player.awardStat(Stats.ITEM_USED.get(item));
             }
@@ -296,15 +_,16 @@
         ResourceLocation resourcelocation = BuiltInRegistries.ITEM.getKey(this.getItem());
         p_41740_.putString("id", resourcelocation == null ? "minecraft:air" : resourcelocation.toString());
         p_41740_.putByte("Count", (byte)this.count);
-        if (this.tag != null) {
-            p_41740_.put("tag", this.tag.copy());
+        var tag = net.neoforged.neoforge.attachment.AttachmentInternals.addAttachmentsToTag(this.tag, this, true);
+        if (tag != null) {
+            p_41740_.put("tag", tag);
         }
 
         return p_41740_;
     }
 
     public int getMaxStackSize() {
-        return this.getItem().getMaxStackSize();
+        return this.getItem().getMaxStackSize(this);
     }
 
     public boolean isStackable() {
@@ -312,7 +_,7 @@
     }
 
     public boolean isDamageableItem() {
-        if (!this.isEmpty() && this.getItem().getMaxDamage() > 0) {
+        if (!this.isEmpty() && this.getItem().isDamageable(this)) {
             CompoundTag compoundtag = this.getTag();
             return compoundtag == null || !compoundtag.getBoolean("Unbreakable");
         } else {
@@ -321,19 +_,19 @@
     }
 
     public boolean isDamaged() {
-        return this.isDamageableItem() && this.getDamageValue() > 0;
+        return this.isDamageableItem() && getItem().isDamaged(this);
     }
 
     public int getDamageValue() {
-        return this.tag == null ? 0 : this.tag.getInt("Damage");
+        return this.getItem().getDamage(this);
     }
 
     public void setDamageValue(int p_41722_) {
-        this.getOrCreateTag().putInt("Damage", Math.max(0, p_41722_));
+        this.getItem().setDamage(this, p_41722_);
     }
 
     public int getMaxDamage() {
-        return this.getItem().getMaxDamage();
+        return this.getItem().getMaxDamage(this);
     }
 
     public boolean hurt(int p_220158_, RandomSource p_220159_, @Nullable ServerPlayer p_220160_) {
@@ -369,6 +_,7 @@
     public <T extends LivingEntity> void hurtAndBreak(int p_41623_, T p_41624_, Consumer<T> p_41625_) {
         if (!p_41624_.level().isClientSide && (!(p_41624_ instanceof Player) || !((Player)p_41624_).getAbilities().instabuild)) {
             if (this.isDamageableItem()) {
+                p_41623_ = this.getItem().damageItem(this, p_41623_, p_41624_, p_41625_);
                 if (this.hurt(p_41623_, p_41624_.getRandom(), p_41624_ instanceof ServerPlayer ? (ServerPlayer)p_41624_ : null)) {
                     p_41625_.accept(p_41624_);
                     Item item = this.getItem();
@@ -418,7 +_,7 @@
     }
 
     public boolean isCorrectToolForDrops(BlockState p_41736_) {
-        return this.getItem().isCorrectToolForDrops(p_41736_);
+        return this.getItem().isCorrectToolForDrops(this, p_41736_);
     }
 
     public InteractionResult interactLivingEntity(Player p_41648_, LivingEntity p_41649_, InteractionHand p_41650_) {
@@ -430,6 +_,7 @@
             return EMPTY;
         } else {
             ItemStack itemstack = new ItemStack(this.getItem(), this.count);
+            net.neoforged.neoforge.attachment.AttachmentInternals.copyStackAttachments(this, itemstack);
             itemstack.setPopTime(this.getPopTime());
             if (this.tag != null) {
                 itemstack.tag = this.tag.copy();
@@ -465,8 +_,17 @@
         if (!p_150943_.is(p_150944_.getItem())) {
             return false;
         } else {
-            return p_150943_.isEmpty() && p_150944_.isEmpty() ? true : Objects.equals(p_150943_.tag, p_150944_.tag);
+            return p_150943_.isEmpty() && p_150944_.isEmpty() ? true : Objects.equals(p_150943_.tag, p_150944_.tag) && p_150943_.areAttachmentsCompatible(p_150944_);
         }
+    }
+
+    /**
+     * Neo: Check if the attachments of this stack and another stack are compatible.
+     * If not, they cannot be stacked.
+     * @see #areAttachmentsCompatible(net.neoforged.neoforge.attachment.AttachmentHolder, net.neoforged.neoforge.attachment.AttachmentHolder)
+     */
+    public boolean areAttachmentsCompatible(ItemStack other) {
+        return areAttachmentsCompatible(this, other);
     }
 
     public String getDescriptionId() {
@@ -560,7 +_,7 @@
 
     public void setTag(@Nullable CompoundTag p_41752_) {
         this.tag = p_41752_;
-        if (this.getItem().canBeDepleted()) {
+        if (this.getItem().isDamageable(this)) {
             this.setDamageValue(this.getDamageValue());
         }
 
@@ -619,7 +_,7 @@
 
     public List<Component> getTooltipLines(@Nullable Player p_41652_, TooltipFlag p_41653_) {
         List<Component> list = Lists.newArrayList();
-        MutableComponent mutablecomponent = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().color);
+        MutableComponent mutablecomponent = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().getStyleModifier());
         if (this.hasCustomHoverName()) {
             mutablecomponent.withStyle(ChatFormatting.ITALIC);
         }
@@ -792,6 +_,7 @@
             list.add(DISABLED_ITEM_TOOLTIP);
         }
 
+        net.neoforged.neoforge.event.EventHooks.onItemTooltip(this, p_41652_, list, p_41653_);
         return list;
     }
 
@@ -800,7 +_,7 @@
     }
 
     private int getHideFlags() {
-        return this.hasTag() && this.tag.contains("HideFlags", 99) ? this.tag.getInt("HideFlags") : 0;
+        return this.hasTag() && this.tag.contains("HideFlags", 99) ? this.tag.getInt("HideFlags") : this.getItem().getDefaultTooltipHideFlags(this);
     }
 
     public void hideTooltipPart(ItemStack.TooltipPart p_41655_) {
@@ -921,9 +_,10 @@
                 }
             }
         } else {
-            multimap = this.getItem().getDefaultAttributeModifiers(p_41639_);
+            multimap = this.getItem().getAttributeModifiers(p_41639_, this);
         }
 
+        multimap = net.neoforged.neoforge.common.CommonHooks.getAttributeModifiers(this, p_41639_, multimap);
         return multimap;
     }
 
@@ -951,7 +_,7 @@
 
         MutableComponent mutablecomponent1 = ComponentUtils.wrapInSquareBrackets(mutablecomponent);
         if (!this.isEmpty()) {
-            mutablecomponent1.withStyle(this.getRarity().color)
+            mutablecomponent1.withStyle(this.getRarity().getStyleModifier())
                 .withStyle(p_220170_ -> p_220170_.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_ITEM, new HoverEvent.ItemStackInfo(this))));
         }
 
@@ -1002,6 +_,8 @@
         this.getItem().onUseTick(p_41732_, p_41733_, this, p_41734_);
     }
 
+    /** @deprecated Forge: Use {@linkplain IItemStackExtension#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
+    @Deprecated
     public void onDestroyed(ItemEntity p_150925_) {
         this.getItem().onDestroyed(p_150925_);
     }
