--- a/net/minecraft/world/item/ItemStackTemplate.java
+++ b/net/minecraft/world/item/ItemStackTemplate.java
@@ -49,6 +_,26 @@
         this(item.builtInRegistryHolder(), 1, patch);
     }
 
+    // Neo: Provide full set of overloads
+    public ItemStackTemplate(Item item, int count, DataComponentPatch patch) {
+        this(item.builtInRegistryHolder(), count, patch);
+    }
+
+    // Neo: Easier integration with DeferredItem
+    public ItemStackTemplate(Holder<Item> item) {
+        this(item, 1, DataComponentPatch.EMPTY);
+    }
+
+    // Neo: Easier integration with DeferredItem
+    public ItemStackTemplate(Holder<Item> item, int count) {
+        this(item, count, DataComponentPatch.EMPTY);
+    }
+
+    // Neo: Easier integration with DeferredItem
+    public ItemStackTemplate(Holder<Item> item, DataComponentPatch patch) {
+        this(item, 1, patch);
+    }
+
     public ItemStackTemplate(Holder<Item> item, int count, DataComponentPatch components) {
         if (count != 0 && !item.is(Items.AIR.builtInRegistryHolder())) {
             this.item = item;
