--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -237,6 +_,10 @@
         return !this.isEmpty() && this.components.hasNonDefault(type);
     }
 
+    public boolean isComponentsPatchEmpty() {
+        return this.isEmpty() || this.components.isPatchEmpty();
+    }
+
     public ItemStack(ItemLike item, int count) {
         this(item.asItem().builtInRegistryHolder(), count);
     }
@@ -356,13 +_,27 @@
     }
 
     public InteractionResult useOn(UseOnContext context) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(context, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_AFTER_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        if (!context.getLevel().isClientSide()) return net.neoforged.neoforge.common.CommonHooks.onPlaceItemIntoWorld(context);
+        return onItemUse(context, (c) -> getItem().useOn(context));
+    }
+
+    @Override
+    public InteractionResult onItemUseFirst(UseOnContext context) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(context, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_BEFORE_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        return onItemUse(context, (c) -> getItem().onItemUseFirst(this, context));
+    }
+
+    private InteractionResult onItemUse(UseOnContext context, java.util.function.Function<UseOnContext, InteractionResult> callback) {
         Player player = context.getPlayer();
         BlockPos pos = context.getClickedPos();
         if (player != null && !player.getAbilities().mayBuild && !this.canPlaceOnBlockInAdventureMode(new BlockInWorld(context.getLevel(), pos, false))) {
             return InteractionResult.PASS;
         } else {
             Item usedItem = this.getItem();
-            InteractionResult result = usedItem.useOn(context);
+            InteractionResult result = callback.apply(context);
             if (player != null && result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                 player.awardStat(Stats.ITEM_USED.get(usedItem));
             }
@@ -410,6 +_,12 @@
         return result;
     }
 
+    // Neo: Delegate max stack size logic to the item to allow it to be stack-sensitive
+    @Override
+    public int getMaxStackSize() {
+        return this.getItem().getMaxStackSize(this);
+    }
+
     public boolean isStackable() {
         return this.getMaxStackSize() > 1 && (!this.isDamageableItem() || !this.isDamaged());
     }
@@ -419,19 +_,19 @@
     }
 
     public boolean isDamaged() {
-        return this.isDamageableItem() && this.getDamageValue() > 0;
+        return this.isDamageableItem() && getItem().isDamaged(this);
     }
 
     public int getDamageValue() {
-        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage());
+        return this.getItem().getDamage(this);
     }
 
     public void setDamageValue(int value) {
-        this.set(DataComponents.DAMAGE, Mth.clamp(value, 0, this.getMaxDamage()));
+        this.getItem().setDamage(this, value);
     }
 
     public int getMaxDamage() {
-        return this.getOrDefault(DataComponents.MAX_DAMAGE, 0);
+        return this.getItem().getMaxDamage(this);
     }
 
     public boolean isBroken() {
@@ -443,6 +_,11 @@
     }
 
     public void hurtAndBreak(int amount, ServerLevel level, @Nullable ServerPlayer player, Consumer<Item> onBreak) {
+        this.hurtAndBreak(amount, level, (LivingEntity) player, onBreak);
+    }
+
+    public void hurtAndBreak(int amount, ServerLevel level, @Nullable LivingEntity player, Consumer<Item> onBreak) {
+        amount = getItem().damageItem(this, amount, player, onBreak);
         int newAmount = this.processDurabilityChange(amount, level, player);
         if (newAmount != 0) {
             this.applyDamage(this.getDamageValue() + newAmount, player, onBreak);
@@ -450,6 +_,10 @@
     }
 
     private int processDurabilityChange(int amount, ServerLevel level, @Nullable ServerPlayer player) {
+        return processDurabilityChange(amount, level, (LivingEntity) player);
+    }
+
+    private int processDurabilityChange(int amount, ServerLevel level, @Nullable LivingEntity player) {
         if (!this.isDamageableItem()) {
             return 0;
         } else if (player != null && player.hasInfiniteMaterials()) {
@@ -460,8 +_,12 @@
     }
 
     private void applyDamage(int newDamage, @Nullable ServerPlayer player, Consumer<Item> onBreak) {
-        if (player != null) {
-            CriteriaTriggers.ITEM_DURABILITY_CHANGED.trigger(player, this, newDamage);
+        applyDamage(newDamage, (LivingEntity) player, onBreak);
+    }
+
+    private void applyDamage(int newDamage, @Nullable LivingEntity player, Consumer<Item> onBreak) {
+        if (player instanceof ServerPlayer serverPlayer) {
+            CriteriaTriggers.ITEM_DURABILITY_CHANGED.trigger(serverPlayer, this, newDamage);
         }
 
         this.setDamageValue(newDamage);
@@ -491,7 +_,7 @@
     public void hurtAndBreak(int amount, LivingEntity owner, EquipmentSlot slot) {
         if (owner.level() instanceof ServerLevel serverLevel) {
             this.hurtAndBreak(
-                amount, serverLevel, owner instanceof ServerPlayer player ? player : null, brokenItem -> owner.onEquippedItemBroken(brokenItem, slot)
+                amount, serverLevel, owner, brokenItem -> owner.onEquippedItemBroken(brokenItem, slot)
             );
         }
     }
@@ -615,6 +_,16 @@
         }
     }
 
+    /**
+     * Compares an itemstack with an {@link ItemStackTemplate} as per {@link #matches(ItemStack, ItemStack)}.
+     */
+    public static boolean matches(ItemStack a, @Nullable ItemStackTemplate b) {
+        if (b == null) {
+            return a.isEmpty();
+        }
+        return a.getCount() == b.count() && isSameItemSameComponents(a, b);
+    }
+
     @Deprecated
     public static boolean listMatches(List<ItemStack> left, List<ItemStack> right) {
         if (left.size() != right.size()) {
@@ -642,6 +_,27 @@
         }
     }
 
+
+    /**
+     * {@return true if a and b refer to the same item, or if a is empty and b is null}
+     */
+    public static boolean isSameItem(ItemStack a, @Nullable ItemStackTemplate b) {
+        return b == null ? a.isEmpty() : a.is(b.item());
+    }
+
+    /**
+     * Compares the item and components of this stack against an {@link ItemStackTemplate}.
+     *
+     * @return True if either this stack is empty and the template is null, or they reference the same item and have equivalent component patches.
+     */
+    public static boolean isSameItemSameComponents(ItemStack a, @Nullable ItemStackTemplate b) {
+        if (a.isEmpty() || b == null) {
+            return a.isEmpty() == (b == null);
+        } else {
+            return a.is(b.item()) && a.components.patchEquals(b.components());
+        }
+    }
+
     public static boolean matchesIgnoringComponents(ItemStack a, ItemStack b, Predicate<DataComponentType<?>> ignoredPredicate) {
         if (a == b) {
             return true;
@@ -818,7 +_,7 @@
     }
 
     public Component getStyledHoverName() {
-        MutableComponent hoverName = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().color());
+        MutableComponent hoverName = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().getStyleModifier());
         if (this.has(DataComponents.CUSTOM_NAME)) {
             hoverName.withStyle(ChatFormatting.ITALIC);
         }
@@ -826,6 +_,7 @@
         return hoverName;
     }
 
+    @Override // Neo: Add override annotation to ensure our extension method signature always matches vanilla
     public <T extends TooltipProvider> void addToTooltip(
         DataComponentType<T> type, Item.TooltipContext context, TooltipDisplay display, Consumer<Component> consumer, TooltipFlag flag
     ) {
@@ -844,6 +_,7 @@
             List<Component> lines = Lists.newArrayList();
             lines.add(this.getStyledHoverName());
             this.addDetailsToTooltip(context, display, player, tooltipFlag, lines::add);
+            net.neoforged.neoforge.event.EventHooks.onItemTooltip(this, player, lines, tooltipFlag, context);
             return lines;
         }
     }
@@ -872,7 +_,8 @@
         this.addToTooltip(DataComponents.DYED_COLOR, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.PROFILE, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.LORE, context, display, builder, tooltipFlag);
-        this.addAttributeTooltips(builder, display, player);
+        // Neo: Replace attribute tooltips with custom handling
+        net.neoforged.neoforge.common.util.AttributeUtil.addAttributeTooltips(this, builder, display, net.neoforged.neoforge.common.util.AttributeTooltipContext.of(player, context, display, tooltipFlag));
         this.addUnitComponentToTooltip(DataComponents.INTANGIBLE_PROJECTILE, INTANGIBLE_TOOLTIP, display, builder);
         this.addUnitComponentToTooltip(DataComponents.UNBREAKABLE, UNBREAKABLE_TOOLTIP, display, builder);
         this.addToTooltip(DataComponents.OMINOUS_BOTTLE_AMPLIFIER, context, display, builder, tooltipFlag);
@@ -925,7 +_,10 @@
             builder.accept(component);
         }
     }
-
+    /**
+     * @deprecated Neo: Use {@link net.neoforged.neoforge.common.util.AttributeUtil#addAttributeTooltips}
+     */
+    @Deprecated
     private void addAttributeTooltips(Consumer<Component> consumer, TooltipDisplay display, @Nullable Player player) {
         if (display.shows(DataComponents.ATTRIBUTE_MODIFIERS)) {
             for (EquipmentSlotGroup slot : EquipmentSlotGroup.values()) {
@@ -980,19 +_,43 @@
         return !this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY).isEmpty();
     }
 
+    /**
+     * Gets all enchantments from NBT. Use {@link ItemStack#getAllEnchantments} for gameplay logic.
+     */
+    @Override
+    public ItemEnchantments getTagEnchantments() {
+        return getEnchantments();
+    }
+
+    /**
+     * @deprecated Neo: Use {@link #getTagEnchantments()} for NBT enchantments, or {@link #getAllEnchantments} for gameplay.
+     */
+    @Deprecated
     public ItemEnchantments getEnchantments() {
         return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
     }
 
     public void forEachModifier(EquipmentSlotGroup slot, TriConsumer<Holder<Attribute>, AttributeModifier, ItemAttributeModifiers.Display> consumer) {
+        // Neo: Reflect real attribute modifiers when doing iteration
+        this.getAttributeModifiers().forEach(slot, consumer);
+        if (false) {
+        // Start disabled vanilla code
         ItemAttributeModifiers modifiers = this.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
         modifiers.forEach(slot, consumer);
+        // end disabled vanilla code
+        }
         EnchantmentHelper.forEachModifier(this, slot, (a, b) -> consumer.accept(a, b, ItemAttributeModifiers.Display.attributeModifiers()));
     }
 
     public void forEachModifier(EquipmentSlot slot, BiConsumer<Holder<Attribute>, AttributeModifier> consumer) {
+        // Neo: Reflect real attribute modifiers when doing iteration
+        this.getAttributeModifiers().forEach(slot, consumer);
+        if (false) {
+        // Start disabled vanilla code
         ItemAttributeModifiers modifiers = this.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
         modifiers.forEach(slot, consumer);
+        // end disabled vanilla code
+        }
         EnchantmentHelper.forEachModifier(this, slot, consumer);
     }
 
@@ -1004,7 +_,7 @@
 
         MutableComponent result = ComponentUtils.wrapInSquareBrackets(hoverName);
         if (!this.isEmpty()) {
-            result.withStyle(this.getRarity().color()).withStyle(s -> s.withHoverEvent(new HoverEvent.ShowItem(ItemStackTemplate.fromNonEmptyStack(this))));
+            result.withStyle(this.getRarity().getStyleModifier()).withStyle(s -> s.withHoverEvent(new HoverEvent.ShowItem(ItemStackTemplate.fromNonEmptyStack(this))));
         }
 
         return result;
@@ -1085,11 +_,14 @@
         }
     }
 
+    /** @deprecated Forge: Use {@linkplain net.neoforged.neoforge.common.extensions.IItemStackExtension#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
+    @Deprecated
     public void onDestroyed(ItemEntity itemEntity) {
         this.getItem().onDestroyed(itemEntity);
     }
 
     public boolean canBeHurtBy(DamageSource source) {
+        if (!getItem().canBeHurtBy(this, source)) return false;
         DamageResistant damageResistant = this.get(DataComponents.DAMAGE_RESISTANT);
         return damageResistant == null || !damageResistant.isResistantTo(source);
     }
