--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -57,6 +_,16 @@
 
     @Nullable
     public static MapItemSavedData getSavedData(ItemStack stack, Level level) {
+        // Neo: Add instance method so that mods can override
+        Item map = stack.getItem();
+        if(map instanceof MapItem) {
+            return ((MapItem)map).getCustomMapData(stack, level);
+        }
+        return null;
+    }
+
+    @Nullable
+    protected MapItemSavedData getCustomMapData(ItemStack stack, Level level) {
         MapId mapid = stack.get(DataComponents.MAP_ID);
         return getSavedData(mapid, level);
     }
