--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -58,6 +_,15 @@
     }
 
     public static @Nullable MapItemSavedData getSavedData(ItemStack itemStack, Level level) {
+        // Neo: Add instance method so that mods can override
+        if (itemStack.getItem() instanceof MapItem mapItem) {
+            return mapItem.getCustomMapData(itemStack, level);
+        }
+        return null;
+    }
+
+    @Nullable
+    protected MapItemSavedData getCustomMapData(ItemStack itemStack, Level level) {
         MapId id = itemStack.get(DataComponents.MAP_ID);
         return getSavedData(id, level);
     }
