--- a/net/minecraft/world/item/MinecartItem.java
+++ b/net/minecraft/world/item/MinecartItem.java
@@ -34,7 +_,7 @@
         } else {
             ItemStack itemStack = context.getItemInHand();
             RailShape shape = blockState.getBlock() instanceof BaseRailBlock
-                ? blockState.getValue(((BaseRailBlock)blockState.getBlock()).getShapeProperty())
+                ? ((BaseRailBlock)blockState.getBlock()).getRailDirection(blockState, level, pos, null)
                 : RailShape.NORTH_SOUTH;
             double offset = 0.0;
             if (shape.isSlope()) {
