--- a/net/minecraft/world/item/PickaxeItem.java
+++ b/net/minecraft/world/item/PickaxeItem.java
@@ -6,4 +_,9 @@
     public PickaxeItem(Tier p_42961_, Item.Properties p_42964_) {
         super(p_42961_, BlockTags.MINEABLE_WITH_PICKAXE, p_42964_);
     }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_PICKAXE_ACTIONS.contains(itemAbility);
+    }
 }
