--- a/net/minecraft/world/item/ProjectileWeaponItem.java
+++ b/net/minecraft/world/item/ProjectileWeaponItem.java
@@ -24,12 +_,42 @@
         super(properties);
     }
 
+    /**
+     * @deprecated Use ItemStack sensitive version {@link ProjectileWeaponItem#getSupportedHeldProjectiles(ItemStack)}
+     */
+    @Deprecated
     public Predicate<ItemStack> getSupportedHeldProjectiles() {
         return this.getAllSupportedProjectiles();
     }
 
+    /**
+     * @deprecated Use ItemStack sensitive version {@link ProjectileWeaponItem#getAllSupportedProjectiles(ItemStack)}
+     */
+    @Deprecated
     public abstract Predicate<ItemStack> getAllSupportedProjectiles();
 
+    /**
+     * Override this method if the weapon stack allows special projectile that would only be used if it's in hand.
+     * The default return value is a union-predicate of {@link ProjectileWeaponItem#getAllSupportedProjectiles(ItemStack)}
+     * and {@link ProjectileWeaponItem#getSupportedHeldProjectiles()}
+     *
+     * @param stack The ProjectileWeapon stack
+     * @return A predicate that returns true for supported projectile stack in hand
+     */
+    public Predicate<ItemStack> getSupportedHeldProjectiles(ItemStack stack) {
+        return getAllSupportedProjectiles(stack).or(getSupportedHeldProjectiles());
+    }
+
+    /**
+     * Override this method if the allowed projectile is weapon stack dependent.
+     *
+     * @param stack The ProjectileWeapon stack
+     * @return A predicate that returns true for all supported projectile stack
+     */
+    public Predicate<ItemStack> getAllSupportedProjectiles(ItemStack stack) {
+        return getAllSupportedProjectiles();
+    }
+
     public static ItemStack getHeldProjectile(LivingEntity entity, Predicate<ItemStack> valid) {
         if (valid.test(entity.getItemInHand(InteractionHand.OFF_HAND))) {
             return entity.getItemInHand(InteractionHand.OFF_HAND);
@@ -97,7 +_,7 @@
             arrow.setCritArrow(true);
         }
 
-        return arrow;
+        return customArrow(arrow, projectile, weapon);
     }
 
     protected static List<ItemStack> draw(ItemStack weapon, ItemStack projectile, LivingEntity shooter) {
@@ -120,7 +_,8 @@
     }
 
     protected static ItemStack useAmmo(ItemStack weapon, ItemStack projectile, LivingEntity holder, boolean forceInfinite) {
-        int ammoToUse = !forceInfinite && !holder.hasInfiniteMaterials() && holder.level() instanceof ServerLevel serverLevel
+        // Neo: Adjust this check to respect ArrowItem#isInfinite, bypassing processAmmoUse if true.
+        int ammoToUse = !forceInfinite && holder.level() instanceof ServerLevel serverLevel && !(holder.hasInfiniteMaterials() || (projectile.getItem() instanceof ArrowItem ai && ai.isInfinite(projectile, weapon, holder)))
             ? EnchantmentHelper.processAmmoUse(serverLevel, weapon, projectile, 1)
             : 0;
         if (ammoToUse > projectile.getCount()) {
@@ -139,5 +_,20 @@
         }
 
         return used;
+    }
+
+    public AbstractArrow customArrow(AbstractArrow arrow, ItemStack projectileStack, ItemStack weaponStack) {
+        return arrow;
+    }
+
+    /**
+     * Neo: Controls what ammo ItemStack that Creative Mode should return if the player has no valid ammo in inventory.
+     * Modded weapons should override this to return their own ammo if they do not use vanilla arrows.
+     * @param player The player (if in context) firing the weapon
+     * @param projectileWeaponItem The weapon ItemStack the ammo is for
+     * @return The default ammo ItemStack for this weapon
+     */
+    public ItemStack getDefaultCreativeAmmo(@Nullable Player player, ItemStack projectileWeaponItem) {
+        return Items.ARROW.getDefaultInstance();
     }
 }
