--- a/net/minecraft/world/item/Rarity.java
+++ b/net/minecraft/world/item/Rarity.java
@@ -9,7 +_,10 @@
 import net.minecraft.util.ByIdMap;
 import net.minecraft.util.StringRepresentable;
 
-public enum Rarity implements StringRepresentable {
+@net.neoforged.fml.common.asm.enumextension.IndexedEnum
+@net.neoforged.fml.common.asm.enumextension.NamedEnum(1)
+@net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.BIDIRECTIONAL)
+public enum Rarity implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
     COMMON(0, "common", ChatFormatting.WHITE),
     UNCOMMON(1, "uncommon", ChatFormatting.YELLOW),
     RARE(2, "rare", ChatFormatting.AQUA),
@@ -21,19 +_,38 @@
     private final int id;
     private final String name;
     private final ChatFormatting color;
+    private final java.util.function.UnaryOperator<net.minecraft.network.chat.Style> styleModifier;
 
     private Rarity(int p_336096_, String p_335642_, ChatFormatting p_43028_) {
         this.id = p_336096_;
         this.name = p_335642_;
         this.color = p_43028_;
-    }
-
+        this.styleModifier = style -> style.withColor(p_43028_);
+    }
+
+    Rarity(int id, String name, java.util.function.UnaryOperator<net.minecraft.network.chat.Style> styleModifier) {
+        this.id = id;
+        this.name = name;
+        this.color = ChatFormatting.BLACK;
+        this.styleModifier = styleModifier;
+    }
+
+    /** @deprecated NeoForge: Use {@link #getStyleModifier()} */
+    @Deprecated
     public ChatFormatting color() {
         return this.color;
     }
 
+    public java.util.function.UnaryOperator<net.minecraft.network.chat.Style> getStyleModifier() {
+        return this.styleModifier;
+    }
+
     @Override
     public String getSerializedName() {
         return this.name;
+    }
+
+    public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+        return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(Rarity.class);
     }
 }
