--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -55,6 +_,29 @@
     }
 
     @Override
+    public InteractionResult interactLivingEntity(ItemStack stack, Player player, LivingEntity entity, net.minecraft.world.InteractionHand hand) {
+        if (entity instanceof net.neoforged.neoforge.common.IShearable target) {
+            BlockPos pos = entity.blockPosition();
+            Level level = entity.level();
+            if (target.isShearable(player, stack, level, pos)) {
+                if (level.isClientSide) return InteractionResult.SUCCESS;
+                for (ItemStack drop : target.onSheared(player, stack, level, pos)) {
+                    target.spawnShearedDrop(level, pos, drop);
+                }
+                entity.gameEvent(GameEvent.SHEAR, player);
+                stack.hurtAndBreak(1, player, LivingEntity.getSlotForHand(hand));
+                return InteractionResult.CONSUME;
+            }
+        }
+        return InteractionResult.PASS;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ToolAction toolAction) {
+        return net.neoforged.neoforge.common.ToolActions.DEFAULT_SHEARS_ACTIONS.contains(toolAction);
+    }
+
+    @Override
     public InteractionResult useOn(UseOnContext p_186371_) {
         Level level = p_186371_.getLevel();
         BlockPos blockpos = p_186371_.getClickedPos();
