--- a/net/minecraft/world/item/ShearsItem.java
+++ b/net/minecraft/world/item/ShearsItem.java
@@ -55,6 +_,27 @@
     }
 
     @Override
+    public InteractionResult interactLivingEntity(ItemStack stack, Player player, LivingEntity entity, net.minecraft.world.InteractionHand hand) {
+        if (entity instanceof net.neoforged.neoforge.common.IShearable target) {
+            if (entity.level().isClientSide) return InteractionResult.CONSUME;
+            BlockPos pos = entity.blockPosition();
+            if (target.isShearable(player, stack, entity.level(), pos)) {
+                target.onSheared(player, stack, entity.level(), pos)
+                      .forEach(drop -> target.spawnShearedDrop(entity.level(), pos, drop));
+                entity.gameEvent(GameEvent.SHEAR, player);
+                stack.hurtAndBreak(1, player, hand == net.minecraft.world.InteractionHand.MAIN_HAND ? net.minecraft.world.entity.EquipmentSlot.MAINHAND : net.minecraft.world.entity.EquipmentSlot.OFFHAND);
+            }
+            return InteractionResult.SUCCESS;
+        }
+        return InteractionResult.PASS;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_SHEARS_ACTIONS.contains(itemAbility);
+    }
+
+    @Override
     public InteractionResult useOn(UseOnContext p_186371_) {
         Level level = p_186371_.getLevel();
         BlockPos blockpos = p_186371_.getClickedPos();
