--- a/net/minecraft/world/item/ShieldItem.java
+++ b/net/minecraft/world/item/ShieldItem.java
@@ -58,4 +_,9 @@
     public EquipmentSlot getEquipmentSlot() {
         return EquipmentSlot.OFFHAND;
     }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_SHIELD_ACTIONS.contains(itemAbility);
+    }
 }
