--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -43,7 +_,7 @@
             return InteractionResult.PASS;
         } else {
             Player player = p_43119_.getPlayer();
-            BlockState blockstate1 = FLATTENABLES.get(blockstate.getBlock());
+            BlockState blockstate1 = blockstate.getToolModifiedState(p_43119_, net.neoforged.neoforge.common.ToolActions.SHOVEL_FLATTEN, false);
             BlockState blockstate2 = null;
             if (blockstate1 != null && level.getBlockState(blockpos.above()).isAir()) {
                 level.playSound(player, blockpos, SoundEvents.SHOVEL_FLATTEN, SoundSource.BLOCKS, 1.0F, 1.0F);
@@ -71,5 +_,22 @@
                 return InteractionResult.PASS;
             }
         }
+    }
+
+    @org.jetbrains.annotations.Nullable
+    public static BlockState getShovelPathingState(BlockState originalState) {
+        BlockState blockState;
+        var flattenable = originalState.getBlock().builtInRegistryHolder().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.FLATTENABLES);
+        if (flattenable != null) {
+            blockState = flattenable.flattened().defaultBlockState();
+        } else {
+            blockState = FLATTENABLES.get(originalState.getBlock());
+        }
+        return blockState;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ToolAction toolAction) {
+        return net.neoforged.neoforge.common.ToolActions.DEFAULT_SHOVEL_ACTIONS.contains(toolAction);
     }
 }
