--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -176,4 +_,24 @@
 
         return false;
     }
+
+    /**
+     * {@return the dispense behavior to register or null to register the default}
+     */
+    protected net.minecraft.core.dispenser.@Nullable DispenseItemBehavior createDispenseBehavior() {
+        return null;
+    }
+
+    // TODO 26.1: It's questionable that this is even needed anymore
+    @net.neoforged.bus.api.SubscribeEvent(priority = net.neoforged.bus.api.EventPriority.LOWEST)
+    private static void registerDispenseBehavior(final net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent event) {
+        event.enqueueWork(() -> BuiltInRegistries.ITEM.forEach(item -> {
+            if (item instanceof SpawnEggItem egg && !net.minecraft.world.level.block.DispenserBlock.DISPENSER_REGISTRY.containsKey(egg)) {
+                var beh = egg.createDispenseBehavior();
+                if (beh != null) {
+                    net.minecraft.world.level.block.DispenserBlock.registerBehavior(egg, beh);
+                }
+            }
+        }));
+    }
 }
