--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -42,11 +_,14 @@
     private final int highlightColor;
     private final EntityType<?> defaultType;
 
+    /** @deprecated Forge: Use {@link net.neoforged.neoforge.common.DeferredSpawnEggItem} instead for suppliers */
+    @Deprecated
     public SpawnEggItem(EntityType<? extends Mob> p_43207_, int p_43208_, int p_43209_, Item.Properties p_43210_) {
         super(p_43210_);
         this.defaultType = p_43207_;
         this.backgroundColor = p_43208_;
         this.highlightColor = p_43209_;
+        if (p_43207_ != null)
         BY_ID.put(p_43207_, this);
     }
 
@@ -134,6 +_,8 @@
 
     @Nullable
     public static SpawnEggItem byId(@Nullable EntityType<?> p_43214_) {
+        var ret = net.neoforged.neoforge.common.DeferredSpawnEggItem.deferredOnlyById(p_43214_);
+        if (ret != null) return ret;
         return BY_ID.get(p_43214_);
     }
 
@@ -143,12 +_,12 @@
 
     public EntityType<?> getType(ItemStack p_330335_) {
         CustomData customdata = p_330335_.getOrDefault(DataComponents.ENTITY_DATA, CustomData.EMPTY);
-        return !customdata.isEmpty() ? customdata.read(ENTITY_TYPE_FIELD_CODEC).result().orElse(this.defaultType) : this.defaultType;
+        return !customdata.isEmpty() ? customdata.read(ENTITY_TYPE_FIELD_CODEC).result().orElse(getDefaultType()) : getDefaultType();
     }
 
     @Override
     public FeatureFlagSet requiredFeatures() {
-        return this.defaultType.requiredFeatures();
+        return this.getDefaultType().requiredFeatures();
     }
 
     public Optional<Mob> spawnOffspringFromSpawnEgg(
@@ -179,5 +_,9 @@
                 }
             }
         }
+    }
+
+    protected EntityType<?> getDefaultType() {
+        return defaultType;
     }
 }
