--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -39,11 +_,14 @@
     private final int highlightColor;
     private final EntityType<?> defaultType;
 
+    /** @deprecated Forge: Use {@link net.neoforged.neoforge.common.DeferredSpawnEggItem} instead for suppliers */
+    @Deprecated
     public SpawnEggItem(EntityType<? extends Mob> p_43207_, int p_43208_, int p_43209_, Item.Properties p_43210_) {
         super(p_43210_);
         this.defaultType = p_43207_;
         this.backgroundColor = p_43208_;
         this.highlightColor = p_43209_;
+        if (p_43207_ != null)
         BY_ID.put(p_43207_, this);
     }
 
@@ -135,6 +_,8 @@
 
     @Nullable
     public static SpawnEggItem byId(@Nullable EntityType<?> p_43214_) {
+        var ret = net.neoforged.neoforge.common.DeferredSpawnEggItem.deferredOnlyById(p_43214_);
+        if (ret != null) return ret;
         return BY_ID.get(p_43214_);
     }
 
@@ -146,16 +_,16 @@
         if (p_43229_ != null && p_43229_.contains("EntityTag", 10)) {
             CompoundTag compoundtag = p_43229_.getCompound("EntityTag");
             if (compoundtag.contains("id", 8)) {
-                return EntityType.byString(compoundtag.getString("id")).orElse(this.defaultType);
+                return EntityType.byString(compoundtag.getString("id")).orElse(this.getDefaultType());
             }
         }
 
-        return this.defaultType;
+        return this.getDefaultType();
     }
 
     @Override
     public FeatureFlagSet requiredFeatures() {
-        return this.defaultType.requiredFeatures();
+        return this.getDefaultType().requiredFeatures();
     }
 
     public Optional<Mob> spawnOffspringFromSpawnEgg(
@@ -192,5 +_,15 @@
                 }
             }
         }
+    }
+
+    protected EntityType<?> getDefaultType() {
+        return defaultType;
+    }
+
+    @Override
+    public boolean isFeatureEnabled() {
+        // NeoForge: Disabled when associated entity type is also disabled
+        return super.isFeatureEnabled() && getDefaultType().isFeatureEnabled();
     }
 }
