--- a/net/minecraft/world/item/StandingAndWallBlockItem.java
+++ b/net/minecraft/world/item/StandingAndWallBlockItem.java
@@ -50,4 +_,9 @@
         super.registerBlocks(blockToItemMap, item);
         blockToItemMap.put(this.wallBlock, item);
     }
+
+    public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+        super.removeFromBlockToItemMap(blockToItemMap, itemIn);
+        blockToItemMap.remove(this.wallBlock);
+    }
 }
