--- a/net/minecraft/world/item/StandingAndWallBlockItem.java
+++ b/net/minecraft/world/item/StandingAndWallBlockItem.java
@@ -50,4 +_,11 @@
         super.registerBlocks(p_43252_, p_43253_);
         p_43252_.put(this.wallBlock, p_43253_);
     }
+
+    /** @deprecated Neo: To be removed without replacement since registry replacement is not a feature anymore. */
+    @Deprecated(forRemoval = true, since = "1.21.1")
+    public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+        super.removeFromBlockToItemMap(blockToItemMap, itemIn);
+        blockToItemMap.remove(this.wallBlock);
+    }
 }
