--- a/net/minecraft/world/item/SwordItem.java
+++ b/net/minecraft/world/item/SwordItem.java
@@ -21,11 +_,25 @@
         super(p_43269_, p_43272_.component(DataComponents.TOOL, createToolProperties()));
     }
 
+    /**
+     * Neo: Allow modded Swords to set exactly what Tool data component to use for their sword.
+     */
+    public SwordItem(Tier p_43269_, Item.Properties p_43272_, Tool toolComponentData) {
+        super(p_43269_, p_43272_.component(DataComponents.TOOL, toolComponentData));
+    }
+
     private static Tool createToolProperties() {
         return new Tool(List.of(Tool.Rule.minesAndDrops(List.of(Blocks.COBWEB), 15.0F), Tool.Rule.overrideSpeed(BlockTags.SWORD_EFFICIENT, 1.5F)), 1.0F, 2);
     }
 
     public static ItemAttributeModifiers createAttributes(Tier p_330371_, int p_331976_, float p_332104_) {
+        return createAttributes(p_330371_, (float)p_331976_, p_332104_);
+    }
+
+    /**
+     * Neo: Method overload to allow giving a float for damage instead of an int.
+     */
+    public static ItemAttributeModifiers createAttributes(Tier p_330371_, float p_331976_, float p_332104_) {
         return ItemAttributeModifiers.builder()
             .add(
                 Attributes.ATTACK_DAMAGE,
@@ -54,5 +_,10 @@
     public boolean hurtEnemy(ItemStack p_43278_, LivingEntity p_43279_, LivingEntity p_43280_) {
         p_43278_.hurtAndBreak(1, p_43280_, EquipmentSlot.MAINHAND);
         return true;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ToolAction toolAction) {
+        return net.neoforged.neoforge.common.ToolActions.DEFAULT_SWORD_ACTIONS.contains(toolAction);
     }
 }
