--- a/net/minecraft/world/item/SwordItem.java
+++ b/net/minecraft/world/item/SwordItem.java
@@ -12,6 +_,13 @@
         super(p_361460_.applySwordProperties(p_43272_, p_362481_, p_364182_));
     }
 
+    /**
+     * Neo: Allow modded Swords to set exactly what properties to use for their sword.
+     */
+    public SwordItem(Item.Properties properties) {
+        super(properties);
+    }
+
     @Override
     public boolean canAttackBlock(BlockState p_43291_, Level p_43292_, BlockPos p_43293_, Player p_43294_) {
         return !p_43294_.isCreative();
@@ -25,5 +_,10 @@
     @Override
     public void postHurtEnemy(ItemStack p_345553_, LivingEntity p_345771_, LivingEntity p_346282_) {
         p_345553_.hurtAndBreak(1, p_346282_, EquipmentSlot.MAINHAND);
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_SWORD_ACTIONS.contains(itemAbility);
     }
 }
