--- a/net/minecraft/world/item/SwordItem.java
+++ b/net/minecraft/world/item/SwordItem.java
@@ -21,11 +_,25 @@
         super(tier, properties.component(DataComponents.TOOL, createToolProperties()));
     }
 
+    /**
+     * Neo: Allow modded Swords to set exactly what Tool data component to use for their sword.
+     */
+    public SwordItem(Tier tier, Item.Properties properties, Tool toolComponentData) {
+        super(tier, properties.component(DataComponents.TOOL, toolComponentData));
+    }
+
     public static Tool createToolProperties() {
         return new Tool(List.of(Tool.Rule.minesAndDrops(List.of(Blocks.COBWEB), 15.0F), Tool.Rule.overrideSpeed(BlockTags.SWORD_EFFICIENT, 1.5F)), 1.0F, 2);
     }
 
     public static ItemAttributeModifiers createAttributes(Tier tier, int attackDamage, float attackSpeed) {
+        return createAttributes(tier, (float)attackDamage, attackSpeed);
+    }
+
+    /**
+     * Neo: Method overload to allow giving a float for damage instead of an int.
+     */
+    public static ItemAttributeModifiers createAttributes(Tier tier, float attackDamage, float attackSpeed) {
         return ItemAttributeModifiers.builder()
             .add(
                 Attributes.ATTACK_DAMAGE,
@@ -55,5 +_,10 @@
     @Override
     public void postHurtEnemy(ItemStack stack, LivingEntity target, LivingEntity attacker) {
         stack.hurtAndBreak(1, attacker, EquipmentSlot.MAINHAND);
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_SWORD_ACTIONS.contains(itemAbility);
     }
 }
