--- a/net/minecraft/world/item/TooltipFlag.java
+++ b/net/minecraft/world/item/TooltipFlag.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.item;
 
-public interface TooltipFlag extends net.neoforged.neoforge.common.extensions.TooltipFlagExtension {
+import net.neoforged.neoforge.common.extensions.TooltipFlagExtension;
+
+public interface TooltipFlag extends TooltipFlagExtension {
     TooltipFlag.Default NORMAL = new TooltipFlag.Default(false, false);
     TooltipFlag.Default ADVANCED = new TooltipFlag.Default(true, false);
 
