--- a/net/minecraft/world/item/TooltipFlag.java
+++ b/net/minecraft/world/item/TooltipFlag.java
@@ -8,6 +_,27 @@
 
     boolean isCreative();
 
+    /**
+     * Neo: Returns the state of the Control key (as reported by Screen) on the client, or {@code false} on the server.
+     */
+    default boolean hasControlDown() {
+        return false;
+    }
+
+    /**
+     * Neo: Returns the state of the Shift key (as reported by Screen) on the client, or {@code false} on the server.
+     */
+    default boolean hasShiftDown() {
+        return false;
+    }
+
+    /**
+     * Neo: Returns the state of the Alt key (as reported by Screen) on the client, or {@code false} on the server.
+     */
+    default boolean hasAltDown() {
+        return false;
+    }
+
     public record Default(boolean advanced, boolean creative) implements TooltipFlag {
         @Override
         public boolean isAdvanced() {
