--- a/net/minecraft/client/gui/font/glyphs/BakedGlyph.java
+++ b/net/minecraft/client/gui/font/glyphs/BakedGlyph.java
@@ -139,6 +_,15 @@
             .setLight(p_382874_);
     }
 
+    /**
+     * Neo: returns the {@link RenderType} to use for the given {@link Font.DisplayMode} and blur setting
+     */
+    public RenderType renderType(Font.DisplayMode p_181388_, boolean blur) {
+        return this.renderTypes.select(p_181388_, blur);
+    }
+
+    /** @deprecated Neo: Use {@link #renderType(Font.DisplayMode, boolean)} instead */
+    @Deprecated
     public RenderType renderType(Font.DisplayMode p_181388_) {
         return this.renderTypes.select(p_181388_);
     }
