--- a/net/minecraft/core/HolderLookup.java
+++ b/net/minecraft/core/HolderLookup.java
@@ -127,6 +139,12 @@
             @Override
             default Stream<HolderSet.Named<T>> listTags() {
                 return this.parent().listTags();
+            }
+
+            @Override
+            @org.jspecify.annotations.Nullable
+            default <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> attachment, ResourceKey<T> key) {
+                return parent().getData(attachment, key);
             }
         }
     }
