--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -93,7 +_,9 @@
     }
 
     protected void doFreezeConversion() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.STRAY, (timer) -> this.conversionTime = timer)) return;
         this.convertTo(EntityTypes.STRAY, ConversionParams.single(this, true, true), stray -> {
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, stray);
             if (!this.isSilent()) {
                 this.level().levelEvent(null, 1048, this.blockPosition(), 0);
             }
