--- a/net/minecraft/client/resources/sounds/SoundInstance.java
+++ b/net/minecraft/client/resources/sounds/SoundInstance.java
@@ -56,4 +_,9 @@
         NONE,
         LINEAR;
     }
+
+    // Neo: Allow SoundInstances to change source of audio stream if needed
+    default java.util.concurrent.CompletableFuture<net.minecraft.client.sounds.AudioStream> getStream(net.minecraft.client.sounds.SoundBufferLibrary soundBuffers, Sound sound, boolean looping) {
+        return soundBuffers.getStream(sound.getPath(), looping);
+    }
 }
