--- a/net/minecraft/client/renderer/texture/AbstractTexture.java
+++ b/net/minecraft/client/renderer/texture/AbstractTexture.java
@@ -12,6 +_,8 @@
     public static final int NOT_ASSIGNED = -1;
     protected int id = -1;
     protected boolean defaultBlur;
+    protected boolean blur;
+    protected boolean mipmap;
     private int wrapS = 10497;
     private int wrapT = 10497;
     private int minFilter = 9986;
@@ -51,6 +_,8 @@
 
     public void setFilter(boolean p_117961_, boolean p_117962_) {
         RenderSystem.assertOnRenderThreadOrInit();
+        this.blur = p_117961_;
+        this.mipmap = p_117962_;
         int i;
         int j;
         if (p_117961_) {
@@ -75,6 +_,20 @@
                 this.magFilter = j;
             }
         }
+    }
+
+    // FORGE: This seems to have been stripped out, but we need it
+    private boolean lastBlur;
+    private boolean lastMipmap;
+
+    public void setBlurMipmap(boolean blur, boolean mipmap) {
+        this.lastBlur = this.blur;
+        this.lastMipmap = this.mipmap;
+        setFilter(blur, mipmap);
+    }
+
+    public void restoreLastBlurMipmap() {
+        setFilter(this.lastBlur, this.lastMipmap);
     }
 
     public int getId() {
