--- a/net/minecraft/client/renderer/texture/SpriteContents.java
+++ b/net/minecraft/client/renderer/texture/SpriteContents.java
@@ -85,6 +_,10 @@
         this.transparency = image.computeTransparency();
     }
 
+    public NativeImage getOriginalImage() {
+        return this.originalImage;
+    }
+
     public void increaseMipLevel(int mipLevel) {
         try {
             this.byMipLevel = MipmapGenerator.generateMipLevels(
