--- a/net/minecraft/client/renderer/texture/TextureAtlas.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlas.java
@@ -87,6 +_,8 @@
             this.sprites = List.copyOf(list);
             this.animatedTextures = List.copyOf(list1);
         }
+
+        net.neoforged.neoforge.client.ClientHooks.onTextureAtlasStitched(this);
     }
 
     @Override
@@ -172,5 +_,9 @@
 
     public void updateFilter(SpriteLoader.Preparations p_251993_) {
         this.setFilter(false, p_251993_.mipLevel() > 0);
+    }
+
+    public Map<ResourceLocation, TextureAtlasSprite> getTextures() {
+        return java.util.Collections.unmodifiableMap(texturesByName);
     }
 }
