--- a/net/minecraft/client/renderer/texture/TextureAtlas.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlas.java
@@ -82,6 +_,8 @@
             this.sprites = List.copyOf(list);
             this.animatedTextures = List.copyOf(list1);
         }
+
+        net.neoforged.neoforge.client.ClientHooks.onTextureAtlasStitched(this);
     }
 
     @Override
@@ -159,5 +_,9 @@
 
     int getHeight() {
         return this.height;
+    }
+
+    public Map<ResourceLocation, TextureAtlasSprite> getTextures() {
+        return java.util.Collections.unmodifiableMap(texturesByName);
     }
 }
