--- a/net/minecraft/client/renderer/texture/TextureAtlas.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlas.java
@@ -102,6 +_,8 @@
                     LOGGER.warn("Failed to dump atlas contents to {}", path);
                 }
             }
+
+            net.neoforged.neoforge.client.ClientHooks.onTextureAtlasStitched(this);
         }
     }
 
@@ -184,5 +_,9 @@
 
     int getHeight() {
         return this.height;
+    }
+
+    public Map<ResourceLocation, TextureAtlasSprite> getTextures() {
+        return java.util.Collections.unmodifiableMap(texturesByName);
     }
 }
