--- a/com/mojang/blaze3d/textures/GpuTexture.java
+++ b/com/mojang/blaze3d/textures/GpuTexture.java
@@ -12,6 +_,11 @@
     public static final int USAGE_TEXTURE_BINDING = 4;
     public static final int USAGE_RENDER_ATTACHMENT = 8;
     public static final int USAGE_CUBEMAP_COMPATIBLE = 16;
+    // Neo: Extended texture usage, some may not always be supported, check backend capabilities before use
+    // Neo: Requires OpenGL 4.2, ARB_shader_image_load_store, or equivalent
+    public static final int USAGE_STORAGE = USAGE_CUBEMAP_COMPATIBLE << 1;
+    // Neo: Top 8 bits reserved for implementation specific usages
+    public static final int USAGE_FIRST_IMPLEMENTATION_DEFINED = 1 << 24;
     private final TextureFormat format;
     private final int width;
     private final int height;
