--- a/com/mojang/blaze3d/textures/TextureFormat.java
+++ b/com/mojang/blaze3d/textures/TextureFormat.java
@@ -4,13 +_,281 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
+/**
+ * Neo: Extended texture formats follow VkFormat naming
+ *      <br>
+ *      Listed GL enums are internalformat/format/type, referencing glTexImage's parameters of the same name
+ *      <br>
+ *      Some formats may not always be supported or may have limitations on specific usages, check backend capabilities before use.
+ *      <p>
+ *      See Vulkan specification for table on how these are laid out in memory.
+ *      <a href="https://registry.khronos.org/vulkan/specs/latest/html/vkspec.html#texel-block-size">Representation and Texel Block Size</a>
+ */
 @OnlyIn(Dist.CLIENT)
 @DontObfuscate
 public enum TextureFormat {
-    RGBA8(4),
+
+    /**
+     * R8_UNORM specifies a one-component, 8-bit unsigned normalized format that has a single 8-bit R component.
+     * <br>
+     * GL_R8/GL_RED/GL_UNSIGNED_BYTE
+     * <p>
+     * Neo: For compatibility reasons, the Vanilla name of RED8 is kept. An alias constant {@link TextureFormat#R8_UNORM} is added for naming consistency.
+     */
     RED8(1),
+    /**
+     * R8_SNORM specifies a one-component, 8-bit signed normalized format that has a single 8-bit R component.
+     * <br>
+     * GL_R8_SNORM/GL_RED/GL_BYTE
+     */
+    R8_SNORM(1),
+    /**
+     * R8_UINT specifies a one-component, 8-bit unsigned integer format that has a single 8-bit R component.
+     * <br>
+     * GL_R8UI/GL_RED/GL_UNSIGNED_BYTE
+     */
+    R8_UINT(1),
+    /**
+     * R8_SINT specifies a one-component, 8-bit signed integer format that has a single 8-bit R component.
+     * <br>
+     * GL_R8I/GL_RED/GL_BYTE
+     * <p>
+     * For compatibility reasons, the Vanilla name of RED8I is kept. An alias constant {@link TextureFormat#R8_SINT} is added for naming consistency.
+     * <p>
+     * Neo: Vanilla RED8I uses GL_UNSIGNED_BYTE for type, Neo corrects this to GL_BYTE
+     */
     RED8I(1),
-    DEPTH32(4);
+    /**
+     * R8G8_UNORM specifies a two-component, 16-bit unsigned normalized format that has an 8-bit R component in byte 0, and an 8-bit G component in byte 1.
+     * <br>
+     * GL_RG8/GL_RG/GL_UNSIGNED_BYTE
+     */
+    R8G8_UNORM(2),
+    /**
+     * R8G8_SNORM specifies a two-component, 16-bit signed normalized format that has an 8-bit R component in byte 0, and an 8-bit G component in byte 1.
+     * <br>
+     * GL_RG8_SNORM/GL_RG/GL_BYTE
+     */
+    R8G8_SNORM(2),
+    /**
+     * R8G8_UINT specifies a two-component, 16-bit unsigned integer format that has an 8-bit R component in byte 0, and an 8-bit G component in byte 1.
+     * <br>
+     * GL_RG8UI/GL_RG/GL_UNSIGNED_BYTE
+     */
+    R8G8_UINT(2),
+    /**
+     * R8G8_SINT specifies a two-component, 16-bit signed integer format that has an 8-bit R component in byte 0, and an 8-bit G component in byte 1.
+     * <br>
+     * GL_RG8I/GL_RG/GL_BYTE
+     */
+    R8G8_SINT(2),
+    /**
+     * R8G8B8A8_UNORM specifies a four-component, 32-bit unsigned normalized format that has an 8-bit R component in byte 0, an 8-bit G component in byte 1, an 8-bit B component in byte 2, and an 8-bit A component in byte 3.
+     * <br>
+     * GL_RGBA8/GL_RGBA/GL_UNSIGNED_BYTE
+     * <p>
+     * Neo: For compatibility reasons, the Vanilla name of RGBA8 is kept. An alias constant {@link TextureFormat#R8G8B8A8_UNORM} is added for naming consistency.
+     */
+    RGBA8(4),
+    /**
+     * R8G8B8A8_SNORM specifies a four-component, 32-bit signed normalized format that has an 8-bit R component in byte 0, an 8-bit G component in byte 1, an 8-bit B component in byte 2, and an 8-bit A component in byte 3.
+     * <br>
+     * GL_RGBA8_SNORM/GL_RGBA/GL_BYTE
+     */
+    R8G8B8A8_SNORM(4),
+    /**
+     * R8G8B8A8_UINT specifies a four-component, 32-bit unsigned integer format that has an 8-bit R component in byte 0, an 8-bit G component in byte 1, an 8-bit B component in byte 2, and an 8-bit A component in byte 3.
+     * <br>
+     * GL_RGBA8UI/GL_RGBA/GL_UNSIGNED_BYTE
+     */
+    R8G8B8A8_UINT(4),
+    /**
+     * R8G8B8A8_SINT specifies a four-component, 32-bit signed integer format that has an 8-bit R component in byte 0, an 8-bit G component in byte 1, an 8-bit B component in byte 2, and an 8-bit A component in byte 3.
+     * <br>
+     * GL_RGBA8I/GL_RGBA/GL_BYTE
+     */
+    R8G8B8A8_SINT(4),
+
+    /**
+     * R16_UNORM specifies a one-component, 16-bit unsigned normalized format that has a single 16-bit R component.
+     * <br>
+     * GL_R16/GL_RED/GL_UNSIGNED_SHORT
+     */
+    R16_UNORM(2),
+    /**
+     * R16_SNORM specifies a one-component, 16-bit signed normalized format that has a single 16-bit R component.
+     * <br>
+     * GL_R16_SNORM/GL_RED/GL_SHORT
+     */
+    R16_SNORM(2),
+    /**
+     * R16_UINT specifies a one-component, 16-bit unsigned integer format that has a single 16-bit R component.
+     * <br>
+     * GL_R16UI/GL_RED/GL_UNSIGNED_SHORT
+     */
+    R16_UINT(2),
+    /**
+     * R16_SINT specifies a one-component, 16-bit signed integer format that has a single 16-bit R component.
+     * <br>
+     * GL_R16I/GL_RED/GL_SHORT
+     */
+    R16_SINT(2),
+    /**
+     * R16_SFLOAT specifies a one-component, 16-bit signed floating-point format that has a single 16-bit R component.
+     * <br>
+     * GL_R16F/GL_RED/GL_HALF_FLOAT
+     */
+    R16_SFLOAT(2),
+    /**
+     * R16G16_UNORM specifies a two-component, 32-bit unsigned normalized format that has a 16-bit R component in bytes 0..1, and a 16-bit G component in bytes 2..3.
+     * <br>
+     * GL_RG16/GL_RG/GL_UNSIGNED_SHORT
+     */
+    R16G16_UNORM(4),
+    /**
+     * R16G16_SNORM specifies a two-component, 32-bit signed normalized format that has a 16-bit R component in bytes 0..1, and a 16-bit G component in bytes 2..3.
+     * <br>
+     * GL_RG16_SNORM/GL_RG/GL_SHORT
+     */
+    R16G16_SNORM(4),
+    /**
+     * R16G16_UINT specifies a two-component, 32-bit unsigned integer format that has a 16-bit R component in bytes 0..1, and a 16-bit G component in bytes 2..3.
+     * <br>
+     * GL_RG16UI/GL_RG/GL_UNSIGNED_SHORT
+     */
+    R16G16_UINT(4),
+    /**
+     * R16G16_SINT specifies a two-component, 32-bit signed integer format that has a 16-bit R component in bytes 0..1, and a 16-bit G component in bytes 2..3.
+     * <br>
+     * GL_RG16I/GL_RG/GL_SHORT
+     */
+    R16G16_SINT(4),
+    /**
+     * R16G16_SFLOAT specifies a two-component, 32-bit signed floating-point format that has a 16-bit R component in bytes 0..1, and a 16-bit G component in bytes 2..3.
+     * <br>
+     * GL_RG16F/GL_RG/GL_HALF_FLOAT
+     */
+    R16G16_SFLOAT(4),
+    /**
+     * R16G16B16A16_UNORM specifies a four-component, 64-bit unsigned normalized format that has a 16-bit R component in bytes 0..1, a 16-bit G component in bytes 2..3, a 16-bit B component in bytes 4..5, and a 16-bit A component in bytes 6..7.
+     * <br>
+     * GL_RGBA16/GL_RGBA/GL_UNSIGNED_SHORT
+     */
+    R16G16B16A16_UNORM(8),
+    /**
+     * R16G16B16A16_SNORM specifies a four-component, 64-bit signed normalized format that has a 16-bit R component in bytes 0..1, a 16-bit G component in bytes 2..3, a 16-bit B component in bytes 4..5, and a 16-bit A component in bytes 6..7.
+     * <br>
+     * GL_RGBA16_SNORM/GL_RGBA/GL_SHORT
+     */
+    R16G16B16A16_SNORM(8),
+    /**
+     * R16G16B16A16_UINT specifies a four-component, 64-bit unsigned integer format that has a 16-bit R component in bytes 0..1, a 16-bit G component in bytes 2..3, a 16-bit B component in bytes 4..5, and a 16-bit A component in bytes 6..7.
+     * <br>
+     * GL_RGBA16UI/GL_RGBA/GL_UNSIGNED_SHORT
+     */
+    R16G16B16A16_UINT(8),
+    /**
+     * R16G16B16A16_SINT specifies a four-component, 64-bit signed integer format that has a 16-bit R component in bytes 0..1, a 16-bit G component in bytes 2..3, a 16-bit B component in bytes 4..5, and a 16-bit A component in bytes 6..7.
+     * <br>
+     * GL_RGBA16I/GL_RGBA/GL_SHORT
+     */
+    R16G16B16A16_SINT(8),
+    /**
+     * R16G16B16A16_SFLOAT specifies a four-component, 64-bit signed floating-point format that has a 16-bit R component in bytes 0..1, a 16-bit G component in bytes 2..3, a 16-bit B component in bytes 4..5, and a 16-bit A component in bytes 6..7.
+     * <br>
+     * GL_RGBA16F/GL_RGBA/GL_HALF_FLOAT
+     */
+    R16G16B16A16_SFLOAT(8),
+
+    /**
+     * R32_UINT specifies a one-component, 32-bit unsigned integer format that has a single 32-bit R component.
+     * <br>
+     * GL_R32UI/GL_RED/GL_UNSIGNED_INT
+     */
+    R32_UINT(4),
+    /**
+     * R32_SINT specifies a one-component, 32-bit signed integer format that has a single 32-bit R component.
+     * <br>
+     * GL_R32I/GL_RED/GL_SIGNED_INT
+     */
+    R32_SINT(4),
+    /**
+     * R32_SFLOAT specifies a one-component, 32-bit signed floating-point format that has a single 32-bit R component.
+     * <br>
+     * GL_R32F/GL_RED/GL_FLOAT
+     */
+    R32_SFLOAT(4),
+    /**
+     * R32G32_UINT specifies a two-component, 64-bit unsigned integer format that has a 32-bit R component in bytes 0..3, and a 32-bit G component in bytes 4..7.
+     * <br>
+     * GL_RG32UI/GL_RG/GL_UNSIGNED_INT
+     */
+    R32G32_UINT(8),
+    /**
+     * R32G32_SINT specifies a two-component, 64-bit signed integer format that has a 32-bit R component in bytes 0..3, and a 32-bit G component in bytes 4..7.
+     * <br>
+     * GL_RG32I/GL_RG/GL_SIGNED_INT
+     */
+    R32G32_SINT(8),
+    /**
+     * R32G32_SFLOAT specifies a two-component, 64-bit signed floating-point format that has a 32-bit R component in bytes 0..3, and a 32-bit G component in bytes 4..7.
+     * <br>
+     * GL_RG32F/GL_RG/GL_FLOAT
+     */
+    R32G32_SFLOAT(8),
+    /**
+     * R32G32B32A32_UINT specifies a four-component, 128-bit unsigned integer format that has a 32-bit R component in bytes 0..3, a 32-bit G component in bytes 4..7, a 32-bit B component in bytes 8..11, and a 32-bit A component in bytes 12..15.
+     * <br>
+     * GL_RGBA32UI/GL_RGBA/GL_UNSIGNED_INT
+     */
+    R32G32B32A32_UINT(16),
+    /**
+     * R32G32B32A32_SINT specifies a four-component, 128-bit signed integer format that has a 32-bit R component in bytes 0..3, a 32-bit G component in bytes 4..7, a 32-bit B component in bytes 8..11, and a 32-bit A component in bytes 12..15.
+     * <br>
+     * GL_RGBA32I/GL_RGBA/GL_SIGNED_INT
+     */
+    R32G32B32A32_SINT(16),
+    /**
+     * R32G32B32A32_SFLOAT specifies a four-component, 128-bit signed floating-point format that has a 32-bit R component in bytes 0..3, a 32-bit G component in bytes 4..7, a 32-bit B component in bytes 8..11, and a 32-bit A component in bytes 12..15.
+     * <br>
+     * GL_RGBA32F/GL_RGBA/GL_FLOAT
+     */
+    R32G32B32A32_SFLOAT(16),
+
+    /**
+     * D32_SFLOAT specifies a one-component, 32-bit signed floating-point format that has 32 bits in the depth component.
+     * <br>
+     * GL_DEPTH_COMPONENT32F/GL_DEPTH_COMPONENT/GL_FLOAT
+     * <p>
+     * Neo: For compatibility reasons, the Vanilla name of DEPTH32 is kept. An alias constant {@link TextureFormat#D32_SFLOAT} is added for naming consistency.
+     * <p>
+     * Neo: Vanilla DEPTH32 uses the not-required by GL32C GL_DEPTH_COMPONENT32 internal format, Neo corrects to the required by GL32C GL_DEPTH_COMPONENT32F internal format.
+     */
+    DEPTH32(4),
+    /**
+     * D32_SFLOAT_S8_UINT specifies a two-component format that has 32 signed float bits in the depth component and 8 unsigned integer bits in the stencil component.
+     * <br>
+     * GL_DEPTH32F_STENCIL8/GL_DEPTH_STENCIL/GL_FLOAT_32_UNSIGNED_INT_24_8_REV
+     */
+    D32_SFLOAT_S8_UINT(8),
+    ;
+
+    /**
+     * @see TextureFormat#RED8
+     */
+    public static final TextureFormat R8_UNORM = RED8;
+    /**
+     * @see TextureFormat#RED8I
+     */
+    public static final TextureFormat R8_SINT = RED8I;
+    /**
+     * @see TextureFormat#RGBA8
+     */
+    public static final TextureFormat R8G8B8A8_UNORM = RGBA8;
+    /**
+     * @see TextureFormat#DEPTH32
+     */
+    public static final TextureFormat D32_SFLOAT = DEPTH32;
 
     private final int pixelSize;
 
@@ -23,10 +_,18 @@
     }
 
     public boolean hasColorAspect() {
-        return this == RGBA8 || this == RED8;
+        // Neo: all color formats do not have depth
+        return !hasDepthAspect();
     }
 
     public boolean hasDepthAspect() {
-        return this == DEPTH32;
+        return switch (this) {
+            case DEPTH32, D32_SFLOAT_S8_UINT -> true;
+            default -> false;
+        };
+    }
+
+    public boolean hasStencilAspect() {
+        return this == D32_SFLOAT_S8_UINT;
     }
 }
