--- a/com/mojang/blaze3d/textures/TextureFormat.java
+++ b/com/mojang/blaze3d/textures/TextureFormat.java
@@ -10,23 +_,152 @@
     RGBA8(4),
     RED8(1),
     RED8I(1),
-    DEPTH32(4);
+    DEPTH32(4),
+    // Neo: Difference notes on the above formats
+    //      Vanilla RED8I uses GL_UNSIGNED_BYTE for type, Neo corrects this to GL_BYTE
+    //      Vanilla DEPTH32 uses the not-required by GL32C GL_DEPTH_COMPONENT32 type, Neo corrects to the required by GL32C GL_DEPTH_COMPONENT32F type
+
+    // Neo: Extended texture formats, these follow VkFormat naming
+    // comments are internalformat/format/type, referencing glTexImage's parameters of the same name
+    // some may not always be supported or have limitations on specific usages, check backend capabilities before use
+
+    // pack8
+    R4G4_UNORM_PACK8(1), // Unsupported by OpenGL
+    // pack16
+    // 4 4 4 4
+    // Requires OpenGL 4.2, or equivalent
+    R4G4B4A4_UNORM_PACK16(2), // GL_RGBA4/GL_RGBA/GL_UNSIGNED_SHORT_4_4_4_4
+    // 5 6 5
+    // Requires OpenGL 4.2, or equivalent
+    R5G6B5_UNORM_PACK16(2), // GL_RGB565/GL_RGB/GL_UNSIGNED_SHORT_5_6_5
+    // 5 5 5 1
+    // Requires OpenGL 4.1, GL_ARB_ES2_compatibility, or equivalent
+    R5G5B5A1_UNORM_PACK16(2), // GL_RGB5_A1/GL_RGBA/GL_UNSIGNED_SHORT_5_5_5_1
+    // pack32
+    // 10 10 10 2
+    A2B10G10R10_UNORM_PACK32(4), // GL_RGB10_A2/GL_RGBA/GL_UNSIGNED_INT_2_10_10_10_REV
+    // 9 9 9 e5
+    E5B9G9R9_UFLOAT_PACK32(4), // GL_RGB9_E5/GL_RGB/GL_UNSIGNED_INT_5_9_9_9_REV
+
+    // R8
+    // RED8 is R8_UNORM(1), // GL_R8/GL_RED/GL_UNSIGNED_BYTE
+    R8_SNORM(1), // GL_R8_SNORM/GL_RED/GL_BYTE
+    R8_UINT(1), // GL_R8UI/GL_RED/GL_UNSIGNED_BYTE
+    // RED8I is R8_SINT(1), // GL_R8I/GL_RED/GL_BYTE
+    // RG8
+    R8G8_UNORM(2), // GL_RG8/GL_RG/GL_UNSIGNED_BYTE
+    R8G8_SNORM(2), // GL_RG8_SNORM/GL_RG/GL_BYTE
+    R8G8_UINT(2), // GL_RG8UI/GL_RG/GL_UNSIGNED_BYTE
+    R8G8_SINT(2), // GL_RG8I/GL_RG/GL_BYTE
+    // RGBA8
+    // RGBA8 is R8G8B8A8_UNORM(4), // GL_RGBA8/GL_RGBA/GL_UNSIGNED_BYTE
+    R8G8B8A8_SNORM(4), // GL_RGBA8_SNORM/GL_RGBA/GL_BYTE
+    R8G8B8A8_UINT(4), // GL_RGBA8UI/GL_RGBA/GL_UNSIGNED_BYTE
+    R8G8B8A8_SINT(4), // GL_RGBA8I/GL_RGBA/GL_BYTE
+
+    // R16
+    R16_UNORM(2), // GL_R16/GL_RED/GL_UNSIGNED_SHORT
+    R16_SNORM(2), // GL_R16_SNORM/GL_RED/GL_SHORT
+    R16_UINT(2), // GL_R16UI/GL_RED/GL_UNSIGNED_SHORT
+    R16_SINT(2), // GL_R16I/GL_RED/GL_SHORT
+    R16_SFLOAT(2), //GL_R16F/GL_RED/GL_HALF_FLOAT
+    // RG16
+    R16G16_UNORM(4), // GL_RG16/GL_RG/GL_UNSIGNED_SHORT
+    R16G16_SNORM(4), // GL_RG16_SNORM/GL_RG/GL_SHORT
+    R16G16_UINT(4), // GL_RG16UI/GL_RG/GL_UNSIGNED_SHORT
+    R16G16_SINT(4), // GL_RG16I/GL_RG/GL_SHORT
+    R16G16_SFLOAT(4), //GL_RG16F/GL_RG/GL_HALF_FLOAT
+    // RGBA16
+    R16G16B16A16_UNORM(8), // GL_RGBA16/GL_RGBA/GL_UNSIGNED_SHORT
+    R16G16B16A16_SNORM(8), // GL_RGBA16_SNORM/GL_RGBA/GL_SHORT
+    R16G16B16A16_UINT(8), // GL_RGBA16UI/GL_RGBA/GL_UNSIGNED_SHORT
+    R16G16B16A16_SINT(8), // GL_RGBA16I/GL_RGBA/GL_SHORT
+    R16G16B16A16_SFLOAT(8), //GL_RGBA16F/GL_RGBA/GL_HALF_FLOAT
+
+    // R32
+    R32_UINT(4), // GL_R32UI/GL_RED/GL_UNSIGNED_INT
+    R32_SINT(4), // GL_R32I/GL_RED/GL_SIGNED_INT
+    R32_SFLOAT(4), // GL_R32F/GL_RED/GL_FLOAT
+    // RG32
+    R32G32_UINT(8), // GL_RG32UI/GL_RG/GL_UNSIGNED_INT
+    R32G32_SINT(8), // GL_RG32I/GL_RG/GL_SIGNED_INT
+    R32G32_SFLOAT(8), // GL_RG32F/GL_RG/GL_FLOAT
+    // RGBA32
+    R32G32B32A32_UINT(16), // GL_RGBA32UI/GL_RGBA/GL_UNSIGNED_INT
+    R32G32B32A32_SINT(16), // GL_RGBA32I/GL_RGBA/GL_SIGNED_INT
+    R32G32B32A32_SFLOAT(16), // GL_RGBA32F/GL_RGBA/GL_FLOAT
+
+    // Depth
+    D16_UNORM(2), // GL_DEPTH_COMPONENT16/GL_DEPTH_COMPONENT/GL_UNSIGNED_SHORT
+    X8_D24_UNORM_PACK32(4), // GL_DEPTH_COMPONENT24/GL_DEPTH_COMPONENT/GL_UNSIGNED_INT_24_8
+    // DEPTH32 is D32_SFLOAT(4), // GL_DEPTH_COMPONENT32F/GL_DEPTH_COMPONENT/GL_FLOAT
+
+    // Stencil
+    // Requires OpenGL 4.3, GL_ARB_ES3_compatibility, or equivalent (render buffer)
+    // Requires OpenGL 4.4, GL_ARB_texture_stencil8, or equivalent (texture)
+    S8_UINT(1), // GL_STENCIL_INDEX8/GL_STENCIL_INDEX/GL_UNSIGNED_BYTE
+
+    // Depth/Stencil
+    D24_UNORM_S8_UINT(4), // GL_DEPTH24_STENCIL8/GL_DEPTH_STENCIL/GL_UNSIGNED_INT_24_8
+    D32_SFLOAT_S8_UINT(8), // GL_DEPTH32F_STENCIL8/GL_DEPTH_STENCIL/GL_FLOAT_32_UNSIGNED_INT_24_8_REV
+
+    // Compressed
+    // internal and external format are identical for these, and there is no type
+    // Requires GL_EXT_texture_compression_s3tc, or equivalent
+    BC1_RGB_UNORM_BLOCK(0, 1), // GL_COMPRESSED_RGB_S3TC_DXT1_EXT
+    BC1_RGB_SRGB_BLOCK(0, 1), // GL_COMPRESSED_SRGB_S3TC_DXT1_EXT
+    BC1_RGBA_UNORM_BLOCK(0, 1), // GL_COMPRESSED_RGBA_S3TC_DXT1_EXT
+    BC1_RGBA_SRGB_BLOCK(0, 1), // GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT1_EXT
+    BC2_UNORM_BLOCK(1), // GL_COMPRESSED_RGBA_S3TC_DXT3_EXT
+    BC2_SRGB_BLOCK(1), // GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT3_EXT
+    BC3_UNORM_BLOCK(1), // GL_COMPRESSED_RGBA_S3TC_DXT5_EXT
+    BC3_SRGB_BLOCK(1), // GL_COMPRESSED_SRGB_ALPHA_S3TC_DXT5_EXT
+    ;
+
+    // VkFormat based name alias for Vanilla defined enums
+    public static final TextureFormat R8_UNORM = RED8; // GL_R8/GL_RED/GL_UNSIGNED_BYTE
+    public static final TextureFormat R8_SINT = RED8I; // GL_R8I/GL_RED/GL_BYTE
+    public static final TextureFormat R8G8B8A8_UNORM = RGBA8; // GL_RGBA8/GL_RGBA/GL_UNSIGNED_BYTE
+    public static final TextureFormat D32_SFLOAT = DEPTH32; // GL_DEPTH_COMPONENT32F/GL_DEPTH_COMPONENT/GL_FLOAT
 
     private final int pixelSize;
+    // Neo: some formats (BC1) are less than a single byte per-pixel
+    private final int pixelSizeNibbles;
 
     private TextureFormat(int p_404990_) {
         this.pixelSize = p_404990_;
+        this.pixelSizeNibbles = p_404990_ * 2;
+    }
+
+    private TextureFormat(int pixelSize, int pixelSizeNibbles) {
+        this.pixelSize = pixelSize;
+        this.pixelSizeNibbles = (pixelSize * 2) + pixelSizeNibbles;
     }
 
     public int pixelSize() {
         return this.pixelSize;
     }
 
+    public int pixelSizeNibbles() {
+        return this.pixelSizeNibbles;
+    }
+
     public boolean hasColorAspect() {
-        return this == RGBA8 || this == RED8;
+        // Neo: all color formats do not have depth and do not have stencil
+        return !hasDepthAspect() && !hasStencilAspect();
     }
 
     public boolean hasDepthAspect() {
-        return this == DEPTH32;
+        return switch (this) {
+            case D16_UNORM, X8_D24_UNORM_PACK32, DEPTH32, D24_UNORM_S8_UINT, D32_SFLOAT_S8_UINT -> true;
+            default -> false;
+        };
+    }
+
+    public boolean hasStencilAspect() {
+        return switch (this) {
+            case S8_UINT, D24_UNORM_S8_UINT, D32_SFLOAT_S8_UINT -> true;
+            default -> false;
+        };
     }
 }
