--- a/net/minecraft/commands/arguments/MessageArgument.java
+++ b/net/minecraft/commands/arguments/MessageArgument.java
@@ -90,7 +_,7 @@
 
     public record Message(String text, MessageArgument.Part[] parts) {
         private Component resolveComponent(CommandSourceStack sender) throws CommandSyntaxException {
-            return this.toComponent(sender, sender.permissions().hasPermission(Permissions.COMMANDS_ENTITY_SELECTORS));
+            return this.toComponent(sender, net.neoforged.neoforge.common.CommonHooks.canUseEntitySelectors(sender));
         }
 
         public Component toComponent(CommandSourceStack sender, boolean allowSelectors) throws CommandSyntaxException {
