--- a/net/minecraft/commands/arguments/MessageArgument.java
+++ b/net/minecraft/commands/arguments/MessageArgument.java
@@ -84,7 +_,7 @@
 
     public static record Message(String text, MessageArgument.Part[] parts) {
         Component resolveComponent(CommandSourceStack source) throws CommandSyntaxException {
-            return this.toComponent(source, source.hasPermission(2));
+            return this.toComponent(source, net.neoforged.neoforge.common.CommonHooks.canUseEntitySelectors(source));
         }
 
         public Component toComponent(CommandSourceStack source, boolean allowSelectors) throws CommandSyntaxException {
