--- a/net/minecraft/commands/arguments/ResourceLocationArgument.java
+++ b/net/minecraft/commands/arguments/ResourceLocationArgument.java
@@ -29,7 +_,7 @@
 
     public static AdvancementHolder getAdvancement(CommandContext<CommandSourceStack> context, String name) throws CommandSyntaxException {
         ResourceLocation resourcelocation = getId(context, name);
-        AdvancementHolder advancementholder = context.getSource().getServer().getAdvancements().get(resourcelocation);
+        AdvancementHolder advancementholder = context.getSource().getAdvancement(resourcelocation);
         if (advancementholder == null) {
             throw ERROR_UNKNOWN_ADVANCEMENT.create(resourcelocation);
         } else {
@@ -38,7 +_,7 @@
     }
 
     public static RecipeHolder<?> getRecipe(CommandContext<CommandSourceStack> context, String name) throws CommandSyntaxException {
-        RecipeManager recipemanager = context.getSource().getServer().getRecipeManager();
+        RecipeManager recipemanager = context.getSource().getRecipeManager();
         ResourceLocation resourcelocation = getId(context, name);
         return recipemanager.byKey(resourcelocation).orElseThrow(() -> ERROR_UNKNOWN_RECIPE.create(resourcelocation));
     }
