--- a/net/minecraft/commands/arguments/TeamArgument.java
+++ b/net/minecraft/commands/arguments/TeamArgument.java
@@ -28,7 +_,7 @@
 
     public static PlayerTeam getTeam(CommandContext<CommandSourceStack> context, String name) throws CommandSyntaxException {
         String s = context.getArgument(name, String.class);
-        Scoreboard scoreboard = context.getSource().getServer().getScoreboard();
+        Scoreboard scoreboard = context.getSource().getScoreboard();
         PlayerTeam playerteam = scoreboard.getPlayerTeam(s);
         if (playerteam == null) {
             throw ERROR_TEAM_NOT_FOUND.create(s);
