--- a/net/minecraft/client/renderer/entity/BoatRenderer.java
+++ b/net/minecraft/client/renderer/entity/BoatRenderer.java
@@ -77,7 +_,7 @@
             poseStack.mulPose(new Quaternionf().setAngleAxis(entity.getBubbleAngle(partialTicks) * (float) (Math.PI / 180.0), 1.0F, 0.0F, 1.0F));
         }
 
-        Pair<ResourceLocation, ListModel<Boat>> pair = this.boatResources.get(entity.getVariant());
+        Pair<ResourceLocation, ListModel<Boat>> pair = getModelWithLocation(entity);
         ResourceLocation resourcelocation = pair.getFirst();
         ListModel<Boat> listmodel = pair.getSecond();
         poseStack.scale(-1.0F, -1.0F, 1.0F);
@@ -96,7 +_,10 @@
         super.render(entity, entityYaw, partialTicks, poseStack, buffer, packedLight);
     }
 
+    @Deprecated // forge: override getModelWithLocation to change the texture / model
     public ResourceLocation getTextureLocation(Boat entity) {
-        return this.boatResources.get(entity.getVariant()).getFirst();
+        return getModelWithLocation(entity).getFirst();
     }
+
+    public Pair<ResourceLocation, ListModel<Boat>> getModelWithLocation(Boat boat) { return this.boatResources.get(boat.getVariant()); }
 }
