--- a/net/minecraft/client/renderer/entity/BoatRenderer.java
+++ b/net/minecraft/client/renderer/entity/BoatRenderer.java
@@ -45,7 +_,7 @@
     private ListModel<Boat> createBoatModel(EntityRendererProvider.Context p_248834_, Boat.Type p_249317_, boolean p_250093_) {
         ModelLayerLocation modellayerlocation = p_250093_ ? ModelLayers.createChestBoatModelName(p_249317_) : ModelLayers.createBoatModelName(p_249317_);
         ModelPart modelpart = p_248834_.bakeLayer(modellayerlocation);
-        if (p_249317_ == Boat.Type.BAMBOO) {
+        if (p_249317_.isRaft()) {
             return (ListModel<Boat>)(p_250093_ ? new ChestRaftModel(modelpart) : new RaftModel(modelpart));
         } else {
             return (ListModel<Boat>)(p_250093_ ? new ChestBoatModel(modelpart) : new BoatModel(modelpart));
@@ -54,8 +_,8 @@
 
     private static ResourceLocation getTextureLocation(Boat.Type p_234566_, boolean p_234567_) {
         return p_234567_
-            ? ResourceLocation.withDefaultNamespace("textures/entity/chest_boat/" + p_234566_.getName() + ".png")
-            : ResourceLocation.withDefaultNamespace("textures/entity/boat/" + p_234566_.getName() + ".png");
+            ? ResourceLocation.parse(p_234566_.getName()).withPrefix("textures/entity/chest_boat/").withSuffix(".png")
+            : ResourceLocation.parse(p_234566_.getName()).withPrefix("textures/entity/boat/").withSuffix(".png");
     }
 
     public void render(Boat p_113929_, float p_113930_, float p_113931_, PoseStack p_113932_, MultiBufferSource p_113933_, int p_113934_) {
@@ -77,7 +_,7 @@
             p_113932_.mulPose(new Quaternionf().setAngleAxis(p_113929_.getBubbleAngle(p_113931_) * (float) (Math.PI / 180.0), 1.0F, 0.0F, 1.0F));
         }
 
-        Pair<ResourceLocation, ListModel<Boat>> pair = this.boatResources.get(p_113929_.getVariant());
+        Pair<ResourceLocation, ListModel<Boat>> pair = getModelWithLocation(p_113929_);
         ResourceLocation resourcelocation = pair.getFirst();
         ListModel<Boat> listmodel = pair.getSecond();
         p_113932_.scale(-1.0F, -1.0F, 1.0F);
@@ -96,7 +_,12 @@
         super.render(p_113929_, p_113930_, p_113931_, p_113932_, p_113933_, p_113934_);
     }
 
+    @Deprecated // forge: override getModelWithLocation to change the texture / model
     public ResourceLocation getTextureLocation(Boat p_113927_) {
-        return this.boatResources.get(p_113927_.getVariant()).getFirst();
+        return getModelWithLocation(p_113927_).getFirst();
+    }
+
+    public Pair<ResourceLocation, ListModel<Boat>> getModelWithLocation(Boat boat) {
+        return this.boatResources.get(boat.getVariant());
     }
 }
