--- a/net/minecraft/client/renderer/entity/EntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderer.java
@@ -105,6 +_,8 @@
         }
 
         if (p_364816_.nameTag != null) {
+            var event = new net.neoforged.neoforge.client.event.RenderNameTagEvent.DoRender(p_364816_, p_364816_.nameTag, this, p_114488_, p_114489_, p_114490_, p_364816_.partialTick);
+            if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled())
             this.renderNameTag(p_364816_, p_364816_.nameTag, p_114488_, p_114489_, p_114490_);
         }
     }
@@ -270,7 +_,12 @@
         }
 
         p_361028_.distanceToCameraSq = this.entityRenderDispatcher.distanceToSqr(p_362104_);
-        boolean flag = p_361028_.distanceToCameraSq < 4096.0 && this.shouldShowName(p_362104_, p_361028_.distanceToCameraSq);
+        boolean flag = p_361028_.distanceToCameraSq < 4096.0;
+        if (flag) {
+            var event = new net.neoforged.neoforge.client.event.RenderNameTagEvent.CanRender(p_362104_, p_361028_, p_362104_.getDisplayName(), this, p_362204_);
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event);
+            flag = event.canRender().isTrue() || (event.canRender().isDefault() && this.shouldShowName(p_362104_, p_361028_.distanceToCameraSq));
+        }
         if (flag) {
             p_361028_.nameTag = this.getNameTag(p_362104_);
             p_361028_.nameTagAttachment = p_362104_.getAttachments().getNullable(EntityAttachment.NAME_TAG, 0, p_362104_.getYRot(p_362204_));
@@ -302,5 +_,7 @@
         }
 
         p_361028_.displayFireAnimation = p_362104_.displayFireAnimation();
+
+        p_361028_.partialTick = p_362204_;
     }
 }
