--- a/net/minecraft/client/renderer/entity/EntityRenderDispatcher.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderDispatcher.java
@@ -222,12 +_,12 @@
     ) {
         AABB aabb = p_entity.getBoundingBox().move(-p_entity.getX(), -p_entity.getY(), -p_entity.getZ());
         LevelRenderer.renderLineBox(poseStack, buffer, aabb, green, blue, alpha, 1.0F);
-        if (p_entity instanceof EnderDragon) {
+        if (p_entity.isMultipartEntity()) {
             double d0 = -Mth.lerp((double)red, p_entity.xOld, p_entity.getX());
             double d1 = -Mth.lerp((double)red, p_entity.yOld, p_entity.getY());
             double d2 = -Mth.lerp((double)red, p_entity.zOld, p_entity.getZ());
 
-            for (EnderDragonPart enderdragonpart : ((EnderDragon)p_entity).getSubEntities()) {
+            for (net.neoforged.neoforge.entity.PartEntity<?> enderdragonpart : p_entity.getParts()) {
                 poseStack.pushPose();
                 double d3 = d0 + Mth.lerp((double)red, enderdragonpart.xOld, enderdragonpart.getX());
                 double d4 = d1 + Mth.lerp((double)red, enderdragonpart.yOld, enderdragonpart.getY());
@@ -467,6 +_,10 @@
         return this.itemInHandRenderer;
     }
 
+    public Map<PlayerSkin.Model, EntityRenderer<? extends Player>> getSkinMap() {
+        return java.util.Collections.unmodifiableMap(playerRenderers);
+    }
+
     @Override
     public void onResourceManagerReload(ResourceManager resourceManager) {
         EntityRendererProvider.Context entityrendererprovider$context = new EntityRendererProvider.Context(
@@ -474,5 +_,6 @@
         );
         this.renderers = EntityRenderers.createEntityRenderers(entityrendererprovider$context);
         this.playerRenderers = EntityRenderers.createPlayerRenderers(entityrendererprovider$context);
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.EntityRenderersEvent.AddLayers(renderers, playerRenderers, entityrendererprovider$context));
     }
 }
