--- a/net/minecraft/client/renderer/entity/EntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderer.java
@@ -132,6 +_,8 @@
 
     protected void submitNameTag(S p_436016_, PoseStack p_433473_, SubmitNodeCollector p_434265_, CameraRenderState p_451278_) {
         if (p_436016_.nameTag != null) {
+            var event = new net.neoforged.neoforge.client.event.RenderNameTagEvent.DoRender(p_436016_, p_436016_.nameTag, this, p_433473_, p_434265_, p_436016_.partialTick);
+            if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled())
             p_434265_.submitNameTag(
                 p_433473_,
                 p_436016_.nameTagAttachment,
@@ -164,6 +_,7 @@
         S s = this.createRenderState();
         this.extractRenderState(p_361382_, s, p_360885_);
         this.finalizeRenderState(p_361382_, s);
+        net.neoforged.neoforge.client.renderstate.RenderStateExtensions.onUpdateEntityRenderState(this, p_361382_, s);
         return s;
     }
 
