--- a/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
+++ b/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
@@ -40,7 +_,7 @@
             p_114637_.translate(-0.5, 0.0, -0.5);
             List<BlockModelPart> list = this.dispatcher
                 .getBlockModel(blockstate)
-                .collectParts(RandomSource.create(blockstate.getSeed(p_361300_.startBlockPos)));
+                .collectParts(p_361300_.level, p_361300_.blockPos, blockstate, RandomSource.create(blockstate.getSeed(p_361300_.startBlockPos)));
             this.dispatcher
                 .getModelRenderer()
                 .tesselateBlock(
@@ -49,7 +_,7 @@
                     blockstate,
                     p_361300_.blockPos,
                     p_114637_,
-                    p_114638_.getBuffer(ItemBlockRenderTypes.getMovingBlockRenderType(blockstate)),
+                    renderType -> p_114638_.getBuffer(net.neoforged.neoforge.client.RenderTypeHelper.getMovingBlockRenderType(renderType)),
                     false,
                     OverlayTexture.NO_OVERLAY
                 );
