--- a/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
+++ b/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
@@ -34,6 +_,8 @@
                 poseStack.pushPose();
                 BlockPos blockpos = BlockPos.containing(entity.getX(), entity.getBoundingBox().maxY, entity.getZ());
                 poseStack.translate(-0.5, 0.0, -0.5);
+                var model = this.dispatcher.getBlockModel(blockstate);
+                for (var renderType : model.getRenderTypes(blockstate, RandomSource.create(blockstate.getSeed(entity.getStartPos())), net.neoforged.neoforge.client.model.data.ModelData.EMPTY))
                 this.dispatcher
                     .getModelRenderer()
                     .tesselateBlock(
@@ -42,11 +_,13 @@
                         blockstate,
                         blockpos,
                         poseStack,
-                        buffer.getBuffer(ItemBlockRenderTypes.getMovingBlockRenderType(blockstate)),
+                        buffer.getBuffer(net.neoforged.neoforge.client.RenderTypeHelper.getMovingBlockRenderType(renderType)),
                         false,
                         RandomSource.create(),
                         blockstate.getSeed(entity.getStartPos()),
-                        OverlayTexture.NO_OVERLAY
+                        OverlayTexture.NO_OVERLAY,
+                        net.neoforged.neoforge.client.model.data.ModelData.EMPTY,
+                        renderType
                     );
                 poseStack.popPose();
                 super.render(entity, entityYaw, partialTicks, poseStack, buffer, packedLight);
