--- a/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
+++ b/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
@@ -36,6 +_,8 @@
         if (blockstate.getRenderShape() == RenderShape.MODEL) {
             p_114637_.pushPose();
             p_114637_.translate(-0.5, 0.0, -0.5);
+            var model = this.dispatcher.getBlockModel(blockstate);
+            for (var renderType : model.getRenderTypes(blockstate, RandomSource.create(blockstate.getSeed(p_361300_.startBlockPos)), net.neoforged.neoforge.client.model.data.ModelData.EMPTY))
             this.dispatcher
                 .getModelRenderer()
                 .tesselateBlock(
@@ -44,11 +_,13 @@
                     blockstate,
                     p_361300_.blockPos,
                     p_114637_,
-                    p_114638_.getBuffer(ItemBlockRenderTypes.getMovingBlockRenderType(blockstate)),
+                    p_114638_.getBuffer(net.neoforged.neoforge.client.RenderTypeHelper.getMovingBlockRenderType(renderType)),
                     false,
                     RandomSource.create(),
                     blockstate.getSeed(p_361300_.startBlockPos),
-                    OverlayTexture.NO_OVERLAY
+                    OverlayTexture.NO_OVERLAY,
+                    net.neoforged.neoforge.client.model.data.ModelData.EMPTY,
+                    renderType
                 );
             p_114637_.popPose();
             super.render(p_361300_, p_114637_, p_114638_, p_114639_);
