--- a/net/minecraft/client/renderer/entity/FishingHookRenderer.java
+++ b/net/minecraft/client/renderer/entity/FishingHookRenderer.java
@@ -65,7 +_,7 @@
     }
 
     public static HumanoidArm getHoldingArm(Player owner) {
-        return owner.getMainHandItem().getItem() instanceof FishingRodItem ? owner.getMainArm() : owner.getMainArm().getOpposite();
+        return owner.getMainHandItem().canPerformAction(net.neoforged.neoforge.common.ItemAbilities.FISHING_ROD_CAST) ? owner.getMainArm() : owner.getMainArm().getOpposite();
     }
 
     private Vec3 getPlayerHandPos(Player owner, float swing, float partialTicks) {
