--- a/net/minecraft/client/renderer/entity/HumanoidMobRenderer.java
+++ b/net/minecraft/client/renderer/entity/HumanoidMobRenderer.java
@@ -61,7 +_,7 @@
         state.isCrouching = entity.isCrouching();
         state.isFallFlying = entity.isFallFlying();
         state.isVisuallySwimming = entity.isVisuallySwimming();
-        state.isPassenger = entity.isPassenger();
+        state.isPassenger = entity.isPassenger() && (entity.getVehicle() != null && entity.getVehicle().shouldRiderSit());
         state.speedValue = 1.0F;
         if (state.isFallFlying) {
             state.speedValue = (float)entity.getDeltaMovement().lengthSqr();
