--- a/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
@@ -77,6 +_,7 @@
             poseStack.translate(0.0F, 0.0F, 0.4375F);
         }
 
+        if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderItemInFrameEvent(state, this, poseStack, submitNodeCollector)).isCanceled()) {
         if (state.mapId != null) {
             int rotation = state.rotation % 4 * 2;
             poseStack.mulPose(Axis.ZP.rotationDegrees(rotation * 360.0F / 8.0F));
@@ -93,6 +_,7 @@
             poseStack.scale(0.5F, 0.5F, 0.5F);
             state.item.submit(poseStack, submitNodeCollector, lightVal, OverlayTexture.NO_OVERLAY, state.outlineColor);
         }
+        }
 
         poseStack.popPose();
     }
@@ -128,7 +_,7 @@
         if (!itemStack.isEmpty()) {
             MapId framedMapId = entity.getFramedMapId(itemStack);
             if (framedMapId != null) {
-                MapItemSavedData mapData = entity.level().getMapData(framedMapId);
+                MapItemSavedData mapData = net.minecraft.world.item.MapItem.getSavedData(itemStack, entity.level());
                 if (mapData != null) {
                     this.mapRenderer.extractRenderState(framedMapId, mapData, state.mapRenderState);
                     state.mapId = framedMapId;
