--- a/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemFrameRenderer.java
@@ -84,25 +_,25 @@
         }
 
         if (!itemstack.isEmpty()) {
-            MapId mapid = entity.getFramedMapId(itemstack);
+            MapItemSavedData mapitemsaveddata = MapItem.getSavedData(itemstack, entity.level());
             if (flag) {
                 poseStack.translate(0.0F, 0.0F, 0.5F);
             } else {
                 poseStack.translate(0.0F, 0.0F, 0.4375F);
             }
 
-            int j = mapid != null ? entity.getRotation() % 4 * 2 : entity.getRotation();
+            int j = mapitemsaveddata != null ? entity.getRotation() % 4 * 2 : entity.getRotation();
             poseStack.mulPose(Axis.ZP.rotationDegrees((float)j * 360.0F / 8.0F));
-            if (mapid != null) {
+            if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderItemInFrameEvent(entity, this, poseStack, buffer, packedLight)).isCanceled()) {
+            if (mapitemsaveddata != null) {
                 poseStack.mulPose(Axis.ZP.rotationDegrees(180.0F));
                 float f = 0.0078125F;
                 poseStack.scale(0.0078125F, 0.0078125F, 0.0078125F);
                 poseStack.translate(-64.0F, -64.0F, 0.0F);
-                MapItemSavedData mapitemsaveddata = MapItem.getSavedData(mapid, entity.level());
                 poseStack.translate(0.0F, 0.0F, -1.0F);
                 if (mapitemsaveddata != null) {
                     int i = this.getLightVal(entity, 15728850, packedLight);
-                    Minecraft.getInstance().gameRenderer.getMapRenderer().render(poseStack, buffer, mapid, mapitemsaveddata, true, i);
+                    Minecraft.getInstance().gameRenderer.getMapRenderer().render(poseStack, buffer, entity.getFramedMapId(itemstack), mapitemsaveddata, true, i);
                 }
             } else {
                 int k = this.getLightVal(entity, 15728880, packedLight);
@@ -110,6 +_,7 @@
                 this.itemRenderer
                     .renderStatic(itemstack, ItemDisplayContext.FIXED, k, OverlayTexture.NO_OVERLAY, poseStack, buffer, entity.level(), entity.getId());
             }
+            }
         }
 
         poseStack.popPose();
@@ -121,7 +_,7 @@
 
     private ModelResourceLocation getFrameModelResourceLoc(T entity, ItemStack item) {
         boolean flag = entity.getType() == EntityType.GLOW_ITEM_FRAME;
-        if (item.is(Items.FILLED_MAP)) {
+        if (item.getItem() instanceof MapItem) {
             return flag ? GLOW_MAP_FRAME_LOCATION : MAP_FRAME_LOCATION;
         } else {
             return flag ? GLOW_FRAME_LOCATION : FRAME_LOCATION;
