--- a/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
@@ -38,6 +_,8 @@
     public void extractRenderState(ItemEntity entity, ItemEntityRenderState state, float partialTicks) {
         super.extractRenderState(entity, state, partialTicks);
         state.bobOffset = entity.bobOffs;
+        state.shouldBob = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(entity.getItem()).shouldBobAsEntity(entity.getItem());
+        state.shouldSpread = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(entity.getItem()).shouldSpreadAsEntity(entity.getItem());
         state.extractItemGroupRenderState(entity, entity.getItem(), this.itemModelResolver);
     }
 
@@ -46,7 +_,7 @@
             poseStack.pushPose();
             AABB boundingBox = state.item.getModelBoundingBox();
             float minOffsetY = -((float)boundingBox.minY) + 0.0625F;
-            float bob = Mth.sin(state.ageInTicks / 10.0F + state.bobOffset) * 0.1F + 0.1F;
+            float bob = state.shouldBob ? Mth.sin(state.ageInTicks / 10.0F + state.bobOffset) * 0.1F + 0.1F : 0;
             poseStack.translate(0.0F, bob + minOffsetY, 0.0F);
             float spin = ItemEntity.getSpin(state.ageInTicks, state.bobOffset);
             poseStack.mulPose(Axis.YP.rotation(spin));
@@ -78,7 +_,9 @@
                     float xo = (random.nextFloat() * 2.0F - 1.0F) * 0.15F;
                     float yo = (random.nextFloat() * 2.0F - 1.0F) * 0.15F;
                     float zo = (random.nextFloat() * 2.0F - 1.0F) * 0.15F;
-                    poseStack.translate(xo, yo, zo);
+                    if (state.shouldSpread) {
+                        poseStack.translate(xo, yo, zo);
+                    }
                     item.submit(poseStack, submitNodeCollector, lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
                     poseStack.popPose();
                 }
@@ -92,7 +_,9 @@
                     poseStack.pushPose();
                     float xo = (random.nextFloat() * 2.0F - 1.0F) * 0.15F * 0.5F;
                     float yo = (random.nextFloat() * 2.0F - 1.0F) * 0.15F * 0.5F;
-                    poseStack.translate(xo, yo, 0.0F);
+                    if (state.shouldSpread) {
+                        poseStack.translate(xo, yo, 0.0F);
+                    }
                     item.submit(poseStack, submitNodeCollector, lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
                     poseStack.popPose();
                     poseStack.translate(0.0F, 0.0F, offsetZ);
