--- a/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
@@ -38,6 +_,8 @@
     public void extractRenderState(ItemEntity p_362393_, ItemEntityRenderState p_361441_, float p_360849_) {
         super.extractRenderState(p_362393_, p_361441_, p_360849_);
         p_361441_.bobOffset = p_362393_.bobOffs;
+        p_361441_.shouldBob = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(p_362393_.getItem()).shouldBobAsEntity(p_362393_.getItem());
+        p_361441_.shouldSpread = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(p_362393_.getItem()).shouldSpreadAsEntity(p_362393_.getItem());
         p_361441_.extractItemGroupRenderState(p_362393_, p_362393_.getItem(), this.itemModelResolver);
     }
 
@@ -46,7 +_,7 @@
             p_435012_.pushPose();
             AABB aabb = p_435798_.item.getModelBoundingBox();
             float f = -((float)aabb.minY) + 0.0625F;
-            float f1 = Mth.sin(p_435798_.ageInTicks / 10.0F + p_435798_.bobOffset) * 0.1F + 0.1F;
+            float f1 = p_435798_.shouldBob ? Mth.sin(p_435798_.ageInTicks / 10.0F + p_435798_.bobOffset) * 0.1F + 0.1F : 0;
             p_435012_.translate(0.0F, f1 + f, 0.0F);
             float f2 = ItemEntity.getSpin(p_435798_.ageInTicks, p_435798_.bobOffset);
             p_435012_.mulPose(Axis.YP.rotation(f2));
@@ -78,7 +_,9 @@
                     float f1 = (p_404894_.nextFloat() * 2.0F - 1.0F) * 0.15F;
                     float f2 = (p_404894_.nextFloat() * 2.0F - 1.0F) * 0.15F;
                     float f3 = (p_404894_.nextFloat() * 2.0F - 1.0F) * 0.15F;
-                    p_404890_.translate(f1, f2, f3);
+                    if (p_405697_.shouldSpread) {
+                        p_404890_.translate(f1, f2, f3);
+                    }
                     itemstackrenderstate.submit(p_404890_, p_433429_, p_404762_, OverlayTexture.NO_OVERLAY, p_405697_.outlineColor);
                     p_404890_.popPose();
                 }
@@ -92,7 +_,9 @@
                     p_404890_.pushPose();
                     float f5 = (p_404894_.nextFloat() * 2.0F - 1.0F) * 0.15F * 0.5F;
                     float f6 = (p_404894_.nextFloat() * 2.0F - 1.0F) * 0.15F * 0.5F;
-                    p_404890_.translate(f5, f6, 0.0F);
+                    if (p_405697_.shouldSpread) {
+                        p_404890_.translate(f5, f6, 0.0F);
+                    }
                     itemstackrenderstate.submit(p_404890_, p_433429_, p_404762_, OverlayTexture.NO_OVERLAY, p_405697_.outlineColor);
                     p_404890_.popPose();
                     p_404890_.translate(0.0F, 0.0F, f4);
