--- a/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
@@ -45,7 +_,8 @@
         BakedModel bakedmodel = this.itemRenderer.getModel(itemstack, entity.level(), null, entity.getId());
         boolean flag = bakedmodel.isGui3d();
         float f = 0.25F;
-        float f1 = Mth.sin(((float)entity.getAge() + partialTicks) / 10.0F + entity.bobOffs) * 0.1F + 0.1F;
+        boolean shouldBob = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(itemstack).shouldBobAsEntity(itemstack);
+        float f1 = shouldBob ? Mth.sin(((float)entity.getAge() + partialTicks) / 10.0F + entity.bobOffs) * 0.1F + 0.1F : 0;
         float f2 = bakedmodel.getTransforms().getTransform(ItemDisplayContext.GROUND).scale.y();
         poseStack.translate(0.0F, f1 + 0.25F * f2, 0.0F);
         float f3 = entity.getSpin(partialTicks);
@@ -100,9 +_,10 @@
             poseStack.translate(f3, f4, f5);
         }
 
+        boolean shouldSpread = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(item).shouldSpreadAsEntity(item);
         for (int j = 0; j < i; j++) {
             poseStack.pushPose();
-            if (j > 0) {
+            if (j > 0 && shouldSpread) {
                 if (isGui3d) {
                     float f7 = (random.nextFloat() * 2.0F - 1.0F) * 0.15F;
                     float f9 = (random.nextFloat() * 2.0F - 1.0F) * 0.15F;
