--- a/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
@@ -73,6 +_,7 @@
     }
 
     public void submit(S state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
+        if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLivingEvent.Pre<T, S, M>(state, this, state.partialTick, poseStack, submitNodeCollector)).isCanceled()) return;
         poseStack.pushPose();
         if (state.hasPose(Pose.SLEEPING)) {
             Direction bedOrientation = state.bedOrientation;
@@ -110,6 +_,7 @@
 
         poseStack.popPose();
         super.submit(state, poseStack, submitNodeCollector, camera);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLivingEvent.Post<T, S, M>(state, this, state.partialTick, poseStack, submitNodeCollector));
     }
 
     protected boolean shouldRenderLayers(S state) {
@@ -278,7 +_,7 @@
 
         state.isFullyFrozen = entity.isFullyFrozen();
         state.isBaby = entity.isBaby();
-        state.isInWater = entity.isInWater();
+        state.isInWater = entity.isInWater() || entity.getFluidInteraction().isInFluidMatching(entity, (e, fluidType, _) -> e.canSwimInFluidType(fluidType));
         state.isAutoSpinAttack = entity.isAutoSpinAttack();
         state.ticksSinceKineticHitFeedback = entity.getTicksSinceLastKineticHitFeedback(partialTicks);
         state.hasRedOverlay = entity.hurtTime > 0 || entity.deathTime > 0;
