--- a/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/LivingEntityRenderer.java
@@ -73,6 +_,7 @@
     }
 
     public void render(S p_361886_, PoseStack p_115311_, MultiBufferSource p_115312_, int p_115313_) {
+        if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLivingEvent.Pre<T, S, M>(p_361886_, this, p_361886_.partialTick, p_115311_, p_115312_, p_115313_)).isCanceled()) return;
         p_115311_.pushPose();
         if (p_361886_.hasPose(Pose.SLEEPING)) {
             Direction direction = p_361886_.bedOrientation;
@@ -108,6 +_,7 @@
 
         p_115311_.popPose();
         super.render(p_361886_, p_115311_, p_115312_, p_115313_);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLivingEvent.Post<T, S, M>(p_361886_, this, p_361886_.partialTick, p_115311_, p_115312_, p_115313_));
     }
 
     protected boolean shouldRenderLayers(S p_364697_) {
@@ -205,7 +_,7 @@
     protected boolean shouldShowName(T p_363517_, double p_365448_) {
         if (p_363517_.isDiscrete()) {
             float f = 32.0F;
-            if (p_365448_ >= 1024.0) {
+            if (!net.neoforged.neoforge.client.ClientHooks.isNameplateInRenderDistance(p_363517_, p_365448_)) {
                 return false;
             }
         }
@@ -289,7 +_,7 @@
         label48: {
             p_360515_.isFullyFrozen = p_362733_.isFullyFrozen();
             p_360515_.isBaby = p_362733_.isBaby();
-            p_360515_.isInWater = p_362733_.isInWater();
+            p_360515_.isInWater = p_362733_.isInWater() || p_362733_.isInFluidType((fluidType, height) -> p_362733_.canSwimInFluidType(fluidType));
             p_360515_.isAutoSpinAttack = p_362733_.isAutoSpinAttack();
             p_360515_.hasRedOverlay = p_362733_.hurtTime > 0 || p_362733_.deathTime > 0;
             ItemStack itemstack = p_362733_.getItemBySlot(EquipmentSlot.HEAD);
