--- a/net/minecraft/client/gui/screens/worldselection/AbstractGameRulesScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/AbstractGameRulesScreen.java
@@ -314,6 +_,12 @@
                             this.addEntry(gameRule, (x$0, x$1, x$2, x$3) -> AbstractGameRulesScreen.this.new IntegerRuleEntry(x$0, x$1, x$2, x$3));
                         }
 
+                        @Override
+                        public <T> void visit(GameRule<T> gameRule) {
+                            // Neo: Append custom entries for none bool/int game rules
+                            net.neoforged.neoforge.client.gamerules.GameRuleEntryFactoryManager.appendGameRuleEntry(AbstractGameRulesScreen.this, gameRule, this::addEntry);
+                        }
+
                         private <T> void addEntry(GameRule<T> gameRule, AbstractGameRulesScreen.EntryFactory<T> factory) {
                             Component readableName = Component.translatable(gameRule.getDescriptionId());
                             if (AbstractGameRulesScreen.RuleList.matchesFilter(
