--- a/net/minecraft/client/gui/screens/worldselection/CreateWorldScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/CreateWorldScreen.java
@@ -171,6 +_,7 @@
         queueLoadScreen(minecraft, PREPARING_WORLD_DATA);
         long start = Util.getMillis();
         PackRepository vanillaOnlyPackRepository = new PackRepository(new ServerPacksSource(minecraft.directoryValidator()));
+        net.neoforged.neoforge.resource.ResourcePackLoader.populatePackRepository(vanillaOnlyPackRepository, net.minecraft.server.packs.PackType.SERVER_DATA, false);
         WorldDataConfiguration dataConfig = SharedConstants.IS_RUNNING_IN_IDE
             ? new WorldDataConfiguration(new DataPackConfig(List.of("vanilla", "tests"), List.of()), FeatureFlags.DEFAULT_FLAGS)
             : WorldDataConfiguration.DEFAULT;
@@ -332,6 +_,10 @@
             SystemToast.onPackCopyFailure(this.minecraft, worldFolder);
             return false;
         } else {
+            if (worldDataAndGenSettings.data().worldGenSettingsLifecycle() != Lifecycle.stable()) {
+                // Neo: set experimental settings confirmation flag so user is not shown warning on next open
+                ((PrimaryLevelData) worldDataAndGenSettings.data()).withConfirmedWarning(true);
+            }
             this.minecraft
                 .createWorldOpenFlows()
                 .createLevelFromExistingSettings(newWorldAccess.get(), context.dataPackResources(), finalLayers, worldDataAndGenSettings, gameRules);
@@ -502,7 +_,7 @@
                                         if (retry) {
                                             onAbort.accept(this.uiState.getSettings().dataConfiguration());
                                         } else {
-                                            onAbort.accept(WorldDataConfiguration.DEFAULT);
+                                            onAbort.accept(new WorldDataConfiguration(new DataPackConfig(ImmutableList.of("vanilla"), ImmutableList.of()), FeatureFlags.VANILLA_SET)); // FORGE: Revert to *actual* vanilla data
                                         }
                                     },
                                     Component.translatable("dataPack.validation.failed"),
@@ -616,6 +_,7 @@
         if (dataPackDir != null) {
             if (this.tempDataPackRepository == null) {
                 this.tempDataPackRepository = ServerPacksSource.createPackRepository(dataPackDir, this.packValidator);
+                net.neoforged.neoforge.resource.ResourcePackLoader.populatePackRepository(this.tempDataPackRepository, net.minecraft.server.packs.PackType.SERVER_DATA, false);
                 this.tempDataPackRepository.reload();
             }
 
