--- a/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
@@ -268,6 +_,12 @@
                         );
                     }
 
+                    @Override
+                    public <T> void visit(GameRule<T> p_461081_) {
+                        // Neo: Append custom entries for none bool/int game rules
+                        addEntry(p_461081_, (label, tooltip, str, gameRule) -> net.neoforged.neoforge.client.gamerules.GameRuleEntryFactoryManager.createEntry(EditGameRulesScreen.this, label, tooltip, str, gameRule));
+                    }
+
                     private <T> void addEntry(GameRule<T> p_461117_, EditGameRulesScreen.EntryFactory<T> p_101226_) {
                         Component component = Component.translatable(p_461117_.getDescriptionId());
                         Component component1 = Component.literal(p_461117_.id()).withStyle(ChatFormatting.YELLOW);
