--- a/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
+++ b/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
@@ -19,7 +_,7 @@
             REGISTRY.inverse()::get
         );
     public static final ContextKeySet EMPTY = register("empty", builder -> {});
-    public static final ContextKeySet CHEST = register("chest", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY));
+    public static final ContextKeySet CHEST = register("chest", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY).optional(LootContextParams.ATTACKING_ENTITY)); //Forge: Chest minecarts can have killer entities
     public static final ContextKeySet COMMAND = register(
         "command", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY)
     );
@@ -33,7 +_,7 @@
             .required(LootContextParams.ADDITIONAL_COST_COMPONENT_ALLOWED)
     );
     public static final ContextKeySet FISHING = register(
-        "fishing", builder -> builder.required(LootContextParams.ORIGIN).required(LootContextParams.TOOL).optional(LootContextParams.THIS_ENTITY)
+        "fishing", builder -> builder.required(LootContextParams.ORIGIN).required(LootContextParams.TOOL).optional(LootContextParams.THIS_ENTITY).optional(LootContextParams.ATTACKING_ENTITY) // Neo: Add the fisher as a killer.
     );
     public static final ContextKeySet ENTITY = register(
         "entity",
