--- a/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
+++ b/net/minecraft/world/level/storage/loot/parameters/LootContextParamSets.java
@@ -19,7 +_,7 @@
             REGISTRY.inverse()::get
         );
     public static final ContextKeySet EMPTY = register("empty", builder -> {});
-    public static final ContextKeySet CHEST = register("chest", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY));
+    public static final ContextKeySet CHEST = register("chest", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY).optional(LootContextParams.ATTACKING_ENTITY)); //Forge: Chest minecarts can have killer entities
     public static final ContextKeySet COMMAND = register(
         "command", builder -> builder.required(LootContextParams.ORIGIN).optional(LootContextParams.THIS_ENTITY)
     );
@@ -33,7 +_,7 @@
             .required(LootContextParams.ADDITIONAL_COST_COMPONENT_ALLOWED)
     );
     public static final ContextKeySet FISHING = register(
-        "fishing", builder -> builder.required(LootContextParams.ORIGIN).required(LootContextParams.TOOL).optional(LootContextParams.THIS_ENTITY)
+        "fishing", builder -> builder.required(LootContextParams.ORIGIN).required(LootContextParams.TOOL).optional(LootContextParams.THIS_ENTITY).optional(LootContextParams.ATTACKING_ENTITY) // Neo: Add the fisher as a killer.
     );
     public static final ContextKeySet ENTITY = register(
         "entity",
@@ -140,10 +_,13 @@
     );
 
     private static ContextKeySet register(String name, Consumer<ContextKeySet.Builder> consumer) {
+        return register(Identifier.withDefaultNamespace(name),  consumer);
+    }
+
+    public static ContextKeySet register(Identifier id, Consumer<ContextKeySet.Builder> consumer) {
         ContextKeySet.Builder builder = new ContextKeySet.Builder();
         consumer.accept(builder);
         ContextKeySet result = builder.build();
-        Identifier id = Identifier.withDefaultNamespace(name);
         ContextKeySet prev = REGISTRY.put(id, result);
         if (prev != null) {
             throw new IllegalStateException("Loot table parameter set " + id + " is already registered");
