--- a/net/minecraft/world/effect/MobEffectInstance.java
+++ b/net/minecraft/world/effect/MobEffectInstance.java
@@ -336,11 +_,11 @@
                 .compareFalseFirst(this.isAmbient(), o.isAmbient())
                 .compareFalseFirst(this.isInfiniteDuration(), o.isInfiniteDuration())
                 .compare(this.getDuration(), o.getDuration())
-                .compare(this.getEffect().value().getColor(), o.getEffect().value().getColor())
+                .compare(this.getEffect().value().getSortOrder(this), o.getEffect().value().getSortOrder(o))
                 .result()
             : ComparisonChain.start()
                 .compare(this.isAmbient(), o.isAmbient())
-                .compare(this.getEffect().value().getColor(), o.getEffect().value().getColor())
+                .compare(this.getEffect().value().getSortOrder(this), o.getEffect().value().getSortOrder(o))
                 .result();
     }
 
