--- a/net/minecraft/world/effect/MobEffect.java
+++ b/net/minecraft/world/effect/MobEffect.java
@@ -35,7 +_,7 @@
 import net.minecraft.world.flag.FeatureFlagSet;
 import net.minecraft.world.flag.FeatureFlags;
 
-public class MobEffect implements FeatureElement {
+public class MobEffect implements FeatureElement, net.neoforged.neoforge.common.extensions.IMobEffectExtension {
     public static final Codec<Holder<MobEffect>> CODEC = BuiltInRegistries.MOB_EFFECT.holderByNameCodec();
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<MobEffect>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.MOB_EFFECT);
     private static final int AMBIENT_ALPHA = Mth.floor(38.25F);
@@ -56,6 +_,7 @@
             int i = p_333517_.isAmbient() ? AMBIENT_ALPHA : 255;
             return ColorParticleOption.create(ParticleTypes.ENTITY_EFFECT, FastColor.ARGB32.color(i, p_19452_));
         };
+        initClient();
     }
 
     protected MobEffect(MobEffectCategory p_333963_, int p_333864_, ParticleOptions p_333716_) {
@@ -180,6 +_,28 @@
     public FeatureFlagSet requiredFeatures() {
         return this.requiredFeatures;
     }
+
+    // NEO START
+    private Object effectRenderer;
+
+    /*
+     * DO NOT CALL, IT WILL DISAPPEAR IN THE FUTURE
+     * Call RenderProperties.getEffectRenderer instead
+     */
+    public Object getEffectRendererInternal() {
+        return effectRenderer;
+    }
+
+    private void initClient() {
+        // Minecraft instance isn't available in datagen, so don't call initializeClient if in datagen
+        if (net.neoforged.fml.loading.FMLEnvironment.dist == net.neoforged.api.distmarker.Dist.CLIENT && !net.neoforged.neoforge.data.loading.DatagenModLoader.isRunningDataGen()) {
+            initializeClient(properties -> this.effectRenderer = properties);
+        }
+    }
+
+    public void initializeClient(java.util.function.Consumer<net.neoforged.neoforge.client.extensions.common.IClientMobEffectExtensions> consumer) {
+    }
+    // END NEO
 
     static record AttributeTemplate(ResourceLocation id, double amount, AttributeModifier.Operation operation) {
         public AttributeModifier create(int p_316614_) {
