--- a/net/minecraft/world/effect/MobEffectInstance.java
+++ b/net/minecraft/world/effect/MobEffectInstance.java
@@ -86,6 +_,7 @@
         this.visible = p_19532_;
         this.showIcon = p_19533_;
         this.hiddenEffect = p_316863_;
+        this.effect.value().fillEffectCures(this.cures, this);
     }
 
     public MobEffectInstance(MobEffectInstance p_19543_) {
@@ -103,6 +_,8 @@
             p_324529_.showIcon(),
             p_324529_.hiddenEffect().map(p_323227_ -> new MobEffectInstance(p_324441_, p_323227_)).orElse(null)
         );
+        this.cures.clear();
+        p_324529_.cures().ifPresent(this.cures::addAll);
     }
 
     private MobEffectInstance.Details asDetails() {
@@ -112,7 +_,8 @@
             this.isAmbient(),
             this.isVisible(),
             this.showIcon(),
-            Optional.ofNullable(this.hiddenEffect).map(MobEffectInstance::asDetails)
+            Optional.ofNullable(this.hiddenEffect).map(MobEffectInstance::asDetails),
+            Optional.of(this.getCures()).filter(cures -> !cures.isEmpty())
         );
     }
 
@@ -130,6 +_,8 @@
         this.ambient = p_19549_.ambient;
         this.visible = p_19549_.visible;
         this.showIcon = p_19549_.showIcon;
+        this.cures.clear();
+        this.cures.addAll(p_19549_.cures);
     }
 
     public boolean update(MobEffectInstance p_19559_) {
@@ -326,11 +_,11 @@
                 .compareFalseFirst(this.isAmbient(), p_19566_.isAmbient())
                 .compareFalseFirst(this.isInfiniteDuration(), p_19566_.isInfiniteDuration())
                 .compare(this.getDuration(), p_19566_.getDuration())
-                .compare(this.getEffect().value().getColor(), p_19566_.getEffect().value().getColor())
+                .compare(this.getEffect().value().getSortOrder(this), p_19566_.getEffect().value().getSortOrder(p_19566_))
                 .result()
             : ComparisonChain.start()
                 .compare(this.isAmbient(), p_19566_.isAmbient())
-                .compare(this.getEffect().value().getColor(), p_19566_.getEffect().value().getColor())
+                .compare(this.getEffect().value().getSortOrder(this), p_19566_.getEffect().value().getSortOrder(p_19566_))
                 .result();
     }
 
@@ -350,6 +_,15 @@
         this.blendState.setImmediate(this);
     }
 
+    private final java.util.Set<net.neoforged.neoforge.common.EffectCure> cures = com.google.common.collect.Sets.newIdentityHashSet();
+
+    /**
+     * {@return the {@link net.neoforged.neoforge.common.EffectCure}s which can cure the {@link MobEffect} held by this {@link MobEffectInstance}}
+     */
+    public java.util.Set<net.neoforged.neoforge.common.EffectCure> getCures() {
+        return cures;
+    }
+
     static class BlendState {
         private float factor;
         private float factorPreviousFrame;
@@ -397,8 +_,7 @@
     }
 
     static record Details(
-        int amplifier, int duration, boolean ambient, boolean showParticles, boolean showIcon, Optional<MobEffectInstance.Details> hiddenEffect
-    ) {
+        int amplifier, int duration, boolean ambient, boolean showParticles, boolean showIcon, Optional<MobEffectInstance.Details> hiddenEffect, Optional<java.util.Set<net.neoforged.neoforge.common.EffectCure>> cures) {
         public static final MapCodec<MobEffectInstance.Details> MAP_CODEC = MapCodec.recursive(
             "MobEffectInstance.Details",
             p_323465_ -> RecordCodecBuilder.mapCodec(
@@ -408,13 +_,14 @@
                                 Codec.BOOL.optionalFieldOf("ambient", Boolean.valueOf(false)).forGetter(MobEffectInstance.Details::ambient),
                                 Codec.BOOL.optionalFieldOf("show_particles", Boolean.valueOf(true)).forGetter(MobEffectInstance.Details::showParticles),
                                 Codec.BOOL.optionalFieldOf("show_icon").forGetter(p_323788_ -> Optional.of(p_323788_.showIcon())),
-                                p_323465_.optionalFieldOf("hidden_effect").forGetter(MobEffectInstance.Details::hiddenEffect)
+                                p_323465_.optionalFieldOf("hidden_effect").forGetter(MobEffectInstance.Details::hiddenEffect),
+                                net.neoforged.neoforge.common.util.NeoForgeExtraCodecs.setOf(net.neoforged.neoforge.common.EffectCure.CODEC).optionalFieldOf("neoforge:cures").forGetter(MobEffectInstance.Details::cures)
                             )
                             .apply(p_324063_, MobEffectInstance.Details::create)
                 )
         );
         public static final StreamCodec<ByteBuf, MobEffectInstance.Details> STREAM_CODEC = StreamCodec.recursive(
-            p_329990_ -> StreamCodec.composite(
+            p_329990_ -> net.neoforged.neoforge.network.codec.NeoForgeStreamCodecs.composite(
                     ByteBufCodecs.VAR_INT,
                     MobEffectInstance.Details::amplifier,
                     ByteBufCodecs.VAR_INT,
@@ -427,14 +_,16 @@
                     MobEffectInstance.Details::showIcon,
                     p_329990_.apply(ByteBufCodecs::optional),
                     MobEffectInstance.Details::hiddenEffect,
+                    net.neoforged.neoforge.common.EffectCure.STREAM_CODEC.<java.util.Set<net.neoforged.neoforge.common.EffectCure>>apply(ByteBufCodecs.collection(java.util.HashSet::new)).apply(ByteBufCodecs::optional),
+                    MobEffectInstance.Details::cures,
                     MobEffectInstance.Details::new
                 )
         );
 
         private static MobEffectInstance.Details create(
-            int p_323657_, int p_324205_, boolean p_324263_, boolean p_324000_, Optional<Boolean> p_323607_, Optional<MobEffectInstance.Details> p_324604_
+            int p_323657_, int p_324205_, boolean p_324263_, boolean p_324000_, Optional<Boolean> p_323607_, Optional<MobEffectInstance.Details> p_324604_, Optional<java.util.Set<net.neoforged.neoforge.common.EffectCure>> cures
         ) {
-            return new MobEffectInstance.Details(p_323657_, p_324205_, p_324263_, p_324000_, p_323607_.orElse(p_324000_), p_324604_);
+            return new MobEffectInstance.Details(p_323657_, p_324205_, p_324263_, p_324000_, p_323607_.orElse(p_324000_), p_324604_, cures);
         }
     }
 }
