--- a/net/minecraft/world/effect/PoisonMobEffect.java
+++ b/net/minecraft/world/effect/PoisonMobEffect.java
@@ -11,7 +_,8 @@
     public void applyEffectTick(LivingEntity p_296276_, int p_296233_) {
         super.applyEffectTick(p_296276_, p_296233_);
         if (p_296276_.getHealth() > 1.0F) {
-            p_296276_.hurt(p_296276_.damageSources().magic(), 1.0F);
+            // Neo: Replace DamageSources#magic() with neoforge:poison to allow differentiating poison damage.
+            p_296276_.hurt(p_296276_.damageSources().source(net.neoforged.neoforge.common.NeoForgeMod.POISON_DAMAGE), 1.0F);
         }
     }
 
