--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -721,8 +736,12 @@
 
                 profiler.push("translucent");
                 chunkSectionsToRender.renderGroup(ChunkSectionLayerGroup.TRANSLUCENT, this.chunkLayerSampler);
+                profiler.popPush("neoforge_render_after_translucent_blocks");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterTranslucentBlocks(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
                 profiler.popPush("string");
                 chunkSectionsToRender.renderGroup(ChunkSectionLayerGroup.TRIPWIRE, this.chunkLayerSampler);
+                profiler.popPush("neoforge_render_after_tripwire_blocks");
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterTripwireBlocks(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
                 if (renderOutline) {
                     this.renderBlockOutline(bufferSource, poseStack, true, levelRenderState);
                 }
@@ -783,6 +834,9 @@
             RenderSystem.setShaderFog(fog);
             MultiBufferSource.BufferSource bufferSource = this.renderBuffers.bufferSource();
             CameraRenderState cameraState = this.levelRenderState.cameraRenderState;
+            Profiler.get().push("neoforge_render_after_weather");
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.RenderLevelStageEvent.AfterWeather(this, this.levelRenderState, null, modelViewMatrix, this.getRenderableSections()));
+            Profiler.get().pop();
             this.weatherEffectRenderer.render(bufferSource, cameraState.pos, this.levelRenderState.weatherRenderState);
             this.worldBorderRenderer.render(this.levelRenderState.worldBorderRenderState, cameraState.pos, renderDistance, depthFar);
             bufferSource.endBatch();
