--- a/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -83,6 +_,13 @@
         ListTag listtag = new ListTag();
 
         for (AttributeInstance attributeinstance : this.attributes.values()) {
+            // Neo: skip saving attributes that are not saveDefault and have their default value and no modifiers
+            if (!attributeinstance.getAttribute().value().isSaveDefault() &&
+                    attributeinstance.getModifiers().isEmpty() &&
+                    attributeinstance.getBaseValue() == attributeinstance.getAttribute().value().getDefaultValue()
+            ) {
+                continue;
+            }
             listtag.add(attributeinstance.save());
         }
 
