--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -190,11 +_,12 @@
         .build();
 
     public static AttributeSupplier getSupplier(EntityType<? extends LivingEntity> type) {
-        return SUPPLIERS.get(type);
+        AttributeSupplier supplier = net.neoforged.neoforge.common.CommonHooks.getAttributesView().get(type);
+        return supplier != null ? supplier : SUPPLIERS.get(type);
     }
 
     public static boolean hasSupplier(EntityType<?> type) {
-        return SUPPLIERS.containsKey(type);
+        return SUPPLIERS.containsKey(type) || net.neoforged.neoforge.common.CommonHooks.getAttributesView().containsKey(type);
     }
 
     public static void validate() {
