--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -168,11 +168,12 @@
       .build();
 
    public static AttributeSupplier getSupplier(EntityType<? extends LivingEntity> p_22298_) {
-      return SUPPLIERS.get(p_22298_);
+      AttributeSupplier supplier = net.neoforged.neoforge.common.ForgeHooks.getAttributesView().get(p_22298_);
+      return supplier != null ? supplier : SUPPLIERS.get(p_22298_);
    }
 
    public static boolean hasSupplier(EntityType<?> p_22302_) {
-      return SUPPLIERS.containsKey(p_22302_);
+      return SUPPLIERS.containsKey(p_22302_) || net.neoforged.neoforge.common.ForgeHooks.getAttributesView().containsKey(p_22302_);
    }
 
    public static void validate() {
