--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -174,11 +_,12 @@
         .build();
 
     public static AttributeSupplier getSupplier(EntityType<? extends LivingEntity> livingEntity) {
-        return SUPPLIERS.get(livingEntity);
+        AttributeSupplier supplier = net.neoforged.neoforge.common.CommonHooks.getAttributesView().get(livingEntity);
+        return supplier != null ? supplier : SUPPLIERS.get(livingEntity);
     }
 
     public static boolean hasSupplier(EntityType<?> entityType) {
-        return SUPPLIERS.containsKey(entityType);
+        return SUPPLIERS.containsKey(entityType) || net.neoforged.neoforge.common.CommonHooks.getAttributesView().containsKey(entityType);
     }
 
     public static void validate() {
