--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithEnchantedBonusCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithEnchantedBonusCondition.java
@@ -41,6 +_,10 @@
     public boolean test(LootContext context) {
         Entity killerEntity = context.getOptionalParameter(LootContextParams.ATTACKING_ENTITY);
         int enchantmentLevel = killerEntity instanceof LivingEntity livingKiller ? EnchantmentHelper.getEnchantmentLevel(this.enchantment, livingKiller) : 0;
+
+        // Neo: Allow mods to modify the enchantment level for loot calculations with relevant context.
+        enchantmentLevel = net.neoforged.neoforge.event.EventHooks.getEntityLootEnchantmentLevel(this.enchantment, enchantmentLevel, context);
+
         float chance = enchantmentLevel > 0 ? this.enchantedChance.calculate(enchantmentLevel) : this.unenchantedChance;
         return context.getRandom().nextFloat() < chance;
     }
