--- a/net/minecraft/client/particle/TerrainParticle.java
+++ b/net/minecraft/client/particle/TerrainParticle.java
@@ -29,7 +29,7 @@
         this.rCol = 0.6F;
         this.gCol = 0.6F;
         this.bCol = 0.6F;
-        if (!blockState.is(Blocks.GRASS_BLOCK)) {
+        if (net.neoforged.neoforge.client.extensions.common.IClientBlockExtensions.of(blockState).areBreakingParticlesTinted(blockState, level, pos)) {
             int col = Minecraft.getInstance().getBlockColors().getColor(blockState, level, pos, 0);
             this.rCol *= (col >> 16 & 0xFF) / 255.0F;
             this.gCol *= (col >> 8 & 0xFF) / 255.0F;
