--- a/net/minecraft/data/tags/BannerPatternTagsProvider.java
+++ b/net/minecraft/data/tags/BannerPatternTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.block.entity.BannerPatterns;
 
 public class BannerPatternTagsProvider extends TagsProvider<BannerPattern> {
+    /** @deprecated Forge: Use the {@linkplain #BannerPatternTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public BannerPatternTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
         super(output, Registries.BANNER_PATTERN, provider);
+    }
+    public BannerPatternTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.BANNER_PATTERN, provider, modId, existingFileHelper);
     }
 
     @Override
