--- a/net/minecraft/data/tags/BiomeTagsProvider.java
+++ b/net/minecraft/data/tags/BiomeTagsProvider.java
@@ -12,8 +_,13 @@
 import net.minecraft.world.level.biome.MultiNoiseBiomeSourceParameterList;
 
 public class BiomeTagsProvider extends KeyTagProvider<Biome> {
+    /** @deprecated Forge: Use the {@linkplain #BiomeTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public BiomeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.BIOME, lookupProvider);
+    }
+    public BiomeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.BIOME, lookupProvider, modId);
     }
 
     @Override
