--- a/net/minecraft/data/tags/CatVariantTagsProvider.java
+++ b/net/minecraft/data/tags/CatVariantTagsProvider.java
@@ -8,8 +_,13 @@
 import net.minecraft.world.entity.animal.CatVariant;
 
 public class CatVariantTagsProvider extends TagsProvider<CatVariant> {
+    /** @deprecated Forge: Use the {@linkplain #CatVariantTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public CatVariantTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
         super(output, Registries.CAT_VARIANT, provider);
+    }
+    public CatVariantTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.CAT_VARIANT, provider, modId, existingFileHelper);
     }
 
     @Override
