--- a/net/minecraft/data/tags/DamageTypeTagsProvider.java
+++ b/net/minecraft/data/tags/DamageTypeTagsProvider.java
@@ -9,8 +_,14 @@
 import net.minecraft.world.damagesource.DamageTypes;
 
 public class DamageTypeTagsProvider extends TagsProvider<DamageType> {
+    /// @deprecated Neo: Use the [mod id variant][#DamageTypeTagsProvider(PackOutput, CompletableFuture, String)]
+    @Deprecated
     public DamageTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.DAMAGE_TYPE, lookupProvider);
+    }
+
+    public DamageTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.DAMAGE_TYPE, lookupProvider, modId);
     }
 
     @Override
