--- a/net/minecraft/data/tags/EnchantmentTagsProvider.java
+++ b/net/minecraft/data/tags/EnchantmentTagsProvider.java
@@ -13,8 +_,13 @@
 import net.minecraft.world.item.enchantment.Enchantment;
 
 public abstract class EnchantmentTagsProvider extends KeyTagProvider<Enchantment> {
+    /** @deprecated Forge: Use the {@linkplain #EnchantmentTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public EnchantmentTagsProvider(PackOutput p_341044_, CompletableFuture<HolderLookup.Provider> p_341146_) {
         super(p_341044_, Registries.ENCHANTMENT, p_341146_);
+    }
+    public EnchantmentTagsProvider(PackOutput p_341044_, CompletableFuture<HolderLookup.Provider> p_341146_, String modId) {
+        super(p_341044_, Registries.ENCHANTMENT, p_341146_, modId);
     }
 
     protected void tooltipOrder(HolderLookup.Provider p_341105_, ResourceKey<Enchantment>... p_345255_) {
