--- a/net/minecraft/data/tags/EnchantmentTagsProvider.java
+++ b/net/minecraft/data/tags/EnchantmentTagsProvider.java
@@ -13,8 +_,12 @@
 import net.minecraft.world.item.enchantment.Enchantment;
 
 public abstract class EnchantmentTagsProvider extends TagsProvider<Enchantment> {
+    /** @deprecated Forge: Use the {@linkplain #EnchantmentTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
     public EnchantmentTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.ENCHANTMENT, lookupProvider);
+    }
+    public EnchantmentTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.ENCHANTMENT, lookupProvider, modId, existingFileHelper);
     }
 
     protected void tooltipOrder(HolderLookup.Provider provider, ResourceKey<Enchantment>... values) {
