--- a/net/minecraft/data/tags/EntityTypeTagsProvider.java
+++ b/net/minecraft/data/tags/EntityTypeTagsProvider.java
@@ -8,8 +_,13 @@
 import net.minecraft.world.entity.EntityType;
 
 public class EntityTypeTagsProvider extends IntrinsicHolderTagsProvider<EntityType<?>> {
+    /** @deprecated Forge: Use the {@linkplain #EntityTypeTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public EntityTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
-        super(output, Registries.ENTITY_TYPE, provider, p_256665_ -> p_256665_.builtInRegistryHolder().key());
+        this(output, provider, "vanilla", null);
+    }
+    public EntityTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.ENTITY_TYPE, provider, p_256665_ -> p_256665_.builtInRegistryHolder().key(), modId, existingFileHelper);
     }
 
     @Override
