--- a/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
+++ b/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.levelgen.flat.FlatLevelGeneratorPresets;
 
 public class FlatLevelGeneratorPresetTagsProvider extends KeyTagProvider<FlatLevelGeneratorPreset> {
+    /** @deprecated Forge: Use the {@linkplain #FlatLevelGeneratorPresetTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public FlatLevelGeneratorPresetTagsProvider(PackOutput p_256604_, CompletableFuture<HolderLookup.Provider> p_255962_) {
         super(p_256604_, Registries.FLAT_LEVEL_GENERATOR_PRESET, p_255962_);
+    }
+    public FlatLevelGeneratorPresetTagsProvider(PackOutput p_256604_, CompletableFuture<HolderLookup.Provider> p_255962_, String modId) {
+        super(p_256604_, Registries.FLAT_LEVEL_GENERATOR_PRESET, p_255962_, modId);
     }
 
     @Override
