--- a/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
+++ b/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.levelgen.flat.FlatLevelGeneratorPresets;
 
 public class FlatLevelGeneratorPresetTagsProvider extends KeyTagProvider<FlatLevelGeneratorPreset> {
+    /** @deprecated Forge: Use the {@linkplain #FlatLevelGeneratorPresetTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public FlatLevelGeneratorPresetTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.FLAT_LEVEL_GENERATOR_PRESET, lookupProvider);
+    }
+    public FlatLevelGeneratorPresetTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.FLAT_LEVEL_GENERATOR_PRESET, lookupProvider, modId);
     }
 
     @Override
