--- a/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
+++ b/net/minecraft/data/tags/FlatLevelGeneratorPresetTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.levelgen.flat.FlatLevelGeneratorPresets;
 
 public class FlatLevelGeneratorPresetTagsProvider extends TagsProvider<FlatLevelGeneratorPreset> {
+    /** @deprecated Forge: Use the {@linkplain #FlatLevelGeneratorPresetTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public FlatLevelGeneratorPresetTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
         super(output, Registries.FLAT_LEVEL_GENERATOR_PRESET, provider);
+    }
+    public FlatLevelGeneratorPresetTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.FLAT_LEVEL_GENERATOR_PRESET, provider, modId, existingFileHelper);
     }
 
     @Override
