--- a/net/minecraft/data/tags/FluidTagsProvider.java
+++ b/net/minecraft/data/tags/FluidTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.material.Fluids;
 
 public class FluidTagsProvider extends IntrinsicHolderTagsProvider<Fluid> {
+    /** @deprecated Forge: Use the {@linkplain #FluidTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public FluidTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
-        super(output, Registries.FLUID, provider, p_256474_ -> p_256474_.builtInRegistryHolder().key());
+        this(output, provider, "vanilla", null);
+    }
+    public FluidTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.FLUID, provider, p_256474_ -> p_256474_.builtInRegistryHolder().key(), modId, existingFileHelper);
     }
 
     @Override
