--- a/net/minecraft/data/tags/InstrumentTagsProvider.java
+++ b/net/minecraft/data/tags/InstrumentTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.item.Instruments;
 
 public class InstrumentTagsProvider extends KeyTagProvider<Instrument> {
+    /** @deprecated Forge: Use the {@linkplain #InstrumentTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public InstrumentTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.INSTRUMENT, lookupProvider);
+    }
+    public InstrumentTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.INSTRUMENT, lookupProvider, modId);
     }
 
     @Override
