--- a/net/minecraft/data/tags/IntrinsicHolderTagsProvider.java
+++ b/net/minecraft/data/tags/IntrinsicHolderTagsProvider.java
@@ -12,16 +_,17 @@
 public abstract class IntrinsicHolderTagsProvider<T> extends TagsProvider<T> {
     private final Function<T, ResourceKey<T>> keyExtractor;
 
+    @Deprecated
     public IntrinsicHolderTagsProvider(
         PackOutput output,
         ResourceKey<? extends Registry<T>> registryKey,
         CompletableFuture<HolderLookup.Provider> lookupProvider,
         Function<T, ResourceKey<T>> keyExtractor
     ) {
-        super(output, registryKey, lookupProvider);
-        this.keyExtractor = keyExtractor;
+        this(output, registryKey, lookupProvider, keyExtractor, "vanilla");
     }
 
+    @Deprecated
     public IntrinsicHolderTagsProvider(
         PackOutput output,
         ResourceKey<? extends Registry<T>> registryKey,
@@ -29,7 +_,16 @@
         CompletableFuture<TagsProvider.TagLookup<T>> parentProvider,
         Function<T, ResourceKey<T>> keyExtractor
     ) {
-        super(output, registryKey, lookupProvider, parentProvider);
+        this(output, registryKey, lookupProvider, parentProvider, keyExtractor, "vanilla");
+    }
+
+    public IntrinsicHolderTagsProvider(PackOutput output, ResourceKey<? extends Registry<T>> registryKey, CompletableFuture<HolderLookup.Provider> lookupProvider, Function<T, ResourceKey<T>> keyExtractor, String modId) {
+        super(output, registryKey, lookupProvider, modId);
+        this.keyExtractor = keyExtractor;
+    }
+
+    public IntrinsicHolderTagsProvider(PackOutput output, ResourceKey<? extends Registry<T>> registryKey, CompletableFuture<HolderLookup.Provider> lookupProvider, CompletableFuture<TagsProvider.TagLookup<T>> parentProvider, Function<T, ResourceKey<T>> keyExtractor, String modId) {
+        super(output, registryKey, lookupProvider, parentProvider, modId);
         this.keyExtractor = keyExtractor;
     }
 
