--- a/net/minecraft/data/tags/IntrinsicHolderTagsProvider.java
+++ b/net/minecraft/data/tags/IntrinsicHolderTagsProvider.java
@@ -12,16 +_,17 @@
 public abstract class IntrinsicHolderTagsProvider<T> extends TagsProvider<T> {
     private final Function<T, ResourceKey<T>> keyExtractor;
 
+    @Deprecated
     public IntrinsicHolderTagsProvider(
         PackOutput p_256164_,
         ResourceKey<? extends Registry<T>> p_256155_,
         CompletableFuture<HolderLookup.Provider> p_256488_,
         Function<T, ResourceKey<T>> p_256168_
     ) {
-        super(p_256164_, p_256155_, p_256488_);
-        this.keyExtractor = p_256168_;
+        this(p_256164_, p_256155_, p_256488_, p_256168_, "vanilla");
     }
 
+    @Deprecated
     public IntrinsicHolderTagsProvider(
         PackOutput p_275304_,
         ResourceKey<? extends Registry<T>> p_275709_,
@@ -29,7 +_,16 @@
         CompletableFuture<TagsProvider.TagLookup<T>> p_275311_,
         Function<T, ResourceKey<T>> p_275566_
     ) {
-        super(p_275304_, p_275709_, p_275227_, p_275311_);
+        this(p_275304_, p_275709_, p_275227_, p_275311_, p_275566_, "vanilla");
+    }
+
+    public IntrinsicHolderTagsProvider(PackOutput p_256164_, ResourceKey<? extends Registry<T>> p_256155_, CompletableFuture<HolderLookup.Provider> p_256488_, Function<T, ResourceKey<T>> p_256168_, String modId) {
+        super(p_256164_, p_256155_, p_256488_, modId);
+        this.keyExtractor = p_256168_;
+    }
+
+    public IntrinsicHolderTagsProvider(PackOutput p_275304_, ResourceKey<? extends Registry<T>> p_275709_, CompletableFuture<HolderLookup.Provider> p_275227_, CompletableFuture<TagsProvider.TagLookup<T>> p_275311_, Function<T, ResourceKey<T>> p_275566_, String modId) {
+        super(p_275304_, p_275709_, p_275227_, p_275311_, modId);
         this.keyExtractor = p_275566_;
     }
 
