--- a/net/minecraft/data/tags/KeyTagProvider.java
+++ b/net/minecraft/data/tags/KeyTagProvider.java
@@ -9,8 +_,16 @@
 import net.minecraft.tags.TagKey;
 
 public abstract class KeyTagProvider<T> extends TagsProvider<T> {
+    /**
+     * @deprecated Neo: Use {@link #KeyTagProvider(PackOutput,ResourceKey,CompletableFuture,String)}
+     */
+    @Deprecated
     protected KeyTagProvider(PackOutput output, ResourceKey<? extends Registry<T>> registryKey, CompletableFuture<HolderLookup.Provider> lookupProvider) {
-        super(output, registryKey, lookupProvider);
+        this(output, registryKey, lookupProvider, "vanilla");
+    }
+
+    protected KeyTagProvider(PackOutput output, ResourceKey<? extends Registry<T>> registryKey, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, registryKey, lookupProvider, modId);
     }
 
     protected TagAppender<ResourceKey<T>, T> tag(TagKey<T> tag) {
