--- a/net/minecraft/data/tags/KeyTagProvider.java
+++ b/net/minecraft/data/tags/KeyTagProvider.java
@@ -9,8 +_,16 @@
 import net.minecraft.tags.TagKey;
 
 public abstract class KeyTagProvider<T> extends TagsProvider<T> {
+    /**
+     * @deprecated Neo: Use {@link #KeyTagProvider(PackOutput,ResourceKey,CompletableFuture,String)}
+     */
+    @Deprecated
     protected KeyTagProvider(PackOutput p_421572_, ResourceKey<? extends Registry<T>> p_421754_, CompletableFuture<HolderLookup.Provider> p_422211_) {
-        super(p_421572_, p_421754_, p_422211_);
+        this(p_421572_, p_421754_, p_422211_, "vanilla");
+    }
+
+    protected KeyTagProvider(PackOutput p_421572_, ResourceKey<? extends Registry<T>> p_421754_, CompletableFuture<HolderLookup.Provider> p_422211_, String modId) {
+        super(p_421572_, p_421754_, p_422211_, modId);
     }
 
     protected TagAppender<ResourceKey<T>, T> tag(TagKey<T> p_421489_) {
