--- a/net/minecraft/data/tags/PaintingVariantTagsProvider.java
+++ b/net/minecraft/data/tags/PaintingVariantTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.entity.decoration.PaintingVariants;
 
 public class PaintingVariantTagsProvider extends TagsProvider<PaintingVariant> {
+    /** @deprecated Forge: Use the {@linkplain #PaintingVariantTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public PaintingVariantTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
         super(output, Registries.PAINTING_VARIANT, provider);
+    }
+    public PaintingVariantTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.PAINTING_VARIANT, provider, modId, existingFileHelper);
     }
 
     @Override
