--- a/net/minecraft/data/tags/PoiTypeTagsProvider.java
+++ b/net/minecraft/data/tags/PoiTypeTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 
 public class PoiTypeTagsProvider extends TagsProvider<PoiType> {
+    /** @deprecated Forge: Use the {@linkplain #PoiTypeTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public PoiTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
         super(output, Registries.POINT_OF_INTEREST_TYPE, provider);
+    }
+    public PoiTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.POINT_OF_INTEREST_TYPE, provider, modId, existingFileHelper);
     }
 
     @Override
