--- a/net/minecraft/data/tags/StructureTagsProvider.java
+++ b/net/minecraft/data/tags/StructureTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.levelgen.structure.Structure;
 
 public class StructureTagsProvider extends TagsProvider<Structure> {
+    /** @deprecated Forge: Use the {@linkplain #StructureTagsProvider(PackOutput, CompletableFuture, String, net.neoforged.neoforge.common.data.ExistingFileHelper) mod id variant} */
+    @Deprecated
     public StructureTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider) {
         super(output, Registries.STRUCTURE, provider);
+    }
+    public StructureTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> provider, String modId, @org.jetbrains.annotations.Nullable net.neoforged.neoforge.common.data.ExistingFileHelper existingFileHelper) {
+        super(output, Registries.STRUCTURE, provider, modId, existingFileHelper);
     }
 
     @Override
