--- a/net/minecraft/data/tags/StructureTagsProvider.java
+++ b/net/minecraft/data/tags/StructureTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.level.levelgen.structure.Structure;
 
 public class StructureTagsProvider extends KeyTagProvider<Structure> {
+    /** @deprecated Forge: Use the {@linkplain #StructureTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public StructureTagsProvider(PackOutput p_256522_, CompletableFuture<HolderLookup.Provider> p_256661_) {
         super(p_256522_, Registries.STRUCTURE, p_256661_);
+    }
+    public StructureTagsProvider(PackOutput p_256522_, CompletableFuture<HolderLookup.Provider> p_256661_, String modId) {
+        super(p_256522_, Registries.STRUCTURE, p_256661_, modId);
     }
 
     @Override
