--- a/net/minecraft/data/tags/WorldPresetTagsProvider.java
+++ b/net/minecraft/data/tags/WorldPresetTagsProvider.java
@@ -9,8 +_,14 @@
 import net.minecraft.world.level.levelgen.presets.WorldPresets;
 
 public class WorldPresetTagsProvider extends TagsProvider<WorldPreset> {
+    /// @deprecated Neo: Use the [mod id variant][#WorldPresetTagsProvider(PackOutput, CompletableFuture, String)]
+    @Deprecated
     public WorldPresetTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.WORLD_PRESET, lookupProvider);
+    }
+
+    public WorldPresetTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.WORLD_PRESET, lookupProvider, modId);
     }
 
     @Override
