--- a/net/minecraft/world/entity/boss/EnderDragonPart.java
+++ b/net/minecraft/world/entity/boss/EnderDragonPart.java
@@ -13,13 +_,13 @@
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
 import net.minecraft.world.item.ItemStack;
 
-public class EnderDragonPart extends Entity {
+public class EnderDragonPart extends net.neoforged.neoforge.entity.PartEntity<EnderDragon> {
     public final EnderDragon parentMob;
     public final String name;
     private final EntityDimensions size;
 
     public EnderDragonPart(EnderDragon parentMob, String name, float width, float height) {
-        super(parentMob.getType(), parentMob.level());
+        super(parentMob);
         this.size = EntityDimensions.scalable(width, height);
         this.refreshDimensions();
         this.parentMob = parentMob;
