--- a/net/minecraft/world/entity/ai/village/VillageSiege.java
+++ b/net/minecraft/world/entity/ai/village/VillageSiege.java
@@ -77,7 +_,9 @@
                         this.spawnX = blockpos.getX() + Mth.floor(Mth.cos(f) * 32.0F);
                         this.spawnY = blockpos.getY();
                         this.spawnZ = blockpos.getZ() + Mth.floor(Mth.sin(f) * 32.0F);
-                        if (this.findRandomSpawnPos(level, new BlockPos(this.spawnX, this.spawnY, this.spawnZ)) != null) {
+                        Vec3 siegeLocation = this.findRandomSpawnPos(level, new BlockPos(this.spawnX, this.spawnY, this.spawnZ));
+                        if (siegeLocation != null) {
+                            if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.village.VillageSiegeEvent(this, level, player, siegeLocation)).isCanceled()) return false;
                             this.nextSpawnTime = 0;
                             this.zombiesToSpawn = 20;
                             break;
@@ -97,7 +_,7 @@
         if (vec3 != null) {
             Zombie zombie;
             try {
-                zombie = new Zombie(level);
+                zombie = EntityType.ZOMBIE.create(level); //Forge: Direct Initialization is deprecated, use EntityType.
                 zombie.finalizeSpawn(level, level.getCurrentDifficultyAt(zombie.blockPosition()), MobSpawnType.EVENT, null);
             } catch (Exception exception) {
                 LOGGER.warn("Failed to create zombie for village siege at {}", vec3, exception);
