--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -68,6 +_,7 @@
                 if (p_175667_.hasTag()) {
                     itemstack.setTag(p_175667_.getTag().copy());
                 }
+                net.neoforged.neoforge.attachment.AttachmentUtils.copyStackAttachments(p_175667_, itemstack);
 
                 p_175665_.setItemInHand(p_175666_, itemstack);
                 p_175665_.awardStat(Stats.CLEAN_SHULKER_BOX);
