--- a/net/minecraft/server/packs/resources/MultiPackResourceManager.java
+++ b/net/minecraft/server/packs/resources/MultiPackResourceManager.java
@@ -28,7 +_,8 @@
         for (PackResources pack : packs) {
             ResourceFilterSection filterSection = this.getPackFilterSection(pack);
             Set<String> providedNamespaces = pack.getNamespaces(type);
-            Predicate<Identifier> pathFilter = filterSection != null ? location -> filterSection.isPathFiltered(location.getPath()) : null;
+            // Neo: Fix MC-271761 by properly coupling namespace and path checks
+            Predicate<Identifier> pathFilter = filterSection != null ? filterSection::isFiltered : null;
 
             for (String namespace : namespaces) {
                 boolean packContainsNamespace = providedNamespaces.contains(namespace);
@@ -58,6 +_,7 @@
         try {
             return pack.getMetadataSection(ResourceFilterSection.TYPE);
         } catch (Exception e) {
+            if (!net.neoforged.neoforge.data.loading.DatagenModLoader.isRunningDataGen())// Neo: Only warn about malformed pack filters outside of datagen in case modders are replacing variables with gradle
             LOGGER.error("Failed to get filter section from pack {}", pack.packId());
             return null;
         }
