--- a/net/minecraft/server/packs/resources/ReloadableResourceManager.java
+++ b/net/minecraft/server/packs/resources/ReloadableResourceManager.java
@@ -32,6 +_,10 @@
         this.resources.close();
     }
 
+    /// @deprecated Neo: Use [net.neoforged.neoforge.client.event.AddClientReloadListenersEvent] or [net.neoforged.neoforge.event.AddServerReloadListenersEvent].
+    ///
+    /// @throws UnsupportedOperationException if called after the event has been fired.
+    @Deprecated
     public void registerReloadListener(PreparableReloadListener listener) {
         this.listeners.add(listener);
     }
@@ -73,5 +_,20 @@
     @Override
     public Stream<PackResources> listPacks() {
         return this.resources.listPacks();
+    }
+
+    /// Neo: Expose the reload listeners so they can be passed to the event.
+    ///
+    /// @return The (immutable) list of reload listeners.
+    public List<PreparableReloadListener> getListeners() {
+        return this.listeners;
+    }
+
+    /// Neo: Updates the [#listeners] with the sorted list from the event.
+    ///
+    /// @implNote The returned list is immutable, so after this method is called, [#registerReloadListener(PreparableReloadListener)] will throw.
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void updateListenersFrom(net.neoforged.neoforge.event.SortedReloadListenerEvent event) {
+        this.listeners = net.neoforged.neoforge.resource.ReloadListenerSort.sort(event);
     }
 }
