--- a/net/minecraft/server/packs/resources/ResourceFilterSection.java
+++ b/net/minecraft/server/packs/resources/ResourceFilterSection.java
@@ -24,4 +_,11 @@
     public boolean isPathFiltered(String path) {
         return this.blockList.stream().anyMatch(p -> p.pathPredicate().test(path));
     }
+
+    // Neo: Fix MC-271761 by properly coupling namespace and path in a single predicate check
+    public boolean isFiltered(net.minecraft.resources.Identifier id) {
+        return this.blockList.stream().anyMatch(pattern ->
+                pattern.namespacePredicate().test(id.getNamespace()) && pattern.pathPredicate().test(id.getPath())
+        );
+    }
 }
