--- a/net/minecraft/server/packs/resources/SimplePreparableReloadListener.java
+++ b/net/minecraft/server/packs/resources/SimplePreparableReloadListener.java
@@ -5,7 +_,7 @@
 import net.minecraft.util.profiling.Profiler;
 import net.minecraft.util.profiling.ProfilerFiller;
 
-public abstract class SimplePreparableReloadListener<T> implements PreparableReloadListener {
+public abstract class SimplePreparableReloadListener<T> extends net.neoforged.neoforge.resource.ContextAwareReloadListener implements PreparableReloadListener {
     @Override
     public final CompletableFuture<Void> reload(
         PreparableReloadListener.SharedState p_434245_, Executor p_10784_, PreparableReloadListener.PreparationBarrier p_10780_, Executor p_10785_
