--- a/net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
+++ b/net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
@@ -42,7 +_,8 @@
 
     protected Map<Identifier, T> prepare(ResourceManager manager, ProfilerFiller profiler) {
         Map<Identifier, T> result = new HashMap<>();
-        scanDirectory(manager, this.lister, this.ops, this.codec, result);
+        // Neo: add condition context
+        scanDirectory(manager, this.lister, this.makeConditionalOps(this.ops), this.codec, result);
         return result;
     }
 
@@ -52,16 +_,29 @@
         scanDirectory(manager, FileToIdConverter.registry(registryKey), ops, codec, result);
     }
 
+    public static <T> void scanDirectoryWithOptionalValues(
+            ResourceManager manager,
+            ResourceKey<? extends Registry<T>> registryKey,
+            DynamicOps<JsonElement> ops,
+            Codec<java.util.Optional<T>> codec,
+            Map<Identifier, java.util.Optional<T>> result
+    ) {
+        scanDirectory(manager, FileToIdConverter.registry(registryKey), ops, codec, result);
+    }
+
     public static <T> void scanDirectory(
         ResourceManager manager, FileToIdConverter lister, DynamicOps<JsonElement> ops, Codec<T> codec, Map<Identifier, T> result
     ) {
+        var conditionalCodec = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodec(codec);
         for (Entry<Identifier, Resource> entry : lister.listMatchingResources(manager).entrySet()) {
             Identifier location = entry.getKey();
             Identifier id = lister.fileToId(location);
 
             try (Reader reader = entry.getValue().openAsReader()) {
-                codec.parse(ops, StrictJsonParser.parse(reader)).ifSuccess(parsed -> {
-                    if (result.putIfAbsent(id, (T)parsed) != null) {
+                conditionalCodec.parse(ops, com.google.gson.JsonParser.parseReader(reader)).ifSuccess(parsed -> {
+                    if (parsed.isEmpty()) {
+                        LOGGER.debug("Skipping loading data file '{}' from '{}' as its conditions were not met", id, location);
+                    } else if (result.putIfAbsent(id, parsed.get()) != null) {
                         throw new IllegalStateException("Duplicate data file ignored with ID " + id);
                     }
                 }).ifError(error -> LOGGER.error("Couldn't parse data file '{}' from '{}': {}", id, location, error));
@@ -69,5 +_,9 @@
                 LOGGER.error("Couldn't parse data file '{}' from '{}'", id, location, var14);
             }
         }
+    }
+
+    protected Identifier getPreparedPath(Identifier rl) {
+        return this.lister.idToFile(rl);
     }
 }
