--- a/net/minecraft/server/packs/resources/SimplePreparableReloadListener.java
+++ b/net/minecraft/server/packs/resources/SimplePreparableReloadListener.java
@@ -4,7 +4,7 @@
 import java.util.concurrent.Executor;
 import net.minecraft.util.profiling.ProfilerFiller;
 
-public abstract class SimplePreparableReloadListener<T> implements PreparableReloadListener {
+public abstract class SimplePreparableReloadListener<T> extends net.neoforged.neoforge.resource.ContextAwareReloadListener implements PreparableReloadListener {
    @Override
    public final CompletableFuture<Void> reload(
       PreparableReloadListener.PreparationBarrier p_10780_,
