--- a/net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
+++ b/net/minecraft/server/packs/resources/SimpleJsonResourceReloadListener.java
@@ -42,7 +_,8 @@
 
     protected Map<Identifier, T> prepare(ResourceManager manager, ProfilerFiller profiler) {
         Map<Identifier, T> result = new HashMap<>();
-        scanDirectory(manager, this.lister, this.ops, this.codec, result);
+        // Neo: add condition context
+        scanDirectory(manager, this.lister, this.makeConditionalOps(this.ops), this.codec, result);
         return result;
     }
 
@@ -52,22 +_,55 @@
         scanDirectory(manager, FileToIdConverter.registry(registryKey), ops, codec, result);
     }
 
+    public static <T> void scanDirectoryWithOptionalValues(
+            ResourceManager manager,
+            ResourceKey<? extends Registry<T>> registryKey,
+            DynamicOps<JsonElement> ops,
+            Codec<java.util.Optional<T>> codec,
+            Map<Identifier, java.util.Optional<T>> result
+    ) {
+        scanDirectory(manager, FileToIdConverter.registry(registryKey), ops, codec, result);
+    }
+
     public static <T> void scanDirectory(
         ResourceManager manager, FileToIdConverter lister, DynamicOps<JsonElement> ops, Codec<T> codec, Map<Identifier, T> result
     ) {
+        // Neo: collect JSON resources before parsing
+        Map<Identifier, JsonElement> jsons = new HashMap<>();
         for (Entry<Identifier, Resource> entry : lister.listMatchingResources(manager).entrySet()) {
             Identifier location = entry.getKey();
             Identifier id = lister.fileToId(location);
 
             try (Reader reader = entry.getValue().openAsReader()) {
-                codec.parse(ops, StrictJsonParser.parse(reader)).ifSuccess(parsed -> {
-                    if (result.putIfAbsent(id, (T)parsed) != null) {
-                        throw new IllegalStateException("Duplicate data file ignored with ID " + id);
-                    }
-                }).ifError(error -> LOGGER.error("Couldn't parse data file '{}' from '{}': {}", id, location, error));
+                jsons.put(id, com.google.gson.JsonParser.parseReader(reader));
             } catch (IllegalArgumentException | IOException | JsonParseException var14) {
                 LOGGER.error("Couldn't parse data file '{}' from '{}'", id, location, var14);
             }
         }
+
+        if (ops instanceof net.neoforged.neoforge.common.conditions.ConditionalOps<JsonElement> conditionalOps) {
+            String prefix = lister.prefix();
+            String extension = lister.extension();
+            var event = new net.neoforged.neoforge.data.event.ModifyJsonDataEvent(prefix, extension, conditionalOps.lookupProvider, jsons);
+            net.neoforged.fml.ModLoader.postEventWrapContainerInModOrder(event);
+        }
+
+        var conditionalCodec = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodec(codec);
+        for (Entry<Identifier, JsonElement> entry : jsons.entrySet()) {
+            Identifier id = entry.getKey();
+            JsonElement json = entry.getValue();
+
+            conditionalCodec.parse(ops, json).ifSuccess(parsed -> {
+                if (parsed.isEmpty()) {
+                    LOGGER.debug("Skipping loading data file '{}' as its conditions were not met", id);
+                } else if (result.putIfAbsent(id, parsed.get()) != null) {
+                    throw new IllegalStateException("Duplicate data file ignored with ID " + id);
+                }
+            }).ifError(error -> LOGGER.error("Couldn't parse data file '{}': {}", id, error));
+        }
+    }
+
+    protected Identifier getPreparedPath(Identifier rl) {
+        return this.lister.idToFile(rl);
     }
 }
