--- a/net/minecraft/client/renderer/state/LevelRenderState.java
+++ b/net/minecraft/client/renderer/state/LevelRenderState.java
@@ -9,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public class LevelRenderState {
+public class LevelRenderState extends net.neoforged.neoforge.client.renderstate.BaseRenderState {
     public CameraRenderState cameraRenderState = new CameraRenderState();
     public final List<EntityRenderState> entityRenderStates = new ArrayList<>();
     public final List<BlockEntityRenderState> blockEntityRenderStates = new ArrayList<>();
@@ -20,6 +_,9 @@
     public final WorldBorderRenderState worldBorderRenderState = new WorldBorderRenderState();
     public final SkyRenderState skyRenderState = new SkyRenderState();
     public long gameTime;
+    public net.neoforged.neoforge.client.@Nullable CustomSkyboxRenderer customSkyboxRenderer;
+    public net.neoforged.neoforge.client.@Nullable CustomCloudsRenderer customCloudsRenderer;
+    public net.neoforged.neoforge.client.@Nullable CustomWeatherEffectRenderer customWeatherEffectRenderer;
 
     public void reset() {
         this.entityRenderStates.clear();
@@ -31,5 +_,9 @@
         this.worldBorderRenderState.reset();
         this.skyRenderState.reset();
         this.gameTime = 0L;
+        this.customSkyboxRenderer = null;
+        this.customCloudsRenderer = null;
+        this.customWeatherEffectRenderer = null;
+        this.resetRenderData();
     }
 }
