--- a/net/minecraft/client/renderer/state/MapRenderState.java
+++ b/net/minecraft/client/renderer/state/MapRenderState.java
@@ -10,11 +_,12 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public class MapRenderState {
+public class MapRenderState extends net.neoforged.neoforge.client.renderstate.BaseRenderState {
     public @Nullable Identifier texture;
     public final List<MapRenderState.MapDecorationRenderState> decorations = new ArrayList<>();
 
-    public static class MapDecorationRenderState {
+    public static class MapDecorationRenderState extends net.neoforged.neoforge.client.renderstate.BaseRenderState {
+        public net.minecraft.core.Holder<net.minecraft.world.level.saveddata.maps.MapDecorationType> type;
         public @Nullable TextureAtlasSprite atlasSprite;
         public byte x;
         public byte y;
