--- a/net/minecraft/server/packs/metadata/pack/PackMetadataSection.java
+++ b/net/minecraft/server/packs/metadata/pack/PackMetadataSection.java
@@ -19,4 +_,8 @@
                 .apply(p_337567_, PackMetadataSection::new)
     );
     public static final MetadataSectionType<PackMetadataSection> TYPE = new MetadataSectionType<>("pack", CODEC);
+
+    public PackMetadataSection(Component description, int packVersion) {
+        this(description, packVersion, Optional.empty());
+    }
 }
