--- a/net/minecraft/util/ExtraCodecs.java
+++ b/net/minecraft/util/ExtraCodecs.java
@@ -558,9 +_,13 @@
 
    static class RecursiveCodec<T> implements Codec<T> {
       private final Supplier<Codec<T>> wrapped;
+      // Neo: cache name in the constructor to make sure that we call wrapped.toString() before the memoized supplier is called.
+      // Neo: as of MC 1.20.2, this is necessary for StructureTemplatePool.DIRECT_CODEC.toString() to not cause a stack overflow after loading a world once.
+      private final String name;
 
       RecursiveCodec(Function<Codec<T>, Codec<T>> p_298813_) {
          this.wrapped = Suppliers.memoize(() -> p_298813_.apply(this));
+         this.name = "RecursiveCodec[" + this.wrapped + "]";
       }
 
       @Override
@@ -575,7 +_,7 @@
 
       @Override
       public String toString() {
-         return "RecursiveCodec[" + this.wrapped + "]";
+         return this.name;
       }
    }
 
