--- a/net/minecraft/util/ExtraCodecs.java
+++ b/net/minecraft/util/ExtraCodecs.java
@@ -474,7 +_,7 @@
 
     public static <T> Codec<HolderSet<T>> nonEmptyHolderSet(Codec<HolderSet<T>> listCodec) {
         return listCodec.validate(
-            list -> list.unwrap().right().filter(List::isEmpty).isPresent() ? DataResult.error(() -> "List must have contents") : DataResult.success(list)
+            list -> list.isImmediatelyResolvable() && list.unwrap().right().filter(List::isEmpty).isPresent() ? DataResult.error(() -> "List must have contents") : DataResult.success(list)
         );
     }
 
@@ -700,6 +_,11 @@
 
         public Set<V> values() {
             return Collections.unmodifiableSet(this.idToValue.values());
+        }
+
+        /// Neo: retrieve a value by ID from this [LateBoundIdMapper]
+        public @Nullable V getValue(I id) {
+            return this.idToValue.get(id);
         }
     }
 
