--- a/net/minecraft/util/SpawnUtil.java
+++ b/net/minecraft/util/SpawnUtil.java
@@ -35,7 +35,7 @@
             && moveToPossibleSpawnPosition(p_216406_, p_216410_, blockpos$mutableblockpos, p_216411_)) {
             T t = p_216404_.create(p_216406_, null, null, blockpos$mutableblockpos, p_216405_, false, false);
             if (t != null) {
-               if (t.checkSpawnRules(p_216406_, p_216405_) && t.checkSpawnObstruction(p_216406_)) {
+               if (net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(t, p_216406_, p_216405_)) {
                   p_216406_.addFreshEntityWithPassengers(t);
                   return Optional.of(t);
                }
