--- a/net/minecraft/util/SpawnUtil.java
+++ b/net/minecraft/util/SpawnUtil.java
@@ -35,7 +_,7 @@
                 && moveToPossibleSpawnPosition(level, yOffset, blockpos$mutableblockpos, strategy)) {
                 T t = (T)entityType.create(level, null, blockpos$mutableblockpos, spawnType, false, false);
                 if (t != null) {
-                    if (t.checkSpawnRules(level, spawnType) && t.checkSpawnObstruction(level)) {
+                    if (net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(t, level, spawnType)) {
                         level.addFreshEntityWithPassengers(t);
                         return Optional.of(t);
                     }
