--- a/net/minecraft/util/SpawnUtil.java
+++ b/net/minecraft/util/SpawnUtil.java
@@ -47,7 +_,7 @@
                 )) {
                 T t = (T)p_216404_.create(p_216406_, null, blockpos$mutableblockpos, p_364255_, false, false);
                 if (t != null) {
-                    if (t.checkSpawnRules(p_216406_, p_364255_) && t.checkSpawnObstruction(p_216406_)) {
+                    if (net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(t, p_216406_, p_364255_)) {
                         p_216406_.addFreshEntityWithPassengers(t);
                         t.playAmbientSound();
                         return Optional.of(t);
