--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -271,8 +_,8 @@
                 .getSchema(DataFixUtils.makeKey(SharedConstants.getCurrentVersion().dataVersion().version()))
                 .getChoiceType(reference, name);
         } catch (IllegalArgumentException var4) {
-            LOGGER.error("No data fixer registered for {}", name);
-            if (SharedConstants.IS_RUNNING_IN_IDE) {
+            LOGGER.debug("No data fixer registered for {}", name);
+            if (SharedConstants.IS_RUNNING_IN_IDE && false) {
                 throw var4;
             }
         }
@@ -682,20 +_,20 @@
 
     public static void logAndPauseIfInIde(String message) {
         LOGGER.error(message);
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             doPause(message);
         }
     }
 
     public static void logAndPauseIfInIde(String message, Throwable throwable) {
         LOGGER.error(message, throwable);
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             doPause(message);
         }
     }
 
     public static <T extends Throwable> T pauseInIde(T t) {
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             LOGGER.error("Trying to throw a fatal exception, pausing in IDE", t);
             doPause(t.getMessage());
         }
