--- a/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
@@ -80,7 +_,7 @@
         .stream()
         .flatMap(block -> block.getStateDefinition().getPossibleStates().stream())
         .collect(ImmutableSet.toImmutableSet());
-    private static final Map<BlockState, Holder<PoiType>> TYPE_BY_STATE = Maps.newHashMap();
+    private static final Map<BlockState, Holder<PoiType>> TYPE_BY_STATE = net.neoforged.neoforge.registries.GameData.getBlockStatePointOfInterestTypeMap();
 
     private static Set<BlockState> getBlockStates(Block block) {
         return ImmutableSet.copyOf(block.getStateDefinition().getPossibleStates());
@@ -98,6 +_,7 @@
     }
 
     private static void registerBlockStates(Holder<PoiType> type, Set<BlockState> matchingStates) {
+        // Neo: we do this automatically for modded PoiTypes in NeoForgeRegistryCallbacks
         matchingStates.forEach(
             blockState -> {
                 Holder<PoiType> previous = TYPE_BY_STATE.put(blockState, type);
