--- a/net/minecraft/world/entity/ai/village/poi/PoiType.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiType.java
@@ -9,7 +_,7 @@
     public static final Predicate<Holder<PoiType>> NONE = poiType -> false;
 
     public PoiType {
-        matchingStates = Set.copyOf(matchingStates);
+        matchingStates = new net.neoforged.neoforge.common.world.poi.PoiStateSet(matchingStates);
     }
 
     public boolean is(BlockState state) {
