--- a/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiTypes.java
@@ -66,7 +_,7 @@
       .stream()
       .flatMap(p_218093_ -> p_218093_.getStateDefinition().getPossibleStates().stream())
       .collect(ImmutableSet.toImmutableSet());
-   private static final Map<BlockState, Holder<PoiType>> TYPE_BY_STATE = Maps.newHashMap();
+   private static final Map<BlockState, Holder<PoiType>> TYPE_BY_STATE = net.neoforged.neoforge.registries.GameData.getBlockStatePointOfInterestTypeMap();
 
    private static Set<BlockState> getBlockStates(Block p_218074_) {
       return ImmutableSet.copyOf(p_218074_.getStateDefinition().getPossibleStates());
@@ -84,6 +_,7 @@
    }
 
    private static void registerBlockStates(Holder<PoiType> p_250815_, Set<BlockState> p_250679_) {
+      // Neo: we do this automatically for modded PoiTypes in NeoForgeRegistryCallbacks
       p_250679_.forEach(
          p_218081_ -> {
             Holder<PoiType> holder = TYPE_BY_STATE.put(p_218081_, p_250815_);
