--- a/net/minecraft/network/CompressionEncoder.java
+++ b/net/minecraft/network/CompressionEncoder.java
@@ -9,6 +_,8 @@
     private final byte[] encodeBuf = new byte[8192];
     private final Deflater deflater;
     private int threshold;
+    private static final boolean DISABLE_PACKET_DEBUG = Boolean.parseBoolean(System.getProperty("neoforge.disablePacketCompressionDebug", "false"));
+    private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger();
 
     public CompressionEncoder(int p_129448_) {
         this.threshold = p_129448_;
@@ -17,6 +_,12 @@
 
     protected void encode(ChannelHandlerContext p_129452_, ByteBuf p_129453_, ByteBuf p_129454_) {
         int i = p_129453_.readableBytes();
+        if (!DISABLE_PACKET_DEBUG && i > net.minecraft.network.CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH) {
+            p_129453_.markReaderIndex();
+            LOGGER.error("Attempted to send packet over maximum protocol size: {} > {}\nData:\n{}", i, net.minecraft.network.CompressionDecoder.MAXIMUM_UNCOMPRESSED_LENGTH,
+                    net.neoforged.neoforge.logging.PacketDump.getContentDump(p_129453_));
+            p_129453_.resetReaderIndex();
+        }
         if (i > 8388608) {
             throw new IllegalArgumentException("Packet too big (is " + i + ", should be less than 8388608)");
         } else {
