--- a/net/minecraft/network/PacketBundleUnpacker.java
+++ b/net/minecraft/network/PacketBundleUnpacker.java
@@ -20,7 +_,7 @@
         if (bundlerinfo$provider == null) {
             throw new EncoderException("Bundler not configured: " + p_265038_);
         } else {
-            bundlerinfo$provider.bundlerInfo().unbundlePacket(p_265038_, p_265735_::add);
+            bundlerinfo$provider.bundlerInfo().unbundlePacket(p_265038_, p_265735_::add, p_265691_);
         }
     }
 }
