--- a/net/minecraft/network/PacketDecoder.java
+++ b/net/minecraft/network/PacketDecoder.java
@@ -23,6 +_,8 @@
     protected void decode(ChannelHandlerContext p_130535_, ByteBuf p_130536_, List<Object> p_130537_) throws Exception {
         int i = p_130536_.readableBytes();
         if (i != 0) {
+            // Neo: Wrap buffer with ChannelHandlerContext
+            p_130536_ = new net.neoforged.neoforge.network.connection.ChannelAwareFriendlyByteBuf(p_130536_, p_130535_);
             Packet<? super T> packet = this.protocolInfo.codec().decode(p_130536_);
             PacketType<? extends Packet<? super T>> packettype = packet.type();
             JvmProfiler.INSTANCE.onPacketReceived(this.protocolInfo.id(), packettype, p_130535_.channel().remoteAddress(), i);
