--- a/net/minecraft/network/PacketBundleUnpacker.java
+++ b/net/minecraft/network/PacketBundleUnpacker.java
@@ -14,7 +_,7 @@
     }
 
     protected void encode(ChannelHandlerContext context, Packet<?> packet, List<Object> p_265735_) throws Exception {
-        this.bundlerInfo.unbundlePacket(packet, p_265735_::add);
+        this.bundlerInfo.unbundlePacket(packet, p_265735_::add, context);
         if (packet.isTerminal()) {
             context.pipeline().remove(context.name());
         }
