--- a/net/minecraft/network/RegistryFriendlyByteBuf.java
+++ b/net/minecraft/network/RegistryFriendlyByteBuf.java
@@ -6,16 +_,40 @@
 
 public class RegistryFriendlyByteBuf extends FriendlyByteBuf {
     private final RegistryAccess registryAccess;
+    private final net.neoforged.neoforge.network.connection.ConnectionType connectionType;
 
+    /**
+     * @deprecated Neo: use overload with ConnectionType context
+     */
+    @Deprecated
     public RegistryFriendlyByteBuf(ByteBuf source, RegistryAccess registryAccess) {
         super(source);
         this.registryAccess = registryAccess;
+        this.connectionType = net.neoforged.neoforge.network.connection.ConnectionType.OTHER;
+    }
+
+    public RegistryFriendlyByteBuf(ByteBuf source, RegistryAccess registryAccess, net.neoforged.neoforge.network.connection.ConnectionType connectionType) {
+        super(source);
+        this.registryAccess = registryAccess;
+        this.connectionType = connectionType;
+    }
+
+    public net.neoforged.neoforge.network.connection.ConnectionType getConnectionType() {
+        return this.connectionType;
     }
 
     public RegistryAccess registryAccess() {
         return this.registryAccess;
     }
 
+    public static Function<ByteBuf, RegistryFriendlyByteBuf> decorator(RegistryAccess registry, net.neoforged.neoforge.network.connection.ConnectionType connectionType) {
+        return p_320793_ -> new RegistryFriendlyByteBuf(p_320793_, registry, connectionType);
+    }
+
+    /**
+     * @deprecated Neo: use overload with ConnectionType context
+     */
+    @Deprecated
     public static Function<ByteBuf, RegistryFriendlyByteBuf> decorator(RegistryAccess registry) {
         return p_320793_ -> new RegistryFriendlyByteBuf(p_320793_, registry);
     }
