--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -222,9 +229,10 @@
 
     private static void consumeFuel(NonNullList<ItemStack> items, ItemStack fuel) {
         Item fuelItem = fuel.getItem();
+        // Neo: Query the crafting remainder prior to stack mutation (shrink) in order to grab the correct value
+        ItemStackTemplate remainder = fuel.getCraftingRemainder();
         fuel.shrink(1);
         if (fuel.isEmpty()) {
-            ItemStackTemplate remainder = fuelItem.getCraftingRemainder();
             items.set(1, remainder != null ? remainder.create() : ItemStack.EMPTY);
         }
     }
