--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -54,6 +_,7 @@
     public static final int BURN_TIME_STANDARD = 200;
     public static final int BURN_COOL_SPEED = 2;
     public static final int UNKNOWN_LIT_DURATION = 0;
+    private final RecipeType<? extends AbstractCookingRecipe> recipeType;
     protected NonNullList<ItemStack> items = NonNullList.withSize(3, ItemStack.EMPTY);
     int litTime;
     int litDuration = 0;
@@ -64,9 +_,14 @@
         public int get(int p_58431_) {
             switch (p_58431_) {
                 case 0:
+                    if (litDuration > Short.MAX_VALUE) {
+                        // Neo: preserve litTime / litDuration ratio on the client as data slots are synced as shorts.
+                        return net.minecraft.util.Mth.floor(((double) litTime / litDuration) * Short.MAX_VALUE);
+                    }
+
                     return AbstractFurnaceBlockEntity.this.litTime;
                 case 1:
-                    return AbstractFurnaceBlockEntity.this.litDuration;
+                    return Math.min(AbstractFurnaceBlockEntity.this.litDuration, Short.MAX_VALUE);
                 case 2:
                     return AbstractFurnaceBlockEntity.this.cookingProgress;
                 case 3:
@@ -106,6 +_,7 @@
     ) {
         super(p_154991_, p_154992_, p_154993_);
         this.quickCheck = RecipeManager.createCheck((RecipeType<AbstractCookingRecipe>)p_154994_);
+        this.recipeType = p_154994_;
     }
 
     private boolean isLit() {
@@ -117,9 +_,9 @@
         super.loadAdditional(p_155025_, p_323468_);
         this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
         ContainerHelper.loadAllItems(p_155025_, this.items, p_323468_);
-        this.litTime = p_155025_.getShort("BurnTime");
-        this.cookingProgress = p_155025_.getShort("CookTime");
-        this.cookingTotalTime = p_155025_.getShort("CookTimeTotal");
+        this.litTime = p_155025_.getInt("BurnTime");
+        this.cookingProgress = p_155025_.getInt("CookTime");
+        this.cookingTotalTime = p_155025_.getInt("CookTimeTotal");
         this.litDuration = 0;
         CompoundTag compoundtag = p_155025_.getCompound("RecipesUsed");
 
@@ -131,9 +_,9 @@
     @Override
     protected void saveAdditional(CompoundTag p_187452_, HolderLookup.Provider p_323656_) {
         super.saveAdditional(p_187452_, p_323656_);
-        p_187452_.putShort("BurnTime", (short)this.litTime);
-        p_187452_.putShort("CookTime", (short)this.cookingProgress);
-        p_187452_.putShort("CookTimeTotal", (short)this.cookingTotalTime);
+        p_187452_.putInt("BurnTime", this.litTime);
+        p_187452_.putInt("CookTime", this.cookingProgress);
+        p_187452_.putInt("CookTimeTotal", this.cookingTotalTime);
         ContainerHelper.saveAllItems(p_187452_, this.items, p_323656_);
         CompoundTag compoundtag = new CompoundTag();
         this.recipesUsed.forEach((p_380898_, p_380899_) -> compoundtag.putInt(p_380898_.location().toString(), p_380899_));
@@ -170,11 +_,15 @@
                 p_155017_.litDuration = p_155017_.litTime;
                 if (p_155017_.isLit()) {
                     flag1 = true;
+                    var remainder = itemstack.getCraftingRemainder();
+                    if (!remainder.isEmpty())
+                        p_155017_.items.set(1, remainder);
+                    else
                     if (flag3) {
                         Item item = itemstack.getItem();
                         itemstack.shrink(1);
                         if (itemstack.isEmpty()) {
-                            p_155017_.items.set(1, item.getCraftingRemainder());
+                            p_155017_.items.set(1, item.getCraftingRemainder()); // Neo: Remainder is handled in the `if` check above.
                         }
                     }
                 }
@@ -227,9 +_,9 @@
                 } else if (!ItemStack.isSameItemSameComponents(itemstack1, itemstack)) {
                     return false;
                 } else {
-                    return itemstack1.getCount() < p_155008_ && itemstack1.getCount() < itemstack1.getMaxStackSize()
+                    return itemstack1.getCount() + itemstack.getCount() <= p_155008_ && itemstack1.getCount() + itemstack.getCount() <= itemstack1.getMaxStackSize() // Neo fix: make furnace respect stack sizes in furnace recipes
                         ? true
-                        : itemstack1.getCount() < itemstack.getMaxStackSize();
+                        : itemstack1.getCount() + itemstack.getCount() <= itemstack.getMaxStackSize(); // Neo fix: make furnace respect stack sizes in furnace recipes
                 }
             }
         } else {
@@ -251,7 +_,7 @@
             if (itemstack2.isEmpty()) {
                 p_267073_.set(2, itemstack1.copy());
             } else if (ItemStack.isSameItemSameComponents(itemstack2, itemstack1)) {
-                itemstack2.grow(1);
+                itemstack2.grow(itemstack1.getCount());
             }
 
             if (itemstack.is(Blocks.WET_SPONGE.asItem()) && !p_267073_.get(1).isEmpty() && p_267073_.get(1).is(Items.BUCKET)) {
@@ -266,7 +_,7 @@
     }
 
     protected int getBurnDuration(FuelValues p_363501_, ItemStack p_58343_) {
-        return p_363501_.burnDuration(p_58343_);
+        return p_58343_.getBurnTime(this.recipeType, p_363501_);
     }
 
     private static int getTotalCookTime(ServerLevel p_380169_, AbstractFurnaceBlockEntity p_222694_) {
@@ -329,7 +_,7 @@
             return true;
         } else {
             ItemStack itemstack = this.items.get(1);
-            return this.level.fuelValues().isFuel(p_58390_) || p_58390_.is(Items.BUCKET) && !itemstack.is(Items.BUCKET);
+            return p_58390_.getBurnTime(this.recipeType, this.level.fuelValues()) > 0 || p_58390_.is(Items.BUCKET) && !itemstack.is(Items.BUCKET);
         }
     }
 
