--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -60,6 +_,7 @@
     public static final int NUM_DATA_VALUES = 4;
     public static final int BURN_TIME_STANDARD = 200;
     public static final int BURN_COOL_SPEED = 2;
+    private final RecipeType<? extends AbstractCookingRecipe> recipeType;
     protected NonNullList<ItemStack> items = NonNullList.withSize(3, ItemStack.EMPTY);
     int litTime;
     int litDuration;
@@ -72,9 +_,14 @@
         public int get(int p_58431_) {
             switch (p_58431_) {
                 case 0:
+                    if (litDuration > Short.MAX_VALUE) {
+                        // Neo: preserve litTime / litDuration ratio on the client as data slots are synced as shorts.
+                        return net.minecraft.util.Mth.floor(((double) litTime / litDuration) * Short.MAX_VALUE);
+                    }
+
                     return AbstractFurnaceBlockEntity.this.litTime;
                 case 1:
-                    return AbstractFurnaceBlockEntity.this.litDuration;
+                    return Math.min(AbstractFurnaceBlockEntity.this.litDuration, Short.MAX_VALUE);
                 case 2:
                     return AbstractFurnaceBlockEntity.this.cookingProgress;
                 case 3:
@@ -114,18 +_,40 @@
     ) {
         super(type, pos, blockState);
         this.quickCheck = RecipeManager.createCheck((RecipeType<AbstractCookingRecipe>)recipeType);
+        this.recipeType = recipeType;
     }
 
     public static void invalidateCache() {
         fuelCache = null;
     }
 
+    /**
+     * @deprecated Neo: get burn times by calling {@link net.neoforged.neoforge.common.extensions.IItemStackExtension#getBurnTime(RecipeType)}
+     */
+    @Deprecated
     public static Map<Item, Integer> getFuel() {
         Map<Item, Integer> map = fuelCache;
         if (map != null) {
             return map;
         } else {
             Map<Item, Integer> map1 = Maps.newLinkedHashMap();
+            buildFuels((e, time) -> e.ifRight(tag -> add(map1, tag, time)).ifLeft(item -> add(map1, item, time)));
+            fuelCache = map1;
+            return map1;
+        }
+    }
+
+    private static void add(java.util.function.ObjIntConsumer<com.mojang.datafixers.util.Either<Item, TagKey<Item>>> consumer, ItemLike item, int time) {
+        consumer.accept(com.mojang.datafixers.util.Either.left(item.asItem()), time);
+    }
+
+    private static void add(java.util.function.ObjIntConsumer<com.mojang.datafixers.util.Either<Item, TagKey<Item>>> consumer, TagKey<Item> tag, int time) {
+        consumer.accept(com.mojang.datafixers.util.Either.right(tag), time);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public static void buildFuels(java.util.function.ObjIntConsumer<com.mojang.datafixers.util.Either<Item, TagKey<Item>>> map1) {
+        {
             add(map1, Items.LAVA_BUCKET, 20000);
             add(map1, Blocks.COAL_BLOCK, 16000);
             add(map1, Items.BLAZE_ROD, 2400);
@@ -185,8 +_,6 @@
             add(map1, Blocks.AZALEA, 100);
             add(map1, Blocks.FLOWERING_AZALEA, 100);
             add(map1, Blocks.MANGROVE_ROOTS, 300);
-            fuelCache = map1;
-            return map1;
         }
     }
 
@@ -226,9 +_,9 @@
         super.loadAdditional(tag, registries);
         this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
         ContainerHelper.loadAllItems(tag, this.items, registries);
-        this.litTime = tag.getShort("BurnTime");
-        this.cookingProgress = tag.getShort("CookTime");
-        this.cookingTotalTime = tag.getShort("CookTimeTotal");
+        this.litTime = tag.getInt("BurnTime");
+        this.cookingProgress = tag.getInt("CookTime");
+        this.cookingTotalTime = tag.getInt("CookTimeTotal");
         this.litDuration = this.getBurnDuration(this.items.get(1));
         CompoundTag compoundtag = tag.getCompound("RecipesUsed");
 
@@ -240,9 +_,9 @@
     @Override
     protected void saveAdditional(CompoundTag tag, HolderLookup.Provider registries) {
         super.saveAdditional(tag, registries);
-        tag.putShort("BurnTime", (short)this.litTime);
-        tag.putShort("CookTime", (short)this.cookingProgress);
-        tag.putShort("CookTimeTotal", (short)this.cookingTotalTime);
+        tag.putInt("BurnTime", this.litTime);
+        tag.putInt("CookTime", this.cookingProgress);
+        tag.putInt("CookTimeTotal", this.cookingTotalTime);
         ContainerHelper.saveAllItems(tag, this.items, registries);
         CompoundTag compoundtag = new CompoundTag();
         this.recipesUsed.forEach((p_187449_, p_187450_) -> compoundtag.putInt(p_187449_.toString(), p_187450_));
@@ -269,28 +_,30 @@
             }
 
             int i = blockEntity.getMaxStackSize();
-            if (!blockEntity.isLit() && canBurn(level.registryAccess(), recipeholder, blockEntity.items, i)) {
+            if (!blockEntity.isLit() && canBurn(level.registryAccess(), recipeholder, blockEntity.items, i, blockEntity)) {
                 blockEntity.litTime = blockEntity.getBurnDuration(itemstack);
                 blockEntity.litDuration = blockEntity.litTime;
                 if (blockEntity.isLit()) {
                     flag1 = true;
+                    if (itemstack.hasCraftingRemainingItem())
+                        blockEntity.items.set(1, itemstack.getCraftingRemainingItem());
+                    else
                     if (flag3) {
                         Item item = itemstack.getItem();
                         itemstack.shrink(1);
                         if (itemstack.isEmpty()) {
-                            Item item1 = item.getCraftingRemainingItem();
-                            blockEntity.items.set(1, item1 == null ? ItemStack.EMPTY : new ItemStack(item1));
+                            blockEntity.items.set(1, itemstack.getCraftingRemainingItem());
                         }
                     }
                 }
             }
 
-            if (blockEntity.isLit() && canBurn(level.registryAccess(), recipeholder, blockEntity.items, i)) {
+            if (blockEntity.isLit() && canBurn(level.registryAccess(), recipeholder, blockEntity.items, i, blockEntity)) {
                 blockEntity.cookingProgress++;
                 if (blockEntity.cookingProgress == blockEntity.cookingTotalTime) {
                     blockEntity.cookingProgress = 0;
                     blockEntity.cookingTotalTime = getTotalCookTime(level, blockEntity);
-                    if (burn(level.registryAccess(), recipeholder, blockEntity.items, i)) {
+                    if (burn(level.registryAccess(), recipeholder, blockEntity.items, i, blockEntity)) {
                         blockEntity.setRecipeUsed(recipeholder);
                     }
 
@@ -314,9 +_,9 @@
         }
     }
 
-    private static boolean canBurn(RegistryAccess registryAccess, @Nullable RecipeHolder<?> recipe, NonNullList<ItemStack> inventory, int maxStackSize) {
+    private static boolean canBurn(RegistryAccess registryAccess, @Nullable RecipeHolder<?> recipe, NonNullList<ItemStack> inventory, int maxStackSize, AbstractFurnaceBlockEntity furnace) {
         if (!inventory.get(0).isEmpty() && recipe != null) {
-            ItemStack itemstack = recipe.value().getResultItem(registryAccess);
+            ItemStack itemstack = ((RecipeHolder<? extends AbstractCookingRecipe>) recipe).value().assemble(new SingleRecipeInput(furnace.getItem(0)), registryAccess);
             if (itemstack.isEmpty()) {
                 return false;
             } else {
@@ -326,9 +_,9 @@
                 } else if (!ItemStack.isSameItemSameComponents(itemstack1, itemstack)) {
                     return false;
                 } else {
-                    return itemstack1.getCount() < maxStackSize && itemstack1.getCount() < itemstack1.getMaxStackSize()
+                    return itemstack1.getCount() + itemstack.getCount() <= maxStackSize && itemstack1.getCount() + itemstack.getCount() <= itemstack1.getMaxStackSize() // Neo fix: make furnace respect stack sizes in furnace recipes
                         ? true
-                        : itemstack1.getCount() < itemstack.getMaxStackSize();
+                        : itemstack1.getCount() + itemstack.getCount() <= itemstack.getMaxStackSize(); // Neo fix: make furnace respect stack sizes in furnace recipes
                 }
             }
         } else {
@@ -336,15 +_,15 @@
         }
     }
 
-    private static boolean burn(RegistryAccess registryAccess, @Nullable RecipeHolder<?> recipe, NonNullList<ItemStack> inventory, int maxStackSize) {
-        if (recipe != null && canBurn(registryAccess, recipe, inventory, maxStackSize)) {
+    private static boolean burn(RegistryAccess registryAccess, @Nullable RecipeHolder<?> recipe, NonNullList<ItemStack> inventory, int maxStackSize, AbstractFurnaceBlockEntity furnace) {
+        if (recipe != null && canBurn(registryAccess, recipe, inventory, maxStackSize, furnace)) {
             ItemStack itemstack = inventory.get(0);
-            ItemStack itemstack1 = recipe.value().getResultItem(registryAccess);
+            ItemStack itemstack1 = ((RecipeHolder<? extends AbstractCookingRecipe>) recipe).value().assemble(new SingleRecipeInput(furnace.getItem(0)), registryAccess);
             ItemStack itemstack2 = inventory.get(2);
             if (itemstack2.isEmpty()) {
                 inventory.set(2, itemstack1.copy());
             } else if (ItemStack.isSameItemSameComponents(itemstack2, itemstack1)) {
-                itemstack2.grow(1);
+                itemstack2.grow(itemstack1.getCount());
             }
 
             if (itemstack.is(Blocks.WET_SPONGE.asItem()) && !inventory.get(1).isEmpty() && inventory.get(1).is(Items.BUCKET)) {
@@ -362,8 +_,7 @@
         if (fuel.isEmpty()) {
             return 0;
         } else {
-            Item item = fuel.getItem();
-            return getFuel().getOrDefault(item, 0);
+            return fuel.getBurnTime(this.recipeType);
         }
     }
 
@@ -373,7 +_,7 @@
     }
 
     public static boolean isFuel(ItemStack stack) {
-        return getFuel().containsKey(stack.getItem());
+        return stack.getBurnTime(null) > 0;
     }
 
     @Override
@@ -431,7 +_,7 @@
             return true;
         } else {
             ItemStack itemstack = this.items.get(1);
-            return isFuel(stack) || stack.is(Items.BUCKET) && !itemstack.is(Items.BUCKET);
+            return stack.getBurnTime(this.recipeType) > 0 || stack.is(Items.BUCKET) && !itemstack.is(Items.BUCKET);
         }
     }
 
