--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -118,9 +_,17 @@
 
     @Override
     public void setItem(int slot, ItemStack itemStack) {
+        setItem(slot, itemStack, false);
+    }
+
+    // Neo: Skip side-effects if insideTransaction is true so the caller can defer them until the transaction commits
+    @Override
+    public void setItem(int slot, ItemStack itemStack, boolean insideTransaction) {
         this.getItems().set(slot, itemStack);
         itemStack.limitSize(this.getMaxStackSize(itemStack));
-        this.setChanged();
+        if (!insideTransaction) {
+            this.setChanged();
+        }
     }
 
     @Override
