--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -124,9 +_,17 @@
 
     @Override
     public void setItem(int slot, ItemStack itemStack) {
+        setItem(slot, itemStack, false);
+    }
+
+    // Neo: Skip side-effects if insideTransaction is true so the caller can defer them until the transaction commits
+    @Override
+    public void setItem(int slot, ItemStack itemStack, boolean insideTransaction) {
         this.getItems().set(slot, itemStack);
         itemStack.limitSize(this.getMaxStackSize(itemStack));
-        this.setChanged();
+        if (!insideTransaction) {
+            this.setChanged();
+        }
     }
 
     @Override
@@ -178,12 +_,18 @@
     }
 
     protected LootContext getLootContext(ServerLevel level) {
+        return this.getLootContext(level, ItemStack.EMPTY);
+    }
+
+    /// Neo: add queried stack for cases where the loot context is used to query a data component which may want to access other components on the stack.
+    protected LootContext getLootContext(ServerLevel level, ItemStack queriedStack) {
         return new LootContext.Builder(
                 new LootParams.Builder(level)
                     .withParameter(LootContextParams.BLOCK_STATE, this.getBlockState())
                     .withParameter(LootContextParams.BLOCK_ENTITY, this)
                     .withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(this.getBlockPos()))
                     .withParameter(LootContextParams.CONTAINER, this)
+                    .withOptionalParameter(net.neoforged.neoforge.common.loot.NeoForgeLootContextParams.QUERIED_STACK, queriedStack.isEmpty() ? null : queriedStack)
                     .create(LootContextParamSets.CONTAINER_PROCESS)
             )
             .create(Optional.empty());
