--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -137,8 +_,8 @@
 
         for (int i1 = 0; i1 < 10 && blockpos.getY() <= l; i1++) {
             BlockState blockstate = level.getBlockState(blockpos);
-            if (blockstate.getBlock() instanceof BeaconBeamBlock beaconbeamblock) {
-                int j1 = beaconbeamblock.getColor().getTextureDiffuseColor();
+            Integer j1 = blockstate.getBeaconColorMultiplier(level, blockpos, pos);
+            if (j1 != null) {
                 if (blockEntity.checkingBeamSections.size() <= 1) {
                     beaconblockentity$beaconbeamsection = new BeaconBlockEntity.BeaconBeamSection(j1);
                     blockEntity.checkingBeamSections.add(beaconblockentity$beaconbeamsection);
