--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -32,9 +_,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.slf4j.Logger;
 
-public abstract class BlockEntity implements net.neoforged.neoforge.common.extensions.IBlockEntityExtension {
+public abstract class BlockEntity extends net.neoforged.neoforge.attachment.AttachmentHolder implements net.neoforged.neoforge.common.extensions.IBlockEntityExtension {
     private static final Codec<BlockEntityType<?>> TYPE_CODEC = BuiltInRegistries.BLOCK_ENTITY_TYPE.byNameCodec();
     private static final Logger LOGGER = LogUtils.getLogger();
+    @Deprecated // Neo: always use getType()
     private final BlockEntityType<?> type;
     @Nullable
     protected Level level;
@@ -42,6 +_,8 @@
     protected boolean remove;
     private BlockState blockState;
     private DataComponentMap components = DataComponentMap.EMPTY;
+    @Nullable
+    private CompoundTag customPersistentData;
 
     public BlockEntity(BlockEntityType<?> p_155228_, BlockPos p_155229_, BlockState p_155230_) {
         this.type = p_155228_;
@@ -57,7 +_,7 @@
     }
 
     public boolean isValidBlockState(BlockState p_353131_) {
-        return this.type.isValid(p_353131_);
+        return this.getType().isValid(p_353131_); // Neo: use getter so correct type is checked for modded subclasses
     }
 
     public static BlockPos getPosFromTag(ChunkPos p_412201_, CompoundTag p_187473_) {
@@ -89,6 +_,8 @@
     }
 
     protected void loadAdditional(CompoundTag p_338466_, HolderLookup.Provider p_338445_) {
+        p_338466_.getCompound("NeoForgeData").ifPresent(neoData -> this.customPersistentData = neoData);
+        p_338466_.getCompound(ATTACHMENTS_NBT_KEY).ifPresent(attachments -> this.deserializeAttachments(p_338445_, attachments));
     }
 
     public final void loadWithComponents(CompoundTag p_338356_, HolderLookup.Provider p_338558_) {
@@ -102,6 +_,9 @@
     }
 
     protected void saveAdditional(CompoundTag p_187471_, HolderLookup.Provider p_323635_) {
+        if (this.customPersistentData != null) p_187471_.put("NeoForgeData", this.customPersistentData.copy());
+        var attachmentsTag = serializeAttachments(p_323635_);
+        if (attachmentsTag != null) p_187471_.put(ATTACHMENTS_NBT_KEY, attachmentsTag);
     }
 
     public final CompoundTag saveWithFullMetadata(HolderLookup.Provider p_323767_) {
@@ -211,10 +_,14 @@
 
     public void setRemoved() {
         this.remove = true;
+        this.invalidateCapabilities();
+        requestModelDataUpdate();
     }
 
     public void clearRemoved() {
         this.remove = false;
+        // Neo: invalidate capabilities on block entity placement
+        invalidateCapabilities();
     }
 
     public void preRemoveSideEffects(BlockPos p_394577_, BlockState p_394161_) {
@@ -244,6 +_,27 @@
 
     public BlockEntityType<?> getType() {
         return this.type;
+    }
+
+    @Override
+    public CompoundTag getPersistentData() {
+        if (this.customPersistentData == null)
+            this.customPersistentData = new CompoundTag();
+        return this.customPersistentData;
+    }
+
+    @Override
+    @Nullable
+    public final <T> T setData(net.neoforged.neoforge.attachment.AttachmentType<T> type, T data) {
+        setChanged();
+        return super.setData(type, data);
+    }
+
+    @Override
+    @Nullable
+    public final <T> T removeData(net.neoforged.neoforge.attachment.AttachmentType<T> type) {
+        setChanged();
+        return super.removeData(type);
     }
 
     @Deprecated
