--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -301,6 +_,13 @@
         return this.validBlocks.contains(p_155263_.getBlock());
     }
 
+    /**
+     * Neo: Add getter for an immutable view of the set of valid blocks.
+     */
+    public Set<Block> getValidBlocks() {
+        return java.util.Collections.unmodifiableSet(this.validBlocks);
+    }
+
     @Nullable
     public Holder.Reference<BlockEntityType<?>> builtInRegistryHolder() {
         return this.builtInRegistryHolder;
