--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -27,7 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.slf4j.Logger;
 
-public abstract class BlockEntity {
+public abstract class BlockEntity extends net.neoforged.neoforge.attachment.AttachmentHolder implements net.neoforged.neoforge.common.extensions.IBlockEntityExtension {
     private static final Logger LOGGER = LogUtils.getLogger();
     private final BlockEntityType<?> type;
     @Nullable
@@ -36,6 +_,8 @@
     protected boolean remove;
     private BlockState blockState;
     private DataComponentMap components = DataComponentMap.EMPTY;
+    @Nullable
+    private CompoundTag customPersistentData;
 
     public BlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState blockState) {
         this.type = type;
@@ -61,6 +_,8 @@
     }
 
     protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
+        if (tag.contains("NeoForgeData", net.minecraft.nbt.Tag.TAG_COMPOUND)) this.customPersistentData = tag.getCompound("NeoForgeData");
+        if (tag.contains(ATTACHMENTS_NBT_KEY, net.minecraft.nbt.Tag.TAG_COMPOUND)) deserializeAttachments(registries, tag.getCompound(ATTACHMENTS_NBT_KEY));
     }
 
     public final void loadWithComponents(CompoundTag tag, HolderLookup.Provider registries) {
@@ -76,6 +_,9 @@
     }
 
     protected void saveAdditional(CompoundTag tag, HolderLookup.Provider registries) {
+        if (this.customPersistentData != null) tag.put("NeoForgeData", this.customPersistentData.copy());
+        var attachmentsTag = serializeAttachments(registries);
+        if (attachmentsTag != null) tag.put(ATTACHMENTS_NBT_KEY, attachmentsTag);
     }
 
     public final CompoundTag saveWithFullMetadata(HolderLookup.Provider registries) {
@@ -205,10 +_,14 @@
 
     public void setRemoved() {
         this.remove = true;
+        this.invalidateCapabilities();
+        requestModelDataUpdate();
     }
 
     public void clearRemoved() {
         this.remove = false;
+        // Neo: invalidate capabilities on block entity placement
+        invalidateCapabilities();
     }
 
     public boolean triggerEvent(int id, int type) {
@@ -231,6 +_,27 @@
         return this.type;
     }
 
+    @Override
+    public CompoundTag getPersistentData() {
+        if (this.customPersistentData == null)
+            this.customPersistentData = new CompoundTag();
+        return this.customPersistentData;
+    }
+
+    @Override
+    @Nullable
+    public final <T> T setData(net.neoforged.neoforge.attachment.AttachmentType<T> type, T data) {
+        setChanged();
+        return super.setData(type, data);
+    }
+
+    @Override
+    @Nullable
+    public final <T> T removeData(net.neoforged.neoforge.attachment.AttachmentType<T> type) {
+        setChanged();
+        return super.removeData(type);
+    }
+
     @Deprecated
     public void setBlockState(BlockState blockState) {
         this.blockState = blockState;
@@ -309,5 +_,15 @@
         <T> T get(DataComponentType<T> component);
 
         <T> T getOrDefault(DataComponentType<? extends T> component, T defaultValue);
+
+        // Neo: Utility for modded component types, to remove the need to invoke '.value()'
+        @Nullable
+        default <T> T get(java.util.function.Supplier<? extends DataComponentType<T>> componentType) {
+            return get(componentType.get());
+        }
+
+        default <T> T getOrDefault(java.util.function.Supplier<? extends DataComponentType<T>> componentType, T value) {
+            return getOrDefault(componentType.get(), value);
+        }
     }
 }
