--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -171,6 +_,7 @@
     }
 
     private static void doBrew(Level level, BlockPos pos, NonNullList<ItemStack> items) {
+        if (net.neoforged.neoforge.event.EventHooks.onPotionAttemptBrew(items)) return;
         ItemStack ingredient = items.get(3);
         PotionBrewing potionBrewing = level.potionBrewing();
 
@@ -178,8 +_,10 @@
             items.set(dest, potionBrewing.mix(ingredient, items.get(dest)));
         }
 
+        net.neoforged.neoforge.event.EventHooks.onPotionBrewed(items);
+        // Neo: Query the crafting remainder prior to stack mutation (shrink) in order to grab the correct value
+        ItemStackTemplate remainder = ingredient.getCraftingRemainder();
         ingredient.shrink(1);
-        ItemStackTemplate remainder = ingredient.getItem().getCraftingRemainder();
         if (remainder != null) {
             if (ingredient.isEmpty()) {
                 ingredient = remainder.create();
@@ -215,13 +_,13 @@
 
     @Override
     public boolean canPlaceItem(int slot, ItemStack itemStack) {
+        PotionBrewing potionBrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
         if (slot == 3) {
-            PotionBrewing potionBrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
             return potionBrewing.isIngredient(itemStack);
         } else {
             return slot == 4
                 ? itemStack.is(ItemTags.BREWING_FUEL)
-                : (itemStack.is(Items.POTION) || itemStack.is(Items.SPLASH_POTION) || itemStack.is(Items.LINGERING_POTION) || itemStack.is(Items.GLASS_BOTTLE))
+                : (potionBrewing.isInput(itemStack) || itemStack.is(Items.GLASS_BOTTLE))
                     && this.getItem(slot).isEmpty();
         }
     }
