--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -164,6 +_,7 @@
     }
 
     private static void doBrew(Level level, BlockPos pos, NonNullList<ItemStack> items) {
+        if (net.neoforged.neoforge.event.EventHooks.onPotionAttemptBrew(items)) return;
         ItemStack itemstack = items.get(3);
         PotionBrewing potionbrewing = level.potionBrewing();
 
@@ -171,15 +_,17 @@
             items.set(i, potionbrewing.mix(itemstack, items.get(i)));
         }
 
-        itemstack.shrink(1);
-        if (itemstack.getItem().hasCraftingRemainingItem()) {
-            ItemStack itemstack1 = new ItemStack(itemstack.getItem().getCraftingRemainingItem());
+        net.neoforged.neoforge.event.EventHooks.onPotionBrewed(items);
+        if (itemstack.hasCraftingRemainingItem()) {
+            ItemStack itemstack1 = itemstack.getCraftingRemainingItem();
+            itemstack.shrink(1);
             if (itemstack.isEmpty()) {
                 itemstack = itemstack1;
             } else {
                 Containers.dropItemStack(level, (double)pos.getX(), (double)pos.getY(), (double)pos.getZ(), itemstack1);
             }
         }
+        else itemstack.shrink(1);
 
         items.set(3, itemstack);
         level.levelEvent(1035, pos, 0);
@@ -208,13 +_,13 @@
 
     @Override
     public boolean canPlaceItem(int index, ItemStack stack) {
+        PotionBrewing potionbrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
         if (index == 3) {
-            PotionBrewing potionbrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
             return potionbrewing.isIngredient(stack);
         } else {
             return index == 4
                 ? stack.is(Items.BLAZE_POWDER)
-                : (stack.is(Items.POTION) || stack.is(Items.SPLASH_POTION) || stack.is(Items.LINGERING_POTION) || stack.is(Items.GLASS_BOTTLE))
+                : (potionbrewing.isInput(stack) || stack.is(Items.GLASS_BOTTLE))
                     && this.getItem(index).isEmpty();
         }
     }
