--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -102,14 +_,14 @@
             setChanged(p_155286_, p_155287_, p_155288_);
         }
 
-        boolean flag = isBrewable(p_155289_.items);
+        boolean flag = net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.canBrew(p_155286_, p_155289_, INGREDIENT_SLOT, new int[]{0, 1, 2}); // Neo: use BrewingRecipeRegistry
         boolean flag1 = p_155289_.brewTime > 0;
         ItemStack itemstack1 = p_155289_.items.get(3);
         if (flag1) {
             --p_155289_.brewTime;
             boolean flag2 = p_155289_.brewTime == 0;
             if (flag2 && flag) {
-                doBrew(p_155286_, p_155287_, p_155289_.items);
+                net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.brewPotions(p_155286_, p_155287_, p_155289_, INGREDIENT_SLOT, new int[]{0, 1, 2}); // Neo: use BrewingRecipeRegistry
                 setChanged(p_155286_, p_155287_, p_155288_);
             } else if (!flag || !itemstack1.is(p_155289_.ingredient)) {
                 p_155289_.brewTime = 0;
@@ -236,11 +_,11 @@
     @Override
     public boolean canPlaceItem(int p_59017_, ItemStack p_59018_) {
         if (p_59017_ == 3) {
-            return PotionBrewing.isIngredient(p_59018_);
+            return net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(this, p_59018_);
         } else if (p_59017_ == 4) {
             return p_59018_.is(Items.BLAZE_POWDER);
         } else {
-            return (p_59018_.is(Items.POTION) || p_59018_.is(Items.SPLASH_POTION) || p_59018_.is(Items.LINGERING_POTION) || p_59018_.is(Items.GLASS_BOTTLE))
+            return net.neoforged.neoforge.common.brewing.BrewingRecipeRegistry.isValidInput(this, p_59018_)
                 && this.getItem(p_59017_).isEmpty();
         }
     }
