--- a/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -152,10 +_,8 @@
                         BlockPos blockpos1 = pos.offset(j1, k1, l1);
                         BlockState blockstate = level.getBlockState(blockpos1);
 
-                        for (Block block : VALID_BLOCKS) {
-                            if (blockstate.is(block)) {
-                                positions.add(blockpos1);
-                            }
+                        if (blockstate.isConduitFrame(level, blockpos1, pos)) {
+                            positions.add(blockpos1);
                         }
                     }
                 }
