--- a/net/minecraft/world/level/block/entity/FuelValues.java
+++ b/net/minecraft/world/level/block/entity/FuelValues.java
@@ -31,6 +_,10 @@
         return Collections.unmodifiableSequencedSet(this.values.keySet());
     }
 
+    /**
+     * @deprecated Neo: use {@link ItemStack#getBurnTime(net.minecraft.world.item.crafting.RecipeType, FuelValues)} instead
+     */
+    @Deprecated
     public int burnDuration(ItemStack itemStack) {
         return itemStack.isEmpty() ? 0 : this.values.getInt(itemStack.getItem());
     }
@@ -40,7 +_,11 @@
     }
 
     public static FuelValues vanillaBurnTimes(HolderLookup.Provider registries, FeatureFlagSet enabledFeatures, int baseUnit) {
-        return new FuelValues.Builder(registries, enabledFeatures)
+        return vanillaBurnTimes(new FuelValues.Builder(registries, enabledFeatures), baseUnit);
+    }
+
+    public static FuelValues vanillaBurnTimes(FuelValues.Builder builder, int baseUnit) {
+        return builder
             .add(Items.LAVA_BUCKET, baseUnit * 100)
             .add(Blocks.COAL_BLOCK, baseUnit * 8 * 10)
             .add(Items.BLAZE_ROD, baseUnit * 12)
