--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -82,6 +_,41 @@
         this.getItems().set(slot, itemStack);
         itemStack.limitSize(this.getMaxStackSize(itemStack));
     }
+    // Neo: Make the hopper cooldown time transactional (it may change when items are inserted)
+    private final net.neoforged.neoforge.transfer.transaction.SnapshotJournal<Integer> cooldownTimeJournal = new net.neoforged.neoforge.transfer.transaction.SnapshotJournal<>() {
+        @Override
+        protected Integer createSnapshot() {
+            return cooldownTime;
+        }
+        @Override
+        protected void revertToSnapshot(Integer snapshot) {
+            cooldownTime = snapshot;
+        }
+    };
+
+    // Neo: Checks if the hopper was empty just before the insertion of `amountChange` at index `slot`.
+    private boolean wasEmpty(int slot, int amountChange) {
+        if (items.get(slot).getCount() != amountChange) return false;
+        for (int i = 0; i < items.size(); ++i) {
+            if (i != slot && !items.get(i).isEmpty()) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public void onTransfer(int slot, int amountChange, net.neoforged.neoforge.transfer.transaction.TransactionContext transaction) {
+        // If we just went from empty to something in the hopper, reset the cooldown.
+        if (amountChange > 0 && !isOnCustomCooldown() && wasEmpty(slot, amountChange)) {
+            cooldownTimeJournal.updateSnapshots(transaction);
+            // This cooldown is always set to 8 in vanilla with one exception:
+            // Hopper -> Hopper transfer sets this cooldown to 7 when this hopper
+            // has not been updated as recently as the one pushing items into it.
+            // Hopper <-> hopper interactions don't use transactions so we don't need to worry about that.
+            setCooldown(MOVE_ITEM_SPEED);
+        }
+    }
 
     @Override
     public void setBlockState(BlockState blockState) {
@@ -139,10 +_,11 @@
     }
 
     private static boolean ejectItems(Level level, BlockPos blockPos, HopperBlockEntity self) {
-        Container container = getAttachedContainer(level, blockPos, self);
-        if (container == null) {
+        var containerOrHandler = getContainerOrHandlerAt(level, blockPos.relative(self.facing), self.facing.getOpposite());
+        if (containerOrHandler.isEmpty()) {
             return false;
-        } else {
+        } else if (containerOrHandler.container() != null) {
+            Container container = containerOrHandler.container();
             Direction direction = self.facing.getOpposite();
             if (isFullContainer(container, direction)) {
                 return false;
@@ -166,6 +_,8 @@
 
                 return false;
             }
+        } else {
+            return net.neoforged.neoforge.transfer.item.VanillaInventoryCodeHooks.insertHook(self, containerOrHandler.itemHandler());
         }
     }
 
@@ -216,8 +_,9 @@
     public static boolean suckInItems(Level level, Hopper hopper) {
         BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
         BlockState blockState = level.getBlockState(blockPos);
-        Container container = getSourceContainer(level, hopper, blockPos, blockState);
-        if (container != null) {
+        var containerOrHandler = getSourceContainerOrHandler(level, hopper, blockPos, blockState);
+        if (containerOrHandler.container() != null) {
+            Container container = containerOrHandler.container();
             Direction direction = Direction.DOWN;
 
             for (int slot : getSlots(container, direction)) {
@@ -227,6 +_,8 @@
             }
 
             return false;
+        } else if (containerOrHandler.itemHandler() != null) {
+            return net.neoforged.neoforge.transfer.item.VanillaInventoryCodeHooks.extractHook(hopper, containerOrHandler.itemHandler());
         } else {
             boolean isBlocked = hopper.isGridAligned()
                 && blockState.isCollisionShapeFullBlock(level, blockPos)
@@ -354,6 +_,8 @@
         return level.getEntitiesOfClass(ItemEntity.class, aabb, EntitySelector.ENTITY_STILL_ALIVE);
     }
 
+    /** @deprecated Use IItemHandler capability instead. To preserve Container-specific interactions, use {@link #getContainerOrHandlerAt} and handle both cases. */
+    @Deprecated
     public static @Nullable Container getContainerAt(Level level, BlockPos pos) {
         return getContainerAt(level, pos, level.getBlockState(pos), pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5);
     }
@@ -389,6 +_,28 @@
         return !entities.isEmpty() ? (Container)entities.get(level.getRandom().nextInt(entities.size())) : null;
     }
 
+    private static net.neoforged.neoforge.transfer.item.ContainerOrHandler getSourceContainerOrHandler(Level level, Hopper hopper, BlockPos pos, BlockState state) {
+        return getContainerOrHandlerAt(level, pos, state, hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ(), Direction.DOWN);
+    }
+
+    public static net.neoforged.neoforge.transfer.item.ContainerOrHandler getContainerOrHandlerAt(Level level, BlockPos pos, @Nullable Direction side) {
+        return getContainerOrHandlerAt(
+                level, pos, level.getBlockState(pos), (double)pos.getX() + 0.5, (double)pos.getY() + 0.5, (double)pos.getZ() + 0.5, side
+        );
+    }
+
+    private static net.neoforged.neoforge.transfer.item.ContainerOrHandler getContainerOrHandlerAt(Level level, BlockPos pos, BlockState state, double x, double y, double z, @Nullable Direction side) {
+        Container container = getBlockContainer(level, pos, state);
+        if (container != null) {
+            return new net.neoforged.neoforge.transfer.item.ContainerOrHandler(container, null);
+        }
+        var blockItemHandler = level.getCapability(net.neoforged.neoforge.capabilities.Capabilities.Item.BLOCK, pos, state, null, side);
+        if (blockItemHandler != null) {
+            return new net.neoforged.neoforge.transfer.item.ContainerOrHandler(null, blockItemHandler);
+        }
+        return net.neoforged.neoforge.transfer.item.VanillaInventoryCodeHooks.getEntityContainerOrHandler(level, x, y, z, side);
+    }
+
     private static boolean canMergeItems(ItemStack a, ItemStack b) {
         return a.getCount() <= a.getMaxStackSize() && ItemStack.isSameItemSameComponents(a, b);
     }
@@ -446,5 +_,9 @@
     @Override
     protected AbstractContainerMenu createMenu(int containerId, Inventory inventory) {
         return new HopperMenu(containerId, inventory, this);
+    }
+
+    public long getLastUpdateTime() {
+        return this.tickedGameTime;
     }
 }
