--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -137,6 +_,7 @@
     }
 
     private static boolean ejectItems(Level level, BlockPos pos, HopperBlockEntity blockEntity) {
+        if (net.neoforged.neoforge.items.VanillaInventoryCodeHooks.insertHook(blockEntity)) return true;
         Container container = getAttachedContainer(level, pos, blockEntity);
         if (container == null) {
             return false;
@@ -214,6 +_,8 @@
     public static boolean suckInItems(Level level, Hopper hopper) {
         BlockPos blockpos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
         BlockState blockstate = level.getBlockState(blockpos);
+        Boolean ret = net.neoforged.neoforge.items.VanillaInventoryCodeHooks.extractHook(level, hopper);
+        if (ret != null) return ret;
         Container container = getSourceContainer(level, hopper, blockpos, blockstate);
         if (container != null) {
             Direction direction = Direction.DOWN;
@@ -470,5 +_,9 @@
     @Override
     protected AbstractContainerMenu createMenu(int id, Inventory player) {
         return new HopperMenu(id, player, this);
+    }
+
+    public long getLastUpdateTime() {
+        return this.tickedGameTime;
     }
 }
