--- a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
@@ -110,9 +_,24 @@
         return itemstack;
     }
 
+    // Neo: Support deferring side effects until end of a transaction
+    @Override
+    public void setItem(int slot, ItemStack stack, boolean performSideEffects) {
+        if (slot == 0) {
+            if (performSideEffects) {
+                setTheItem(stack);
+            } else {
+                this.item = stack;
+            }
+        }
+    }
     @Override
     public void setTheItem(ItemStack p_304781_) {
         this.item = p_304781_;
+        performSideEffects(0, /* unused, pass empty */ItemStack.EMPTY);
+    }
+    @Override
+    public void performSideEffects(int slot, ItemStack originalStack) {
         boolean flag = !this.item.isEmpty();
         Optional<Holder<JukeboxSong>> optional = JukeboxSong.fromStack(this.level.registryAccess(), this.item);
         this.notifyItemChangedInJukebox(flag);
