--- a/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/JukeboxBlockEntity.java
@@ -113,9 +_,27 @@
         return itemstack;
     }
 
+    // Neo: Support deferring "final" actions until end of a transaction
+    @Override
+    public void setItem(int slot, ItemStack stack, boolean insideTransaction) {
+        if (slot != 0) return;
+
+        if (!insideTransaction) {
+            setTheItem(stack);
+        } else {
+            this.item = stack;
+        }
+    }
+
+
     @Override
     public void setTheItem(ItemStack p_304781_) {
         this.item = p_304781_;
+        onCommit(0, /* ignored */ ItemStack.EMPTY);
+    }
+
+    @Override
+    public void onCommit(int slot, ItemStack ignored) {
         boolean flag = !this.item.isEmpty();
         Optional<Holder<JukeboxSong>> optional = JukeboxSong.fromStack(this.level.registryAccess(), this.item);
         this.notifyItemChangedInJukebox(flag);
