--- a/net/minecraft/world/level/block/entity/ShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShelfBlockEntity.java
@@ -87,6 +_,20 @@
         return itemstack;
     }
 
+    @Override
+    public void setItem(int p_256610_, ItemStack p_255789_) {
+        setItem(p_256610_, p_255789_, false);
+    }
+
+    // Neo: Skip side effects if insideTransaction is true so the caller can defer them until the transaction commits
+    @Override
+    public void setItem(int p_256610_, ItemStack p_255789_, boolean insideTransaction) {
+        this.setItemNoUpdate(p_256610_, p_255789_);
+        if (!insideTransaction) {
+            this.setChanged();
+        }
+    }
+
     public void setChanged(Holder.@Nullable Reference<GameEvent> p_435920_) {
         super.setChanged();
         if (this.level != null) {
