--- a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -237,7 +_,8 @@
 
     @Override
     public boolean canPlaceItemThroughFace(int p_59663_, ItemStack p_59664_, @Nullable Direction p_59665_) {
-        return !(Block.byItem(p_59664_.getItem()) instanceof ShulkerBoxBlock);
+        // Neo: Make shulker boxes respect IItemStackExtension#canFitInsideContainerItems
+        return !(Block.byItem(p_59664_.getItem()) instanceof ShulkerBoxBlock) && p_59664_.canFitInsideContainerItems();
     }
 
     @Override
