--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -287,4 +287,9 @@
          p_277656_.setAllowedPlayerEditor(null);
       }
    }
+
+   @Override
+   public net.minecraft.world.phys.AABB getRenderBoundingBox() {
+      return new net.minecraft.world.phys.AABB(this.getBlockPos(), this.getBlockPos().offset(1, 1, 1));
+   }
 }
