--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -202,7 +_,7 @@
                     hasAnyClickCommand = true;
                     break;
                 case ClickEvent.Custom custom:
-                    level.getServer().handleCustomClickAction(custom.id(), custom.payload());
+                    level.getServer().handleCustomClickAction(custom.id(), custom.payload(), (net.minecraft.server.level.ServerPlayer) player, player.getGameProfile());
                     hasAnyClickCommand = true;
                     break;
                 case null:
