--- a/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SpawnerBlockEntity.java
@@ -32,6 +_,11 @@
                 level.sendBlockUpdated(pos, state, state, 260);
             }
         }
+
+        @Override
+        public com.mojang.datafixers.util.Either<net.minecraft.world.level.block.entity.BlockEntity, net.minecraft.world.entity.Entity> getOwner() {
+            return com.mojang.datafixers.util.Either.left(SpawnerBlockEntity.this);
+        }
     };
 
     public SpawnerBlockEntity(BlockPos worldPosition, BlockState blockState) {
