--- a/net/minecraft/client/data/models/BlockModelGenerators.java
+++ b/net/minecraft/client/data/models/BlockModelGenerators.java
@@ -98,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class BlockModelGenerators {
+public class BlockModelGenerators implements net.neoforged.neoforge.client.extensions.IBlockModelGeneratorExtensions {
     public final Consumer<BlockStateGenerator> blockStateOutput;
     public final ItemModelOutput itemModelOutput;
     public final BiConsumer<ResourceLocation, ModelInstance> modelOutput;
