--- a/net/minecraft/client/data/models/EquipmentAssetProvider.java
+++ b/net/minecraft/client/data/models/EquipmentAssetProvider.java
@@ -26,6 +_,10 @@
         this.pathProvider = p_387559_.createPathProvider(PackOutput.Target.RESOURCE_PACK, "equipment");
     }
 
+    protected void registerModels(BiConsumer<ResourceKey<EquipmentAsset>, EquipmentClientInfo> output) {
+        bootstrap(output);
+    }
+
     private static void bootstrap(BiConsumer<ResourceKey<EquipmentAsset>, EquipmentClientInfo> p_387865_) {
         p_387865_.accept(
             EquipmentAssets.LEATHER,
@@ -121,21 +_,21 @@
     }
 
     public static EquipmentClientInfo onlyHumanoid(String p_388505_) {
-        return EquipmentClientInfo.builder().addHumanoidLayers(Identifier.withDefaultNamespace(p_388505_)).build();
+        return EquipmentClientInfo.builder().addHumanoidLayers(Identifier.parse(p_388505_)).build();
     }
 
     public static EquipmentClientInfo humanoidAndMountArmor(String p_456144_) {
         return EquipmentClientInfo.builder()
-            .addHumanoidLayers(Identifier.withDefaultNamespace(p_456144_))
-            .addLayers(EquipmentClientInfo.LayerType.HORSE_BODY, EquipmentClientInfo.Layer.leatherDyeable(Identifier.withDefaultNamespace(p_456144_), false))
-            .addLayers(EquipmentClientInfo.LayerType.NAUTILUS_BODY, EquipmentClientInfo.Layer.leatherDyeable(Identifier.withDefaultNamespace(p_456144_), false))
+            .addHumanoidLayers(Identifier.parse(p_456144_))
+            .addLayers(EquipmentClientInfo.LayerType.HORSE_BODY, EquipmentClientInfo.Layer.leatherDyeable(Identifier.parse(p_456144_), false))
+            .addLayers(EquipmentClientInfo.LayerType.NAUTILUS_BODY, EquipmentClientInfo.Layer.leatherDyeable(Identifier.parse(p_456144_), false))
             .build();
     }
 
     @Override
     public CompletableFuture<?> run(CachedOutput p_387304_) {
         Map<ResourceKey<EquipmentAsset>, EquipmentClientInfo> map = new HashMap<>();
-        bootstrap((p_386976_, p_388942_) -> {
+        registerModels((p_386976_, p_388942_) -> {
             if (map.putIfAbsent(p_386976_, p_388942_) != null) {
                 throw new IllegalStateException("Tried to register equipment asset twice for id: " + p_386976_);
             }
