--- a/net/minecraft/client/data/models/ItemModelGenerators.java
+++ b/net/minecraft/client/data/models/ItemModelGenerators.java
@@ -203,6 +_,8 @@
         ModelTemplates.THREE_LAYERED_ITEM.create(target, TextureMapping.layered(layer0, layer1, layer2), this.modelOutput);
     }
 
+    /// @deprecated Neo: Use [#generateDynamicTrimmableArmorSet] instead.
+    @Deprecated
     public void generateTrimmableArmorSet(
         Item helmet,
         Item chestplate,
@@ -217,6 +_,8 @@
         this.generateTrimmableItem(boots, TRIM_PREFIX_BOOTS, hasDyedLayer, trimPaletteReplacements);
     }
 
+    /// @deprecated Neo: Use [#generateDynamicTrimmableItem] instead.
+    @Deprecated
     public void generateTrimmableItem(
         Item armor, Identifier slotTrimPrefix, boolean hasDyedLayer, Map<TrimMaterials.Palette, TrimMaterials.Palette> trimPaletteReplacements
     ) {
