--- a/net/minecraft/client/data/models/MultiVariant.java
+++ b/net/minecraft/client/data/models/MultiVariant.java
@@ -11,19 +_,59 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
+/// @param customBlockStateModels Neo-added parameter to allow using custom blockstate models with MultiVariantGenerator and MultiPartGenerator (only used during datagen)
 @OnlyIn(Dist.CLIENT)
-public record MultiVariant(WeightedList<Variant> variants) {
+public record MultiVariant(WeightedList<Variant> variants, WeightedList<net.neoforged.neoforge.client.model.generators.blockstate.CustomBlockStateModelBuilder> customBlockStateModels) {
     public MultiVariant {
+        if (!customBlockStateModels.isEmpty()) {
+            if (!variants.isEmpty()) {
+                throw new IllegalArgumentException("Cannot specify both a variant list and a custom blockstate model");
+            }
+        } else
         if (variants.isEmpty()) {
             throw new IllegalArgumentException("Variant list must contain at least one element");
         }
     }
 
+    public MultiVariant(WeightedList<Variant> variants) {
+        this(variants, WeightedList.of());
+    }
+
+    // Neo: convenience functions for datagen with custom blockstate models
+    public static MultiVariant of(WeightedList<net.neoforged.neoforge.client.model.generators.blockstate.CustomBlockStateModelBuilder> customBlockStateModels) {
+        return new MultiVariant(WeightedList.of(), customBlockStateModels);
+    }
+
+    public static MultiVariant of(net.neoforged.neoforge.client.model.generators.blockstate.CustomBlockStateModelBuilder customBlockStateModel) {
+        return new MultiVariant(WeightedList.of(), WeightedList.of(customBlockStateModel));
+    }
+
     public MultiVariant with(VariantMutator mutator) {
+        if (!this.customBlockStateModels.isEmpty()) {
+            return new MultiVariant(this.variants, this.customBlockStateModels.map(model -> model.with(mutator)));
+        }
         return new MultiVariant(this.variants.map(mutator));
     }
 
+    /// Applies the given unbaked mutator.
+    ///
+    /// If this multi variant consists of plain Vanilla [variants][Variant], each variant will be
+    /// temporarily converted to a [SingleVariant.Unbaked], mutated and then converted back to a plain variant.
+    public MultiVariant with(net.neoforged.neoforge.client.model.generators.blockstate.UnbakedMutator mutator) {
+        if (!this.customBlockStateModels.isEmpty()) {
+            return new MultiVariant(this.variants, this.customBlockStateModels.map(model -> model.with(mutator)));
+        }
+        return new MultiVariant(this.variants.map(v -> mutator.apply(new SingleVariant.Unbaked(v)).variant()));
+    }
+
     public BlockStateModel.Unbaked toUnbaked() {
+        if (!this.customBlockStateModels.isEmpty()) {
+            var builders = this.customBlockStateModels.unwrap();
+            if (builders.size() == 1) {
+                return builders.getFirst().value().toUnbaked();
+            }
+            return new WeightedVariants.Unbaked(this.customBlockStateModels.map(net.neoforged.neoforge.client.model.generators.blockstate.CustomBlockStateModelBuilder::toUnbaked));
+        }
         List<Weighted<Variant>> entries = this.variants.unwrap();
         return entries.size() == 1
             ? new SingleVariant.Unbaked(entries.getFirst().value())
