--- a/net/minecraft/world/entity/decoration/HangingEntity.java
+++ b/net/minecraft/world/entity/decoration/HangingEntity.java
@@ -83,7 +_,7 @@
         if (!this.level().noCollision(this, this.getPopBox())) {
             return false;
         } else {
-            boolean flag = BlockPos.betweenClosedStream(this.calculateSupportBox()).allMatch(p_423325_ -> {
+            boolean flag = BlockPos.betweenClosedStream(this.calculateSupportBox()).filter(pos -> !net.minecraft.world.level.block.Block.canSupportCenter(this.level(), pos, this.getDirection())).allMatch(p_423325_ -> {
                 BlockState blockstate = this.level().getBlockState(p_423325_);
                 return blockstate.isSolid() || DiodeBlock.isDiode(blockstate);
             });
