--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -248,7 +_,7 @@
     private void removeFramedMap(ItemStack itemStack) {
         MapId mapId = this.getFramedMapId(itemStack);
         if (mapId != null) {
-            MapItemSavedData mapItemSavedData = MapItem.getSavedData(mapId, this.level());
+            MapItemSavedData mapItemSavedData = MapItem.getSavedData(itemStack, this.level());
             if (mapItemSavedData != null) {
                 mapItemSavedData.removedFromFrame(this.pos, this.getId());
             }
