--- a/net/minecraft/world/level/chunk/storage/SectionStorage.java
+++ b/net/minecraft/world/level/chunk/storage/SectionStorage.java
@@ -238,4 +_,12 @@
     public void close() throws IOException {
         this.simpleRegionStorage.close();
     }
+
+    /**
+     * Neo: Removes the data for the given chunk position.
+     * See PR #937
+     */
+    public void remove(long sectionPosAsLong) {
+        this.storage.remove(sectionPosAsLong);
+    }
 }
