--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -91,7 +_,9 @@
     List<SerializableChunkData.SectionData> sectionData,
     List<CompoundTag> entities,
     List<CompoundTag> blockEntities,
-    CompoundTag structureData
+    CompoundTag structureData,
+    @Nullable CompoundTag attachmentData,
+    @Nullable ListTag auxLightData
 ) {
     private static final Codec<List<SavedTick<Block>>> BLOCK_TICKS_CODEC = SavedTick.codec(BuiltInRegistries.BLOCK.byNameCodec()).listOf();
     private static final Codec<List<SavedTick<Fluid>>> FLUID_TICKS_CODEC = SavedTick.codec(BuiltInRegistries.FLUID.byNameCodec()).listOf();
@@ -107,6 +_,33 @@
     public static final String BLOCK_LIGHT_TAG = "BlockLight";
     public static final String SKY_LIGHT_TAG = "SkyLight";
 
+    /**
+     * @deprecated Neo: use constructor with additional data instead
+     */
+    @Deprecated
+    SerializableChunkData(
+            PalettedContainerFactory containerFactory,
+            ChunkPos chunkPos,
+            int minSectionY,
+            long lastUpdateTime,
+            long inhabitedTime,
+            ChunkStatus chunkStatus,
+            BlendingData.@Nullable Packed blendingData,
+            @Nullable BelowZeroRetrogen belowZeroRetrogen,
+            UpgradeData upgradeData,
+            long @Nullable [] carvingMask,
+            Map<Heightmap.Types, long[]> heightmaps,
+            ChunkAccess.PackedTicks packedTicks,
+            ShortList[] postProcessingSections,
+            boolean lightCorrect,
+            List<SerializableChunkData.SectionData> sectionData,
+            List<CompoundTag> entities,
+            List<CompoundTag> blockEntities,
+            CompoundTag structureData
+    ) {
+        this(containerFactory, chunkPos, minSectionY, lastUpdateTime, inhabitedTime, chunkStatus, blendingData, belowZeroRetrogen, upgradeData, carvingMask, heightmaps, packedTicks, postProcessingSections, lightCorrect, sectionData, entities, blockEntities, structureData, null, null);
+    }
+
     public static SerializableChunkData parse(LevelHeightAccessor levelHeight, PalettedContainerFactory containerFactory, CompoundTag chunkData) {
         if (chunkData.getString("Status").isEmpty()) {
             return null;
@@ -122,7 +_,7 @@
             long[] carvingMask = chunkData.getLongArray("carving_mask").orElse(null);
             Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
             chunkData.getCompound("Heightmaps").ifPresent(heightmapsTag -> {
-                for (Heightmap.Types type : status.heightmapsAfter()) {
+                for (Heightmap.Types type : status.getChunkSaveHeightmaps()) {
                     heightmapsTag.getLongArray(type.getSerializationKey()).ifPresent(longs -> heightmaps.put(type, longs));
                 }
             });
@@ -185,6 +_,9 @@
                 }
             }
 
+            CompoundTag attachmentData = chunkData.getCompound(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY).orElse(null);
+            ListTag auxLightData = chunkData.getList(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY).orElse(null);
+
             return new SerializableChunkData(
                 containerFactory,
                 chunkPos,
@@ -203,7 +_,9 @@
                 sectionData,
                 entities,
                 blockEntities,
-                structureData
+                structureData,
+                attachmentData,
+                auxLightData
             );
         }
     }
@@ -263,6 +_,9 @@
                 postLoadChunk(level, this.entities, this.blockEntities),
                 BlendingData.unpack(this.blendingData)
             );
+            if (this.auxLightData != null) {
+                ((LevelChunk) chunk).getAuxLightManager(chunkPos).deserializeNBT(this.auxLightData);
+            }
         } else {
             ProtoChunkTicks<Block> blockTicks = ProtoChunkTicks.load(this.packedTicks.blocks());
             ProtoChunkTicks<Fluid> fluidTicks = ProtoChunkTicks.load(this.packedTicks.fluids());
@@ -284,7 +_,7 @@
         chunk.setLightCorrect(this.lightCorrect);
         EnumSet<Heightmap.Types> toPrime = EnumSet.noneOf(Heightmap.Types.class);
 
-        for (Heightmap.Types type : chunk.getPersistedStatus().heightmapsAfter()) {
+        for (Heightmap.Types type : chunk.getPersistedStatus().getChunkSaveHeightmaps()) {
             long[] heightmap = this.heightmaps.get(type);
             if (heightmap != null) {
                 chunk.setHeightmap(type, heightmap);
@@ -304,6 +_,10 @@
             }
         }
 
+        if (this.attachmentData != null) {
+            chunk.readAttachmentsFromNBT(level.registryAccess(), this.attachmentData);
+        }
+
         if (chunkType == ChunkType.LEVELCHUNK) {
             return new ImposterProtoChunk((LevelChunk)chunk, false);
         } else {
@@ -374,7 +_,7 @@
             Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
 
             for (Entry<Heightmap.Types, Heightmap> entry : chunk.getHeightmaps()) {
-                if (chunk.getPersistedStatus().heightmapsAfter().contains(entry.getKey())) {
+                if (chunk.getPersistedStatus().getChunkSaveHeightmaps().contains(entry.getKey())) {
                     long[] data = entry.getValue().getRawData();
                     heightmaps.put(entry.getKey(), (long[])data.clone());
                 }
@@ -387,6 +_,18 @@
             CompoundTag structureData = packStructureData(
                 StructurePieceSerializationContext.fromLevel(level), pos, chunk.getAllStarts(), chunk.getAllReferences()
             );
+
+            CompoundTag attachmentData = null;
+            try {
+                attachmentData = chunk.writeAttachmentsToNBT(level.registryAccess());
+            } catch (Exception exception) {
+                LOGGER.error("Failed to write chunk attachments. An attachment has likely thrown an exception trying to write state. It will not persist. Report this to the mod author", exception);
+            }
+            ListTag auxLightData = null;
+            if (chunk instanceof LevelChunk levelChunk) {
+                auxLightData = levelChunk.getAuxLightManager(pos).serializeNBT();
+            }
+
             return new SerializableChunkData(
                 level.palettedContainerFactory(),
                 pos,
@@ -405,7 +_,9 @@
                 sectionData,
                 entities,
                 blockEntities,
-                structureData
+                structureData,
+                attachmentData,
+                auxLightData
             );
         }
     }
@@ -473,6 +_,14 @@
         this.heightmaps.forEach((type, data) -> heightmapsTag.put(type.getSerializationKey(), new LongArrayTag(data)));
         tag.put("Heightmaps", heightmapsTag);
         tag.put("structures", this.structureData);
+
+        if (attachmentData != null) {
+            tag.put(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY, attachmentData);
+        }
+        if (auxLightData != null) {
+            tag.put(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, auxLightData);
+        }
+
         return tag;
     }
 
