--- a/com/mojang/blaze3d/platform/TextureUtil.java
+++ b/com/mojang/blaze3d/platform/TextureUtil.java
@@ -54,6 +_,10 @@
 
     public static void writeAsPNG(Path dir, String prefix, GpuTexture texture, int maxMipLevel, IntUnaryOperator pixelModifier) {
         RenderSystem.assertOnRenderThread();
+        // Neo: Textures with packed depth/stencil formats cannot be exported as the driver implementation is free to store this data in whatever layout it prefers
+        if (texture.getFormat() == com.mojang.blaze3d.textures.TextureFormat.DEPTH24_STENCIL8 || texture.getFormat() == com.mojang.blaze3d.textures.TextureFormat.DEPTH32_STENCIL8) {
+            throw new IllegalArgumentException("Textures with packed depth/stencil formats cannot be exported");
+        }
         long bufferLength = 0L;
 
         for (int i = 0; i <= maxMipLevel; i++) {
@@ -80,6 +_,7 @@
                         };
                         case RGBA8 -> byteOffset -> data.getInt(byteOffset);
                         case DEPTH32 -> byteOffset -> ARGB.gray(data.getFloat(byteOffset));
+                        case DEPTH24_STENCIL8, DEPTH32_STENCIL8 -> _ -> -1;
                     };
                     int offsetx = 0;
 
