--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -108,6 +_,16 @@
             debugOptions
         );
         this.handle = windowAndDevice.window();
+        // Neo: If the window was taken over from EarlyLoadingScreen, inherit a potentially resized window size
+        var earlyLoadingScreen = net.neoforged.fml.loading.EarlyLoadingScreenController.current();
+        if (earlyLoadingScreen != null) {
+            var width = new int[1];
+            var height = new int[1];
+            GLFW.glfwGetWindowSize(this.handle, width, height);
+            // The window height and width can be 0 if minimized
+            this.width = this.windowedWidth = Math.max(width[0], 1);
+            this.height = this.windowedHeight = Math.max(height[0], 1);
+        }
         RenderSystem.initRenderer(windowAndDevice.device());
         if (initialMonitor != null) {
             VideoMode mode = initialMonitor.getPreferredVidMode(this.fullscreen ? this.preferredFullscreenVideoMode : Optional.empty());
@@ -237,8 +_,10 @@
     }
 
     public void defaultErrorCallback(int errorCode, long description) {
-        RenderSystem.assertOnRenderThread();
         String errorString = MemoryUtil.memUTF8(description);
+        if (!RenderSystem.isOnRenderThread()) {
+            throw new IllegalStateException("Encountered GL error off-thread @ " + errorSection + ": " + errorCode + ": " + errorString);
+        }
         LOGGER.error("########## GL ERROR ##########");
         LOGGER.error("@ {}", this.errorSection);
         LOGGER.error("{}: {}", errorCode, errorString);
