--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -237,11 +_,11 @@
     }
 
     public void defaultErrorCallback(int errorCode, long description) {
+        String errorString = MemoryUtil.memUTF8(description);
         if (!RenderSystem.isOnRenderThread()) {
-            LOGGER.error("Error callback from invalid thread ({})", Thread.currentThread().getName());
+            LOGGER.error("Error callback from invalid thread ({}): {}: {}", Thread.currentThread().getName(), errorCode, errorString);
         }
 
-        String errorString = MemoryUtil.memUTF8(description);
         LOGGER.error("########## GL ERROR ##########");
         LOGGER.error("@ {}", this.errorSection);
         LOGGER.error("{}: {}", errorCode, errorString);
