--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -92,6 +_,16 @@
         this.windowedWidth = this.width = allowedWindowMinSize(displayData.width());
         this.windowedHeight = this.height = allowedWindowMinSize(displayData.height());
         this.handle = this.createWindow(backend, this.width, this.height, title, this.fullscreen && initialMonitor != null ? initialMonitor.getMonitor() : 0L);
+        // Neo: If the window was taken over from EarlyLoadingScreen, inherit a potentially resized window size
+        var earlyLoadingScreen = net.neoforged.fml.loading.EarlyLoadingScreenController.current();
+        if (earlyLoadingScreen != null) {
+            var width = new int[1];
+            var height = new int[1];
+            GLFW.glfwGetWindowSize(this.handle, width, height);
+            // The window height and width can be 0 if minimized
+            this.width = this.windowedWidth = Math.max(width[0], 1);
+            this.height = this.windowedHeight = Math.max(height[0], 1);
+        }
         this.backend = backend;
         if (initialMonitor != null) {
             VideoMode mode = initialMonitor.getPreferredVidMode(this.fullscreen ? this.preferredFullscreenVideoMode : Optional.empty());
@@ -121,6 +_,11 @@
         long windowHandle;
         try (GLFWErrorScope var9 = new GLFWErrorScope(glfwErrors)) {
             backend.setWindowHints();
+            var earlyLoadingScreen = net.neoforged.fml.loading.EarlyLoadingScreenController.current();
+            if (earlyLoadingScreen != null) {
+                windowHandle = earlyLoadingScreen.takeOverGlfwWindow();
+                GLFW.glfwSetWindowTitle(windowHandle, title);
+            } else
             windowHandle = GLFW.glfwCreateWindow(width, height, title, monitor, 0L);
             if (windowHandle == 0L) {
                 backend.handleWindowCreationErrors(glfwErrors.firstError());
@@ -226,8 +_,10 @@
     }
 
     public void defaultErrorCallback(int errorCode, long description) {
-        RenderSystem.assertOnRenderThread();
         String errorString = MemoryUtil.memUTF8(description);
+        if (!RenderSystem.isOnRenderThread()) {
+            throw new IllegalStateException("Encountered GL error off-thread @ " + errorSection + ": " + errorCode + ": " + errorString);
+        }
         LOGGER.error("########## GL ERROR ##########");
         LOGGER.error("@ {}", this.errorSection);
         LOGGER.error("{}: {}", errorCode, errorString);
