--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -120,6 +_,11 @@
         long windowHandle;
         try (GLFWErrorScope var9 = new GLFWErrorScope(glfwErrors)) {
             backend.setWindowHints();
+            var earlyLoadingScreen = net.neoforged.fml.loading.EarlyLoadingScreenController.current();
+            if (earlyLoadingScreen != null) {
+                windowHandle = earlyLoadingScreen.takeOverGlfwWindow();
+                GLFW.glfwSetWindowTitle(windowHandle, title);
+            } else
             windowHandle = GLFW.glfwCreateWindow(width, height, title, monitor, 0L);
             if (windowHandle == 0L) {
                 backend.handleWindowCreationErrors(glfwErrors.firstError());
@@ -225,8 +_,10 @@
     }
 
     public void defaultErrorCallback(int errorCode, long description) {
-        RenderSystem.assertOnRenderThread();
         String errorString = MemoryUtil.memUTF8(description);
+        if (!RenderSystem.isOnRenderThread()) {
+            throw new IllegalStateException("Encountered GL error off-thread @ " + errorSection + ": " + errorCode + ": " + errorString);
+        }
         LOGGER.error("########## GL ERROR ##########");
         LOGGER.error("@ {}", this.errorSection);
         LOGGER.error("{}: {}", errorCode, errorString);
