--- a/net/minecraft/client/renderer/texture/atlas/SpriteSources.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteSources.java
@@ -25,5 +_,7 @@
         ID_MAPPER.put(ResourceLocation.withDefaultNamespace("filter"), SourceFilter.MAP_CODEC);
         ID_MAPPER.put(ResourceLocation.withDefaultNamespace("unstitch"), Unstitcher.MAP_CODEC);
         ID_MAPPER.put(ResourceLocation.withDefaultNamespace("paletted_permutations"), PalettedPermutations.MAP_CODEC);
+
+        net.neoforged.fml.ModLoader.postEvent(new net.neoforged.neoforge.client.event.RegisterSpriteSourcesEvent(ID_MAPPER));
     }
 }
