--- a/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
@@ -45,6 +45,9 @@
             .orElse(AnimationMetadataSection.EMPTY);
          FrameSize framesize = animationmetadatasection.calculateFrameSize(nativeimage.getWidth(), nativeimage.getHeight());
          if (Mth.isMultipleOf(nativeimage.getWidth(), framesize.width()) && Mth.isMultipleOf(nativeimage.getHeight(), framesize.height())) {
+            SpriteContents contents = net.neoforged.neoforge.client.ClientHooks.loadSpriteContents(p_293680_, p_293681_, framesize, nativeimage, animationmetadatasection);
+            if (contents != null)
+               return contents;
             return new SpriteContents(p_293680_, framesize, nativeimage, resourcemetadata);
          } else {
             LOGGER.error(
