--- a/net/minecraft/client/renderer/texture/atlas/SpriteSourceList.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteSourceList.java
@@ -34,7 +_,15 @@
         this.sources = sources;
     }
 
+    /**
+     * @deprecated Neo: use {@link #list(ResourceManager, java.util.Set)} instead
+     */
+    @Deprecated
     public List<SpriteSource.Loader> list(ResourceManager resourceManager) {
+        return this.list(resourceManager, java.util.Set.of());
+    }
+
+    public List<SpriteSource.Loader> list(ResourceManager resourceManager, java.util.Set<net.minecraft.server.packs.metadata.MetadataSectionType<?>> additionalMetadata) {
         final Map<Identifier, SpriteSource.DiscardableLoader> sprites = new HashMap<>();
         SpriteSource.Output output = new SpriteSource.Output() {
             @Override
@@ -58,7 +_,7 @@
                 }
             }
         };
-        this.sources.forEach(s -> s.run(resourceManager, output));
+        this.sources.forEach(s -> s.run(resourceManager, output, additionalMetadata));
         Builder<SpriteSource.Loader> result = ImmutableList.builder();
         result.add(loader -> MissingTextureAtlasSprite.create());
         result.addAll(sprites.values());
