--- a/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteResourceLoader.java
@@ -26,7 +_,7 @@
     Logger LOGGER = LogUtils.getLogger();
 
     static SpriteResourceLoader create(Set<MetadataSectionType<?>> p_434463_) {
-        return (p_432311_, p_432312_) -> {
+        return (p_432311_, p_432312_, constructor) -> {
             Optional<AnimationMetadataSection> optional;
             List<MetadataSectionType.WithValue<?>> list;
             try {
@@ -65,10 +_,15 @@
                 framesize = new FrameSize(nativeimage.getWidth(), nativeimage.getHeight());
             }
 
-            return new SpriteContents(p_432311_, framesize, nativeimage, optional, list);
+            return constructor.create(p_432311_, framesize, nativeimage, optional, list);
         };
     }
 
     @Nullable
-    SpriteContents loadSprite(ResourceLocation p_295581_, Resource p_294329_);
+    default SpriteContents loadSprite(ResourceLocation p_295581_, Resource p_294329_) {
+        return loadSprite(p_295581_, p_294329_, SpriteContents::new);
+    }
+
+    @Nullable
+    SpriteContents loadSprite(ResourceLocation id, Resource resource, net.neoforged.neoforge.client.textures.SpriteContentsConstructor constructor);
 }
