--- a/net/minecraft/world/item/BucketItem.java
+++ b/net/minecraft/world/item/BucketItem.java
@@ -115,9 +120,16 @@
             boolean shiftKeyDown = user != null && user.isShiftKeyDown();
             boolean placeLiquid = mayReplace
                 || block instanceof LiquidBlockContainer container && container.canPlaceLiquid(user, level, pos, blockState, this.content);
+            // TODO: The stack is only used as a parameter in the vaporization check. What if the fluidType is different? And should the rest of the method be made stack-aware as well?
+            var containedFluidStack = containerItem != null ? net.neoforged.neoforge.transfer.fluid.FluidUtil.getFirstStackContained(containerItem) : net.neoforged.neoforge.fluids.FluidStack.EMPTY;
             boolean canPlaceFluidInsideBlock = blockState.isAir() || placeLiquid && (!shiftKeyDown || hitResult == null);
             if (!canPlaceFluidInsideBlock) {
-                return hitResult != null && this.emptyContents(user, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null);
+                return hitResult != null && this.emptyContents(user, level, hitResult.getBlockPos().relative(hitResult.getDirection()), null, containerItem);
+            }
+
+            if (!containedFluidStack.isEmpty() && this.content.getFluidType().isVaporizedOnPlacement(level, pos, containedFluidStack)) {
+                this.content.getFluidType().onVaporize(user, level, pos, containedFluidStack);
+                return true;
             }
 
             if (level.environmentAttributes().getValue(EnvironmentAttributes.WATER_EVAPORATES, pos) && this.content.is(FluidTags.WATER)) {
@@ -132,7 +144,7 @@
                 }
 
                 return true;
-            } else if (block instanceof LiquidBlockContainer container && this.content == Fluids.WATER) {
+            } else if (block instanceof LiquidBlockContainer container && container.canPlaceLiquid(user, level, pos, blockState, content)) {
                 container.placeLiquid(level, pos, blockState, flowingFluid.getSource(false));
                 this.playEmptySound(user, level, pos);
                 return true;
