--- a/com/mojang/renderpearl/backend/api/CommandEncoderBackend.java
+++ b/com/mojang/renderpearl/backend/api/CommandEncoderBackend.java
@@ -40,6 +_,12 @@
 
     void clearDepthTexture(GpuTexture depthTexture, double clearDepth);
 
+    /// Neo: clears the stencil bits of the given texture to the given value.
+    ///
+    /// @param texture The texture to clear
+    /// @param value   The value to set the stencil bits to
+    void clearStencilTexture(GpuTexture texture, int value);
+
     void writeToBuffer(GpuBufferSlice destination, ByteBuffer data);
 
     void copyToBuffer(GpuBufferSlice source, GpuBufferSlice target);
