--- a/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
+++ b/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
@@ -69,6 +_,17 @@
 public class ArgumentTypeInfos {
     private static final Map<Class<?>, ArgumentTypeInfo<?, ?>> BY_CLASS = Maps.newHashMap();
 
+    /// Neo: Use this in conjunction with a [`DeferredRegister#register(String, Supplier)`][net.neoforged.neoforge.registries.DeferredRegister#register(String, java.util.function.Supplier)]
+    /// call to both populate the [#BY_CLASS] map and register the argument type info so it can be used in commands.
+    ///
+    /// @param infoClass the class type of the argument type info
+    /// @param argumentTypeInfo the argument type info instance
+    /// @return the provided argument type info instance for chaining
+    public static synchronized <A extends ArgumentType<?>, T extends ArgumentTypeInfo.Template<A>, I extends ArgumentTypeInfo<A, T>> I registerByClass(Class<A> infoClass, I argumentTypeInfo) {
+        BY_CLASS.put(infoClass, argumentTypeInfo);
+        return argumentTypeInfo;
+    }
+
     private static <A extends ArgumentType<?>, T extends ArgumentTypeInfo.Template<A>> ArgumentTypeInfo<A, T> register(
         Registry<ArgumentTypeInfo<?, ?>> registry, String id, Class<? extends A> brigadierType, ArgumentTypeInfo<A, T> info
     ) {
