--- a/net/minecraft/data/metadata/PackMetadataGenerator.java
+++ b/net/minecraft/data/metadata/PackMetadataGenerator.java
@@ -19,7 +_,7 @@
 import net.minecraft.server.packs.metadata.pack.PackMetadataSection;
 import net.minecraft.world.flag.FeatureFlagSet;
 
-public class PackMetadataGenerator implements DataProvider {
+public class PackMetadataGenerator implements DataProvider, net.neoforged.neoforge.common.conditions.IConditionBuilder {
     private final PackOutput output;
     private final Map<String, Supplier<JsonElement>> elements = new HashMap<>();
 
