--- a/net/minecraft/client/model/Model.java
+++ b/net/minecraft/client/model/Model.java
@@ -23,6 +_,10 @@
         this.allParts = p_368583_.getAllParts();
     }
 
+    protected static net.neoforged.neoforge.client.entity.animation.json.AnimationHolder getAnimation(ResourceLocation key) {
+        return net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.INSTANCE.getAnimationHolder(key);
+    }
+
     public final RenderType renderType(ResourceLocation p_103120_) {
         return this.renderType.apply(p_103120_);
     }
