--- a/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -157,8 +_,8 @@
         if (this.level() != null && !this.level().isClientSide) {
             this.goalSelector.removeGoal(this.meleeGoal);
             this.goalSelector.removeGoal(this.bowGoal);
-            ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
-            if (itemstack.is(Items.BOW)) {
+            ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
+            if (itemstack.getItem() instanceof net.minecraft.world.item.BowItem) {
                 int i = this.getHardAttackInterval();
                 if (this.level().getDifficulty() != Difficulty.HARD) {
                     i = this.getAttackInterval();
@@ -182,16 +_,28 @@
 
     @Override
     public void performRangedAttack(LivingEntity p_32141_, float p_32142_) {
-        ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
-        ItemStack itemstack1 = this.getProjectile(itemstack);
-        AbstractArrow abstractarrow = this.getArrow(itemstack1, p_32142_, itemstack);
+        ItemStack weapon = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
+        ItemStack itemstack1 = this.getProjectile(weapon);
+        AbstractArrow abstractarrow = this.getArrow(itemstack1, p_32142_, weapon);
+        if (weapon.getItem() instanceof net.minecraft.world.item.ProjectileWeaponItem weaponItem)
+            abstractarrow = weaponItem.customArrow(abstractarrow, itemstack1, weapon);
         double d0 = p_32141_.getX() - this.getX();
         double d1 = p_32141_.getY(0.3333333333333333) - abstractarrow.getY();
         double d2 = p_32141_.getZ() - this.getZ();
         double d3 = Math.sqrt(d0 * d0 + d2 * d2);
-        abstractarrow.shoot(d0, d1 + d3 * 0.2F, d2, 1.6F, (float)(14 - this.level().getDifficulty().getId() * 4));
+        var event = net.neoforged.neoforge.event.EventHooks.onProjectileWeaponShootPre(this, abstractarrow, p_32141_, weapon, itemstack1, 1.6F, (float)(14 - this.level().getDifficulty().getId() * 4));
+        if (event.isCanceled()) return;
+        net.minecraft.world.entity.projectile.Projectile projectile = event.getProjectile();
+        if (projectile instanceof AbstractArrow) {
+            abstractarrow = (AbstractArrow) projectile;
+            // Neo: keep the original AbstractArrow#shoot call, so we don't break old/multiloader mixins
+            abstractarrow.shoot(d0, d1 + d3 * 0.2F, d2, event.getPower(), event.getDivergence());
+        } else {
+            projectile.shoot(d0, d1 + d3 * 0.2F, d2, event.getPower(), event.getDivergence());
+        }
         this.playSound(SoundEvents.SKELETON_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
-        this.level().addFreshEntity(abstractarrow);
+        this.level().addFreshEntity(projectile);
+        net.neoforged.neoforge.event.EventHooks.onProjectileWeaponShootPost(this, projectile, p_32141_, weapon, itemstack1, 0);
     }
 
     protected AbstractArrow getArrow(ItemStack p_32156_, float p_32157_, @Nullable ItemStack p_346155_) {
