--- a/net/minecraft/world/entity/monster/CrossbowAttackMob.java
+++ b/net/minecraft/world/entity/monster/CrossbowAttackMob.java
@@ -16,7 +_,7 @@
     void onCrossbowAttackPerformed();
 
     default void performCrossbowAttack(LivingEntity body, float crossbowPower) {
-        InteractionHand hand = ProjectileUtil.getWeaponHoldingHand(body, Items.CROSSBOW);
+        InteractionHand hand = ProjectileUtil.getWeaponHoldingHand(body, item -> item instanceof CrossbowItem);
         ItemStack usedItem = body.getItemInHand(hand);
         if (usedItem.getItem() instanceof CrossbowItem crossbow) {
             crossbow.performShooting(body.level(), body, hand, usedItem, crossbowPower, 14 - body.level().getDifficulty().getId() * 4, this.getTarget());
