--- a/net/minecraft/world/entity/monster/CrossbowAttackMob.java
+++ b/net/minecraft/world/entity/monster/CrossbowAttackMob.java
@@ -17,7 +_,7 @@
     void onCrossbowAttackPerformed();
 
     default void performCrossbowAttack(LivingEntity user, float velocity) {
-        InteractionHand interactionhand = ProjectileUtil.getWeaponHoldingHand(user, Items.CROSSBOW);
+        InteractionHand interactionhand = ProjectileUtil.getWeaponHoldingHand(user, item -> item instanceof CrossbowItem);
         ItemStack itemstack = user.getItemInHand(interactionhand);
         if (itemstack.getItem() instanceof CrossbowItem crossbowitem) {
             crossbowitem.performShooting(
