--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -121,7 +_,6 @@
 
     @Override
     public void setTarget(@Nullable LivingEntity target) {
-        super.setTarget(target);
         AttributeInstance movementSpeed = this.getAttribute(Attributes.MOVEMENT_SPEED);
         if (target == null) {
             this.targetChangeTime = 0;
@@ -135,6 +_,7 @@
                 movementSpeed.addTransientModifier(SPEED_MODIFIER_ATTACKING);
             }
         }
+        super.setTarget(target); //Forge: Moved down to allow event handlers to write data manager values.
     }
 
     @Override
@@ -207,7 +_,7 @@
     }
 
     private boolean isBeingStaredBy(Player player) {
-        return !LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM.test(player) ? false : this.isLookingAtMe(player, 0.025, true, false, this.getEyeY());
+        return !LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET.test(player, this) ? false : this.isLookingAtMe(player, 0.025, true, false, this.getEyeY()) && !net.neoforged.neoforge.common.CommonHooks.shouldSuppressEnderManAnger(this, player);
     }
 
     @Override
@@ -441,7 +_,7 @@
             if (this.enderman.getCarriedBlock() == null) {
                 return false;
             } else {
-                return !getServerLevel(this.enderman).getGameRules().get(GameRules.MOB_GRIEFING)
+                return !net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.enderman.level()), this.enderman)
                     ? false
                     : this.enderman.getRandom().nextInt(reducedTickDelay(2000)) == 0;
             }
@@ -461,7 +_,7 @@
             BlockState carried = this.enderman.getCarriedBlock();
             if (carried != null) {
                 carried = Block.updateFromNeighbourShapes(carried, this.enderman.level(), pos);
-                if (this.canPlaceBlock(level, pos, carried, targetState, belowState, below)) {
+                if (this.canPlaceBlock(level, pos, carried, targetState, belowState, below) && !net.neoforged.neoforge.event.EventHooks.onBlockPlace(enderman, net.neoforged.neoforge.common.util.BlockSnapshot.create(level.dimension(), level, below), net.minecraft.core.Direction.UP)) {
                     level.setBlock(pos, carried, 3);
                     level.gameEvent(GameEvent.BLOCK_PLACE, pos, GameEvent.Context.of(this.enderman, carried));
                     this.enderman.setCarriedBlock(null);
@@ -473,6 +_,7 @@
             return targetState.isAir()
                 && !belowState.isAir()
                 && !belowState.is(Blocks.BEDROCK)
+                && !belowState.is(net.neoforged.neoforge.common.Tags.Blocks.ENDERMAN_PLACE_ON_BLACKLIST)
                 && belowState.isCollisionShapeFullBlock(level, below)
                 && carried.canSurvive(level, pos)
                 && level.getEntities(this.enderman, AABB.unitCubeFromLowerCorner(Vec3.atLowerCornerOf(pos))).isEmpty();
@@ -583,7 +_,7 @@
             if (this.enderman.getCarriedBlock() != null) {
                 return false;
             } else {
-                return !getServerLevel(this.enderman).getGameRules().get(GameRules.MOB_GRIEFING)
+                return !net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.enderman.level()), this.enderman)
                     ? false
                     : this.enderman.getRandom().nextInt(reducedTickDelay(20)) == 0;
             }
