--- a/net/minecraft/world/entity/monster/Illusioner.java
+++ b/net/minecraft/world/entity/monster/Illusioner.java
@@ -174,9 +_,11 @@
 
     @Override
     public void performRangedAttack(LivingEntity target, float distanceFactor) {
-        ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
+        ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
         ItemStack itemstack1 = this.getProjectile(itemstack);
         AbstractArrow abstractarrow = ProjectileUtil.getMobArrow(this, itemstack1, distanceFactor, itemstack);
+        if (itemstack.getItem() instanceof net.minecraft.world.item.BowItem bowItem)
+            abstractarrow = bowItem.customArrow(abstractarrow, itemstack1, itemstack);
         double d0 = target.getX() - this.getX();
         double d1 = target.getY(0.3333333333333333) - abstractarrow.getY();
         double d2 = target.getZ() - this.getZ();
