--- a/net/minecraft/world/entity/monster/Pillager.java
+++ b/net/minecraft/world/entity/monster/Pillager.java
@@ -96,7 +_,7 @@
 
     @Override
     public boolean canFireProjectileWeapon(ProjectileWeaponItem p_33280_) {
-        return p_33280_ == Items.CROSSBOW;
+        return p_33280_.builtInRegistryHolder().is(net.neoforged.neoforge.common.Tags.Items.PILLAGER_USABLE_CROSSBOWS);
     }
 
     public boolean isChargingCrossbow() {
@@ -128,7 +_,7 @@
     public AbstractIllager.IllagerArmPose getArmPose() {
         if (this.isChargingCrossbow()) {
             return AbstractIllager.IllagerArmPose.CROSSBOW_CHARGE;
-        } else if (this.isHolding(Items.CROSSBOW)) {
+        } else if (this.isHolding(is -> is.getItem() instanceof net.minecraft.world.item.CrossbowItem)) {
             return AbstractIllager.IllagerArmPose.CROSSBOW_HOLD;
         } else {
             return this.isAggressive() ? AbstractIllager.IllagerArmPose.ATTACKING : AbstractIllager.IllagerArmPose.NEUTRAL;
