--- a/net/minecraft/world/entity/monster/Ravager.java
+++ b/net/minecraft/world/entity/monster/Ravager.java
@@ -141,7 +_,7 @@
 
             if (this.level() instanceof ServerLevel serverlevel
                 && this.horizontalCollision
-                && serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                && net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverlevel, this)) {
                 boolean flag = false;
                 AABB aabb = this.getBoundingBox().inflate(0.2);
 
@@ -224,7 +_,7 @@
     private void roar() {
         if (this.isAlive()) {
             if (this.level() instanceof ServerLevel serverlevel) {
-                Predicate<Entity> predicate = serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+                Predicate<Entity> predicate = net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverlevel, this)
                     ? ROAR_TARGET_WITH_GRIEFING
                     : ROAR_TARGET_WITHOUT_GRIEFING;
 
