--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -89,7 +_,9 @@
     }
 
     protected void doFreezeConversion() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.STRAY, (timer) -> this.conversionTime = timer)) return;
         this.convertTo(EntityType.STRAY, ConversionParams.single(this, true, true), p_390244_ -> {
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_390244_);
             if (!this.isSilent()) {
                 this.level().levelEvent(null, 1048, this.blockPosition(), 0);
             }
