--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 
@@ -86,7 +_,9 @@
     }
 
     protected void doFreezeConversion() {
-        this.convertTo(EntityType.STRAY, true);
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.STRAY, (timer) -> this.conversionTime = timer)) return;
+        Stray stray = this.convertTo(EntityType.STRAY, true);
+        if (stray != null) net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, stray);
         if (!this.isSilent()) {
             this.level().levelEvent(null, 1048, this.blockPosition(), 0);
         }
