--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -257,6 +_,7 @@
     }
 
     protected void doUnderWaterConversion() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.DROWNED, (timer) -> this.conversionTime = timer)) return;
         this.convertToZombieType(EntityType.DROWNED);
         if (!this.isSilent()) {
             this.level().levelEvent(null, 1040, this.blockPosition(), 0);
@@ -267,7 +_,10 @@
         this.convertTo(
             p_34311_,
             ConversionParams.single(this, true, true),
-            p_390253_ -> p_390253_.handleAttributes(p_390253_.level().getCurrentDifficultyAt(p_390253_.blockPosition()).getSpecialMultiplier())
+            p_381517_ -> {
+                p_381517_.handleAttributes(p_381517_.level().getCurrentDifficultyAt(p_381517_.blockPosition()).getSpecialMultiplier());
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_381517_);
+            }
         );
     }
 
@@ -287,6 +_,7 @@
                 p_370686_.setGossips(p_371444_.getGossips().store(NbtOps.INSTANCE));
                 p_370686_.setTradeOffers(p_371444_.getOffers().copy());
                 p_370686_.setVillagerXp(p_371444_.getVillagerXp());
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(p_371444_, p_370686_);
                 if (!this.isSilent()) {
                     p_371466_.levelEvent(null, 1026, this.blockPosition(), 0);
                 }
@@ -437,7 +_,7 @@
     @Override
     public boolean killedEntity(ServerLevel p_219160_, LivingEntity p_219161_) {
         boolean flag = super.killedEntity(p_219160_, p_219161_);
-        if ((p_219160_.getDifficulty() == Difficulty.NORMAL || p_219160_.getDifficulty() == Difficulty.HARD) && p_219161_ instanceof Villager villager) {
+        if ((p_219160_.getDifficulty() == Difficulty.NORMAL || p_219160_.getDifficulty() == Difficulty.HARD) && p_219161_ instanceof Villager villager && net.neoforged.neoforge.event.EventHooks.canLivingConvert(p_219161_, EntityType.ZOMBIE_VILLAGER, (timer) -> {})) {
             if (p_219160_.getDifficulty() != Difficulty.HARD && this.random.nextBoolean()) {
                 return flag;
             }
