--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -136,7 +136,7 @@
       if (!this.level().isClientSide && this.isAlive() && this.isConverting()) {
          int i = this.getConversionProgress();
          this.villagerConversionTime -= i;
-         if (this.villagerConversionTime <= 0) {
+         if (this.villagerConversionTime <= 0 && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.VILLAGER, (timer) -> this.villagerConversionTime = timer)) {
             this.finishConversion((ServerLevel)this.level());
          }
       }
@@ -251,6 +251,7 @@
       if (!this.isSilent()) {
          p_34399_.levelEvent(null, 1027, this.blockPosition(), 0);
       }
+      net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, villager);
    }
 
    private int getConversionProgress() {
