--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -203,7 +203,7 @@
       if (!this.level().isClientSide && this.isAlive() && !this.isNoAi()) {
          if (this.isUnderWaterConverting()) {
             --this.conversionTime;
-            if (this.conversionTime < 0) {
+            if (this.conversionTime < 0 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.DROWNED, (timer) -> this.conversionTime = timer)) {
                this.doUnderWaterConversion();
             }
          } else if (this.convertsInWater()) {
@@ -265,9 +265,10 @@
       if (zombie != null) {
          zombie.handleAttributes(zombie.level().getCurrentDifficultyAt(zombie.blockPosition()).getSpecialMultiplier());
          zombie.setCanBreakDoors(zombie.supportsBreakDoorGoal() && this.canBreakDoors());
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombie);
       }
-   }
 
+   }
    protected boolean isSunSensitive() {
       return true;
    }
@@ -285,14 +286,19 @@
             livingentity = (LivingEntity)p_34288_.getEntity();
          }
 
-         if (livingentity != null
-            && this.level().getDifficulty() == Difficulty.HARD
-            && (double)this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE)
-            && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
-            int i = Mth.floor(this.getX());
-            int j = Mth.floor(this.getY());
-            int k = Mth.floor(this.getZ());
-            Zombie zombie = new Zombie(this.level());
+         int i = Mth.floor(this.getX());
+         int j = Mth.floor(this.getY());
+         int k = Mth.floor(this.getZ());
+
+         net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent event = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(this, level(), i, j, k, livingentity, this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).getValue());
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.DENY) return true;
+         if (event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW  ||
+                 (livingentity != null
+                 && this.level().getDifficulty() == Difficulty.HARD
+                 && (double)this.random.nextFloat() < this.getAttributeValue(Attributes.SPAWN_REINFORCEMENTS_CHANCE)
+                 && this.level().getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING))) {
+
+            Zombie zombie = event.getCustomSummonedAid() != null && event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW ? event.getCustomSummonedAid() : EntityType.ZOMBIE.create(this.level());
 
             for(int l = 0; l < 50; ++l) {
                int i1 = i + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
@@ -308,7 +314,9 @@
                      && this.level().isUnobstructed(zombie)
                      && this.level().noCollision(zombie)
                      && !this.level().containsAnyLiquid(zombie.getBoundingBox())) {
-                     zombie.setTarget(livingentity);
+                     if (livingentity != null) {
+                         zombie.setTarget(livingentity);
+                     }
                      zombie.finalizeSpawn(serverlevel, this.level().getCurrentDifficultyAt(zombie.blockPosition()), MobSpawnType.REINFORCEMENT, null, null);
                      serverlevel.addFreshEntityWithPassengers(zombie);
                      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE)
@@ -403,7 +411,7 @@
    @Override
    public boolean killedEntity(ServerLevel p_219160_, LivingEntity p_219161_) {
       boolean flag = super.killedEntity(p_219160_, p_219161_);
-      if ((p_219160_.getDifficulty() == Difficulty.NORMAL || p_219160_.getDifficulty() == Difficulty.HARD) && p_219161_ instanceof Villager villager) {
+      if ((p_219160_.getDifficulty() == Difficulty.NORMAL || p_219160_.getDifficulty() == Difficulty.HARD) && p_219161_ instanceof Villager villager && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(p_219161_, EntityType.ZOMBIE_VILLAGER, (timer) -> {})) {
          if (p_219160_.getDifficulty() != Difficulty.HARD && this.random.nextBoolean()) {
             return flag;
          }
@@ -421,6 +429,7 @@
             zombievillager.setGossips(villager.getGossips().store(NbtOps.INSTANCE));
             zombievillager.setTradeOffers(villager.getOffers().createTag());
             zombievillager.setVillagerXp(villager.getVillagerXp());
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(p_219161_, zombievillager);
             if (!this.isSilent()) {
                p_219160_.levelEvent(null, 1026, this.blockPosition(), 0);
             }
@@ -506,7 +515,7 @@
    }
 
    public static boolean getSpawnAsBabyOdds(RandomSource p_219163_) {
-      return p_219163_.nextFloat() < 0.05F;
+      return p_219163_.nextFloat() < net.minecraftforge.common.ForgeConfig.SERVER.zombieBabyChance.get();
    }
 
    protected void handleAttributes(float p_34340_) {
@@ -531,7 +540,7 @@
    }
 
    protected void randomizeReinforcementsChance() {
-      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.random.nextDouble() * 0.1F);
+      this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.random.nextDouble() * net.minecraftforge.common.ForgeConfig.SERVER.zombieBaseSummonChance.get());
    }
 
    @Override
