--- a/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
+++ b/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
@@ -56,12 +_,13 @@
             BlockStateModel blockstatemodel = this.blockModelShaper.getBlockModel(p_110919_);
             this.singleThreadRandom.setSeed(p_110919_.getSeed(p_110920_));
             this.singleThreadPartList.clear();
-            blockstatemodel.collectParts(this.singleThreadRandom, this.singleThreadPartList);
+            blockstatemodel.collectParts(p_110921_, p_110920_, p_110919_, this.singleThreadRandom, this.singleThreadPartList);
             this.modelRenderer
-                .tesselateBlock(p_110921_, this.singleThreadPartList, p_110919_, p_110920_, p_110922_, p_110923_, true, OverlayTexture.NO_OVERLAY);
+                .tesselateBlock(p_110921_, this.singleThreadPartList, p_110919_, p_110920_, p_110922_, type -> p_110923_, true, OverlayTexture.NO_OVERLAY);
         }
     }
 
+    @Deprecated // Neo: Buffer lookup parameter
     public void renderBatched(
         BlockState p_234356_,
         BlockPos p_234357_,
@@ -71,8 +_,20 @@
         boolean p_234361_,
         List<BlockModelPart> p_410643_
     ) {
+        this.renderBatched(p_234356_, p_234357_, p_234358_, p_234359_, type -> p_234360_, p_234361_, p_410643_);
+    }
+
+    public void renderBatched(
+        BlockState p_234356_,
+        BlockPos p_234357_,
+        BlockAndTintGetter p_234358_,
+        PoseStack p_234359_,
+        java.util.function.Function<net.minecraft.client.renderer.chunk.ChunkSectionLayer, VertexConsumer> bufferLookup,
+        boolean p_234361_,
+        List<BlockModelPart> p_410643_
+    ) {
         try {
-            this.modelRenderer.tesselateBlock(p_234358_, p_410643_, p_234356_, p_234357_, p_234359_, p_234360_, p_234361_, OverlayTexture.NO_OVERLAY);
+            this.modelRenderer.tesselateBlock(p_234358_, p_410643_, p_234356_, p_234357_, p_234359_, bufferLookup, p_234361_, OverlayTexture.NO_OVERLAY);
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Tesselating block in world");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Block being tesselated");
@@ -83,6 +_,7 @@
 
     public void renderLiquid(BlockPos p_234364_, BlockAndTintGetter p_234365_, VertexConsumer p_234366_, BlockState p_234367_, FluidState p_234368_) {
         try {
+            if (net.neoforged.neoforge.client.extensions.common.IClientFluidTypeExtensions.of(p_234368_).renderFluid(p_234368_, p_234365_, p_234364_, p_234366_, p_234367_)) return;
             this.liquidBlockRenderer.tesselate(p_234365_, p_234364_, p_234366_, p_234367_, p_234368_);
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Tesselating liquid in world");
@@ -100,7 +_,12 @@
         return this.blockModelShaper.getBlockModel(p_110911_);
     }
 
+    @Deprecated // Neo: Level and pos parameters
     public void renderSingleBlock(BlockState p_110913_, PoseStack p_110914_, MultiBufferSource p_110915_, int p_110916_, int p_110917_) {
+        renderSingleBlock(p_110913_, p_110914_, p_110915_, p_110916_, p_110917_, net.minecraft.world.level.EmptyBlockAndTintGetter.INSTANCE, BlockPos.ZERO);
+    }
+
+    public void renderSingleBlock(BlockState p_110913_, PoseStack p_110914_, MultiBufferSource p_110915_, int p_110916_, int p_110917_, BlockAndTintGetter level, BlockPos pos) {
         RenderShape rendershape = p_110913_.getRenderShape();
         if (rendershape != RenderShape.INVISIBLE) {
             BlockStateModel blockstatemodel = this.getBlockModel(p_110913_);
@@ -109,7 +_,7 @@
             float f1 = (i >> 8 & 0xFF) / 255.0F;
             float f2 = (i & 0xFF) / 255.0F;
             ModelBlockRenderer.renderModel(
-                p_110914_.last(), p_110915_.getBuffer(ItemBlockRenderTypes.getRenderType(p_110913_)), blockstatemodel, f, f1, f2, p_110916_, p_110917_
+                    p_110914_.last(), p_110915_, blockstatemodel, f, f1, f2, p_110916_, p_110917_, level, pos, p_110913_
             );
         }
     }
@@ -117,5 +_,9 @@
     @Override
     public void onResourceManagerReload(ResourceManager p_110909_) {
         this.liquidBlockRenderer.setupSprites(this.blockModelShaper, this.materials);
+    }
+
+    public LiquidBlockRenderer getLiquidBlockRenderer() {
+        return this.liquidBlockRenderer;
     }
 }
