--- a/net/minecraft/client/renderer/block/BlockStateModelSet.java
+++ b/net/minecraft/client/renderer/block/BlockStateModelSet.java
@@ -26,6 +_,10 @@
     }
 
     public Material.Baked getParticleMaterial(BlockState blockState) {
-        return this.get(blockState).particleMaterial();
+        return this.getParticleMaterial(blockState, net.minecraft.client.renderer.block.BlockAndTintGetter.EMPTY, net.minecraft.core.BlockPos.ZERO);
+    }
+
+    public Material.Baked getParticleMaterial(BlockState blockState, net.minecraft.client.renderer.block.BlockAndTintGetter level, net.minecraft.core.BlockPos pos) {
+        return this.get(blockState).particleMaterial(level, pos, blockState);
     }
 }
