--- a/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
+++ b/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
@@ -80,4 +_,8 @@
     public void onResourceManagerReload(ResourceManager resourceManager) {
         this.liquidBlockRenderer = new LiquidBlockRenderer(this.sprites);
     }
+
+    public LiquidBlockRenderer getLiquidBlockRenderer() {
+        return Objects.requireNonNull(this.liquidBlockRenderer);
+    }
 }
