--- a/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
+++ b/net/minecraft/client/renderer/block/BlockRenderDispatcher.java
@@ -40,14 +_,20 @@
         this.sprites = sprites;
     }
 
+    /// @deprecated Neo: use [#renderBreakingTexture(BlockAndTintGetter, BlockState, BlockPos, PoseStack, MultiBufferSource, int)] instead
+    @Deprecated
     public void renderBreakingTexture(BlockState state, BlockPos pos, PoseStack poseStack, MultiBufferSource bufferSource, int progress) {
+        renderBreakingTexture(BlockAndTintGetter.EMPTY, state, pos, poseStack, bufferSource, progress);
+    }
+
+    public void renderBreakingTexture(BlockAndTintGetter level, BlockState state, BlockPos pos, PoseStack poseStack, MultiBufferSource bufferSource, int progress) {
         if (state.getRenderShape() == RenderShape.MODEL) {
             PoseStack.Pose pose = poseStack.last();
             VertexConsumer buffer = new SheetedDecalTextureGenerator(bufferSource.getBuffer(ModelBakery.DESTROY_TYPES.get(progress)), pose, 1.0F);
             BlockStateModel model = this.modelManager.getBlockStateModelSet().get(state);
             this.singleThreadRandom.setSeed(state.getSeed(pos));
             this.singleThreadPartList.clear();
-            model.collectParts(this.singleThreadRandom, this.singleThreadPartList);
+            model.collectParts(level, pos, state, this.singleThreadRandom, this.singleThreadPartList);
             QuadInstance instance = new QuadInstance();
 
             for (BlockStateModelPart part : this.singleThreadPartList) {
