--- a/net/minecraft/client/renderer/block/BlockModelShaper.java
+++ b/net/minecraft/client/renderer/block/BlockModelShaper.java
@@ -17,8 +_,14 @@
         this.modelManager = p_110880_;
     }
 
+    @Deprecated
     public TextureAtlasSprite getParticleIcon(BlockState p_110883_) {
-        return this.getBlockModel(p_110883_).particleIcon();
+        return getParticleIcon(p_110883_, net.minecraft.world.level.EmptyBlockAndTintGetter.INSTANCE, net.minecraft.core.BlockPos.ZERO);
+    }
+
+    public TextureAtlasSprite getParticleIcon(BlockState p_110883_, net.minecraft.world.level.BlockAndTintGetter level, net.minecraft.core.BlockPos pos) {
+        var model = this.getBlockModel(p_110883_);
+        return model.particleIcon(level, pos, p_110883_);
     }
 
     public BlockStateModel getBlockModel(BlockState p_110894_) {
