--- a/net/minecraft/client/renderer/block/FluidStateModelSet.java
+++ b/net/minecraft/client/renderer/block/FluidStateModelSet.java
@@ -17,7 +_,7 @@
         new Material(Identifier.withDefaultNamespace("block/water_still")),
         new Material(Identifier.withDefaultNamespace("block/water_flow")),
         new Material(Identifier.withDefaultNamespace("block/water_overlay")),
-        BlockTintSources.water()
+        net.neoforged.neoforge.client.fluid.FluidTintSources.water()
     );
     private static final FluidModel.Unbaked LAVA_MODEL = new FluidModel.Unbaked(
         new Material(Identifier.withDefaultNamespace("block/lava_still")), new Material(Identifier.withDefaultNamespace("block/lava_flow")), null, null
@@ -33,7 +_,8 @@
     public static Map<Fluid, FluidModel> bake(MaterialBaker materials) {
         FluidModel waterModel = WATER_MODEL.bake(materials, () -> "Water");
         FluidModel lavaModel = LAVA_MODEL.bake(materials, () -> "Lava");
-        return Map.of(Fluids.WATER, waterModel, Fluids.FLOWING_WATER, waterModel, Fluids.LAVA, lavaModel, Fluids.FLOWING_LAVA, lavaModel);
+        Map<Fluid, FluidModel> models = Map.of(Fluids.WATER, waterModel, Fluids.FLOWING_WATER, waterModel, Fluids.LAVA, lavaModel, Fluids.FLOWING_LAVA, lavaModel);
+        return net.neoforged.neoforge.client.ClientHooks.gatherFluidModels(models, materials);
     }
 
     public FluidModel get(FluidState state) {
