--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -259,8 +_,7 @@
                     BlockPos blockpos = p_234371_.relative(direction);
                     TextureAtlasSprite textureatlassprite2 = atextureatlassprite[1];
                     if (!flag) {
-                        Block block = p_234370_.getBlockState(blockpos).getBlock();
-                        if (block instanceof HalfTransparentBlock || block instanceof LeavesBlock) {
+                        if (p_234370_.getBlockState(blockpos).shouldDisplayFluidOverlay(p_234370_, blockpos, p_234374_)) {
                             textureatlassprite2 = this.waterOverlay;
                         }
                     }
