--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -41,12 +_,17 @@
         this.waterStill = sprites.get(ModelBakery.WATER_STILL);
         this.waterFlowing = sprites.get(ModelBakery.WATER_FLOW);
         this.waterOverlay = sprites.get(ModelBakery.WATER_OVERLAY);
+        net.neoforged.neoforge.client.textures.FluidSpriteCache.reload();
     }
 
     private static boolean isNeighborSameFluid(FluidState fluidState, FluidState neighborFluidState) {
         return neighborFluidState.getType().isSame(fluidState.getType());
     }
 
+    private static boolean isNeighborStateHidingOverlay(FluidState selfState, BlockState otherState, Direction neighborFace) {
+        return otherState.shouldHideAdjacentFluidFace(neighborFace, selfState);
+    }
+
     private static boolean isFaceOccludedByState(Direction direction, float height, BlockState state) {
         VoxelShape occluder = state.getFaceOcclusionShape(direction.getOpposite());
         if (occluder == Shapes.empty()) {
@@ -68,10 +_,16 @@
         return isFaceOccludedByState(direction.getOpposite(), 1.0F, state);
     }
 
+    /** @deprecated Neo: use overload that accepts BlockState */
+    @Deprecated
     public static boolean shouldRenderFace(FluidState fluidState, BlockState blockState, Direction direction, FluidState neighborFluidState) {
         return !isNeighborSameFluid(fluidState, neighborFluidState) && !isFaceOccludedBySelf(blockState, direction);
     }
 
+    public static boolean shouldRenderFace(FluidState fluidState, BlockState selfState, Direction direction, BlockState otherState) {
+        return !isNeighborStateHidingOverlay(fluidState, otherState, direction.getOpposite()) && !isFaceOccludedBySelf(selfState, direction);
+    }
+
     public void tesselate(BlockAndTintGetter level, BlockPos pos, VertexConsumer builder, BlockState blockState, FluidState fluidState) {
         BlockState blockStateDown = level.getBlockState(pos.relative(Direction.DOWN));
         FluidState fluidStateDown = blockStateDown.getFluidState();
@@ -85,18 +_,18 @@
         FluidState fluidStateWest = blockStateWest.getFluidState();
         BlockState blockStateEast = level.getBlockState(pos.relative(Direction.EAST));
         FluidState fluidStateEast = blockStateEast.getFluidState();
-        boolean renderUp = !isNeighborSameFluid(fluidState, fluidStateUp);
-        boolean renderDown = shouldRenderFace(fluidState, blockState, Direction.DOWN, fluidStateDown)
+        boolean renderUp = !isNeighborStateHidingOverlay(fluidState, blockStateUp, Direction.UP);
+        boolean renderDown = shouldRenderFace(fluidState, blockState, Direction.DOWN, blockStateDown)
             && !isFaceOccludedByNeighbor(Direction.DOWN, 0.8888889F, blockStateDown);
-        boolean renderNorth = shouldRenderFace(fluidState, blockState, Direction.NORTH, fluidStateNorth);
-        boolean renderSouth = shouldRenderFace(fluidState, blockState, Direction.SOUTH, fluidStateSouth);
-        boolean renderWest = shouldRenderFace(fluidState, blockState, Direction.WEST, fluidStateWest);
-        boolean renderEast = shouldRenderFace(fluidState, blockState, Direction.EAST, fluidStateEast);
+        boolean renderNorth = shouldRenderFace(fluidState, blockState, Direction.NORTH, blockStateNorth);
+        boolean renderSouth = shouldRenderFace(fluidState, blockState, Direction.SOUTH, blockStateSouth);
+        boolean renderWest = shouldRenderFace(fluidState, blockState, Direction.WEST, blockStateWest);
+        boolean renderEast = shouldRenderFace(fluidState, blockState, Direction.EAST, blockStateEast);
         if (renderUp || renderDown || renderEast || renderWest || renderNorth || renderSouth) {
-            boolean isLava = fluidState.is(FluidTags.LAVA);
-            TextureAtlasSprite stillSprite = isLava ? this.lavaStill : this.waterStill;
-            TextureAtlasSprite flowingSprite = isLava ? this.lavaFlowing : this.waterFlowing;
-            int color = isLava ? -1 : BiomeColors.getAverageWaterColor(level, pos);
+            TextureAtlasSprite[] fluidSprites = net.neoforged.neoforge.client.textures.FluidSpriteCache.getFluidSprites(level, pos, fluidState);
+            TextureAtlasSprite stillSprite = fluidSprites[0]; // Neo: Custom still sprite
+            TextureAtlasSprite flowingSprite = fluidSprites[1]; // Neo: Custom flowing sprite
+            int color = net.neoforged.neoforge.client.extensions.common.IClientFluidTypeExtensions.of(fluidState).getTintColor(fluidState, level, pos);
             float shadeDown = level.getShade(Direction.DOWN, true);
             float shadeUp = level.getShade(Direction.UP, true);
             float shadeNorth = level.getShade(Direction.NORTH, true);
@@ -255,10 +_,9 @@
                 if (renderCondition && !isFaceOccludedByNeighbor(faceDir, Math.max(hh0, hh1), level.getBlockState(pos.relative(faceDir)))) {
                     BlockPos tPos = pos.relative(faceDir);
                     TextureAtlasSprite sprite = flowingSprite;
-                    if (!isLava) {
-                        Block relativeBlock = level.getBlockState(tPos).getBlock();
-                        if (relativeBlock instanceof HalfTransparentBlock || relativeBlock instanceof LeavesBlock) {
-                            sprite = this.waterOverlay;
+                    if (fluidSprites[2] != null) {
+                        if (level.getBlockState(tPos).shouldDisplayFluidOverlay(level, tPos, fluidState)) {
+                            sprite = fluidSprites[2];
                         }
                     }
 
@@ -318,6 +_,26 @@
     private float getHeight(BlockAndTintGetter level, Fluid fluidType, BlockPos pos) {
         BlockState state = level.getBlockState(pos);
         return this.getHeight(level, fluidType, pos, state, state.getFluidState());
+    }
+
+    private void vertex(
+            VertexConsumer builder,
+            float x,
+            float y,
+            float z,
+            float red,
+            float green,
+            float blue,
+            float alpha,
+            float u,
+            float v,
+            int light
+    ) {
+        builder.addVertex(x, y, z)
+                .setColor(red, green, blue, alpha)
+                .setUv(u, v)
+                .setLight(light)
+                .setNormal(0.0F, 1.0F, 0.0F);
     }
 
     private float getHeight(BlockAndTintGetter level, Fluid fluidType, BlockPos pos, BlockState state, FluidState fluidState) {
