--- /dev/null
+++ b/net/minecraft/client/renderer/block/ModelBlockRenderer.java.rej
@@ -1,0 +_,107 @@
+diff a/projects/neoforge/src/main/java/net/minecraft/client/renderer/block/ModelBlockRenderer.java b/projects/neoforge/src/main/java/net/minecraft/client/renderer/block/ModelBlockRenderer.java	(rejected hunks)
+@@ -130,7 +130,8 @@ public class ModelBlockRenderer {
+         boolean p_234397_,
+         int p_234400_
+     ) {
+-        ModelBlockRenderer.AmbientOcclusionRenderStorage modelblockrenderer$ambientocclusionrenderstorage = new ModelBlockRenderer.AmbientOcclusionRenderStorage();
++        // Neo: Inject enhanced AO pipeline
++        ModelBlockRenderer.AmbientOcclusionRenderStorage modelblockrenderer$ambientocclusionrenderstorage = net.neoforged.neoforge.client.model.ao.EnhancedAoRenderStorage.newInstance();
+         boolean perPartAO = net.neoforged.neoforge.client.config.NeoForgeClientConfig.INSTANCE.handleAmbientOcclusionPerPart.getAsBoolean();
+         int lightEmission = -1;
+         int i = 0;
+@@ -275,8 +276,12 @@ public class ModelBlockRenderer {
+         int p_111022_
+     ) {
+         for (BakedQuad bakedquad : p_111018_) {
++            if (!bakedquad.hasAmbientOcclusion()) {
++                renderModelQuadFlat(p_111013_, p_111014_, p_111015_, -1, p_111022_, true, p_111016_, p_111017_, bakedquad, p_412172_);
++                continue;
++            }
+             calculateShape(p_111013_, p_111014_, p_111015_, bakedquad.vertices(), bakedquad.direction(), p_412172_);
+-            if (!net.neoforged.neoforge.client.ClientHooks.calculateFaceWithoutAO(p_111013_, p_111014_, p_111015_, bakedquad, p_412172_.cache, p_412172_.faceCubic, p_412172_.brightness, p_412172_.lightmap))
++            p_412172_.captureQuad(bakedquad);
+             p_412172_.calculate(p_111013_, p_111014_, p_111015_, bakedquad.direction(), bakedquad.shade());
+             this.putQuadData(p_111013_, p_111014_, p_111015_, p_111017_, p_111016_.last(), bakedquad, p_412172_, p_111022_);
+         }
+@@ -392,17 +397,30 @@ public class ModelBlockRenderer {
+         ModelBlockRenderer.CommonRenderStorage p_412163_
+     ) {
+         for (BakedQuad bakedquad : p_111010_) {
++            renderModelQuadFlat(p_111002_, p_111003_, p_111004_, p_111005_, p_111006_, p_111007_, p_111008_, p_111009_, bakedquad, p_412163_);
++        }
++    }
++
++    private void renderModelQuadFlat(
++        BlockAndTintGetter p_111002_,
++        BlockState p_111003_,
++        BlockPos p_111004_,
++        int p_111005_,
++        int p_111006_,
++        boolean p_111007_,
++        PoseStack p_111008_,
++        VertexConsumer p_111009_,
++        BakedQuad bakedquad,
++        ModelBlockRenderer.CommonRenderStorage p_412163_
++    ) {
++        {
+             if (p_111007_) {
+                 calculateShape(p_111002_, p_111003_, p_111004_, bakedquad.vertices(), bakedquad.direction(), p_412163_);
+                 BlockPos blockpos = (BlockPos)(p_412163_.faceCubic ? p_412163_.scratchPos.setWithOffset(p_111004_, bakedquad.direction()) : p_111004_);
+                 p_111005_ = p_412163_.cache.getLightColor(p_111003_, p_111002_, blockpos);
+             }
+ 
+-            float f = p_111002_.getShade(bakedquad.direction(), bakedquad.shade());
+-            p_412163_.brightness[0] = f;
+-            p_412163_.brightness[1] = f;
+-            p_412163_.brightness[2] = f;
+-            p_412163_.brightness[3] = f;
++            net.neoforged.neoforge.client.model.ao.EnhancedAoRenderStorage.applyFlatQuadBrightness(p_111002_, bakedquad, p_412163_);
+             p_412163_.lightmap[0] = p_111005_;
+             p_412163_.lightmap[1] = p_111005_;
+             p_412163_.lightmap[2] = p_111005_;
+@@ -795,12 +813,15 @@ public class ModelBlockRenderer {
+     }
+ 
+     @OnlyIn(Dist.CLIENT)
+-    static class AmbientOcclusionRenderStorage extends ModelBlockRenderer.CommonRenderStorage {
+-        final float[] faceShape = new float[ModelBlockRenderer.SizeInfo.COUNT];
++    public static class AmbientOcclusionRenderStorage extends ModelBlockRenderer.CommonRenderStorage {
++        protected final float[] faceShape = new float[ModelBlockRenderer.SizeInfo.COUNT];
+ 
+         public AmbientOcclusionRenderStorage() {
+         }
+ 
++        // Neo: Call this before calling calculate if the render storage is an EnhancedAoRenderStorage!
++        public void captureQuad(BakedQuad quad) {}
++
+         public void calculate(BlockAndTintGetter p_412286_, BlockState p_412102_, BlockPos p_412730_, Direction p_412336_, boolean p_412430_) {
+             BlockPos blockpos = this.faceCubic ? p_412730_.relative(p_412336_) : p_412730_;
+             ModelBlockRenderer.AdjacencyInfo modelblockrenderer$adjacencyinfo = ModelBlockRenderer.AdjacencyInfo.fromFacing(p_412336_);
+@@ -983,7 +1004,7 @@ public class ModelBlockRenderer {
+             return p_412676_ + p_412469_ + p_412736_ + p_412253_ >> 2 & 16711935;
+         }
+ 
+-        private static int blend(int p_412759_, int p_412518_, int p_412118_, int p_412415_, float p_412783_, float p_412051_, float p_412530_, float p_412587_) {
++        protected static int blend(int p_412759_, int p_412518_, int p_412118_, int p_412415_, float p_412783_, float p_412051_, float p_412530_, float p_412587_) {
+             int i = (int)(
+                     (p_412759_ >> 16 & 0xFF) * p_412783_
+                         + (p_412518_ >> 16 & 0xFF) * p_412051_
+@@ -1110,7 +1131,8 @@ public class ModelBlockRenderer {
+     }
+ 
+     @OnlyIn(Dist.CLIENT)
+-    static class CommonRenderStorage {
++    public static class CommonRenderStorage {
++        CommonRenderStorage() {}
+         public final BlockPos.MutableBlockPos scratchPos = new BlockPos.MutableBlockPos();
+         public boolean faceCubic;
+         public boolean facePartial;
+@@ -1137,7 +1159,7 @@ public class ModelBlockRenderer {
+         FLIP_EAST(11);
+ 
+         public static final int COUNT = values().length;
+-        final int index;
++        public final int index;
+ 
+         private SizeInfo(int p_412421_) {
+             this.index = p_412421_;
