--- a/net/minecraft/client/renderer/block/ModelBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/ModelBlockRenderer.java
@@ -78,7 +_,7 @@
                     List<BakedQuad> culledQuads = part.getQuads(direction);
                     if (!culledQuads.isEmpty()) {
                         if (!validCacheForDirection) {
-                            shouldRenderFace = this.shouldRenderFace(level, state, direction, this.scratchPos.setWithOffset(pos, direction));
+                            shouldRenderFace = shouldRenderFace(level, pos, state, cull, direction, scratch.scratchPos.setWithOffset(pos, direction));
                             cacheValid |= cacheMask;
                             if (shouldRenderFace) {
                                 shouldRenderFaceCache |= cacheMask;
@@ -118,7 +_,7 @@
                     if (!culledQuads.isEmpty()) {
                         BlockPos relativePos = this.scratchPos.setWithOffset(pos, direction);
                         if (!validCacheForDirection) {
-                            shouldRenderFace = this.shouldRenderFace(level, state, direction, relativePos);
+                            shouldRenderFace = shouldRenderFace(level, pos, state, cull, direction, relativePos);
                             cacheValid |= cacheMask;
                             if (shouldRenderFace) {
                                 shouldRenderFaceCache |= cacheMask;
