--- a/net/minecraft/network/protocol/BundlePacket.java
+++ b/net/minecraft/network/protocol/BundlePacket.java
@@ -3,14 +_,18 @@
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.PacketListener;
 
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
 public abstract class BundlePacket<T extends PacketListener> implements Packet<T> {
-    private final Iterable<Packet<T>> packets;
+    private final Iterable<Packet<? super T>> packets;
 
-    protected BundlePacket(Iterable<Packet<T>> p_265290_) {
-        this.packets = p_265290_;
+    protected BundlePacket(Iterable<Packet<? super T>> p_265290_) {
+        this.packets = net.neoforged.neoforge.network.bundle.BundlePacketUtils.flatten(p_265290_);
     }
 
-    public final Iterable<Packet<T>> subPackets() {
+    public final Iterable<Packet<? super T>> subPackets() {
         return this.packets;
     }
 
