--- a/net/minecraft/client/ClientClockManager.java
+++ b/net/minecraft/client/ClientClockManager.java
@@ -24,7 +24,11 @@
         for (ClientClockManager.ClockInstance instance : this.clocks.values()) {
             if (!instance.paused) {
-                instance.totalTicks += gameTimeDelta;
+                // Neo: support variable-speed clocks
+                float timeStep = instance.fractionalTick + gameTimeDelta * instance.speed;
+                long result = (long)timeStep;
+                instance.fractionalTick = (timeStep - result);
+                instance.totalTicks += result;
             }
         }
     }
