--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -81,6 +_,13 @@
         this.key = p_265775_;
         this.fallback = p_265204_;
         this.args = p_265752_;
+        if (!net.neoforged.fml.loading.FMLEnvironment.production) {
+            for (Object arg : this.args) {
+                if (!(arg instanceof Component) && !isAllowedPrimitiveArgument(arg)) {
+                    throw new IllegalArgumentException("TranslatableContents' arguments must be either a Component, Number, Boolean, or a String. Was given " + arg + " for " + this.key);
+                }
+            }
+        }
     }
 
     @Override
