--- a/net/minecraft/network/chat/contents/TranslatableContents.java
+++ b/net/minecraft/network/chat/contents/TranslatableContents.java
@@ -81,6 +_,13 @@
         this.key = p_265775_;
         this.fallback = p_265204_;
         this.args = p_265752_;
+        if (!net.neoforged.fml.loading.FMLEnvironment.production) {
+            for (Object arg : this.args) {
+                if (!(arg instanceof Component) && !isAllowedPrimitiveArgument(arg)) {
+                    throw new IllegalArgumentException("TranslatableContents' arguments must be either a Component, Number, Boolean, or a String. Was given " + arg + " for " + this.key);
+                }
+            }
+        }
     }
 
     @Override
@@ -92,6 +_,13 @@
         Language language = Language.getInstance();
         if (language != this.decomposedWith) {
             this.decomposedWith = language;
+
+            Component langComponent = language.getComponent(this.key);
+            if (langComponent != null) {
+                this.decomposedParts = ImmutableList.of(langComponent);
+                return;
+            }
+
             String s = this.fallback != null ? language.getOrDefault(this.key, this.fallback) : language.getOrDefault(this.key);
 
             try {
@@ -170,28 +_,46 @@
     public <T> Optional<T> visit(FormattedText.StyledContentConsumer<T> p_237521_, Style p_237522_) {
         this.decompose();
 
-        for (FormattedText formattedtext : this.decomposedParts) {
-            Optional<T> optional = formattedtext.visit(p_237521_, p_237522_);
-            if (optional.isPresent()) {
-                return optional;
+        if (!net.neoforged.neoforge.common.util.InsertingContents.pushTranslation(this)) {
+            // Reference cycle.
+            return Optional.empty();
+        }
+
+        try {
+            for (FormattedText formattedtext : this.decomposedParts) {
+                Optional<T> optional = formattedtext.visit(p_237521_, p_237522_);
+                if (optional.isPresent()) {
+                    return optional;
+                }
             }
-        }
 
-        return Optional.empty();
+            return Optional.empty();
+        } finally {
+            net.neoforged.neoforge.common.util.InsertingContents.popTranslation();
+        }
     }
 
     @Override
     public <T> Optional<T> visit(FormattedText.ContentConsumer<T> p_237519_) {
         this.decompose();
 
-        for (FormattedText formattedtext : this.decomposedParts) {
-            Optional<T> optional = formattedtext.visit(p_237519_);
-            if (optional.isPresent()) {
-                return optional;
+        if (!net.neoforged.neoforge.common.util.InsertingContents.pushTranslation(this)) {
+            // Reference cycle.
+            return Optional.empty();
+        }
+
+        try {
+            for (FormattedText formattedtext : this.decomposedParts) {
+                Optional<T> optional = formattedtext.visit(p_237519_);
+                if (optional.isPresent()) {
+                    return optional;
+                }
             }
-        }
 
-        return Optional.empty();
+            return Optional.empty();
+        } finally {
+            net.neoforged.neoforge.common.util.InsertingContents.popTranslation();
+        }
     }
 
     @Override
