--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
@@ -38,19 +_,20 @@
       }
 
       if (!list.isEmpty()) {
+         var eventProvider = net.neoforged.neoforge.event.EventHooks.alterGround(p_225969_, list, this.provider::getState);
          int i = list.get(0).getY();
          list.stream().filter(p_69310_ -> p_69310_.getY() == i).forEach(p_225978_ -> {
-            this.placeCircle(p_225969_, p_225978_.west().north());
-            this.placeCircle(p_225969_, p_225978_.east(2).north());
-            this.placeCircle(p_225969_, p_225978_.west().south(2));
-            this.placeCircle(p_225969_, p_225978_.east(2).south(2));
+            this.placeCircle(p_225969_, p_225978_.west().north(), eventProvider);
+            this.placeCircle(p_225969_, p_225978_.east(2).north(), eventProvider);
+            this.placeCircle(p_225969_, p_225978_.west().south(2), eventProvider);
+            this.placeCircle(p_225969_, p_225978_.east(2).south(2), eventProvider);
 
             for(int j = 0; j < 5; ++j) {
                int k = p_225969_.random().nextInt(64);
                int l = k % 8;
                int i1 = k / 8;
                if (l == 0 || l == 7 || i1 == 0 || i1 == 7) {
-                  this.placeCircle(p_225969_, p_225978_.offset(-3 + l, 0, -3 + i1));
+                  this.placeCircle(p_225969_, p_225978_.offset(-3 + l, 0, -3 + i1), eventProvider);
                }
             }
          });
@@ -58,20 +_,28 @@
    }
 
    private void placeCircle(TreeDecorator.Context p_225971_, BlockPos p_225972_) {
+      placeCircle(p_225971_, p_225972_, this.provider::getState);
+   }
+
+   private void placeCircle(TreeDecorator.Context p_225971_, BlockPos p_225972_, net.neoforged.neoforge.event.level.AlterGroundEvent.StateProvider eventProvider) {
       for(int i = -2; i <= 2; ++i) {
          for(int j = -2; j <= 2; ++j) {
             if (Math.abs(i) != 2 || Math.abs(j) != 2) {
-               this.placeBlockAt(p_225971_, p_225972_.offset(i, 0, j));
+               this.placeBlockAt(p_225971_, p_225972_.offset(i, 0, j), eventProvider);
             }
          }
       }
    }
 
    private void placeBlockAt(TreeDecorator.Context p_225974_, BlockPos p_225975_) {
+      placeCircle(p_225974_, p_225975_, this.provider::getState);
+   }
+
+   private void placeBlockAt(TreeDecorator.Context p_225974_, BlockPos p_225975_, net.neoforged.neoforge.event.level.AlterGroundEvent.StateProvider eventProvider) {
       for(int i = 2; i >= -3; --i) {
          BlockPos blockpos = p_225975_.above(i);
          if (Feature.isGrassOrDirt(p_225974_.level(), blockpos)) {
-            p_225974_.setBlock(blockpos, this.provider.getState(p_225974_.random(), p_225975_));
+            p_225974_.setBlock(blockpos, eventProvider.getState(p_225974_.random(), p_225975_));
             break;
          }
 
