--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AlterGroundDecorator.java
@@ -26,19 +_,20 @@
     public void place(TreeDecorator.Context context) {
         List<BlockPos> blockPositions = TreeFeature.getLowestTrunkOrRootOfTree(context);
         if (!blockPositions.isEmpty()) {
+            var eventProvider = net.neoforged.neoforge.event.EventHooks.alterGround(context, blockPositions, this.provider::getState);
             int minY = blockPositions.getFirst().getY();
             blockPositions.stream().filter(pos -> pos.getY() == minY).forEach(pos -> {
-                this.placeCircle(context, pos.west().north());
-                this.placeCircle(context, pos.east(2).north());
-                this.placeCircle(context, pos.west().south(2));
-                this.placeCircle(context, pos.east(2).south(2));
+                this.placeCircle(context, pos.west().north(), eventProvider);
+                this.placeCircle(context, pos.east(2).north(), eventProvider);
+                this.placeCircle(context, pos.west().south(2), eventProvider);
+                this.placeCircle(context, pos.east(2).south(2), eventProvider);
 
                 for (int i = 0; i < 5; i++) {
                     int placement = context.random().nextInt(64);
                     int xx = placement % 8;
                     int zz = placement / 8;
                     if (xx == 0 || xx == 7 || zz == 0 || zz == 7) {
-                        this.placeCircle(context, pos.offset(-3 + xx, 0, -3 + zz));
+                        this.placeCircle(context, pos.offset(-3 + xx, 0, -3 + zz), eventProvider);
                     }
                 }
             });
