--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -58,14 +_,12 @@
         if (this.uncategorizedStackTrace != null && this.uncategorizedStackTrace.length > 0) {
             p_127520_.append("-- Head --\n");
             p_127520_.append("Thread: ").append(Thread.currentThread().getName()).append("\n");
-            p_127520_.append("Stacktrace:\n");
+            p_127520_.append("Stacktrace:");
 
             for (StackTraceElement stacktraceelement : this.uncategorizedStackTrace) {
                 p_127520_.append("\t").append("at ").append(stacktraceelement);
                 p_127520_.append("\n");
             }
-
-            p_127520_.append("\n");
         }
 
         for (CrashReportCategory crashreportcategory : this.details) {
@@ -73,6 +_,7 @@
             p_127520_.append("\n\n");
         }
 
+        net.neoforged.neoforge.logging.CrashReportExtender.extendSystemReport(systemReport);
         this.systemReport.appendToCrashReportString(p_127520_);
     }
 
