--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -58,14 +_,8 @@
         if (this.uncategorizedStackTrace != null && this.uncategorizedStackTrace.length > 0) {
             builder.append("-- Head --\n");
             builder.append("Thread: ").append(Thread.currentThread().getName()).append("\n");
-            builder.append("Stacktrace:\n");
-
-            for (StackTraceElement stacktraceelement : this.uncategorizedStackTrace) {
-                builder.append("\t").append("at ").append(stacktraceelement);
-                builder.append("\n");
-            }
-
-            builder.append("\n");
+            builder.append("Stacktrace:");
+            builder.append(net.neoforged.neoforge.logging.CrashReportExtender.generateEnhancedStackTrace(this.uncategorizedStackTrace));
         }
 
         for (CrashReportCategory crashreportcategory : this.details) {
@@ -73,6 +_,7 @@
             builder.append("\n\n");
         }
 
+        net.neoforged.neoforge.logging.CrashReportExtender.extendSystemReport(systemReport);
         this.systemReport.appendToCrashReportString(builder);
     }
 
@@ -92,18 +_,7 @@
             throwable.setStackTrace(this.exception.getStackTrace());
         }
 
-        String s;
-        try {
-            stringwriter = new StringWriter();
-            printwriter = new PrintWriter(stringwriter);
-            throwable.printStackTrace(printwriter);
-            s = stringwriter.toString();
-        } finally {
-            IOUtils.closeQuietly((Writer)stringwriter);
-            IOUtils.closeQuietly((Writer)printwriter);
-        }
-
-        return s;
+        return net.neoforged.neoforge.logging.CrashReportExtender.generateEnhancedStackTrace(throwable);
     }
 
     public String getFriendlyReport(ReportType type, List<String> links) {
