--- a/net/minecraft/FileUtil.java
+++ b/net/minecraft/FileUtil.java
@@ -18,7 +_,7 @@
 public class FileUtil {
     private static final Pattern COPY_COUNTER_PATTERN = Pattern.compile("(<name>.*) \\((<count>\\d*)\\)", 66);
     private static final int MAX_FILE_NAME = 255;
-    private static final Pattern RESERVED_WINDOWS_FILENAMES = Pattern.compile(".*\\.|(?:COM|CLOCK\\$|CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])(?:\\..*)?", 2);
+    private static final Pattern RESERVED_WINDOWS_FILENAMES = Pattern.compile(".*\\.|(?:COM¹|COM²|COM³|LPT¹|LPT²|LPT³|CLOCK\\$|CON|PRN|AUX|NUL|COM[0-9]|LPT[0-9])(?:\\..*)?", 2); // NEO: Fix MC-268617
     private static final Pattern STRICT_PATH_SEGMENT_CHECK = Pattern.compile("[-._a-z0-9]+");
 
     public static String findAvailableName(Path p_133731_, String p_133732_, String p_133733_) throws IOException {
