--- a/net/minecraft/ReportType.java
+++ b/net/minecraft/ReportType.java
@@ -131,6 +_,7 @@
         builder.append("---- ");
         builder.append(this.header());
         builder.append(" ----\n");
+        net.neoforged.fml.CrashReportCallables.getHeaders().forEach(header -> builder.append(header).append('\n'));
         builder.append("// ");
         builder.append(this.getErrorComment());
         builder.append('\n');
