--- a/net/minecraft/Util.java
+++ b/net/minecraft/Util.java
@@ -267,8 +_,8 @@
                 .getSchema(DataFixUtils.makeKey(SharedConstants.getCurrentVersion().getDataVersion().getVersion()))
                 .getChoiceType(p_type, choiceName);
         } catch (IllegalArgumentException illegalargumentexception) {
-            LOGGER.error("No data fixer registered for {}", choiceName);
-            if (SharedConstants.IS_RUNNING_IN_IDE) {
+            LOGGER.debug("No data fixer registered for {}", choiceName);
+            if (SharedConstants.IS_RUNNING_IN_IDE && false) {
                 throw illegalargumentexception;
             }
         }
@@ -526,20 +_,20 @@
 
     public static void logAndPauseIfInIde(String error) {
         LOGGER.error(error);
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             doPause(error);
         }
     }
 
     public static void logAndPauseIfInIde(String message, Throwable error) {
         LOGGER.error(message, error);
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             doPause(message);
         }
     }
 
     public static <T extends Throwable> T pauseInIde(T throwable) {
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_WITH_JDWP) {
             LOGGER.error("Trying to throw a fatal exception, pausing in IDE", throwable);
             doPause(throwable.getMessage());
         }
