--- a/net/minecraft/world/entity/animal/fish/AbstractFish.java
+++ b/net/minecraft/world/entity/animal/fish/AbstractFish.java
@@ -167,7 +_,7 @@
 
         @Override
         public void tick() {
-            if (this.mob.isEyeInFluid(FluidTags.WATER)) {
+            if (this.mob.getFluidInteraction().isEyeInFluidMatching(this.mob, (_, type, _) -> type.getIsWaterLike())) {
                 this.mob.setDeltaMovement(this.mob.getDeltaMovement().add(0.0, 0.005, 0.0));
             }
 
