--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -390,6 +_,7 @@
             this.modified = true;
         }
 
+        net.neoforged.neoforge.event.EventHooks.fireChunkTicketLevelUpdated(this.level, node, oldLevel, level, chunk);
         return chunk;
     }
 
@@ -524,9 +_,13 @@
 
                     this.save(chunk);
                     if (chunk instanceof LevelChunk levelChunk) {
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkEvent.Unload(levelChunk));
                         this.level.unload(levelChunk);
                     }
 
+                    net.neoforged.neoforge.common.CommonHooks.onChunkUnload(this.poiManager, chunk); // Neo: Must be called for all chunk unloading. Not just LevelChunks.
+                    this.chunkTypeCache.remove(chunk.getPos().pack()); // Neo: Prevent chunk type cache from permanently retaining data for unloaded chunks
+
                     this.lightEngine.updateChunkStatus(chunk.getPos());
                     this.lightEngine.tryScheduleUpdate();
                     this.nextChunkSaveTime.remove(chunk.getPos().pack());
@@ -566,6 +_,7 @@
                 Profiler.get().incrementCounter("chunkLoad");
                 if (chunkData.isPresent()) {
                     ChunkAccess chunk = chunkData.get().read(this.level, this.poiManager, this.storageInfo(), pos);
+                    net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkDataEvent.Load(chunk, chunkData.get()));
                     this.markPosition(pos, chunk.getPersistedStatus().getChunkType());
                     return chunk;
                 } else {
@@ -762,6 +_,7 @@
             Profiler.get().incrementCounter("chunkSave");
             this.activeChunkWrites.incrementAndGet();
             SerializableChunkData data = SerializableChunkData.copyOf(this.level, chunk);
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkDataEvent.Save(chunk, this.level, data));
             CompletableFuture<CompoundTag> encodedData = CompletableFuture.supplyAsync(data::write, Util.backgroundExecutor());
             this.write(pos, encodedData::join).handle((ignored, throwable) -> {
                 if (throwable != null) {
@@ -827,9 +_,11 @@
 
     private static void markChunkPendingToSend(ServerPlayer player, LevelChunk chunk) {
         player.connection.chunkSender.markChunkPendingToSend(chunk);
+        net.neoforged.neoforge.event.EventHooks.fireChunkWatch(player, chunk, player.level());
     }
 
     private static void dropChunk(ServerPlayer player, ChunkPos pos) {
+        net.neoforged.neoforge.event.EventHooks.fireChunkUnWatch(player, pos, player.level());
         player.connection.chunkSender.dropChunk(player, pos);
     }
 
@@ -937,7 +_,7 @@
             ChunkHolder holder = this.visibleChunkMap.get(spawnCandidateChunks.nextLong());
             if (holder != null) {
                 LevelChunk chunk = holder.getTickingChunk();
-                if (chunk != null && this.anyPlayerCloseEnoughForSpawningInternal(holder.getPos())) {
+                if (chunk != null && (this.anyPlayerCloseEnoughForSpawningInternal(holder.getPos()) || ticketStorage.shouldForceNaturalSpawning(holder.getPos()))) {// Neo: Force natural spawning in other chunks
                     output.add(chunk);
                 }
             }
@@ -961,6 +_,19 @@
         return triState == TriState.DEFAULT ? this.anyPlayerCloseEnoughForSpawningInternal(pos) : triState.toBoolean(true);
     }
 
+    // TODO 1.21.11 unobfuscated - seems unused and requires an AT for ticketStorage
+//    void collectForceTickingChunks(List<LevelChunk> list) {
+//        for (long l : distanceManager.ticketStorage.getForceLoadedChunks()) {
+//            ChunkHolder holder = this.visibleChunkMap.get(l);
+//            if (holder != null) {
+//                var chunk = holder.getTickingChunk();
+//                if (chunk != null) {
+//                    list.add(chunk);
+//                }
+//            }
+//        }
+//    }
+
     public boolean anyPlayerCloseEnoughTo(BlockPos pos, int maxDistance) {
         Vec3 target = new Vec3(pos);
 
@@ -1090,6 +_,7 @@
                 this.playerMap.unIgnorePlayer(player);
             }
 
+            //PATCH 1.20.2: Figure out the firing of the watch and unwatch events when chunk tracking updates.
             this.updateChunkTracking(player);
         }
     }
@@ -1138,7 +_,7 @@
     }
 
     protected void addEntity(Entity entity) {
-        if (!(entity instanceof EnderDragonPart)) {
+        if (!(entity instanceof net.neoforged.neoforge.entity.PartEntity)) {
             EntityType<?> type = entity.getType();
             int range = type.clientTrackingRange() * 16;
             if (range != 0) {
@@ -1296,6 +_,20 @@
         }
     }
 
+    // Neo: Getter for players watching an entity.
+    public List<ServerPlayer> getPlayersWatching(Entity entity) {
+        var trackedEntity = entityMap.get(entity.getId());
+        if (trackedEntity != null) {
+            var ret = new java.util.ArrayList<ServerPlayer>(trackedEntity.seenBy.size());
+            for (var connection : trackedEntity.seenBy) {
+                ret.add(connection.getPlayer());
+            }
+            return java.util.Collections.unmodifiableList(ret);
+        } else {
+            return List.of();
+        }
+    }
+
     private class DistanceManager extends net.minecraft.server.level.DistanceManager {
         protected DistanceManager(TicketStorage ticketStorage, Executor executor, Executor mainThreadExecutor) {
             super(ticketStorage, executor, mainThreadExecutor);
@@ -1427,5 +_,20 @@
                 this.updatePlayer(player);
             }
         }
+    }
+
+    /**
+     * Neo: PR #937
+     * This is for mainly pre-generation usage such as Neoforge's generate command.
+     * Use this to schedule chunk load tasks into ChunkTaskPriorityQueueSorter so a chunk is fully finished all of their tasks before scheduling more chunks to load.
+     * Reason for this is when scheduling a huge ton of Full Status chunk tasks to the server (to load chunks),
+     * you could cause the server to only process those loading tasks and never reach the two chunk tasks that are
+     * automatically scheduled to run after the chunk is loaded to Full. As a result of flooding the system with Full Status chunk tasks,
+     * the queue for the two kind of successor chunk tasks will grow and become a memory leak of lambdas and chunk references.
+     * Use this method to schedule tasks for loading chunks in your whenCompleteAsync method call so the tasks gets processed properly over time and not leak.
+     * See {@link net.neoforged.neoforge.server.command.generation.GenerationTask#enqueueChunks} as an example usage of this method.
+     */
+    public void scheduleOnMainThreadMailbox(Runnable runnable) {
+        mainThreadExecutor.schedule(runnable);
     }
 }
