--- a/net/minecraft/server/level/GenerationChunkHolder.java
+++ b/net/minecraft/server/level/GenerationChunkHolder.java
@@ -32,6 +_,7 @@
     private final AtomicReference<@Nullable ChunkGenerationTask> task = new AtomicReference<>();
     private final AtomicInteger generationRefCount = new AtomicInteger();
     private volatile CompletableFuture<Void> generationSaveSyncFuture = CompletableFuture.completedFuture(null);
+    public net.minecraft.world.level.chunk.LevelChunk currentlyLoading; // Forge: Used to bypass future chain when loading chunks.
 
     public GenerationChunkHolder(ChunkPos pos) {
         this.pos = pos;
