--- /dev/null
+++ b/net/minecraft/server/level/ServerPlayer.java.rej
@@ -1,0 +_,34 @@
+diff a/projects/neoforge/src/main/java/net/minecraft/server/level/ServerPlayer.java b/projects/neoforge/src/main/java/net/minecraft/server/level/ServerPlayer.java	(rejected hunks)
+@@ -1135,12 +1135,28 @@ public class ServerPlayer extends Player {
+ 
+                 return OptionalInt.empty();
+             } else {
+-                if (extraDataWriter == null) {
++                // Neo: Support sending additional arbitrary data to menu factories on the client-side
++                var extraData = net.neoforged.neoforge.common.util.FriendlyByteBufUtil.writeCustomData(
++                        buffer -> {
++                            p_9033_.writeClientSideData(abstractcontainermenu, buffer);
++                            if (extraDataWriter != null) {
++                                extraDataWriter.accept(buffer);
++                            }
++                        },
++                        registryAccess()
++                );
++                if (extraData.length != 0) {
++                    this.connection.send(
++                        new net.neoforged.neoforge.network.payload.AdvancedOpenScreenPayload(
++                            abstractcontainermenu.containerId,
++                            abstractcontainermenu.getType(),
++                            p_9033_.getDisplayName(),
++                            extraData
++                        )
++                    );
++                } else {
+                 this.connection
+                     .send(new ClientboundOpenScreenPacket(abstractcontainermenu.containerId, abstractcontainermenu.getType(), p_9033_.getDisplayName()));
+-                } else {
+-                    this.connection
+-                        .send(new net.neoforged.neoforge.network.payload.AdvancedOpenScreenPayload(abstractcontainermenu.containerId, abstractcontainermenu.getType(), p_9033_.getDisplayName(), net.neoforged.neoforge.common.util.FriendlyByteBufUtil.writeCustomData(extraDataWriter, registryAccess())));
+                 }
+                 this.initMenu(abstractcontainermenu);
+                 this.containerMenu = abstractcontainermenu;
