--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -80,7 +_,10 @@
         this.previousGameModeForPlayer = previousGameModeForPlayer;
         this.gameModeForPlayer = gameModeForPlayer;
         Abilities abilities = this.player.getAbilities();
+        // Neo: preserve flying state, removed on tick if Attribute or ability no longer applies
+        boolean wasFlying = abilities.flying;
         gameModeForPlayer.updatePlayerAbilities(abilities);
+        abilities.flying = wasFlying || abilities.flying;
     }
 
     private boolean isInRangeOfGround() {
@@ -148,6 +_,10 @@
     }
 
     public void handleBlockBreakAction(BlockPos pos, ServerboundPlayerActionPacket.Action action, Direction direction, int maxY, int sequence) {
+        net.neoforged.neoforge.event.entity.player.PlayerInteractEvent.LeftClickBlock event = net.neoforged.neoforge.common.CommonHooks.onLeftClickBlock(player, pos, direction, action);
+        if (event.isCanceled()) {
+            return;
+        }
         if (!this.player.isWithinBlockInteractionRange(pos, 1.0)) {
             this.debugLogging(pos, false, sequence, "too far");
         } else if (pos.getY() > maxY) {
@@ -192,6 +_,7 @@
                         blockState,
                         item -> this.player.onEquippedItemBroken(item, EquipmentSlot.MAINHAND)
                     );
+                    if (event.getUseBlock() != net.minecraft.util.TriState.FALSE)
                     blockState.attack(this.level, pos, this.player);
                     progress = blockState.getDestroyProgress(this.player, this.player.level(), pos);
                 }
@@ -258,8 +_,10 @@
     }
 
     public boolean destroyBlock(BlockPos pos) {
+        // Neo: Fire the BlockBreakEvent, and ignore the original ItemStack#canDestroyBlock check since the break event manages the status of it.
         BlockState state = this.level.getBlockState(pos);
-        if (!this.player.getMainHandItem().canDestroyBlock(state, this.level, pos, this.player)) {
+        var event = net.neoforged.neoforge.common.CommonHooks.fireBlockBreak(level, gameModeForPlayer, player, pos, state);
+        if (event.isCanceled()) {
             return false;
         }
 
@@ -275,30 +_,46 @@
         }
 
         BlockState adjustedState = block.playerWillDestroy(this.level, pos, state, this.player);
-        boolean changed = this.level.removeBlock(pos, false);
-        if (SharedConstants.DEBUG_BLOCK_BREAK) {
-            LOGGER.info("server broke {} {} -> {}", pos, adjustedState, this.level.getBlockState(pos));
-        }
-
-        if (changed) {
-            block.destroy(this.level, pos, adjustedState);
-        }
 
         if (this.player.preventsBlockDrops()) {
+            removeBlock(pos, adjustedState, false, player.getMainHandItem().copy());
             return true;
         }
 
         ItemStack itemStack = this.player.getMainHandItem();
         ItemStack destroyedWith = itemStack.copy();
-        boolean canDestroy = this.player.hasCorrectToolForDrops(adjustedState);
+        boolean canDestroy = adjustedState.canHarvestBlock(this.level, pos, this.player); // previously player.hasCorrectToolForDrops(blockstate)
         itemStack.mineBlock(this.level, adjustedState, pos, this.player);
+        boolean changed = removeBlock(pos, adjustedState, canDestroy, destroyedWith.copy());
+
         if (changed && canDestroy) {
             block.playerDestroy(this.level, this.player, pos, adjustedState, blockEntity, destroyedWith);
         }
 
+        // Neo: Fire the PlayerDestroyItemEvent if the tool was broken at any point during the break process
+        if (itemStack.isEmpty() && !destroyedWith.isEmpty()) {
+            net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem(this.player, destroyedWith, InteractionHand.MAIN_HAND);
+        }
+
         return true;
     }
 
+    /**
+     * Neo: Patched-in method that handles actual removal of blocks for {@link #destroyBlock(BlockPos)}.
+     *
+     * @param pos The block pos of the destroyed block
+     * @param state The state of the destroyed block
+     * @param canHarvest If the player breaking the block can harvest the drops of the block
+     * @param toolStack The players main-hand prior to destroying the block and applying damage to the tool.
+     * @return If the block was removed, as reported by {@link BlockState#onDestroyedByPlayer}.
+     */
+    private boolean removeBlock(BlockPos pos, BlockState state, boolean canHarvest, ItemStack toolStack) {
+        boolean removed = state.onDestroyedByPlayer(this.level, pos, this.player, toolStack, canHarvest, this.level.getFluidState(pos));
+        if (removed)
+            state.getBlock().destroy(this.level, pos, state);
+        return removed;
+    }
+
     public InteractionResult useItem(ServerPlayer player, Level level, ItemStack itemStack, InteractionHand hand) {
         if (this.gameModeForPlayer == GameType.SPECTATOR) {
             return InteractionResult.PASS;
@@ -308,6 +_,9 @@
             return InteractionResult.PASS;
         }
 
+        InteractionResult cancelResult = net.neoforged.neoforge.common.CommonHooks.onItemRightClick(player, hand);
+        if (cancelResult != null) return cancelResult;
+
         int oldCount = itemStack.getCount();
         int oldDamage = itemStack.getDamageValue();
         InteractionResult result = itemStack.use(level, player, hand);
@@ -351,6 +_,8 @@
             return InteractionResult.FAIL;
         }
 
+        net.neoforged.neoforge.event.entity.player.PlayerInteractEvent.RightClickBlock event = net.neoforged.neoforge.common.CommonHooks.onRightClickBlock(player, hand, pos, hitResult);
+        if (event.isCanceled()) return event.getCancellationResult();
         if (this.gameModeForPlayer == GameType.SPECTATOR) {
             MenuProvider menuProvider = state.getMenuProvider(level, pos);
             if (menuProvider != null) {
@@ -360,10 +_,15 @@
                 return InteractionResult.PASS;
             }
         } else {
+            UseOnContext context = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem() != net.minecraft.util.TriState.FALSE) {
+                InteractionResult result = itemStack.onItemUseFirst(context);
+                if (result != InteractionResult.PASS) return result;
+            }
             boolean haveSomethingInOurHands = !player.getMainHandItem().isEmpty() || !player.getOffhandItem().isEmpty();
-            boolean suppressUsingBlock = player.isSecondaryUseActive() && haveSomethingInOurHands;
+            boolean suppressUsingBlock = (player.isSecondaryUseActive() && haveSomethingInOurHands) && !(player.getMainHandItem().doesSneakBypassUse(level, pos, player) && player.getOffhandItem().doesSneakBypassUse(level, pos, player));
             ItemStack usedItemStack = itemStack.copy();
-            if (!suppressUsingBlock) {
+            if (event.getUseBlock().isTrue() || (event.getUseBlock().isDefault() && !suppressUsingBlock)) {
                 InteractionResult itemUse = state.useItemOn(player.getItemInHand(hand), level, player, hand, hitResult);
                 if (itemUse.consumesAction()) {
                     CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger(player, pos, usedItemStack);
@@ -379,8 +_,8 @@
                 }
             }
 
-            if (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack)) {
-                UseOnContext context = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem().isTrue() || (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack))) {
+                if (event.getUseItem().isFalse()) return InteractionResult.PASS;
                 InteractionResult success;
                 if (player.hasInfiniteMaterials()) {
                     int count = itemStack.getCount();
