--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -99,7 +_,7 @@
             LOGGER.error("Failed to create dimension data storage directory", e);
         }
 
-        this.savedDataStorage = new SavedDataStorage(dataFolder, fixerUpper, level.registryAccess());
+        this.savedDataStorage = new SavedDataStorage(level, dataFolder, fixerUpper, level.registryAccess());
         this.ticketStorage = this.savedDataStorage.computeIfAbsent(TicketStorage.TYPE);
         this.chunkMap = new ChunkMap(
             level,
@@ -160,6 +_,12 @@
             }
         }
 
+        ChunkHolder chunkholder = this.getVisibleChunkIfPresent(pos);
+        if (chunkholder != null && chunkholder.currentlyLoading != null) {
+            // Neo: If the requested chunk is loading, bypass the future chain to prevent a deadlock.
+            return chunkholder.currentlyLoading;
+        }
+
         profiler.incrementCounter("getChunkCacheMiss");
         CompletableFuture<ChunkResult<ChunkAccess>> serverFuture = this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate);
         this.mainThreadProcessor.managedBlock(serverFuture::isDone);
@@ -191,6 +_,11 @@
         ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(pos);
         if (chunkHolder == null) {
             return null;
+        }
+
+        if (chunkHolder.currentlyLoading != null) {
+            // Neo: If the requested chunk is loading, bypass the future chain to prevent a deadlock.
+            return chunkHolder.currentlyLoading;
         }
 
         ChunkAccess chunk = chunkHolder.getChunkIfPresent(ChunkStatus.FULL);
