--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -3,7 +_,11 @@
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.BuiltInRegistries;
 
-public record TicketType(long timeout, int flags) {
+public record TicketType(long timeout, int flags, boolean forceNaturalSpawning) {
+    public TicketType(long timeout, int flags) {
+        this(timeout, flags, false);
+    }
+
     public static final long NO_TIMEOUT = 0L;
     public static final int FLAG_PERSIST = 1;
     public static final int FLAG_LOADING = 2;
