--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -85,4 +_,26 @@
         category.setDetail("Derived", true);
         this.wrapped.fillCrashReportCategory(category, levelHeightAccessor);
     }
+
+    // Neo: Variable day time code
+
+    @Override
+    public float getDayTimeFraction() {
+        return wrapped.getDayTimeFraction();
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return wrapped.getDayTimePerTick();
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        wrapped.setDayTimeFraction(dayTimeFraction);
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        wrapped.setDayTimePerTick(dayTimePerTick);
+    }
 }
