--- a/net/minecraft/world/level/storage/DimensionDataStorage.java
+++ b/net/minecraft/world/level/storage/DimensionDataStorage.java
@@ -73,7 +_,7 @@
     }
 
     @Nullable
-    private <T extends SavedData> T readSavedData(BiFunction<CompoundTag, HolderLookup.Provider, T> p_324609_, DataFixTypes p_295832_, String p_164870_) {
+    private <T extends SavedData> T readSavedData(BiFunction<CompoundTag, HolderLookup.Provider, T> p_324609_, @Nullable DataFixTypes p_295832_, String p_164870_) {
         try {
             Path path = this.getDataFile(p_164870_);
             if (Files.exists(path)) {
@@ -92,7 +_,7 @@
         p_164857_.setDirty();
     }
 
-    public CompoundTag readTagFromDisk(String p_78159_, DataFixTypes p_295038_, int p_78160_) throws IOException {
+    public CompoundTag readTagFromDisk(String p_78159_, @Nullable DataFixTypes p_295038_, int p_78160_) throws IOException {
         CompoundTag compoundtag1;
         try (
             InputStream inputstream = Files.newInputStream(this.getDataFile(p_78159_));
@@ -107,9 +_,16 @@
                 }
             }
 
-            int i = NbtUtils.getDataVersion(compoundtag, 1343);
-            compoundtag1 = p_295038_.update(this.fixerUpper, compoundtag, i, p_78160_);
+            if (p_295038_ != null) {
+                int i = NbtUtils.getDataVersion(compoundtag, 1343);
+                compoundtag1 = p_295038_.update(this.fixerUpper, compoundtag, i, p_78160_);
+            } else {
+                compoundtag1 = compoundtag;
+            }
         }
+
+        // Neo: delete any temporary files so that we don't inflate disk space unnecessarily.
+        net.neoforged.neoforge.common.IOUtilities.cleanupTempFiles(this.dataFolder, p_78159_);
 
         return compoundtag1;
     }
@@ -182,7 +_,11 @@
 
     private static void tryWrite(Path p_390658_, CompoundTag p_390657_) {
         try {
-            NbtIo.writeCompressed(p_390657_, p_390658_);
+            // Neo: ensure parent directories exist if the SavedData's path contains slashes
+            if (!Files.exists(p_390658_)) {
+                Files.createDirectories(p_390658_.getParent());
+            }
+            net.neoforged.neoforge.common.IOUtilities.writeNbtCompressed(p_390657_, p_390658_);
         } catch (IOException ioexception) {
             LOGGER.error("Could not save data to {}", p_390658_.getFileName(), ioexception);
         }
