--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -39,6 +_,7 @@
             Path path2 = path.resolve(player.getStringUUID() + ".dat");
             Path path3 = path.resolve(player.getStringUUID() + ".dat_old");
             Util.safeReplaceFile(path2, path1, path3);
+            net.neoforged.neoforge.event.EventHooks.firePlayerSavingEvent(player, playerDir, player.getStringUUID());
         } catch (Exception exception) {
             LOGGER.warn("Failed to save player data for {}", player.getName().getString());
         }
@@ -80,7 +_,12 @@
             int i = NbtUtils.getDataVersion(p_316252_, -1);
             p_316252_ = DataFixTypes.PLAYER.updateToCurrentVersion(this.fixerUpper, p_316252_, i);
             player.load(p_316252_);
+            net.neoforged.neoforge.event.EventHooks.firePlayerLoadingEvent(player, playerDir, player.getStringUUID());
             return p_316252_;
         });
+    }
+
+    public File getPlayerDir() {
+        return playerDir;
     }
 }
