--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -320,10 +_,58 @@
         return this.settings.copy();
     }
 
+    public boolean hasConfirmedExperimentalWarning() {
+        return this.confirmedExperimentalWarning;
+    }
+
+    public PrimaryLevelData withConfirmedWarning(boolean confirmedWarning) { // Builder-like to not patch ctor
+        this.confirmedExperimentalWarning = confirmedWarning;
+        return this;
+    }
+
     @Deprecated
     public static enum SpecialWorldProperty {
         NONE,
         FLAT,
         DEBUG;
+    }
+
+    // Neo: Variable day time code
+
+    private float dayTimeFraction = 0.0f;
+    private float dayTimePerTick = -1.0f;
+
+    @Override
+    public float getDayTimeFraction() {
+        return dayTimeFraction;
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return dayTimePerTick;
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        this.dayTimeFraction = dayTimeFraction;
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        this.dayTimePerTick = dayTimePerTick;
+    }
+
+    private static java.util.stream.Stream<String> updateRemovedFeatureFlags(java.util.stream.Stream<String> removedFeatures, java.util.stream.Stream<String> enabledFeatures) {
+        var unknownFeatureFlags = new HashSet<net.minecraft.resources.Identifier>();
+        // parses the incoming Stream<String> and spits out unknown flag names (Identifier)
+        // we do not care about the returned FeatureFlagSet, only the flags which do not exist
+        net.minecraft.world.flag.FeatureFlags.REGISTRY.fromNames(enabledFeatures.map(net.minecraft.resources.Identifier::parse).collect(Collectors.toSet()), unknownFeatureFlags::add);
+        // concat the received removed flags with our new additions
+        return java.util.stream.Stream.concat(removedFeatures, unknownFeatureFlags.stream()
+                // we only want modded flags, mojang has datafixers for vanilla flags
+                .filter(java.util.function.Predicate.not(name -> name.getNamespace().equals(net.minecraft.resources.Identifier.DEFAULT_NAMESPACE)))
+                .map(net.minecraft.resources.Identifier::toString))
+                // no duplicates should exist in this stream
+                .distinct();
     }
 }
