--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -42,6 +_,7 @@
     private final Set<String> knownServerBrands;
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
+    private boolean confirmedExperimentalWarning = false;
 
     private PrimaryLevelData(
         @Nullable UUID singlePlayerUUID,
@@ -90,7 +_,7 @@
     ) {
         long gameTime = input.get("Time").asLong(0L);
         LevelVersion levelVersion = LevelVersion.parse(input);
-        return new PrimaryLevelData(
+        var result = new PrimaryLevelData(
             input.get("singleplayer_uuid").flatMap(UUIDUtil.CODEC::parse).result().orElse(null),
             input.get("WasModded").asBoolean(false),
             input.get("spawn").read(LevelData.RespawnData.CODEC).result().orElse(LevelData.RespawnData.DEFAULT),
@@ -98,11 +_,16 @@
             levelVersion.levelDataVersion(),
             input.get("initialized").asBoolean(true),
             input.get("ServerBrands").asStream().flatMap(b -> b.asString().result().stream()).collect(Collectors.toCollection(Sets::newLinkedHashSet)),
-            input.get("removed_features").asStream().flatMap(b -> b.asString().result().stream()).collect(Collectors.toSet()),
+            // Neo: Append removed modded feature flags
+            updateRemovedFeatureFlags(input.get("removed_features").asStream().flatMap(b -> b.asString().result().stream()), input.get("enabled_features").asStream().flatMap(features -> features.asString().result().stream())).collect(Collectors.toSet()),
             settings,
             specialWorldProperty,
             worldGenSettingsLifecycle
-        );
+        ).withConfirmedWarning(worldGenSettingsLifecycle != Lifecycle.stable() && input.get("confirmedExperimentalSettings").asBoolean(false));
+        // Neo:
+        result.setDayTimeFraction(input.get("neoDayTimeFraction").asFloat(0f));
+        result.setDayTimePerTick(input.get("neoDayTimePerTick").asFloat(-1f));
+        return result;
     }
 
     @Override
@@ -139,6 +_,12 @@
         }
 
         tag.store(WorldDataConfiguration.MAP_CODEC, this.settings.dataConfiguration());
+
+        // Neo:
+        tag.putString("forgeLifecycle", net.neoforged.neoforge.common.CommonHooks.encodeLifecycle(this.settings.getLifecycle()));
+        tag.putBoolean("confirmedExperimentalSettings", this.confirmedExperimentalWarning);
+        tag.putFloat("neoDayTimeFraction", dayTimeFraction);
+        tag.putFloat("neoDayTimePerTick", dayTimePerTick);
     }
 
     public static void writeLastPlayed(CompoundTag tag) {
@@ -325,10 +_,58 @@
         return this.settings.copy();
     }
 
+    public boolean hasConfirmedExperimentalWarning() {
+        return this.confirmedExperimentalWarning;
+    }
+
+    public PrimaryLevelData withConfirmedWarning(boolean confirmedWarning) { // Builder-like to not patch ctor
+        this.confirmedExperimentalWarning = confirmedWarning;
+        return this;
+    }
+
     @Deprecated
     public enum SpecialWorldProperty {
         NONE,
         FLAT,
         DEBUG;
+    }
+
+    // Neo: Variable day time code
+
+    private float dayTimeFraction = 0.0f;
+    private float dayTimePerTick = -1.0f;
+
+    @Override
+    public float getDayTimeFraction() {
+        return dayTimeFraction;
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return dayTimePerTick;
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        this.dayTimeFraction = dayTimeFraction;
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        this.dayTimePerTick = dayTimePerTick;
+    }
+
+    private static java.util.stream.Stream<String> updateRemovedFeatureFlags(java.util.stream.Stream<String> removedFeatures, java.util.stream.Stream<String> enabledFeatures) {
+        var unknownFeatureFlags = new HashSet<net.minecraft.resources.Identifier>();
+        // parses the incoming Stream<String> and spits out unknown flag names (Identifier)
+        // we do not care about the returned FeatureFlagSet, only the flags which do not exist
+        net.minecraft.world.flag.FeatureFlags.REGISTRY.fromNames(enabledFeatures.map(net.minecraft.resources.Identifier::parse).collect(Collectors.toSet()), unknownFeatureFlags::add);
+        // concat the received removed flags with our new additions
+        return java.util.stream.Stream.concat(removedFeatures, unknownFeatureFlags.stream()
+                // we only want modded flags, mojang has datafixers for vanilla flags
+                .filter(java.util.function.Predicate.not(name -> name.getNamespace().equals(net.minecraft.resources.Identifier.DEFAULT_NAMESPACE)))
+                .map(net.minecraft.resources.Identifier::toString))
+                // no duplicates should exist in this stream
+                .distinct();
     }
 }
