--- a/net/minecraft/world/level/storage/SavedDataStorage.java
+++ b/net/minecraft/world/level/storage/SavedDataStorage.java
@@ -46,8 +_,17 @@
     private final Path dataFolder;
     private CompletableFuture<?> pendingWriteFuture = CompletableFuture.completedFuture(null);
     private boolean closed;
+    // Neo: Needed to make level context available to modded saved data
+    private final net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerLevel level;
 
+    /** @deprecated Neo: Use the version that has a level */
+    @Deprecated
     public SavedDataStorage(Path dataFolder, DataFixer fixerUpper, HolderLookup.Provider registries) {
+        this(null, dataFolder, fixerUpper, registries);
+    }
+
+    public SavedDataStorage(net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerLevel level, Path dataFolder, DataFixer fixerUpper, HolderLookup.Provider registries) {
+        this.level = level;
         this.fixerUpper = fixerUpper;
         this.dataFolder = dataFolder;
         this.registries = registries;
@@ -68,7 +_,7 @@
             return data;
         }
 
-        T newData = (T)type.constructor().get();
+        T newData = (T)type.factory().create(this.level);
         this.set(type, newData);
         return newData;
     }
@@ -89,7 +_,7 @@
             if (Files.exists(file)) {
                 CompoundTag tag = this.readTagFromDisk(file, type.dataFixType(), SharedConstants.getCurrentVersion().dataVersion().version());
                 RegistryOps<Tag> ops = this.registries.createSerializationContext(NbtOps.INSTANCE);
-                return type.codec()
+                return type.codecFactory().create(this.level)
                     .parse(ops, tag.get("data"))
                     .resultOrPartial(error -> LOGGER.error("Failed to parse saved data for '{}': {}", type, error))
                     .orElse(null);
@@ -106,7 +_,7 @@
         data.setDirty();
     }
 
-    public CompoundTag readTagFromDisk(Path dataFile, DataFixTypes type, int newVersion) throws IOException {
+    public CompoundTag readTagFromDisk(Path dataFile, @Nullable DataFixTypes type, int newVersion) throws IOException {
         try (
             InputStream in = Files.newInputStream(dataFile);
             PushbackInputStream inputStream = new PushbackInputStream(new FastBufferedInputStream(in), 2);
@@ -121,7 +_,15 @@
             }
 
             int version = NbtUtils.getDataVersion(tag, 1343);
+            if (type != null)
             return type.update(this.fixerUpper, tag, version, newVersion);
+            else return tag;
+        } finally {
+            // Neo: delete any temporary files so that we don't inflate disk space unnecessarily.
+            this.pendingWriteFuture = this.pendingWriteFuture.thenRunAsync(
+                () -> net.neoforged.neoforge.common.IOUtilities.tryCleanupTempFiles(this.dataFolder, dataFile.getFileName().toString()),
+                Util.ioPool()
+            );
         }
     }
 
@@ -196,7 +_,7 @@
     }
 
     private <T extends SavedData> CompoundTag encodeUnchecked(SavedDataType<T> type, SavedData data, RegistryOps<Tag> ops) {
-        Codec<T> codec = type.codec();
+        Codec<T> codec = type.codecFactory().create(this.level);
         CompoundTag tag = new CompoundTag();
         tag.put("data", codec.encodeStart(ops, (T)data).getOrThrow());
         NbtUtils.addCurrentDataVersion(tag);
@@ -208,7 +_,11 @@
 
         try {
             FileUtil.createDirectoriesSafe(path.getParent());
-            NbtIo.writeCompressed(tag, path);
+            // Neo: ensure parent directories exist if the SavedData's path contains slashes
+            if (!Files.exists(path)) {
+                Files.createDirectories(path.getParent());
+            }
+            net.neoforged.neoforge.common.IOUtilities.writeNbtCompressed(tag, path);
         } catch (IOException e) {
             LOGGER.error("Could not save data to {}", path.getFileName(), e);
         }
