--- a/net/minecraft/world/level/storage/ServerLevelData.java
+++ b/net/minecraft/world/level/storage/ServerLevelData.java
@@ -87,4 +_,10 @@
     void setGameTime(long p_78617_);
 
     void setDayTime(long p_78624_);
+
+    //Neo
+    float getDayTimeFraction();
+    float getDayTimePerTick();
+    void setDayTimeFraction(float dayTimeFraction);
+    void setDayTimePerTick(float dayTimePerTick);
 }
