--- a/net/minecraft/world/level/storage/ServerLevelData.java
+++ b/net/minecraft/world/level/storage/ServerLevelData.java
@@ -38,4 +_,11 @@
     void setGameType(GameType gameType);
 
     void setGameTime(final long time);
+
+    //Neo
+    // TODO 26.1: I believe these may be outdated and replaced by the clock managers persistence
+    float getDayTimeFraction();
+    float getDayTimePerTick();
+    void setDayTimeFraction(float dayTimeFraction);
+    void setDayTimePerTick(float dayTimePerTick);
 }
