--- a/net/minecraft/client/renderer/chunk/RenderSectionRegion.java
+++ b/net/minecraft/client/renderer/chunk/RenderSectionRegion.java
@@ -90,6 +_,21 @@
         return this.level.getHeight();
     }
 
+    @Override
+    public float getShade(float normalX, float normalY, float normalZ, boolean shade) {
+        return this.level.getShade(normalX, normalY, normalZ, shade);
+    }
+
+    @Override
+    public net.neoforged.neoforge.model.data.ModelData getModelData(BlockPos pos) {
+        return modelDataSnapshot.get(pos.asLong());
+    }
+
+    @Override
+    public net.neoforged.neoforge.common.world.AuxiliaryLightManager getAuxLightManager(net.minecraft.world.level.ChunkPos pos) {
+        return this.getSection(pos.x(), this.minSectionY, pos.z()).wrapped.getAuxLightManager(pos);
+    }
+
     public static int index(int minSectionX, int minSectionY, int minSectionZ, int sectionX, int sectionY, int sectionZ) {
         return sectionX - minSectionX + (sectionY - minSectionY) * 3 + (sectionZ - minSectionZ) * 3 * 3;
     }
