--- a/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
@@ -22,12 +_,18 @@
     private final int minChunkZ;
     protected final RenderChunk[] chunks;
     protected final Level level;
+    private final it.unimi.dsi.fastutil.longs.Long2ObjectFunction<net.neoforged.neoforge.client.model.data.ModelData> modelDataSnapshot;
 
+    @Deprecated
     RenderChunkRegion(Level p_200456_, int p_200457_, int p_200458_, RenderChunk[] p_350680_) {
+        this(p_200456_, p_200457_, p_200458_, p_350680_, net.neoforged.neoforge.client.model.data.ModelDataManager.EMPTY_SNAPSHOT);
+    }
+    RenderChunkRegion(Level p_200456_, int p_200457_, int p_200458_, RenderChunk[] p_350680_, it.unimi.dsi.fastutil.longs.Long2ObjectFunction<net.neoforged.neoforge.client.model.data.ModelData> modelDataSnapshot) {
         this.level = p_200456_;
         this.minChunkX = p_200457_;
         this.minChunkZ = p_200458_;
         this.chunks = p_350680_;
+        this.modelDataSnapshot = modelDataSnapshot;
     }
 
     @Override
@@ -75,6 +_,21 @@
     @Override
     public int getHeight() {
         return this.level.getHeight();
+    }
+
+    @Override
+    public float getShade(float normalX, float normalY, float normalZ, boolean shade) {
+        return this.level.getShade(normalX, normalY, normalZ, shade);
+    }
+
+    @Override
+    public net.neoforged.neoforge.client.model.data.ModelData getModelData(BlockPos pos) {
+        return modelDataSnapshot.get(pos.asLong());
+    }
+
+    @Override
+    public net.neoforged.neoforge.common.world.AuxiliaryLightManager getAuxLightManager(net.minecraft.world.level.ChunkPos pos) {
+        return this.getChunk(pos.x, pos.z).wrapped.getAuxLightManager(pos);
     }
 
     public static int index(int p_350911_, int p_350842_, int p_350891_, int p_350833_) {
