--- a/net/minecraft/client/renderer/chunk/RenderRegionCache.java
+++ b/net/minecraft/client/renderer/chunk/RenderRegionCache.java
@@ -34,7 +_,8 @@
             }
         }
 
-        return new RenderSectionRegion(level, minSectionX, minSectionY, minSectionZ, regionSections);
+        var modelDataManager = level.getModelDataManager().snapshotSectionRegion(minSectionX, minSectionY, minSectionZ, maxSectionX, maxSectionY, maxSectionZ);
+        return new RenderSectionRegion(level, minSectionX, minSectionY, minSectionZ, regionSections, modelDataManager);
     }
 
     private SectionCopy getSectionDataCopy(Level level, int sectionX, int sectionY, int sectionZ) {
