--- a/net/minecraft/client/renderer/chunk/SectionCompiler.java
+++ b/net/minecraft/client/renderer/chunk/SectionCompiler.java
@@ -49,7 +_,13 @@
         this.blockColors = blockColors;
     }
 
+    /// @deprecated Neo: use [#compile(SectionPos, RenderSectionRegion, VertexSorting, SectionBufferBuilderPack, java.util.List)] instead
+    @Deprecated
     public SectionCompiler.Results compile(SectionPos sectionPos, RenderSectionRegion region, VertexSorting vertexSorting, SectionBufferBuilderPack builders) {
+        return compile(sectionPos, region, vertexSorting, builders, List.of());
+    }
+
+    public SectionCompiler.Results compile(SectionPos sectionPos, RenderSectionRegion region, VertexSorting vertexSorting, SectionBufferBuilderPack builders, List<net.neoforged.neoforge.client.event.AddSectionGeometryEvent.AdditionalSectionRenderer> additionalRenderers) {
         SectionCompiler.Results results = new SectionCompiler.Results();
         BlockPos minPos = sectionPos.origin();
         BlockPos maxPos = minPos.offset(15, 15, 15);
@@ -85,6 +_,8 @@
 
                     FluidState fluidState = blockState.getFluidState();
                     if (!fluidState.isEmpty()) {
+                        var customRenderer = this.fluidModelSet.get(fluidState).customRenderer();
+                        if (customRenderer == null || !customRenderer.renderFluid(fluidRenderer, fluidState, region, pos, fluidOutput, blockState))
                         fluidRenderer.tesselate(region, pos, fluidOutput, blockState, fluidState);
                     }
 
@@ -110,6 +_,7 @@
             }
         }
 
+        net.neoforged.neoforge.client.ClientHooks.addAdditionalGeometry(additionalRenderers, layer -> this.getOrBeginLayer(startedLayers, builders, layer), region, blockRenderer);
         for (Entry<ChunkSectionLayer, BufferBuilder> entry : startedLayers.entrySet()) {
             ChunkSectionLayer layer = entry.getKey();
             MeshData mesh = entry.getValue().build();
