--- a/net/minecraft/client/renderer/chunk/SectionCopy.java
+++ b/net/minecraft/client/renderer/chunk/SectionCopy.java
@@ -25,8 +_,10 @@
     private final @Nullable PalettedContainer<BlockState> section;
     private final boolean debug;
     private final LevelHeightAccessor levelHeightAccessor;
+    final LevelChunk wrapped;
 
     SectionCopy(LevelChunk levelChunk, int sectionIndex) {
+        this.wrapped = levelChunk;
         this.levelHeightAccessor = levelChunk;
         this.debug = levelChunk.getLevel().isDebug();
         this.blockEntities = ImmutableMap.copyOf(levelChunk.getBlockEntities());
