--- a/net/minecraft/client/renderer/chunk/SectionRenderDispatcher.java
+++ b/net/minecraft/client/renderer/chunk/SectionRenderDispatcher.java
@@ -377,9 +_,10 @@
 
         public SectionRenderDispatcher.RenderSection.CompileTask createCompileTask(RenderRegionCache cache) {
             this.cancelTasks();
+            var additionalRenderers = net.neoforged.neoforge.client.ClientHooks.gatherAdditionalRenderers(this.renderOrigin, SectionRenderDispatcher.this.level);
             RenderSectionRegion region = cache.createRegion(SectionRenderDispatcher.this.level, this.sectionNode);
             boolean isRecompile = this.sectionMesh.get() != CompiledSectionMesh.UNCOMPILED;
-            this.lastRebuildTask = new SectionRenderDispatcher.RenderSection.RebuildTask(region, isRecompile);
+            this.lastRebuildTask = new SectionRenderDispatcher.RenderSection.RebuildTask(region, isRecompile, additionalRenderers);
             return this.lastRebuildTask;
         }
 
@@ -492,6 +_,15 @@
             return success;
         }
 
+        // Neo: start
+
+        @Override
+        public boolean isEmpty() {
+            return !getSectionMesh().hasRenderableLayers();
+        }
+
+        // Neo: end
+
         @OnlyIn(Dist.CLIENT)
         public abstract class CompileTask {
             protected final AtomicBoolean isCancelled;
@@ -530,11 +_,19 @@
         @OnlyIn(Dist.CLIENT)
         private class RebuildTask extends SectionRenderDispatcher.RenderSection.CompileTask {
             protected final RenderSectionRegion region;
+            private final java.util.List<net.neoforged.neoforge.client.event.AddSectionGeometryEvent.AdditionalSectionRenderer> additionalRenderers;
 
+            /** @deprecated Neo: use {@link #RebuildTask(RenderSectionRegion, boolean, java.util.List)} instead */
+            @Deprecated
             public RebuildTask(RenderSectionRegion region, boolean isRecompile) {
+                this(region, isRecompile, java.util.List.of());
+            }
+
+            public RebuildTask(RenderSectionRegion region, boolean isRecompile, java.util.List<net.neoforged.neoforge.client.event.AddSectionGeometryEvent.AdditionalSectionRenderer> additionalRenderers) {
                 Objects.requireNonNull(RenderSection.this);
                 super(isRecompile);
                 this.region = region;
+                this.additionalRenderers = additionalRenderers;
             }
 
             @Override
@@ -557,7 +_,7 @@
                         SectionCompiler.Results results;
                         try (Zone ignored = Profiler.get().zone("Compile Section")) {
                             results = SectionRenderDispatcher.this.sectionCompiler
-                                .compile(sectionPos, this.region, RenderSection.this.createVertexSorting(sectionPos, cameraPos), buffers);
+                                .compile(sectionPos, this.region, RenderSection.this.createVertexSorting(sectionPos, cameraPos), buffers, this.additionalRenderers);
                         }
 
                         if (results.renderedLayers.isEmpty()) {
