--- a/net/minecraft/client/renderer/chunk/SectionCompiler.java
+++ b/net/minecraft/client/renderer/chunk/SectionCompiler.java
@@ -43,6 +_,10 @@
     }
 
     public SectionCompiler.Results compile(SectionPos sectionPos, RenderSectionRegion region, VertexSorting vertexSorting, SectionBufferBuilderPack builders) {
+        return compile(sectionPos, region, vertexSorting, builders, List.of());
+    }
+
+    public SectionCompiler.Results compile(SectionPos sectionPos, RenderSectionRegion region, VertexSorting vertexSorting, SectionBufferBuilderPack builders, List<net.neoforged.neoforge.client.event.AddSectionGeometryEvent.AdditionalSectionRenderer> additionalRenderers) {
         SectionCompiler.Results results = new SectionCompiler.Results();
         BlockPos minPos = sectionPos.origin();
         BlockPos maxPos = minPos.offset(15, 15, 15);
@@ -82,19 +_,19 @@
             }
 
             if (blockState.getRenderShape() == RenderShape.MODEL) {
-                boolean forceOpaque = ItemBlockRenderTypes.forceOpaque(blockState);
                 random.setSeed(blockState.getSeed(pos));
-                this.blockRenderer.getBlockModel(blockState).collectParts(random, parts);
+                this.blockRenderer.getBlockModel(blockState).collectParts(region, pos, blockState, random, parts);
                 poseStack.pushPose();
                 poseStack.translate(
                     (float)SectionPos.sectionRelative(pos.getX()), (float)SectionPos.sectionRelative(pos.getY()), (float)SectionPos.sectionRelative(pos.getZ())
                 );
-                this.blockRenderer.renderBatched(blockState, pos, region, poseStack, forceOpaque ? opaqueQuadOutput : quadOutput, true, parts);
+                this.blockRenderer.renderBatched(blockState, pos, region, poseStack, bufferLookup, true, parts);
                 poseStack.popPose();
                 parts.clear();
             }
         }
 
+        net.neoforged.neoforge.client.ClientHooks.addAdditionalGeometry(additionalRenderers, bufferLookup, region, poseStack);
         for (Entry<ChunkSectionLayer, BufferBuilder> entry : startedLayers.entrySet()) {
             ChunkSectionLayer layer = entry.getKey();
             MeshData mesh = entry.getValue().build();
