--- a/net/minecraft/client/renderer/chunk/SectionCompiler.java
+++ b/net/minecraft/client/renderer/chunk/SectionCompiler.java
@@ -43,6 +_,10 @@
     }
 
     public SectionCompiler.Results compile(SectionPos p_350790_, RenderChunkRegion p_350782_, VertexSorting p_350635_, SectionBufferBuilderPack p_350612_) {
+        return compile(p_350790_, p_350782_, p_350635_, p_350612_, List.of());
+    }
+
+    public SectionCompiler.Results compile(SectionPos p_350790_, RenderChunkRegion p_350782_, VertexSorting p_350635_, SectionBufferBuilderPack p_350612_, List<net.neoforged.neoforge.client.event.AddSectionGeometryEvent.AdditionalSectionRenderer> additionalRenderers) {
         SectionCompiler.Results sectioncompiler$results = new SectionCompiler.Results();
         BlockPos blockpos = p_350790_.origin();
         BlockPos blockpos1 = blockpos.offset(15, 15, 15);
@@ -52,6 +_,7 @@
         Map<RenderType, BufferBuilder> map = new Reference2ObjectArrayMap<>(RenderType.chunkBufferLayers().size());
         RandomSource randomsource = RandomSource.create();
         List<BlockModelPart> list = new ObjectArrayList<>();
+        java.util.function.Function<RenderType, com.mojang.blaze3d.vertex.VertexConsumer> bufferLookup = renderType -> this.getOrBeginLayer(map, p_350612_, renderType);
 
         for (BlockPos blockpos2 : BlockPos.betweenClosed(blockpos, blockpos1)) {
             BlockState blockstate = p_350782_.getBlockState(blockpos2);
@@ -74,22 +_,25 @@
             }
 
             if (blockstate.getRenderShape() == RenderShape.MODEL) {
-                RenderType rendertype2 = ItemBlockRenderTypes.getChunkRenderType(blockstate);
-                BufferBuilder bufferbuilder1 = this.getOrBeginLayer(map, p_350612_, rendertype2);
                 randomsource.setSeed(blockstate.getSeed(blockpos2));
-                this.blockRenderer.getBlockModel(blockstate).collectParts(randomsource, list);
+                this.blockRenderer.getBlockModel(blockstate).collectParts(p_350782_, blockpos2, blockstate, randomsource, list);
                 posestack.pushPose();
                 posestack.translate(
                     (float)SectionPos.sectionRelative(blockpos2.getX()),
                     (float)SectionPos.sectionRelative(blockpos2.getY()),
                     (float)SectionPos.sectionRelative(blockpos2.getZ())
                 );
-                this.blockRenderer.renderBatched(blockstate, blockpos2, p_350782_, posestack, bufferbuilder1, true, list);
+                this.blockRenderer.renderBatched(blockstate, blockpos2, p_350782_, posestack, bufferLookup, true, list);
                 posestack.popPose();
                 list.clear();
             }
         }
-
+        net.neoforged.neoforge.client.ClientHooks.addAdditionalGeometry(
+                additionalRenderers,
+                type -> this.getOrBeginLayer(map, p_350612_, type),
+                p_350782_,
+                posestack
+        );
         for (Entry<RenderType, BufferBuilder> entry : map.entrySet()) {
             RenderType rendertype1 = entry.getKey();
             MeshData meshdata = entry.getValue().build();
@@ -121,9 +_,10 @@
     private <E extends BlockEntity> void handleBlockEntity(SectionCompiler.Results p_350754_, E p_350386_) {
         BlockEntityRenderer<E> blockentityrenderer = this.blockEntityRenderer.getRenderer(p_350386_);
         if (blockentityrenderer != null) {
-            p_350754_.blockEntities.add(p_350386_);
             if (blockentityrenderer.shouldRenderOffScreen(p_350386_)) {
                 p_350754_.globalBlockEntities.add(p_350386_);
+            } else {
+                p_350754_.blockEntities.add(p_350386_); // Neo: Fix MC-112730
             }
         }
     }
