--- a/net/minecraft/world/level/biome/BiomeSpecialEffects.java
+++ b/net/minecraft/world/level/biome/BiomeSpecialEffects.java
@@ -71,6 +_,8 @@
         }
     }
 
+    @net.neoforged.fml.common.asm.enumextension.NamedEnum
+    @net.neoforged.fml.common.asm.enumextension.NetworkedEnum(net.neoforged.fml.common.asm.enumextension.NetworkedEnum.NetworkCheck.CLIENTBOUND)
     public static enum GrassColorModifier implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
         NONE("none") {
             @Override
@@ -93,12 +_,22 @@
         };
 
         private final String name;
+        private final ColorModifier delegate;
         public static final Codec<BiomeSpecialEffects.GrassColorModifier> CODEC = StringRepresentable.fromEnum(BiomeSpecialEffects.GrassColorModifier::values);
 
-        public abstract int modifyColor(final double x, final double z, final int baseColor);
+        public int modifyColor(double x, double z, int baseColor) {
+            return delegate.modifyGrassColor(x, z, baseColor);
+        }
 
+        @net.neoforged.fml.common.asm.enumextension.ReservedConstructor
         private GrassColorModifier(String name) {
             this.name = name;
+            this.delegate = null;
+        }
+
+        GrassColorModifier(String name, ColorModifier delegate) {
+            this.name = name;
+            this.delegate = delegate;
         }
 
         public String getName() {
@@ -108,6 +_,15 @@
         @Override
         public String getSerializedName() {
             return this.name;
+        }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(BiomeSpecialEffects.GrassColorModifier.class);
+        }
+
+        @FunctionalInterface
+        public interface ColorModifier {
+            int modifyGrassColor(double x, double z, int color);
         }
     }
 }
