--- a/net/minecraft/world/level/biome/Biome.java
+++ b/net/minecraft/world/level/biome/Biome.java
@@ -195,7 +_,7 @@
     }
 
     public BiomeGenerationSettings getGenerationSettings() {
-        return this.generationSettings;
+        return this.modifiableBiomeInfo().get().generationSettings();
     }
 
     public int getGrassColor(double x, double z) {
@@ -419,5 +_,32 @@
         public String getSerializedName() {
             return this.name;
         }
+    }
+
+    // Neo: Grant ability to modify Biomes in specific ways such as adding new features or change temperature
+    private final net.neoforged.neoforge.common.world.ModifiableBiomeInfo modifiableBiomeInfo;
+
+    /**
+     * {@return Cache of original biome data and biome data modified by biome modifiers}
+     * Modified biome data is set by server after datapacks and serverconfigs load.
+     * Climate and effects field reads are coremodded to redirect to this.
+     **/
+    public net.neoforged.neoforge.common.world.ModifiableBiomeInfo modifiableBiomeInfo()
+    {
+         return this.modifiableBiomeInfo;
+    }
+
+    /**
+     * {@return The biome's climate settings, with modifications if called after modifiers are applied in server init.}
+     */
+    public ClimateSettings getModifiedClimateSettings() {
+        return this.modifiableBiomeInfo().get().climateSettings();
+    }
+
+    /**
+     * {@return The biome's client effects, with modifications if called after modifiers are applied in server init.}
+     */
+    public BiomeSpecialEffects getModifiedSpecialEffects() {
+        return this.modifiableBiomeInfo().get().effects();
     }
 }
