--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -450,7 +451,8 @@
             minecraft.gui.hud.getChat().restoreState(cookie.chatState());
         }
 
-        this.potionBrewing = PotionBrewing.bootstrap(this.enabledFeatures);
+        this.connectionType = cookie.connectionType();
+        this.potionBrewing = PotionBrewing.bootstrap(this.enabledFeatures, this.registryAccess);
         this.fuelValues = FuelValues.vanillaBurnTimes(cookie.receivedRegistries(), this.enabledFeatures);
         this.levelLoadTracker = cookie.levelLoadTracker();
         this.clockManager = new ClientClockManager();
@@ -1820,8 +1834,11 @@
         }
 
         this.fuelValues = FuelValues.vanillaBurnTimes(this.registryAccess, this.enabledFeatures);
-        List<ItemStack> searchItems = List.copyOf(CreativeModeTabs.searchTab().getDisplayItems());
-        this.searchTrees.updateCreativeTags(searchItems);
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.TagsUpdatedEvent.ClientPacketReceived(this.registryAccess, this.connection.isMemoryConnection()));
+        CreativeModeTabs.allTabs().stream().filter(net.minecraft.world.item.CreativeModeTab::hasSearchBar).forEach(tab -> {
+            List<ItemStack> stacks = List.copyOf(tab.getDisplayItems());
+            this.searchTrees.updateCreativeTags(stacks, net.neoforged.neoforge.client.CreativeModeTabSearchRegistry.getTagSearchKey(tab));
+        });
     }
 
     @Override
