--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -100,6 +_,14 @@
         this.setHealth(this.getMaxHealth());
         this.noPhysics = true;
         this.phaseManager = new EnderDragonPhaseManager(this);
+        this.setId(ENTITY_COUNTER.getAndAdd(this.subEntities.length + 1) + 1); // Forge: Fix MC-158205: Make sure part ids are successors of parent mob id
+    }
+
+    @Override
+    public void setId(int id) {
+        super.setId(id);
+        for (int i = 0; i < this.subEntities.length; i++) // Forge: Fix MC-158205: Set part ids to successors of parent mob id
+            this.subEntities[i].setId(id + i + 1);
     }
 
     public void setDragonFight(EnderDragonFight fight) {
@@ -148,8 +_,19 @@
         entityData.define(DATA_PHASE, EnderDragonPhase.HOVERING.getId());
     }
 
+    private net.minecraft.world.entity.@Nullable EntityReference<Player> unlimitedLastHurtByPlayer = null;
+
     @Override
     public void aiStep() {
+        // lastHurtByPlayer is cleared after 100 ticks, capture it indefinitely in unlimitedLastHurtByPlayer for LivingExperienceDropEvent
+        if (this.lastHurtByPlayer != null) this.unlimitedLastHurtByPlayer = lastHurtByPlayer;
+        if (this.unlimitedLastHurtByPlayer != null) {
+            Player p = this.unlimitedLastHurtByPlayer.getEntity(this.level(), Player.class);
+            if (p == null || p.isRemoved()) {
+                this.unlimitedLastHurtByPlayer = null;
+            }
+        }
+
         this.processFlappingMovement();
         if (this.level().isClientSide()) {
             this.setHealth(this.getHealth());
@@ -434,7 +_,7 @@
                     BlockPos blockPos = new BlockPos(x, y, z);
                     BlockState state = level.getBlockState(blockPos);
                     if (!state.isAir() && !state.is(BlockTags.DRAGON_TRANSPARENT)) {
-                        if (level.getGameRules().get(GameRules.MOB_GRIEFING) && !state.is(BlockTags.DRAGON_IMMUNE)) {
+                        if (net.neoforged.neoforge.common.CommonHooks.canEntityDestroy(level, blockPos, this) && !state.is(BlockTags.DRAGON_IMMUNE)) {
                             destroyedBlock = level.removeBlock(blockPos, false) || destroyedBlock;
                         } else {
                             hitWall = true;
@@ -535,7 +_,8 @@
 
         if (this.level() instanceof ServerLevel level) {
             if (this.dragonDeathTime > 150 && this.dragonDeathTime % 5 == 0 && level.getGameRules().get(GameRules.MOB_DROPS)) {
-                ExperienceOrb.award(level, this.position(), Mth.floor(xpCount * 0.08F));
+                int award = net.neoforged.neoforge.event.EventHooks.getExperienceDrop(this, net.minecraft.world.entity.EntityReference.get(this.unlimitedLastHurtByPlayer, level, Player.class), Mth.floor((float) xpCount * 0.08F));
+                ExperienceOrb.award((ServerLevel) this.level(), this.position(), award);
             }
 
             if (this.dragonDeathTime == 1 && !this.isSilent()) {
@@ -553,7 +_,8 @@
 
         if (this.dragonDeathTime >= 200 && this.level() instanceof ServerLevel level) {
             if (level.getGameRules().get(GameRules.MOB_DROPS)) {
-                ExperienceOrb.award(level, this.position(), Mth.floor(xpCount * 0.2F));
+                int award = net.neoforged.neoforge.event.EventHooks.getExperienceDrop(this, net.minecraft.world.entity.EntityReference.get(this.unlimitedLastHurtByPlayer, level, Player.class), Mth.floor((float) xpCount * 0.2F));
+                ExperienceOrb.award((ServerLevel) this.level(), this.position(), award);
             }
 
             if (this.dragonFight != null) {
@@ -859,8 +_,19 @@
     }
 
     @Override
+    public boolean isMultipartEntity() {
+        return true;
+    }
+
+    @Override
+    public net.neoforged.neoforge.entity.PartEntity<?>[] getParts() {
+        return this.subEntities;
+    }
+
+    @Override
     public void recreateFromPacket(ClientboundAddEntityPacket packet) {
         super.recreateFromPacket(packet);
+        if (true) return; // Forge: Fix MC-158205: Moved into setId()
         EnderDragonPart[] subEntities = this.getSubEntities();
 
         for (int i = 0; i < subEntities.length; i++) {
