--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -230,7 +_,8 @@
 
             return true;
         } else {
-            Predicate<LivingEntity> predicate = this.isActive() ? LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM : p_380174_ -> true;
+            // Neo: provide entity being looked at to disguise check
+            java.util.function.BiPredicate<LivingEntity, LivingEntity> predicate = this.isActive() ? LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET : (p_380174_, target) -> true;
 
             for (Player player : list) {
                 if (!player.isCreative()
