--- a/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -51,7 +_,7 @@
         this.eraseCache = eraseCache;
         this.dataFixer = dataFixer;
         this.levelStorage = levelSource;
-        this.overworldSavedDataStorage = new SavedDataStorage(this.levelStorage.getDimensionPath(Level.OVERWORLD).resolve("data"), dataFixer, registryAccess);
+        this.overworldSavedDataStorage = new SavedDataStorage(null, this.levelStorage.getDimensionPath(Level.OVERWORLD).resolve("data"), dataFixer, registryAccess);
         this.recreateRegionFiles = recreateRegionFiles;
         this.thread = THREAD_FACTORY.newThread(this::work);
         this.thread.setUncaughtExceptionHandler((t, e) -> {
