--- a/net/minecraft/world/entity/projectile/hurtingprojectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/AbstractHurtingProjectile.java
@@ -18,11 +18,17 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public abstract class AbstractHurtingProjectile extends Projectile {
 
     public static final double INITAL_ACCELERATION_POWER = 0.1D;
     public static final double DEFLECTION_SCALE = 0.5D;
     public double accelerationPower;
+    public float bukkitYield = 1; // CraftBukkit
+    public boolean isIncendiary = true; // CraftBukkit
 
     protected AbstractHurtingProjectile(EntityType<? extends AbstractHurtingProjectile> type, Level level) {
         super(type, level);
@@ -90,12 +96,12 @@
             }
 
             if (hitresult.getType() != HitResult.Type.MISS && this.isAlive()) {
-                this.hitTargetOrDeflectSelf(hitresult);
+                this.preHitTargetOrDeflectSelf(hitresult); // CraftBukkit - projectile hit event
             }
 
             this.createParticleTrail();
         } else {
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
     }
 
