--- a/net/minecraft/client/resources/model/EquipmentClientInfo.java
+++ b/net/minecraft/client/resources/model/EquipmentClientInfo.java
@@ -113,6 +_,7 @@
     }
 
     @OnlyIn(Dist.CLIENT)
+    @net.neoforged.fml.common.asm.enumextension.NamedEnum
     public static enum LayerType implements StringRepresentable, net.neoforged.fml.common.asm.enumextension.IExtensibleEnum {
         HUMANOID("humanoid"),
         HUMANOID_LEGGINGS("humanoid_leggings"),
@@ -134,7 +_,7 @@
         private final String id;
 
         private LayerType(String p_387059_) {
-            this.id = p_387059_;
+            this.id = p_387059_.replaceAll(":", "/");
         }
 
         @Override
@@ -144,6 +_,10 @@
 
         public String trimAssetPrefix() {
             return "trims/entity/" + this.id;
+        }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(LayerType.class);
         }
     }
 }
