--- a/net/minecraft/client/resources/model/EquipmentClientInfo.java
+++ b/net/minecraft/client/resources/model/EquipmentClientInfo.java
@@ -108,6 +_,7 @@
         }
     }
 
+    @net.neoforged.fml.common.asm.enumextension.NamedEnum
     public enum LayerType implements StringRepresentable, IExtensibleEnum {
         HUMANOID("humanoid"),
         HUMANOID_LEGGINGS("humanoid_leggings"),
@@ -133,7 +_,7 @@
         private final String id;
 
         LayerType(String id) {
-            this.id = id;
+            this.id = id.replaceAll(":", "/");
         }
 
         @Override
@@ -143,6 +_,10 @@
 
         public String trimAssetPrefix() {
             return "trims/entity/" + this.id;
+        }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(LayerType.class);
         }
     }
 }
