--- a/net/minecraft/client/resources/model/EquipmentClientInfo.java
+++ b/net/minecraft/client/resources/model/EquipmentClientInfo.java
@@ -126,6 +_,7 @@
         }
     }
 
+    @net.neoforged.fml.common.asm.enumextension.NamedEnum
     public enum LayerType implements StringRepresentable, IExtensibleEnum {
         HUMANOID("humanoid"),
         HUMANOID_LEGGINGS("humanoid_leggings"),
@@ -151,7 +_,7 @@
         private final String id;
 
         LayerType(String id) {
-            this.id = id;
+            this.id = id.replaceAll(":", "/");
         }
 
         @Override
