--- a/net/minecraft/client/resources/model/ItemModel.java
+++ b/net/minecraft/client/resources/model/ItemModel.java
@@ -29,11 +_,11 @@
 
     @Override
     public BakedModel bake(ModelBaker p_371426_, Function<Material, TextureAtlasSprite> p_371750_, ModelState p_371674_) {
-        BakedModel bakedmodel = p_371426_.bake(this.id, p_371674_);
+        BakedModel bakedmodel = p_371426_.bake(this.id, p_371674_, p_371750_);
         if (this.overrides.isEmpty()) {
             return bakedmodel;
         } else {
-            BakedOverrides bakedoverrides = new BakedOverrides(p_371426_, this.overrides);
+            BakedOverrides bakedoverrides = new BakedOverrides(p_371426_, this.overrides, p_371750_);
             return new ItemModel.BakedModelWithOverrides(bakedmodel, bakedoverrides);
         }
     }
