--- a/net/minecraft/client/resources/model/ModelDiscovery.java
+++ b/net/minecraft/client/resources/model/ModelDiscovery.java
@@ -60,7 +_,13 @@
     }
 
     private ModelDiscovery.ModelWrapper getOrCreateModel(Identifier id) {
-        return this.modelWrappers.computeIfAbsent(id, this.uncachedResolver);
+        // Neo: Remove computeIfAbsent, might cause problems with nested model discovery.
+        var wrapper = this.modelWrappers.get(id);
+        if (wrapper == null) {
+            wrapper = this.uncachedResolver.get(id);
+            this.modelWrappers.put(id, wrapper);
+        }
+        return wrapper;
     }
 
     private ModelDiscovery.ModelWrapper createAndQueueWrapper(Identifier id, UnbakedModel rawModel) {
@@ -69,6 +_,8 @@
         if (!isRoot) {
             this.parentDiscoveryQueue.add(result);
         }
+        // Neo: Allow UnbakedModels to resolve additional dependencies
+        rawModel.resolveDependencies(this.resolver);
 
         return result;
     }
@@ -148,16 +_,18 @@
         private static final ModelDiscovery.Slot<TextureSlots> KEY_TEXTURE_SLOTS = slot(4);
         private static final ModelDiscovery.Slot<Material.Baked> KEY_PARTICLE_SPRITE = slot(5);
         private static final ModelDiscovery.Slot<QuadCollection> KEY_DEFAULT_GEOMETRY = slot(6);
-        private static final int SLOT_COUNT = 7;
+        // Neo: Cache additional properties too
+        private static final ModelDiscovery.Slot<net.minecraft.util.context.ContextMap> KEY_ADDITIONAL_PROPERTIES = slot(7);
+        private static final int SLOT_COUNT = 8;
         private final Identifier id;
         private boolean valid;
         private ModelDiscovery.@Nullable ModelWrapper parent;
         private final UnbakedModel wrapped;
-        private final AtomicReferenceArray<@Nullable Object> fixedSlots = new AtomicReferenceArray<>(7);
+        private final AtomicReferenceArray<@Nullable Object> fixedSlots = new AtomicReferenceArray<>(SLOT_COUNT);
         private final Map<ModelState, QuadCollection> modelBakeCache = new ConcurrentHashMap<>();
 
         private static <T> ModelDiscovery.Slot<T> slot(int index) {
-            Objects.checkIndex(index, 7);
+            Objects.checkIndex(index, SLOT_COUNT);
             return new ModelDiscovery.Slot<>(index);
         }
 
@@ -229,15 +_,20 @@
 
         private QuadCollection bakeDefaultState(TextureSlots textureSlots, ModelBaker baker, ModelState state) {
             QuadCollection result = this.getSlot(KEY_DEFAULT_GEOMETRY);
-            return result != null ? result : this.updateSlot(KEY_DEFAULT_GEOMETRY, this.getTopGeometry().bake(textureSlots, baker, state, this));
+            return result != null ? result : this.updateSlot(KEY_DEFAULT_GEOMETRY, this.getTopGeometry().bake(textureSlots, baker, state, this, this.getTopAdditionalProperties()));
         }
 
         @Override
         public QuadCollection bakeTopGeometry(TextureSlots textureSlots, ModelBaker baker, ModelState state) {
             return state == BlockModelRotation.IDENTITY ? this.bakeDefaultState(textureSlots, baker, state) : this.modelBakeCache.computeIfAbsent(state, s -> {
                 UnbakedGeometry topGeometry = this.getTopGeometry();
-                return topGeometry.bake(textureSlots, baker, s, this);
+                return topGeometry.bake(textureSlots, baker, s, this, this.getTopAdditionalProperties());
             });
+        }
+
+        @Override
+        public net.minecraft.util.context.ContextMap getTopAdditionalProperties() {
+            return this.getSimpleProperty(KEY_ADDITIONAL_PROPERTIES, net.neoforged.neoforge.client.extensions.ResolvedModelExtension::findTopAdditionalProperties);
         }
     }
 
