--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -117,6 +_,13 @@
         p_252014_.popPush("special");
         this.loadSpecialItemModelAndDependencies(ItemRenderer.TRIDENT_IN_HAND_MODEL);
         this.loadSpecialItemModelAndDependencies(ItemRenderer.SPYGLASS_IN_HAND_MODEL);
+        Set<ModelResourceLocation> additionalModels = new HashSet<>();
+        net.neoforged.neoforge.client.ClientHooks.onRegisterAdditionalModels(additionalModels);
+        for (ModelResourceLocation rl : additionalModels) {
+            UnbakedModel unbakedmodel = this.getModel(rl.id());
+            this.unbakedCache.put(rl.id(), unbakedmodel);
+            this.registerModelAndLoadDependencies(rl, unbakedmodel);
+        }
         this.topLevelModels.values().forEach(p_247954_ -> p_247954_.resolveParents(this::getModel));
         p_252014_.pop();
     }
@@ -246,14 +_,24 @@
         }
 
         @Override
+        public Function<Material, TextureAtlasSprite> getModelTextureGetter() {
+            return this.modelTextureGetter;
+        }
+
+        @Override
         public BakedModel bake(ResourceLocation p_252176_, ModelState p_249765_) {
+            return bake(p_252176_, p_249765_, this.modelTextureGetter);
+        }
+
+        @Override
+        public BakedModel bake(ResourceLocation p_252176_, ModelState p_249765_, Function<Material, TextureAtlasSprite> sprites) {
             ModelBakery.BakedCacheKey modelbakery$bakedcachekey = new ModelBakery.BakedCacheKey(p_252176_, p_249765_.getRotation(), p_249765_.isUvLocked());
             BakedModel bakedmodel = ModelBakery.this.bakedCache.get(modelbakery$bakedcachekey);
             if (bakedmodel != null) {
                 return bakedmodel;
             } else {
                 UnbakedModel unbakedmodel = this.getModel(p_252176_);
-                BakedModel bakedmodel1 = this.bakeUncached(unbakedmodel, p_249765_);
+                BakedModel bakedmodel1 = this.bakeUncached(unbakedmodel, p_249765_, sprites);
                 ModelBakery.this.bakedCache.put(modelbakery$bakedcachekey, bakedmodel1);
                 return bakedmodel1;
             }
@@ -261,13 +_,18 @@
 
         @Nullable
         BakedModel bakeUncached(UnbakedModel p_352386_, ModelState p_352194_) {
+            return bakeUncached(p_352386_, p_352194_, this.modelTextureGetter);
+        }
+
+        @Nullable
+        BakedModel bakeUncached(UnbakedModel p_352386_, ModelState p_352194_, Function<Material, TextureAtlasSprite> sprites) {
             if (p_352386_ instanceof BlockModel blockmodel && blockmodel.getRootModel() == ModelBakery.GENERATION_MARKER) {
                 return ModelBakery.ITEM_MODEL_GENERATOR
-                    .generateBlockModel(this.modelTextureGetter, blockmodel)
-                    .bake(this, blockmodel, this.modelTextureGetter, p_352194_, false);
+                    .generateBlockModel(sprites, blockmodel)
+                    .bake(this, blockmodel, sprites, p_352194_, false);
             }
 
-            return p_352386_.bake(this, this.modelTextureGetter, p_352194_);
+            return p_352386_.bake(this, sprites, p_352194_);
         }
     }
 
