--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -73,6 +_,20 @@
         Map<Identifier, ResolvedModel> resolvedModels,
         ResolvedModel missingModel
     ) {
+        this(entityModelSet, materials, playerSkinRenderCache, unbakedBlockStateModels, clientInfos, resolvedModels, missingModel, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.LoadedModels.EMPTY, net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations.EMPTY);
+    }
+
+    public ModelBakery(
+        EntityModelSet entityModelSet,
+        MaterialSet materials,
+        PlayerSkinRenderCache playerSkinRenderCache,
+        Map<BlockState, BlockStateModel.UnbakedRoot> unbakedBlockStateModels,
+        Map<Identifier, ClientItem> clientInfos,
+        Map<Identifier, ResolvedModel> resolvedModels,
+        ResolvedModel missingModel,
+        net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.LoadedModels standaloneModels,
+        net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations pendingAnimations
+    ) {
         this.entityModelSet = entityModelSet;
         this.sprites = sprites;
         this.playerSkinRenderCache = playerSkinRenderCache;
@@ -80,6 +_,8 @@
         this.clientInfos = clientInfos;
         this.resolvedModels = resolvedModels;
         this.missingModel = missingModel;
+        this.standaloneModels = standaloneModels;
+        this.pendingAnimations = pendingAnimations;
     }
 
     public CompletableFuture<ModelBakery.BakingResult> bakeModels(MaterialBaker materials, Executor taskExecutor) {
@@ -103,7 +_,7 @@
                     return clientInfo.model()
                         .bake(
                             new ItemModel.BakingContext(
-                                baker, this.entityModelSet, this.sprites, this.playerSkinRenderCache, missingModels.item, clientInfo.registrySwapper()
+                                baker, this.entityModelSet, this.materials, this.playerSkinRenderCache, missingModels.item, clientInfo.registrySwapper(), this.pendingAnimations
                             )
                         );
                 } catch (Exception var6x) {
@@ -113,6 +_,8 @@
             },
             taskExecutor
         );
+        CompletableFuture<net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels> standaloneModelsFuture =
+                net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.bake(this.standaloneModels, baker, taskExecutor);
         Map<Identifier, ClientItem.Properties> itemStackModelProperties = new HashMap<>(this.clientInfos.size());
         this.clientInfos.forEach((id, clientInfo) -> {
             ClientItem.Properties properties = clientInfo.properties();
@@ -120,13 +_,15 @@
                 itemStackModelProperties.put(id, properties);
             }
         });
+        bakedBlockStateModelFuture = bakedBlockStateModelFuture.thenCombine(standaloneModelsFuture, (stateModels, standaloneModels) -> stateModels);
         return bakedBlockStateModelFuture.thenCombine(
             bakedItemStackModelFuture,
             (bakedBlockStateModels, bakedItemStateModels) -> new ModelBakery.BakingResult(
                 missingModels,
                 (Map<BlockState, BlockStateModel>)bakedBlockStateModels,
                 (Map<Identifier, ItemModel>)bakedItemStateModels,
-                itemStackModelProperties
+                itemStackModelProperties,
+                standaloneModelsFuture.join()
             )
         );
     }
@@ -137,7 +_,20 @@
         Map<BlockState, BlockStateModel> blockStateModels,
         Map<Identifier, ItemModel> itemStackModels,
         Map<Identifier, ClientItem.Properties> itemProperties
+        , net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels standaloneModels
     ) {
+        /**
+         * @deprecated Neo: use {@link #BakingResult(ModelBakery.MissingModels, Map, Map, Map, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels)} instead
+         */
+        @Deprecated
+        public BakingResult(
+                ModelBakery.MissingModels missingModels,
+                Map<BlockState, BlockStateModel> blockStateModels,
+                Map<Identifier, ItemModel> itemStackModels,
+                Map<Identifier, ClientItem.Properties> itemProperties
+        ) {
+            this(missingModels, blockStateModels, itemStackModels, itemProperties, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels.EMPTY);
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
