--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -67,7 +_,11 @@
     private final Map<Identifier, ClientItem> clientInfos;
     private final Map<Identifier, ResolvedModel> resolvedModels;
     private final ResolvedModel missingModel;
+    private final net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.LoadedModels standaloneModels;
+    private final net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations pendingAnimations;
 
+    /// @deprecated Neo: use [#ModelBakery(EntityModelSet, SpriteGetter, PlayerSkinRenderCache, Map, Map, Map, ResolvedModel, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.LoadedModels, net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations)] instead
+    @Deprecated
     public ModelBakery(
         EntityModelSet entityModelSet,
         SpriteGetter sprites,
@@ -77,6 +_,20 @@
         Map<Identifier, ResolvedModel> resolvedModels,
         ResolvedModel missingModel
     ) {
+        this(entityModelSet, sprites, playerSkinRenderCache, unbakedBlockStateModels, clientInfos, resolvedModels, missingModel, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.LoadedModels.EMPTY, net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations.EMPTY);
+    }
+
+    public ModelBakery(
+        EntityModelSet entityModelSet,
+        SpriteGetter sprites,
+        PlayerSkinRenderCache playerSkinRenderCache,
+        Map<BlockState, BlockStateModel.UnbakedRoot> unbakedBlockStateModels,
+        Map<Identifier, ClientItem> clientInfos,
+        Map<Identifier, ResolvedModel> resolvedModels,
+        ResolvedModel missingModel,
+        net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.LoadedModels standaloneModels,
+        net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations pendingAnimations
+    ) {
         this.entityModelSet = entityModelSet;
         this.sprites = sprites;
         this.playerSkinRenderCache = playerSkinRenderCache;
@@ -84,6 +_,8 @@
         this.clientInfos = clientInfos;
         this.resolvedModels = resolvedModels;
         this.missingModel = missingModel;
+        this.standaloneModels = standaloneModels;
+        this.pendingAnimations = pendingAnimations;
     }
 
     public CompletableFuture<ModelBakery.BakingResult> bakeModels(MaterialBaker materials, Executor taskExecutor) {
@@ -107,7 +_,7 @@
                     return clientInfo.model()
                         .bake(
                             new ItemModel.BakingContext(
-                                baker, this.entityModelSet, this.sprites, this.playerSkinRenderCache, missingModels.item, clientInfo.registrySwapper()
+                                baker, this.entityModelSet, this.sprites, this.playerSkinRenderCache, missingModels.item, clientInfo.registrySwapper(), this.pendingAnimations
                             ),
                             IDENTITY
                         );
@@ -118,6 +_,8 @@
             },
             taskExecutor
         );
+        CompletableFuture<net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels> standaloneModelsFuture =
+                net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.bake(this.standaloneModels, baker, taskExecutor);
         Map<Identifier, ClientItem.Properties> itemStackModelProperties = new HashMap<>(this.clientInfos.size());
         this.clientInfos.forEach((id, clientInfo) -> {
             ClientItem.Properties properties = clientInfo.properties();
@@ -125,13 +_,15 @@
                 itemStackModelProperties.put(id, properties);
             }
         });
+        bakedBlockStateModelFuture = bakedBlockStateModelFuture.thenCombine(standaloneModelsFuture, (stateModels, standaloneModels) -> stateModels);
         return bakedBlockStateModelFuture.thenCombine(
             bakedItemStackModelFuture,
             (bakedBlockStateModels, bakedItemStateModels) -> new ModelBakery.BakingResult(
                 missingModels,
                 (Map<BlockState, BlockStateModel>)bakedBlockStateModels,
                 (Map<Identifier, ItemModel>)bakedItemStateModels,
-                itemStackModelProperties
+                itemStackModelProperties,
+                standaloneModelsFuture.join()
             )
         );
     }
@@ -142,7 +_,19 @@
         Map<BlockState, BlockStateModel> blockStateModels,
         Map<Identifier, ItemModel> itemStackModels,
         Map<Identifier, ClientItem.Properties> itemProperties
+        , net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels standaloneModels
     ) {
+        /// @deprecated Neo: use [#BakingResult(ModelBakery.MissingModels, Map, Map, Map, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels)] instead
+        @Deprecated
+        public BakingResult(
+                ModelBakery.MissingModels missingModels,
+                Map<BlockState, BlockStateModel> blockStateModels,
+                Map<Identifier, ItemModel> itemStackModels,
+                Map<Identifier, ClientItem.Properties> itemProperties
+        ) {
+            this(missingModels, blockStateModels, itemStackModels, itemProperties, net.neoforged.neoforge.client.model.standalone.StandaloneModelLoader.BakedModels.EMPTY);
+        }
+
         public BlockStateModel getBlockStateModel(BlockState blockState) {
             return this.blockStateModels.getOrDefault(blockState, this.missingModels.block);
         }
@@ -152,6 +_,8 @@
     private static class InternerImpl implements ModelBaker.Interner {
         private final Interner<Vector3fc> vectors = Interners.newStrongInterner();
         private final Interner<BakedQuad.MaterialInfo> materialInfos = Interners.newStrongInterner();
+        private final Interner<net.neoforged.neoforge.client.model.quad.BakedNormals> normals = Interners.newStrongInterner();
+        private final Interner<net.neoforged.neoforge.client.model.quad.BakedColors> colors = Interners.newStrongInterner();
 
         @Override
         public Vector3fc vector(Vector3fc v) {
@@ -161,6 +_,16 @@
         @Override
         public BakedQuad.MaterialInfo materialInfo(BakedQuad.MaterialInfo material) {
             return this.materialInfos.intern(material);
+        }
+
+        @Override
+        public net.neoforged.neoforge.client.model.quad.BakedNormals normals(net.neoforged.neoforge.client.model.quad.BakedNormals normals) {
+            return this.normals.intern(normals);
+        }
+
+        @Override
+        public net.neoforged.neoforge.client.model.quad.BakedColors colors(net.neoforged.neoforge.client.model.quad.BakedColors colors) {
+            return this.colors.intern(colors);
         }
     }
 
