--- a/net/minecraft/client/resources/model/ModelBaker.java
+++ b/net/minecraft/client/resources/model/ModelBaker.java
@@ -6,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface ModelBaker {
+public interface ModelBaker extends net.neoforged.neoforge.client.extensions.IModelBakerExtension {
     BakedModel bake(ResourceLocation p_250776_, ModelState p_251280_);
 
     SpriteGetter sprites();
