--- a/net/minecraft/client/resources/model/ModelDiscovery.java
+++ b/net/minecraft/client/resources/model/ModelDiscovery.java
@@ -22,6 +_,7 @@
     final UnbakedModel missingModel;
     private final List<ResolvableModel> topModels = new ArrayList<>();
     private final Map<ResourceLocation, UnbakedModel> referencedModels = new HashMap<>();
+    final Map<ResourceLocation, UnbakedModel> standaloneModels = new HashMap<>();
 
     public ModelDiscovery(Map<ResourceLocation, UnbakedModel> p_360750_, UnbakedModel p_365355_) {
         this.inputModels = p_360750_;
@@ -31,6 +_,12 @@
 
     public void registerSpecialModels() {
         this.referencedModels.put(ItemModelGenerator.GENERATED_ITEM_MODEL_ID, new ItemModelGenerator());
+
+        net.neoforged.neoforge.client.ClientHooks.onRegisterAdditionalModels(path -> {
+            UnbakedModel model = getBlockModel(path);
+            this.addRoot(model);
+            this.standaloneModels.put(path, model);
+        });
     }
 
     public void addRoot(ResolvableModel p_388596_) {
