--- a/net/minecraft/client/resources/model/SimpleModelWrapper.java
+++ b/net/minecraft/client/resources/model/SimpleModelWrapper.java
@@ -25,6 +_,11 @@
 
     public static BlockStateModelPart bake(ModelBaker modelBakery, Identifier location, ModelState state) {
         ResolvedModel model = modelBakery.getModel(location);
+        return bake(modelBakery, model, state);
+    }
+
+    // Neo: split off to allow baking an existing ResolvedModel into a BlockModelPart
+    public static BlockStateModelPart bake(ModelBaker modelBakery, ResolvedModel model, ModelState state) {
         TextureSlots textureSlots = model.getTopTextureSlots();
         boolean hasAmbientOcclusion = model.getTopAmbientOcclusion();
         Material.Baked particleMaterial = model.resolveParticleMaterial(textureSlots, modelBakery);
@@ -43,7 +_,7 @@
         }
 
         if (forbiddenSprites != null) {
-            LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", location, forbiddenSprites);
+            LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", model.debugName(), forbiddenSprites);
             return modelBakery.missingBlockModelPart();
         } else {
             return new SimpleModelWrapper(geometry, hasAmbientOcclusion, particleMaterial);
