--- a/net/minecraft/client/resources/model/UnbakedGeometry.java
+++ b/net/minecraft/client/resources/model/UnbakedGeometry.java
@@ -9,5 +_,9 @@
 public interface UnbakedGeometry extends net.neoforged.neoforge.client.extensions.UnbakedGeometryExtension {
     UnbakedGeometry EMPTY = (p_405112_, p_405354_, p_405672_, p_405151_) -> QuadCollection.EMPTY;
 
+    /**
+     * @deprecated Neo: Use {@link #bake(TextureSlots, ModelBaker, ModelState, ModelDebugName, net.minecraft.util.context.ContextMap)}.
+     */
+    @Deprecated
     QuadCollection bake(TextureSlots p_405831_, ModelBaker p_405026_, ModelState p_405122_, ModelDebugName p_405635_);
 }
