--- a/net/minecraft/client/resources/model/UnbakedModel.java
+++ b/net/minecraft/client/resources/model/UnbakedModel.java
@@ -9,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface UnbakedModel {
+public interface UnbakedModel extends net.neoforged.neoforge.client.resource.model.extensions.IUnbakedModelExtension {
     Collection<ResourceLocation> getDependencies();
 
     void resolveParents(Function<ResourceLocation, UnbakedModel> p_119538_);
