--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -123,7 +123,7 @@
             cachedusernametoidresolver_gameprofileinfo.setLastAccess(this.getNextOperation());
             optional = Optional.of(cachedusernametoidresolver_gameprofileinfo.nameAndId());
         } else {
-            Optional<NameAndId> optional1 = this.lookupGameProfile(this.profileRepository, s1);
+            Optional<NameAndId> optional1 = this.lookupGameProfile(this.profileRepository, name); // CraftBukkit - use correct case for offline players
 
             if (optional1.isPresent()) {
                 optional = Optional.of(this.addInternal((NameAndId) optional1.get()).nameAndId());
@@ -169,7 +169,7 @@
             DateFormat dateformat = createDateFormat();
 
             jsonarray.forEach((jsonelement) -> {
-                Optional optional = readGameProfile(jsonelement, dateformat);
+                Optional<CachedUserNameToIdResolver.GameProfileInfo> optional = readGameProfile(jsonelement, dateformat); // CraftBukkit - decompile error
 
                 Objects.requireNonNull(list);
                 optional.ifPresent(list::add);
