--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -27,6 +27,12 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.nbt.NbtIo;
+import net.minecraft.nbt.NbtAccounter;
+import net.minecraft.nbt.CompoundTag;
+// CraftBukkit end
+
 public class OldUsersConverter {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -77,7 +83,7 @@
                 try {
                     userbanlist.load();
                 } catch (IOException ioexception) {
-                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", userbanlist.getFile().getName(), ioexception);
+                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", userbanlist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -137,7 +143,7 @@
                 try {
                     ipbanlist.load();
                 } catch (IOException ioexception) {
-                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", ipbanlist.getFile().getName(), ioexception);
+                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", ipbanlist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -176,7 +182,7 @@
                 try {
                     serveroplist.load();
                 } catch (IOException ioexception) {
-                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", serveroplist.getFile().getName(), ioexception);
+                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", serveroplist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -222,7 +228,7 @@
                 try {
                     userwhitelist.load();
                 } catch (IOException ioexception) {
-                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", userwhitelist.getFile().getName(), ioexception);
+                    OldUsersConverter.LOGGER.warn("Could not load existing file {}", userwhitelist.getFile().getName()); // CraftBukkit - don't print stacktrace
                 }
             }
 
@@ -341,6 +347,30 @@
                         File file5 = new File(file, oldName + ".dat");
                         File file6 = new File(directory, newName + ".dat");
 
+                        // CraftBukkit start - Use old file name to seed lastKnownName
+                        CompoundTag root = null;
+
+                        try {
+                            root = NbtIo.readCompressed(new java.io.FileInputStream(file5), NbtAccounter.unlimitedHeap());
+                        } catch (Exception exception) {
+                            exception.printStackTrace();
+                        }
+
+                        if (root != null) {
+                            if (!root.contains("bukkit")) {
+                                root.put("bukkit", new CompoundTag());
+                            }
+                            CompoundTag data = root.getCompoundOrEmpty("bukkit");
+                            data.putString("lastKnownName", oldName);
+
+                            try {
+                                NbtIo.writeCompressed(root, new java.io.FileOutputStream(file2));
+                            } catch (Exception exception) {
+                                exception.printStackTrace();
+                            }
+                       }
+                        // CraftBukkit end
+
                         OldUsersConverter.ensureDirectoryExists(directory);
                         if (!file5.renameTo(file6)) {
                             throw new OldUsersConverter.ConversionError("Could not convert file for " + oldName);
