--- a/net/minecraft/client/resources/model/cuboid/CuboidModelElement.java
+++ b/net/minecraft/client/resources/model/cuboid/CuboidModelElement.java
@@ -71,7 +_,8 @@
                 }
             }
 
-            return new CuboidModelElement(from, to, faces, rotation, shadeDirectionOverride, lightEmission);
+            var faceData = net.neoforged.neoforge.client.model.ExtraFaceData.read(object.get("neoforge_data"), net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT);
+            return new CuboidModelElement(from, to, faces, rotation, shade, lightEmission, faceData);
         }
 
         private @Nullable Direction getShadeDirectionOverride(JsonObject object) {
