--- a/net/minecraft/client/resources/model/cuboid/FaceBakery.java
+++ b/net/minecraft/client/resources/model/cuboid/FaceBakery.java
@@ -70,9 +_,11 @@
         Transparency transparency = computeMaterialTransparency(material, uvs);
         ModelBaker.Interner interner = modelBaker.interner();
         BakedQuad.SpriteInfo spriteInfo = interner.spriteInfo(BakedQuad.SpriteInfo.of(material, transparency));
-        return bakeQuad(interner, from, to, uvs, face.rotation(), face.tintIndex(), spriteInfo, facing, modelState, elementRotation, shade, lightEmission);
+        return bakeQuad(interner, from, to, uvs, face.rotation(), face.tintIndex(), spriteInfo, facing, modelState, elementRotation, shade, lightEmission, face.faceData());
     }
 
+    /** @deprecated Neo: use {@link #bakeQuad(ModelBaker.Interner, Vector3fc, Vector3fc, CuboidFace.UVs, Quadrant, int, BakedQuad.SpriteInfo, Direction, ModelState, CuboidRotation, boolean, int, net.neoforged.neoforge.client.model.ExtraFaceData)} instead */
+    @Deprecated
     public static BakedQuad bakeQuad(
         ModelBaker.Interner interner,
         Vector3fc from,
@@ -87,6 +_,24 @@
         boolean shade,
         int lightEmission
     ) {
+        return bakeQuad(interner, from, to, uvs, uvRotation,tintIndex, spriteInfo, facing, modelState, elementRotation, shade, lightEmission, net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT);
+    }
+
+    public static BakedQuad bakeQuad(
+            ModelBaker.Interner interner,
+            Vector3fc from,
+            Vector3fc to,
+            CuboidFace.UVs uvs,
+            Quadrant uvRotation,
+            int tintIndex,
+            BakedQuad.SpriteInfo spriteInfo,
+            Direction facing,
+            ModelState modelState,
+            @Nullable CuboidRotation elementRotation,
+            boolean shade,
+            int lightEmission,
+            net.neoforged.neoforge.client.model.ExtraFaceData faceData
+    ) {
         Matrix4fc uvTransform = modelState.inverseFaceTransformation(facing);
         Vector3fc[] vertexPositions = new Vector3fc[4];
         long[] vertexPackedUvs = new long[4];
@@ -112,6 +_,8 @@
 
         Direction finalDirection = calculateFacing(vertexPositions);
         if (elementRotation == null && finalDirection != null) {
+            // Neo: Suppress winding re-calculation when the quads may not be axis-aligned due to root transforms
+            if (!modelState.mayApplyArbitraryRotation())
             recalculateWinding(vertexPositions, vertexPackedUvs, finalDirection);
         }
 
@@ -128,7 +_,10 @@
             Objects.requireNonNullElse(finalDirection, Direction.UP),
             spriteInfo,
             shade,
-            lightEmission
+            Math.max(lightEmission, faceData.lightEmission()),
+            interner.normals(net.neoforged.neoforge.client.model.quad.BakedNormals.of(net.neoforged.neoforge.client.model.quad.BakedNormals.computeQuadNormal(vertexPositions[0], vertexPositions[1], vertexPositions[2], vertexPositions[3]))),
+            interner.colors(net.neoforged.neoforge.client.model.quad.BakedColors.of(faceData.color())),
+            faceData.ambientOcclusion()
         );
     }
 
@@ -214,6 +_,11 @@
     }
 
     public static void recalculateWinding(Vector3fc[] positions, long[] uvs, Direction direction) {
+        recalculateWinding(positions, uvs, direction, null, null);
+    }
+
+    // Neo: Allow swapping colors and normals alongside the positions/uvs
+    public static void recalculateWinding(Vector3fc[] positions, long[] uvs, Direction direction, int @Nullable [] colors, int @Nullable [] normals) {
         float minX = 999.0F;
         float minY = 999.0F;
         float minZ = 999.0F;
@@ -266,6 +_,12 @@
             if (vertexToSwap != vertex) {
                 swap(positions, vertexToSwap, vertex);
                 swap(uvs, vertexToSwap, vertex);
+                if (colors != null) {
+                    swap(colors, vertexToSwap, vertex);
+                }
+                if (normals != null) {
+                    swap(normals, vertexToSwap, vertex);
+                }
             }
         }
     }
@@ -289,6 +_,13 @@
 
     private static void swap(long[] array, int indexA, int indexB) {
         long tmp = array[indexA];
+        array[indexA] = array[indexB];
+        array[indexB] = tmp;
+    }
+
+    // Neo: Used by the winding patch to swap colors/normals
+    private static void swap(int[] array, int indexA, int indexB) {
+        int tmp = array[indexA];
         array[indexA] = array[indexB];
         array[indexB] = tmp;
     }
