--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderDispatcher.java
@@ -73,13 +_,25 @@
         this.cameraPos = cameraPos;
     }
 
+    /**
+     * @deprecated Neo: use {@link #tryExtractRenderState(BlockEntity, float, ModelFeatureRenderer.CrumblingOverlay, net.minecraft.client.renderer.culling.Frustum)} instead
+     */
     public <E extends BlockEntity, S extends BlockEntityRenderState> @Nullable S tryExtractRenderState(
         E blockEntity, float partialTicks, ModelFeatureRenderer.@Nullable CrumblingOverlay breakProgress
     ) {
+        return tryExtractRenderState(blockEntity, partialTicks, breakProgress, null);
+    }
+
+    @Nullable
+    public <E extends BlockEntity, S extends BlockEntityRenderState> S tryExtractRenderState(
+            E blockEntity, float partialTicks, ModelFeatureRenderer.@Nullable CrumblingOverlay breakProgress, net.minecraft.client.renderer.culling.@Nullable Frustum frustum
+    ) {
         BlockEntityRenderer<E, S> renderer = this.getRenderer(blockEntity);
         if (renderer == null) {
             return null;
         } else if (!blockEntity.hasLevel() || !blockEntity.getType().isValid(blockEntity.getBlockState())) {
+            return null;
+        } else if (frustum != null && !frustum.isVisible(renderer.getRenderBoundingBox(blockEntity))) {
             return null;
         } else if (!renderer.shouldRender(blockEntity, this.cameraPos)) {
             return null;
