--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderer.java
@@ -8,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface BlockEntityRenderer<T extends BlockEntity> {
+public interface BlockEntityRenderer<T extends BlockEntity> extends net.neoforged.neoforge.client.extensions.IBlockEntityRendererExtension<T> {
     void render(T blockEntity, float partialTick, PoseStack poseStack, MultiBufferSource bufferSource, int packedLight, int packedOverlay);
 
     default boolean shouldRenderOffScreen(T blockEntity) {
