--- a/net/minecraft/client/renderer/blockentity/ChestRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/ChestRenderer.java
@@ -77,6 +_,8 @@
         if (state.type != ChestType.SINGLE) {
             state.lightCoords = combineResult.apply(new BrightnessCombiner<>()).applyAsInt(state.lightCoords);
         }
+
+        state.customSprite = getCustomSprite(blockEntity, state);
     }
 
     public void submit(ChestRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
@@ -87,7 +_,7 @@
         float open = state.open;
         open = 1.0F - open;
         open = 1.0F - open * open * open;
-        SpriteId spriteId = Sheets.chooseSprite(state.material, state.type);
+        SpriteId spriteId = state.customSprite != null ? state.customSprite : Sheets.chooseSprite(state.material, state.type);
         RenderType renderType = spriteId.renderType(RenderTypes::entityCutoutCull);
         TextureAtlasSprite sprite = this.sprites.get(spriteId);
         if (state.type != ChestType.SINGLE) {
@@ -124,5 +_,20 @@
         } else {
             return entity instanceof TrappedChestBlockEntity ? ChestRenderState.ChestMaterialType.TRAPPED : ChestRenderState.ChestMaterialType.REGULAR;
         }
+    }
+
+    /**
+     * Neo: Return a custom {@link SpriteId} to render the chest with or {@code null} to
+     * fall back to the vanilla material selection.
+     */
+    @Nullable
+    protected SpriteId getCustomSprite(T blockEntity, ChestRenderState renderState) {
+        return null;
+    }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(T blockEntity) {
+        net.minecraft.core.BlockPos pos = blockEntity.getBlockPos();
+        return net.minecraft.world.phys.AABB.encapsulatingFullBlocks(pos.offset(-1, 0, -1), pos.offset(1, 1, 1));
     }
 }
