--- a/net/minecraft/client/renderer/blockentity/ConduitRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/ConduitRenderer.java
@@ -131,4 +_,10 @@
             p_112401_.popPose();
         }
     }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(ConduitBlockEntity blockEntity) {
+        net.minecraft.core.BlockPos pos = blockEntity.getBlockPos();
+        return new net.minecraft.world.phys.AABB(pos.getX(), pos.getY() - .25, pos.getZ(), pos.getX() + 1.0, pos.getY() + 1.25, pos.getZ() + 1.0);
+    }
 }
