--- a/net/minecraft/client/renderer/blockentity/DecoratedPotRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/DecoratedPotRenderer.java
@@ -239,4 +_,10 @@
         this.top.getExtentsForGui(posestack, p_428219_);
         this.bottom.getExtentsForGui(posestack, p_428219_);
     }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(DecoratedPotBlockEntity blockEntity) {
+        net.minecraft.core.BlockPos pos = blockEntity.getBlockPos();
+        return new net.minecraft.world.phys.AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1.0, pos.getY() + 1.3, pos.getZ() + 1.0);
+    }
 }
