--- a/net/minecraft/client/renderer/blockentity/EnchantTableRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/EnchantTableRenderer.java
@@ -83,4 +_,10 @@
         );
         p_439439_.popPose();
     }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(EnchantingTableBlockEntity blockEntity) {
+        net.minecraft.core.BlockPos pos = blockEntity.getBlockPos();
+        return new net.minecraft.world.phys.AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1., pos.getY() + 1.5, pos.getZ() + 1.);
+    }
 }
