--- a/net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
@@ -91,4 +_,9 @@
     public int getViewDistance() {
         return 68;
     }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(PistonMovingBlockEntity blockEntity) {
+        return net.minecraft.world.phys.AABB.INFINITE;
+    }
 }
