--- a/net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
@@ -65,6 +_,10 @@
     private void renderBlock(
         BlockPos p_112459_, BlockState p_112460_, PoseStack p_112461_, MultiBufferSource p_112462_, Level p_112463_, boolean p_112464_, int p_112465_
     ) {
+        if (true) {
+            net.neoforged.neoforge.client.ClientHooks.renderPistonMovedBlocks(p_112459_, p_112460_, p_112461_, p_112462_, p_112463_, p_112464_, p_112465_, blockRenderer == null ? blockRenderer = net.minecraft.client.Minecraft.getInstance().getBlockRenderer() : blockRenderer);
+            return;
+        }
         RenderType rendertype = ItemBlockRenderTypes.getMovingBlockRenderType(p_112460_);
         VertexConsumer vertexconsumer = p_112462_.getBuffer(rendertype);
         this.blockRenderer
