--- a/net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/PistonHeadRenderer.java
@@ -65,6 +_,10 @@
     private void renderBlock(
         BlockPos pos, BlockState state, PoseStack poseStack, MultiBufferSource bufferSource, Level level, boolean extended, int packedOverlay
     ) {
+        if (true) {
+            net.neoforged.neoforge.client.ClientHooks.renderPistonMovedBlocks(pos, state, poseStack, bufferSource, level, extended, packedOverlay, blockRenderer);
+            return;
+        }
         RenderType rendertype = ItemBlockRenderTypes.getMovingBlockRenderType(state);
         VertexConsumer vertexconsumer = bufferSource.getBuffer(rendertype);
         this.blockRenderer
@@ -86,5 +_,10 @@
     @Override
     public int getViewDistance() {
         return 68;
+    }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(PistonMovingBlockEntity blockEntity) {
+        return net.minecraft.world.phys.AABB.INFINITE;
     }
 }
