--- a/net/minecraft/client/renderer/blockentity/ShulkerBoxRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/ShulkerBoxRenderer.java
@@ -56,4 +_,10 @@
         this.model.renderToBuffer(p_112480_, vertexconsumer, p_112482_, p_112483_);
         p_112480_.popPose();
     }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(ShulkerBoxBlockEntity blockEntity) {
+        net.minecraft.core.BlockPos pos = blockEntity.getBlockPos();
+        return new net.minecraft.world.phys.AABB(pos.getX() - 0.5, pos.getY() - 0.5, pos.getZ() - 0.5, pos.getX() + 1.5, pos.getY() + 1.5, pos.getZ() + 1.5);
+    }
 }
