--- a/net/minecraft/client/renderer/blockentity/SkullBlockRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/SkullBlockRenderer.java
@@ -61,7 +_,7 @@
                 case PIGLIN -> new PiglinHeadModel(p_387840_.bakeLayer(ModelLayers.PIGLIN_HEAD));
             });
         } else {
-            return null;
+            return net.neoforged.neoforge.client.ClientHooks.getModdedSkullModel(p_387840_, p_388801_); // Neo: Lookup model for modded skull types
         }
     }
 
@@ -150,5 +_,15 @@
 
     public static RenderType getPlayerSkinRenderType(ResourceLocation p_428466_) {
         return RenderType.entityTranslucent(p_428466_);
+    }
+
+    @Override
+    public net.minecraft.world.phys.AABB getRenderBoundingBox(SkullBlockEntity blockEntity) {
+        SkullBlock.Type type = ((AbstractSkullBlock) blockEntity.getBlockState().getBlock()).getType();
+        if (type == SkullBlock.Types.DRAGON) {
+            net.minecraft.core.BlockPos pos = blockEntity.getBlockPos();
+            return new net.minecraft.world.phys.AABB(pos.getX() - .75, pos.getY() - .35, pos.getZ() - .75, pos.getX() + 1.75, pos.getY() + 1.0, pos.getZ() + 1.75);
+        }
+        return BlockEntityRenderer.super.getRenderBoundingBox(blockEntity);
     }
 }
