--- a/net/minecraft/client/gui/Gui.java
+++ b/net/minecraft/client/gui/Gui.java
@@ -79,6 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.apache.commons.lang3.tuple.Pair;
 import org.jspecify.annotations.Nullable;
+// Neo: Exceptionally add a static wildcard import to make the patch bearable, and comments to avoid the detection by spotless rules.
+/* space for import change */ import static net.neoforged.neoforge.client.gui.VanillaGuiLayers.* /* space for wildcard import */;
 
 @OnlyIn(Dist.CLIENT)
 public class Gui {
@@ -172,6 +_,15 @@
     private Pair<Gui.ContextualInfo, ContextualBarRenderer> contextualInfoBar = Pair.of(Gui.ContextualInfo.EMPTY, ContextualBarRenderer.EMPTY);
     private final Map<Gui.ContextualInfo, Supplier<ContextualBarRenderer>> contextualInfoBarRenderers;
     private float scopeScale;
+    private final net.neoforged.neoforge.client.gui.GuiLayerManager layerManager = new net.neoforged.neoforge.client.gui.GuiLayerManager();
+    /**
+     * Neo: This variable controls the height of overlays on the left of the hotbar (e.g. health, armor).
+     */
+    public int leftHeight;
+    /**
+     * Neo: This variable controls the height of overlays on the right of the hotbar (e.g. food, vehicle health, air).
+     */
+    public int rightHeight;
 
     public Gui(Minecraft minecraft) {
         this.minecraft = minecraft;
@@ -192,6 +_,7 @@
             () -> new JumpableVehicleBarRenderer(minecraft)
         );
         this.resetTitleTimes();
+        this.registerVanillaLayers();
     }
 
     public void resetTitleTimes() {
@@ -202,28 +_,47 @@
 
     public void render(GuiGraphics graphics, DeltaTracker deltaTracker) {
         if (!(this.minecraft.screen instanceof LevelLoadingScreen)) {
-            if (!this.minecraft.options.hideGui) {
-                this.renderCameraOverlays(graphics, deltaTracker);
-                this.renderCrosshair(graphics, deltaTracker);
-                graphics.nextStratum();
-                this.renderHotbarAndDecorations(graphics, deltaTracker);
-                this.renderEffects(graphics, deltaTracker);
-                this.renderBossOverlay(graphics, deltaTracker);
-            }
+            leftHeight = 39;
+            rightHeight = 39;
+            updateContextualBarRenderer();
+            layerManager.render(graphics, deltaTracker);
+        }
+    }
 
-            this.renderSleepOverlay(graphics, deltaTracker);
+    private void registerVanillaLayers() {
+        java.util.function.BooleanSupplier guiVisible = () -> !this.minecraft.options.hideGui;
+        java.util.function.BooleanSupplier survivalVisible = () -> this.minecraft.gameMode.canHurtPlayer() && guiVisible.getAsBoolean();
+        layerManager.add(CAMERA_OVERLAYS, this::renderCameraOverlays, guiVisible);
+        layerManager.add(CROSSHAIR, this::renderCrosshair, guiVisible);
+        layerManager.add(AFTER_CAMERA_DECORATIONS, (guiGraphics, deltaTracker) -> guiGraphics.nextStratum(), guiVisible);
+        // TODO: Split this up again into its pieces
+        layerManager.add(HOTBAR, this::renderHotbar, guiVisible);
+        layerManager.add(PLAYER_HEALTH, (guiGraphics, partialTick) -> renderHealthLevel(guiGraphics), survivalVisible);
+        layerManager.add(ARMOR_LEVEL, (guiGraphics, partialTick) -> renderArmorLevel(guiGraphics), survivalVisible);
+        layerManager.add(FOOD_LEVEL, (guiGraphics, partialTick) -> renderFoodLevel(guiGraphics), survivalVisible);
+        layerManager.add(VEHICLE_HEALTH, (guiGraphics, partialTick) -> renderVehicleHealth(guiGraphics), guiVisible);
+        layerManager.add(AIR_LEVEL, (guiGraphics, partialTick) -> renderAirLevel(guiGraphics), survivalVisible);
+        layerManager.add(CONTEXTUAL_INFO_BAR_BACKGROUND, this::renderContextualInfoBarBackground, guiVisible);
+        layerManager.add(EXPERIENCE_LEVEL, this::renderExperienceLevel, guiVisible);
+        layerManager.add(CONTEXTUAL_INFO_BAR, this::renderContextualInfoBar, guiVisible);
+        layerManager.add(SELECTED_ITEM_NAME, this::maybeRenderSelectedItemName, guiVisible);
+        layerManager.add(SPECTATOR_TOOLTIP, this::maybeRenderSpectatorTooltip, guiVisible);
+        layerManager.add(EFFECTS, this::renderEffects, guiVisible);
+        layerManager.add(BOSS_OVERLAY, this::renderBossOverlay, guiVisible);
+        layerManager.add(SLEEP_OVERLAY, this::renderSleepOverlay);
+        layerManager.add(DEMO_OVERLAY, this::renderDemoOverlay, guiVisible);
+        layerManager.add(SCOREBOARD_SIDEBAR, this::renderScoreboardSidebar, guiVisible);
+        layerManager.add(OVERLAY_MESSAGE, this::renderOverlayMessage, guiVisible);
+        layerManager.add(TITLE, this::renderTitle, guiVisible);
+        layerManager.add(CHAT, this::renderChat, guiVisible);
+        layerManager.add(TAB_LIST, this::renderTabList, guiVisible);
+        layerManager.add(SUBTITLE_OVERLAY, (graphics, deltaTracker) -> {
             if (!this.minecraft.options.hideGui) {
-                this.renderDemoOverlay(graphics, deltaTracker);
-                this.renderScoreboardSidebar(graphics, deltaTracker);
-                this.renderOverlayMessage(graphics, deltaTracker);
-                this.renderTitle(graphics, deltaTracker);
-                this.renderChat(graphics, deltaTracker);
-                this.renderTabList(graphics, deltaTracker);
                 this.renderSubtitleOverlay(graphics, this.minecraft.screen == null || this.minecraft.screen.isInGameUi());
             } else if (this.minecraft.screen != null && this.minecraft.screen.isInGameUi()) {
                 this.renderSubtitleOverlay(graphics, true);
             }
-        }
+        });
     }
 
     private void renderBossOverlay(GuiGraphics graphics, DeltaTracker deltaTracker) {
@@ -267,8 +_,12 @@
                 for (EquipmentSlot slot : EquipmentSlot.values()) {
                     ItemStack item = player.getItemBySlot(slot);
                     Equippable equippable = item.get(DataComponents.EQUIPPABLE);
-                    if (equippable != null && equippable.slot() == slot && equippable.cameraOverlay().isPresent()) {
-                        this.renderTextureOverlay(graphics, equippable.cameraOverlay().get().withPath(p -> "textures/" + p + ".png"), 1.0F);
+                    if (equippable != null && equippable.slot() == slot) {
+                        if (equippable.cameraOverlay().isPresent()) {
+                            this.renderTextureOverlay(graphics, equippable.cameraOverlay().get().withPath(p -> "textures/" + p + ".png"), 1.0F);
+                        }
+
+                        net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(item).renderFirstPersonOverlay(item, slot, this.minecraft.player, graphics, deltaTracker);
                     }
                 }
             }
@@ -321,7 +_,11 @@
             if (alpha > 0) {
                 graphics.nextStratum();
                 graphics.pose().pushMatrix();
-                graphics.pose().translate(graphics.guiWidth() / 2, graphics.guiHeight() - 68);
+                // Include a shift based on the bar height plus the difference between the height that renderSelectedItemName
+                // renders at (59) and the height that the overlay/status bar renders at (68) by default
+                int yShift = Math.max(leftHeight, rightHeight) + (68 - 59);
+                // If y shift is less than the default y level, just render it at the base y level
+                graphics.pose().translate(graphics.guiWidth() / 2, graphics.guiHeight() - Math.max(yShift, 68));
                 int color;
                 if (this.animateOverlayMessageColor) {
                     color = Mth.hsvToArgb(t / 50.0F, 0.7F, 0.6F, alpha);
@@ -382,10 +_,19 @@
     private void renderChat(GuiGraphics graphics, DeltaTracker deltaTracker) {
         if (this.minecraft.player != null && !this.chat.isChatFocused()) {
             Window window = this.minecraft.getWindow();
+            // Neo: Allow customizing position of chat component
+            var chatBottomMargin = 40; // See ChatComponent#BOTTOM_MARGIN (used in translate calls in ChatComponent#render)
+            var event = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(
+                    new net.neoforged.neoforge.client.event.CustomizeGuiOverlayEvent.Chat(window, graphics, deltaTracker, 0, graphics.guiHeight() - chatBottomMargin)
+            );
+            // The event is given the absolute Y position; account for chat component's own offsetting here
+            graphics.pose().pushMatrix();
+            graphics.pose().translate(event.getPosX(), (event.getPosY() - graphics.guiHeight() + chatBottomMargin) / (float) this.chat.getScale());
             int mouseX = Mth.floor(this.minecraft.mouseHandler.getScaledXPos(window));
             int mouseY = Mth.floor(this.minecraft.mouseHandler.getScaledYPos(window));
             graphics.nextStratum();
             this.chat.render(graphics, this.getFont(), this.tickCount, mouseX, mouseY, ChatComponent.DisplayMode.BACKGROUND, false);
+            graphics.pose().popMatrix();
         }
     }
 
@@ -477,6 +_,8 @@
             int harmfulCount = 0;
 
             for (MobEffectInstance instance : Ordering.natural().reverse().sortedCopy(activeEffects)) {
+                var renderer = net.neoforged.neoforge.client.extensions.common.IClientMobEffectExtensions.of(instance);
+                if (!renderer.isVisibleInGui(instance)) continue;
                 Holder<MobEffect> effect = instance.getEffect();
                 if (instance.showIcon()) {
                     int x = graphics.guiWidth();
@@ -508,6 +_,7 @@
                         }
                     }
 
+                    if (renderer.renderGuiIcon(instance, this, graphics, x, y, 0, alpha)) continue;
                     graphics.blitSprite(RenderPipelines.GUI_TEXTURED, getMobEffectSprite(effect), x + 3, y + 3, 18, 18, ARGB.white(alpha));
                 }
             }
@@ -518,32 +_,52 @@
         return effect.unwrapKey().map(ResourceKey::identifier).map(id -> id.withPrefix("mob_effect/")).orElseGet(MissingTextureAtlasSprite::getLocation);
     }
 
-    private void renderHotbarAndDecorations(GuiGraphics graphics, DeltaTracker deltaTracker) {
+    // Neo: Split off into separate method to wrap in a layer
+    private void renderHotbar(GuiGraphics graphics, DeltaTracker deltaTracker) {
         if (this.minecraft.gameMode.getPlayerMode() == GameType.SPECTATOR) {
             this.spectatorGui.renderHotbar(graphics);
         } else {
             this.renderItemHotbar(graphics, deltaTracker);
         }
-
-        if (this.minecraft.gameMode.canHurtPlayer()) {
-            this.renderPlayerHealth(graphics);
-        }
-
-        this.renderVehicleHealth(graphics);
+    }
+
+    // Neo: Wrap in a method to call it when re-rendering
+    private void updateContextualBarRenderer() {
         Gui.ContextualInfo nextContextualInfo = this.nextContextualInfoState();
         if (nextContextualInfo != this.contextualInfoBar.getKey()) {
             this.contextualInfoBar = Pair.of(nextContextualInfo, this.contextualInfoBarRenderers.get(nextContextualInfo).get());
         }
+    }
 
+    // Neo: Split off into separate method to wrap in a layer
+    private void renderContextualInfoBarBackground(GuiGraphics graphics, DeltaTracker deltaTracker)
+    {
         this.contextualInfoBar.getValue().renderBackground(graphics, deltaTracker);
+    }
+
+    // Neo: Split off into separate method to wrap in a layer
+    private void renderExperienceLevel(GuiGraphics graphics, DeltaTracker deltaTracker) {
         if (this.minecraft.gameMode.hasExperience() && this.minecraft.player.experienceLevel > 0) {
             ContextualBarRenderer.renderExperienceLevel(graphics, this.minecraft.font, this.minecraft.player.experienceLevel);
         }
+    }
 
+    // Neo: Split off into separate method to wrap in a layer
+    private void renderContextualInfoBar(GuiGraphics graphics, DeltaTracker deltaTracker) {
         this.contextualInfoBar.getValue().render(graphics, deltaTracker);
+    }
+
+    // Neo: Split off into separate method to wrap in a layer
+    private void maybeRenderSelectedItemName(GuiGraphics graphics, DeltaTracker deltaTracker) {
         if (this.minecraft.gameMode.getPlayerMode() != GameType.SPECTATOR) {
-            this.renderSelectedItemName(graphics);
-        } else if (this.minecraft.player.isSpectator()) {
+            this.renderSelectedItemName(graphics, Math.max(this.leftHeight, this.rightHeight));
+        }
+    }
+
+    // Neo: Split off into separate method to wrap in a layer
+    private void maybeRenderSpectatorTooltip(GuiGraphics graphics, DeltaTracker deltaTracker) {
+        // Neo: Note this is originally mutually exclusive with the condition in maybeRenderSelectedItemName
+        if (this.minecraft.gameMode.getPlayerMode() == GameType.SPECTATOR && this.minecraft.player.isSpectator()) {
             this.spectatorGui.renderAction(graphics);
         }
     }
@@ -610,16 +_,21 @@
     }
 
     private void renderSelectedItemName(GuiGraphics graphics) {
+        renderSelectedItemName(graphics, 0);
+    }
+
+    private void renderSelectedItemName(GuiGraphics graphics, int yShift) {
         Profiler.get().push("selectedItemName");
         if (this.toolHighlightTimer > 0 && !this.lastToolHighlight.isEmpty()) {
-            MutableComponent str = Component.empty().append(this.lastToolHighlight.getHoverName()).withStyle(this.lastToolHighlight.getRarity().color());
+            MutableComponent str = Component.empty().append(this.lastToolHighlight.getHoverName()).withStyle(this.lastToolHighlight.getRarity().getStyleModifier());
             if (this.lastToolHighlight.has(DataComponents.CUSTOM_NAME)) {
                 str.withStyle(ChatFormatting.ITALIC);
             }
 
-            int strWidth = this.getFont().width(str);
+            Component highlightTip = this.lastToolHighlight.getHighlightTip(str);
+            int strWidth = this.getFont().width(highlightTip);
             int x = (graphics.guiWidth() - strWidth) / 2;
-            int y = graphics.guiHeight() - 59;
+            int y = graphics.guiHeight() -  Math.max(yShift, 59);
             if (!this.minecraft.gameMode.canHurtPlayer()) {
                 y += 14;
             }
@@ -630,7 +_,13 @@
             }
 
             if (alpha > 0) {
-                graphics.drawStringWithBackdrop(this.getFont(), str, x, y, strWidth, ARGB.white(alpha));
+                Font font = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(lastToolHighlight).getFont(lastToolHighlight, net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.FontContext.SELECTED_ITEM_NAME);
+                if (font == null) {
+                    graphics.drawStringWithBackdrop(this.getFont(), highlightTip, x, y, strWidth, ARGB.color(alpha, -1));
+                } else {
+                    x = (graphics.guiWidth() - font.width(highlightTip)) / 2;
+                    graphics.drawStringWithBackdrop(font, highlightTip, x, y, strWidth, ARGB.color(alpha, -1));
+                }
             }
         }
 
@@ -749,7 +_,15 @@
         return (int)Math.ceil(hearts / 10.0);
     }
 
+    @Deprecated // Neo: Split up into different layers
     private void renderPlayerHealth(GuiGraphics graphics) {
+        renderHealthLevel(graphics);
+        renderArmorLevel(graphics);
+        renderFoodLevel(graphics);
+        renderAirLevel(graphics);
+    }
+
+    private void renderHealthLevel(GuiGraphics graphics) {
         Player player = this.getCameraPlayer();
         if (player != null) {
             int currentHealth = Mth.ceil(player.getHealth());
@@ -773,33 +_,62 @@
             this.random.setSeed(this.tickCount * 312871);
             int xLeft = graphics.guiWidth() / 2 - 91;
             int xRight = graphics.guiWidth() / 2 + 91;
-            int yLineBase = graphics.guiHeight() - 39;
+            int yLineBase = graphics.guiHeight() - leftHeight;
             float maxHealth = Math.max((float)player.getAttributeValue(Attributes.MAX_HEALTH), (float)Math.max(oldHealth, currentHealth));
             int totalAbsorption = Mth.ceil(player.getAbsorptionAmount());
             int numHealthRows = Mth.ceil((maxHealth + totalAbsorption) / 2.0F / 10.0F);
             int healthRowHeight = Math.max(10 - (numHealthRows - 2), 3);
             int yLineAir = yLineBase - 10;
+            leftHeight += (numHealthRows - 1) * healthRowHeight + 10;
             int heartOffsetIndex = -1;
             if (player.hasEffect(MobEffects.REGENERATION)) {
                 heartOffsetIndex = this.tickCount % Mth.ceil(maxHealth + 5.0F);
             }
 
-            Profiler.get().push("armor");
-            renderArmor(graphics, player, yLineBase, numHealthRows, healthRowHeight, xLeft);
-            Profiler.get().popPush("health");
+            Profiler.get().push("health");
             this.renderHearts(
                 graphics, player, xLeft, yLineBase, healthRowHeight, heartOffsetIndex, maxHealth, currentHealth, oldHealth, totalAbsorption, blink
             );
+            Profiler.get().pop();
+        }
+    }
+
+    private void renderArmorLevel(GuiGraphics graphics) {
+        Player player = this.getCameraPlayer();
+        if (player != null) {
+            int l = graphics.guiWidth() / 2 - 91;
+            Profiler.get().push("armor");
+            renderArmor(graphics, player, graphics.guiHeight() - leftHeight + 10, 1, 0, l);
+            Profiler.get().pop();
+            if (player.getArmorValue() > 0) {
+                leftHeight += 10;
+            }
+        }
+    }
+
+    private void renderFoodLevel(GuiGraphics graphics) {
+        Player player = this.getCameraPlayer();
+        if (player != null) {
             LivingEntity vehicleWithHearts = this.getPlayerVehicleWithHealth();
             int vehicleHearts = this.getVehicleMaxHearts(vehicleWithHearts);
             if (vehicleHearts == 0) {
-                Profiler.get().popPush("food");
+                Profiler.get().push("food");
+                int xRight = graphics.guiWidth() / 2 + 91;
+                int yLineBase = graphics.guiHeight() - rightHeight;
                 this.renderFood(graphics, player, yLineBase, xRight);
-                yLineAir -= 10;
+                rightHeight += 10;
+                Profiler.get().pop();
             }
+        }
+    }
 
-            Profiler.get().popPush("air");
-            this.renderAirBubbles(graphics, player, vehicleHearts, yLineAir, xRight);
+    private void renderAirLevel(GuiGraphics graphics) {
+        Player player = this.getCameraPlayer();
+        if (player != null) {
+            int xRight = graphics.guiWidth() / 2 + 91;
+            int yLineAir = graphics.guiHeight() - rightHeight;
+            Profiler.get().push("air");
+            this.renderAirBubbles(graphics, player, 10, yLineAir, xRight);
             Profiler.get().pop();
         }
     }
@@ -912,6 +_,8 @@
                     graphics.blitSprite(RenderPipelines.GUI_TEXTURED, AIR_EMPTY_SPRITE, airBubbleXPos, yLineAir + wobbleYOffset, 9, 9);
                 }
             }
+
+            rightHeight += 10;
         }
     }
 
@@ -979,7 +_,7 @@
             if (hearts != 0) {
                 int currentHealth = (int)Math.ceil(vehicleWithHearts.getHealth());
                 Profiler.get().popPush("mountHealth");
-                int yLine1 = graphics.guiHeight() - 39;
+                int yLine1 = graphics.guiHeight() - rightHeight;
                 int xRight = graphics.guiWidth() / 2 + 91;
                 int yo = yLine1;
 
@@ -1000,6 +_,7 @@
                     }
 
                     yo -= 10;
+                    rightHeight += 10;
                 }
             }
         }
@@ -1031,7 +_,9 @@
         int top = (graphics.guiHeight() - height) / 2;
         int right = left + width;
         int bottom = top + height;
-        graphics.blit(RenderPipelines.GUI_TEXTURED, SPYGLASS_SCOPE_LOCATION, left, top, 0.0F, 0.0F, width, height, width, height);
+        ItemStack useItem = this.minecraft.player.getUseItem();
+        var clientItemExtensions = net.neoforged.neoforge.client.extensions.common.IClientItemExtensions.of(useItem);
+        graphics.blit(RenderPipelines.GUI_TEXTURED, clientItemExtensions.getScopeOverlayTexture(useItem), left, top, 0.0F, 0.0F, width, height, width, height);
         graphics.fill(RenderPipelines.GUI, 0, bottom, graphics.guiWidth(), graphics.guiHeight(), -16777216);
         graphics.fill(RenderPipelines.GUI, 0, 0, graphics.guiWidth(), top, -16777216);
         graphics.fill(RenderPipelines.GUI, 0, top, left, bottom, -16777216);
@@ -1180,7 +_,7 @@
                 this.toolHighlightTimer = 0;
             } else if (this.lastToolHighlight.isEmpty()
                 || !selected.is(this.lastToolHighlight.getItem())
-                || !selected.getHoverName().equals(this.lastToolHighlight.getHoverName())) {
+                || (!selected.getHoverName().equals(this.lastToolHighlight.getHoverName()) || !selected.getHighlightTip(selected.getHoverName()).equals(this.lastToolHighlight.getHighlightTip(this.lastToolHighlight.getHoverName())))) {
                 this.toolHighlightTimer = (int)(40.0 * this.minecraft.options.notificationDisplayTime().get());
             } else if (this.toolHighlightTimer > 0) {
                 this.toolHighlightTimer--;
@@ -1455,12 +_,27 @@
                 type = NORMAL;
             }
 
+            type = net.neoforged.neoforge.client.ClientHooks.firePlayerHeartTypeEvent(player, type);
+
             return type;
         }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(Gui.HeartType.class);
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
     public interface RenderFunction {
         void render(GuiGraphics graphics, final DeltaTracker deltaTracker);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void initModdedOverlays() {
+        this.layerManager.initModdedLayers();
+    }
+
+    public int getLayerCount() {
+        return this.layerManager.getLayerCount();
     }
 }
