--- a/net/minecraft/client/gui/Gui.java
+++ b/net/minecraft/client/gui/Gui.java
@@ -177,7 +_,8 @@
             profiler.push("screen");
 
             try {
-                this.screen.extractRenderStateWithTooltipAndSubtitles(graphics, xMouse, yMouse, deltaTracker.getGameTimeDeltaTicks());
+                // Neo: Wrap Screen#extractRenderStateWithTooltipAndSubtitles() to allow for GUI Layers and ScreenEvent.Render.[Pre/Post]
+                net.neoforged.neoforge.client.ClientHooks.drawScreen(this.screen, graphics, xMouse, yMouse, deltaTracker.getGameTimeDeltaTicks());
             } catch (Throwable t) {
                 CrashReport report = CrashReport.forThrowable(t, "Rendering screen");
                 CrashReportCategory category = report.addCategory("Screen render details");
@@ -226,9 +_,7 @@
             LOGGER.error("setScreen called from non-game thread");
         }
 
-        if (this.screen != null) {
-            this.screen.removed();
-        } else {
+        if (this.screen == null) {
             this.minecraft.setLastInputType(InputType.NONE);
         }
 
@@ -250,6 +_,19 @@
             }
         }
 
+        net.neoforged.neoforge.client.ClientHooks.clearGuiLayers(this);
+        Screen old = this.screen;
+        if (screen != null) {
+            var event = new net.neoforged.neoforge.client.event.ScreenEvent.Opening(old, screen);
+            if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(event).isCanceled()) return;
+            screen = event.getNewScreen();
+        }
+
+        if (old != null && screen != old) {
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.client.event.ScreenEvent.Closing(old));
+            old.removed();
+        }
+
         this.screen = screen;
         if (this.screen != null) {
             this.screen.added();
@@ -376,6 +_,8 @@
             Screen screen = function.apply(nextStep);
             nextStep = () -> this.setScreen(screen);
         }
+
+        nextStep = net.neoforged.neoforge.client.loading.ClientModLoader.completeModLoading(nextStep);
 
         return nextStep;
     }
