--- a/net/minecraft/client/renderer/block/model/BakedQuad.java
+++ b/net/minecraft/client/renderer/block/model/BakedQuad.java
@@ -11,6 +_,11 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.joml.Vector3fc;
 
+/**
+ * @param hasAmbientOcclusion {@code false} to force-disable AO for this quad
+ *                            or {@code true} to use the behavior dictated by
+ *                            {@link net.neoforged.neoforge.client.extensions.BlockModelPartExtension#ambientOcclusion}.
+ */
 @OnlyIn(Dist.CLIENT)
 public record BakedQuad(
     Vector3fc position0,
