--- a/net/minecraft/client/renderer/block/model/BakedQuad.java
+++ b/net/minecraft/client/renderer/block/model/BakedQuad.java
@@ -12,13 +_,19 @@
     protected final Direction direction;
     protected final TextureAtlasSprite sprite;
     private final boolean shade;
+    private final boolean hasAmbientOcclusion;
 
     public BakedQuad(int[] vertices, int tintIndex, Direction direction, TextureAtlasSprite sprite, boolean shade) {
+        this(vertices, tintIndex, direction, sprite, shade, true);
+    }
+
+    public BakedQuad(int[] vertices, int tintIndex, Direction direction, TextureAtlasSprite sprite, boolean shade, boolean hasAmbientOcclusion) {
         this.vertices = vertices;
         this.tintIndex = tintIndex;
         this.direction = direction;
         this.sprite = sprite;
         this.shade = shade;
+        this.hasAmbientOcclusion = hasAmbientOcclusion;
     }
 
     public TextureAtlasSprite getSprite() {
@@ -43,5 +_,13 @@
 
     public boolean isShade() {
         return this.shade;
+    }
+
+    /**
+     * {@return false to force-disable AO for this quad or true to use the behavior dictated by
+     * {@link net.neoforged.neoforge.client.extensions.IBakedModelExtension#useAmbientOcclusion(net.minecraft.world.level.block.state.BlockState, net.neoforged.neoforge.client.model.data.ModelData, net.minecraft.client.renderer.RenderType)}}
+     */
+    public boolean hasAmbientOcclusion() {
+        return this.hasAmbientOcclusion;
     }
 }
