--- a/net/minecraft/client/renderer/block/model/BlockModel.java
+++ b/net/minecraft/client/renderer/block/model/BlockModel.java
@@ -23,7 +_,19 @@
         PlayerSkinRenderCache playerSkinRenderCache,
         Function<BlockState, BlockStateModel> modelGetter,
         BlockModel missingBlockModel
+        , net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations pendingAnimations
     ) implements SpecialModelRenderer.BakingContext {
+        /// @deprecated Neo: use [#BakingContext(EntityModelSet, SpriteGetter, PlayerSkinRenderCache, Function, BlockModel, net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations)] instead
+        @Deprecated
+        public BakingContext(
+                EntityModelSet entityModelSet,
+                SpriteGetter sprites,
+                PlayerSkinRenderCache playerSkinRenderCache,
+                Function<BlockState, BlockStateModel> modelGetter,
+                BlockModel missingBlockModel
+        ) {
+            this(entityModelSet, sprites, playerSkinRenderCache, modelGetter, missingBlockModel, net.neoforged.neoforge.client.entity.animation.json.AnimationLoader.PendingAnimations.EMPTY);
+        }
     }
 
     @OnlyIn(Dist.CLIENT)
