--- a/net/minecraft/client/renderer/block/model/BlockModelPart.java
+++ b/net/minecraft/client/renderer/block/model/BlockModelPart.java
@@ -12,6 +_,8 @@
 public interface BlockModelPart extends net.neoforged.neoforge.client.extensions.BlockModelPartExtension {
     List<BakedQuad> getQuads(@Nullable Direction direction);
 
+    /** @deprecated Neo: Use {@link #ambientOcclusion()} instead. */
+    @Deprecated
     boolean useAmbientOcclusion();
 
     Material.Baked particleMaterial();
