--- a/net/minecraft/client/renderer/block/model/BlockModel.java
+++ b/net/minecraft/client/renderer/block/model/BlockModel.java
@@ -77,9 +_,10 @@
     public BlockModel parent;
     @Nullable
     protected ResourceLocation parentLocation;
+    public final net.neoforged.neoforge.client.model.geometry.BlockGeometryBakingContext customData = new net.neoforged.neoforge.client.model.geometry.BlockGeometryBakingContext(this);
 
     public static BlockModel fromStream(Reader reader) {
-        return GsonHelper.fromJson(GSON, reader, BlockModel.class);
+        return GsonHelper.fromJson(net.neoforged.neoforge.client.model.ExtendedBlockModelDeserializer.INSTANCE, reader, BlockModel.class);
     }
 
     public static BlockModel fromString(String jsonString) {
@@ -104,10 +_,15 @@
         this.overrides = overrides;
     }
 
+    @Deprecated
     public List<BlockElement> getElements() {
+        if (customData.hasCustomGeometry()) return java.util.Collections.emptyList();
         return this.elements.isEmpty() && this.parent != null ? this.parent.getElements() : this.elements;
     }
 
+    @Nullable
+    public ResourceLocation getParentLocation() { return parentLocation; }
+
     public boolean hasAmbientOcclusion() {
         if (this.hasAmbientOcclusion != null) {
             return this.hasAmbientOcclusion;
@@ -136,6 +_,10 @@
         return this.overrides.isEmpty() ? ItemOverrides.EMPTY : new ItemOverrides(baker, model, this.overrides);
     }
 
+    public ItemOverrides getOverrides(ModelBaker baker, BlockModel model, Function<Material, TextureAtlasSprite> spriteGetter) {
+        return this.overrides.isEmpty() ? ItemOverrides.EMPTY : new ItemOverrides(baker, model, this.overrides, spriteGetter);
+    }
+
     @Override
     public Collection<ResourceLocation> getDependencies() {
         Set<ResourceLocation> set = Sets.newHashSet();
@@ -184,6 +_,10 @@
             blockmodel.parent = (BlockModel)unbakedmodel;
         }
 
+        if (customData.hasCustomGeometry()) {
+            customData.getCustomGeometry().resolveParents(resolver, customData);
+        }
+
         this.overrides.forEach(p_247932_ -> {
             UnbakedModel unbakedmodel1 = resolver.apply(p_247932_.getModel());
             if (!Objects.equals(unbakedmodel1, this)) {
@@ -198,6 +_,12 @@
     }
 
     public BakedModel bake(
+            ModelBaker baker, BlockModel model, Function<Material, TextureAtlasSprite> spriteGetter, ModelState state, boolean guiLight3d
+    ) {
+        return net.neoforged.neoforge.client.model.geometry.UnbakedGeometryHelper.bake(this, baker, model, spriteGetter, state, guiLight3d);
+    }
+
+    public BakedModel bakeVanilla(
         ModelBaker baker, BlockModel model, Function<Material, TextureAtlasSprite> spriteGetter, ModelState state, boolean guiLight3d
     ) {
         TextureAtlasSprite textureatlassprite = spriteGetter.apply(this.getMaterial("particle"));
@@ -288,7 +_,18 @@
         ItemTransform itemtransform5 = this.getTransform(ItemDisplayContext.GUI);
         ItemTransform itemtransform6 = this.getTransform(ItemDisplayContext.GROUND);
         ItemTransform itemtransform7 = this.getTransform(ItemDisplayContext.FIXED);
-        return new ItemTransforms(itemtransform, itemtransform1, itemtransform2, itemtransform3, itemtransform4, itemtransform5, itemtransform6, itemtransform7);
+
+        var builder = com.google.common.collect.ImmutableMap.<ItemDisplayContext, ItemTransform>builder();
+        for(ItemDisplayContext type : ItemDisplayContext.values()) {
+            if (type.isModded()) {
+                var transform = this.getTransform(type);
+                if (transform != ItemTransform.NO_TRANSFORM) {
+                    builder.put(type, transform);
+                }
+            }
+        }
+
+        return new ItemTransforms(itemtransform, itemtransform1, itemtransform2, itemtransform3, itemtransform4, itemtransform5, itemtransform6, itemtransform7, builder.build());
     }
 
     private ItemTransform getTransform(ItemDisplayContext displayContext) {
@@ -407,6 +_,8 @@
         public boolean lightLikeBlock() {
             return this == SIDE;
         }
+
+        public String getSerializedName() { return name; }
     }
 
     @OnlyIn(Dist.CLIENT)
