--- a/net/minecraft/client/renderer/block/model/BlockModelPart.java
+++ b/net/minecraft/client/renderer/block/model/BlockModelPart.java
@@ -13,6 +_,8 @@
 public interface BlockModelPart extends net.neoforged.neoforge.client.extensions.BlockModelPartExtension {
     List<BakedQuad> getQuads(@Nullable Direction p_409878_);
 
+    /** @deprecated Neo: Use {@link #ambientOcclusion()} instead. */
+    @Deprecated
     boolean useAmbientOcclusion();
 
     TextureAtlasSprite particleIcon();
