--- a/net/minecraft/client/renderer/block/model/BlockElement.java
+++ b/net/minecraft/client/renderer/block/model/BlockElement.java
@@ -31,6 +_,7 @@
     public final BlockElementRotation rotation;
     public final boolean shade;
     public final int lightEmission;
+    private net.neoforged.neoforge.client.model.ExtraFaceData faceData;
 
     public BlockElement(Vector3f p_253626_, Vector3f p_254426_, Map<Direction, BlockElementFace> p_254454_) {
         this(p_253626_, p_254426_, p_254454_, null, true, 0);
@@ -44,6 +_,10 @@
         boolean p_361372_,
         int p_361908_
     ) {
+        this(p_361324_, p_363867_, p_362722_, p_360926_, p_361372_, p_361908_, net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT);
+    }
+
+    public BlockElement(Vector3f p_361324_, Vector3f p_363867_, Map<Direction, BlockElementFace> p_362722_, @Nullable BlockElementRotation p_360926_, boolean p_361372_, int p_361908_, net.neoforged.neoforge.client.model.ExtraFaceData faceData) {
         this.from = p_361324_;
         this.to = p_363867_;
         this.faces = p_362722_;
@@ -51,6 +_,8 @@
         this.shade = p_361372_;
         this.lightEmission = p_361908_;
         this.fillUvs();
+        this.setFaceData(faceData);
+        this.faces.values().forEach(face -> face.parent().setValue(this));
     }
 
     private void fillUvs() {
@@ -98,7 +_,8 @@
                     }
                 }
 
-                return new BlockElement(vector3f, vector3f1, map, blockelementrotation, flag, i);
+                var faceData = net.neoforged.neoforge.client.model.ExtraFaceData.read(jsonobject.get("neoforge_data"), net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT);
+                return new BlockElement(vector3f, vector3f1, map, blockelementrotation, flag, i, faceData);
             }
         }
 
@@ -209,5 +_,13 @@
                 return new Vector3f(afloat[0], afloat[1], afloat[2]);
             }
         }
+    }
+
+    public net.neoforged.neoforge.client.model.ExtraFaceData getFaceData() {
+        return this.faceData;
+    }
+
+    public void setFaceData(net.neoforged.neoforge.client.model.ExtraFaceData faceData) {
+        this.faceData = java.util.Objects.requireNonNull(faceData);
     }
 }
