--- a/net/minecraft/client/renderer/block/model/BlockElementFace.java
+++ b/net/minecraft/client/renderer/block/model/BlockElementFace.java
@@ -15,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public record BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, @Nullable BlockElementFace.UVs uvs, Quadrant rotation) {
+public record BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, @Nullable BlockElementFace.UVs uvs, Quadrant rotation, @Nullable net.neoforged.neoforge.client.model.ExtraFaceData faceData, org.apache.commons.lang3.mutable.MutableObject<BlockElement> parent) {
     public static final int NO_TINT = -1;
 
     public static float getU(BlockElementFace.UVs p_405199_, Quadrant p_404689_, int p_404785_) {
@@ -26,6 +_,21 @@
         return p_405500_.getVertexV(p_404709_.rotateVertexIndex(p_405113_)) / 16.0F;
     }
 
+    public BlockElementFace(@Nullable Direction cullForDirection, int tintIndex, String texture, @Nullable BlockElementFace.UVs uvs, Quadrant rotation) {
+        this(cullForDirection, tintIndex, texture, uvs, rotation, null, new org.apache.commons.lang3.mutable.MutableObject<>());
+    }
+
+    @Override
+    public net.neoforged.neoforge.client.model.ExtraFaceData faceData() {
+        if (this.faceData != null) {
+            return this.faceData;
+        } else if (this.parent.getValue() != null) {
+            return this.parent.getValue().faceData();
+        } else {
+            return net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT;
+        }
+    }
+
     @OnlyIn(Dist.CLIENT)
     public static class Deserializer implements JsonDeserializer<BlockElementFace> {
         private static final int DEFAULT_TINT_INDEX = -1;
@@ -38,7 +_,9 @@
             String s = getTexture(jsonobject);
             BlockElementFace.UVs blockelementface$uvs = getUVs(jsonobject);
             Quadrant quadrant = getRotation(jsonobject);
-            return new BlockElementFace(direction, i, s, blockelementface$uvs, quadrant);
+            if (jsonobject.has("forge_data")) throw new JsonParseException("forge_data should be replaced by neoforge_data"); // TODO 1.22: Remove
+            var faceData = net.neoforged.neoforge.client.model.ExtraFaceData.read(jsonobject.get("neoforge_data"), null);
+            return new BlockElementFace(direction, i, s, blockelementface$uvs, quadrant, faceData, new org.apache.commons.lang3.mutable.MutableObject<>());
         }
 
         private static int getTintIndex(JsonObject p_111369_) {
