--- a/net/minecraft/client/renderer/block/model/FaceBakery.java
+++ b/net/minecraft/client/renderer/block/model/FaceBakery.java
@@ -77,6 +_,8 @@
 
         Direction direction = calculateFacing(avector3fc);
         if (p_111607_ == null && direction != null) {
+            // Neo: Suppress winding re-calculation when the quads may not be axis-aligned due to root transforms
+            if (!p_111606_.mayApplyArbitraryRotation())
             recalculateWinding(avector3fc, along, direction);
         }
 
@@ -93,7 +_,10 @@
             Objects.requireNonNullElse(direction, Direction.UP),
             p_111604_,
             p_111608_,
-            p_364857_
+            Math.max(p_364857_, p_111603_.faceData().lightEmission()),
+            net.neoforged.neoforge.client.ClientHooks.fillNormal(p_470769_, avector3fc[0], avector3fc[1], avector3fc[2], avector3fc[3]),
+            p_470769_.colors(net.neoforged.neoforge.client.model.quad.BakedColors.of(p_111603_.faceData().color())),
+            p_111603_.faceData().ambientOcclusion()
         );
     }
 
