--- a/net/minecraft/client/renderer/block/model/FaceBakery.java
+++ b/net/minecraft/client/renderer/block/model/FaceBakery.java
@@ -57,7 +_,14 @@
             this.recalculateWinding(aint, direction);
         }
 
-        return new BakedQuad(aint, face.tintIndex(), direction, sprite, shade);
+        net.neoforged.neoforge.client.ClientHooks.fillNormal(aint, direction);
+        var data = face.faceData();
+        var quad = new BakedQuad(aint, face.tintIndex(), direction, sprite, shade, data.ambientOcclusion());
+        if (!net.neoforged.neoforge.client.model.ExtraFaceData.DEFAULT.equals(data)) {
+            net.neoforged.neoforge.client.model.QuadTransformers.applyingLightmap(data.blockLight(), data.skyLight()).processInPlace(quad);
+            net.neoforged.neoforge.client.model.QuadTransformers.applyingColor(data.color()).processInPlace(quad);
+        }
+        return quad;
     }
 
     public static BlockFaceUV recomputeUVs(BlockFaceUV uv, Direction facing, Transformation rotation) {
