--- a/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
+++ b/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
@@ -60,8 +_,8 @@
                 break;
             }
 
-            SpriteContents spritecontents = p_405238_.get(material, p_404999_).contents();
-            list.addAll(processFrames(i, s, spritecontents));
+            net.minecraft.client.renderer.texture.TextureAtlasSprite sprite = p_405238_.get(material, p_404999_);
+            list.addAll(net.neoforged.neoforge.client.ClientHooks.fixItemModelSeams(processFrames(i, s, sprite.contents()), sprite));
         }
 
         return SimpleUnbakedGeometry.bake(list, p_386773_, p_405238_, p_386536_, p_404999_);
