--- a/net/minecraft/client/renderer/block/model/ItemTransforms.java
+++ b/net/minecraft/client/renderer/block/model/ItemTransforms.java
@@ -20,8 +_,24 @@
     ItemTransform gui,
     ItemTransform ground,
     ItemTransform fixed,
-    ItemTransform fixedFromBottom
+    ItemTransform fixedFromBottom,
+    com.google.common.collect.ImmutableMap<ItemDisplayContext, ItemTransform> moddedTransforms
 ) {
+    @Deprecated
+    public ItemTransforms(
+        ItemTransform thirdPersonLeftHand,
+        ItemTransform thirdPersonRightHand,
+        ItemTransform firstPersonLeftHand,
+        ItemTransform firstPersonRightHand,
+        ItemTransform head,
+        ItemTransform gui,
+        ItemTransform ground,
+        ItemTransform fixed,
+        ItemTransform fixedFromBottom
+    ) {
+        this(thirdPersonLeftHand, thirdPersonRightHand, firstPersonLeftHand, firstPersonRightHand, head, gui, ground, fixed, fixedFromBottom, com.google.common.collect.ImmutableMap.of());
+    }
+
     public static final ItemTransforms NO_TRANSFORMS = new ItemTransforms(
         ItemTransform.NO_TRANSFORM,
         ItemTransform.NO_TRANSFORM,
@@ -35,6 +_,17 @@
     );
 
     public ItemTransform getTransform(ItemDisplayContext p_270619_) {
+        if (p_270619_.isModded()) {
+            ItemTransform moddedTransform = moddedTransforms.get(p_270619_);
+            if (moddedTransform != null) {
+                return moddedTransform;
+            }
+            ItemDisplayContext moddedFallback = p_270619_.fallback();
+            if (moddedFallback == null) {
+                return ItemTransform.NO_TRANSFORM;
+            }
+            p_270619_ = moddedFallback;
+        }
         return switch (p_270619_) {
             case THIRD_PERSON_LEFT_HAND -> this.thirdPersonLeftHand;
             case THIRD_PERSON_RIGHT_HAND -> this.thirdPersonRightHand;
@@ -70,9 +_,23 @@
             ItemTransform itemtransform6 = this.getTransform(p_111822_, jsonobject, ItemDisplayContext.GROUND);
             ItemTransform itemtransform7 = this.getTransform(p_111822_, jsonobject, ItemDisplayContext.FIXED);
             ItemTransform itemtransform8 = this.getTransform(p_111822_, jsonobject, ItemDisplayContext.ON_SHELF);
-            return new ItemTransforms(
-                itemtransform1, itemtransform, itemtransform3, itemtransform2, itemtransform4, itemtransform5, itemtransform6, itemtransform7, itemtransform8
-            );
+
+            var builder = com.google.common.collect.ImmutableMap.<ItemDisplayContext, ItemTransform>builder();
+            for (ItemDisplayContext type : ItemDisplayContext.values()) {
+                if (type.isModded()) {
+                    var transform = this.getTransform(p_111822_, jsonobject, type);
+                    var fallbackType = type;
+                    while (transform == ItemTransform.NO_TRANSFORM && fallbackType.fallback() != null) {
+                        fallbackType = fallbackType.fallback();
+                        transform = this.getTransform(p_111822_, jsonobject, fallbackType);
+                    }
+                    if (transform != ItemTransform.NO_TRANSFORM){
+                        builder.put(type, transform);
+                    }
+                }
+            }
+
+            return new ItemTransforms(itemtransform1, itemtransform, itemtransform3, itemtransform2, itemtransform4, itemtransform5, itemtransform6, itemtransform7, itemtransform8, builder.build());
         }
 
         private ItemTransform getTransform(JsonDeserializationContext p_270385_, JsonObject p_270436_, ItemDisplayContext p_270100_) {
