--- a/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
+++ b/net/minecraft/client/renderer/block/model/ItemModelGenerator.java
@@ -32,13 +_,16 @@
 
             Material material = model.getMaterial(s);
             map.put(s, Either.left(material));
-            SpriteContents spritecontents = spriteGetter.apply(material).contents();
-            list.addAll(this.processFrames(i, s, spritecontents));
+            TextureAtlasSprite sprite = spriteGetter.apply(material);
+            // Neo: fix MC-73186 on generated item models
+            list.addAll(net.neoforged.neoforge.client.ClientHooks.fixItemModelSeams(this.processFrames(i, s, sprite.contents()), sprite));
         }
 
         map.put("particle", model.hasTexture("particle") ? Either.left(model.getMaterial("particle")) : map.get("layer0"));
         BlockModel blockmodel = new BlockModel(null, list, map, false, model.getGuiLight(), model.getTransforms(), model.getOverrides());
         blockmodel.name = model.name;
+        blockmodel.customData.copyFrom(model.customData);
+        blockmodel.customData.setGui3d(false);
         return blockmodel;
     }
 
