--- a/net/minecraft/client/renderer/block/model/MultiVariant.java
+++ b/net/minecraft/client/renderer/block/model/MultiVariant.java
@@ -45,12 +_,12 @@
     public BakedModel bake(ModelBaker p_249016_, Function<Material, TextureAtlasSprite> p_111851_, ModelState p_111852_) {
         if (this.variants.size() == 1) {
             Variant variant1 = this.variants.getFirst();
-            return p_249016_.bake(variant1.getModelLocation(), variant1);
+            return p_249016_.bake(variant1.getModelLocation(), variant1, p_111851_);
         } else {
             SimpleWeightedRandomList.Builder<BakedModel> builder = SimpleWeightedRandomList.builder();
 
             for (Variant variant : this.variants) {
-                BakedModel bakedmodel = p_249016_.bake(variant.getModelLocation(), variant);
+                BakedModel bakedmodel = p_249016_.bake(variant.getModelLocation(), variant, p_111851_);
                 builder.add(bakedmodel, variant.getWeight());
             }
 
