--- a/net/minecraft/client/renderer/block/model/SimpleModelWrapper.java
+++ b/net/minecraft/client/renderer/block/model/SimpleModelWrapper.java
@@ -18,11 +_,21 @@
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
-public record SimpleModelWrapper(QuadCollection quads, boolean useAmbientOcclusion, TextureAtlasSprite particleIcon) implements BlockModelPart {
+public record SimpleModelWrapper(QuadCollection quads, boolean useAmbientOcclusion, TextureAtlasSprite particleIcon, net.minecraft.client.renderer.chunk.@Nullable ChunkSectionLayer renderType) implements BlockModelPart {
     private static final Logger LOGGER = LogUtils.getLogger();
 
+    @Deprecated // Neo: Use render type aware version
+    public SimpleModelWrapper(QuadCollection quads, boolean useAmbientOcclusion, TextureAtlasSprite particleIcon) {
+        this(quads, useAmbientOcclusion, particleIcon, null);
+    }
+
     public static BlockModelPart bake(ModelBaker p_405335_, Identifier p_468685_, ModelState p_405869_) {
         ResolvedModel resolvedmodel = p_405335_.getModel(p_468685_);
+        return bake(p_405335_, resolvedmodel, p_405869_);
+    }
+
+    // Neo: split off to allow baking an existing ResolvedModel into a BlockModelPart
+    public static BlockModelPart bake(ModelBaker p_405335_, ResolvedModel resolvedmodel, ModelState p_405869_) {
         TextureSlots textureslots = resolvedmodel.getTopTextureSlots();
         boolean flag = resolvedmodel.getTopAmbientOcclusion();
         TextureAtlasSprite textureatlassprite = resolvedmodel.resolveParticleSprite(textureslots, p_405335_);
@@ -41,15 +_,22 @@
         }
 
         if (multimap != null) {
-            LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", p_468685_, multimap);
+            LOGGER.warn("Rejecting block model {}, since it contains sprites from outside of supported atlas: {}", resolvedmodel.debugName(), multimap);
             return p_405335_.missingBlockModelPart();
         } else {
-            return new SimpleModelWrapper(quadcollection, flag, textureatlassprite);
+            var renderTypeGroup = resolvedmodel.getTopAdditionalProperties().getOptional(net.neoforged.neoforge.client.model.NeoForgeModelProperties.RENDER_TYPE);
+            var renderTypes = renderTypeGroup == null || renderTypeGroup.isEmpty() ? null : renderTypeGroup.block();
+            return new SimpleModelWrapper(quadcollection, flag, textureatlassprite, renderTypes);
         }
     }
 
     @Override
     public List<BakedQuad> getQuads(@Nullable Direction p_405263_) {
         return this.quads.getQuads(p_405263_);
+    }
+
+    @Override
+    public net.minecraft.client.renderer.chunk.ChunkSectionLayer getRenderType(net.minecraft.world.level.block.state.BlockState state) {
+        return this.renderType != null ? this.renderType : BlockModelPart.super.getRenderType(state);
     }
 }
