--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -98,11 +99,16 @@
             levelVersion.levelDataVersion(),
             input.get("initialized").asBoolean(true),
             input.get("ServerBrands").asStream().flatMap(b -> b.asString().result().stream()).collect(Collectors.toCollection(Sets::newLinkedHashSet)),
-            input.get("removed_features").asStream().flatMap(b -> b.asString().result().stream()).collect(Collectors.toSet()),
+            // Neo: Append removed modded feature flags
+            updateRemovedFeatureFlags(input.get("removed_features").asStream().flatMap(b -> b.asString().result().stream()), input.get("enabled_features").asStream().flatMap(features -> features.asString().result().stream())).collect(Collectors.toSet()),
             settings,
             specialWorldProperty,
             worldGenSettingsLifecycle
-        );
+        ).withConfirmedWarning(worldGenSettingsLifecycle != Lifecycle.stable() && input.get("confirmedExperimentalSettings").asBoolean(false));
+        // Neo:
+        result.setDayTimeFraction(input.get("neoDayTimeFraction").asFloat(0f));
+        result.setDayTimePerTick(input.get("neoDayTimePerTick").asFloat(-1f));
+        return result;
     }
 
     @Override
