--- a/net/minecraft/world/phys/shapes/MinecartCollisionContext.java
+++ b/net/minecraft/world/phys/shapes/MinecartCollisionContext.java
@@ -23,7 +_,7 @@
         boolean onRails = BaseRailBlock.isRail(currentState);
         if (onRails) {
             this.ingoreBelow = currentRailPos.below();
-            RailShape shape = currentState.getValue(((BaseRailBlock)currentState.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)currentState.getBlock()).getRailDirection(currentState, entity.level(), currentRailPos, entity);
             if (shape.isSlope()) {
                 this.slopeIgnore = switch (shape) {
                     case ASCENDING_EAST -> currentRailPos.east();
