--- a/net/minecraft/world/phys/shapes/MinecartCollisionContext.java
+++ b/net/minecraft/world/phys/shapes/MinecartCollisionContext.java
@@ -25,7 +_,7 @@
         boolean flag = BaseRailBlock.isRail(blockstate);
         if (flag) {
             this.ingoreBelow = blockpos.below();
-            RailShape railshape = blockstate.getValue(((BaseRailBlock)blockstate.getBlock()).getShapeProperty());
+            RailShape railshape = ((BaseRailBlock)blockstate.getBlock()).getRailDirection(blockstate, p_366450_.level(), blockpos, p_366450_);
             if (railshape.isSlope()) {
                 this.slopeIgnore = switch (railshape) {
                     case ASCENDING_EAST -> blockpos.east();
