--- a/net/minecraft/world/level/lighting/BlockLightEngine.java
+++ b/net/minecraft/world/level/lighting/BlockLightEngine.java
@@ -109,7 +_,7 @@
     }
 
     private int getEmission(long packedPos, BlockState state) {
-        int i = state.getLightEmission();
+        int i = state.getLightEmission(chunkSource.getLevel(), mutablePos);
         return i > 0 && this.storage.lightOnInSection(SectionPos.blockToSection(packedPos)) ? i : 0;
     }
 
@@ -119,7 +_,7 @@
         LightChunk lightchunk = this.chunkSource.getChunkForLighting(chunkPos.x, chunkPos.z);
         if (lightchunk != null) {
             lightchunk.findBlockLightSources((p_285266_, p_285452_) -> {
-                int i = p_285452_.getLightEmission();
+                int i = p_285452_.getLightEmission(chunkSource.getLevel(), p_285266_);
                 this.enqueueIncrease(p_285266_.asLong(), LightEngine.QueueEntry.increaseLightFromEmission(i, isEmptyShape(p_285452_)));
             });
         }
