--- a/net/minecraft/world/level/lighting/BlockLightEngine.java
+++ b/net/minecraft/world/level/lighting/BlockLightEngine.java
@@ -109,7 +_,7 @@
     }
 
     private int getEmission(long p_285243_, BlockState p_284973_) {
-        int i = p_284973_.getLightEmission();
+        int i = p_284973_.getLightEmission(chunkSource.getLevel(), mutablePos);
         return i > 0 && this.storage.lightOnInSection(SectionPos.blockToSection(p_285243_)) ? i : 0;
     }
 
@@ -119,7 +_,7 @@
         LightChunk lightchunk = this.chunkSource.getChunkForLighting(p_285274_.x, p_285274_.z);
         if (lightchunk != null) {
             lightchunk.findBlockLightSources((p_360252_, p_360253_) -> {
-                int i = p_360253_.getLightEmission();
+                int i = p_360253_.getLightEmission(chunkSource.getLevel(), p_360252_);
                 this.enqueueIncrease(p_360252_.asLong(), LightEngine.QueueEntry.increaseLightFromEmission(i, isEmptyShape(p_360253_)));
             });
         }
