--- a/net/minecraft/world/level/lighting/BlockLightEngine.java
+++ b/net/minecraft/world/level/lighting/BlockLightEngine.java
@@ -110,7 +_,7 @@
     }
 
     private int getEmission(long blockNode, BlockState state) {
-        int emission = state.getLightEmission();
+        int emission = state.getLightEmission(chunkSource.getLevel(), mutablePos);
         return emission > 0 && this.storage.lightOnInSection(SectionPos.blockToSection(blockNode)) ? emission : 0;
     }
 
@@ -120,7 +_,7 @@
         LightChunk chunk = this.chunkSource.getChunkForLighting(pos.x(), pos.z());
         if (chunk != null) {
             chunk.findBlockLightSources((lightPos, state) -> {
-                int lightEmission = state.getLightEmission();
+                int lightEmission = state.getLightEmission(chunkSource.getLevel(), lightPos);
                 this.enqueueIncrease(lightPos.asLong(), LightEngine.QueueEntry.increaseLightFromEmission(lightEmission, isEmptyShape(state)));
             });
         }
