--- a/net/minecraft/world/level/lighting/LightEngine.java
+++ b/net/minecraft/world/level/lighting/LightEngine.java
@@ -44,7 +_,7 @@
         return state2 == state1
             ? false
             : state2.getLightBlock(level, pos) != state1.getLightBlock(level, pos)
-                || state2.getLightEmission() != state1.getLightEmission()
+                || state2.getLightEmission(level, pos) != state1.getLightEmission(level, pos)
                 || state2.useShapeForLightOcclusion()
                 || state1.useShapeForLightOcclusion();
     }
