--- a/net/minecraft/world/level/lighting/LightEngine.java
+++ b/net/minecraft/world/level/lighting/LightEngine.java
@@ -38,11 +_,19 @@
         this.clearChunkCache();
     }
 
+    /**
+     * @deprecated Neo: use overload with level context instead
+     */
+    @Deprecated
     public static boolean hasDifferentLightProperties(BlockState oldState, BlockState newState) {
+        return hasDifferentLightProperties(net.minecraft.world.level.EmptyBlockGetter.INSTANCE, BlockPos.ZERO, oldState, newState);
+    }
+
+    public static boolean hasDifferentLightProperties(net.minecraft.world.level.BlockGetter level, BlockPos pos, BlockState oldState, BlockState newState) {
         return newState == oldState
             ? false
             : newState.getLightDampening() != oldState.getLightDampening()
-                || newState.getLightEmission() != oldState.getLightEmission()
+                || newState.getLightEmission(level, pos) != oldState.getLightEmission(level, pos)
                 || newState.useShapeForLightOcclusion()
                 || oldState.useShapeForLightOcclusion();
     }
