--- a/net/minecraft/util/datafix/fixes/MobEffectIdFix.java
+++ b/net/minecraft/util/datafix/fixes/MobEffectIdFix.java
@@ -76,8 +_,20 @@
         return updateMobEffectIdField(input, oldFieldName, input, newFieldName);
     }
 
+    private static <T> Dynamic<T> updateMobEffectIdFieldConsideringForge(Dynamic<T> input, String oldFieldName, Dynamic<T> output, String newFieldName, String forgeFieldId) {
+        final var forgeField = input.get(forgeFieldId).result();
+        if (forgeField.isPresent()) {
+                return setFieldIfPresent((input == output ? input.remove(forgeFieldId) : output), newFieldName, forgeField);
+            }
+        return updateMobEffectIdField(input, oldFieldName, output, newFieldName);
+    }
+
+    private static <T> Dynamic<T> setFieldIfPresent(Dynamic<T> dynamic, String s, Optional<Dynamic<T>> optional) {
+        return optional.isEmpty() ? dynamic : dynamic.set(s, optional.get());
+    }
+
     private static <T> Dynamic<T> updateMobEffectInstance(Dynamic<T> input) {
-        input = updateMobEffectIdField(input, "Id", "id");
+        input = updateMobEffectIdFieldConsideringForge(input, "Id", input, "id", "forge:id");
         input = input.renameField("Ambient", "ambient");
         input = input.renameField("Amplifier", "amplifier");
         input = input.renameField("Duration", "duration");
@@ -96,7 +_,7 @@
     }
 
     private static <T> Dynamic<T> updateSuspiciousStewEntry(Dynamic<T> input, Dynamic<T> output) {
-        output = updateMobEffectIdField(input, "EffectId", output, "id");
+        output = updateMobEffectIdFieldConsideringForge(input, "EffectId", output, "id", "forge:effect_id");
         Optional<Dynamic<T>> duration = input.get("EffectDuration").result();
         return output.replaceField("EffectDuration", "duration", duration);
     }
