--- a/net/minecraft/server/commands/DataPackCommand.java
+++ b/net/minecraft/server/commands/DataPackCommand.java
@@ -186,7 +_,7 @@
     private static int listEnabledPacks(CommandSourceStack source) {
         PackRepository packrepository = source.getServer().getPackRepository();
         packrepository.reload();
-        Collection<? extends Pack> collection = packrepository.getSelectedPacks();
+        Collection<? extends Pack> collection = packrepository.getSelectedPacks().stream().filter(p -> !p.isHidden()).toList();
         if (collection.isEmpty()) {
             source.sendSuccess(() -> Component.translatable("commands.datapack.list.enabled.none"), false);
         } else {
