--- a/net/minecraft/server/commands/EnchantCommand.java
+++ b/net/minecraft/server/commands/EnchantCommand.java
@@ -74,7 +_,7 @@
                 if (entity instanceof LivingEntity target) {
                     ItemStack item = target.getMainHandItem();
                     if (!item.isEmpty()) {
-                        if (enchantment.canEnchant(item)
+                        if (item.supportsEnchantment(enchantmentHolder) // Neo: Respect IItemExtension#supportsEnchantment
                             && EnchantmentHelper.isEnchantmentCompatible(EnchantmentHelper.getEnchantmentsForCrafting(item).keySet(), enchantmentHolder)) {
                             item.enchant(enchantmentHolder, level);
                             success++;
