--- a/net/minecraft/server/commands/SetSpawnCommand.java
+++ b/net/minecraft/server/commands/SetSpawnCommand.java
@@ -29,7 +_,7 @@
             respawnData.pos().getZ(),
             respawnData.yaw(),
             respawnData.pitch(),
-            Component.translationArg(respawnData.dimension().identifier()),
+            net.neoforged.neoforge.common.CommonHooks.getDimensionDescription(respawnData.dimension()), // Neo: Use dimension translation, if one exists
             player.getDisplayName()
         ),
         (playerCount, var1, respawnData) -> Component.translatable(
@@ -39,7 +_,7 @@
             respawnData.pos().getZ(),
             respawnData.yaw(),
             respawnData.pitch(),
-            Component.translationArg(respawnData.dimension().identifier()),
+            net.neoforged.neoforge.common.CommonHooks.getDimensionDescription(respawnData.dimension()), // Neo: Use dimension translation, if one exists
             playerCount
         )
     );
