--- a/net/minecraft/server/commands/SpreadPlayersCommand.java
+++ b/net/minecraft/server/commands/SpreadPlayersCommand.java
@@ -252,16 +_,23 @@
                 position = positions[positionIndex++];
             }
 
-            entity.teleportTo(
-                level,
-                Mth.floor(position.x) + 0.5,
-                position.getSpawnY(level, maxHeight),
-                Mth.floor(position.z) + 0.5,
-                Set.of(),
-                entity.getYRot(),
-                entity.getXRot(),
-                true
+            net.neoforged.neoforge.event.entity.EntityTeleportEvent.SpreadPlayersCommand event = net.neoforged.neoforge.event.EventHooks.onEntityTeleportSpreadPlayersCommand(entity, level,
+                    (double)Mth.floor(position.x) + 0.5,
+                    position.getSpawnY(level, maxHeight),
+                    (double)Mth.floor(position.z) + 0.5
             );
+            if (!event.isCanceled()) {
+                entity.teleportTo(
+                        level,
+                        event.getTargetX(),
+                        event.getTargetY(),
+                        event.getTargetZ(),
+                        Set.of(),
+                        entity.getYRot(),
+                        entity.getXRot(),
+                        true
+                );
+            }
             double closest = Double.MAX_VALUE;
 
             for (SpreadPlayersCommand.Position testPosition : positions) {
