--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -244,6 +_,14 @@
         float xRot,
         @Nullable LookAt lookAt
     ) throws CommandSyntaxException {
+        net.neoforged.neoforge.event.entity.EntityTeleportEvent.TeleportCommand event = net.neoforged.neoforge.event.EventHooks.onEntityTeleportCommand(victim, x, y, z);
+        if (event.isCanceled()) {
+             return;
+        }
+        x = event.getTargetX();
+        y = event.getTargetY();
+        z = event.getTargetZ();
+
         BlockPos blockPos = BlockPos.containing(x, y, z);
         if (!Level.isInSpawnableBounds(blockPos)) {
             throw INVALID_POSITION.create();
