--- a/net/minecraft/server/commands/TimeCommand.java
+++ b/net/minecraft/server/commands/TimeCommand.java
@@ -188,7 +_,9 @@
         return wrapTime(clockManager.getTotalTicks(clock));
     }
 
-    private static int setPaused(CommandSourceStack source, Holder<WorldClock> clock, boolean paused) {
+    private static int setPaused(CommandSourceStack source, Holder<WorldClock> clock, boolean paused) throws CommandSyntaxException {
+        // Neo: Disallow using 'pause and resume' sub commands if clocks ignore pausing
+        net.neoforged.neoforge.common.CommonHooks.throwIfIgnoresPausing(clock);
         source.getServer().clockManager().setPaused(clock, paused);
         source.sendSuccess(() -> Component.translatable(paused ? "commands.time.pause" : "commands.time.resume", clock.getRegisteredName()), true);
         return 1;
