--- a/net/minecraft/server/commands/TimeCommand.java
+++ b/net/minecraft/server/commands/TimeCommand.java
@@ -215,6 +_,9 @@
     }
 
     private static int setPaused(CommandSourceStack source, Holder<WorldClock> clock, boolean paused) throws CommandSyntaxException {
+        // Neo: Disallow using 'pause and resume' sub commands if clocks ignore pausing
+        net.neoforged.neoforge.common.CommonHooks.throwIfIgnoresPausing(clock);
+
         ServerClockManager clockManager = source.getServer().clockManager();
         String clockName = clock.getRegisteredName();
         if (clockManager.getInstance(clock).isPaused() == paused) {
