--- a/net/minecraft/core/particles/BlockParticleOption.java
+++ b/net/minecraft/core/particles/BlockParticleOption.java
@@ -15,18 +_,38 @@
     );
     private final ParticleType<BlockParticleOption> type;
     private final BlockState state;
+    /** Neo: Position of the block this particle was spawned for, if available, to provide model data for the particle texture selection */
+    private final net.minecraft.core.@org.jspecify.annotations.Nullable BlockPos pos;
 
     public static MapCodec<BlockParticleOption> codec(ParticleType<BlockParticleOption> type) {
         return BLOCK_STATE_CODEC.xmap(state -> new BlockParticleOption(type, state), o -> o.state).fieldOf("block_state");
     }
 
     public static StreamCodec<? super RegistryFriendlyByteBuf, BlockParticleOption> streamCodec(ParticleType<BlockParticleOption> type) {
-        return ByteBufCodecs.idMapper(Block.BLOCK_STATE_REGISTRY).map(state -> new BlockParticleOption(type, state), o -> o.state);
+        return StreamCodec.composite(
+                ByteBufCodecs.idMapper(Block.BLOCK_STATE_REGISTRY),
+                option -> option.state,
+                net.neoforged.neoforge.network.codec.NeoForgeStreamCodecs.connectionAware(
+                        ByteBufCodecs.optional(net.minecraft.core.BlockPos.STREAM_CODEC),
+                        net.neoforged.neoforge.network.codec.NeoForgeStreamCodecs.uncheckedUnit(java.util.Optional.empty())
+                ),
+                option -> java.util.Optional.ofNullable(option.pos),
+                (state, pos) -> new BlockParticleOption(type, state, pos.orElse(null))
+        );
     }
 
     public BlockParticleOption(ParticleType<BlockParticleOption> type, BlockState state) {
+        this(type, state, null);
+    }
+
+    /**
+     * Neo: construct a {@link BlockParticleOption} for the given type and {@link BlockState} and optionally the position
+     * of the block this particle is being spawned for
+     */
+    public BlockParticleOption(ParticleType<BlockParticleOption> type, BlockState state, net.minecraft.core.@org.jspecify.annotations.Nullable BlockPos pos) {
         this.type = type;
         this.state = state;
+        this.pos = pos;
     }
 
     @Override
@@ -36,5 +_,12 @@
 
     public BlockState getState() {
         return this.state;
+    }
+
+    /**
+     * Neo: returns the position of the block this particle was spawned for, if available
+     */
+    public net.minecraft.core.@org.jspecify.annotations.Nullable BlockPos getPos() {
+        return pos;
     }
 }
