--- a/net/minecraft/core/particles/ItemParticleOption.java
+++ b/net/minecraft/core/particles/ItemParticleOption.java
@@ -24,7 +_,7 @@
             throw new IllegalArgumentException("Empty stacks are not allowed");
         } else {
             this.type = type;
-            this.itemStack = itemStack;
+            this.itemStack = itemStack.copy(); //Forge: Fix stack updating after the fact causing particle changes.
         }
     }
 
